{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2f0efef7-574d-47c7-a374-7a0e9f249999",
   "metadata": {},
   "source": [
    "# Extract annotations and data from OMIM\n",
    "\n",
    "The purpose of this notebook is to extract and format OMIM data for subsequent model annotation.\n",
    "\n",
    "Additionally, the purpose of this notebook is to extract relevant data about morbid SNPs and phenotypes from the database.\n",
    "\n",
    "## Notebook Requirements:\n",
    "*  Model genes **must** have the at least one of following annotations stored in the `object.annotation`. Values are expected to be seperated by semicolons. Accepted keys currently include:\n",
    "    * `\"ncbigene\"`\n",
    "    * `\"hgnc.symbol\"` \n",
    "    * `\"ensembl\"`\n",
    "    * `\"mim\"`\n",
    "\n",
    "## OMIM® - Online Mendelian Inheritance in Man®\n",
    "To utilize this notebook: \n",
    "\n",
    "1. Go to [OMIM database](https://omim.org/) and submit a request to download files.\n",
    "2. Follow the instructions to obtain download access for MIM files.\n",
    "3. After receiving access (check email), download the following files:\n",
    "    * `mim2gene.txt`: A tab-delimited file linking MIM numbers with NCBI Gene IDs, Ensembl Gene IDs, and HGNC Approved Gene Symbols.\n",
    "    * `mimTitles.txt`A tab-delimited file of MIM numbers and titles.\n",
    "    * `genemap2.txt`A tab-delimited file containing OMIM's Synopsis of the Human Gene Map including additional information such as genomic coordinates and inheritance.\n",
    "    * `morbidmap.txt`A tab-delimited file of OMIM's Synopsis of the Human Gene Map (same as `genemap2.txt` above) sorted alphabetically by disorder.\n",
    "4. Remember clear out any personal account information and ensure the downloaded OMIM file(s) remains local!\n",
    "\n",
    "### Citations\n",
    "Hamosh A, Scott AF, Amberger JS, Bocchini CA, McKusick VA. Online Mendelian Inheritance in Man (OMIM), a knowledgebase of human genes and genetic disorders. Nucleic Acids Res. 2005 Jan 1;33(Database issue):D514-7. doi: 10.1093/nar/gki033. PMID: 15608251; PMCID: PMC539987.\n",
    "\n",
    "\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "677865f9-6af2-46f6-bfa4-69a67cc560d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    GEM_NAME,\n",
    "    build_string,\n",
    "    compare_tables,\n",
    "    get_annotation_df,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    split_string,\n",
    "    visualize_comparison,\n",
    ")\n",
    "from rbc_gem_utils.database.mim import (\n",
    "    MIM_DB_TAG,\n",
    "    MIM_NUMBER_RE,\n",
    "    get_last_updated_dates_MIM,\n",
    "    load_data_MIM,\n",
    ")\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "334d903e-ff3d-4fff-b2f3-d086f4abacd5",
   "metadata": {},
   "source": [
    "## Set notebook options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c36db2fe-9854-4ae5-837b-fd5b73745bf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_tag = MIM_DB_TAG\n",
    "\n",
    "# Compare to/Overwrite previous files\n",
    "compare_figsize = (5, 5)\n",
    "compare = True\n",
    "display_nunique = True\n",
    "overwrite = True\n",
    "print_footer_notes = False\n",
    "\n",
    "# Best mapping keys are hgnc.symbol or mim\n",
    "mapping_key = \"hgnc.symbol\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a1d8b2b-67ec-4950-a607-fe3c39bf9608",
   "metadata": {},
   "source": [
    "## Check OMIM version\n",
    "If the version does not match the expected version, it is because database has been updated since the last time this code was utilized. \n",
    "* Major updates appear to be monthly. See last update [here](https://omim.org/statistics/update).\n",
    "* Data files are updated nightly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02c80874-4657-464e-b25d-1c9613f79451",
   "metadata": {},
   "outputs": [],
   "source": [
    "use_interim = False\n",
    "database_dirpath = get_dirpath(\n",
    "    \"database\", db_tag, use_temp=\"interim\" if use_interim else None\n",
    ")\n",
    "annotation_dirpath = get_dirpath(\n",
    "    \"annotation\", use_temp=\"interim\" if use_interim else None\n",
    ")\n",
    "\n",
    "# Ensure directories exist\n",
    "database_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "annotation_dirpath.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da25493-fb63-482b-839d-6b51c444f77f",
   "metadata": {},
   "outputs": [],
   "source": [
    "last_updated = get_last_updated_dates_MIM(database_dirpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d71de780-88d2-49a3-8161-975506858cd0",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "464fd8e9-6166-4075-9301-e55876d64edd",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27a1bcc5-4189-4783-82a8-d81f97f25e84",
   "metadata": {},
   "source": [
    "### Determine mapping method\n",
    "#### Load model annotations of relevance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1cfa205-82c3-4f38-a4a1-89fa716274bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"genes\"\n",
    "annotation_cols = [\"ncbigene\", \"hgnc.symbol\", \"ensembl\", \"mim\"]\n",
    "\n",
    "df_model_mappings = (\n",
    "    get_annotation_df(model.genes, annotation_cols)\n",
    "    .rename({\"id\": annotation_type}, axis=1)\n",
    "    .dropna(subset=[mapping_key])\n",
    ")\n",
    "for col in df_model_mappings.columns:\n",
    "    df_model_mappings[col] = df_model_mappings[col].apply(lambda x: split_string(x))\n",
    "    df_model_mappings = df_model_mappings.explode(col).drop_duplicates()\n",
    "df_model_mappings = df_model_mappings.sort_values(annotation_type)\n",
    "\n",
    "print(df_model_mappings.nunique(dropna=True))\n",
    "df_model_mappings = df_model_mappings.reset_index(drop=True)\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23e92897-2f20-4206-9e4b-7ecfd634dbaf",
   "metadata": {},
   "source": [
    "#### Get query IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d848c0c1-99d1-4ce0-81bf-d955d9bb7872",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Mapping key: {mapping_key}\")\n",
    "query_ids = df_model_mappings[mapping_key].dropna().unique()\n",
    "assert len(set(query_ids)) == len(query_ids), \"Duplicate IDs in list to query\"\n",
    "model_search_mapping = df_model_mappings.set_index(annotation_type)[\n",
    "    mapping_key\n",
    "].to_dict()\n",
    "print(f\"Number of model genes associated with query: {len(model_search_mapping)}\")\n",
    "print(f\"Number of unique IDs to query: {len(query_ids)}\")\n",
    "df_model_mappings[[annotation_type, mapping_key]].drop_duplicates()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6adba09a-cc64-42a8-9126-d6bf36646a8e",
   "metadata": {},
   "source": [
    "## Load MIM data\n",
    "### MIM Numbers and types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2b93834-56e5-46d9-acb0-2571cda7b4e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mim = load_data_MIM(\"mim2gene\", print_footer_notes=print_footer_notes).fillna(\"\")\n",
    "df_mim = df_mim.rename(\n",
    "    {\n",
    "        \"# MIM Number\": \"mim\",\n",
    "        \"Entrez Gene ID (NCBI)\": \"ncbigene\",\n",
    "        \"Approved Gene Symbol (HGNC)\": \"hgnc.symbol\",\n",
    "        \"Ensembl Gene ID (Ensembl)\": \"ensembl\",\n",
    "        \"MIM Entry Type (see FAQ 1.3 at https://omim.org/help/faq)\": \"type\",\n",
    "    },\n",
    "    axis=1,\n",
    ")\n",
    "df_mim.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e44ab388-0fa9-4097-978d-7cebf9423c93",
   "metadata": {},
   "source": [
    "### Morbid Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c98c13e-3eae-4604-becd-e4e1d5687153",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_morbidmap = load_data_MIM(\"morbidmap\", print_footer_notes=print_footer_notes).fillna(\n",
    "    \"\"\n",
    ")\n",
    "df_morbidmap[\"Gene/Locus And Other Related Symbols\"] = df_morbidmap[\n",
    "    \"Gene/Locus And Other Related Symbols\"\n",
    "].apply(lambda x: x.split(\", \")[0])\n",
    "rename_mapping = {\n",
    "    \"MIM Number\": \"mim\",\n",
    "    \"# Phenotype\": \"phenotype\",\n",
    "    \"Cyto Location\": \"cyto.location\",\n",
    "    \"Gene/Locus And Other Related Symbols\": \"morbid.symbol\",\n",
    "}\n",
    "df_morbidmap = df_morbidmap.loc[:, list(rename_mapping)].rename(rename_mapping, axis=1)\n",
    "df_morbidmap[\"mim.phenotype\"] = df_morbidmap[\"phenotype\"].apply(\n",
    "    lambda phenotype: build_string(\n",
    "        [\n",
    "            MIM_NUMBER_RE.search(s).group()\n",
    "            for s in split_string(phenotype, sep=\" \")\n",
    "            if MIM_NUMBER_RE.search(s)\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "df_morbidmap.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "380b1770-c0a1-4aa8-98e2-38107a7874bc",
   "metadata": {},
   "source": [
    "### Gene Map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "629a9bc0-8033-43bd-9cb8-47415f02f3b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genemap = load_data_MIM(\"genemap2\", print_footer_notes=print_footer_notes).fillna(\"\")\n",
    "# # Seperate mouse genes IDs\n",
    "# df_model_mim_final[\"Mouse Gene Symbol\"] = df_model_mim_final[\"Mouse Gene Symbol/ID\"].fillna(\"\").apply(lambda x: x.split(\" \")[0])\n",
    "# df_model_mim_final[\"Mouse Gene Symbol\"] = df_model_mim_final[\"Mouse Gene Symbol\"].str.replace(\",\", \";\")\n",
    "df_genemap[\"Mouse Gene ID\"] = (\n",
    "    df_genemap[\"Mouse Gene Symbol/ID\"]\n",
    "    .fillna(\"\")\n",
    "    .apply(lambda x: x.split(\" \")[-1].lstrip(\"(\").rstrip(\")\"))\n",
    ")\n",
    "\n",
    "df_genemap[\"Mouse Gene ID\"] = df_genemap[\"Mouse Gene ID\"].str.replace(\",\", \";\")\n",
    "# df_genemap[\"Mouse Gene ID\"] = df_genemap[\"Mouse Gene ID\"].str.replace(\"MGI:\", \"\")\n",
    "\n",
    "rename_mapping = {\n",
    "    \"# Chromosome\": \"chromosome\",\n",
    "    \"Cyto Location\": \"cyto.location\",\n",
    "    \"MIM Number\": \"mim\",\n",
    "    \"Approved Gene Symbol\": \"hgnc.symbol\",\n",
    "    \"Entrez Gene ID\": \"ncbigene\",\n",
    "    \"Ensembl Gene ID\": \"ensemble\",\n",
    "    \"Phenotypes\": \"phenotype\",\n",
    "    \"Mouse Gene ID\": \"mgi\",\n",
    "}\n",
    "df_genemap = df_genemap.loc[:, list(rename_mapping)].rename(rename_mapping, axis=1)\n",
    "df_genemap[\"mim.phenotype\"] = df_genemap[\"phenotype\"].apply(\n",
    "    lambda phenotype: build_string(\n",
    "        [\n",
    "            MIM_NUMBER_RE.search(s).group()\n",
    "            for s in split_string(phenotype, sep=\" \")\n",
    "            if MIM_NUMBER_RE.search(s)\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "df_genemap = df_genemap[\n",
    "    df_genemap[mapping_key].isin(list(model_search_mapping.values()))\n",
    "].reset_index(drop=True)\n",
    "df_genemap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c467c969-5f1b-4752-9e33-11164fce1d3a",
   "metadata": {},
   "source": [
    "### Titles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "616a2fe4-1c37-4fc2-96eb-350d07665573",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mimtitles = load_data_MIM(\"mimTitles\", print_footer_notes=print_footer_notes).fillna(\n",
    "    \"\"\n",
    ")\n",
    "rename_mapping = {\n",
    "    \"MIM Number\": \"mim\",\n",
    "    \"Preferred Title; symbol\": \"title\",\n",
    "}\n",
    "df_mimtitles = df_mimtitles.loc[:, list(rename_mapping)].rename(rename_mapping, axis=1)\n",
    "df_mimtitles[\"title\"] = df_mimtitles[\"title\"].apply(lambda x: split_string(x))\n",
    "df_mimtitles[\"title.symbol\"] = df_mimtitles[\"title\"].apply(\n",
    "    lambda x: x[-1] if len(x) > 1 else \"\"\n",
    ")\n",
    "df_mimtitles[\"title\"] = df_mimtitles[\"title\"].apply(lambda x: x[0])\n",
    "\n",
    "df_mimtitles.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f9475fb-c6b0-4b3a-a162-16cdb08d4a09",
   "metadata": {},
   "source": [
    "## Map to model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a57c96e-33a8-4f77-b24f-68579921ae47",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_model_mim = (\n",
    "    pd.merge(\n",
    "        df_model_mappings[[annotation_type, mapping_key]].drop_duplicates(),\n",
    "        df_mim,\n",
    "        left_on=mapping_key,\n",
    "        right_on=mapping_key,\n",
    "        how=\"left\",\n",
    "        suffixes=(\"\", \"_drop\"),\n",
    "    )\n",
    "    .merge(\n",
    "        df_morbidmap, left_on=\"mim\", right_on=\"mim\", how=\"left\", suffixes=(\"\", \"_drop\")\n",
    "    )\n",
    "    .merge(\n",
    "        df_genemap, left_on=\"mim\", right_on=\"mim\", how=\"left\", suffixes=(\"\", \"_drop\")\n",
    "    )\n",
    "    .merge(\n",
    "        df_mimtitles, left_on=\"mim\", right_on=\"mim\", how=\"left\", suffixes=(\"\", \"_drop\")\n",
    "    )\n",
    ")\n",
    "df_removed_mim = df_model_mim[df_model_mim[\"type\"] == \"moved/removed\"]\n",
    "df_model_mim = df_model_mim[df_model_mim[\"type\"] != \"moved/removed\"].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "df_model_mim = df_model_mim.drop(\n",
    "    [c for c in df_model_mim.columns if c.endswith(\"_drop\")], axis=1\n",
    ").drop_duplicates()\n",
    "\n",
    "df_model_mim[\"mim.all\"] = df_model_mim[[\"mim\", \"mim.phenotype\"]].apply(\n",
    "    lambda x: build_string(sorted(set(x.dropna().unique()))), axis=1\n",
    ")\n",
    "df_model_mim = df_model_mim[df_model_mim[\"type\"].str.find(\"gene\") == 0]\n",
    "df_model_mim = df_model_mim.sort_values(annotation_type).reset_index(drop=True)\n",
    "\n",
    "df = df_mimtitles[\n",
    "    df_mimtitles[\"mim\"].isin(df_model_mim[\"mim.phenotype\"].dropna())\n",
    "].rename(\n",
    "    {\n",
    "        \"mim\": \"mim.phenotype\",\n",
    "        \"title\": \"mim.phenotype.title\",\n",
    "        \"title.symbol\": \"mim.phenotype.title.symbol\",\n",
    "    },\n",
    "    axis=1,\n",
    ")\n",
    "\n",
    "df_model_mim = (\n",
    "    df_model_mim.merge(\n",
    "        df,\n",
    "        left_on=\"mim.phenotype\",\n",
    "        right_on=\"mim.phenotype\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .groupby([\"genes\", \"mim\", \"mim.phenotype\"], as_index=False)\n",
    "    .agg(lambda x: build_string(x.dropna().unique()))\n",
    ")\n",
    "df_model_mim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7adbc209-b955-4196-91ae-2b175e6ae685",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_model_mim = (\n",
    "    pd.merge(\n",
    "        df_model_mappings[[annotation_type, mapping_key]].drop_duplicates(),\n",
    "        df_mim,\n",
    "        left_on=mapping_key,\n",
    "        right_on=mapping_key,\n",
    "        how=\"left\",\n",
    "        suffixes=(\"\", \"_drop\"),\n",
    "    )\n",
    "    .merge(\n",
    "        df_morbidmap, left_on=\"mim\", right_on=\"mim\", how=\"left\", suffixes=(\"\", \"_drop\")\n",
    "    )\n",
    "    .merge(\n",
    "        df_genemap, left_on=\"mim\", right_on=\"mim\", how=\"left\", suffixes=(\"\", \"_drop\")\n",
    "    )\n",
    "    .merge(\n",
    "        df_mimtitles, left_on=\"mim\", right_on=\"mim\", how=\"left\", suffixes=(\"\", \"_drop\")\n",
    "    )\n",
    ")\n",
    "df_removed_mim = df_model_mim[df_model_mim[\"type\"] == \"moved/removed\"]\n",
    "df_model_mim = df_model_mim[df_model_mim[\"type\"] != \"moved/removed\"].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "df_model_mim = df_model_mim.drop(\n",
    "    [c for c in df_model_mim.columns if c.endswith(\"_drop\")], axis=1\n",
    ").drop_duplicates()\n",
    "\n",
    "df_model_mim = df_model_mim.replace(\"\", float(\"nan\"))\n",
    "\n",
    "\n",
    "df_model_mim[\"mim.all\"] = df_model_mim[[\"mim\", \"mim.phenotype\"]].apply(\n",
    "    lambda x: build_string(sorted(set(x.dropna().unique()))), axis=1\n",
    ")\n",
    "df_model_mim = df_model_mim[df_model_mim[\"type\"].str.find(\"gene\") == 0]\n",
    "df_model_mim = df_model_mim.sort_values(annotation_type).reset_index(drop=True)\n",
    "\n",
    "df_model_mim = (\n",
    "    df_model_mim.merge(\n",
    "        df_mimtitles[\n",
    "            df_mimtitles[\"mim\"].isin(df_model_mim[\"mim.phenotype\"].dropna())\n",
    "        ].rename(\n",
    "            {\n",
    "                \"mim\": \"mim.phenotype\",\n",
    "                \"title\": \"mim.phenotype.title\",\n",
    "                \"title.symbol\": \"mim.phenotype.title.symbol\",\n",
    "            },\n",
    "            axis=1,\n",
    "        ),\n",
    "        left_on=\"mim.phenotype\",\n",
    "        right_on=\"mim.phenotype\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    .fillna(\"\")\n",
    "    .groupby([\"genes\", \"mim\", \"mim.phenotype\"], as_index=False)\n",
    "    .agg(lambda x: build_string(x.unique()))\n",
    ")\n",
    "if overwrite:\n",
    "    df_model_mim.to_csv(\n",
    "        database_dirpath / f\"{db_tag}_{GEM_NAME}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_model_mim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "252ea025-a26c-4a31-9e4f-a8c4c0092847",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_annotations = df_model_mim.loc[\n",
    "    :, [\"genes\", \"mim.all\", \"ncbigene\", \"hgnc.symbol\", \"ensembl\", \"mgi\"]\n",
    "]\n",
    "df_annotations = df_annotations.rename({\"mim.all\": \"mim\"}, axis=1)\n",
    "df_annotations = df_annotations.drop_duplicates()\n",
    "df_annotations = df_annotations.groupby(\"genes\", as_index=False).agg(\n",
    "    lambda x: build_string(sorted(x.dropna()))\n",
    ")\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "if compare:\n",
    "    compare_on_index = [annotation_type]\n",
    "    try:\n",
    "        df_previous = pd.read_csv(\n",
    "            annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "        df_previous = df_previous.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "    except FileNotFoundError:\n",
    "        df_previous = pd.DataFrame([], columns=compare_on_index)\n",
    "    df_comparision = compare_tables(\n",
    "        df_previous.set_index(compare_on_index),\n",
    "        df_annotations.set_index(compare_on_index),\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=compare_figsize)\n",
    "    ax.yaxis.set_tick_params(labelsize=8)\n",
    "    ax = visualize_comparison(df_comparision)\n",
    "\n",
    "if display_nunique:\n",
    "    for col in df_annotations.columns:\n",
    "        df = (\n",
    "            df_annotations[col]\n",
    "            .apply(lambda x: split_string(x))\n",
    "            .explode(col)\n",
    "            .drop_duplicates()\n",
    "        )\n",
    "        print(f\"{df.name}: {df.nunique()}\")\n",
    "if overwrite:\n",
    "    df_annotations.to_csv(\n",
    "        annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "879f97ff-c76f-44f6-b7bd-afe3778b1dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_model_mim = df_annotations[[\"genes\", \"mim\"]].copy()\n",
    "df_model_mim[\"mim\"] = df_model_mim[\"mim\"].apply(lambda x: split_string(x))\n",
    "df_model_mim = df_model_mim.explode(\"mim\").drop_duplicates()\n",
    "df_model_mim\n",
    "\n",
    "df_model_mim = pd.merge(df_model_mim, df_mim[[\"mim\", \"type\"]])\n",
    "df_model_mim = df_model_mim.merge(df_mimtitles).drop_duplicates().reset_index(drop=True)\n",
    "df_duplicated = df_model_mim[df_model_mim[[\"mim\", \"type\"]].duplicated()].reset_index(\n",
    "    drop=True\n",
    ")\n",
    "print(df_model_mim[[\"mim\", \"type\"]].drop_duplicates()[\"type\"].value_counts())\n",
    "print(\"Duplicated:\\t\", df_duplicated[\"mim\"].count())\n",
    "df_model_mim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af6b9e5a-ea00-486c-852e-e273e3e0748d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_model_mim = df_annotations.copy()\n",
    "df_model_mim[\"mim\"] = df_model_mim[\"mim\"].str.split(\";\")\n",
    "df_model_mim = df_model_mim.explode(\"mim\").drop_duplicates()\n",
    "df_model_morbidmap = df_model_mim.merge(df_morbidmap)\n",
    "df_model_morbidmap = df_model_morbidmap.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"genes\",\n",
    "        \"hgnc.symbol\",\n",
    "        \"ncbigene\",\n",
    "        \"mim\",\n",
    "        \"mim.phenotype\",\n",
    "        \"phenotype\",\n",
    "        \"cyto.location\",\n",
    "    ],\n",
    "]\n",
    "if overwrite:\n",
    "    df_model_morbidmap.to_csv(\n",
    "        database_dirpath / f\"{db_tag}_{GEM_NAME}_morbidmap.tsv\", sep=\"\\t\"\n",
    "    )\n",
    "\n",
    "df_model_morbidmap"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
