{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "13bddfd6-19f9-44de-9062-5462c031dfb5",
   "metadata": {},
   "source": [
    "# Extract relevant data from dbSNP\n",
    "\n",
    "The purpose of this notebook is to extract relevant data about SNPs from the database.\n",
    "\n",
    "## Notebook Requirements:\n",
    "*  Model genes **must** have at least one of the following annotations stored in the `object.annotation`. Values are expected to be seperated by semicolons. Accepted keys currently include:\n",
    "    * `\"dbsnp\"`\n",
    "*  Note: Requires internet connection to download information from [dbSNP](https://www.ncbi.nlm.nih.gov/snp/).\n",
    "*  Utilizes the [Bio.Entrez package](https://biopython.org/docs/latest/api/Bio.Entrez.html).\n",
    "\n",
    "### Citations\n",
    "Phan L, Zhang H, Wang Q, Villamarin R, Hefferon T, Ramanathan A, Kattman B. The evolution of dbSNP: 25 years of impact in genomic research. Nucleic Acids Res. 2025 Jan 6;53(D1):D925-D931. doi: 10.1093/nar/gkae977. PMID: 39530225; PMCID: PMC11701571.\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49fed61a-e9eb-47f9-b97e-8edeff408452",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.1\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.12.3\n",
      "bio                                   1.6.2\n",
      "cobra                                0.29.0\n",
      "depinfo                               2.2.0\n",
      "kaleido                               0.2.1\n",
      "matplotlib                            3.8.2\n",
      "memote                               0.17.0\n",
      "networkx                              3.2.1\n",
      "notebook                              7.0.7\n",
      "openpyxl                              3.1.2\n",
      "pandas                                2.2.0\n",
      "pre-commit                            3.6.0\n",
      "pyvis                                 0.3.2\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.31.0\n",
      "scipy                                1.12.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip        23.3.1\n",
      "setuptools 68.2.2\n",
      "wheel      0.41.2\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Darwin  24.4.0-x86_64\n",
      "CPython        3.11.7\n"
     ]
    }
   ],
   "source": [
    "from collections import defaultdict\n",
    "from xml.etree import ElementTree\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import Entrez\n",
    "from rbc_gem_utils import (\n",
    "    GEM_NAME,\n",
    "    build_string,\n",
    "    get_annotation_df,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    split_string,\n",
    ")\n",
    "from rbc_gem_utils.util import has_value_type\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8477f24-711b-4c58-a5eb-bb1f0fc7c460",
   "metadata": {},
   "source": [
    "## Set notebook options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d646579b-539a-4f70-a93d-72107d6f3d81",
   "metadata": {},
   "outputs": [],
   "source": [
    "overwrite = True\n",
    "\n",
    "# Make sure to use your own email!\n",
    "email = \"EMAILADDRESS\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6da6a907-3339-4f93-868a-01622f9e6de6",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "13d4753e-2d6d-44d9-bad0-7a477220b3a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14e1a8f10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14e1a8f10>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1a26d35-de9c-4390-aca2-a6c8481eac4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes: 820\n",
      "dbsnp: 5070\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPE</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPIA</td>\n",
       "      <td>rs121918591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs145813597;rs1042079;rs930337;rs149975952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKR7A2</td>\n",
       "      <td>rs6670759;rs1043657;rs859208;rs859210;rs223120...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRM</td>\n",
       "      <td>rs1049932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>USP5</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>VCPIP1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>VPS4B</td>\n",
       "      <td>rs17688948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>WDR77</td>\n",
       "      <td>rs7416672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819</th>\n",
       "      <td>YES1</td>\n",
       "      <td>rs34580680;rs35126906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>820 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes                                              dbsnp\n",
       "0       RPE                                               <NA>\n",
       "1      RPIA                                        rs121918591\n",
       "2      SORD         rs145813597;rs1042079;rs930337;rs149975952\n",
       "3    AKR7A2  rs6670759;rs1043657;rs859208;rs859210;rs223120...\n",
       "4       SRM                                          rs1049932\n",
       "..      ...                                                ...\n",
       "815    USP5                                               <NA>\n",
       "816  VCPIP1                                               <NA>\n",
       "817   VPS4B                                         rs17688948\n",
       "818   WDR77                                          rs7416672\n",
       "819    YES1                              rs34580680;rs35126906\n",
       "\n",
       "[820 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_type = \"genes\"\n",
    "df_model_mappings = get_annotation_df(\n",
    "    getattr(model, annotation_type), [\"dbsnp\"]\n",
    ").rename({\"id\": annotation_type}, axis=1)\n",
    "for col in df_model_mappings.columns:\n",
    "    df = (\n",
    "        df_model_mappings[col]\n",
    "        .apply(lambda x: split_string(x))\n",
    "        .explode(col)\n",
    "        .drop_duplicates()\n",
    "    )\n",
    "    print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21905f9a-fb31-466a-84c5-cf9ee6681287",
   "metadata": {},
   "source": [
    "### Get IDs for query from model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ae646a47-2606-4d32-bb97-bafb1413a2ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPIA</td>\n",
       "      <td>rs121918591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs145813597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs1042079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs930337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs149975952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>814</th>\n",
       "      <td>UFD1</td>\n",
       "      <td>rs17744624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>VPS4B</td>\n",
       "      <td>rs17688948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>WDR77</td>\n",
       "      <td>rs7416672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819</th>\n",
       "      <td>YES1</td>\n",
       "      <td>rs34580680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819</th>\n",
       "      <td>YES1</td>\n",
       "      <td>rs35126906</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5070 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genes        dbsnp\n",
       "1     RPIA  rs121918591\n",
       "2     SORD  rs145813597\n",
       "2     SORD    rs1042079\n",
       "2     SORD     rs930337\n",
       "2     SORD  rs149975952\n",
       "..     ...          ...\n",
       "814   UFD1   rs17744624\n",
       "817  VPS4B   rs17688948\n",
       "818  WDR77    rs7416672\n",
       "819   YES1   rs34580680\n",
       "819   YES1   rs35126906\n",
       "\n",
       "[5070 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_snps = df_model_mappings.loc[:, [\"genes\", \"dbsnp\"]].copy()\n",
    "df_model_snps[\"dbsnp\"] = df_model_mappings[\"dbsnp\"].apply(split_string)\n",
    "df_model_snps = df_model_snps.explode(\"dbsnp\")\n",
    "df_model_snps = df_model_snps.dropna().drop_duplicates()\n",
    "rs_ids = df_model_snps[\"dbsnp\"].unique()\n",
    "df_model_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cdc8c330-2057-4c12-820d-4aa9cdf72b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_results_dbSNP(email, rs_ids):\n",
    "    Entrez.email = email\n",
    "    search = Entrez.efetch(db=\"snp\", retmode=\"text\", id=rs_ids)\n",
    "    all_lines = search.readlines()\n",
    "    search.close()\n",
    "    return all_lines"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e996f2bd-5293-4f6d-99b5-345a172f0840",
   "metadata": {},
   "source": [
    "### Get results from SNP as a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e253df0-8aee-48f6-87a2-6f9198e5ad18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching results for batch 1  (500/5070)\n",
      "Fetching results for batch 2  (1000/5070)\n",
      "Fetching results for batch 3  (1500/5070)\n",
      "Fetching results for batch 4  (2000/5070)\n",
      "Fetching results for batch 5  (2500/5070)\n",
      "Fetching results for batch 6  (3000/5070)\n",
      "Fetching results for batch 7  (3500/5070)\n",
      "Fetching results for batch 8  (4000/5070)\n",
      "Fetching results for batch 9  (4500/5070)\n",
      "Fetching results for batch 10  (5000/5070)\n",
      "Fetching results for batch 11  (5070/5070)\n",
      "\n",
      "Number of failed IDs: 3\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GLOBAL_SAMPLESIZE</th>\n",
       "      <th>CLINICAL_SIGNIFICANCE</th>\n",
       "      <th>ACC</th>\n",
       "      <th>CHR</th>\n",
       "      <th>HANDLE</th>\n",
       "      <th>SPDI</th>\n",
       "      <th>FXN_CLASS</th>\n",
       "      <th>VALIDATED</th>\n",
       "      <th>DOCSUM</th>\n",
       "      <th>...</th>\n",
       "      <th>SS</th>\n",
       "      <th>ALLELE</th>\n",
       "      <th>SNP_CLASS</th>\n",
       "      <th>CHRPOS</th>\n",
       "      <th>CHRPOS_PREV_ASSM</th>\n",
       "      <th>SNP_ID_SORT</th>\n",
       "      <th>CLINICAL_SORT</th>\n",
       "      <th>CHRPOS_SORT</th>\n",
       "      <th>MERGED_SORT</th>\n",
       "      <th>error</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbsnp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs121918591</th>\n",
       "      <td>121918591</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>PAGE_CC,TOPMED,GNOMAD,OMIM-CURATED-RECORDS,ILL...</td>\n",
       "      <td>NC_000002.12:88729278:C:T</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000002.12:g.88729279C&gt;T,NC_000002.11:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>275518424,1958438289,2732810607,3021993357,365...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>2:88729279</td>\n",
       "      <td>2:89028797</td>\n",
       "      <td>0121918591</td>\n",
       "      <td>1</td>\n",
       "      <td>0088729279</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs145813597</th>\n",
       "      <td>145813597</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic-likely-pathogenic,uncertain-signifi...</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>TOPMED,EVA_EXAC,SWEGEN,HUGCELL_USP,ILLUMINA,GN...</td>\n",
       "      <td>NC_000015.10:45065302:C:A,NC_000015.10:4506530...</td>\n",
       "      <td>missense_variant,coding_sequence_variant,non_c...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45065303C&gt;A,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>342402496,464140601,491493966,491700043,991737...</td>\n",
       "      <td>H</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45065303</td>\n",
       "      <td>15:45357501</td>\n",
       "      <td>0145813597</td>\n",
       "      <td>1</td>\n",
       "      <td>0045065303</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1042079</th>\n",
       "      <td>1042079</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>ILLUMINA,BUSHMAN,GNOMAD,GMI,KHV_HUMAN_GENOMES,...</td>\n",
       "      <td>NC_000015.10:45068981:A:G,NC_000015.10:4506898...</td>\n",
       "      <td>missense_variant,non_coding_transcript_variant...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45068982A&gt;G,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1509008,4403349,76874733,95213358,156480828,15...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45068982</td>\n",
       "      <td>15:45361180</td>\n",
       "      <td>0001042079</td>\n",
       "      <td>1</td>\n",
       "      <td>0045068982</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs930337</th>\n",
       "      <td>930337</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>ILLUMINA,SYSTEMSBIOZJU,GENOMICARE,BUSHMAN,EVA_...</td>\n",
       "      <td>NC_000015.10:45072335:A:C,NC_000015.10:4507233...</td>\n",
       "      <td>missense_variant,non_coding_transcript_variant...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45072336A&gt;C,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1368774,2558685,3214527,5252928,21245444,24802...</td>\n",
       "      <td>H</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45072336</td>\n",
       "      <td>15:45364534</td>\n",
       "      <td>0000930337</td>\n",
       "      <td>1</td>\n",
       "      <td>0045072336</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs149975952</th>\n",
       "      <td>149975952</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>WEILL_CORNELL_DGM,EVA_EXAC,ILLUMINA,SGDP_PRJ,G...</td>\n",
       "      <td>NC_000015.10:45073419:G:A,NC_000015.10:4507341...</td>\n",
       "      <td>missense_variant,coding_sequence_variant,non_c...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45073420G&gt;A,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>342402507,489048604,491493972,1353273008,16918...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45073420</td>\n",
       "      <td>15:45365618</td>\n",
       "      <td>0149975952</td>\n",
       "      <td>1</td>\n",
       "      <td>0045073420</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs17744624</th>\n",
       "      <td>17744624</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>USC_VALOUEV,JJLAB,EXOME_CHIP,KRGDB,EVA_EXAC,GO...</td>\n",
       "      <td>NC_000022.11:19467906:G:A,NC_000022.11:1946790...</td>\n",
       "      <td>missense_variant,coding_sequence_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.19467907G&gt;A,NC_000022.11:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>24701832,48413642,69277235,69376500,74818968,8...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:19467907</td>\n",
       "      <td>22:19455430</td>\n",
       "      <td>0017744624</td>\n",
       "      <td>0</td>\n",
       "      <td>0019467907</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs17688948</th>\n",
       "      <td>17688948</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>18</td>\n",
       "      <td>EVA_UK10K_ALSPAC,EVA-GONL,EVA_DECODE,EVA_EXAC,...</td>\n",
       "      <td>NC_000018.10:63410411:G:C</td>\n",
       "      <td>missense_variant,coding_sequence_variant,genic...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000018.10:g.63410412G&gt;C,NC_000018.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>24492533,69217429,74820330,160453165,237511803...</td>\n",
       "      <td>S</td>\n",
       "      <td>snv</td>\n",
       "      <td>18:63410412</td>\n",
       "      <td>18:61077645</td>\n",
       "      <td>0017688948</td>\n",
       "      <td>0</td>\n",
       "      <td>0063410412</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs7416672</th>\n",
       "      <td>7416672</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>TOPMED,SC_JCM,GNOMAD,ILLUMINA</td>\n",
       "      <td>NC_000001.11:111448776:C:A,NC_000001.11:111448...</td>\n",
       "      <td>missense_variant,2KB_upstream_variant,coding_s...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000001.11:g.111448777C&gt;A,NC_000001.11:...</td>\n",
       "      <td>...</td>\n",
       "      <td>11204945,160878283,482563863,2746395432,275957...</td>\n",
       "      <td>H</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:111448777</td>\n",
       "      <td>1:111991399</td>\n",
       "      <td>0007416672</td>\n",
       "      <td>0</td>\n",
       "      <td>0111448777</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs34580680</th>\n",
       "      <td>34580680</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>18</td>\n",
       "      <td>EVA_EXAC,KOGIC,GRF,PERLEGEN,GNOMAD,1000G_HIGH_...</td>\n",
       "      <td>NC_000018.10:745839:T:C</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000018.10:g.745840T&gt;C,NC_000018.9:g.74...</td>\n",
       "      <td>...</td>\n",
       "      <td>48533661,69204557,160622497,339897863,48170461...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>18:745840</td>\n",
       "      <td>18:745840</td>\n",
       "      <td>0034580680</td>\n",
       "      <td>0</td>\n",
       "      <td>0000745840</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs35126906</th>\n",
       "      <td>35126906</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000018.10</td>\n",
       "      <td>18</td>\n",
       "      <td>GNOMAD,CANCER-GENOME,ILLUMINA,AFFY,EVA_EXAC,PE...</td>\n",
       "      <td>NC_000018.10:743294:T:C</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000018.10:g.743295T&gt;C,NC_000018.9:g.74...</td>\n",
       "      <td>...</td>\n",
       "      <td>48533660,69204555,1693062780,2743159113,298510...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>18:743295</td>\n",
       "      <td>18:743295</td>\n",
       "      <td>0035126906</td>\n",
       "      <td>0</td>\n",
       "      <td>0000743295</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5067 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNP_ID GLOBAL_SAMPLESIZE  \\\n",
       "dbsnp                                      \n",
       "rs121918591  121918591                 0   \n",
       "rs145813597  145813597                 0   \n",
       "rs1042079      1042079                 0   \n",
       "rs930337        930337                 0   \n",
       "rs149975952  149975952                 0   \n",
       "...                ...               ...   \n",
       "rs17744624    17744624                 0   \n",
       "rs17688948    17688948                 0   \n",
       "rs7416672      7416672                 0   \n",
       "rs34580680    34580680                 0   \n",
       "rs35126906    35126906                 0   \n",
       "\n",
       "                                         CLINICAL_SIGNIFICANCE           ACC  \\\n",
       "dbsnp                                                                          \n",
       "rs121918591                                         pathogenic  NC_000002.12   \n",
       "rs145813597  pathogenic-likely-pathogenic,uncertain-signifi...  NC_000015.10   \n",
       "rs1042079                                               benign  NC_000015.10   \n",
       "rs930337                                                benign  NC_000015.10   \n",
       "rs149975952                                             benign  NC_000015.10   \n",
       "...                                                        ...           ...   \n",
       "rs17744624                                                 NaN  NC_000022.11   \n",
       "rs17688948                                                 NaN  NC_000018.10   \n",
       "rs7416672                                                  NaN  NC_000001.11   \n",
       "rs34580680                                                 NaN  NC_000018.10   \n",
       "rs35126906                                                 NaN  NC_000018.10   \n",
       "\n",
       "            CHR                                             HANDLE  \\\n",
       "dbsnp                                                                \n",
       "rs121918591   2  PAGE_CC,TOPMED,GNOMAD,OMIM-CURATED-RECORDS,ILL...   \n",
       "rs145813597  15  TOPMED,EVA_EXAC,SWEGEN,HUGCELL_USP,ILLUMINA,GN...   \n",
       "rs1042079    15  ILLUMINA,BUSHMAN,GNOMAD,GMI,KHV_HUMAN_GENOMES,...   \n",
       "rs930337     15  ILLUMINA,SYSTEMSBIOZJU,GENOMICARE,BUSHMAN,EVA_...   \n",
       "rs149975952  15  WEILL_CORNELL_DGM,EVA_EXAC,ILLUMINA,SGDP_PRJ,G...   \n",
       "...          ..                                                ...   \n",
       "rs17744624   22  USC_VALOUEV,JJLAB,EXOME_CHIP,KRGDB,EVA_EXAC,GO...   \n",
       "rs17688948   18  EVA_UK10K_ALSPAC,EVA-GONL,EVA_DECODE,EVA_EXAC,...   \n",
       "rs7416672     1                      TOPMED,SC_JCM,GNOMAD,ILLUMINA   \n",
       "rs34580680   18  EVA_EXAC,KOGIC,GRF,PERLEGEN,GNOMAD,1000G_HIGH_...   \n",
       "rs35126906   18  GNOMAD,CANCER-GENOME,ILLUMINA,AFFY,EVA_EXAC,PE...   \n",
       "\n",
       "                                                          SPDI  \\\n",
       "dbsnp                                                            \n",
       "rs121918591                          NC_000002.12:88729278:C:T   \n",
       "rs145813597  NC_000015.10:45065302:C:A,NC_000015.10:4506530...   \n",
       "rs1042079    NC_000015.10:45068981:A:G,NC_000015.10:4506898...   \n",
       "rs930337     NC_000015.10:45072335:A:C,NC_000015.10:4507233...   \n",
       "rs149975952  NC_000015.10:45073419:G:A,NC_000015.10:4507341...   \n",
       "...                                                        ...   \n",
       "rs17744624   NC_000022.11:19467906:G:A,NC_000022.11:1946790...   \n",
       "rs17688948                           NC_000018.10:63410411:G:C   \n",
       "rs7416672    NC_000001.11:111448776:C:A,NC_000001.11:111448...   \n",
       "rs34580680                             NC_000018.10:745839:T:C   \n",
       "rs35126906                             NC_000018.10:743294:T:C   \n",
       "\n",
       "                                                     FXN_CLASS  \\\n",
       "dbsnp                                                            \n",
       "rs121918591           coding_sequence_variant,missense_variant   \n",
       "rs145813597  missense_variant,coding_sequence_variant,non_c...   \n",
       "rs1042079    missense_variant,non_coding_transcript_variant...   \n",
       "rs930337     missense_variant,non_coding_transcript_variant...   \n",
       "rs149975952  missense_variant,coding_sequence_variant,non_c...   \n",
       "...                                                        ...   \n",
       "rs17744624            missense_variant,coding_sequence_variant   \n",
       "rs17688948   missense_variant,coding_sequence_variant,genic...   \n",
       "rs7416672    missense_variant,2KB_upstream_variant,coding_s...   \n",
       "rs34580680            coding_sequence_variant,missense_variant   \n",
       "rs35126906            coding_sequence_variant,missense_variant   \n",
       "\n",
       "                                   VALIDATED  \\\n",
       "dbsnp                                          \n",
       "rs121918591  by-frequency,by-alfa,by-cluster   \n",
       "rs145813597  by-frequency,by-alfa,by-cluster   \n",
       "rs1042079    by-frequency,by-alfa,by-cluster   \n",
       "rs930337     by-frequency,by-alfa,by-cluster   \n",
       "rs149975952  by-frequency,by-alfa,by-cluster   \n",
       "...                                      ...   \n",
       "rs17744624   by-frequency,by-alfa,by-cluster   \n",
       "rs17688948   by-frequency,by-alfa,by-cluster   \n",
       "rs7416672    by-frequency,by-alfa,by-cluster   \n",
       "rs34580680   by-frequency,by-alfa,by-cluster   \n",
       "rs35126906   by-frequency,by-alfa,by-cluster   \n",
       "\n",
       "                                                        DOCSUM  ...  \\\n",
       "dbsnp                                                           ...   \n",
       "rs121918591  HGVS=NC_000002.12:g.88729279C>T,NC_000002.11:g...  ...   \n",
       "rs145813597  HGVS=NC_000015.10:g.45065303C>A,NC_000015.10:g...  ...   \n",
       "rs1042079    HGVS=NC_000015.10:g.45068982A>G,NC_000015.10:g...  ...   \n",
       "rs930337     HGVS=NC_000015.10:g.45072336A>C,NC_000015.10:g...  ...   \n",
       "rs149975952  HGVS=NC_000015.10:g.45073420G>A,NC_000015.10:g...  ...   \n",
       "...                                                        ...  ...   \n",
       "rs17744624   HGVS=NC_000022.11:g.19467907G>A,NC_000022.11:g...  ...   \n",
       "rs17688948   HGVS=NC_000018.10:g.63410412G>C,NC_000018.9:g....  ...   \n",
       "rs7416672    HGVS=NC_000001.11:g.111448777C>A,NC_000001.11:...  ...   \n",
       "rs34580680   HGVS=NC_000018.10:g.745840T>C,NC_000018.9:g.74...  ...   \n",
       "rs35126906   HGVS=NC_000018.10:g.743295T>C,NC_000018.9:g.74...  ...   \n",
       "\n",
       "                                                            SS ALLELE  \\\n",
       "dbsnp                                                                   \n",
       "rs121918591  275518424,1958438289,2732810607,3021993357,365...      Y   \n",
       "rs145813597  342402496,464140601,491493966,491700043,991737...      H   \n",
       "rs1042079    1509008,4403349,76874733,95213358,156480828,15...      D   \n",
       "rs930337     1368774,2558685,3214527,5252928,21245444,24802...      H   \n",
       "rs149975952  342402507,489048604,491493972,1353273008,16918...      D   \n",
       "...                                                        ...    ...   \n",
       "rs17744624   24701832,48413642,69277235,69376500,74818968,8...      N   \n",
       "rs17688948   24492533,69217429,74820330,160453165,237511803...      S   \n",
       "rs7416672    11204945,160878283,482563863,2746395432,275957...      H   \n",
       "rs34580680   48533661,69204557,160622497,339897863,48170461...      Y   \n",
       "rs35126906   48533660,69204555,1693062780,2743159113,298510...      Y   \n",
       "\n",
       "            SNP_CLASS       CHRPOS CHRPOS_PREV_ASSM SNP_ID_SORT CLINICAL_SORT  \\\n",
       "dbsnp                                                                           \n",
       "rs121918591       snv   2:88729279       2:89028797  0121918591             1   \n",
       "rs145813597       snv  15:45065303      15:45357501  0145813597             1   \n",
       "rs1042079         snv  15:45068982      15:45361180  0001042079             1   \n",
       "rs930337          snv  15:45072336      15:45364534  0000930337             1   \n",
       "rs149975952       snv  15:45073420      15:45365618  0149975952             1   \n",
       "...               ...          ...              ...         ...           ...   \n",
       "rs17744624        snv  22:19467907      22:19455430  0017744624             0   \n",
       "rs17688948        snv  18:63410412      18:61077645  0017688948             0   \n",
       "rs7416672         snv  1:111448777      1:111991399  0007416672             0   \n",
       "rs34580680        snv    18:745840        18:745840  0034580680             0   \n",
       "rs35126906        snv    18:743295        18:743295  0035126906             0   \n",
       "\n",
       "            CHRPOS_SORT MERGED_SORT error  \n",
       "dbsnp                                      \n",
       "rs121918591  0088729279           0   NaN  \n",
       "rs145813597  0045065303           0   NaN  \n",
       "rs1042079    0045068982           0   NaN  \n",
       "rs930337     0045072336           0   NaN  \n",
       "rs149975952  0045073420           0   NaN  \n",
       "...                 ...         ...   ...  \n",
       "rs17744624   0019467907           0   NaN  \n",
       "rs17688948   0063410412           0   NaN  \n",
       "rs7416672    0111448777           0   NaN  \n",
       "rs34580680   0000745840           0   NaN  \n",
       "rs35126906   0000743295           0   NaN  \n",
       "\n",
       "[5067 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = defaultdict(dict)\n",
    "batch_size = 500\n",
    "for batch_idx, batch in enumerate(np.arange(0, len(rs_ids), batch_size), start=0):\n",
    "    query_ids = rs_ids[batch : batch + batch_size]\n",
    "    print(\n",
    "        f\"Fetching results for batch {batch_idx + 1}  ({batch + len(query_ids)}/{len(rs_ids)})\"\n",
    "    )\n",
    "    all_lines = fetch_results_dbSNP(email=email, rs_ids=query_ids)\n",
    "    for entry_idx, xml_str in enumerate(all_lines, start=batch_idx * batch_size):\n",
    "        entry = ElementTree.fromstring(xml_str)\n",
    "        data[rs_ids[entry_idx]].update(\n",
    "            {e.tag: e.text for e in entry if has_value_type(e)}\n",
    "        )\n",
    "\n",
    "# Make into DataFrame\n",
    "df_snps_all = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "failed_ids = list(df_snps_all[\"error\"].dropna().index)\n",
    "print(f\"\\nNumber of failed IDs: {len(failed_ids)}\")\n",
    "df_snps_all.index.name = \"dbsnp\"\n",
    "df_snps_all = df_snps_all[~df_snps_all[\"error\"].notna()]\n",
    "df_snps_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "901fb1a6-0667-4b1b-a90a-657395466b5b",
   "metadata": {},
   "source": [
    "### Parse and reformat results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "be8c543e-c172-4d1c-bd42-8bbdf88d8ff8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>handle</th>\n",
       "      <th>spdi</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>...</th>\n",
       "      <th>ss</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "      <th>sequence_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>rs334</td>\n",
       "      <td>334</td>\n",
       "      <td>not-provided;protective;likely-benign;pathogen...</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>ILLUMINA;PERLEGEN;CLINSEQ_SNP;1000G_HIGH_COVER...</td>\n",
       "      <td>NC_000011.10:5227001:T:A;NC_000011.10:5227001:...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>335;24811263;48419811;49850538;49850540;498505...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>11:5248232</td>\n",
       "      <td>0000000334</td>\n",
       "      <td>1</td>\n",
       "      <td>0005227002</td>\n",
       "      <td>0</td>\n",
       "      <td>[T:A, T:C, T:G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AGT</td>\n",
       "      <td>rs699</td>\n",
       "      <td>699</td>\n",
       "      <td>risk-factor;benign</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>LEE;CGAP-GAI;EVA-GONL;WEILL_CORNELL_DGM;KRGDB;...</td>\n",
       "      <td>NC_000001.11:230710047:A:C;NC_000001.11:230710...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>704;5428;6484;1510443;3240877;3474289;4404260;...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>1:230710048</td>\n",
       "      <td>1:230845794</td>\n",
       "      <td>0000000699</td>\n",
       "      <td>1</td>\n",
       "      <td>0230710048</td>\n",
       "      <td>0</td>\n",
       "      <td>[A:C, A:G, A:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSTP1</td>\n",
       "      <td>rs1695</td>\n",
       "      <td>1695</td>\n",
       "      <td>benign;not-provided</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>SSMP;YEGNASUBRAMANIAN_LAB;PERLEGEN;CLINSEQ_SNP...</td>\n",
       "      <td>NC_000011.10:67585217:A:G;NC_000011.10:6758521...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>1720;5330;9110;1390210;1545478;3178003;5586573...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:67585218</td>\n",
       "      <td>11:67352689</td>\n",
       "      <td>0000001695</td>\n",
       "      <td>1</td>\n",
       "      <td>0067585218</td>\n",
       "      <td>0</td>\n",
       "      <td>[A:G, A:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MPO</td>\n",
       "      <td>rs2759</td>\n",
       "      <td>2759</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>1000GENOMES;UAEU_GENOMICS_LAB;EVA_MGP;EVA_DECO...</td>\n",
       "      <td>NC_000017.11:58270744:T:C</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>2789;16338615;23791841;38350281;69199282;74862...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>17:58270745</td>\n",
       "      <td>17:56348106</td>\n",
       "      <td>0000002759</td>\n",
       "      <td>0</td>\n",
       "      <td>0058270745</td>\n",
       "      <td>0</td>\n",
       "      <td>[T:C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FKBP2</td>\n",
       "      <td>rs4672</td>\n",
       "      <td>4672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>BIOINF_KMB_FNS_UNIBA;ILLUMINA;JJLAB;EVA_DECODE...</td>\n",
       "      <td>NC_000011.10:64242406:G:A;NC_000011.10:6424240...</td>\n",
       "      <td>intron_variant;coding_sequence_variant;3_prime...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>1555019;4436250;20825221;24790172;39886556;752...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:64242407</td>\n",
       "      <td>11:64009879</td>\n",
       "      <td>0000004672</td>\n",
       "      <td>0</td>\n",
       "      <td>0064242407</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:A, G:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5062</th>\n",
       "      <td>CLCN3</td>\n",
       "      <td>rs2150238934</td>\n",
       "      <td>2150238934</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2150238934</td>\n",
       "      <td>0</td>\n",
       "      <td>99999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5063</th>\n",
       "      <td>CLCN3</td>\n",
       "      <td>rs2150254146</td>\n",
       "      <td>2150254146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2150254146</td>\n",
       "      <td>0</td>\n",
       "      <td>99999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5064</th>\n",
       "      <td>CLCN3</td>\n",
       "      <td>rs2150267036</td>\n",
       "      <td>2150267036</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2150267036</td>\n",
       "      <td>0</td>\n",
       "      <td>99999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5065</th>\n",
       "      <td>CLCN3</td>\n",
       "      <td>rs2150274994</td>\n",
       "      <td>2150274994</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2150274994</td>\n",
       "      <td>0</td>\n",
       "      <td>99999999999</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5066</th>\n",
       "      <td>IDH2</td>\n",
       "      <td>rs2151549649</td>\n",
       "      <td>2151549649</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>EVA</td>\n",
       "      <td>NC_000015.10:90088647:G:A;NC_000015.10:9008864...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>8936221570</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:90088648</td>\n",
       "      <td>15:90631880</td>\n",
       "      <td>2151549649</td>\n",
       "      <td>0</td>\n",
       "      <td>0090088648</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:A, G:C, G:T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5067 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes         dbsnp      snp_id  \\\n",
       "0       HBB         rs334         334   \n",
       "1       AGT         rs699         699   \n",
       "2     GSTP1        rs1695        1695   \n",
       "3       MPO        rs2759        2759   \n",
       "4     FKBP2        rs4672        4672   \n",
       "...     ...           ...         ...   \n",
       "5062  CLCN3  rs2150238934  2150238934   \n",
       "5063  CLCN3  rs2150254146  2150254146   \n",
       "5064  CLCN3  rs2150267036  2150267036   \n",
       "5065  CLCN3  rs2150274994  2150274994   \n",
       "5066   IDH2  rs2151549649  2151549649   \n",
       "\n",
       "                                  clinical_significance           acc  chr  \\\n",
       "0     not-provided;protective;likely-benign;pathogen...  NC_000011.10   11   \n",
       "1                                    risk-factor;benign  NC_000001.11    1   \n",
       "2                                   benign;not-provided  NC_000011.10   11   \n",
       "3                                                   NaN  NC_000017.11   17   \n",
       "4                                                   NaN  NC_000011.10   11   \n",
       "...                                                 ...           ...  ...   \n",
       "5062                                                NaN           NaN  NaN   \n",
       "5063                                                NaN           NaN  NaN   \n",
       "5064                                                NaN           NaN  NaN   \n",
       "5065                                                NaN           NaN  NaN   \n",
       "5066                                                NaN  NC_000015.10   15   \n",
       "\n",
       "                                                 handle  \\\n",
       "0     ILLUMINA;PERLEGEN;CLINSEQ_SNP;1000G_HIGH_COVER...   \n",
       "1     LEE;CGAP-GAI;EVA-GONL;WEILL_CORNELL_DGM;KRGDB;...   \n",
       "2     SSMP;YEGNASUBRAMANIAN_LAB;PERLEGEN;CLINSEQ_SNP...   \n",
       "3     1000GENOMES;UAEU_GENOMICS_LAB;EVA_MGP;EVA_DECO...   \n",
       "4     BIOINF_KMB_FNS_UNIBA;ILLUMINA;JJLAB;EVA_DECODE...   \n",
       "...                                                 ...   \n",
       "5062                                                NaN   \n",
       "5063                                                NaN   \n",
       "5064                                                NaN   \n",
       "5065                                                NaN   \n",
       "5066                                                EVA   \n",
       "\n",
       "                                                   spdi  \\\n",
       "0     NC_000011.10:5227001:T:A;NC_000011.10:5227001:...   \n",
       "1     NC_000001.11:230710047:A:C;NC_000001.11:230710...   \n",
       "2     NC_000011.10:67585217:A:G;NC_000011.10:6758521...   \n",
       "3                             NC_000017.11:58270744:T:C   \n",
       "4     NC_000011.10:64242406:G:A;NC_000011.10:6424240...   \n",
       "...                                                 ...   \n",
       "5062                                                NaN   \n",
       "5063                                                NaN   \n",
       "5064                                                NaN   \n",
       "5065                                                NaN   \n",
       "5066  NC_000015.10:90088647:G:A;NC_000015.10:9008864...   \n",
       "\n",
       "                                              fxn_class  \\\n",
       "0              coding_sequence_variant;missense_variant   \n",
       "1              coding_sequence_variant;missense_variant   \n",
       "2              coding_sequence_variant;missense_variant   \n",
       "3              coding_sequence_variant;missense_variant   \n",
       "4     intron_variant;coding_sequence_variant;3_prime...   \n",
       "...                                                 ...   \n",
       "5062                                                NaN   \n",
       "5063                                                NaN   \n",
       "5064                                                NaN   \n",
       "5065                                                NaN   \n",
       "5066           missense_variant;coding_sequence_variant   \n",
       "\n",
       "                            validated  ...  \\\n",
       "0     by-frequency;by-alfa;by-cluster  ...   \n",
       "1     by-frequency;by-alfa;by-cluster  ...   \n",
       "2     by-frequency;by-alfa;by-cluster  ...   \n",
       "3     by-frequency;by-alfa;by-cluster  ...   \n",
       "4     by-frequency;by-alfa;by-cluster  ...   \n",
       "...                               ...  ...   \n",
       "5062                              NaN  ...   \n",
       "5063                              NaN  ...   \n",
       "5064                              NaN  ...   \n",
       "5065                              NaN  ...   \n",
       "5066                              NaN  ...   \n",
       "\n",
       "                                                     ss allele snp_class  \\\n",
       "0     335;24811263;48419811;49850538;49850540;498505...      N       snv   \n",
       "1     704;5428;6484;1510443;3240877;3474289;4404260;...      N       snv   \n",
       "2     1720;5330;9110;1390210;1545478;3178003;5586573...      D       snv   \n",
       "3     2789;16338615;23791841;38350281;69199282;74862...      Y       snv   \n",
       "4     1555019;4436250;20825221;24790172;39886556;752...      D       snv   \n",
       "...                                                 ...    ...       ...   \n",
       "5062                                                NaN    NaN       NaN   \n",
       "5063                                                NaN    NaN       NaN   \n",
       "5064                                                NaN    NaN       NaN   \n",
       "5065                                                NaN    NaN       NaN   \n",
       "5066                                         8936221570      N       snv   \n",
       "\n",
       "           chrpos chrpos_prev_assm snp_id_sort clinical_sort  chrpos_sort  \\\n",
       "0      11:5227002       11:5248232  0000000334             1   0005227002   \n",
       "1     1:230710048      1:230845794  0000000699             1   0230710048   \n",
       "2     11:67585218      11:67352689  0000001695             1   0067585218   \n",
       "3     17:58270745      17:56348106  0000002759             0   0058270745   \n",
       "4     11:64242407      11:64009879  0000004672             0   0064242407   \n",
       "...           ...              ...         ...           ...          ...   \n",
       "5062          NaN              NaN  2150238934             0  99999999999   \n",
       "5063          NaN              NaN  2150254146             0  99999999999   \n",
       "5064          NaN              NaN  2150267036             0  99999999999   \n",
       "5065          NaN              NaN  2150274994             0  99999999999   \n",
       "5066  15:90088648      15:90631880  2151549649             0   0090088648   \n",
       "\n",
       "     merged_sort  sequence_change  \n",
       "0              0  [T:A, T:C, T:G]  \n",
       "1              0  [A:C, A:G, A:T]  \n",
       "2              0       [A:G, A:T]  \n",
       "3              0            [T:C]  \n",
       "4              0       [G:A, G:T]  \n",
       "...          ...              ...  \n",
       "5062           0              NaN  \n",
       "5063           0              NaN  \n",
       "5064           0              NaN  \n",
       "5065           0              NaN  \n",
       "5066           0  [G:A, G:C, G:T]  \n",
       "\n",
       "[5067 rows x 26 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_snps = df_model_mappings.loc[:, [\"genes\", \"dbsnp\"]].copy()\n",
    "df_model_snps[\"dbsnp\"] = df_model_mappings[\"dbsnp\"].apply(split_string)\n",
    "df_model_snps = df_model_snps.explode(\"dbsnp\")\n",
    "df_model_snps = df_model_snps.dropna().drop_duplicates()\n",
    "rs_ids = df_model_snps[\"dbsnp\"].unique()\n",
    "df_model_snps\n",
    "\n",
    "rename_mapping = {\n",
    "    key: key.lower()\n",
    "    for key in df_snps_all.columns\n",
    "    if key not in {\"error\", \"GLOBAL_SAMPLESIZE\"}\n",
    "}\n",
    "\n",
    "df_model_snps = df_model_snps.merge(\n",
    "    df_snps_all.loc[:, list(rename_mapping)].rename(rename_mapping, axis=1),\n",
    "    left_on=\"dbsnp\",\n",
    "    right_on=\"dbsnp\",\n",
    "    how=\"inner\",\n",
    ")\n",
    "for col in df_model_snps.columns:\n",
    "    df_model_snps[col] = df_model_snps[col].str.split(\",\")\n",
    "    # Switch to default seperator (currently semicolon)\n",
    "    df_model_snps[col] = df_model_snps[col].apply(lambda x: build_string(x))\n",
    "df_model_snps = df_model_snps.sort_values([\"snp_id_sort\"]).reset_index(drop=True)\n",
    "df_model_snps[\"sequence_change\"] = (\n",
    "    df_model_snps[\"spdi\"]\n",
    "    .dropna()\n",
    "    .apply(lambda value: [s.split(\":\", 2)[-1] for s in split_string(value)])\n",
    ")\n",
    "\n",
    "if overwrite:\n",
    "    df_model_snps.to_csv(get_dirpath(\"database\") / f\"snps_{GEM_NAME}.tsv\", sep=\"\\t\")\n",
    "df_model_snps"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
