{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "26efcbdc-2dce-4fb4-8778-dff2287f6817",
   "metadata": {},
   "source": [
    "# Extract annotations and data from EBI Complex Portal\n",
    "\n",
    "The purpose of this notebook is to extract and format ComplexPortal data for subsequent model annotation. \n",
    "\n",
    "Additionally, the purpose of this notebook is to extract relevant data about complex stoichiometry from the database.\n",
    "\n",
    "## Notebook Requirements:\n",
    "*  Model genes **must** have the at least one of following annotations stored in the `object.annotation`. Values are expected to be seperated by semicolons. Accepted keys currently include:\n",
    "    * `\"complexportal\"`\n",
    "    * `\"uniprot\"`\n",
    "*  Note: Requires internet connection to download information from [Complex Portal](https://www.ebi.ac.uk/complexportal/home).\n",
    "\n",
    "### Citations\n",
    "Meldal BHM, Bye-A-Jee H, Gajdoš L, Hammerová Z, Horácková A, Melicher F, Perfetto L, Pokorný D, Lopez MR, Türková A, Wong ED, Xie Z, Casanova EB, Del-Toro N, Koch M, Porras P, Hermjakob H, Orchard S. Complex Portal 2018: extended content and enhanced visualization tools for macromolecular complexes. Nucleic Acids Res. 2019 Jan 8;47(D1):D550-D558. doi: 10.1093/nar/gky1001. PMID: 30357405; PMCID: PMC6323931.\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463131c7-4073-4b0f-8b34-ce350f5f8c5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "from warnings import warn\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    GEM_NAME,\n",
    "    build_string,\n",
    "    check_database_release_online,\n",
    "    compare_tables,\n",
    "    explode_column,\n",
    "    get_annotation_df,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    split_string,\n",
    "    visualize_comparison,\n",
    ")\n",
    "from rbc_gem_utils.database.complexportal import (\n",
    "    COMPLEXPORTAL_DB_TAG,\n",
    "    COMPLEXPORTAL_RELEASE_EXPECTED,\n",
    "    COMPLEXPORTAL_URL,\n",
    "    parse_complex_participants,\n",
    ")\n",
    "from rbc_gem_utils.database.uniprot import UNIPROT_ISOFORM_ID_RE\n",
    "from rbc_gem_utils.util import ensure_iterable\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "022896c7-aa8a-484a-854d-22ce46a45f66",
   "metadata": {},
   "source": [
    "## Set notebook options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aee3473-7047-4c06-a655-16264f75933e",
   "metadata": {},
   "outputs": [],
   "source": [
    "db_tag = COMPLEXPORTAL_DB_TAG\n",
    "expected_release = COMPLEXPORTAL_RELEASE_EXPECTED\n",
    "database_url = COMPLEXPORTAL_URL\n",
    "download_database = False\n",
    "\n",
    "compare_figsize = (5, 5)\n",
    "compare = True\n",
    "display_nunique = True\n",
    "overwrite = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7e1dec-fc60-4f45-9a59-b0667f0f83fc",
   "metadata": {},
   "source": [
    "## Check Complex Portal release\n",
    "* According to [Complex Portal](https://www.ebi.ac.uk/complexportal/download), database is released monthly. \n",
    "* All files are available from the ftp site  and will be updated as new complexes with every data release.\n",
    "* If the current release does not match the expected release, it is because database has been updated since the last time this code was utilized.\n",
    "    * If the notebook works without needing any significant modifications, the only update needed is to the release in the [complex.py](../../src/rbc_gem_utils/database/complexportal.py) source code file to resolve the issue."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63740a51-f91b-477e-9ed7-3c174a77a4c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxomony_int = 9606\n",
    "use_interim = not check_database_release_online(\n",
    "    db_tag, verbose=True, **{\"taxomony_int\": taxomony_int}\n",
    ")\n",
    "# Use different directory paths for unexpected behavior\n",
    "if use_interim:\n",
    "    warn(\n",
    "        \"Online release of database has been updated since the last time notebook was used.\"\n",
    "    )\n",
    "\n",
    "\n",
    "database_dirpath = get_dirpath(\n",
    "    \"database\", db_tag, use_temp=\"interim\" if use_interim else None\n",
    ")\n",
    "annotation_dirpath = get_dirpath(\n",
    "    \"annotation\", use_temp=\"interim\" if use_interim else None\n",
    ")\n",
    "# Ensure directories exist\n",
    "database_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "annotation_dirpath.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af15880b-be08-4c23-9b7d-9971939a5146",
   "metadata": {},
   "source": [
    "## Load ComplexPortal data\n",
    "### Download new files and update database\n",
    "If an argument is not provided (`arg=None`), its default value for the repository used. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4eb650f9-94d6-4646-b4ab-82a51079b256",
   "metadata": {},
   "outputs": [],
   "source": [
    "if download_database:\n",
    "    df_complexportal_data = pd.read_csv(f\"{database_url}/{taxomony_int}.tsv\", sep=\"\\t\")\n",
    "    df_complexportal_data.to_csv(\n",
    "        database_dirpath / f\"{taxomony_int}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "else:\n",
    "    df_complexportal_data = pd.read_csv(\n",
    "        database_dirpath / f\"{taxomony_int}.tsv\",\n",
    "        sep=\"\\t\",\n",
    "    )\n",
    "df_complexportal_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec59ce9b-afc7-4dd0-9a1d-b5a9b77a3bb7",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f10c12-e5d4-4d50-a027-3bc88505aaf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "113349dc-dff8-4396-93ba-82eab2983e1b",
   "metadata": {},
   "source": [
    "### Extract current annotations from model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4295adb0-dd43-4773-bbf9-0248d4f1f3b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"genes\"\n",
    "annotation_cols = [\"uniprot\", \"complexportal\"]\n",
    "mapping_key = \"complexportal\"\n",
    "\n",
    "df_model_mappings = (\n",
    "    get_annotation_df(model.genes, annotation_cols)\n",
    "    .rename({\"id\": annotation_type}, axis=1)\n",
    "    .dropna(subset=[mapping_key])\n",
    ")\n",
    "for col in df_model_mappings.columns:\n",
    "    df_model_mappings = explode_column(df_model_mappings, name=col, sep=\";\")\n",
    "df_model_mappings = df_model_mappings.sort_values(annotation_type)\n",
    "\n",
    "print(df_model_mappings.nunique(dropna=True))\n",
    "df_model_mappings = df_model_mappings.reset_index(drop=True)\n",
    "\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b199576-127b-452f-bd94-c1cb1ac92841",
   "metadata": {},
   "source": [
    "### Map to model genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26581b36-ddb5-4701-8f3d-2a14b17e1346",
   "metadata": {},
   "outputs": [],
   "source": [
    "col = \"Expanded participant list\"\n",
    "df = parse_complex_participants(\n",
    "    df_complexportal_data.set_index(\"#Complex ac\")[col]\n",
    ").reset_index(drop=False)\n",
    "df = df.rename({\"#Complex ac\": \"complexportal\"}, axis=1)\n",
    "\n",
    "df[\"uniprot\"] = df[\"participants\"].apply(lambda x: ensure_iterable(x.split(\"-\")[0]))\n",
    "df = df.explode(\"uniprot\")\n",
    "\n",
    "if mapping_key == \"complexportal\":\n",
    "    df = df[df[\"complexportal\"].isin(df_model_mappings[\"complexportal\"].values)]\n",
    "else:\n",
    "    df[\"participants\"] = df[\"participants\"].apply(lambda x: x.split(\"-\")[0])\n",
    "    df = df[df[\"participants\"].isin(df_model_mappings[mapping_key].values)]\n",
    "df_complex_participants = df.copy()\n",
    "df_complex_participants = df_complex_participants.sort_values(\n",
    "    [\"complexportal\", \"uniprot\"], ascending=[True, True]\n",
    ")\n",
    "df_complex_participants = df_complex_participants.reset_index(drop=True)\n",
    "print(df_complex_participants.nunique())\n",
    "df_complex_participants"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0d49e41-a1db-4ff8-a5b1-8c2f3e712236",
   "metadata": {},
   "source": [
    "## Format ComplexPortal information for all files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf804317-09c9-4423-bcfe-a515e9f1b8df",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_model_complex_annotations = df_complexportal_data[\n",
    "    df_complexportal_data[\"#Complex ac\"].isin(df_complex_participants[\"complexportal\"])\n",
    "].copy()\n",
    "df_model_complex_annotations = df_model_complex_annotations.reset_index(drop=True)\n",
    "rename_mapping = {\n",
    "    \"#Complex ac\": \"complexportal\",\n",
    "    \"Recommended name\": \"complexportal.name\",\n",
    "    \"Aliases for complex\": \"complexportal.alias\",\n",
    "    \"Identifiers (and stoichiometry) of molecules in complex\": \"complexportal.stoichiometry.all\",\n",
    "    \"Experimental evidence\": \"complexportal.evidence\",\n",
    "    \"Cross references\": \"complexportal.cross_references\",\n",
    "    \"Description\": \"complexportal.description\",\n",
    "    \"Complex properties\": \"complexportal.properties\",\n",
    "    \"Complex assembly\": \"complexportal.assembly\",\n",
    "    \"Ligand\": \"complexportal.ligand\",\n",
    "    \"Disease\": \"complexportal.disease\",\n",
    "    \"Agonist\": \"complexportal.agonist\",\n",
    "    \"Antagonist\": \"complexportal.antagonist\",\n",
    "    \"Comment\": \"complexportal.comment\",\n",
    "    \"Source\": \"complexportal.source\",\n",
    "    \"Taxonomy identifier\": \"taxomony\",\n",
    "    \"Evidence Code\": \"eco\",\n",
    "    \"Go Annotations\": \"go\",\n",
    "    \"Expanded participant list\": \"uniprot\",\n",
    "}\n",
    "df_model_complex_annotations = df_model_complex_annotations.replace(\"-\", \"\")\n",
    "df_model_complex_annotations = df_model_complex_annotations.loc[\n",
    "    :, list(rename_mapping)\n",
    "].rename(rename_mapping, axis=1)\n",
    "df_model_complex_annotations\n",
    "for key in df_model_complex_annotations.columns:\n",
    "    df_model_complex_annotations[key] = df_model_complex_annotations[key].apply(\n",
    "        lambda x: build_string(x.split(\"|\")) if isinstance(x, str) else x\n",
    "    )\n",
    "\n",
    "    if key == \"complexportal.cross_references\":\n",
    "        for subkey in {\"reactome\", \"complex portal\", \"pubmed\", \"intenz\"}:\n",
    "            df_model_complex_annotations[subkey] = df_model_complex_annotations[\n",
    "                key\n",
    "            ].apply(\n",
    "                lambda x: build_string(\n",
    "                    [\n",
    "                        s.split(\"(\")[0].replace(\" \", \"\").replace(f\"{subkey}:\", \"\")\n",
    "                        for s in x.split(\";\")\n",
    "                        if s.startswith(subkey)\n",
    "                    ]\n",
    "                )\n",
    "            )\n",
    "        for subkey in {\"rhea\", \"efo\"}:\n",
    "            df_model_complex_annotations[subkey] = df_model_complex_annotations[\n",
    "                key\n",
    "            ].apply(\n",
    "                lambda x: build_string(\n",
    "                    [\n",
    "                        s.split(\"(\")[0]\n",
    "                        .replace(\" \", \"\")\n",
    "                        .lstrip(f\":{subkey}\")\n",
    "                        .lower()\n",
    "                        .replace(f\"{subkey}:\", \"\")\n",
    "                        for s in x.split(\";\")\n",
    "                        if s.startswith(subkey)\n",
    "                    ]\n",
    "                )\n",
    "            )\n",
    "            if subkey == \"efo\":\n",
    "                for subsubkey in [\"orphanet\", \"efo\"]:  # EFO must be last\n",
    "                    df_model_complex_annotations[\n",
    "                        subsubkey\n",
    "                    ] = df_model_complex_annotations[subkey].apply(\n",
    "                        lambda x: build_string(\n",
    "                            [s for s in x.split(\";\") if s.startswith(subsubkey)]\n",
    "                        )\n",
    "                    )\n",
    "\n",
    "    if key == \"complexportal.evidence\":\n",
    "        for subkey in {\"intact\", \"wwpdb\"}:\n",
    "            df_model_complex_annotations[\n",
    "                subkey.replace(\"wwpdb\", \"pdb\")\n",
    "            ] = df_model_complex_annotations[key].apply(\n",
    "                lambda x: build_string(\n",
    "                    [\n",
    "                        s.split(\"(\")[0].replace(f\"{subkey}:\", \"\")\n",
    "                        for s in x.split(\";\")\n",
    "                        if s.startswith(subkey)\n",
    "                    ]\n",
    "                )\n",
    "            )\n",
    "    if key in {\"eco\", \"go\", \"uniprot\"}:\n",
    "        if key == \"uniprot\":\n",
    "            df_model_complex_annotations[\n",
    "                f\"{key}.stoichiometry\"\n",
    "            ] = df_model_complex_annotations[key].apply(\n",
    "                lambda x: \";\".join([s.split(\"(\")[1].rstrip(\")\") for s in x.split(\";\")])\n",
    "            )\n",
    "        df_model_complex_annotations[key] = df_model_complex_annotations[key].apply(\n",
    "            lambda x: build_string([s.split(\"(\")[0] for s in x.split(\";\")])\n",
    "        )\n",
    "    if key in {\"complexportal.disease\"}:\n",
    "        for subkey in [\"mondo\", \"mim\"]:\n",
    "            df_model_complex_annotations[subkey] = df_model_complex_annotations[\n",
    "                key\n",
    "            ].apply(\n",
    "                lambda x: build_string(\n",
    "                    re.findall(rf\"\\[({subkey}.+?(?=]))\", x, re.IGNORECASE)\n",
    "                )\n",
    "            )\n",
    "    if key == \"complexportal.stoichiometry.all\":\n",
    "        for subkey in [\"urs\", \"chebi\"]:\n",
    "            search_re = re.compile(f\"({subkey}.+?(?=;))\", re.IGNORECASE)\n",
    "            if subkey == \"urs\":\n",
    "                subkey = \"rnacentral\"\n",
    "            df_model_complex_annotations[subkey] = df_model_complex_annotations[\n",
    "                key\n",
    "            ].apply(lambda x: build_string(search_re.findall(x)))\n",
    "            df_model_complex_annotations[\n",
    "                f\"{subkey}.stoichiometry\"\n",
    "            ] = df_model_complex_annotations[subkey].apply(\n",
    "                lambda x: build_string(\n",
    "                    [s.split(\"(\")[1].rstrip(\")\") for s in x.split(\";\") if s]\n",
    "                )\n",
    "            )\n",
    "            df_model_complex_annotations[f\"{subkey}\"] = df_model_complex_annotations[\n",
    "                subkey\n",
    "            ].apply(\n",
    "                lambda x: build_string([s.split(\"(\")[0] for s in x.split(\";\") if s])\n",
    "            )\n",
    "df_model_complex_annotations = df_model_complex_annotations.replace(\n",
    "    float(\"nan\"), pd.NA\n",
    ").replace(\"\", pd.NA)\n",
    "df_model_complex_annotations = df_model_complex_annotations.replace(\"\", float(\"nan\"))\n",
    "df_model_complex_annotations = df_model_complex_annotations.sort_values(\n",
    "    [\"complexportal\", \"uniprot\"], ascending=[True, True]\n",
    ")\n",
    "df_model_complex_annotations = df_model_complex_annotations.reset_index(drop=True)\n",
    "df_model_complex_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d45bcff9-b2d0-4b81-818a-b2ba33981ee9",
   "metadata": {},
   "source": [
    "## Format Complex Tables\n",
    "### Complexes and stoichiometry\n",
    "Information extracted here can be useful for protein constrained modeling. All possible complexes are extracted here. Data is further formatted and refined in the curation notebooks [HarmonizeProteinsComplexes](../curation/HarmonizeProteinsComplexes.ipynb#Complex-Portal)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f26c99f-cbc8-40e6-af97-26668edaa429",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_model_complex_data = df_complex_participants.loc[\n",
    "    :, [\"complexportal\", \"participants\", \"stoichiometry\", \"uniprot\"]\n",
    "]\n",
    "df_model_complex_data[\"uniprot.isoform\"] = df_model_complex_data[\"participants\"].apply(\n",
    "    lambda x: (\n",
    "        UNIPROT_ISOFORM_ID_RE.search(x).group()\n",
    "        if UNIPROT_ISOFORM_ID_RE.search(x)\n",
    "        else \"\"\n",
    "    )\n",
    ")\n",
    "\n",
    "df_model_complex_data = df_model_complex_data.merge(\n",
    "    df_model_complex_annotations.loc[\n",
    "        :,\n",
    "        [\n",
    "            \"complexportal\",\n",
    "            \"complexportal.name\",\n",
    "            \"complexportal.alias\",\n",
    "            \"complexportal.stoichiometry.all\",\n",
    "            \"reactome\",\n",
    "        ],\n",
    "    ],\n",
    "    left_on=\"complexportal\",\n",
    "    right_on=\"complexportal\",\n",
    "    how=\"left\",\n",
    "    suffixes=(\"\", \"_drop\"),\n",
    ")\n",
    "df_model_complex_data = df_model_complex_data.drop(\n",
    "    [c for c in df_model_complex_data.columns if c.endswith(\"_drop\")], axis=1\n",
    ")\n",
    "if overwrite:\n",
    "    df_model_complex_data.to_csv(\n",
    "        database_dirpath / f\"{db_tag}_{GEM_NAME}_complexes.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "df_model_complex_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86812ae8-df6e-45c7-8ea7-52337ccbcc1b",
   "metadata": {},
   "source": [
    "### Annotation tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "528129cd-2934-46b0-8e6d-cc6cd8cca676",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_annotations = df_model_complex_annotations.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"complexportal\",\n",
    "        \"eco\",\n",
    "        \"go\",\n",
    "        \"uniprot\",\n",
    "        \"rnacentral\",\n",
    "        \"chebi\",\n",
    "        \"pdb\",\n",
    "        \"intact\",\n",
    "        \"reactome\",\n",
    "        \"pubmed\",\n",
    "        \"efo\",\n",
    "        \"rhea\",\n",
    "        \"orphanet\",\n",
    "        \"mondo\",\n",
    "        \"mim\",\n",
    "    ],\n",
    "].copy()\n",
    "df_annotations[\"uniprot\"] = df_annotations[\"uniprot\"].apply(\n",
    "    lambda x: [s.split(\"-\")[0] for s in split_string(x)]\n",
    ")\n",
    "df_annotations = df_annotations.explode(\"uniprot\")\n",
    "df_annotations = df_model_mappings.merge(\n",
    "    df_annotations,\n",
    "    left_on=mapping_key,\n",
    "    right_on=mapping_key,\n",
    "    how=\"left\",\n",
    "    suffixes=(\"\", \"_drop\"),\n",
    ")\n",
    "to_drop = [c for c in df_annotations.columns if c.endswith(\"_drop\")]\n",
    "for k in to_drop:\n",
    "    df_annotations = df_annotations[\n",
    "        df_annotations[k.strip(\"_drop\")] == df_annotations[k]\n",
    "    ].drop(k, axis=1)\n",
    "    df_annotations = df_annotations.drop_duplicates()\n",
    "\n",
    "df_annotations = df_annotations.sort_values(\n",
    "    [\"genes\", \"complexportal\", \"uniprot\"]\n",
    ").reset_index(drop=True)\n",
    "df_annotations = df_annotations.reset_index(drop=True)\n",
    "\n",
    "df_gene_annotations = df_annotations.copy()\n",
    "df_cplx_annotations = df_annotations.copy()\n",
    "df_cplx_annotations[\"complexes\"] = df_cplx_annotations[\"complexportal\"].str.replace(\n",
    "    \"-\", \"_\"\n",
    ")\n",
    "\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42dedaa8-db6b-406e-b835-a541e4decd75",
   "metadata": {},
   "source": [
    "#### Complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8e90f18-b4b2-4634-a57f-343cef62acd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"complexes\"\n",
    "df_annotations = df_cplx_annotations.groupby(annotation_type, as_index=False).agg(\n",
    "    lambda x: build_string(sorted(x.dropna().unique()))\n",
    ")\n",
    "\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "if compare:\n",
    "    compare_on_index = [annotation_type]\n",
    "    try:\n",
    "        df_previous = pd.read_csv(\n",
    "            annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "        df_previous = df_previous.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "    except FileNotFoundError:\n",
    "        df_previous = pd.DataFrame([], columns=compare_on_index)\n",
    "    df_comparision = compare_tables(\n",
    "        df_previous.set_index(compare_on_index),\n",
    "        df_annotations.set_index(compare_on_index),\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=compare_figsize)\n",
    "    ax.yaxis.set_tick_params(labelsize=8)\n",
    "    ax = visualize_comparison(df_comparision)\n",
    "\n",
    "if display_nunique:\n",
    "    for col in df_annotations.columns:\n",
    "        df = explode_column(df_annotations, name=col, sep=\";\")\n",
    "        df = df[col].drop_duplicates()\n",
    "        print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "if overwrite:\n",
    "    df_annotations.to_csv(\n",
    "        annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f61f0b61-03cf-4510-a04d-d509b44bd9a1",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91377d2e-477e-4fae-b747-49688bc9c2af",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"genes\"\n",
    "df_annotations = df_gene_annotations.groupby(annotation_type, as_index=False).agg(\n",
    "    lambda x: build_string(list(x.dropna().unique()))\n",
    ")\n",
    "\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "if compare:\n",
    "    compare_on_index = [annotation_type]\n",
    "    try:\n",
    "        df_previous = pd.read_csv(\n",
    "            annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "        df_previous = df_previous.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "    except FileNotFoundError:\n",
    "        df_previous = pd.DataFrame([], columns=compare_on_index)\n",
    "    df_comparision = compare_tables(\n",
    "        df_previous.set_index(compare_on_index),\n",
    "        df_annotations.set_index(compare_on_index),\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=compare_figsize)\n",
    "    ax.yaxis.set_tick_params(labelsize=8)\n",
    "    ax = visualize_comparison(df_comparision)\n",
    "\n",
    "if display_nunique:\n",
    "    for col in df_annotations.columns:\n",
    "        df = explode_column(df_annotations, name=col, sep=\";\")\n",
    "        df = df[col].drop_duplicates()\n",
    "        print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "if overwrite:\n",
    "    df_annotations.to_csv(\n",
    "        annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_annotations"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
