{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ecb8f14e-15a3-4fb2-8b3f-5939b7e6d018",
   "metadata": {},
   "source": [
    "# Harmonize annotations for the RBC-GEM\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b3144c1-38d3-4040-97d1-a2d2fe2208e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    GEM_NAME,\n",
    "    MODEL_PATH,\n",
    "    ROOT_PATH,\n",
    "    compare_tables,\n",
    "    explode_column,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    visualize_comparison,\n",
    ")\n",
    "from rbc_gem_utils.database import (\n",
    "    COMPLEXPORTAL_DB_TAG,\n",
    "    DRUGBANK_DB_TAG,\n",
    "    DRUGCENTRAL_DB_TAG,\n",
    "    HUMANGEM_DB_TAG,\n",
    "    MGI_DB_TAG,\n",
    "    MIM_DB_TAG,\n",
    "    TCDB_DB_TAG,\n",
    "    UNIPROT_DB_TAG,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3efb3df5-48d7-453f-be50-71f78c347dbf",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de1a4842-65ae-451c-b775-29b109d127aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "\n",
    "model.reactions.sort()\n",
    "model.genes.sort()\n",
    "model.metabolites.sort()\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f96f8818-0a0b-4518-8f2a-91ad84df9a7e",
   "metadata": {},
   "source": [
    "## Create annotation tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45db517a-4e4f-4c9e-bd9b-11e6b1da9951",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_dirpath = get_dirpath(\"annotation\")\n",
    "\n",
    "compare = True\n",
    "display_nunique = True\n",
    "overwrite = True\n",
    "compare_figsize = None\n",
    "\n",
    "REFERENCES_TAG = \"References\"\n",
    "\n",
    "preserve_model_previous = False\n",
    "model_annotation_dfs = {\n",
    "    \"reactions\": None,\n",
    "    \"metabolites\": None,\n",
    "    \"genes\": None,\n",
    "}\n",
    "to_double_check = {\n",
    "    \"reactions\": {},\n",
    "    \"metabolites\": {},\n",
    "    \"genes\": {},\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cca1276e-cccf-4421-ad97-675c7eb7987a",
   "metadata": {},
   "source": [
    "### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8109f221-d551-4519-a85f-467e54fd8eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"reactions\"\n",
    "databases = [HUMANGEM_DB_TAG, REFERENCES_TAG]\n",
    "df_annotations_all = pd.read_csv(\n",
    "    model_dirpath / f\"{annotation_type}.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "df_annotations_all = df_annotations_all.set_index(annotation_type)\n",
    "df_annotations_all = df_annotations_all.rename(\n",
    "    {c: f\"{c}_MODEL\" for c in df_annotations_all.columns}, axis=1\n",
    ")\n",
    "\n",
    "for db_tag in databases:\n",
    "    df = pd.read_csv(\n",
    "        annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\",\n",
    "        sep=\"\\t\",\n",
    "    ).set_index(annotation_type)\n",
    "    df.columns = [f\"{c}_{db_tag}\" for c in df.columns]\n",
    "    df_annotations_all = df_annotations_all.merge(\n",
    "        df,\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "        how=\"outer\",\n",
    "    )\n",
    "\n",
    "# Keep database tags on duplicate columns for reconcilliation, remove for unique columns\n",
    "unique_columns = {name.rsplit(\"_\", 1)[0] for name in df_annotations_all.columns}\n",
    "to_double_check[annotation_type] = {\n",
    "    unique: [\n",
    "        name for name in df_annotations_all.columns if name.rsplit(\"_\", 1)[0] == unique\n",
    "    ]\n",
    "    for unique in unique_columns\n",
    "}\n",
    "unique_mapping = {\n",
    "    to_double_check[annotation_type].pop(unique).pop(): unique\n",
    "    for unique, dbs in to_double_check[annotation_type].copy().items()\n",
    "    if len(dbs) == 1\n",
    "}\n",
    "print(f\"Annotation columns to reconcile: {to_double_check[annotation_type]}\")\n",
    "df_annotations_all = df_annotations_all.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "df_annotations_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3104d723-0aa1-4c72-8524-cc9aca27472a",
   "metadata": {},
   "source": [
    "#### Reconcile cross-referenced from multiple sources\n",
    "These columns had multiple mappings and need to be reconciled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b4173e8-49ea-4558-af6c-8a057b79f1b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "reconcilliation = {\n",
    "    # Overwrite with MetAtlas\n",
    "    \"metanetx.reaction\": HUMANGEM_DB_TAG,\n",
    "    \"kegg.reaction\": HUMANGEM_DB_TAG,\n",
    "    \"vmhreaction\": HUMANGEM_DB_TAG,\n",
    "    \"metatlas\": HUMANGEM_DB_TAG,\n",
    "    \"reactome\": HUMANGEM_DB_TAG,\n",
    "    \"tcdb\": HUMANGEM_DB_TAG,\n",
    "    \"ec-code\": HUMANGEM_DB_TAG,\n",
    "    \"rhea\": HUMANGEM_DB_TAG,\n",
    "    # Preserve from model, some BiGG mappings in MetAtlas are off.\n",
    "    \"bigg.reaction\": \"MODEL\",\n",
    "    # Non-miriam fields, spontaneous comes from References\n",
    "    \"spontaneous\": \"MODEL\",\n",
    "    \"references\": REFERENCES_TAG,\n",
    "}\n",
    "\n",
    "df_annotations = df_annotations_all.copy()\n",
    "for key, chosen_col in reconcilliation.items():\n",
    "    if key not in to_double_check[annotation_type]:\n",
    "        continue\n",
    "    df_base_column = df_annotations_all.loc[:, f\"{key}_{chosen_col}\"].copy()\n",
    "    df_base_column.name = key\n",
    "    df_columns = df_annotations_all.loc[:, to_double_check[annotation_type][key]]\n",
    "    for idx, row in df_columns.iterrows():\n",
    "        if preserve_model_previous:\n",
    "            # Preserve old ID to prevent loss of ID\n",
    "            try:\n",
    "                if not row.isna()[f\"{key}_MODEL\"] and row.isna()[f\"{key}_{chosen_col}\"]:\n",
    "                    df_base_column[idx] = row[f\"{key}_MODEL\"]\n",
    "            except KeyError as e:\n",
    "                pass\n",
    "\n",
    "    df_annotations = df_annotations.drop(to_double_check[annotation_type][key], axis=1)\n",
    "    to_double_check[annotation_type].pop(key)\n",
    "    df_annotations[key] = df_base_column\n",
    "\n",
    "\n",
    "# Overwrite\n",
    "df_annotations = df_annotations.rename(unique_mapping, axis=1)\n",
    "df_annotations = df_annotations.reset_index(drop=False).fillna(\"\")\n",
    "model_annotation_dfs[annotation_type] = df_annotations\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bd12578-22a2-4d13-bec1-a603f3a7b52f",
   "metadata": {},
   "source": [
    "#### Save annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66c3be4c-c0e4-47fe-8752-3ba9db3bd6d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (\n",
    "    len(to_double_check[annotation_type]) == 0\n",
    "), f\"There are still columns with data that need reconcilliation: {to_double_check[annotation_type]}\"\n",
    "df_annotations = model_annotation_dfs[annotation_type]\n",
    "\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "if compare:\n",
    "    compare_on_index = [annotation_type]\n",
    "    try:\n",
    "        df_previous = pd.read_csv(\n",
    "            model_dirpath / f\"{annotation_type}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "        df_previous = df_previous.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "    except FileNotFoundError:\n",
    "        df_previous = pd.DataFrame([], columns=compare_on_index)\n",
    "    df_comparision = compare_tables(\n",
    "        df_previous.set_index(compare_on_index),\n",
    "        df_annotations.set_index(compare_on_index),\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=compare_figsize)\n",
    "    ax.yaxis.set_tick_params(labelsize=8)\n",
    "    ax = visualize_comparison(df_comparision)\n",
    "\n",
    "if display_nunique:\n",
    "    for col in df_annotations.columns:\n",
    "        df = explode_column(df_annotations, name=col, sep=\";\")\n",
    "        df = df[col].drop_duplicates()\n",
    "        print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "if overwrite:\n",
    "    df_annotations.to_csv(\n",
    "        model_dirpath / f\"{annotation_type}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f033de86-ec65-4a90-bbaf-7f44bbb3ca35",
   "metadata": {},
   "source": [
    "### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95012e1-b6e6-4f12-a99d-0292dc70f805",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"metabolites\"\n",
    "databases = [HUMANGEM_DB_TAG]\n",
    "df_annotations_all = pd.read_csv(\n",
    "    model_dirpath / f\"{annotation_type}.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "df_annotations_all = df_annotations_all.set_index(annotation_type)\n",
    "df_annotations_all = df_annotations_all.rename(\n",
    "    {c: f\"{c}_MODEL\" for c in df_annotations_all.columns}, axis=1\n",
    ")\n",
    "\n",
    "for db_tag in databases:\n",
    "    df = pd.read_csv(\n",
    "        annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\",\n",
    "        sep=\"\\t\",\n",
    "    ).set_index(annotation_type)\n",
    "    df.columns = [f\"{c}_{db_tag}\" for c in df.columns]\n",
    "    df_annotations_all = df_annotations_all.merge(\n",
    "        df,\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "        how=\"outer\",\n",
    "    )\n",
    "\n",
    "# Keep database tags on duplicate columns for reconcilliation, remove for unique columns\n",
    "unique_columns = {name.rsplit(\"_\", 1)[0] for name in df_annotations_all.columns}\n",
    "to_double_check[annotation_type] = {\n",
    "    unique: [\n",
    "        name for name in df_annotations_all.columns if name.rsplit(\"_\", 1)[0] == unique\n",
    "    ]\n",
    "    for unique in unique_columns\n",
    "}\n",
    "unique_mapping = {\n",
    "    to_double_check[annotation_type].pop(unique).pop(): unique\n",
    "    for unique, dbs in to_double_check[annotation_type].copy().items()\n",
    "    if len(dbs) == 1\n",
    "}\n",
    "print(f\"Annotation columns to reconcile: {to_double_check[annotation_type]}\")\n",
    "df_annotations_all = df_annotations_all.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "df_annotations_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31d3dfa8-6240-4826-bb78-7e2effef42c5",
   "metadata": {},
   "source": [
    "#### Reconcile cross-referenced from multiple sources\n",
    "These columns had multiple mappings and need to be reconciled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07745a9f-e7f8-4a72-adef-be9b60936a65",
   "metadata": {},
   "outputs": [],
   "source": [
    "reconcilliation = {\n",
    "    # Overwrite with MetAtlas\n",
    "    \"kegg.compound\": HUMANGEM_DB_TAG,\n",
    "    \"lipidmaps\": HUMANGEM_DB_TAG,\n",
    "    \"hmdb\": HUMANGEM_DB_TAG,\n",
    "    \"vmhmetabolite\": HUMANGEM_DB_TAG,\n",
    "    \"metatlas\": HUMANGEM_DB_TAG,\n",
    "    \"chebi\": HUMANGEM_DB_TAG,\n",
    "    \"pubchem.compound\": HUMANGEM_DB_TAG,\n",
    "    \"metanetx.chemical\": HUMANGEM_DB_TAG,\n",
    "    \"inchi\": HUMANGEM_DB_TAG,\n",
    "    # Preserve from model, some BiGG mappings in MetAtlas are off.\n",
    "    \"bigg.metabolite\": \"MODEL\",\n",
    "    # # Non-miriam fields\n",
    "    \"references\": REFERENCES_TAG,\n",
    "}\n",
    "\n",
    "df_annotations = df_annotations_all.copy()\n",
    "for key, chosen_col in reconcilliation.items():\n",
    "    if key not in to_double_check[annotation_type]:\n",
    "        continue\n",
    "    df_base_column = df_annotations_all.loc[:, f\"{key}_{chosen_col}\"].copy()\n",
    "    df_base_column.name = key\n",
    "    df_columns = df_annotations_all.loc[:, to_double_check[annotation_type][key]]\n",
    "    for idx, row in df_columns.iterrows():\n",
    "        if preserve_model_previous:\n",
    "            # Preserve old ID to prevent loss of ID\n",
    "            try:\n",
    "                if not row.isna()[f\"{key}_MODEL\"] and row.isna()[f\"{key}_{chosen_col}\"]:\n",
    "                    df_base_column[idx] = row[f\"{key}_MODEL\"]\n",
    "            except KeyError as e:\n",
    "                pass\n",
    "    df_annotations[key] = df_base_column\n",
    "    df_annotations = df_annotations.drop(to_double_check[annotation_type][key], axis=1)\n",
    "    to_double_check[annotation_type].pop(key)\n",
    "\n",
    "\n",
    "# Overwrite\n",
    "df_annotations = df_annotations.rename(unique_mapping, axis=1)\n",
    "df_annotations = df_annotations.reset_index(drop=False).fillna(\"\")\n",
    "model_annotation_dfs[annotation_type] = df_annotations\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd82608e-7f2e-4bfc-aa55-c401584a8354",
   "metadata": {},
   "source": [
    "#### Save annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ed65a02-5c80-4f18-9449-680987cd26a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (\n",
    "    len(to_double_check[annotation_type]) == 0\n",
    "), f\"There are still columns with data that need reconcilliation: {to_double_check[annotation_type]}\"\n",
    "df_annotations = model_annotation_dfs[annotation_type]\n",
    "\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "if compare:\n",
    "    compare_on_index = [annotation_type]\n",
    "    try:\n",
    "        df_previous = pd.read_csv(\n",
    "            ROOT_PATH / MODEL_PATH / f\"{annotation_type}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "        df_previous = df_previous.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "    except FileNotFoundError:\n",
    "        df_previous = pd.DataFrame([], columns=compare_on_index)\n",
    "    df_comparision = compare_tables(\n",
    "        df_previous.set_index(compare_on_index),\n",
    "        df_annotations.set_index(compare_on_index),\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=compare_figsize)\n",
    "    ax.yaxis.set_tick_params(labelsize=8)\n",
    "    ax = visualize_comparison(df_comparision)\n",
    "\n",
    "if display_nunique:\n",
    "    for col in df_annotations.columns:\n",
    "        df = explode_column(df_annotations, name=col, sep=\";\")\n",
    "        df = df[col].drop_duplicates()\n",
    "        print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "if overwrite:\n",
    "    df_annotations.to_csv(\n",
    "        model_dirpath / f\"{annotation_type}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06afa071-e133-4c32-a4dd-733fe229b371",
   "metadata": {},
   "source": [
    "### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a422324f-56b9-4680-9b3b-d599d5cc3325",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"genes\"\n",
    "databases = [\n",
    "    HUMANGEM_DB_TAG,\n",
    "    UNIPROT_DB_TAG,\n",
    "    MGI_DB_TAG,\n",
    "    MIM_DB_TAG,\n",
    "    TCDB_DB_TAG,\n",
    "    DRUGBANK_DB_TAG,\n",
    "    DRUGCENTRAL_DB_TAG,\n",
    "    # CDCDB_DB_TAG,\n",
    "    COMPLEXPORTAL_DB_TAG,\n",
    "    REFERENCES_TAG,\n",
    "]\n",
    "df_annotations_all = pd.read_csv(\n",
    "    model_dirpath / f\"{annotation_type}.tsv\", sep=\"\\t\", index_col=None\n",
    ")\n",
    "\n",
    "df_annotations_all = df_annotations_all.set_index(annotation_type)\n",
    "df_annotations_all = df_annotations_all.rename(\n",
    "    {c: f\"{c}_MODEL\" for c in df_annotations_all.columns}, axis=1\n",
    ")\n",
    "\n",
    "for db_tag in databases:\n",
    "    df = pd.read_csv(\n",
    "        annotation_dirpath / f\"{annotation_type}_{db_tag}.tsv\",\n",
    "        sep=\"\\t\",\n",
    "    ).set_index(annotation_type)\n",
    "    df.columns = [f\"{c}_{db_tag}\" for c in df.columns]\n",
    "    df_annotations_all = df_annotations_all.merge(\n",
    "        df,\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "        how=\"outer\",\n",
    "    )\n",
    "\n",
    "# Keep database tags on duplicate columns for reconcilliation, remove for unique columns\n",
    "unique_columns = {name.rsplit(\"_\", 1)[0] for name in df_annotations_all.columns}\n",
    "to_double_check[annotation_type] = {\n",
    "    unique: [\n",
    "        name for name in df_annotations_all.columns if name.rsplit(\"_\", 1)[0] == unique\n",
    "    ]\n",
    "    for unique in unique_columns\n",
    "}\n",
    "unique_mapping = {\n",
    "    to_double_check[annotation_type].pop(unique).pop(): unique\n",
    "    for unique, dbs in to_double_check[annotation_type].copy().items()\n",
    "    if len(dbs) == 1\n",
    "}\n",
    "print(f\"Annotation columns to reconcile: {to_double_check[annotation_type]}\")\n",
    "df_annotations_all = df_annotations_all.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "df_annotations_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b7669cf-2779-4ca9-930e-df5cf06ad269",
   "metadata": {},
   "source": [
    "#### Reconcile cross-referenced from multiple sources\n",
    "These columns had multiple mappings and need to be reconciled.\n",
    "\n",
    "Currently, overwrites with one database and will preserve previous model annotations if nothing exists for a field."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cdea21f-5e83-44be-aec9-3cdc1e217182",
   "metadata": {},
   "outputs": [],
   "source": [
    "reconcilliation = {\n",
    "    # Overwrite with UniProt\n",
    "    \"uniprot\": UNIPROT_DB_TAG,\n",
    "    \"uniprot.isoform\": UNIPROT_DB_TAG,\n",
    "    \"uniprot.chain\": UNIPROT_DB_TAG,\n",
    "    \"refseq\": UNIPROT_DB_TAG,\n",
    "    \"ncbigene\": UNIPROT_DB_TAG,\n",
    "    \"hgnc.symbol\": UNIPROT_DB_TAG,\n",
    "    \"ccds\": UNIPROT_DB_TAG,\n",
    "    \"dbsnp\": UNIPROT_DB_TAG,\n",
    "    \"ec-code\": UNIPROT_DB_TAG,\n",
    "    \"pdb\": UNIPROT_DB_TAG,\n",
    "    \"pfam\": UNIPROT_DB_TAG,\n",
    "    \"go\": UNIPROT_DB_TAG,\n",
    "    \"pharmgkb.gene\": UNIPROT_DB_TAG,\n",
    "    \"taxonomy\": UNIPROT_DB_TAG,\n",
    "    \"smr\": UNIPROT_DB_TAG,\n",
    "    \"iuphar.receptor\": UNIPROT_DB_TAG,\n",
    "    \"hgnc\": UNIPROT_DB_TAG,\n",
    "    \"intact\": UNIPROT_DB_TAG,  # COMPLEXPORTAL_DB_TAG\n",
    "    \"tcdb\": UNIPROT_DB_TAG,  # TCDB_DB_TAG,\n",
    "    # Overwrite with DrugCentral\n",
    "    \"drugcentral\": DRUGCENTRAL_DB_TAG,\n",
    "    # Overwrite with DrugBank\n",
    "    \"drugbank\": DRUGBANK_DB_TAG,\n",
    "    \"smpdb\": DRUGBANK_DB_TAG,\n",
    "    # Overwrite with MetAtlas\n",
    "    \"ensembl\": HUMANGEM_DB_TAG,\n",
    "    # Overwrite with MGI\n",
    "    \"mgi\": MGI_DB_TAG,\n",
    "    # Overwrite with MIM\n",
    "    \"mim\": MIM_DB_TAG,\n",
    "    # Overwrite with Complex Portal\n",
    "    \"complexportal\": COMPLEXPORTAL_DB_TAG,\n",
    "    # Keep model\n",
    "    \"pubchem.compound\": \"MODEL\",\n",
    "    \"pubmed\": \"MODEL\",\n",
    "    \"efo\": \"MODEL\",\n",
    "    \"eco\": \"MODEL\",\n",
    "    \"chebi\": \"MODEL\",\n",
    "    \"rnacentral\": \"MODEL\",\n",
    "    \"mondo\": \"MODEL\",\n",
    "    # Non-miriam fields\n",
    "    \"chromosome\": UNIPROT_DB_TAG,\n",
    "    \"references\": \"References\",\n",
    "    # To double check eventually for formatting within model\n",
    "    \"string\": UNIPROT_DB_TAG,\n",
    "    \"hogenom\": UNIPROT_DB_TAG,\n",
    "    \"genewiki\": UNIPROT_DB_TAG,\n",
    "    \"prints\": UNIPROT_DB_TAG,\n",
    "    \"supfam\": UNIPROT_DB_TAG,\n",
    "    \"bindingdb\": UNIPROT_DB_TAG,\n",
    "    \"gxa\": UNIPROT_DB_TAG,\n",
    "    \"sasbdb\": UNIPROT_DB_TAG,\n",
    "    \"merops\": UNIPROT_DB_TAG,\n",
    "    \"allergome\": UNIPROT_DB_TAG,\n",
    "    \"ideal\": UNIPROT_DB_TAG,\n",
    "    \"cazy\": UNIPROT_DB_TAG,\n",
    "    \"bgee.gene\": UNIPROT_DB_TAG,\n",
    "    \"panther.family\": UNIPROT_DB_TAG,\n",
    "    \"ena.embl\": UNIPROT_DB_TAG,\n",
    "    \"nextprot\": UNIPROT_DB_TAG,\n",
    "    \"reactome\": UNIPROT_DB_TAG,\n",
    "    \"pirsf\": UNIPROT_DB_TAG,\n",
    "    \"smart\": UNIPROT_DB_TAG,\n",
    "    \"dip\": UNIPROT_DB_TAG,\n",
    "    \"genetree\": UNIPROT_DB_TAG,\n",
    "    \"rhea\": UNIPROT_DB_TAG,\n",
    "    \"pathwaycommons\": UNIPROT_DB_TAG,\n",
    "    \"depod\": UNIPROT_DB_TAG,\n",
    "    \"kegg.genes\": UNIPROT_DB_TAG,\n",
    "    \"disprot\": UNIPROT_DB_TAG,\n",
    "    \"massive\": UNIPROT_DB_TAG,\n",
    "    \"peptideatlas\": UNIPROT_DB_TAG,\n",
    "    \"elm\": UNIPROT_DB_TAG,\n",
    "    \"maxqb\": \"UniProt\",\n",
    "    \"brenda\": \"UniProt\",\n",
    "    \"hamap\": \"UniProt\",\n",
    "    \"slm\": \"UniProt\",\n",
    "    \"phosphosite.protein\": \"UniProt\",\n",
    "    \"cdd\": \"UniProt\",\n",
    "    \"pride\": \"UniProt\",\n",
    "    \"ctd\": \"UniProt\",\n",
    "    \"phylomedb\": \"UniProt\",\n",
    "    \"paxdb.protein\": \"UniProt\",\n",
    "    \"proteomicsdb.protein\": \"UniProt\",\n",
    "    \"epd\": \"UniProt\",\n",
    "    \"hpa\": \"UniProt\",\n",
    "    \"interpro\": \"UniProt\",\n",
    "    \"mint\": \"UniProt\",\n",
    "    \"signor\": \"UniProt\",\n",
    "    \"orthodb\": \"UniProt\",\n",
    "    \"biocyc\": \"UniProt\",\n",
    "    \"chembl.target\": \"UniProt\",\n",
    "    \"orphanet\": \"UniProt\",\n",
    "    \"genecards\": \"UniProt\",\n",
    "    \"biogrid\": \"UniProt\",\n",
    "    \"uniparc\": \"UniProt\",\n",
    "    \"prosite\": \"UniProt\",\n",
    "    \"corum\": \"UniProt\",\n",
    "    \"bmrb\": \"UniProt\",\n",
    "    \"peroxibase\": \"UniProt\",\n",
    "    \"treefam\": \"UniProt\",\n",
    "    \"oma.grp\": \"UniProt\",\n",
    "    \"eggnog\": \"UniProt\",\n",
    "}\n",
    "\n",
    "df_annotations = df_annotations_all.copy()\n",
    "for key, chosen_col in reconcilliation.items():\n",
    "    if key not in to_double_check[annotation_type]:\n",
    "        continue\n",
    "    df_base_column = df_annotations_all.loc[:, f\"{key}_{chosen_col}\"].copy()\n",
    "    df_base_column.name = key\n",
    "    df_columns = df_annotations_all.loc[:, to_double_check[annotation_type][key]]\n",
    "    for idx, row in df_columns.iterrows():\n",
    "        if preserve_model_previous:\n",
    "            # Preserve old ID to prevent loss of ID\n",
    "            try:\n",
    "                if not row.isna()[f\"{key}_MODEL\"] and row.isna()[f\"{key}_{chosen_col}\"]:\n",
    "                    df_base_column[idx] = row[f\"{key}_MODEL\"]\n",
    "            except KeyError as e:\n",
    "                pass\n",
    "\n",
    "    df_annotations = df_annotations.drop(to_double_check[annotation_type][key], axis=1)\n",
    "    to_double_check[annotation_type].pop(key)\n",
    "    df_annotations[key] = df_base_column\n",
    "\n",
    "\n",
    "# Overwrite\n",
    "df_annotations = df_annotations.rename(unique_mapping, axis=1)\n",
    "df_annotations = df_annotations.reset_index(drop=False).fillna(\"\")\n",
    "model_annotation_dfs[annotation_type] = df_annotations\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "df_annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f16c5f1f-d802-4483-a17f-aff5c9f60f08",
   "metadata": {},
   "source": [
    "#### Save annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a1e11a3-d5b7-4d80-b239-3ba97a2e307a",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (\n",
    "    len(to_double_check[annotation_type]) == 0\n",
    "), f\"There are still columns with data that need reconcilliation: {to_double_check[annotation_type]}\"\n",
    "df_annotations = model_annotation_dfs[annotation_type]\n",
    "\n",
    "df_annotations = df_annotations.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "if compare:\n",
    "    compare_on_index = [annotation_type]\n",
    "    try:\n",
    "        df_previous = pd.read_csv(\n",
    "            ROOT_PATH / MODEL_PATH / f\"{annotation_type}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "        df_previous = df_previous.replace(float(\"nan\"), pd.NA).replace(\"\", pd.NA)\n",
    "    except FileNotFoundError:\n",
    "        df_previous = pd.DataFrame([], columns=compare_on_index)\n",
    "    df_comparision = compare_tables(\n",
    "        df_previous.set_index(compare_on_index),\n",
    "        df_annotations.set_index(compare_on_index),\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(5, 20))\n",
    "    ax.yaxis.set_tick_params(labelsize=8)\n",
    "    ax = visualize_comparison(df_comparision)\n",
    "\n",
    "if display_nunique:\n",
    "    for col in df_annotations.columns:\n",
    "        df = explode_column(df_annotations, name=col, sep=\";\")\n",
    "        df = df[col].drop_duplicates()\n",
    "        print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "if overwrite:\n",
    "    df_annotations.to_csv(\n",
    "        model_dirpath / f\"{annotation_type}.tsv\", sep=\"\\t\", index=False\n",
    "    )\n",
    "\n",
    "df_annotations"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
