{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2537d45b-6871-43b9-81f8-e3639df15ac5",
   "metadata": {},
   "source": [
    "# Simulate models using pcFVA - REDS Recall, ATP11C V972M\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2255cec1-1c92-48cd-b35a-01065975dc9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Set parameter LicenseID to value 2664191\n",
      "Academic license - for non-commercial use only - expires 2026-05-12\n",
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.3\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.13.4\n",
      "bio                                   1.8.0\n",
      "cobra                                0.29.1\n",
      "depinfo                               2.2.0\n",
      "gurobipy                             12.0.3\n",
      "matplotlib                           3.10.3\n",
      "matplotlib-venn                       1.1.2\n",
      "memote                               0.17.0\n",
      "networkx                                3.5\n",
      "notebook                              7.4.4\n",
      "openpyxl                              3.1.5\n",
      "pandas                                2.3.1\n",
      "pre-commit                            4.2.0\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.32.4\n",
      "scikit-learn                          1.7.0\n",
      "scipy                                1.16.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip          25.1\n",
      "setuptools 78.1.1\n",
      "wheel      0.45.1\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Windows 10-AMD64\n",
      "CPython  3.11.13\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import io\n",
    "import shutil\n",
    "import zipfile\n",
    "import tempfile\n",
    "import numpy as np\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "import zipfile\n",
    "import tempfile\n",
    "from pathlib import Path\n",
    "import gurobipy as gp\n",
    "import pandas as pd\n",
    "from cobra.core import get_solution\n",
    "from cobra.exceptions import OptimizationError\n",
    "from cobra.flux_analysis.variability import flux_variability_analysis, find_blocked_reactions\n",
    "from cobra.manipulation import remove_genes\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    write_cobra_model,\n",
    "    show_versions,\n",
    "    handle_msg\n",
    ")\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    DEFAULT_PREFIX_SUFFIX_VALUES,\n",
    "    DEFAULT_PROTEOME_COMPARTMENT,\n",
    "    BudgetDilution,\n",
    "    ProteinDilution,\n",
    "    add_relaxation_budget,\n",
    "    load_overlay_model,\n",
    ")\n",
    "\n",
    "gp.setParam(\"OutputFlag\", 0)\n",
    "gp.setParam(\"LogToConsole\", 0)\n",
    "\n",
    "# Show versions of notebook\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d1f0e62-ca1f-4201-a29d-dc3221be59c1",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c9d9475-8441-4f92-b46e-6257c5a7d843",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>100000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-09\n",
       "lower_bound: -1e-08\n",
       "upper_bound: 100000000.0\n",
       "processes: 30\n",
       "cache_directory: C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "# Set bound defaults much larger to prevent model loading issues\n",
    "COBRA_CONFIGURATION.bounds = (-1e-8, 1e8)\n",
    "COBRA_CONFIGURATION.tolerance = 1e-9\n",
    "COBRA_CONFIGURATION.processes = 30 # Cannot exceed number of cores - 1\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abae3e11-3327-4a6c-96db-59d70b0d7313",
   "metadata": {},
   "source": [
    "### Define organism, model, and dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c6b691ad-10c0-4081-aa10-8bceceec4d00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATP11C_V972M_Sample'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism = \"Human\"\n",
    "model_id = \"RBC_GEM\"\n",
    "dataset_name = \"REDSRecall\"\n",
    "genotype = \"ATP11C_V972M\"\n",
    "grouped_data_key = f'{genotype}_Sample'\n",
    "grouped_data_key"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17a0dfee-d2ec-4da0-98e4-ee9277432008",
   "metadata": {},
   "source": [
    "### Set variables for sample identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4c122056-7d54-4f63-873b-41dff10f0ad4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_key = \"SAMPLE ID\"\n",
    "donor_key = \"PUBLIC RECALL DONOR ID\"\n",
    "\n",
    "# For sample IDs\n",
    "donor_re = re.compile(rf\"(?P<donor>S(?P<num>\\d\\d\\d))\")\n",
    "operations = \"|\".join([x.capitalize() for x in [\"mean\", \"median\"]])\n",
    "operation_re = re.compile(r\"(?P<op>\" + operations + r\")\\_(?P<group>\\w+)\")\n",
    "sample_id_re = re.compile(\n",
    "    r\"(?!\" + operations + r\")\" + donor_re.pattern + r\"\\_\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f627bc-4414-4fb6-a7eb-106c7c10fd84",
   "metadata": {},
   "source": [
    "### Set computation options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2ec77319-ed41-46d3-b22c-0bb35d203e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"xml\"  # In our experience, SBML/XML loads faster, but will take up to 4x more space uncompressed as compared to JSON\n",
    "run_computations = True  # Keep off to use previously computed results\n",
    "overwrite = True  # Whether to allow overwriting of previous simulation results\n",
    "verbose = True\n",
    "\n",
    "# Objective reactions\n",
    "objective_reactions = [\"NaKt\"] \n",
    "# Reactions that must have the capability to carry flux, sort for consistency\n",
    "required_flux_reactions = ['PSFLIPt'] # Add reactions to this list\n",
    "required_flux_reactions = sorted(set(objective_reactions + required_flux_reactions))\n",
    "\n",
    "only_flux_abundance_reactions = False # Only simulate reactions necessary for flux-abundance correlations\n",
    "min_relax_budget_for_objectives = True\n",
    "# Remove blocked reactions before pcFVA simulation.\n",
    "# For large models and/or multiple runs at different optimums, will speed up computation and potentially improve results.\n",
    "remove_blocked_reactions = True\n",
    "# Relaxation reactions that should be restricted to inactive\n",
    "protein_relaxations_to_restrict = [\"ATP11A\", \"ATP11B\", \"ATP11C\", \"TMEM30A\"]\n",
    "# Protein constraints that need lower bounds relaxed to prevent increase of \n",
    "# associated subunits in direct conflict with physiologically known. \n",
    "# Setting the lower bound prevents from being required.\n",
    "protein_constraints_lb_to_relax = [\"TMEM30A\"]\n",
    "# Expected as percentages (e.g., 0, 50, 90, 99, 100)\n",
    "optimum_percents = [0]\n",
    "# RUn parsimonius FBA\n",
    "pfba_factor = None\n",
    "# Whether to run loopless pcFVA\n",
    "loopless = False\n",
    "\n",
    "zip_kwargs = dict(\n",
    "    compression=zipfile.ZIP_DEFLATED,\n",
    "    compresslevel=None\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b42e25bb-d247-43bc-aae9-94b980787bfb",
   "metadata": {},
   "source": [
    "#### Set prefixes/suffixes to expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ee90ae93-49fc-42ad-8de9-60ff06c8e179",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.dilution\"]\n",
    "protein_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.metabolite\"]\n",
    "relaxation_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.relaxation\"]\n",
    "enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]\n",
    "enzyme_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.dilution\"]\n",
    "enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.metabolite\"]\n",
    "budget_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"budgets\"][\"prefix.dilution\"]\n",
    "budget_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"budgets\"][\"prefix.metabolite\"]\n",
    "comp_suffix = f\"_{DEFAULT_PROTEOME_COMPARTMENT}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6e1dacf-75e8-4548-90d6-bb51c9cf1ffa",
   "metadata": {},
   "source": [
    "### Set figure options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3b6a15e9-9848-4c2e-b78d-8538f0d0dcd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fd11ff1-4805-4860-b346-1f5c993b54ac",
   "metadata": {},
   "source": [
    "### Set paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ca3a0645-ab37-434d-af5c-d6f3d385b5f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set paths\n",
    "overlay_dirpath = get_dirpath(\"analysis\") / \"OVERLAY\" / organism\n",
    "model_dirpath = overlay_dirpath / model_id\n",
    "results_dirpath = (get_dirpath(use_temp=\"processed\") / model_id / \"OVERLAY\" / organism / dataset_name / grouped_data_key)\n",
    "pcfva_results_dirpath = (results_dirpath / \"pcFVA\" / \"_\".join((\"REQ\", *required_flux_reactions)) /  \"_\".join((\"OBJ\", *objective_reactions)))\n",
    "pcfva_results_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "# ZIP directories\n",
    "sample_pcmodels_dirpath = results_dirpath / \"pcmodels\"\n",
    "reduced_models_dirpath = pcfva_results_dirpath.parent / \"reduced_pcmodels\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50dc6fa4-161e-4511-bf31-0b75d3b93dcb",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6f8349c9-7146-48d4-864a-fd56990c670b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>188043e6110</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>7815</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>15687</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>723</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>68</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC at 0x188043e6110>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_cobra_model(filename=model_dirpath / f\"{model_id}.xml\")\n",
    "pcmodel = load_overlay_model(filename=model_dirpath / f\"{model_id}_PC.xml\")\n",
    "\n",
    "# Add relaxation budget to initial PC model to get names of relaxation reactions\n",
    "add_relaxation_budget(pcmodel, 0, verbose=False)\n",
    "pcmodel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8eb2bdf-4c8a-4f73-a445-bfadab293359",
   "metadata": {},
   "source": [
    "### Check settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "653a080b-a7bb-4941-854f-05525df7e27a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulation Options\n",
      "================================\n",
      "min_relax_budget_for_objectives:\n",
      "\tTrue\n",
      "only_flux_abundance_reactions:\n",
      "\tFalse\n",
      "remove_blocked_reactions:\n",
      "\tTrue\n",
      "pFBA_factor:\n",
      "\tNone\n",
      "loopless:\n",
      "\tFalse\n",
      "objective_reactions:\n",
      "\t['NaKt']\n",
      "required_flux_reactions:\n",
      "\t['NaKt', 'PSFLIPt']\n",
      "protein_relaxations_to_restrict:\n",
      "\t['ATP11A', 'ATP11B', 'ATP11C', 'TMEM30A']\n",
      "protein_constraints_lb_to_relax:\n",
      "\t['TMEM30A']\n"
     ]
    }
   ],
   "source": [
    "# Check settings here\n",
    "invalid = sorted(set([x for x in objective_reactions if x not in pcmodel.reactions]))\n",
    "if invalid:\n",
    "    raise ValueError(f\"Objective reactions not found in model: {invalid}\")\n",
    "invalid = sorted(set([x for x in required_flux_reactions if x not in pcmodel.reactions]))\n",
    "if invalid:\n",
    "    raise ValueError(f\"Required flux capable reactions not found in model: {invalid}\")\n",
    "invalid = sorted(set([x for x in protein_relaxations_to_restrict + protein_constraints_lb_to_relax if x not in pcmodel.genes]))\n",
    "if invalid:\n",
    "    raise ValueError(f\"Genes/Proteins not found in model: {invalid}\")\n",
    "invalid = sorted(set([x for x in optimum_percents if x < 0]))\n",
    "if invalid:\n",
    "    raise ValueError(f\"Optimum values must be non-negative: {invalid}\")\n",
    "if pfba_factor is not None and pfba_factor < 1:\n",
    "    raise ValueError(f\"The pFBA factor should be greater than 1 if set.\")\n",
    "\n",
    "simulation_options_text = \"\\n\".join((\n",
    "    \"Simulation Options\",\n",
    "    \"================================\",\n",
    "    f\"min_relax_budget_for_objectives:\\n\\t{min_relax_budget_for_objectives}\",\n",
    "    f\"only_flux_abundance_reactions:\\n\\t{only_flux_abundance_reactions}\",\n",
    "    f\"remove_blocked_reactions:\\n\\t{remove_blocked_reactions}\",\n",
    "    f\"pFBA_factor:\\n\\t{pfba_factor}\",\n",
    "    F\"loopless:\\n\\t{loopless}\",\n",
    "    f\"objective_reactions:\\n\\t{objective_reactions}\",\n",
    "    f\"required_flux_reactions:\\n\\t{required_flux_reactions}\",\n",
    "    f\"protein_relaxations_to_restrict:\\n\\t{protein_relaxations_to_restrict}\",\n",
    "    f\"protein_constraints_lb_to_relax:\\n\\t{protein_constraints_lb_to_relax}\",\n",
    "))\n",
    "print(simulation_options_text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b97e3c2a-a931-48ff-b5ff-0c5e306524f9",
   "metadata": {},
   "source": [
    "### Define list of PC-models to load for simulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9e404c61-7c0f-479f-9a34-8a8d4279a1bf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of models to simulate: 38\n"
     ]
    }
   ],
   "source": [
    "with zipfile.ZipFile(f'{sample_pcmodels_dirpath}.zip', mode=\"r\") as zfile:\n",
    "    pcmodel_names = sorted([Path(x).name.replace(Path(x).suffix, \"\") for x in zfile.namelist() if x])\n",
    "if sample_pcmodels_dirpath.exists():\n",
    "    pcmodel_names += [filename.name for filename in sample_pcmodels_dirpath.iterdir()]\n",
    "# Check sample directory for existing files\n",
    "if not len(pcmodel_names) == len(set(pcmodel_names)):\n",
    "    raise ValueError(f\"Duplicates found: {[k for k, v in Counter(pcmodel_names).items() if v > 1]}\")\n",
    "\n",
    "models_to_simulate = sorted(set(pcmodel_names))\n",
    "handle_msg(f\"Number of models to simulate: {len(models_to_simulate)}\", print_msg=True)\n",
    "models_to_simulate;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dab4e5e8-eac9-4cbe-871a-56a754751001",
   "metadata": {},
   "source": [
    "### Generate results using pcFVA for context specific models\n",
    "Note that this can take a signficiant amount of time depending on the number of models and their sizes. Best to use a targeted approach in generating results. \n",
    "Alternatively, skip result generation and load the previously generated results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "02cc23c8-ffb9-4e1e-bab3-a2765d851180",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reactions in addition to the minimum for flux-abundance correlations\n",
    "list_of_reactions = []\n",
    "# # Use to get ALL reactions in the original model\n",
    "list_of_reactions += model.reactions.list_attr(\"id\")\n",
    "# # Use to get ALL reactions in the PC model\n",
    "# list_of_reactions += pcmodel.reactions.list_attr(\"id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f95de14b-afc3-4ab7-9e3c-2474a801459f",
   "metadata": {},
   "source": [
    "#### Generate results for subset of PC model reactions\n",
    "##### Reactions necessary for all flux-abundance correlation computations.\n",
    "To reduce computation time, a subset of reactions can be defined. \n",
    "For flux-abundance correlations, the minimum reaction set are reactions associated with genes associated and the corresponding enzyme dilution reaction for total enzyme."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4e9b8d5b-dcb8-4e49-aba8-290006484ac3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Minimum number of reactions minimize/maximize (minimum): 4711 / 15687\n"
     ]
    }
   ],
   "source": [
    "min_reaction_list = model.reactions.query(lambda x: x.gene_reaction_rule).list_attr(\n",
    "    \"id\"\n",
    ")\n",
    "# Add protein dilutons to see effective protein concentrations used\n",
    "min_reaction_list += pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, ProteinDilution)\n",
    ").list_attr(\"id\")\n",
    "min_reaction_list += pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, BudgetDilution)\n",
    ").list_attr(\"id\")\n",
    "\n",
    "# Already limited to reactions with gene reaction rules\n",
    "enzyme_totals_list = pcmodel.metabolites.query(\n",
    "    lambda x: x.id.startswith(f\"{enzyme_met_prefix}\")\n",
    "    and enzyme_met_suffix_total in x.id\n",
    ")\n",
    "enzyme_reaction_map = {\n",
    "    f\"{enzyme_rxn_prefix}{x}\": x.id.replace(f\"{enzyme_met_prefix}\", \"\").replace(\n",
    "        f\"{enzyme_met_suffix_total}_{x.compartment}\", \"\"\n",
    "    )\n",
    "    for x in enzyme_totals_list\n",
    "}\n",
    "# Combine lists\n",
    "min_reaction_list += list(enzyme_reaction_map)\n",
    "handle_msg(f\"Minimum number of reactions minimize/maximize (minimum): {len(min_reaction_list)} / {len(pcmodel.reactions)}\", print_msg=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54897e9c-f167-4d4d-84ad-e1d1d340a621",
   "metadata": {},
   "source": [
    "##### Refined set of PC model reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ee1f77e8-fada-451c-876d-fb3af50ee78b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions minimize/maximize (chosen): 5675 / 15687\n"
     ]
    }
   ],
   "source": [
    "if only_flux_abundance_reactions:\n",
    "    reaction_list = min_reaction_list.copy()\n",
    "else:\n",
    "    list_of_reactions = [getattr(rid, \"_id\", rid) for rid in list_of_reactions]\n",
    "    reaction_list = sorted(\n",
    "        [getattr(x, \"_id\", x) for x in set(min_reaction_list).union(list_of_reactions)]\n",
    "    )\n",
    "handle_msg(f\"Number of reactions minimize/maximize (chosen): {len(reaction_list)} / {len(pcmodel.reactions)}\", print_msg=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d04226f-5e51-4413-a886-7259f2e04873",
   "metadata": {},
   "source": [
    "### Determine previously existing solutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9f2c2dc7-cc28-4f5d-985c-cec28c599029",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Format: (zipped + unzipped)/total\n"
     ]
    }
   ],
   "source": [
    "existing_files = defaultdict(set)\n",
    "handle_msg(\"Format: (zipped + unzipped)/total\", print_msg=True)\n",
    "if remove_blocked_reactions:\n",
    "    reduced_models_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "    if (Path(f'{reduced_models_dirpath}.zip').exists() or Path(f'{reduced_models_dirpath}').exists()) and not overwrite:\n",
    "        # Search reduced model directory and zip file\n",
    "        if Path(f'{reduced_models_dirpath}.zip').exists():\n",
    "            with zipfile.ZipFile(f'{reduced_models_dirpath}.zip', mode=\"r\") as zfile:\n",
    "                existing_files[reduced_models_dirpath.name].update([Path(x).name.replace(f\".{ftype}\", \"\") for x in zfile.namelist() if x])\n",
    "        in_the_zip = len(existing_files[reduced_models_dirpath.name])\n",
    "        existing_files[reduced_models_dirpath.name].update([x.name.replace(f\".{ftype}\", \"\") for x in list(reduced_models_dirpath.iterdir())])\n",
    "        total = len(existing_files[reduced_models_dirpath.name])\n",
    "        handle_msg(f\"Number of existing reduced models: ({in_the_zip} + {total - in_the_zip}) / {total}\\n\", print_msg=True)\n",
    "    else:\n",
    "        existing_files[reduced_models_dirpath.name].update([])\n",
    "\n",
    "for optimum in optimum_percents:\n",
    "    # Replace decimals with hyphens\n",
    "    dirname = (\n",
    "        f\"Opt{optimum}\"\n",
    "        + (f\"_pFBA{pfba_factor}\" if pfba_factor else \"\")\n",
    "        + (\"loopless\" if loopless else \"\")\n",
    "    ).replace(\".\", \"-\")\n",
    "    optimum_dirpath = (pcfva_results_dirpath.parent if optimum == 0 else pcfva_results_dirpath) / dirname\n",
    "    optimum_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "    if (Path(f'{optimum_dirpath}.zip').exists() or Path(f'{optimum_dirpath}').exists()) and not overwrite:\n",
    "        # Search reduced model directory and zip file\n",
    "        if Path(f'{optimum_dirpath}.zip').exists():\n",
    "            with zipfile.ZipFile(f'{optimum_dirpath}.zip', mode=\"r\") as zfile:\n",
    "                existing_files[optimum_dirpath.name].update([Path(x).name for x in zfile.namelist() if x])\n",
    "        in_the_zip = len(existing_files[optimum_dirpath.name])\n",
    "        existing_files[optimum_dirpath.name].update([x.name for x in list(optimum_dirpath.iterdir())])\n",
    "        total = len(existing_files[optimum_dirpath.name])\n",
    "        handle_msg(\n",
    "                '\\t'.join((\n",
    "                    f\"Optimum: {optimum}\" \n",
    "                    + ('' if pfba_factor is None else '\\tpFBA factor: '+ str(pfba_factor)),\n",
    "                    f\"Number of existing solution: ({in_the_zip} + {total - in_the_zip}) / {total}\"\n",
    "                )), \n",
    "            print_msg=True\n",
    "            )\n",
    "    else:\n",
    "        existing_files[optimum_dirpath.name].update([])\n",
    "existing_files;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1803779-e5a4-4608-a09d-4c9ad5b27df5",
   "metadata": {},
   "source": [
    "## Run pcFVA\n",
    "### Define helper functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "459e3692-7ba7-4cfc-89bf-ea59a4a6ad6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_and_remove_blocked_reactions(pcmodel_sample, relaxation_rxns_required=None, prevent_removal=None, verbose=False):\n",
    "    if prevent_removal is None:\n",
    "        prevent_removal = []\n",
    "    pcmodel_sample = pcmodel_sample.copy()\n",
    "    n_reactions_original = len(pcmodel_sample.reactions)\n",
    "    n_genes_original = len(pcmodel_sample.genes)\n",
    "    pcmodel_sample.objective = (\n",
    "        sum(\n",
    "            [\n",
    "                r.flux_expression\n",
    "                for r in pcmodel_sample.reactions.get_by_any(\n",
    "                    required_flux_reactions\n",
    "                )\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "    for relax_rxn in pcmodel_sample.reactions.query(lambda x: x.id.startswith(relaxation_rxn_prefix)):\n",
    "        if relax_rxn.id in relaxation_rxns_required:\n",
    "            continue\n",
    "        else:\n",
    "            relax_rxn.bounds = (0, 0)\n",
    "\n",
    "    reactions_to_remove = find_blocked_reactions(\n",
    "        model=pcmodel_sample,\n",
    "        reaction_list=None,\n",
    "        zero_cutoff=COBRA_CONFIGURATION.tolerance,\n",
    "        open_exchanges=True,\n",
    "        processes=min(60, COBRA_CONFIGURATION.processes),\n",
    "    )\n",
    "    reactions_to_remove = sorted(set(reactions_to_remove).difference([rxn for prot in prevent_removal for rxn in [f\"{protein_rxn_prefix}{protein_met_prefix}{prot}{comp_suffix}\", f\"{relaxation_rxn_prefix}{protein_met_prefix}{prot}{comp_suffix}\"]]))\n",
    "    pcmodel_sample.remove_reactions(reactions_to_remove, remove_orphans=True)\n",
    "    genes_to_remove = [\n",
    "        gene.id for gene in pcmodel_sample.genes\n",
    "        if not pcmodel_sample.reactions.has_id(f\"{protein_rxn_prefix}{protein_met_prefix}{gene.id}{comp_suffix}\") \n",
    "        and not pcmodel_sample.reactions.has_id(f\"{relaxation_rxn_prefix}{protein_met_prefix}{gene.id}{comp_suffix}\")\n",
    "    ]\n",
    "    genes_to_remove = sorted(set(genes_to_remove).difference(prevent_removal))\n",
    "    remove_genes(pcmodel_sample, gene_list=genes_to_remove, remove_reactions=True)\n",
    "    handle_msg(f\"Number of blocked reactions removed: {n_reactions_original - len(pcmodel_sample.reactions)}\", print_msg=verbose)\n",
    "    handle_msg(f\"Number of associated genes removed: {n_genes_original - len(pcmodel_sample.genes)}\", print_msg=verbose)\n",
    "    return pcmodel_sample"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf338f15-6879-4c60-a336-09db7a2d4ffd",
   "metadata": {},
   "source": [
    "### Simulate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dbc8e0a4-d283-456f-a250-1d768978b207",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C0\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C0\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6327\n",
      "Number of associated genes removed: 249\n",
      "Number of reactions minimize/maximize for sample: 3249 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C0\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C0\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C1\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C1\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6304\n",
      "Number of associated genes removed: 246\n",
      "Number of reactions minimize/maximize for sample: 3257 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C1\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C1\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C10\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C10\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6310\n",
      "Number of associated genes removed: 247\n",
      "Number of reactions minimize/maximize for sample: 3255 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C10\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C10\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C11\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C11\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6348\n",
      "Number of associated genes removed: 253\n",
      "Number of reactions minimize/maximize for sample: 3243 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C11\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C11\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C12\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C12\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6296\n",
      "Number of associated genes removed: 244\n",
      "Number of reactions minimize/maximize for sample: 3259 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C12\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C12\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C13\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C13\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6313\n",
      "Number of associated genes removed: 247\n",
      "Number of reactions minimize/maximize for sample: 3255 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C13\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C13\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C14\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C14\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6304\n",
      "Number of associated genes removed: 246\n",
      "Number of reactions minimize/maximize for sample: 3257 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C14\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C14\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C15\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C15\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6316\n",
      "Number of associated genes removed: 247\n",
      "Number of reactions minimize/maximize for sample: 3251 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C15\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C15\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C16\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C16\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6306\n",
      "Number of associated genes removed: 245\n",
      "Number of reactions minimize/maximize for sample: 3254 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C16\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C16\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C17\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C17\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6310\n",
      "Number of associated genes removed: 247\n",
      "Number of reactions minimize/maximize for sample: 3255 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C17\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C17\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C2\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C2\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6300\n",
      "Number of associated genes removed: 245\n",
      "Number of reactions minimize/maximize for sample: 3258 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C2\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C2\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C3\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C3\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6331\n",
      "Number of associated genes removed: 248\n",
      "Number of reactions minimize/maximize for sample: 3251 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C3\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C3\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C4\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C4\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6331\n",
      "Number of associated genes removed: 250\n",
      "Number of reactions minimize/maximize for sample: 3248 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C4\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C4\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C5\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C5\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6296\n",
      "Number of associated genes removed: 244\n",
      "Number of reactions minimize/maximize for sample: 3259 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C5\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C5\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C6\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C6\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6320\n",
      "Number of associated genes removed: 248\n",
      "Number of reactions minimize/maximize for sample: 3250 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C6\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C6\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C7\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C7\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6445\n",
      "Number of associated genes removed: 253\n",
      "Number of reactions minimize/maximize for sample: 3230 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C7\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C7\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C8\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C8\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6314\n",
      "Number of associated genes removed: 247\n",
      "Number of reactions minimize/maximize for sample: 3252 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C8\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C8\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele0_C9\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele0_C9\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6320\n",
      "Number of associated genes removed: 248\n",
      "Number of reactions minimize/maximize for sample: 3250 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele0_C9\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele0_C9\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C18\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C18\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6789\n",
      "Number of associated genes removed: 274\n",
      "Number of reactions minimize/maximize for sample: 3072 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C18\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C18\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C19\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C19\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6584\n",
      "Number of associated genes removed: 266\n",
      "Number of reactions minimize/maximize for sample: 3149 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C19\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C19\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C20\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C20\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6937\n",
      "Number of associated genes removed: 296\n",
      "Number of reactions minimize/maximize for sample: 3053 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C20\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C20\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C21\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C21\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6767\n",
      "Number of associated genes removed: 272\n",
      "Number of reactions minimize/maximize for sample: 3059 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C21\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C21\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C22\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C22\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6604\n",
      "Number of associated genes removed: 265\n",
      "Number of reactions minimize/maximize for sample: 3152 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C22\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C22\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C23\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C23\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6487\n",
      "Number of associated genes removed: 256\n",
      "Number of reactions minimize/maximize for sample: 3184 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C23\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C23\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C24\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C24\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7441\n",
      "Number of associated genes removed: 292\n",
      "Number of reactions minimize/maximize for sample: 2734 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C24\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C24\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele1_C25\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele1_C25\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6794\n",
      "Number of associated genes removed: 288\n",
      "Number of reactions minimize/maximize for sample: 3067 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele1_C25\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele1_C25\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele2_S379_D10\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele2_S379_D10\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7977\n",
      "Number of associated genes removed: 308\n",
      "Number of reactions minimize/maximize for sample: 2484 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele2_S379_D10\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele2_S379_D10\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele2_S379_D23\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele2_S379_D23\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7201\n",
      "Number of associated genes removed: 305\n",
      "Number of reactions minimize/maximize for sample: 2937 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele2_S379_D23\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele2_S379_D23\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele2_S379_D42\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele2_S379_D42\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7765\n",
      "Number of associated genes removed: 303\n",
      "Number of reactions minimize/maximize for sample: 2582 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele2_S379_D42\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele2_S379_D42\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele2_S417_D10\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele2_S417_D10\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7418\n",
      "Number of associated genes removed: 323\n",
      "Number of reactions minimize/maximize for sample: 2864 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele2_S417_D10\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele2_S417_D10\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele2_S417_D23\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele2_S417_D23\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7981\n",
      "Number of associated genes removed: 319\n",
      "Number of reactions minimize/maximize for sample: 2500 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele2_S417_D23\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele2_S417_D23\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Allele2_S417_D42\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Allele2_S417_D42\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 8013\n",
      "Number of associated genes removed: 322\n",
      "Number of reactions minimize/maximize for sample: 2504 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Allele2_S417_D42\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Allele2_S417_D42\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Mean_Allele0\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Mean_Allele0\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6304\n",
      "Number of associated genes removed: 246\n",
      "Number of reactions minimize/maximize for sample: 3257 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Mean_Allele0\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Mean_Allele0\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Mean_Allele1\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Mean_Allele1\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6310\n",
      "Number of associated genes removed: 246\n",
      "Number of reactions minimize/maximize for sample: 3253 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Mean_Allele1\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Mean_Allele1\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Mean_Allele2\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Mean_Allele2\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6800\n",
      "Number of associated genes removed: 269\n",
      "Number of reactions minimize/maximize for sample: 3069 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Mean_Allele2\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Mean_Allele2\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Median_Allele0\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Median_Allele0\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6320\n",
      "Number of associated genes removed: 248\n",
      "Number of reactions minimize/maximize for sample: 3250 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Median_Allele0\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Median_Allele0\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Median_Allele1\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Median_Allele1\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 6534\n",
      "Number of associated genes removed: 266\n",
      "Number of reactions minimize/maximize for sample: 3164 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Median_Allele1\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Median_Allele1\n",
      "====================================================\n",
      "Computing pcFVA results for RBC_GEM_PC_Median_Allele2\n",
      "====================================================\n",
      "Loading PC-model for RBC_GEM_PC_Median_Allele2\n",
      "Determining reactions and protein constraints to remove\n",
      "Number of blocked reactions removed: 7746\n",
      "Number of associated genes removed: 304\n",
      "Number of reactions minimize/maximize for sample: 2567 / 15687\n",
      "Starting simulations for RBC_GEM_PC_Median_Allele2\n",
      "Finished pcFVA for optimum: 0\n",
      "Finished pcFVA for RBC_GEM_PC_Median_Allele2\n",
      "Compressing results\n",
      "Appending model files to temporary zip file\n",
      "Setting temporary zip file as the new zip file\n",
      "Finished compression, cleaning up files\n",
      "Finished cleanup\n",
      "\n",
      "Appending model files to temporary zip file\n",
      "Setting temporary zip file as the new zip file\n",
      "Finished compression, cleaning up files\n",
      "Finished cleanup\n",
      "\n"
     ]
    }
   ],
   "source": [
    "columns = [\"model\", \"reactions\", \"optimum\", \"min\", \"max\"]\n",
    "zip_kwargs = dict(\n",
    "    compression=zipfile.ZIP_DEFLATED,\n",
    "    compresslevel=None\n",
    ")\n",
    "if run_computations:\n",
    "    for idx, pcmodel_sample_id in enumerate(models_to_simulate, start=1):\n",
    "        handle_msg(\n",
    "            \"\\n\".join((\n",
    "                \"====================================================\",\n",
    "                f\"Computing pcFVA results for {pcmodel_sample_id}\",\n",
    "                \"====================================================\",\n",
    "                f\"Loading PC-model for {pcmodel_sample_id}\",\n",
    "            )),\n",
    "            print_msg=verbose\n",
    "        )\n",
    "        # See if a reduced model has already been created for simulation.\n",
    "        if (not overwrite and remove_blocked_reactions and pcmodel_sample_id in existing_files[reduced_models_dirpath.name]):\n",
    "            handle_msg(f\"Previously generated reduced model found.\", print_msg=True)\n",
    "            # Check if all simulations have been performed with reduced modelk\n",
    "            dirnames = [\n",
    "                (\n",
    "                    f\"Opt{optimum}\" \n",
    "                    + (f\"_pFBA{pfba_factor}\" if pfba_factor else \"\")\n",
    "                    + (\"loopless\" if loopless else \"\")\n",
    "                ).replace(\".\", \"-\")\n",
    "                for optimum in optimum_percents\n",
    "            ]\n",
    "            if all([f\"{pcmodel_sample_id}_FVAsol.csv\" in existing_files[dirname] for dirname in dirnames]):\n",
    "                handle_msg(\"All simulations already performed for model\", print_msg=verbose)\n",
    "                continue\n",
    "            else:\n",
    "                # Not all simulations performed, therefore load existing reduced model\n",
    "                try:\n",
    "                    with zipfile.ZipFile(f\"{reduced_models_dirpath}.zip\", 'r') as zfile:\n",
    "                        with zfile.open(f\"{pcmodel_sample_id}.{ftype}\", 'r') as model_file:\n",
    "                            pcmodel_sample = load_overlay_model(filename=io.StringIO(model_file.read().decode(\"utf-8\")), filetype=ftype)\n",
    "                except (KeyError, FileNotFoundError):\n",
    "                    # If file not found, model hasn't been put into zip file yet\n",
    "                    pcmodel_sample = load_overlay_model(filename=reduced_models_dirpath /f\"{pcmodel_sample_id}.{ftype}\")\n",
    "        else:\n",
    "            # Load full model\n",
    "            try:\n",
    "                with zipfile.ZipFile(f'{sample_pcmodels_dirpath}.zip', 'r') as zfile:\n",
    "                    with zfile.open(f\"{pcmodel_sample_id}.{ftype}\", 'r') as model_file:\n",
    "                        pcmodel_sample = load_overlay_model(filename=io.StringIO(model_file.read().decode(\"utf-8\")), filetype=ftype)\n",
    "            except (FileNotFoundError, KeyError):\n",
    "                # If file not found, model hasn't been put into zip file yet\n",
    "                pcmodel_sample = load_overlay_model(filename=sample_pcmodels_dirpath /f\"{pcmodel_sample_id}.{ftype}\")\n",
    "            # Restrict relaxation reactions for specific proteins\n",
    "            for protein in protein_relaxations_to_restrict:\n",
    "                protein_met = pcmodel_sample.metabolites.get_by_id(f\"{protein_met_prefix}{protein}{comp_suffix}\")\n",
    "                relax_prot_rxn = pcmodel_sample.reactions.get_by_id(f\"{relaxation_rxn_prefix}{protein_met.id}\")\n",
    "                relax_prot_rxn.bounds = (0, 0)\n",
    "            # Relax lower bound constraint for specific proteins\n",
    "            for protein in protein_constraints_lb_to_relax:\n",
    "                protein_met = pcmodel_sample.metabolites.get_by_id(f\"{protein_met_prefix}{protein}{comp_suffix}\")\n",
    "                protein_rxn = pcmodel_sample.reactions.get_by_id(f\"{protein_rxn_prefix}{protein_met.id}\")\n",
    "                protein_rxn.lower_bound = 0\n",
    "        \n",
    "            budget_rxn_relaxation = pcmodel_sample.reactions.get_by_id(f\"{budget_rxn_prefix}{budget_met_prefix}relaxation\")\n",
    "            # Determine smallest allowable relxation budget that allows flux through objectives and set as upper bound\n",
    "            with pcmodel_sample:\n",
    "                pcmodel_sample.objective = (sum([r.flux_expression for r in pcmodel_sample.reactions.get_by_any(required_flux_reactions)]) - budget_rxn_relaxation.flux_expression)\n",
    "                pcmodel_sample.objective_direction = \"max\"\n",
    "                 # Fail loudly, should not occur unless a restricted relaxation proteins are absolutely necessary\n",
    "                pcmodel_sample.slim_optimize(error_value=None)\n",
    "                relaxation_rxns_required = get_solution(pcmodel_sample, reactions=pcmodel_sample.reactions.query(lambda x: x.id.startswith(relaxation_rxn_prefix)))\n",
    "                relaxation_rxns_required = set(relaxation_rxns_required.fluxes[relaxation_rxns_required.fluxes != 0].index.to_list())\n",
    "                budget_min = budget_rxn_relaxation.flux\n",
    "            if min_relax_budget_for_objectives:\n",
    "                budget_rxn_relaxation.upper_bound = budget_min\n",
    "            \n",
    "            if remove_blocked_reactions:\n",
    "                handle_msg(\"Determining reactions and protein constraints to remove\", print_msg=verbose)\n",
    "                with warnings.catch_warnings(action=\"ignore\"):\n",
    "                    pcmodel_sample = find_and_remove_blocked_reactions(\n",
    "                        pcmodel_sample, \n",
    "                        relaxation_rxns_required=relaxation_rxns_required, \n",
    "                        prevent_removal=list(set(protein_relaxations_to_restrict).union(protein_constraints_lb_to_relax)),\n",
    "                        verbose=verbose\n",
    "                    )\n",
    "                pcmodel_sample.slim_optimize()\n",
    "                # Write reduced model into directory\n",
    "                write_cobra_model(pcmodel_sample, filename=reduced_models_dirpath / f\"{pcmodel_sample.id}.{ftype}\")\n",
    "        # Set objective reaction(s)\n",
    "        pcmodel_sample.objective = sum([r.flux_expression for r in pcmodel_sample.reactions.get_by_any(objective_reactions)])\n",
    "        # Not all reactions exist for reduced models, regenerate list per model\n",
    "        rxn_list = [x for x in reaction_list if x in pcmodel_sample.reactions]\n",
    "        handle_msg(f\"Number of reactions minimize/maximize for sample: {len(rxn_list)} / {len(pcmodel.reactions)}\", print_msg=verbose)\n",
    "        handle_msg(f\"Starting simulations for {pcmodel_sample}\", print_msg=verbose)\n",
    "        for optimum in optimum_percents:\n",
    "            dirname = (f\"Opt{optimum}\" + (f\"_pFBA{pfba_factor}\" if pfba_factor else \"\") + (\"loopless\" if loopless else \"\")).replace(\".\", \"-\")\n",
    "            filename = f\"{pcmodel_sample}_FVAsol.csv\"\n",
    "            if not overwrite and filename in existing_files[dirname]:\n",
    "                handle_msg(f\"Already simulated at optimum: {optimum}\" + ('' if pfba_factor is None else ' pFBA factor: '+ str(pfba_factor)), print_msg=verbose)\n",
    "                continue\n",
    "            fraction_of_optimum = round(optimum / 100, 4)\n",
    "            try:\n",
    "                pcfva_sol = flux_variability_analysis(\n",
    "                    pcmodel_sample,\n",
    "                    reaction_list=rxn_list,\n",
    "                    loopless=loopless,\n",
    "                    pfba_factor=pfba_factor,\n",
    "                    fraction_of_optimum=fraction_of_optimum,\n",
    "                    processes=min(60, COBRA_CONFIGURATION.processes),\n",
    "                )\n",
    "            except OptimizationError as e:\n",
    "                msg = f\"{pcmodel_sample_id} failed due to an exception.\"\n",
    "                handle_msg(msg, print_msg=verbose)\n",
    "                with open(pcfva_results_dirpath / f\"pcFVA-errors-{dirname}.log\", \"a\") as file:\n",
    "                    file.write(f\"{msg} {str(e)}\\n\")\n",
    "            else:\n",
    "                pcfva_sol.index.name = \"reactions\"\n",
    "                pcfva_sol = pcfva_sol.reset_index(drop=False)\n",
    "                pcfva_sol[\"model\"] = pcmodel_sample.id\n",
    "                pcfva_sol[\"optimum\"] = fraction_of_optimum\n",
    "                pcfva_sol = pcfva_sol.rename({\"minimum\": \"min\", \"maximum\": \"max\"}, axis=1)\n",
    "                optimum_dirpath = (pcfva_results_dirpath.parent if optimum == 0 else pcfva_results_dirpath) / dirname\n",
    "                # Save simulation results\n",
    "                pcfva_sol.to_csv(optimum_dirpath / filename, index=False)\n",
    "                handle_msg(f\"Finished pcFVA for optimum: {optimum}\", print_msg=verbose)\n",
    "        handle_msg(f\"Finished pcFVA for {pcmodel_sample_id}\", print_msg=verbose)\n",
    "handle_msg(\"Compressing results\", print_msg=verbose)\n",
    "dirpaths = [reduced_models_dirpath]\n",
    "for optimum in optimum_percents:\n",
    "    dirname = (f\"Opt{optimum}\" + (f\"_pFBA{pfba_factor}\" if pfba_factor else \"\") + (\"loopless\" if loopless else \"\")).replace(\".\", \"-\")\n",
    "    dirpaths += [(pcfva_results_dirpath.parent if optimum == 0 else pcfva_results_dirpath) / dirname]\n",
    "\n",
    "for dirpath in dirpaths:\n",
    "    with tempfile.TemporaryDirectory() as tmpdir:\n",
    "        if not overwrite and Path(f\"{dirpath}.zip\").exists():\n",
    "            handle_msg(\"Copying original zip to temporary directory\", print_msg=verbose)\n",
    "            shutil.copy(f\"{dirpath}.zip\", tmpdir)\n",
    "        handle_msg(\"Appending model files to temporary zip file\", print_msg=verbose)\n",
    "        with zipfile.ZipFile(f\"{tmpdir}/{dirpath.name}.zip\", 'a', **zip_kwargs) as zfile:\n",
    "            existing_files = set([Path(x).name for x in zfile.namelist() if x])\n",
    "            for filename in list(dirpath.iterdir()):\n",
    "                if filename.name in existing_files:\n",
    "                    continue\n",
    "                zfile.write(f\"{filename}\", arcname=f\"{filename.name}\")\n",
    "        # Replacing original directory\n",
    "        handle_msg(\"Setting temporary zip file as the new zip file\", print_msg=verbose)\n",
    "        shutil.copy(f\"{tmpdir}/{dirpath.name}.zip\", dirpath.parent)\n",
    "    handle_msg(\"Finished compression, cleaning up files\", print_msg=verbose)\n",
    "    shutil.rmtree(str(dirpath))\n",
    "    handle_msg(\"Finished cleanup\", print_msg=verbose)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c282559-19d0-4d8b-a7b3-2a112c398c7f",
   "metadata": {},
   "source": [
    "### Combine all simulation results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "51f1f78b-545b-4e67-a612-58a30263a1c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>model</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123875</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Mean_Allele1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.040705</td>\n",
       "      <td>0.040236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123876</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Mean_Allele2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.053735</td>\n",
       "      <td>0.052435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123877</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.037787</td>\n",
       "      <td>0.036772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123878</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.040307</td>\n",
       "      <td>0.040021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123879</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123880 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              reactions                      model  optimum       min  \\\n",
       "0       1SGTH2OHE1ABCte      RBC_GEM_PC_Allele0_C0      0.0  0.000000   \n",
       "1       1SGTH2OHE1ABCte      RBC_GEM_PC_Allele0_C1      0.0  0.000000   \n",
       "2       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C10      0.0  0.000000   \n",
       "3       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C11      0.0  0.000000   \n",
       "4       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C12      0.0  0.000000   \n",
       "...                 ...                        ...      ...       ...   \n",
       "123875          XYLU_Dt    RBC_GEM_PC_Mean_Allele1      0.0 -0.040705   \n",
       "123876          XYLU_Dt    RBC_GEM_PC_Mean_Allele2      0.0 -0.053735   \n",
       "123877          XYLU_Dt  RBC_GEM_PC_Median_Allele0      0.0 -0.037787   \n",
       "123878          XYLU_Dt  RBC_GEM_PC_Median_Allele1      0.0 -0.040307   \n",
       "123879          XYLU_Dt  RBC_GEM_PC_Median_Allele2      0.0 -0.056075   \n",
       "\n",
       "             max  \n",
       "0       0.014688  \n",
       "1       0.015410  \n",
       "2       0.014478  \n",
       "3       0.013896  \n",
       "4       0.011391  \n",
       "...          ...  \n",
       "123875  0.040236  \n",
       "123876  0.052435  \n",
       "123877  0.036772  \n",
       "123878  0.040021  \n",
       "123879  0.056075  \n",
       "\n",
       "[123880 rows x 5 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load combine solutions into single DataFrame\n",
    "def read_csv_from_zip(csv_file):\n",
    "    with zfile.open(csv_file) as f:\n",
    "        return pd.read_csv(f)\n",
    "\n",
    "dfs_to_concat = []\n",
    "for optimum in optimum_percents:\n",
    "    dirname = [f\"Opt{optimum}\"]\n",
    "    dirname += [f\"pFBA{pfba_factor}\"] if pfba_factor is not None else []\n",
    "    dirname = \"_\".join(dirname).replace(\".\", \"-\")\n",
    "    dirpath = (pcfva_results_dirpath.parent if optimum == 0 else pcfva_results_dirpath)\n",
    "    with zipfile.ZipFile(dirpath / f'{dirname}.zip', 'a', **zip_kwargs) as zfile:\n",
    "        dfs_to_concat += [read_csv_from_zip(csv_file) for csv_file in zfile.namelist() if csv_file.endswith('.csv')]\n",
    "\n",
    "df_pcfva_all = pd.concat(dfs_to_concat, ignore_index=True)\n",
    "\n",
    "# Fill missing (blocked) reactions with zero values to have all reactions for all models at all optimums\n",
    "df_all_reactions = pd.DataFrame(\n",
    "    index=pd.Index(df_pcfva_all[\"reactions\"].unique(), name=\"reactions\"),\n",
    ")\n",
    "\n",
    "\n",
    "columns_to_fill = [\"reactions\", \"model\", \"optimum\"]\n",
    "for idx, col in enumerate(columns_to_fill[1:], start=2):\n",
    "    # Create initial column\n",
    "    df_all_reactions.loc[:, col] = pd.NA\n",
    "    # Fill unique values\n",
    "    df_all_reactions.loc[:, col] = set(df_pcfva_all[col].unique())\n",
    "    # Explode to propogate\n",
    "    df_all_reactions = df_all_reactions.explode(col).reset_index(drop=False)\n",
    "    df_all_reactions = df_all_reactions.set_index(columns_to_fill[:idx])\n",
    "df_all_reactions = df_all_reactions.sort_index()\n",
    "df_pcfva_all = df_all_reactions.merge(df_pcfva_all.set_index(columns_to_fill), left_index=True, right_index=True, how=\"left\")\n",
    "df_pcfva_all = df_pcfva_all.reset_index(drop=False)\n",
    "# Ensure fill worked as expected\n",
    "assert len(df_pcfva_all.index) == np.prod([df_pcfva_all[col].nunique() for col in columns_to_fill])\n",
    "df_pcfva_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3222db91-41bc-46ab-a4d4-84ab95b98021",
   "metadata": {},
   "source": [
    "## Export results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4ee9f17a-a845-4e6f-8eab-d83aa10bdc68",
   "metadata": {},
   "outputs": [],
   "source": [
    "compression_opts = dict(method='zip', archive_name=f\"{pcmodel.id}_All_FVAsols.csv\")\n",
    "compression_opts.update(zip_kwargs)\n",
    "df_pcfva_all.to_csv(\n",
    "    pcfva_results_dirpath / f\"{pcmodel.id}_All_FVAsols.zip\", \n",
    "    index=False,\n",
    "    compression=compression_opts,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adf7cefd-e3bb-4240-8b90-d69db8635f42",
   "metadata": {},
   "source": [
    "## Test load pcFVA generated results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "1b3a12d9-85cb-4ea7-b2a5-5882debb1e1e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>model</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123875</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Mean_Allele1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.040705</td>\n",
       "      <td>0.040236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123876</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Mean_Allele2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.053735</td>\n",
       "      <td>0.052435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123877</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.037787</td>\n",
       "      <td>0.036772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123878</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.040307</td>\n",
       "      <td>0.040021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123879</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123880 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              reactions                      model  optimum       min  \\\n",
       "0       1SGTH2OHE1ABCte      RBC_GEM_PC_Allele0_C0      0.0  0.000000   \n",
       "1       1SGTH2OHE1ABCte      RBC_GEM_PC_Allele0_C1      0.0  0.000000   \n",
       "2       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C10      0.0  0.000000   \n",
       "3       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C11      0.0  0.000000   \n",
       "4       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C12      0.0  0.000000   \n",
       "...                 ...                        ...      ...       ...   \n",
       "123875          XYLU_Dt    RBC_GEM_PC_Mean_Allele1      0.0 -0.040705   \n",
       "123876          XYLU_Dt    RBC_GEM_PC_Mean_Allele2      0.0 -0.053735   \n",
       "123877          XYLU_Dt  RBC_GEM_PC_Median_Allele0      0.0 -0.037787   \n",
       "123878          XYLU_Dt  RBC_GEM_PC_Median_Allele1      0.0 -0.040307   \n",
       "123879          XYLU_Dt  RBC_GEM_PC_Median_Allele2      0.0 -0.056075   \n",
       "\n",
       "             max  \n",
       "0       0.014688  \n",
       "1       0.015410  \n",
       "2       0.014478  \n",
       "3       0.013896  \n",
       "4       0.011391  \n",
       "...          ...  \n",
       "123875  0.040236  \n",
       "123876  0.052435  \n",
       "123877  0.036772  \n",
       "123878  0.040021  \n",
       "123879  0.056075  \n",
       "\n",
       "[123880 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test to see if results were recently generated in this run, otherwise load DataFrame of generated results\n",
    "df_pcfva_all = pd.read_csv(\n",
    "    pcfva_results_dirpath / f\"{pcmodel.id}_All_FVAsols.zip\",\n",
    "    index_col=None,\n",
    ")\n",
    "df_pcfva_all"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
