{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc9114c6-b2d1-48a1-8ba3-6686a1ea2449",
   "metadata": {},
   "source": [
    "# Compute statistically significant fluxes between groups - REDS Recall, ATP11C V972M\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d694e76-31c0-4aba-add6-a07a00a5d04c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.3\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.13.4\n",
      "bio                                   1.8.0\n",
      "cobra                                0.29.1\n",
      "depinfo                               2.2.0\n",
      "gurobipy                             12.0.3\n",
      "matplotlib                           3.10.3\n",
      "matplotlib-venn                       1.1.2\n",
      "memote                               0.17.0\n",
      "networkx                                3.5\n",
      "notebook                              7.4.4\n",
      "openpyxl                              3.1.5\n",
      "pandas                                2.3.1\n",
      "pre-commit                            4.2.0\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.32.4\n",
      "scikit-learn                          1.7.0\n",
      "scipy                                1.16.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip          25.1\n",
      "setuptools 78.1.1\n",
      "wheel      0.45.1\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Windows 10-AMD64\n",
      "CPython  3.11.13\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import defaultdict\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import kruskal, mannwhitneyu, false_discovery_control\n",
    "from itertools import combinations\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    ensure_iterable,\n",
    ")\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    DEFAULT_PROTEOME_COMPARTMENT,\n",
    "    DEFAULT_PREFIX_SUFFIX_VALUES,\n",
    "    add_relaxation_budget,\n",
    "    load_overlay_model,\n",
    "    EnzymeDilution,\n",
    ")\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "384c972a-9f2f-4b88-a65c-017d1a6561d6",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "418ddc9c-b48e-400a-b5a1-e4e6bfe73bd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>100000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-09\n",
       "lower_bound: -1e-08\n",
       "upper_bound: 100000000.0\n",
       "processes: 127\n",
       "cache_directory: C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "# Set bound defaults much larger to prevent model loading issues\n",
    "COBRA_CONFIGURATION.bounds = (-1e-8, 1e8)\n",
    "COBRA_CONFIGURATION.tolerance = 1e-9\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c3720a-d637-4115-8960-f48671ee725a",
   "metadata": {},
   "source": [
    "### Define organism, model, and dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5b08faa-240f-44ee-9d16-ddbdc556852f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATP11C_V972M_Sample'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "organism = \"Human\"\n",
    "model_id = \"RBC_GEM\"\n",
    "dataset_name = \"REDSRecall\"\n",
    "genotype = \"ATP11C_V972M\"\n",
    "grouped_data_key = f'{genotype}_Sample'\n",
    "grouped_data_key"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0ccb7e-c784-4b63-bfe9-7ebdce0e24b5",
   "metadata": {},
   "source": [
    "### Set variables for sample identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6435bfd2-9c1f-4f09-8a7a-3d0a0efba741",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_key = \"SAMPLE ID\"\n",
    "donor_key = \"PUBLIC RECALL DONOR ID\"\n",
    "\n",
    "# For sample IDs\n",
    "donor_re = re.compile(rf\"(?P<donor>S(?P<num>\\d\\d\\d))\")\n",
    "operations = \"|\".join([x.capitalize() for x in [\"mean\", \"median\"]])\n",
    "operation_re = re.compile(r\"(?P<op>\" + operations + r\")\\_(?P<group>\\w+)\")\n",
    "sample_id_re = re.compile(\n",
    "    r\"(?!\" + operations + r\")\" + donor_re.pattern + r\"\\_\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e478c85d-b0ed-466c-b6d3-e8dead52f018",
   "metadata": {},
   "source": [
    "### Set computation options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52e59494-b23d-4775-8863-1baeb1f46194",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"xml\"  # In our experience, SBML/XML loads faster, but will take up to 4x more space uncompressed as compared to JSON\n",
    "run_computations = True  # Keep off to use previously computed results\n",
    "overwrite = False  # Whether to allow overwriting of previous simulation results\n",
    "verbose = True\n",
    "\n",
    "# Objective reactions\n",
    "objective_reactions = [\"NaKt\"] \n",
    "# Reactions that must have the capability to carry flux, sort for consistency\n",
    "required_flux_reactions = [\"PSFLIPt\"] # Add reactions to this list\n",
    "required_flux_reactions = sorted(set(objective_reactions + required_flux_reactions))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c42e70c-e09e-40cb-bbd5-41e578efa775",
   "metadata": {},
   "source": [
    "#### Set prefixes/suffixes to expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "04209050-5c8f-405c-bd38-c890ddbc6da5",
   "metadata": {},
   "outputs": [],
   "source": [
    "enzyme_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.dilution\"]\n",
    "enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.metabolite\"]\n",
    "enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]\n",
    "comp_suffix = f\"_{DEFAULT_PROTEOME_COMPARTMENT}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "295bb5e1-67d6-4436-af4a-0f34befaa111",
   "metadata": {},
   "source": [
    "### Set figure options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "06ad3312-69a9-4bd9-92c4-a2d467bed985",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc2d514-cb81-4968-95b6-95e758cd3e73",
   "metadata": {},
   "source": [
    "### Set paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0ccd462b-0c65-4b33-b745-6a4e74996f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set paths\n",
    "overlay_dirpath = get_dirpath(\"analysis\") / \"OVERLAY\" / organism\n",
    "model_dirpath = overlay_dirpath / model_id\n",
    "results_dirpath = (get_dirpath(use_temp=\"processed\") / model_id / \"OVERLAY\" / organism / dataset_name / grouped_data_key)\n",
    "pcfva_results_dirpath = (results_dirpath / \"pcFVA\" / \"_\".join((\"REQ\", *required_flux_reactions)) /  \"_\".join((\"OBJ\", *objective_reactions)))\n",
    "corr_results_dirpath = results_dirpath / \"correlations\"\n",
    "# Ensure directory  exists\n",
    "corr_results_dirpath.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "806b8d0f-8ca2-48e4-9da1-7db464f01407",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "541f7b2c-ad0d-4404-899b-fa05a0bd7fa1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Set parameter LicenseID to value 2664191\n",
      "Academic license - for non-commercial use only - expires 2026-05-12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>1e2288ae350</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>7815</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>15687</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>723</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>68</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC at 0x1e2288ae350>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_cobra_model(filename=model_dirpath / f\"{model_id}.xml\")\n",
    "pcmodel = load_overlay_model(filename=model_dirpath / f\"{model_id}_PC.xml\")\n",
    "\n",
    "# Add relaxation budget to initial PC model to get names of relaxation reactions\n",
    "add_relaxation_budget(pcmodel, 0, verbose=False)\n",
    "pcmodel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49cf0307-bf28-4f10-8784-143d62c669a7",
   "metadata": {},
   "source": [
    "## Load pcFVA generated results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "537fccd8-52b6-44c9-83a0-af5c62b7fd2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>model</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.015410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C11</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.013896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>RBC_GEM_PC_Allele0_C12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123875</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Mean_Allele1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.040705</td>\n",
       "      <td>0.040236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123876</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Mean_Allele2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.053735</td>\n",
       "      <td>0.052435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123877</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.037787</td>\n",
       "      <td>0.036772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123878</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.040307</td>\n",
       "      <td>0.040021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123879</th>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>123880 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              reactions                      model  optimum       min  \\\n",
       "0       1SGTH2OHE1ABCte      RBC_GEM_PC_Allele0_C0      0.0  0.000000   \n",
       "1       1SGTH2OHE1ABCte      RBC_GEM_PC_Allele0_C1      0.0  0.000000   \n",
       "2       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C10      0.0  0.000000   \n",
       "3       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C11      0.0  0.000000   \n",
       "4       1SGTH2OHE1ABCte     RBC_GEM_PC_Allele0_C12      0.0  0.000000   \n",
       "...                 ...                        ...      ...       ...   \n",
       "123875          XYLU_Dt    RBC_GEM_PC_Mean_Allele1      0.0 -0.040705   \n",
       "123876          XYLU_Dt    RBC_GEM_PC_Mean_Allele2      0.0 -0.053735   \n",
       "123877          XYLU_Dt  RBC_GEM_PC_Median_Allele0      0.0 -0.037787   \n",
       "123878          XYLU_Dt  RBC_GEM_PC_Median_Allele1      0.0 -0.040307   \n",
       "123879          XYLU_Dt  RBC_GEM_PC_Median_Allele2      0.0 -0.056075   \n",
       "\n",
       "             max  \n",
       "0       0.014688  \n",
       "1       0.015410  \n",
       "2       0.014478  \n",
       "3       0.013896  \n",
       "4       0.011391  \n",
       "...          ...  \n",
       "123875  0.040236  \n",
       "123876  0.052435  \n",
       "123877  0.036772  \n",
       "123878  0.040021  \n",
       "123879  0.056075  \n",
       "\n",
       "[123880 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load DataFrame of generated results\n",
    "df_pcfva_all = pd.read_csv(\n",
    "    pcfva_results_dirpath / f\"{pcmodel.id}_All_FVAsols.zip\",\n",
    "    index_col=None,\n",
    ").fillna(0)\n",
    "\n",
    "df_pcfva_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63a4524-c95b-4980-a95f-5ba109e75025",
   "metadata": {},
   "source": [
    "## Create DataFrame for calculations and visualizations\n",
    "### Get maximum reaction fluxes and associated abundance values\n",
    "#### Get maximum reaction fluxes and ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3efaec1d-6db8-4b96-975b-67303d06d9ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [min, max, range]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxns = model.reactions.list_attr(\"id\")\n",
    "df_max_flux_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "df_max_flux_per_model = df_max_flux_per_model.groupby([\"model\", \"reactions\", \"optimum\"])[[\"min\", \"max\"]].agg(\n",
    "    {\n",
    "        \"min\": \"min\", # Minimum reaction flux per model\n",
    "        \"max\": \"max\", # Maximum reaction flux per model\n",
    "    }\n",
    ")\n",
    "# Address issues possibly caused by floating point precision, ideally a value that prevents any negative ranges\n",
    "df_max_flux_per_model.loc[df_max_flux_per_model[\"max\"] < df_max_flux_per_model[\"min\"], [\"max\", \"min\"]] = [0, 0]\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_flux_per_model[\"max\"] = df_max_flux_per_model[\"max\"].apply(lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol))))\n",
    "df_max_flux_per_model[\"min\"] = df_max_flux_per_model[\"min\"].apply(lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol))))\n",
    "df_max_flux_per_model[\"range\"] = df_max_flux_per_model[\"max\"] - df_max_flux_per_model[\"min\"]\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_flux_per_model[df_max_flux_per_model[\"range\"] < 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e000d2d8-7a7c-4b02-b088-867e04ce7dbe",
   "metadata": {},
   "source": [
    "#### Get maximum \"enzyme\" abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4d6feeea-cfe4-4da9-99db-47f2729bd4e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [abundance]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxns = pcmodel.reactions.query(lambda x: isinstance(x, EnzymeDilution) and x.id.endswith(f\"{enzyme_met_suffix_total}{comp_suffix}\")).list_attr(\"id\")\n",
    "df_max_abundance_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "# Rename dilution reactions to match \n",
    "reaction_enzyme_map = {\n",
    "    enzyme_rid: enzyme_rid.replace(\n",
    "        f\"{enzyme_rxn_prefix}{enzyme_met_prefix}\", \"\"\n",
    "    ).replace(\n",
    "        f\"{enzyme_met_suffix_total}{comp_suffix}\", \"\"\n",
    "    )\n",
    "    for enzyme_rid in df_max_abundance_per_model[\"reactions\"]\n",
    "}\n",
    "df_max_abundance_per_model[\"reactions\"] = df_max_abundance_per_model[\"reactions\"].replace(reaction_enzyme_map)\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.groupby([\"model\", \"reactions\", \"optimum\"])[[\"max\"]].max()\n",
    "# Address issues possibly caused by floating point precision, atol is ideally a value that prevents any negative ranges\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(lambda x: 0 if x < 0 else x)\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol))))\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.rename({\"max\": \"abundance\"}, axis=1)\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_abundance_per_model[(df_max_abundance_per_model < 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95269b45-4f16-4d09-95c5-af4734b23d82",
   "metadata": {},
   "source": [
    "#### Merge DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "60602d4b-89ba-493b-ae29-55a35c669a31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>1SGTH2OHE2ABCte</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1QN1GST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1QN4GST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1SQOX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63569</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XPPT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>1.137815</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63570</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLK</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63571</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLTD_Dx</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "      <td>0.112151</td>\n",
       "      <td>0.239638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63572</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLTt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "      <td>0.112151</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63573</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "      <td>0.112151</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63574 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           model        reactions  optimum       min  \\\n",
       "0          RBC_GEM_PC_Allele0_C0  1SGTH2OHE1ABCte      0.0  0.000000   \n",
       "1          RBC_GEM_PC_Allele0_C0  1SGTH2OHE2ABCte      0.0  0.000000   \n",
       "2          RBC_GEM_PC_Allele0_C0       23E1QN1GST      0.0  0.000000   \n",
       "3          RBC_GEM_PC_Allele0_C0       23E1QN4GST      0.0  0.000000   \n",
       "4          RBC_GEM_PC_Allele0_C0         23E1SQOX      0.0  0.000000   \n",
       "...                          ...              ...      ...       ...   \n",
       "63569  RBC_GEM_PC_Median_Allele2             XPPT      0.0  0.000000   \n",
       "63570  RBC_GEM_PC_Median_Allele2             XYLK      0.0  0.000000   \n",
       "63571  RBC_GEM_PC_Median_Allele2         XYLTD_Dx      0.0 -0.056075   \n",
       "63572  RBC_GEM_PC_Median_Allele2            XYLTt      0.0 -0.056075   \n",
       "63573  RBC_GEM_PC_Median_Allele2          XYLU_Dt      0.0 -0.056075   \n",
       "\n",
       "            max     range  abundance  \n",
       "0      0.014688  0.014688   0.062771  \n",
       "1      0.014688  0.014688   0.062771  \n",
       "2      0.014688  0.014688  21.663839  \n",
       "3      0.014688  0.014688  21.663839  \n",
       "4      1.296063  1.296063        NaN  \n",
       "...         ...       ...        ...  \n",
       "63569  0.001258  0.001258   1.137815  \n",
       "63570  0.000000  0.000000   0.000000  \n",
       "63571  0.056075  0.112151   0.239638  \n",
       "63572  0.056075  0.112151        NaN  \n",
       "63573  0.056075  0.112151        NaN  \n",
       "\n",
       "[63574 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_all = pd.merge(\n",
    "    df_max_flux_per_model,\n",
    "    df_max_abundance_per_model,\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    ")\n",
    "df_data_all = df_data_all.reset_index(drop=False)\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3094239-cea4-4899-8cba-84b0f3969ba3",
   "metadata": {},
   "source": [
    "### Identify genotypes for results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dfe7da7a-4f23-4fc8-a0fa-7ad337c5f6cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>sample</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>1SGTH2OHE2ABCte</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1QN1GST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1QN4GST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1SQOX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63569</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XPPT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>0.001258</td>\n",
       "      <td>1.137815</td>\n",
       "      <td>Median_Allele2</td>\n",
       "      <td>Median</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63570</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLK</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>Median_Allele2</td>\n",
       "      <td>Median</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63571</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLTD_Dx</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "      <td>0.112151</td>\n",
       "      <td>0.239638</td>\n",
       "      <td>Median_Allele2</td>\n",
       "      <td>Median</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63572</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLTt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "      <td>0.112151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Median_Allele2</td>\n",
       "      <td>Median</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63573</th>\n",
       "      <td>RBC_GEM_PC_Median_Allele2</td>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.056075</td>\n",
       "      <td>0.056075</td>\n",
       "      <td>0.112151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Median_Allele2</td>\n",
       "      <td>Median</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>63574 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           model        reactions  optimum       min  \\\n",
       "0          RBC_GEM_PC_Allele0_C0  1SGTH2OHE1ABCte      0.0  0.000000   \n",
       "1          RBC_GEM_PC_Allele0_C0  1SGTH2OHE2ABCte      0.0  0.000000   \n",
       "2          RBC_GEM_PC_Allele0_C0       23E1QN1GST      0.0  0.000000   \n",
       "3          RBC_GEM_PC_Allele0_C0       23E1QN4GST      0.0  0.000000   \n",
       "4          RBC_GEM_PC_Allele0_C0         23E1SQOX      0.0  0.000000   \n",
       "...                          ...              ...      ...       ...   \n",
       "63569  RBC_GEM_PC_Median_Allele2             XPPT      0.0  0.000000   \n",
       "63570  RBC_GEM_PC_Median_Allele2             XYLK      0.0  0.000000   \n",
       "63571  RBC_GEM_PC_Median_Allele2         XYLTD_Dx      0.0 -0.056075   \n",
       "63572  RBC_GEM_PC_Median_Allele2            XYLTt      0.0 -0.056075   \n",
       "63573  RBC_GEM_PC_Median_Allele2          XYLU_Dt      0.0 -0.056075   \n",
       "\n",
       "            max     range  abundance          sample ATP11C_V972M  \n",
       "0      0.014688  0.014688   0.062771      Allele0_C0            0  \n",
       "1      0.014688  0.014688   0.062771      Allele0_C0            0  \n",
       "2      0.014688  0.014688  21.663839      Allele0_C0            0  \n",
       "3      0.014688  0.014688  21.663839      Allele0_C0            0  \n",
       "4      1.296063  1.296063        NaN      Allele0_C0            0  \n",
       "...         ...       ...        ...             ...          ...  \n",
       "63569  0.001258  0.001258   1.137815  Median_Allele2       Median  \n",
       "63570  0.000000  0.000000   0.000000  Median_Allele2       Median  \n",
       "63571  0.056075  0.112151   0.239638  Median_Allele2       Median  \n",
       "63572  0.056075  0.112151        NaN  Median_Allele2       Median  \n",
       "63573  0.056075  0.112151        NaN  Median_Allele2       Median  \n",
       "\n",
       "[63574 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_all[\"sample\"] = df_data_all[\"model\"].apply(lambda x: x.replace(f\"{pcmodel.id}_\", \"\"))\n",
    "df_data_all[genotype] = df_data_all[\"sample\"].apply(lambda x: x.split(\"_\", 1)[0].replace(\"Allele\", \"\"))\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3902ff15-bf6a-43a9-ad82-d81742231410",
   "metadata": {},
   "source": [
    "## Compute statistically significant results between groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "76ce7d4c-834d-4b2f-b705-059839846fea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>sample</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>1SGTH2OHE1ABCte</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>1SGTH2OHE2ABCte</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1QN1GST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1QN4GST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>23E1SQOX</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele0_C0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53531</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>XPPT</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>1.189980</td>\n",
       "      <td>Allele2_S417_D42</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53532</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>XYLK</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>Allele2_S417_D42</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53533</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>XYLTD_Dx</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.065284</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.130568</td>\n",
       "      <td>0.278992</td>\n",
       "      <td>Allele2_S417_D42</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53534</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>XYLTt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.065284</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.130568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele2_S417_D42</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53535</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>XYLU_Dt</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.071349</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.136633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Allele2_S417_D42</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53536 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             model        reactions  optimum       min  \\\n",
       "0            RBC_GEM_PC_Allele0_C0  1SGTH2OHE1ABCte      0.0  0.000000   \n",
       "1            RBC_GEM_PC_Allele0_C0  1SGTH2OHE2ABCte      0.0  0.000000   \n",
       "2            RBC_GEM_PC_Allele0_C0       23E1QN1GST      0.0  0.000000   \n",
       "3            RBC_GEM_PC_Allele0_C0       23E1QN4GST      0.0  0.000000   \n",
       "4            RBC_GEM_PC_Allele0_C0         23E1SQOX      0.0  0.000000   \n",
       "...                            ...              ...      ...       ...   \n",
       "53531  RBC_GEM_PC_Allele2_S417_D42             XPPT      0.0  0.000000   \n",
       "53532  RBC_GEM_PC_Allele2_S417_D42             XYLK      0.0  0.000000   \n",
       "53533  RBC_GEM_PC_Allele2_S417_D42         XYLTD_Dx      0.0 -0.065284   \n",
       "53534  RBC_GEM_PC_Allele2_S417_D42            XYLTt      0.0 -0.065284   \n",
       "53535  RBC_GEM_PC_Allele2_S417_D42          XYLU_Dt      0.0 -0.071349   \n",
       "\n",
       "            max     range  abundance            sample ATP11C_V972M  \n",
       "0      0.014688  0.014688   0.062771        Allele0_C0            0  \n",
       "1      0.014688  0.014688   0.062771        Allele0_C0            0  \n",
       "2      0.014688  0.014688  21.663839        Allele0_C0            0  \n",
       "3      0.014688  0.014688  21.663839        Allele0_C0            0  \n",
       "4      1.296063  1.296063        NaN        Allele0_C0            0  \n",
       "...         ...       ...        ...               ...          ...  \n",
       "53531  0.000715  0.000715   1.189980  Allele2_S417_D42            2  \n",
       "53532  0.006065  0.006065   0.025918  Allele2_S417_D42            2  \n",
       "53533  0.065284  0.130568   0.278992  Allele2_S417_D42            2  \n",
       "53534  0.065284  0.130568        NaN  Allele2_S417_D42            2  \n",
       "53535  0.065284  0.136633        NaN  Allele2_S417_D42            2  \n",
       "\n",
       "[53536 rows x 9 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_for_analyses = df_data_all[[not bool(operation_re.search(x)) for x in df_data_all[\"model\"]]].reset_index(drop=True)\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7f4389-594f-4c01-8743-94cbef5ffe54",
   "metadata": {},
   "source": [
    "### Create groups of models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cf91c097-af01-425f-a72e-9c3a5df6450b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_key = grouped_data_key\n",
    "id_key = \"sample\"\n",
    "model_groups = {all_key: list(df_data_for_analyses[id_key].unique())}\n",
    "\n",
    "def create_group_of_models(df, id_key, groupby, verbose=False):\n",
    "    grouped = df.groupby(groupby)[id_key].agg(lambda x: list(x.unique()))\n",
    "    grouped = {\"_\".join([str(x) for x in ensure_iterable(k)]): v for k, v in grouped.to_dict().items()}\n",
    "    if verbose:\n",
    "        max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "        for group_name, model_list in grouped.items():\n",
    "            spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "            print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "    return grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c290b56e-0285-4f7a-84b1-0c6376dc545d",
   "metadata": {},
   "source": [
    "#### Based on allele count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bd8b83f1-abda-4ade-b5bd-098b6885dd13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATP11C_V972M_0:\t18 samples\n",
      "ATP11C_V972M_1:\t8 samples\n",
      "ATP11C_V972M_2:\t6 samples\n",
      "\n"
     ]
    }
   ],
   "source": [
    "grouped = create_group_of_models(df_data_for_analyses, id_key=id_key, groupby=genotype, verbose=False)\n",
    "grouped = {f\"{genotype}_{str(k)}\": v for k, v in grouped.items()}\n",
    "if verbose:\n",
    "    max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "    for group_name, model_list in grouped.items():\n",
    "        spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "        print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "model_groups.update(grouped)\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f798fa7e-2a9a-4664-afbe-b30fd94b1df0",
   "metadata": {},
   "source": [
    "### View groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a1152679-d912-4768-802c-5d365e05c481",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Possible groups for analyses\n",
      "============================\n",
      "ATP11C_V972M_Sample:\t32 samples\n",
      "ATP11C_V972M_0:     \t18 samples\n",
      "ATP11C_V972M_1:     \t8 samples\n",
      "ATP11C_V972M_2:     \t6 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1SGTH2OHE1ABCte</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1SGTH2OHE2ABCte</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23E1QN1GST</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23E1QN4GST</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23E1SQOX</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPPT</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>1.189980</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLK</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLTD_Dx</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.065284</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.130568</td>\n",
       "      <td>0.278992</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLTt</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.065284</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.130568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLU_Dt</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.071349</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.136633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53536 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        model  optimum  \\\n",
       "reactions       sample                                                   \n",
       "1SGTH2OHE1ABCte Allele0_C0              RBC_GEM_PC_Allele0_C0      0.0   \n",
       "1SGTH2OHE2ABCte Allele0_C0              RBC_GEM_PC_Allele0_C0      0.0   \n",
       "23E1QN1GST      Allele0_C0              RBC_GEM_PC_Allele0_C0      0.0   \n",
       "23E1QN4GST      Allele0_C0              RBC_GEM_PC_Allele0_C0      0.0   \n",
       "23E1SQOX        Allele0_C0              RBC_GEM_PC_Allele0_C0      0.0   \n",
       "...                                                       ...      ...   \n",
       "XPPT            Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42      0.0   \n",
       "XYLK            Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42      0.0   \n",
       "XYLTD_Dx        Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42      0.0   \n",
       "XYLTt           Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42      0.0   \n",
       "XYLU_Dt         Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42      0.0   \n",
       "\n",
       "                                       min       max     range  abundance  \\\n",
       "reactions       sample                                                      \n",
       "1SGTH2OHE1ABCte Allele0_C0        0.000000  0.014688  0.014688   0.062771   \n",
       "1SGTH2OHE2ABCte Allele0_C0        0.000000  0.014688  0.014688   0.062771   \n",
       "23E1QN1GST      Allele0_C0        0.000000  0.014688  0.014688  21.663839   \n",
       "23E1QN4GST      Allele0_C0        0.000000  0.014688  0.014688  21.663839   \n",
       "23E1SQOX        Allele0_C0        0.000000  1.296063  1.296063        NaN   \n",
       "...                                    ...       ...       ...        ...   \n",
       "XPPT            Allele2_S417_D42  0.000000  0.000715  0.000715   1.189980   \n",
       "XYLK            Allele2_S417_D42  0.000000  0.006065  0.006065   0.025918   \n",
       "XYLTD_Dx        Allele2_S417_D42 -0.065284  0.065284  0.130568   0.278992   \n",
       "XYLTt           Allele2_S417_D42 -0.065284  0.065284  0.130568        NaN   \n",
       "XYLU_Dt         Allele2_S417_D42 -0.071349  0.065284  0.136633        NaN   \n",
       "\n",
       "                                 ATP11C_V972M  \n",
       "reactions       sample                         \n",
       "1SGTH2OHE1ABCte Allele0_C0                  0  \n",
       "1SGTH2OHE2ABCte Allele0_C0                  0  \n",
       "23E1QN1GST      Allele0_C0                  0  \n",
       "23E1QN4GST      Allele0_C0                  0  \n",
       "23E1SQOX        Allele0_C0                  0  \n",
       "...                                       ...  \n",
       "XPPT            Allele2_S417_D42            2  \n",
       "XYLK            Allele2_S417_D42            2  \n",
       "XYLTD_Dx        Allele2_S417_D42            2  \n",
       "XYLTt           Allele2_S417_D42            2  \n",
       "XYLU_Dt         Allele2_S417_D42            2  \n",
       "\n",
       "[53536 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Possible groups for analyses\\n============================\")\n",
    "max_name_len = max([len(group_name) for group_name in list(model_groups)])\n",
    "for group_name, model_list in model_groups.items():\n",
    "    spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "    print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "\n",
    "df_data_for_analyses = df_data_for_analyses.set_index([\"reactions\", id_key])\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2684b9-37c0-462b-98a1-5d20012b4e93",
   "metadata": {},
   "source": [
    "#### Ensure groups exist and setup directory structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e89c9396-815f-450e-a20c-d58be95ac66d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected directory structure\n",
      "============================\n",
      "ATP11C_V972M_Sample\n",
      "└── correlations\n"
     ]
    }
   ],
   "source": [
    "group_results_dirpath_dict = {all_key: corr_results_dirpath}\n",
    "header = \"Expected directory structure\"\n",
    "print(\"\\n\".join((header, \"=\" * len(header), all_key)))\n",
    "print(u\"\\u2514\\u2500\\u2500\" + f\" {group_results_dirpath_dict[all_key].name}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8fc2e6-5be8-4ac8-98ad-92d69335d316",
   "metadata": {},
   "source": [
    "#### Load subsystems and metabolic categories to enrich results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f7fff7e1-e706-4abf-adc9-6ee5a72bfd5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subsystem</th>\n",
       "      <th>category</th>\n",
       "      <th>kegg.pathway.name</th>\n",
       "      <th>kegg.pathway</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>hsa00250</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>hsa00330</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>hsa00270</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>hsa00260</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>hsa00340</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>Other</td>\n",
       "      <td>Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>has00970</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Representative subsystem for all transport rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>hsa00983</td>\n",
       "      <td>Subnetwork of KEGG pathway \"Drug metabolism - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Azathioprine and 6-mercaptopurine metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>hsa00983</td>\n",
       "      <td>Subnetwork of KEGG pathway \"Drug metabolism - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Ribavirin metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Curated category, from reactome R-HSA-9755088,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subsystem               category  \\\n",
       "0    Alanine, aspartate and glutamate metabolism  Amino acid metabolism   \n",
       "1                Arginine and proline metabolism  Amino acid metabolism   \n",
       "2             Cysteine and methionine metabolism  Amino acid metabolism   \n",
       "3       Glycine, serine and threonine metabolism  Amino acid metabolism   \n",
       "4                           Histidine metabolism  Amino acid metabolism   \n",
       "..                                           ...                    ...   \n",
       "73                   Aminoacyl-tRNA biosynthesis                  Other   \n",
       "74                      Transport, extracellular    Transport reactions   \n",
       "75                     5-fluorouracil metabolism                  Other   \n",
       "76  Azathioprine and 6-mercaptopurine metabolism                  Other   \n",
       "77                          Ribavirin metabolism                  Other   \n",
       "\n",
       "                              kegg.pathway.name kegg.pathway  \\\n",
       "0   Alanine, aspartate and glutamate metabolism     hsa00250   \n",
       "1               Arginine and proline metabolism     hsa00330   \n",
       "2            Cysteine and methionine metabolism     hsa00270   \n",
       "3      Glycine, serine and threonine metabolism     hsa00260   \n",
       "4                          Histidine metabolism     hsa00340   \n",
       "..                                          ...          ...   \n",
       "73                  Aminoacyl-tRNA biosynthesis     has00970   \n",
       "74                                                             \n",
       "75              Drug metabolism - other enzymes     hsa00983   \n",
       "76              Drug metabolism - other enzymes     hsa00983   \n",
       "77                                                             \n",
       "\n",
       "                                                notes  \n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "..                                                ...  \n",
       "73                                                     \n",
       "74  Representative subsystem for all transport rea...  \n",
       "75  Subnetwork of KEGG pathway \"Drug metabolism - ...  \n",
       "76  Subnetwork of KEGG pathway \"Drug metabolism - ...  \n",
       "77  Curated category, from reactome R-HSA-9755088,...  \n",
       "\n",
       "[77 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystems_to_exclude = {\"Pseudoreactions\"}\n",
    "use_abbrevs = True\n",
    "abbreviations = {\n",
    "    \"Amino acid metabolism\": \"A\",\n",
    "    \"Carbohydrate metabolism\": \"C\",\n",
    "    \"Lipid metabolism\": \"L\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"V\",\n",
    "    \"Nucleotide metabolism\": \"N\",\n",
    "    \"Reactive species\": \"R\",\n",
    "    \"Transport reactions\": \"T\",\n",
    "    \"Other\": \"O\",\n",
    "}\n",
    "categories_to_keep = list(abbreviations)\n",
    "\n",
    "df_pathways = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / \"subsystems.tsv\", sep=\"\\t\", dtype=str\n",
    ").fillna(\"\")\n",
    "\n",
    "# Rename \"name\" to subsystem to match reaction attribute\n",
    "df_pathways = df_pathways.rename({\"name\": \"subsystem\"}, axis=1)\n",
    "# Group \"Metabolism of other amino acids\" with amino acids rather than treat as \"other\"\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].replace(\n",
    "    \"Metabolism of other amino acids\", \"Amino acid metabolism\"\n",
    ")\n",
    "\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].apply(lambda x: (\"Other\" if x not in categories_to_keep else x))\n",
    "df_pathways = df_pathways[~df_pathways[\"subsystem\"].isin(subsystems_to_exclude)].copy()\n",
    "subsystem_to_category_dict = df_pathways.set_index(\"subsystem\")[\"category\"].to_dict()\n",
    "df_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1281f9fb-3124-4f6b-95c6-ccc02b3c6733",
   "metadata": {},
   "source": [
    "## Compute significant results between groups\n",
    "#### Compare all subgroups at once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f5bffd48-9cbd-41c7-9994-7128a2878d68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Groups to compare\n",
      "=================\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_1', 'ATP11C_V972M_2')\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_1')\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_2')\n",
      "('ATP11C_V972M_1', 'ATP11C_V972M_2')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1SGTH2OHE1ABCte</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1SGTH2OHE2ABCte</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.062771</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23E1QN1GST</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23E1QN4GST</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>0.014688</td>\n",
       "      <td>21.663839</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23E1SQOX</th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>RBC_GEM_PC_Allele0_C0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>1.296063</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XPPT</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>0.000715</td>\n",
       "      <td>1.189980</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLK</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.006065</td>\n",
       "      <td>0.025918</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLTD_Dx</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>-0.065284</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.130568</td>\n",
       "      <td>0.278992</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLTt</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>-0.065284</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.130568</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>XYLU_Dt</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>RBC_GEM_PC_Allele2_S417_D42</td>\n",
       "      <td>-0.071349</td>\n",
       "      <td>0.065284</td>\n",
       "      <td>0.136633</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53536 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        model       min  \\\n",
       "reactions       sample                                                    \n",
       "1SGTH2OHE1ABCte Allele0_C0              RBC_GEM_PC_Allele0_C0  0.000000   \n",
       "1SGTH2OHE2ABCte Allele0_C0              RBC_GEM_PC_Allele0_C0  0.000000   \n",
       "23E1QN1GST      Allele0_C0              RBC_GEM_PC_Allele0_C0  0.000000   \n",
       "23E1QN4GST      Allele0_C0              RBC_GEM_PC_Allele0_C0  0.000000   \n",
       "23E1SQOX        Allele0_C0              RBC_GEM_PC_Allele0_C0  0.000000   \n",
       "...                                                       ...       ...   \n",
       "XPPT            Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42  0.000000   \n",
       "XYLK            Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42  0.000000   \n",
       "XYLTD_Dx        Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42 -0.065284   \n",
       "XYLTt           Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42 -0.065284   \n",
       "XYLU_Dt         Allele2_S417_D42  RBC_GEM_PC_Allele2_S417_D42 -0.071349   \n",
       "\n",
       "                                       max     range  abundance ATP11C_V972M  \n",
       "reactions       sample                                                        \n",
       "1SGTH2OHE1ABCte Allele0_C0        0.014688  0.014688   0.062771            0  \n",
       "1SGTH2OHE2ABCte Allele0_C0        0.014688  0.014688   0.062771            0  \n",
       "23E1QN1GST      Allele0_C0        0.014688  0.014688  21.663839            0  \n",
       "23E1QN4GST      Allele0_C0        0.014688  0.014688  21.663839            0  \n",
       "23E1SQOX        Allele0_C0        1.296063  1.296063        NaN            0  \n",
       "...                                    ...       ...        ...          ...  \n",
       "XPPT            Allele2_S417_D42  0.000715  0.000715   1.189980            2  \n",
       "XYLK            Allele2_S417_D42  0.006065  0.006065   0.025918            2  \n",
       "XYLTD_Dx        Allele2_S417_D42  0.065284  0.130568   0.278992            2  \n",
       "XYLTt           Allele2_S417_D42  0.065284  0.130568        NaN            2  \n",
       "XYLU_Dt         Allele2_S417_D42  0.065284  0.136633        NaN            2  \n",
       "\n",
       "[53536 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "compare_group = genotype\n",
    "optimum = 0\n",
    "value_to_compare = \"range\"\n",
    "group_timepoints_by = None\n",
    "compare_pairwise = True\n",
    "compare_all_groups = True\n",
    "ordered_group_to_compare = [f\"{compare_group}_{alleles}\" for alleles in [0, 1, 2]]\n",
    "all_samples_for_comparison = [value for g in ordered_group_to_compare for value in np.array(model_groups[g])]\n",
    "df_data_for_correlations = df_data_for_analyses.loc[pd.IndexSlice[:, all_samples_for_comparison], :]\n",
    "df_data_for_correlations = df_data_for_correlations[df_data_for_correlations[\"optimum\"] == optimum].drop(\"optimum\", axis=1)\n",
    "\n",
    "print(\"Groups to compare\\n=================\")\n",
    "pairwise_group_combos = []\n",
    "if compare_all_groups:\n",
    "    print(tuple(ordered_group_to_compare))\n",
    "if compare_pairwise:\n",
    "    pairwise_group_combos += list(combinations(ordered_group_to_compare, 2))\n",
    "    for group in pairwise_group_combos:\n",
    "        print(group)\n",
    "df_data_for_correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bba2921-371b-4f2a-83ac-52eab0b6ae13",
   "metadata": {},
   "source": [
    "### Kruskal Wallis H-test (3 or more groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "e2e98727-a180-40bf-a3bc-0795a3c4ae10",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_dict = defaultdict(dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ba5742b8-10fe-42da-a85f-2df06d9dfc45",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if len(ordered_group_to_compare) > 2:\n",
    "    for rid in df_data_for_correlations.index.get_level_values(\"reactions\").unique():\n",
    "        df_data_rxn = df_data_for_correlations.loc[rid]\n",
    "        df_data_rxn_opt_value = df_data_rxn[value_to_compare].copy()\n",
    "        data_arrays = {\n",
    "            group_name: df_data_rxn_opt_value.loc[model_groups[group_name]].values \n",
    "            for group_name in ordered_group_to_compare\n",
    "        }\n",
    "        values = list(data_arrays.values())\n",
    "        unique_values = set([v for value_list in values for v in value_list if not np.isnan(v)])\n",
    "        if len(unique_values) <= 1:\n",
    "            # Skip variables that do not have any differences\n",
    "            results_dict[tuple(ordered_group_to_compare)][rid] = dict(zip([\"statistic\", \"pvalue\"], [pd.NA, pd.NA]))\n",
    "        else:\n",
    "            result = kruskal(*values, nan_policy=\"omit\")\n",
    "            results_dict[tuple(ordered_group_to_compare)][rid] = {\n",
    "                attr: getattr(result, attr) \n",
    "                for attr in [\"statistic\", \"pvalue\"]\n",
    "            }\n",
    "\n",
    "dataframes = {key: pd.DataFrame.from_dict(values, orient=\"index\") for key, values in results_dict.items()}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d05f036-101b-4e2f-9301-e71c65e81f67",
   "metadata": {},
   "source": [
    "### Mann Whiteney U test (2 groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8136fe70-1abd-4b17-a351-c5ca42a969be",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(ordered_group_to_compare) == 2 or compare_pairwise:\n",
    "    for rid in df_data_for_correlations.index.get_level_values(\"reactions\").unique():\n",
    "        df_data_rxn = df_data_for_correlations.loc[rid]\n",
    "        df_data_rxn_opt_value = df_data_rxn[value_to_compare].copy()\n",
    "        data_arrays = {\n",
    "            group_name: df_data_rxn_opt_value.loc[model_groups[group_name]].values \n",
    "            for group_name in ordered_group_to_compare\n",
    "        }\n",
    "        combos = [tuple(ordered_group_to_compare)] if (len(ordered_group_to_compare) == 2) else pairwise_group_combos\n",
    "        for combo in combos:\n",
    "            values = [data_arrays[group] for group in combo]\n",
    "            unique_values = set([v for value_list in values for v in value_list if not np.isnan(v)])\n",
    "            if len(unique_values) <= 1:\n",
    "                # Skip variables that do not have any differences\n",
    "                result = dict(zip([\"statistic\", \"pvalue\"], [pd.NA, pd.NA]))\n",
    "            else:\n",
    "                result = mannwhitneyu(*values, nan_policy=\"omit\")\n",
    "                result = {\n",
    "                    attr: getattr(result, attr) \n",
    "                    for attr in [\"statistic\", \"pvalue\"]\n",
    "                }\n",
    "            results_dict[combo][rid] = result\n",
    "dataframes = {key: pd.DataFrame.from_dict(values, orient=\"index\") for key, values in results_dict.items()}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c5a0b8c-3e3f-440e-94fc-a3d5387615fe",
   "metadata": {},
   "source": [
    "### View comparisons made"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7345d02a-b866-4834-98a2-f20db832d1f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different comparisons made: 4\n",
      "Groups compared\n",
      "===============\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_1', 'ATP11C_V972M_2')\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_1')\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_2')\n",
      "('ATP11C_V972M_1', 'ATP11C_V972M_2')\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of different comparisons made: {len(dataframes)}\")\n",
    "print(\"Groups compared\\n===============\")\n",
    "for key in list(dataframes):\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24818e90-9bcd-47ee-ae9b-179aee5b3910",
   "metadata": {},
   "source": [
    "### Create ordered metadata for visualization of clustered samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "af490bc1-3f9b-474d-a731-56a420d05ab0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Allele0_C0</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C1</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C2</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C3</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C4</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C5</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C6</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C7</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C8</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C9</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C10</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C11</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C12</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C13</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C14</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C15</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C16</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele0_C17</th>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C18</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C19</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C20</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C21</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C22</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C23</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C24</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele1_C25</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele2_S379_D10</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele2_S379_D23</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele2_S379_D42</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele2_S417_D10</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele2_S417_D23</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ATP11C_V972M\n",
       "sample                        \n",
       "Allele0_C0                   0\n",
       "Allele0_C1                   0\n",
       "Allele0_C2                   0\n",
       "Allele0_C3                   0\n",
       "Allele0_C4                   0\n",
       "Allele0_C5                   0\n",
       "Allele0_C6                   0\n",
       "Allele0_C7                   0\n",
       "Allele0_C8                   0\n",
       "Allele0_C9                   0\n",
       "Allele0_C10                  0\n",
       "Allele0_C11                  0\n",
       "Allele0_C12                  0\n",
       "Allele0_C13                  0\n",
       "Allele0_C14                  0\n",
       "Allele0_C15                  0\n",
       "Allele0_C16                  0\n",
       "Allele0_C17                  0\n",
       "Allele1_C18                  1\n",
       "Allele1_C19                  1\n",
       "Allele1_C20                  1\n",
       "Allele1_C21                  1\n",
       "Allele1_C22                  1\n",
       "Allele1_C23                  1\n",
       "Allele1_C24                  1\n",
       "Allele1_C25                  1\n",
       "Allele2_S379_D10             2\n",
       "Allele2_S379_D23             2\n",
       "Allele2_S379_D42             2\n",
       "Allele2_S417_D10             2\n",
       "Allele2_S417_D23             2\n",
       "Allele2_S417_D42             2"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_metadata = df_data_for_analyses.reset_index(drop=False)[[\"sample\", \"ATP11C_V972M\"]].drop_duplicates()\n",
    "df_metadata = df_metadata[~df_metadata[\"sample\"].apply(lambda x: bool(operation_re.search(x)))]\n",
    "df_metadata[\"ATP11C_V972M\"] = df_metadata[\"ATP11C_V972M\"].astype(int)\n",
    "\n",
    "to_concat = []\n",
    "for allele_count in df_metadata[\"ATP11C_V972M\"].unique():\n",
    "    df = df_metadata[df_metadata[\"ATP11C_V972M\"] == allele_count]\n",
    "    df.loc[:, \"sample\"] = df[\"sample\"].apply(lambda x: (allele_count, x.split(\"_\", maxsplit=1)[-1] if \"_C\" not in x else int(x.split(\"_C\")[-1])) )\n",
    "    to_concat += [df]\n",
    "df_metadata = pd.concat(to_concat, axis=0).sort_values(by=[\"ATP11C_V972M\", \"sample\"])\n",
    "# Reformat as original clustered sample IDs\n",
    "df_metadata[\"sample\"] = df_metadata[\"sample\"].apply(lambda x: \"Allele{}_C{}\".format(*x) if isinstance(x[1], int) else \"Allele{}_{}\".format(*x))\n",
    "df_metadata = df_metadata.set_index(\"sample\")\n",
    "df_metadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a16a3b4-ff30-4ede-a251-5ef488ed13ab",
   "metadata": {},
   "source": [
    "### Determine significance using p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0864a7a5-d2c0-4b80-addf-830f9f36a424",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0024778499408964127\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_1', 'ATP11C_V972M_2')\n",
      "Min & Max values: (-3.2095, 4.9987)\n",
      "\n",
      "0.9783104767086463\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_1')\n",
      "Min & Max values: (-2.5928, 2.4905)\n",
      "\n",
      "0.000357291836140036\n",
      "('ATP11C_V972M_0', 'ATP11C_V972M_2')\n",
      "Min & Max values: (-3.0402, 3.1123)\n",
      "\n",
      "0.026472236641991963\n",
      "('ATP11C_V972M_1', 'ATP11C_V972M_2')\n",
      "Min & Max values: (-2.2483, 3.0754)\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele0_C0</th>\n",
       "      <th>Allele0_C1</th>\n",
       "      <th>Allele0_C10</th>\n",
       "      <th>Allele0_C11</th>\n",
       "      <th>Allele0_C12</th>\n",
       "      <th>Allele0_C13</th>\n",
       "      <th>Allele0_C14</th>\n",
       "      <th>Allele0_C15</th>\n",
       "      <th>Allele0_C16</th>\n",
       "      <th>Allele0_C17</th>\n",
       "      <th>...</th>\n",
       "      <th>Allele2_S379_D42</th>\n",
       "      <th>Allele2_S417_D10</th>\n",
       "      <th>Allele2_S417_D23</th>\n",
       "      <th>Allele2_S417_D42</th>\n",
       "      <th>name</th>\n",
       "      <th>stoichiometry</th>\n",
       "      <th>proteins</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PROTSPP</th>\n",
       "      <td>-0.144614</td>\n",
       "      <td>-0.929996</td>\n",
       "      <td>-0.736722</td>\n",
       "      <td>0.108218</td>\n",
       "      <td>-1.304976</td>\n",
       "      <td>-0.760306</td>\n",
       "      <td>-0.360157</td>\n",
       "      <td>-0.514591</td>\n",
       "      <td>-1.023557</td>\n",
       "      <td>-0.292863</td>\n",
       "      <td>...</td>\n",
       "      <td>1.280868</td>\n",
       "      <td>1.801966</td>\n",
       "      <td>1.798204</td>\n",
       "      <td>1.699028</td>\n",
       "      <td>O-phospho-L-seryl-[protein] phosphatase</td>\n",
       "      <td>H2O + O-phospho-L-seryl-[protein] --&gt; Orthopho...</td>\n",
       "      <td>CPPED1;DUSP23;PPP1CA;PPP1CB;PPP1CC;PPP2CA;PPP2...</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Protein modification</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROTTPP</th>\n",
       "      <td>-0.144614</td>\n",
       "      <td>-0.929996</td>\n",
       "      <td>-0.736722</td>\n",
       "      <td>0.108218</td>\n",
       "      <td>-1.304976</td>\n",
       "      <td>-0.760306</td>\n",
       "      <td>-0.360157</td>\n",
       "      <td>-0.514591</td>\n",
       "      <td>-1.023557</td>\n",
       "      <td>-0.292863</td>\n",
       "      <td>...</td>\n",
       "      <td>1.280868</td>\n",
       "      <td>1.801966</td>\n",
       "      <td>1.798204</td>\n",
       "      <td>1.699028</td>\n",
       "      <td>O-phospho-L-threonyl-[protein] phosphatase</td>\n",
       "      <td>H2O + O-phospho-L-threonyl-[protein] --&gt; Ortho...</td>\n",
       "      <td>CPPED1;DUSP23;PPP1CA;PPP1CB;PPP1CC;PPP2CA;PPP2...</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Protein modification</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GALK</th>\n",
       "      <td>0.960666</td>\n",
       "      <td>0.754957</td>\n",
       "      <td>0.203758</td>\n",
       "      <td>0.145185</td>\n",
       "      <td>0.481374</td>\n",
       "      <td>-0.453299</td>\n",
       "      <td>1.220174</td>\n",
       "      <td>0.461633</td>\n",
       "      <td>-0.026970</td>\n",
       "      <td>-0.336114</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.775643</td>\n",
       "      <td>-3.040196</td>\n",
       "      <td>-0.616288</td>\n",
       "      <td>-0.824819</td>\n",
       "      <td>Galactokinase</td>\n",
       "      <td>ATP + galactose --&gt; ADP + Alpha-D-Galactose 1-...</td>\n",
       "      <td>GALK1;GALK2</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAL1PP</th>\n",
       "      <td>0.496131</td>\n",
       "      <td>0.716346</td>\n",
       "      <td>0.071459</td>\n",
       "      <td>-0.543827</td>\n",
       "      <td>0.574790</td>\n",
       "      <td>1.354124</td>\n",
       "      <td>0.995176</td>\n",
       "      <td>0.076097</td>\n",
       "      <td>0.753239</td>\n",
       "      <td>-0.546921</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.172473</td>\n",
       "      <td>-1.396024</td>\n",
       "      <td>-0.687117</td>\n",
       "      <td>-1.147408</td>\n",
       "      <td>alpha-D-galactose 1-phosphatase</td>\n",
       "      <td>Alpha-D-Galactose 1-phosphate + H2O --&gt; galact...</td>\n",
       "      <td>IMPA1;IMPA2</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SER_Cltex</th>\n",
       "      <td>0.728571</td>\n",
       "      <td>0.607369</td>\n",
       "      <td>-0.080595</td>\n",
       "      <td>1.297377</td>\n",
       "      <td>0.462513</td>\n",
       "      <td>0.247351</td>\n",
       "      <td>1.034612</td>\n",
       "      <td>0.085226</td>\n",
       "      <td>-0.255215</td>\n",
       "      <td>0.171550</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.162790</td>\n",
       "      <td>-1.365677</td>\n",
       "      <td>-1.785034</td>\n",
       "      <td>-2.222901</td>\n",
       "      <td>Serine transport via chloride antiport</td>\n",
       "      <td>Chloride + L-serine (extracellular) &lt;=&gt; Chlori...</td>\n",
       "      <td>SLC4A1</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELGTHGTHR</th>\n",
       "      <td>1.464823</td>\n",
       "      <td>1.125985</td>\n",
       "      <td>0.046722</td>\n",
       "      <td>-1.050288</td>\n",
       "      <td>-0.405483</td>\n",
       "      <td>0.779571</td>\n",
       "      <td>1.042202</td>\n",
       "      <td>-1.076011</td>\n",
       "      <td>-0.240062</td>\n",
       "      <td>-0.223394</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011250</td>\n",
       "      <td>-0.062291</td>\n",
       "      <td>-1.890291</td>\n",
       "      <td>-1.773124</td>\n",
       "      <td>Hydrogen selenide:glutathione oxidoreductase</td>\n",
       "      <td>Reduced glutathione + Glutathioselenol --&gt; Oxi...</td>\n",
       "      <td>GPX1</td>\n",
       "      <td>0.00186</td>\n",
       "      <td>Selenocompound metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SELDIGTHOXy</th>\n",
       "      <td>1.464823</td>\n",
       "      <td>1.125985</td>\n",
       "      <td>0.046722</td>\n",
       "      <td>-1.050288</td>\n",
       "      <td>-0.405483</td>\n",
       "      <td>0.779571</td>\n",
       "      <td>1.042202</td>\n",
       "      <td>-1.076011</td>\n",
       "      <td>-0.240062</td>\n",
       "      <td>-0.223394</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.011250</td>\n",
       "      <td>-0.062291</td>\n",
       "      <td>-1.890291</td>\n",
       "      <td>-1.773124</td>\n",
       "      <td>Glutathioselenol:NADP oxidoreductase</td>\n",
       "      <td>H+ + Nicotinamide adenine dinucleotide phospha...</td>\n",
       "      <td>GPX1</td>\n",
       "      <td>0.00186</td>\n",
       "      <td>Selenocompound metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPM</th>\n",
       "      <td>1.329435</td>\n",
       "      <td>0.861124</td>\n",
       "      <td>-1.025625</td>\n",
       "      <td>0.711761</td>\n",
       "      <td>0.634871</td>\n",
       "      <td>0.312986</td>\n",
       "      <td>0.888676</td>\n",
       "      <td>-0.544850</td>\n",
       "      <td>-0.521783</td>\n",
       "      <td>-0.923030</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.768069</td>\n",
       "      <td>0.385402</td>\n",
       "      <td>-0.954955</td>\n",
       "      <td>-1.154751</td>\n",
       "      <td>Phosphopentomutase</td>\n",
       "      <td>Alpha-D-Ribose 1-phosphate &lt;=&gt; Alpha-D-Ribose ...</td>\n",
       "      <td>PGM2</td>\n",
       "      <td>0.00186</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGSA2</th>\n",
       "      <td>1.467234</td>\n",
       "      <td>-1.064964</td>\n",
       "      <td>-0.795996</td>\n",
       "      <td>1.360563</td>\n",
       "      <td>0.668676</td>\n",
       "      <td>1.329622</td>\n",
       "      <td>0.556424</td>\n",
       "      <td>0.702329</td>\n",
       "      <td>-0.164324</td>\n",
       "      <td>-0.466903</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.530425</td>\n",
       "      <td>-0.294791</td>\n",
       "      <td>-0.736715</td>\n",
       "      <td>-0.396364</td>\n",
       "      <td>Erroneous formation of methylgloxal (glycerald...</td>\n",
       "      <td>Glyceraldehyde 3-phosphate --&gt; Methylglyoxal +...</td>\n",
       "      <td></td>\n",
       "      <td>0.00186</td>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGSA</th>\n",
       "      <td>1.467234</td>\n",
       "      <td>-1.064964</td>\n",
       "      <td>-0.795996</td>\n",
       "      <td>1.360563</td>\n",
       "      <td>0.668676</td>\n",
       "      <td>1.329622</td>\n",
       "      <td>0.556424</td>\n",
       "      <td>0.702329</td>\n",
       "      <td>-0.164324</td>\n",
       "      <td>-0.466903</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.530425</td>\n",
       "      <td>-0.294791</td>\n",
       "      <td>-0.736715</td>\n",
       "      <td>-0.396364</td>\n",
       "      <td>Erroneous formation of methylgloxal (dihydroxy...</td>\n",
       "      <td>Dihydroxyacetone phosphate --&gt; Methylglyoxal +...</td>\n",
       "      <td></td>\n",
       "      <td>0.00186</td>\n",
       "      <td>Pyruvate metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>150 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Allele0_C0  Allele0_C1  Allele0_C10  Allele0_C11  Allele0_C12  \\\n",
       "reactions                                                                    \n",
       "PROTSPP       -0.144614   -0.929996    -0.736722     0.108218    -1.304976   \n",
       "PROTTPP       -0.144614   -0.929996    -0.736722     0.108218    -1.304976   \n",
       "GALK           0.960666    0.754957     0.203758     0.145185     0.481374   \n",
       "GAL1PP         0.496131    0.716346     0.071459    -0.543827     0.574790   \n",
       "SER_Cltex      0.728571    0.607369    -0.080595     1.297377     0.462513   \n",
       "...                 ...         ...          ...          ...          ...   \n",
       "SELGTHGTHR     1.464823    1.125985     0.046722    -1.050288    -0.405483   \n",
       "SELDIGTHOXy    1.464823    1.125985     0.046722    -1.050288    -0.405483   \n",
       "PPM            1.329435    0.861124    -1.025625     0.711761     0.634871   \n",
       "MGSA2          1.467234   -1.064964    -0.795996     1.360563     0.668676   \n",
       "MGSA           1.467234   -1.064964    -0.795996     1.360563     0.668676   \n",
       "\n",
       "             Allele0_C13  Allele0_C14  Allele0_C15  Allele0_C16  Allele0_C17  \\\n",
       "reactions                                                                      \n",
       "PROTSPP        -0.760306    -0.360157    -0.514591    -1.023557    -0.292863   \n",
       "PROTTPP        -0.760306    -0.360157    -0.514591    -1.023557    -0.292863   \n",
       "GALK           -0.453299     1.220174     0.461633    -0.026970    -0.336114   \n",
       "GAL1PP          1.354124     0.995176     0.076097     0.753239    -0.546921   \n",
       "SER_Cltex       0.247351     1.034612     0.085226    -0.255215     0.171550   \n",
       "...                  ...          ...          ...          ...          ...   \n",
       "SELGTHGTHR      0.779571     1.042202    -1.076011    -0.240062    -0.223394   \n",
       "SELDIGTHOXy     0.779571     1.042202    -1.076011    -0.240062    -0.223394   \n",
       "PPM             0.312986     0.888676    -0.544850    -0.521783    -0.923030   \n",
       "MGSA2           1.329622     0.556424     0.702329    -0.164324    -0.466903   \n",
       "MGSA            1.329622     0.556424     0.702329    -0.164324    -0.466903   \n",
       "\n",
       "             ...  Allele2_S379_D42  Allele2_S417_D10  Allele2_S417_D23  \\\n",
       "reactions    ...                                                         \n",
       "PROTSPP      ...          1.280868          1.801966          1.798204   \n",
       "PROTTPP      ...          1.280868          1.801966          1.798204   \n",
       "GALK         ...         -0.775643         -3.040196         -0.616288   \n",
       "GAL1PP       ...         -1.172473         -1.396024         -0.687117   \n",
       "SER_Cltex    ...         -1.162790         -1.365677         -1.785034   \n",
       "...          ...               ...               ...               ...   \n",
       "SELGTHGTHR   ...         -0.011250         -0.062291         -1.890291   \n",
       "SELDIGTHOXy  ...         -0.011250         -0.062291         -1.890291   \n",
       "PPM          ...         -1.768069          0.385402         -0.954955   \n",
       "MGSA2        ...         -1.530425         -0.294791         -0.736715   \n",
       "MGSA         ...         -1.530425         -0.294791         -0.736715   \n",
       "\n",
       "             Allele2_S417_D42  \\\n",
       "reactions                       \n",
       "PROTSPP              1.699028   \n",
       "PROTTPP              1.699028   \n",
       "GALK                -0.824819   \n",
       "GAL1PP              -1.147408   \n",
       "SER_Cltex           -2.222901   \n",
       "...                       ...   \n",
       "SELGTHGTHR          -1.773124   \n",
       "SELDIGTHOXy         -1.773124   \n",
       "PPM                 -1.154751   \n",
       "MGSA2               -0.396364   \n",
       "MGSA                -0.396364   \n",
       "\n",
       "                                                          name  \\\n",
       "reactions                                                        \n",
       "PROTSPP                O-phospho-L-seryl-[protein] phosphatase   \n",
       "PROTTPP             O-phospho-L-threonyl-[protein] phosphatase   \n",
       "GALK                                             Galactokinase   \n",
       "GAL1PP                         alpha-D-galactose 1-phosphatase   \n",
       "SER_Cltex               Serine transport via chloride antiport   \n",
       "...                                                        ...   \n",
       "SELGTHGTHR        Hydrogen selenide:glutathione oxidoreductase   \n",
       "SELDIGTHOXy               Glutathioselenol:NADP oxidoreductase   \n",
       "PPM                                         Phosphopentomutase   \n",
       "MGSA2        Erroneous formation of methylgloxal (glycerald...   \n",
       "MGSA         Erroneous formation of methylgloxal (dihydroxy...   \n",
       "\n",
       "                                                 stoichiometry  \\\n",
       "reactions                                                        \n",
       "PROTSPP      H2O + O-phospho-L-seryl-[protein] --> Orthopho...   \n",
       "PROTTPP      H2O + O-phospho-L-threonyl-[protein] --> Ortho...   \n",
       "GALK         ATP + galactose --> ADP + Alpha-D-Galactose 1-...   \n",
       "GAL1PP       Alpha-D-Galactose 1-phosphate + H2O --> galact...   \n",
       "SER_Cltex    Chloride + L-serine (extracellular) <=> Chlori...   \n",
       "...                                                        ...   \n",
       "SELGTHGTHR   Reduced glutathione + Glutathioselenol --> Oxi...   \n",
       "SELDIGTHOXy  H+ + Nicotinamide adenine dinucleotide phospha...   \n",
       "PPM          Alpha-D-Ribose 1-phosphate <=> Alpha-D-Ribose ...   \n",
       "MGSA2        Glyceraldehyde 3-phosphate --> Methylglyoxal +...   \n",
       "MGSA         Dihydroxyacetone phosphate --> Methylglyoxal +...   \n",
       "\n",
       "                                                      proteins   pvalue  \\\n",
       "reactions                                                                 \n",
       "PROTSPP      CPPED1;DUSP23;PPP1CA;PPP1CB;PPP1CC;PPP2CA;PPP2...  0.00001   \n",
       "PROTTPP      CPPED1;DUSP23;PPP1CA;PPP1CB;PPP1CC;PPP2CA;PPP2...  0.00001   \n",
       "GALK                                               GALK1;GALK2  0.00001   \n",
       "GAL1PP                                             IMPA1;IMPA2  0.00001   \n",
       "SER_Cltex                                               SLC4A1  0.00001   \n",
       "...                                                        ...      ...   \n",
       "SELGTHGTHR                                                GPX1  0.00186   \n",
       "SELDIGTHOXy                                               GPX1  0.00186   \n",
       "PPM                                                       PGM2  0.00186   \n",
       "MGSA2                                                           0.00186   \n",
       "MGSA                                                            0.00186   \n",
       "\n",
       "                             subsystem                 category  \n",
       "reactions                                                        \n",
       "PROTSPP           Protein modification                    Other  \n",
       "PROTTPP           Protein modification                    Other  \n",
       "GALK              Galactose metabolism  Carbohydrate metabolism  \n",
       "GAL1PP            Galactose metabolism  Carbohydrate metabolism  \n",
       "SER_Cltex     Transport, extracellular      Transport reactions  \n",
       "...                                ...                      ...  \n",
       "SELGTHGTHR   Selenocompound metabolism    Amino acid metabolism  \n",
       "SELDIGTHOXy  Selenocompound metabolism    Amino acid metabolism  \n",
       "PPM          Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "MGSA2              Pyruvate metabolism  Carbohydrate metabolism  \n",
       "MGSA               Pyruvate metabolism  Carbohydrate metabolism  \n",
       "\n",
       "[150 rows x 30 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pvalue_sig = {\n",
    "    ('ATP11C_V972M_0', 'ATP11C_V972M_1', 'ATP11C_V972M_2'): 0.0025,\n",
    "    ('ATP11C_V972M_0', 'ATP11C_V972M_1'): 0.001,\n",
    "    ('ATP11C_V972M_0', 'ATP11C_V972M_2'): 0.002,\n",
    "    ('ATP11C_V972M_1', 'ATP11C_V972M_2'): 0.05,\n",
    "}\n",
    "\n",
    "enzyme_reactions_only = False\n",
    "include_boundary_reactions = False\n",
    "sort_by_subsystem = False\n",
    "standardize_by = \"mean\"\n",
    "use_group_means = False\n",
    "fdr_method = None\n",
    "\n",
    "significant_dataframes = {}\n",
    "for met in model.metabolites.query(lambda x: x.compartment == \"e\"):\n",
    "    met.name += \" (extracellular)\"\n",
    "metadata_columns = [\n",
    "    \"name\", \n",
    "    \"stoichiometry\", \n",
    "    \"proteins\", \n",
    "    \"pvalue\" if not fdr_method else \"adj_pvalue\",\n",
    "    \"subsystem\", \n",
    "    \"category\",\n",
    "]\n",
    "for key, df in dataframes.items():\n",
    "    df = df.dropna().copy()\n",
    "    df[\"pvalue\"] = df[\"pvalue\"].astype(float)\n",
    "    if fdr_method is not None and not fdr_method in {\"bon\", \"bh\", \"by\"}:\n",
    "        raise ValueError(f\"Unrecognized FDR correction method : {fdr_method}\")\n",
    "    elif fdr_method == \"bon\":\n",
    "        pvalue_key = \"adj_pvalue\"\n",
    "        df[pvalue_key] = df[\"pvalue\"] * len(df[\"pvalue\"])\n",
    "    elif fdr_method in {\"bh\", \"by\"}:\n",
    "        pvalue_key = \"adj_pvalue\"\n",
    "        df[pvalue_key] = false_discovery_control(df[\"pvalue\"].astype(float), method=fdr_method)\n",
    "    else:\n",
    "        pvalue_key = \"pvalue\"\n",
    "    print(df.loc[\"PSFLIPt\", \"pvalue\"])\n",
    "    pvalue = pvalue_sig if isinstance(pvalue_sig, (float, int)) else pvalue_sig[key]\n",
    "    df = df[df[pvalue_key] <= pvalue].drop(\"statistic\", axis=1)\n",
    "    if enzyme_reactions_only:\n",
    "        df_pivot = df_data_for_correlations.loc[df.index, [\"abundance\", value_to_compare]].dropna(subset=\"abundance\")\n",
    "        df_pivot = df_pivot.drop(\"abundance\", axis=1)\n",
    "    else:\n",
    "        df_pivot = df_data_for_correlations.loc[df.index, value_to_compare]\n",
    "    if not include_boundary_reactions:\n",
    "        df_pivot = df_pivot[~df_pivot.index.isin(model.reactions.query(lambda x: x.boundary).list_attr(\"id\"), level=\"reactions\")]\n",
    "    df_pivot = df_pivot.reset_index(drop=False)\n",
    "    df_pivot = df_pivot.pivot(columns=id_key, index=\"reactions\", values=value_to_compare)\n",
    "    df = pd.merge(df, df_pivot, left_index=True, right_index=True).sort_values(pvalue_key)\n",
    "    df.index.name = \"reactions\"\n",
    "    df = df.reset_index(drop=False).set_index([\"reactions\", pvalue_key]).T\n",
    "    if df.empty:\n",
    "        df = pd.DataFrame([], columns=metadata_columns)\n",
    "    else:    \n",
    "        df = pd.concat(\n",
    "            [\n",
    "                 # Sort index by donor number and subgroup while concatenating\n",
    "                df.loc[model_groups[g]].sort_index() \n",
    "                for g in key\n",
    "            ],\n",
    "            axis=0\n",
    "        )\n",
    "    \n",
    "        df = df.T.reset_index(drop=False)\n",
    "        # Enrich results\n",
    "        # df[\"name\"] = [r.name.replace(\",\", \"\") for r in model.reactions.get_by_any(list(df[\"reactions\"].values))]\n",
    "        df[\"name\"] = [r.name for r in model.reactions.get_by_any(list(df[\"reactions\"].values))]\n",
    "        df[\"stoichiometry\"] = [r.build_reaction_string(use_metabolite_names=True) for r in model.reactions.get_by_any(list(df[\"reactions\"].values))]\n",
    "        df[\"subsystem\"] = [r.subsystem for r in model.reactions.get_by_any(list(df[\"reactions\"].values))]\n",
    "        df[\"category\"] = df[\"subsystem\"].replace(subsystem_to_category_dict)\n",
    "        df[\"proteins\"] = [\";\".join(sorted([g.id for g in r.genes])) for r in model.reactions.get_by_any(list(df[\"reactions\"].values))]\n",
    "        # Replace commas to prevent issues with CSV export\n",
    "        # df[\"subsystem\"] = df[\"subsystem\"].apply(lambda x: x.replace(\",\", \"\"))\n",
    "        # df[\"category\"] = df[\"category\"].apply(lambda x: x.replace(\",\", \"\"))\n",
    "        df[pvalue_key] = df[pvalue_key].apply(lambda x: round(x, 5))\n",
    "    \n",
    "        df = df.set_index(\"reactions\")\n",
    "        if sort_by_subsystem:\n",
    "            df = df.sort_values(by=[\"category\", \"subsystem\", \"proteins\"])\n",
    "    \n",
    "        df_meta = df.loc[:, metadata_columns].copy()\n",
    "        df_data = df.loc[:, ~df.columns.isin(df_meta.columns)].copy()\n",
    "        if use_group_means:\n",
    "            df_data = pd.concat([df_data.loc[:, model_groups[g]].mean(axis=1) for g in key], axis=1)\n",
    "            df_data.columns = list(key)\n",
    "        if standardize_by == \"mean\":\n",
    "            df_data = df_data.sub(df_data.mean(axis=1), axis=0).div(df_data.std(axis=1), axis=0).dropna(how=\"all\", axis=0)\n",
    "        elif standardize_by == \"median\":\n",
    "            df_data = ((df_data.T - df_data.median(axis=1)) / (df_data.quantile(q=0.75, axis=1) - df_data.quantile(q=0.25, axis=1))).T\n",
    "        else:\n",
    "            df_data = df_data.loc[:, [x for x in df_metadata.index if x in df_data.columns]]\n",
    "        # Put dataframes back together for custom reordering\n",
    "        df = df_data.merge(df_meta, left_index=True, right_index=True)\n",
    "    significant_dataframes[key] = df\n",
    "    print(key)\n",
    "    print(f\"Min & Max values: ({df_data.min().min():.4f}, {df_data.max().max():.4f})\")\n",
    "    print()\n",
    "key = tuple(ordered_group_to_compare)\n",
    "key = ('ATP11C_V972M_0', 'ATP11C_V972M_2')\n",
    "df = significant_dataframes[key]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5464e18a-646c-436f-b672-af4e1841a91e",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = ('ATP11C_V972M_0', 'ATP11C_V972M_2')\n",
    "df = significant_dataframes[key].copy()\n",
    "\n",
    "df.loc[[\"56DH5FUt\", \"56DTHMt\", \"56DURAt\"], \"subsystem\"] = \"Nucleotide transport\"\n",
    "df.loc[[\"PAFH\", \"PAFS\"], \"subsystem\"] = \"Ether lipid metabolism\"\n",
    "df.loc[[\"D5FURADy\"], \"category\"] = \"Nucleotide metabolism\"\n",
    "df.loc[[\"CYSTL2\"], \"subsystem\"] = \"Cysteine and methionine metabolism\"\n",
    "df.loc[[\n",
    "    'ETHAt',\n",
    "    'FAt_16_2E',\n",
    "    \"SPHGNSte\", \n",
    "    \"SPHGNte\", \n",
    "    \"SPHS1PABCte\", \n",
    "    \"PEFLIPt\", \n",
    "    \"PSFLIPt\",\n",
    "    'DHET1112te',\n",
    "    'DHET1314te',\n",
    "    'DHET56te',\n",
    "    'DHET89te',\n",
    "    'DIHDPA1011te',\n",
    "    'DIHDPA1314te',\n",
    "    'DIHDPA1617te',\n",
    "    'DIHDPA1920te',\n",
    "    'DIHDPA78te',\n",
    "    'DIHETE1112te',\n",
    "    'DIHETE1718te',\n",
    "    'DIHETE56te',\n",
    "    'DIHETE78te',\n",
    "    'DIHETEt1314e',\n",
    "    'DIHOME910te',\n",
    "    'DIHOMEt1213e',\n",
    "], \"subsystem\" ] = \"Lipid transport\"\n",
    "df[\"subsystem\"] = df[\"subsystem\"].replace({\n",
    "    '5-fluorouracil metabolism': 'Nucleotide metabolism',\n",
    "    'Glycolysis / Gluconeogenesis': 'Glycolysis and pyruvate metabolism',\n",
    "    'Pyruvate metabolism': 'Glycolysis and pyruvate metabolism',\n",
    "    'Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism': 'Glycolysis and pyruvate metabolism',\n",
    "    'Purine metabolism': 'Nucleotide metabolism',\n",
    "    'Pyrimidine metabolism': 'Nucleotide metabolism',\n",
    "    'Transport, extracellular': 'Other transport'\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "df = df.sort_values([\"category\", \"subsystem\"])\n",
    "\n",
    "ordered_subsystems = [\n",
    "    'Alanine, aspartate and glutamate metabolism',\n",
    "    'Glycine, serine and threonine metabolism',\n",
    "    'Cysteine and methionine metabolism',\n",
    "    'Selenocompound metabolism',\n",
    "    'Amino sugar and nucleotide sugar metabolism',    \n",
    "    'Galactose metabolism',\n",
    "    'Glycolysis and pyruvate metabolism',\n",
    "    'Pentose and glucuronate interconversions',\n",
    "    'Pentose phosphate pathway',\n",
    "    'Nicotinate and nicotinamide metabolism',\n",
    "    'Pantothenate and CoA metabolism',\n",
    "    'Thiamine metabolism',\n",
    "    'Nucleotide metabolism',\n",
    "    'Protein modification',\n",
    "    'Linoleate metabolism',\n",
    "    'Arachidonic acid metabolism',\n",
    "    'Eicosapentaenoic acid metabolism',\n",
    "    'Docosahexaenoic acid metabolism',\n",
    "    'Ether lipid metabolism',\n",
    "    'Glycerophospholipid metabolism',\n",
    "    'Sphingolipid metabolism',\n",
    "    \"Lipid transport\",\n",
    "    \"Nucleotide transport\",\n",
    "    \"Other transport\",\n",
    "]\n",
    "df = pd.concat(\n",
    "    [\n",
    "        df[df[\"subsystem\"] == subsystem].sort_values(by=\"pvalue\")\n",
    "        for subsystem in ordered_subsystems\n",
    "    ],\n",
    "    axis=0,\n",
    ")\n",
    "significant_dataframes[key] = df.copy()\n",
    "set(df[\"subsystem\"]).difference(ordered_subsystems)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610dc29f-f123-457b-8d29-0d7461a189f7",
   "metadata": {},
   "source": [
    "## Export results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6cd72da-0f5b-47c3-bdf4-0cfdd3c54bd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"tsv\"\n",
    "for key, df_main in significant_dataframes.items(): \n",
    "    df_meta_row = df_main.loc[:, metadata_columns].copy()\n",
    "    df_data = df_main.loc[:, ~df_main.columns.isin(df_meta.columns)].copy()\n",
    "    df_meta_col = df_metadata.copy()\n",
    "    if use_group_means:\n",
    "        df_meta_col = df_meta_col.groupby(compare_group, as_index=False).mean()\n",
    "        df_meta_col.index = [f\"{genotype}_{x}\" for x in df_meta_col[\"ATP11C_V972M\"].values]\n",
    "    \n",
    "    df_meta_col = df_meta_col.loc[list(df_data.columns)].copy()\n",
    "    for df_type, df in zip([\"data\", \"meta_row\", \"meta_col\"], [df_data, df_meta_row, df_meta_col]):\n",
    "        filename = \"_\".join(\n",
    "            [\"MannWhiteney\" if len(key) == 2 else \"Kruskal\"]\n",
    "            + [g.split(\"_\")[-1] for g in key]\n",
    "            + [id_key]\n",
    "            + [df_type]\n",
    "        )\n",
    "        if use_group_means:\n",
    "            filename += \"_mean\"\n",
    "        group_results_dirpath_dict[grouped_data_key].mkdir(exist_ok=True)\n",
    "        filename = group_results_dirpath_dict[grouped_data_key] / filename\n",
    "        df.to_csv(\n",
    "            f\"{filename}.{ftype}\",\n",
    "            sep=\"\\t\" if ftype == \"tsv\" else ',',\n",
    "            index=True\n",
    "        )\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
