{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc9114c6-b2d1-48a1-8ba3-6686a1ea2449",
   "metadata": {},
   "source": [
    "# Visualize ranges of fluxes across samples - REDS Recall, ATP11C V972M\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d694e76-31c0-4aba-add6-a07a00a5d04c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    ensure_iterable,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    ")\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    DEFAULT_PREFIX_SUFFIX_VALUES,\n",
    "    DEFAULT_PROTEOME_COMPARTMENT,\n",
    "    EnzymeDilution,\n",
    "    add_relaxation_budget,\n",
    "    load_overlay_model,\n",
    ")\n",
    "from rbc_gem_utils.visualization import cmap_map\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "384c972a-9f2f-4b88-a65c-017d1a6561d6",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "418ddc9c-b48e-400a-b5a1-e4e6bfe73bd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "# Set bound defaults much larger to prevent model loading issues\n",
    "COBRA_CONFIGURATION.bounds = (-1e-8, 1e8)\n",
    "COBRA_CONFIGURATION.tolerance = 1e-9\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c3720a-d637-4115-8960-f48671ee725a",
   "metadata": {},
   "source": [
    "### Define organism, model, and dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5b08faa-240f-44ee-9d16-ddbdc556852f",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = \"Human\"\n",
    "model_id = \"RBC_GEM\"\n",
    "dataset_name = \"REDSRecall\"\n",
    "genotype = \"ATP11C_V972M\"\n",
    "grouped_data_key = f\"{genotype}_Sample\"\n",
    "grouped_data_key"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0ccb7e-c784-4b63-bfe9-7ebdce0e24b5",
   "metadata": {},
   "source": [
    "### Set variables for sample identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "664f7049-077b-42f1-be94-0d0cd511ec60",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_key = \"SAMPLE ID\"\n",
    "donor_key = \"PUBLIC RECALL DONOR ID\"\n",
    "\n",
    "# For sample IDs\n",
    "donor_re = re.compile(rf\"(?P<donor>S(?P<num>\\d\\d\\d))\")\n",
    "operations = \"|\".join([x.capitalize() for x in [\"mean\", \"median\"]])\n",
    "operation_re = re.compile(r\"(?P<op>\" + operations + r\")\\_(?P<group>\\w+)\")\n",
    "sample_id_re = re.compile(r\"(?!\" + operations + r\")\" + donor_re.pattern + r\"\\_\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e478c85d-b0ed-466c-b6d3-e8dead52f018",
   "metadata": {},
   "source": [
    "### Set computation options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52e59494-b23d-4775-8863-1baeb1f46194",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"xml\"  # In our experience, SBML/XML loads faster, but will take up to 4x more space uncompressed as compared to JSON\n",
    "run_computations = True  # Keep off to use previously computed results\n",
    "overwrite = False  # Whether to allow overwriting of previous simulation results\n",
    "verbose = True\n",
    "\n",
    "# Objective reactions\n",
    "objective_reactions = [\"NaKt\"]\n",
    "# Reactions that must have the capability to carry flux, sort for consistency\n",
    "required_flux_reactions = [\"PSFLIPt\"]  # Add reactions to this list\n",
    "required_flux_reactions = sorted(set(objective_reactions + required_flux_reactions))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9120af5-9f98-425a-80a8-80e1905e4380",
   "metadata": {},
   "source": [
    "#### Set prefixes/suffixes to expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4740d4e4-7534-4150-b366-a3afc551b68a",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.dilution\"]\n",
    "protein_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.metabolite\"]\n",
    "relaxation_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.relaxation\"]\n",
    "enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]\n",
    "enzyme_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.dilution\"]\n",
    "enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.metabolite\"]\n",
    "budget_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"budgets\"][\"prefix.dilution\"]\n",
    "budget_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"budgets\"][\"prefix.metabolite\"]\n",
    "comp_suffix = f\"_{DEFAULT_PROTEOME_COMPARTMENT}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "295bb5e1-67d6-4436-af4a-0f34befaa111",
   "metadata": {},
   "source": [
    "### Set figure options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ad3312-69a9-4bd9-92c4-a2d467bed985",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc2d514-cb81-4968-95b6-95e758cd3e73",
   "metadata": {},
   "source": [
    "### Set paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ccd462b-0c65-4b33-b745-6a4e74996f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set paths\n",
    "overlay_dirpath = get_dirpath(\"analysis\") / \"OVERLAY\" / organism\n",
    "model_dirpath = overlay_dirpath / model_id\n",
    "results_dirpath = (\n",
    "    get_dirpath(use_temp=\"processed\")\n",
    "    / model_id\n",
    "    / \"OVERLAY\"\n",
    "    / organism\n",
    "    / dataset_name\n",
    "    / grouped_data_key\n",
    ")\n",
    "pcfva_results_dirpath = (\n",
    "    results_dirpath\n",
    "    / \"pcFVA\"\n",
    "    / \"_\".join((\"REQ\", *required_flux_reactions))\n",
    "    / \"_\".join((\"OBJ\", *objective_reactions))\n",
    ")\n",
    "# Objective reaction does not matter since correlations are computed\n",
    "# based on min and max fluxes and abundance, which are obtained when optimum is 0.\n",
    "corr_results_dirpath = results_dirpath / \"correlations\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "806b8d0f-8ca2-48e4-9da1-7db464f01407",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "541f7b2c-ad0d-4404-899b-fa05a0bd7fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = read_cobra_model(filename=model_dirpath / f\"{model_id}.xml\")\n",
    "pcmodel = load_overlay_model(filename=model_dirpath / f\"{model_id}_PC.xml\")\n",
    "\n",
    "# Add relaxation budget to initial PC model to get names of relaxation reactions\n",
    "add_relaxation_budget(pcmodel, 0, verbose=False)\n",
    "pcmodel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49cf0307-bf28-4f10-8784-143d62c669a7",
   "metadata": {},
   "source": [
    "## Load pcFVA generated results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "537fccd8-52b6-44c9-83a0-af5c62b7fd2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load DataFrame of generated results\n",
    "df_pcfva_all = pd.read_csv(\n",
    "    pcfva_results_dirpath / f\"{pcmodel.id}_All_FVAsols.zip\",\n",
    "    index_col=None,\n",
    ").fillna(0)\n",
    "\n",
    "df_pcfva_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63a4524-c95b-4980-a95f-5ba109e75025",
   "metadata": {},
   "source": [
    "## Create DataFrame for visualization\n",
    "### Get maximum reaction fluxes and associated abundance values\n",
    "#### Get maximum reaction fluxes and flux ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73dbb8d3-7fc4-4168-84ad-57f84ee4824f",
   "metadata": {},
   "outputs": [],
   "source": [
    "rxns = model.reactions.list_attr(\"id\")\n",
    "df_max_flux_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "df_max_flux_per_model = df_max_flux_per_model.groupby(\n",
    "    [\"model\", \"reactions\", \"optimum\"]\n",
    ")[[\"min\", \"max\"]].agg(\n",
    "    {\n",
    "        \"min\": \"min\",  # Minimum reaction flux per model\n",
    "        \"max\": \"max\",  # Maximum reaction flux per model\n",
    "    }\n",
    ")\n",
    "# Address issues possibly caused by floating point precision, ideally a value that prevents any negative ranges\n",
    "df_max_flux_per_model.loc[\n",
    "    df_max_flux_per_model[\"max\"] < df_max_flux_per_model[\"min\"], [\"max\", \"min\"]\n",
    "] = [0, 0]\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_flux_per_model[\"max\"] = df_max_flux_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_flux_per_model[\"min\"] = df_max_flux_per_model[\"min\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_flux_per_model[\"range\"] = (\n",
    "    df_max_flux_per_model[\"max\"] - df_max_flux_per_model[\"min\"]\n",
    ")\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_flux_per_model[df_max_flux_per_model[\"range\"] < 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ea38c15-2a68-42a4-b9b8-6772a7a8fc2f",
   "metadata": {},
   "source": [
    "#### Get maximum \"enzyme\" abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e766d0df-120f-4968-9b8f-b610af420a7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "rxns = pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, EnzymeDilution)\n",
    "    and x.id.endswith(f\"{enzyme_met_suffix_total}{comp_suffix}\")\n",
    ").list_attr(\"id\")\n",
    "df_max_abundance_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "# Rename dilution reactions to match\n",
    "reaction_enzyme_map = {\n",
    "    enzyme_rid: enzyme_rid.replace(\n",
    "        f\"{enzyme_rxn_prefix}{enzyme_met_prefix}\", \"\"\n",
    "    ).replace(f\"{enzyme_met_suffix_total}{comp_suffix}\", \"\")\n",
    "    for enzyme_rid in df_max_abundance_per_model[\"reactions\"]\n",
    "}\n",
    "df_max_abundance_per_model[\"reactions\"] = df_max_abundance_per_model[\n",
    "    \"reactions\"\n",
    "].replace(reaction_enzyme_map)\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.groupby(\n",
    "    [\"model\", \"reactions\", \"optimum\"]\n",
    ")[[\"max\"]].max()\n",
    "# Address issues possibly caused by floating point precision, atol is ideally a value that prevents any negative ranges\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if x < 0 else x\n",
    ")\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.rename(\n",
    "    {\"max\": \"abundance\"}, axis=1\n",
    ")\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_abundance_per_model[(df_max_abundance_per_model < 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a1b613-8a09-4f72-8342-909fd2b02bff",
   "metadata": {},
   "source": [
    "#### Merge DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be484e01-700b-47b1-9b3e-8b037fcfd223",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_data_all = pd.merge(\n",
    "    df_max_flux_per_model,\n",
    "    df_max_abundance_per_model,\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    ")\n",
    "df_data_all = df_data_all.reset_index(drop=False)\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6afec28e-5b30-42a1-bee5-278ac78af25f",
   "metadata": {},
   "source": [
    "### Identify genotypes for results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e605cb3-eb57-416c-83dd-09de89657388",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_data_all[\"sample\"] = df_data_all[\"model\"].apply(\n",
    "    lambda x: x.replace(f\"{pcmodel.id}_\", \"\")\n",
    ")\n",
    "df_data_all[genotype] = df_data_all[\"sample\"].apply(\n",
    "    lambda x: x.split(\"_\", 1)[0].replace(\"Allele\", \"\")\n",
    ")\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3902ff15-bf6a-43a9-ad82-d81742231410",
   "metadata": {},
   "source": [
    "## Visualize flux ranges\n",
    "### Remove models based on data operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76ce7d4c-834d-4b2f-b705-059839846fea",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_data_for_analyses = df_data_all[\n",
    "    [not bool(operation_re.search(x)) for x in df_data_all[\"model\"]]\n",
    "].reset_index(drop=True)\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7f4389-594f-4c01-8743-94cbef5ffe54",
   "metadata": {},
   "source": [
    "### Create groups of models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "099f41d9-3091-46ef-9fb9-94e0431bfdf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_key = grouped_data_key\n",
    "id_key = \"sample\"\n",
    "model_groups = {all_key: list(df_data_for_analyses[id_key].unique())}\n",
    "\n",
    "\n",
    "def create_group_of_models(df, id_key, groupby, verbose=False):\n",
    "    grouped = df.groupby(groupby)[id_key].agg(lambda x: list(x.unique()))\n",
    "    grouped = {\n",
    "        \"_\".join([str(x) for x in ensure_iterable(k)]): v\n",
    "        for k, v in grouped.to_dict().items()\n",
    "    }\n",
    "    if verbose:\n",
    "        max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "        for group_name, model_list in grouped.items():\n",
    "            spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "            print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "    return grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c12b01a5-a699-4cfa-a715-b88703734644",
   "metadata": {},
   "source": [
    "#### Based on allele count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2a0144d-84bf-4d48-a86a-d420e7c77863",
   "metadata": {},
   "outputs": [],
   "source": [
    "grouped = create_group_of_models(\n",
    "    df_data_for_analyses, id_key=id_key, groupby=genotype, verbose=False\n",
    ")\n",
    "grouped = {f\"{genotype}_{str(k)}\": v for k, v in grouped.items()}\n",
    "if verbose:\n",
    "    max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "    for group_name, model_list in grouped.items():\n",
    "        spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "        print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "model_groups.update(grouped)\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5df95db3-cb7b-487f-9159-66d1d8922f67",
   "metadata": {},
   "source": [
    "### View groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ced7361e-6c04-4a94-9a2b-5246cc2cad2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"Possible groups for analyses\\n============================\")\n",
    "max_name_len = max([len(group_name) for group_name in list(model_groups)])\n",
    "for group_name, model_list in model_groups.items():\n",
    "    spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "    print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "\n",
    "df_data_for_analyses = df_data_for_analyses.set_index([\"reactions\", id_key])\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf1b2cd0-3cfc-4ff7-8f7e-bb259ed1ce00",
   "metadata": {},
   "source": [
    "### Ensure groups exist and setup directory structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6eb4d7fb-5c55-4752-a3f1-7c9b259962cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "group_results_dirpath_dict = {all_key: corr_results_dirpath}\n",
    "header = \"Expected directory structure\"\n",
    "print(\"\\n\".join((header, \"=\" * len(header), all_key)))\n",
    "print(\"\\u2514\\u2500\\u2500\" + f\" {group_results_dirpath_dict[all_key].name}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d642c3b8-d529-4da8-9b11-715718adcd4c",
   "metadata": {},
   "source": [
    "### Visualize selected flux results\n",
    "#### Set groups to compare and filter data accordingly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13cffec8-c36d-447f-bcba-1ad43af758e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_group_name = genotype\n",
    "ordered_groups_to_compare = [\n",
    "    f\"{main_group_name}_{allele_count}\"\n",
    "    for allele_count in [\n",
    "        0,\n",
    "        # 1,\n",
    "        2,\n",
    "    ]\n",
    "]\n",
    "# # Set list of samples for comparison, filter out irrelevant samples, and sort samples by sample number\n",
    "df_data_groups_to_compare = pd.concat(\n",
    "    [\n",
    "        df_data_for_analyses.loc[pd.IndexSlice[:, model_groups[group]], :].sort_index(\n",
    "            level=1\n",
    "        )\n",
    "        for group in ordered_groups_to_compare\n",
    "    ],\n",
    "    axis=0,\n",
    ")\n",
    "df_data_groups_to_compare"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcf0386d-90f9-44f8-88cd-c6fd5f6fbf12",
   "metadata": {},
   "source": [
    "#### Load previous correlations if they exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0494f69-9d88-4be7-ab2e-e003981e1632",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlations_pair = (\"flux\", \"abundance\")\n",
    "filepath = (\n",
    "    group_results_dirpath_dict[all_key]\n",
    "    / f\"{correlations_pair[0]}_{correlations_pair[1]}.tsv\"\n",
    ")\n",
    "try:\n",
    "    df_correlations = pd.read_csv(\n",
    "        filepath, sep=\"\\t\", index_col=\"reactions\", usecols=[\"reactions\", \"rho\"]\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    reaction_correlations = {}\n",
    "    print(\n",
    "        \"No previously computed correlations between '{}' and '{}' for group '{}'.\".format(\n",
    "            *correlations_pair, main_group_name\n",
    "        )\n",
    "    )\n",
    "else:\n",
    "    reaction_correlations = df_correlations.squeeze().to_dict()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd47afd-a4d3-4b01-85db-6638d94b9285",
   "metadata": {},
   "source": [
    "### Set common visualization options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b88ade39-2fca-466f-aaea-b329fd916470",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A dict where keys are columns to sort by and values are whether to sort as ascending or descending,\n",
    "# or a list where values are columns to sort by, and all values are sorted as ascending.\n",
    "rank_order_results = {col: True for col in [\"range\", \"max\", \"min\", \"abundance\"]}\n",
    "optimum_colors = {\n",
    "    0.00: \"xkcd:green\",\n",
    "    # 0.50: \"xkcd:dark yellow\",\n",
    "    # 0.90: \"xkcd:orange\",\n",
    "    # 0.99: \"xkcd:red\",\n",
    "}\n",
    "\n",
    "plot_as_bar = True\n",
    "ypad_percent = 0.1\n",
    "# Kwargs for each individual panel\n",
    "panel_kwargs = dict(\n",
    "    # Only for bar plot\n",
    "    width=1,  # Bar width\n",
    "    linewidth=0.0,  # Edgewidth of bars\n",
    "    edgecolor=\"black\",  # Color of edge width\n",
    "    # Both plots\n",
    "    alpha=0.5,\n",
    "    zorder=2,\n",
    "    maxnum_yticks=3,\n",
    "    # Equilibrium line\n",
    "    equilibrium_lineprops=dict(\n",
    "        color=\"black\",\n",
    "        linestyle=\":\",\n",
    "        linewidth=1,\n",
    "    ),\n",
    ")\n",
    "flux_label_color = \"xkcd:dark red\"\n",
    "abundance_label_color = \"xkcd:dark blue\"\n",
    "abundance_kwargs = dict(\n",
    "    maxnum_yticks=panel_kwargs.get(\"maxnum_yticks\", 3),\n",
    "    lineprops=dict(\n",
    "        color=\"black\",\n",
    "        linestyle=\"--\",\n",
    "        marker=\"\",\n",
    "        linewidth=1,\n",
    "        markersize=4,\n",
    "        zorder=2,\n",
    "    ),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20a54078-801c-4892-b6d7-f6caa9e7c74b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_flux_range_panel(\n",
    "    x, min_values, max_values, ax, bar=True, equilibrium_lineprops=None, **kwargs\n",
    "):\n",
    "    # Pop problematic kwargs\n",
    "    xlim_pad = kwargs.pop(\"xlim_pad\") if kwargs.get(\"xlim_pad\") is not None else 1\n",
    "    ylim_pad = kwargs.pop(\"ylim_pad\") if kwargs.get(\"ylim_pad\") is not None else 0\n",
    "    maxnum_yticks = (\n",
    "        kwargs.pop(\"maxnum_yticks\") if kwargs.get(\"maxnum_yticks\") is not None else 0\n",
    "    )\n",
    "    width = kwargs.pop(\"width\") if kwargs.get(\"width\") else 1\n",
    "    if bar:\n",
    "        ax.bar(\n",
    "            x, height=max_values - min_values, bottom=min_values, width=width, **kwargs\n",
    "        )\n",
    "    else:\n",
    "        # Fill between limits\n",
    "        ax.fill_between(x, min_values, max_values, **kwargs)\n",
    "\n",
    "    if equilibrium_lineprops:\n",
    "        defaults = dict(\n",
    "            linestyle=\":\",\n",
    "            linewidth=1.5,\n",
    "            color=\"black\",\n",
    "            zorder=3,\n",
    "            label=\"equilibrium\",\n",
    "        )\n",
    "        equilibrium_lineprops.update(\n",
    "            {k: v for k, v in defaults.items() if k not in equilibrium_lineprops}\n",
    "        )\n",
    "        ax.hlines(0, x[0] - 2 * xlim_pad, x[-1] + 2 * xlim_pad, **equilibrium_lineprops)\n",
    "\n",
    "    if maxnum_yticks:\n",
    "        ax.yaxis.set_major_locator(mpl.ticker.MaxNLocator(maxnum_yticks))\n",
    "\n",
    "    return ax\n",
    "\n",
    "\n",
    "def plot_abundance_line(x, y, ax, xlims=None, ylims=None, lineprops=None, **kwargs):\n",
    "    default_lineprops = dict(\n",
    "        color=\"black\",\n",
    "        linestyle=\"--\",\n",
    "        marker=\"o\",\n",
    "        linewidth=1,\n",
    "        zorder=3,\n",
    "    )\n",
    "    lineprops = {} if lineprops is None else lineprops\n",
    "    lineprops.update({k: v for k, v in default_lineprops.items() if k not in lineprops})\n",
    "\n",
    "    ax.plot(x, y, **lineprops)\n",
    "    ax.set_xlim(xlims)\n",
    "    ax.set_ylim(ylims)\n",
    "    if kwargs.get(\"maxnum_yticks\"):\n",
    "        ax.yaxis.set_major_locator(mpl.ticker.MaxNLocator(kwargs.get(\"maxnum_yticks\")))\n",
    "    return ax\n",
    "\n",
    "\n",
    "def format_panel_for_group(\n",
    "    ax_panel, xaxis_dict, yaxis_dict, fontsizes_dict, panel_idx=0, **kwargs\n",
    "):\n",
    "    # Format x axis\n",
    "    ax_panel.set_xticks([])\n",
    "    ax_panel.set_xlim(xaxis_dict.get(\"lim\"))\n",
    "    ax_panel.set_xlabel(\n",
    "        xaxis_dict.get(\"label\"), fontsize=fontsizes_dict.get(\"xlabel\", \"large\")\n",
    "    )\n",
    "    ax_panel.xaxis.set_tick_params(\n",
    "        labelsize=fontsizes_dict.get(\"xticks\", \"large\"), labelbottom=False\n",
    "    )\n",
    "    # Format y axis\n",
    "    ax_panel.set_ylim(yaxis_dict.get(\"lim\"))\n",
    "    if panel_idx == 0:\n",
    "        ax_panel.set_ylabel(\n",
    "            yaxis_dict.get(\"label\"),\n",
    "            fontsize=fontsizes_dict.get(\"ylabel\", \"large\"),\n",
    "            va=yaxis_dict.get(\"va\", \"center\"),\n",
    "            labelpad=yaxis_dict.get(\"labelpad\"),\n",
    "            color=yaxis_dict.get(\"color\", \"black\"),\n",
    "        )\n",
    "        if kwargs.get(\"min_max_norm_flux\"):\n",
    "            ax_panel.set_yticks(np.linspace(0, 1, kwargs.get(\"maxnum_yticks\", 3)))\n",
    "        elif yaxis_dict.get(\"maxnum_yticks\"):\n",
    "            ax_panel.set_yticks(\n",
    "                np.linspace(\n",
    "                    ax_panel.get_ybound()[0],\n",
    "                    ax_panel.get_ybound()[1],\n",
    "                    panel_kwargs.get(\"maxnum_yticks\", 3),\n",
    "                )\n",
    "            )\n",
    "        ax_panel.yaxis.set_tick_params(\n",
    "            labelsize=fontsizes_dict.get(\"yticks\", \"large\"),\n",
    "            labelleft=True,\n",
    "            colors=yaxis_dict.get(\"color\", \"black\"),\n",
    "        )\n",
    "        if yaxis_dict.get(\"ticklabel_format\"):\n",
    "            ax_panel.ticklabel_format(\n",
    "                axis=\"y\", **yaxis_dict.get(\"ticklabel_format\", {})\n",
    "            )\n",
    "\n",
    "    seperator_spine_color = kwargs.get(\"seperator_spine_color\")\n",
    "    if panel_idx != 0:\n",
    "        ax_panel.set_ylabel(None)\n",
    "        ax_panel.yaxis.set_tick_params(labelleft=False)\n",
    "        ax_panel.set_yticks([])\n",
    "        (\n",
    "            ax_panel.spines[\"left\"].set_color(seperator_spine_color)\n",
    "            if seperator_spine_color\n",
    "            else ax_panel.spines[\"left\"].set_visible(False)\n",
    "        )\n",
    "\n",
    "    if panel_idx != (len(ordered_model_groups) - 1):\n",
    "        (\n",
    "            ax_panel.spines[\"right\"].set_color(seperator_spine_color)\n",
    "            if seperator_spine_color\n",
    "            else ax_panel.spines[\"right\"].set_visible(False)\n",
    "        )\n",
    "\n",
    "\n",
    "def create_flux_range_figure_axes(nrows, ncols, ngroups, **kwargs):\n",
    "    fig = mpl.figure.Figure(\n",
    "        figsize=kwargs.get(\n",
    "            \"figsize\",\n",
    "            (\n",
    "                (ncols + kwargs.get(\"wspace\", 1))\n",
    "                * ngroups\n",
    "                * kwargs.get(\"width_scalar\", 1),\n",
    "                (nrows + kwargs.get(\"hspace\", 1)) * kwargs.get(\"height_scalar\", 1),\n",
    "            ),\n",
    "        )\n",
    "    )\n",
    "    outer_grid = fig.add_gridspec(\n",
    "        nrows, ncols, hspace=kwargs.get(\"hspace\", 1), wspace=kwargs.get(\"wspace\", 1)\n",
    "    )\n",
    "    axes = np.zeros((nrows, ncols, ngroups)).astype(object)\n",
    "    for row_idx in range(nrows):\n",
    "        for col_idx in range(ncols):\n",
    "            inner_grid = outer_grid[row_idx, col_idx].subgridspec(\n",
    "                1, ngroups, hspace=0, wspace=0\n",
    "            )\n",
    "            axes[row_idx, col_idx, :] = inner_grid.subplots()\n",
    "\n",
    "    return fig, axes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c815c63-e305-488c-96c4-a3e746f39ffc",
   "metadata": {},
   "source": [
    "#### Visualize groups for one or more reaction fluxes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66342a9a-fad0-41fb-8e8e-45ee26992cff",
   "metadata": {},
   "outputs": [],
   "source": [
    "fluxes_to_plot = np.atleast_2d(\n",
    "    np.array(\n",
    "        [\n",
    "            # ['SERHL', \"CYSL2\"],\n",
    "            # ['G6PDH2'],\n",
    "            # ['PPM'],\n",
    "            # ['ADK1'],\n",
    "            # ['DTHYMy'],\n",
    "            # ['PAFS', 'LPCOAT', 'LPCPAT'],\n",
    "            # ['PROTSPP', 'PROTTPP'],\n",
    "            # ['PSFLIPt', 'PEFLIPt'],\n",
    "            # ['PCFLOPt', \"SMFLOPt\"],\n",
    "            # ['DTHYMy', 'DURADy'],\n",
    "            [\"PSSCRt\", \"PCSCRt\", \"PESCRt\"],\n",
    "        ],\n",
    "    )\n",
    ")\n",
    "optimums = [0]\n",
    "sharex = True\n",
    "sharey = \"row\"\n",
    "supylabel = False\n",
    "plot_as_bar = False\n",
    "min_max_norm_flux = False\n",
    "min_max_norm_abundance = True\n",
    "include_abundance_line = True\n",
    "width_scalar = 1\n",
    "height_scalar = 0.8\n",
    "fontsizes_dict = dict(\n",
    "    xlabel=10,\n",
    "    xticks=10,\n",
    "    ylabel=10,\n",
    "    yticks=10,\n",
    "    title=12,\n",
    ")\n",
    "panel_kwargs = dict(\n",
    "    # Only for bar plot\n",
    "    width=1,  # Bar width\n",
    "    linewidth=0.5,  # Edgewidth of bars\n",
    "    edgecolor=\"black\",  # Color of edge width\n",
    "    # Both plots\n",
    "    alpha=0.5,\n",
    "    zorder=2,\n",
    "    maxnum_yticks=3,\n",
    "    # Equilibrium line\n",
    "    equilibrium_lineprops=dict(\n",
    "        color=\"black\",\n",
    "        linestyle=\":\",\n",
    "        linewidth=1,\n",
    "    ),\n",
    ")\n",
    "flux_str = \"Normalized\\nFlux\" if min_max_norm_flux else \"Flux\\n(mmol/gDW/hr)\"\n",
    "abundance_str = (\n",
    "    \"Normalized\\nAbundance\" if min_max_norm_abundance else \"Abundance\\n(nmol/gDW)\"\n",
    ")\n",
    "ordered_model_groups = {\n",
    "    group: model_groups[group] for group in ordered_groups_to_compare\n",
    "}\n",
    "xlabels_dict = {\n",
    "    group: group.replace(main_group_name, \"\").strip(\"_\")\n",
    "    for group in list(ordered_model_groups)\n",
    "}\n",
    "ylabels_dict = (\n",
    "    {reaction: flux_str for reaction in fluxes_to_plot.flatten() if reaction}\n",
    "    if not supylabel\n",
    "    else {}\n",
    ")\n",
    "ylabels_dict = {}\n",
    "titles_dict = {\n",
    "    reaction: reaction\n",
    "    + (\n",
    "        \" ($\\\\rho$ = \" + \"{})\".format(round(reaction_correlations[reaction], 4))\n",
    "        if reaction_correlations and reaction_correlations.get(reaction)\n",
    "        else \"\"\n",
    "    )\n",
    "    for reaction in fluxes_to_plot.flatten()\n",
    "    if reaction\n",
    "}\n",
    "# titles_dict = {}\n",
    "fig, all_axes = create_flux_range_figure_axes(\n",
    "    *fluxes_to_plot.shape,\n",
    "    len(ordered_model_groups),\n",
    "    width_scalar=width_scalar,  # * (fluxes_to_plot.shape[0]),\n",
    "    height_scalar=height_scalar,  # * (fluxes_to_plot.shape[1] + hspace),\n",
    "    hspace=0.5 if sharex else 1,\n",
    "    wspace=0.05 if sharey else 0.5,\n",
    ")\n",
    "\n",
    "for row_idx, (row_rxns, row_axes) in enumerate(zip(fluxes_to_plot, all_axes)):\n",
    "    for col_idx, (reaction, axes) in enumerate(zip(row_rxns, row_axes)):\n",
    "        df_data_reaction = df_data_groups_to_compare.loc[reaction].copy()\n",
    "        if min_max_norm_flux:\n",
    "            df_data_reaction[\"norm_range\"] = (\n",
    "                (df_data_reaction[\"range\"] - df_data_reaction[\"range\"].min())\n",
    "                / (df_data_reaction[\"range\"].max() - df_data_reaction[\"range\"].min())\n",
    "            ).fillna(0)\n",
    "            ylims = (0, 1 + ypad_percent)\n",
    "        elif sharey:\n",
    "            rxn_list = row_rxns if sharey == \"row\" else fluxes_to_plot.flatten()\n",
    "            df_ylim = df_data_groups_to_compare.loc[[x for x in rxn_list if x]]\n",
    "            ylims = (\n",
    "                df_ylim[\"min\"].min() - abs(df_ylim[\"min\"].min() * ypad_percent),\n",
    "                df_ylim[\"max\"].max() + abs(df_ylim[\"max\"].max() * ypad_percent),\n",
    "            )\n",
    "        else:\n",
    "            ylims = (\n",
    "                df_data_reaction[\"min\"].min()\n",
    "                - abs(df_data_reaction[\"min\"].min() * ypad_percent),\n",
    "                df_data_reaction[\"max\"].max()\n",
    "                + abs(df_data_reaction[\"max\"].max() * ypad_percent),\n",
    "            )\n",
    "        for panel_idx, ((group, sample_list), ax_panel) in enumerate(\n",
    "            zip(ordered_model_groups.items(), axes)\n",
    "        ):\n",
    "            # Get samples for group and order by donor number\n",
    "            df_data_group = df_data_reaction.loc[sample_list]\n",
    "            # Get data for ooptimums\n",
    "            xlims = (0, len(df_data_group.index.unique()))\n",
    "            for optimum in optimums:\n",
    "                df = df_data_group[df_data_group[\"optimum\"] == optimum].drop(\n",
    "                    \"optimum\", axis=1\n",
    "                )\n",
    "                # Rank order results within the optimum if desired\n",
    "                df = df.sort_values(\n",
    "                    by=list(rank_order_results),\n",
    "                    ascending=(\n",
    "                        list(rank_order_results.values())\n",
    "                        if isinstance(rank_order_results, dict)\n",
    "                        else True\n",
    "                    ),\n",
    "                )\n",
    "                ax_panel = plot_flux_range_panel(\n",
    "                    x=np.arange(*xlims),\n",
    "                    min_values=(\n",
    "                        df[\"min\"].values\n",
    "                        if not min_max_norm_flux\n",
    "                        else df[\"min\"].values - df[\"min\"].values\n",
    "                    ),\n",
    "                    max_values=(\n",
    "                        df[\"max\"].values\n",
    "                        if not min_max_norm_flux\n",
    "                        else df[\"norm_range\"].values\n",
    "                    ),\n",
    "                    ax=ax_panel,\n",
    "                    bar=plot_as_bar,\n",
    "                    color=optimum_colors.get(optimum, \"xkcd:light green\"),\n",
    "                    **panel_kwargs,\n",
    "                )\n",
    "\n",
    "            if include_abundance_line:\n",
    "                abundance_values = df_data_reaction.groupby(level=0)[\"abundance\"].max()\n",
    "                if min_max_norm_abundance:\n",
    "                    abundance_values = (\n",
    "                        (abundance_values - abundance_values.min())\n",
    "                        / (abundance_values.max() - abundance_values.min())\n",
    "                    ).fillna(0)\n",
    "                    aylims = (0, 1)\n",
    "                elif sharey:\n",
    "                    rxn_list = row_rxns if sharey == \"row\" else fluxes_to_plot.flatten()\n",
    "                    aylims = (\n",
    "                        df_data_groups_to_compare.loc[[x for x in rxn_list if x]][\n",
    "                            \"abundance\"\n",
    "                        ].min(),\n",
    "                        df_data_groups_to_compare.loc[[x for x in rxn_list if x]][\n",
    "                            \"abundance\"\n",
    "                        ].max(),\n",
    "                    )\n",
    "                    aylims = (\n",
    "                        aylims[0] - (aylims[0] * ypad_percent),\n",
    "                        aylims[1] + (aylims[1] * ypad_percent),\n",
    "                    )\n",
    "                else:\n",
    "                    aylims = (\n",
    "                        0 - (df_data_reaction[\"abundance\"].max() * ypad_percent),\n",
    "                        df_data_reaction[\"abundance\"].max()\n",
    "                        + (df_data_reaction[\"abundance\"].max() * ypad_percent),\n",
    "                    )\n",
    "\n",
    "                ax_abundance = plot_abundance_line(\n",
    "                    x=np.arange(*xlims),\n",
    "                    y=abundance_values.loc[df.index].copy(),\n",
    "                    ax=ax_panel.twinx(),\n",
    "                    xlims=(\n",
    "                        xlims[0] - panel_kwargs.get(\"xlim_pad\", 1),\n",
    "                        xlims[1] + panel_kwargs.get(\"xlim_pad\", 1),\n",
    "                    ),\n",
    "                    ylims=aylims,\n",
    "                    **abundance_kwargs,\n",
    "                )\n",
    "                if (panel_idx == len(ordered_model_groups) - 1) and (\n",
    "                    (sharey and col_idx == fluxes_to_plot.shape[1] - 1) or not sharey\n",
    "                ):\n",
    "                    ylabel = abundance_str if not supylabel else None\n",
    "                    if min_max_norm_abundance:\n",
    "                        ax_abundance.set_yticks(\n",
    "                            np.linspace(0, 1, panel_kwargs.get(\"maxnum_yticks\", 3))\n",
    "                        )\n",
    "                    elif panel_kwargs.get(\"maxnum_yticks\"):\n",
    "                        ax_abundance.set_yticks(\n",
    "                            np.linspace(\n",
    "                                ax_abundance.get_ybound()[0],\n",
    "                                ax_abundance.get_ybound()[1],\n",
    "                                panel_kwargs.get(\"maxnum_yticks\", 3),\n",
    "                            )\n",
    "                        )\n",
    "                    ax_abundance.set_ylabel(\n",
    "                        ylabel,\n",
    "                        fontsize=fontsizes_dict.get(\"ylabel\", \"large\"),\n",
    "                        va=\"center\",\n",
    "                        labelpad=15,\n",
    "                        color=abundance_label_color,\n",
    "                    )\n",
    "                    ax_abundance.yaxis.set_tick_params(\n",
    "                        labelsize=fontsizes_dict.get(\"yticks\", \"large\"),\n",
    "                        labelright=True,\n",
    "                        colors=abundance_label_color,\n",
    "                    )\n",
    "                else:\n",
    "                    ax_abundance.set_ylabel(None)\n",
    "                    ax_abundance.set_yticks([])\n",
    "                    ax_abundance.yaxis.set_tick_params(labelright=False)\n",
    "\n",
    "            # Format panel, lot of conditional inputs depending on where panel is and whether multiple rows/columns are present.\n",
    "            format_panel_for_group(\n",
    "                ax_panel,\n",
    "                panel_idx=panel_idx,\n",
    "                xaxis_dict=dict(\n",
    "                    # Add extra padding for limits\n",
    "                    lim=(\n",
    "                        xlims[0] - panel_kwargs.get(\"xlim_pad\", 1),\n",
    "                        xlims[1] + panel_kwargs.get(\"xlim_pad\", 1),\n",
    "                    ),\n",
    "                    label=(\n",
    "                        xlabels_dict.get(group)\n",
    "                        if (\n",
    "                            (row_idx == fluxes_to_plot.shape[0] - 1 and sharex)\n",
    "                            or not sharex\n",
    "                        )\n",
    "                        else None\n",
    "                    ),\n",
    "                ),\n",
    "                yaxis_dict=dict(\n",
    "                    lim=ylims,\n",
    "                    label=(\n",
    "                        ylabels_dict.get(reaction, flux_str)\n",
    "                        if (not supylabel and ((col_idx == 0 and sharey) or not sharey))\n",
    "                        else None\n",
    "                    ),\n",
    "                    va=\"center\",\n",
    "                    labelpad=15,\n",
    "                    color=flux_label_color if include_abundance_line else \"black\",\n",
    "                    maxnum_yticks=panel_kwargs.get(\"maxnum_yticks\", 3),\n",
    "                    ticklabel_format=dict(style=\"sci\", scilimits=(0, 0)),\n",
    "                ),\n",
    "                fontsizes_dict=fontsizes_dict,\n",
    "                seperator_spine_color=\"xkcd:grey\",\n",
    "                min_max_norm_flux=min_max_norm_flux,\n",
    "            )\n",
    "            if panel_idx == 0 and ((col_idx != 0 and sharey) or not sharey):\n",
    "                ax_panel.yaxis.set_tick_params(labelleft=False)\n",
    "                ax_panel.set_yticks([])\n",
    "            if panel_idx == (len(ordered_model_groups) - 1) / 2 or (\n",
    "                len(ordered_model_groups) % 2 == 0 and panel_idx % 2 == 0\n",
    "            ):\n",
    "                if (\n",
    "                    panel_idx == (len(ordered_model_groups) - 1) / 2\n",
    "                ):  # Odd number of panels\n",
    "                    ax_panel.set_title(\n",
    "                        titles_dict.get(reaction, reaction),\n",
    "                        fontsize=fontsizes_dict.get(\"title\", \"xx-large\"),\n",
    "                    )\n",
    "                else:  # Even number of panels\n",
    "                    ax_panel.text(\n",
    "                        x=1,\n",
    "                        y=1.1,\n",
    "                        s=titles_dict.get(reaction, reaction),\n",
    "                        transform=ax_panel.transAxes,\n",
    "                        ha=\"center\",\n",
    "                    )\n",
    "if supylabel:\n",
    "    # Adjust manually\n",
    "    text = fig.text(\n",
    "        x=0.02 + 0.025 * fluxes_to_plot.shape[1],\n",
    "        y=0.5,\n",
    "        s=flux_str,\n",
    "        ha=\"center\",\n",
    "        va=\"center\",\n",
    "        fontsize=fontsizes_dict.get(\"supylabel\", \"xx-large\"),\n",
    "        color=flux_label_color if include_abundance_line else \"black\",\n",
    "        rotation=\"vertical\",\n",
    "    )\n",
    "    if include_abundance_line:\n",
    "        fig.text(\n",
    "            x=1.0 - 0.025 * fluxes_to_plot.shape[1],\n",
    "            y=0.5,\n",
    "            s=abundance_str,\n",
    "            ha=\"center\",\n",
    "            va=\"center\",\n",
    "            fontsize=fontsizes_dict.get(\"supylabel\", \"xx-large\"),\n",
    "            color=abundance_label_color,\n",
    "            rotation=\"vertical\",\n",
    "        )\n",
    "\n",
    "fig.align_xlabels()\n",
    "fig.align_ylabels()\n",
    "plt.show()\n",
    "if save_figures:\n",
    "    rxns = \"_\".join(fluxes_to_plot.flatten())\n",
    "    fig_dirpath = (\n",
    "        group_results_dirpath_dict[all_key]\n",
    "        / \"fluxranges\"\n",
    "        / \"_\".join(ordered_groups_to_compare)\n",
    "    )\n",
    "    fig_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "    fig.savefig(\n",
    "        fig_dirpath / f\"fluxes_{rxns}.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16d9216a-94bd-4638-a9b6-40e389609b1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.metabolites.sphings_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "893cb410-eefc-4a1c-bdf9-98ac3e348df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.reactions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c0ba7d0-4dfa-4e7c-9a83-d0cd77de2f38",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
