{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e3208d0-39f0-4d23-8d86-6de6fa19fd8b",
   "metadata": {},
   "source": [
    "# Create Protein-Constrained RBC model via OVERLAY workflow \n",
    "This notebook facilitates the construction of a proteome constrained model (\"pcModel\") via the OVERLAY methodology.\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d968a45f-c732-43c5-bfc4-1114f8d8cb01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Set parameter LicenseID to value 2664191\n",
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.3\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.13.4\n",
      "bio                                   1.8.0\n",
      "cobra                                0.29.1\n",
      "depinfo                               2.2.0\n",
      "gurobipy                             12.0.3\n",
      "matplotlib                           3.10.3\n",
      "matplotlib-venn                       1.1.2\n",
      "memote                               0.17.0\n",
      "networkx                                3.5\n",
      "notebook                              7.4.4\n",
      "openpyxl                              3.1.5\n",
      "pandas                                2.3.1\n",
      "pre-commit                            4.2.0\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.32.4\n",
      "scikit-learn                          1.7.0\n",
      "scipy                                1.16.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip          25.1\n",
      "setuptools 78.1.1\n",
      "wheel      0.45.1\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Windows 10-AMD64\n",
      "CPython  3.11.13\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "from collections import defaultdict\n",
    "\n",
    "import gurobipy as gp\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    build_string,\n",
    "    get_annotation_df,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    split_string,\n",
    "    write_cobra_model,\n",
    ")\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    ATTR_SUBCLASS_DICT,\n",
    "    DEFAULT_KEFF,\n",
    "    DEFAULT_PREFIX_SUFFIX_VALUES,\n",
    "    DEFAULT_PROTEOME_COMPARTMENT,\n",
    "    Budget,\n",
    "    BudgetDilution,\n",
    "    Complex,\n",
    "    ComplexDilution,\n",
    "    Enzyme,\n",
    "    EnzymeDilution,\n",
    "    Protein,\n",
    "    ProteinDilution,\n",
    "    construct_pcmodel_from_tables,\n",
    "    create_complex_table,\n",
    "    create_enzyme_table,\n",
    "    create_protein_table,\n",
    "    create_sequence_table,\n",
    ")\n",
    "from rbc_gem_utils.database import MGI_DB_TAG, UNIPROT_DB_TAG\n",
    "from rbc_gem_utils.util import strip_plural\n",
    "\n",
    "gp.setParam(\"OutputFlag\", 0)\n",
    "gp.setParam(\"LogToConsole\", 0)\n",
    "\n",
    "# Show versions of notebook\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c19f01-8e9f-448e-bcbb-64fb84358e1b",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28d822cd-8d55-40c1-81d3-34165bf0f24b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 127\n",
       "cache_directory: C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "COBRA_CONFIGURATION.bounds = (-1e3, 1e3)\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9c1895d-964d-4590-8308-d1bf64a21cc7",
   "metadata": {},
   "source": [
    "### Define organism and model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "63cfcfc4-804b-4a04-a38d-5366faa6a66b",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = \"Mouse\"\n",
    "model_id = \"RBC_GEM\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b333100-6a8a-49a3-8e25-d95c840c5e65",
   "metadata": {},
   "source": [
    "### Set paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "eaa426d8-681c-44ea-9246-9f6b53f292e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "database_dirpath = get_dirpath(\"database\", MGI_DB_TAG)\n",
    "overlay_dirpath = get_dirpath(\"analysis\") / \"OVERLAY\" / organism\n",
    "model_dirpath = overlay_dirpath / model_id"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b34b89b1-f3bb-418f-81de-1438c31cf7a0",
   "metadata": {},
   "source": [
    "### Define hemoglobin proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a6d07d8-c823-459e-95f7-fd357cb4c45b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Identify hemoglobin proteins\n",
    "hemoglobin_proteins = {\n",
    "    k.replace(\"-\", \"_\"): v\n",
    "    for k, v in {\n",
    "        \"Hba\": \"P01942\",  # Hemoglobin subunit alpha\n",
    "        \"Hba-a1\": \"P01942\",\n",
    "        \"Hbb-b1\": \"P02088\",  # Hemoglobin subunit beta-1\n",
    "        \"Hbb-b2\": \"P02089\",  # Hemoglobin subunit beta-2\n",
    "        \"Hbb-bh0\": \"P04443\",  # Hemoglobin subunit beta-H0\n",
    "        \"Hbb-bh1\": \"P04444\",  # Hemoglobin subunit beta-H1\n",
    "        \"Hbz\": \"P06467\",  # Hemoglobin subunit zeta\n",
    "        \"Hba-x\": \"P06467\",\n",
    "        \"Hbz1\": \"P06467\",\n",
    "        \"Hbb-y\": \"P02104\",  # Hemoglobin subunit epsilon-Y2\n",
    "    }.items()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae972b5d-74b6-4530-a55f-14b45fefd4c6",
   "metadata": {},
   "source": [
    "## Load RBC model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "66ac203d-e751-4070-9920-52277287b79d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>22fd5338550</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3274</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>826</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x22fd5338550>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_cobra_model(filename=model_dirpath / f\"{model_id}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c38a3d73-7dee-4165-ac00-a21f45d98ea1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes      826\n",
      "uniprot    826\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4galt</td>\n",
       "      <td>Q67BJ4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aars1</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aasdhppt</td>\n",
       "      <td>Q9CQF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abca1</td>\n",
       "      <td>P41233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Abca7</td>\n",
       "      <td>Q91V24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>821</th>\n",
       "      <td>Yes1</td>\n",
       "      <td>Q04736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>822</th>\n",
       "      <td>Zdhhc2</td>\n",
       "      <td>P59267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>823</th>\n",
       "      <td>Zdhhc20</td>\n",
       "      <td>Q5Y5T1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>Q8R173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>Zdhhc5</td>\n",
       "      <td>Q8VDZ4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genes uniprot\n",
       "0      A4galt  Q67BJ4\n",
       "1       Aars1  Q8BGQ7\n",
       "2    Aasdhppt  Q9CQF6\n",
       "3       Abca1  P41233\n",
       "4       Abca7  Q91V24\n",
       "..        ...     ...\n",
       "821      Yes1  Q04736\n",
       "822    Zdhhc2  P59267\n",
       "823   Zdhhc20  Q5Y5T1\n",
       "824    Zdhhc3  Q8R173\n",
       "825    Zdhhc5  Q8VDZ4\n",
       "\n",
       "[826 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_type = \"genes\"\n",
    "mapping_key = \"uniprot\"\n",
    "annotation_cols = [mapping_key]\n",
    "\n",
    "df_model_mappings = (\n",
    "    get_annotation_df(model.genes, annotation_cols)\n",
    "    .rename({\"id\": annotation_type}, axis=1)\n",
    "    .dropna(subset=[mapping_key])\n",
    ")\n",
    "for col in df_model_mappings.columns:\n",
    "    df_model_mappings[col] = df_model_mappings[col].apply(lambda x: split_string(x))\n",
    "    df_model_mappings = df_model_mappings.explode(col).drop_duplicates().dropna()\n",
    "df_model_mappings = df_model_mappings.sort_values(annotation_type)\n",
    "\n",
    "print(df_model_mappings.nunique(dropna=True))\n",
    "df_model_mappings = df_model_mappings.reset_index(drop=True)\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb9eff5e-0f50-463e-a82c-40c4565e16b6",
   "metadata": {},
   "source": [
    "## Assemble data for PC-model\n",
    "### Load protein data\n",
    "#### Protein amino acid sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e168b414-e397-4144-9283-ade328ab6a35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P04919' 'P22907' 'P49282' 'P53657']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot</th>\n",
       "      <th>uniprot.isoform</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>canonical</th>\n",
       "      <th>erythroid</th>\n",
       "      <th>backup</th>\n",
       "      <th>avoid</th>\n",
       "      <th>keywords.erythroid</th>\n",
       "      <th>keywords.backup</th>\n",
       "      <th>keywords.avoid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sequence.length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1L3P4</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>A1L3P4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MAGARRGWRLAPVRRGVCGPRARPLMRPLWLLFAVSFFGWTGALDG...</td>\n",
       "      <td>702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A2A432</td>\n",
       "      <td>A2A432-1</td>\n",
       "      <td>A2A432-1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MSRSTRSKERRENDTDSEDNSSETSNQERRRCRQGPPRPPYPPLLP...</td>\n",
       "      <td>970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A2A432</td>\n",
       "      <td>A2A432-2</td>\n",
       "      <td>A2A432-2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MSRSTRSKERRENDTDSEDNSSETSNQERRRCRQGPPRPPYPPLLP...</td>\n",
       "      <td>941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A2ALK8</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>A2ALK8</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MTSRLRALGGRINNTRTSELPKEKTRSEVICSIRFLDGLVQTFKVN...</td>\n",
       "      <td>913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A2AN08</td>\n",
       "      <td>A2AN08-1</td>\n",
       "      <td>A2AN08-1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MATSGGEEAAAAAPAPGAPATGQDTTPGWEVAVRPLLSASYSAFEM...</td>\n",
       "      <td>5180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1103</th>\n",
       "      <td>Q9Z2L6</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Q9Z2L6</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MLRGARSHLPASVAPAAVLAAALLSSFARCSLPGRGDPVASVLSPY...</td>\n",
       "      <td>481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104</th>\n",
       "      <td>Q9Z2M7</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Q9Z2M7</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MATLCLFDMDGTLTAPRQKITEEMDGFLQKLRQKTKIGVVGGSDFE...</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>Q9Z2V4</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Q9Z2V4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MPPQLHNGLDFSAKVIQGSLDSLPQAVRKFVEGNAQLCQPEYIHIC...</td>\n",
       "      <td>622</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1106</th>\n",
       "      <td>Q9Z2W0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Q9Z2W0</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MAMNGRARKEAIQATARELLKFVNRSPSPFHVVAECRSRLLQAGFR...</td>\n",
       "      <td>473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1107</th>\n",
       "      <td>Q9Z306</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Q9Z306</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MRDYDEVIAFLGEWGPFQRLIFFLLSASIIPNGFNGMSVVFLAGTP...</td>\n",
       "      <td>553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1108 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     uniprot uniprot.isoform sequence.id  canonical  erythroid  backup  avoid  \\\n",
       "0     A1L3P4            <NA>      A1L3P4       True      False   False  False   \n",
       "1     A2A432        A2A432-1    A2A432-1       True      False   False  False   \n",
       "2     A2A432        A2A432-2    A2A432-2      False      False   False  False   \n",
       "3     A2ALK8            <NA>      A2ALK8       True      False   False  False   \n",
       "4     A2AN08        A2AN08-1    A2AN08-1       True      False   False  False   \n",
       "...      ...             ...         ...        ...        ...     ...    ...   \n",
       "1103  Q9Z2L6            <NA>      Q9Z2L6       True      False   False  False   \n",
       "1104  Q9Z2M7            <NA>      Q9Z2M7       True      False   False  False   \n",
       "1105  Q9Z2V4            <NA>      Q9Z2V4       True      False   False  False   \n",
       "1106  Q9Z2W0            <NA>      Q9Z2W0       True      False   False  False   \n",
       "1107  Q9Z306            <NA>      Q9Z306       True      False   False  False   \n",
       "\n",
       "     keywords.erythroid keywords.backup keywords.avoid  \\\n",
       "0                  <NA>            <NA>           <NA>   \n",
       "1                  <NA>            <NA>           <NA>   \n",
       "2                  <NA>            <NA>           <NA>   \n",
       "3                  <NA>            <NA>           <NA>   \n",
       "4                  <NA>            <NA>           <NA>   \n",
       "...                 ...             ...            ...   \n",
       "1103               <NA>            <NA>           <NA>   \n",
       "1104               <NA>            <NA>           <NA>   \n",
       "1105               <NA>            <NA>           <NA>   \n",
       "1106               <NA>            <NA>           <NA>   \n",
       "1107               <NA>            <NA>           <NA>   \n",
       "\n",
       "                                               sequence  sequence.length  \n",
       "0     MAGARRGWRLAPVRRGVCGPRARPLMRPLWLLFAVSFFGWTGALDG...              702  \n",
       "1     MSRSTRSKERRENDTDSEDNSSETSNQERRRCRQGPPRPPYPPLLP...              970  \n",
       "2     MSRSTRSKERRENDTDSEDNSSETSNQERRRCRQGPPRPPYPPLLP...              941  \n",
       "3     MTSRLRALGGRINNTRTSELPKEKTRSEVICSIRFLDGLVQTFKVN...              913  \n",
       "4     MATSGGEEAAAAAPAPGAPATGQDTTPGWEVAVRPLLSASYSAFEM...             5180  \n",
       "...                                                 ...              ...  \n",
       "1103  MLRGARSHLPASVAPAAVLAAALLSSFARCSLPGRGDPVASVLSPY...              481  \n",
       "1104  MATLCLFDMDGTLTAPRQKITEEMDGFLQKLRQKTKIGVVGGSDFE...              242  \n",
       "1105  MPPQLHNGLDFSAKVIQGSLDSLPQAVRKFVEGNAQLCQPEYIHIC...              622  \n",
       "1106  MAMNGRARKEAIQATARELLKFVNRSPSPFHVVAECRSRLLQAGFR...              473  \n",
       "1107  MRDYDEVIAFLGEWGPFQRLIFFLLSASIIPNGFNGMSVVFLAGTP...              553  \n",
       "\n",
       "[1108 rows x 12 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_isoforms_sequences = pd.read_csv(\n",
    "    database_dirpath / f\"{UNIPROT_DB_TAG}_isoforms_sequences.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(pd.NA)\n",
    "print(df_isoforms_sequences[df_isoforms_sequences[\"erythroid\"]][\"uniprot\"].unique())\n",
    "df_isoforms_sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0ed08ce-df40-4451-afc0-a6cf6c720d5f",
   "metadata": {},
   "source": [
    "#### Determine protein isoforms and associated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "88cd6736-a2da-48ea-be6a-a946219eeee2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canonical    826\n",
      "erythroid      4\n",
      "backup        36\n",
      "avoid         14\n",
      "dtype: int64\n",
      "Total: 854\n",
      "\n",
      "canonical    818\n",
      "erythroid      4\n",
      "backup        19\n",
      "avoid          3\n",
      "dtype: int64\n",
      "Total: 829\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4galt</td>\n",
       "      <td>Q67BJ4</td>\n",
       "      <td>Q67BJ4</td>\n",
       "      <td>MGISCSHLEETMSKPPDCLLRMLRGTPRQRVFTFFIISFKFMFLIS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Aars1</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "      <td>MDATLTAREIRERFINFFRRNEHTYVHSSATIPLDDPTLLFANAGM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Aasdhppt</td>\n",
       "      <td>Q9CQF6</td>\n",
       "      <td>Q9CQF6-1</td>\n",
       "      <td>MVFPAKRLCVVPSMEGVRWAFSCGTWLPSRAEWLLAMRSIQPEEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abca1</td>\n",
       "      <td>P41233</td>\n",
       "      <td>P41233</td>\n",
       "      <td>MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Abca7</td>\n",
       "      <td>Q91V24</td>\n",
       "      <td>Q91V24</td>\n",
       "      <td>MALGTQLMLLLWKNYTYRRRQPIQLLVELLWPLFLFFILVAVRHSH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>Yes1</td>\n",
       "      <td>Q04736</td>\n",
       "      <td>Q04736</td>\n",
       "      <td>MGCIKSKENKSPAIKYTPENLTEPVSPSASHYGVEHATVAPTSSTK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>Zdhhc2</td>\n",
       "      <td>P59267</td>\n",
       "      <td>P59267</td>\n",
       "      <td>MAPSGSGGVRRRCRRVLYWIPVVFISLLLGWSYYAYAIQLCIVSME...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>Zdhhc20</td>\n",
       "      <td>Q5Y5T1</td>\n",
       "      <td>Q5Y5T1-1</td>\n",
       "      <td>MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVSTISRTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>Q8R173</td>\n",
       "      <td>Q8R173</td>\n",
       "      <td>MMLIPTHHFRDIERKPEYLQPEKCAPPPFPGPAGAMWFIRDGCGIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>Zdhhc5</td>\n",
       "      <td>Q8VDZ4</td>\n",
       "      <td>Q8VDZ4-1</td>\n",
       "      <td>MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLNVSPA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>829 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genes uniprot sequence.id  \\\n",
       "0      A4galt  Q67BJ4      Q67BJ4   \n",
       "1       Aars1  Q8BGQ7      Q8BGQ7   \n",
       "2    Aasdhppt  Q9CQF6    Q9CQF6-1   \n",
       "3       Abca1  P41233      P41233   \n",
       "4       Abca7  Q91V24      Q91V24   \n",
       "..        ...     ...         ...   \n",
       "824      Yes1  Q04736      Q04736   \n",
       "825    Zdhhc2  P59267      P59267   \n",
       "826   Zdhhc20  Q5Y5T1    Q5Y5T1-1   \n",
       "827    Zdhhc3  Q8R173      Q8R173   \n",
       "828    Zdhhc5  Q8VDZ4    Q8VDZ4-1   \n",
       "\n",
       "                                              sequence  \n",
       "0    MGISCSHLEETMSKPPDCLLRMLRGTPRQRVFTFFIISFKFMFLIS...  \n",
       "1    MDATLTAREIRERFINFFRRNEHTYVHSSATIPLDDPTLLFANAGM...  \n",
       "2    MVFPAKRLCVVPSMEGVRWAFSCGTWLPSRAEWLLAMRSIQPEEKE...  \n",
       "3    MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...  \n",
       "4    MALGTQLMLLLWKNYTYRRRQPIQLLVELLWPLFLFFILVAVRHSH...  \n",
       "..                                                 ...  \n",
       "824  MGCIKSKENKSPAIKYTPENLTEPVSPSASHYGVEHATVAPTSSTK...  \n",
       "825  MAPSGSGGVRRRCRRVLYWIPVVFISLLLGWSYYAYAIQLCIVSME...  \n",
       "826  MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVSTISRTG...  \n",
       "827  MMLIPTHHFRDIERKPEYLQPEKCAPPPFPGPAGAMWFIRDGCGIA...  \n",
       "828  MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLNVSPA...  \n",
       "\n",
       "[829 rows x 4 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Erythroid first, then canonical to enable drop_duplicates to keep erythroid over canonical\n",
    "df_model_isoforms_sequences = (\n",
    "    pd.concat(\n",
    "        (\n",
    "            df_isoforms_sequences[df_isoforms_sequences[\"erythroid\"]],\n",
    "            df_isoforms_sequences[df_isoforms_sequences[\"canonical\"]],\n",
    "            df_isoforms_sequences[df_isoforms_sequences[\"backup\"]],\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "    .fillna(pd.NA)\n",
    "    .drop_duplicates()\n",
    "    .sort_values(\n",
    "        [\"uniprot\", \"erythroid\", \"uniprot.isoform\"], ascending=[True, False, True]\n",
    "    )\n",
    ")\n",
    "\n",
    "print(\n",
    "    df_model_isoforms_sequences[[\"canonical\", \"erythroid\", \"backup\", \"avoid\"]].sum(\n",
    "        axis=0\n",
    "    )\n",
    ")\n",
    "print(f\"Total: {len(df_model_isoforms_sequences)}\")\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"uniprot\",\n",
    "        \"uniprot.isoform\",\n",
    "        \"sequence.id\",\n",
    "        \"sequence\",\n",
    "        \"sequence.length\",\n",
    "        \"canonical\",\n",
    "        \"erythroid\",\n",
    "        \"backup\",\n",
    "        \"avoid\",\n",
    "    ],\n",
    "].reset_index(drop=True)\n",
    "\n",
    "\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences.copy()\n",
    "df_model_isoforms_sequences[\"keep\"] = df_model_isoforms_sequences[\"canonical\"].values\n",
    "to_avoid = df_model_isoforms_sequences[df_model_isoforms_sequences[\"avoid\"]][\n",
    "    \"uniprot\"\n",
    "].to_dict()\n",
    "df_model_isoforms_sequences.loc[\n",
    "    list(to_avoid),\n",
    "    \"keep\",\n",
    "] = False\n",
    "\n",
    "df_possible_backups = df_model_isoforms_sequences[\n",
    "    df_model_isoforms_sequences[\"uniprot\"].isin(list(to_avoid.values()))\n",
    "]\n",
    "df_possible_backups = df_possible_backups[~df_possible_backups[\"avoid\"]]\n",
    "df_model_isoforms_sequences.loc[\n",
    "    list(df_possible_backups.index),\n",
    "    \"keep\",\n",
    "] = True\n",
    "df_model_isoforms_sequences.loc[\n",
    "    df_model_isoforms_sequences[df_model_isoforms_sequences[\"erythroid\"]].index,\n",
    "    \"keep\",\n",
    "] = True\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences[\n",
    "    df_model_isoforms_sequences[\"keep\"]\n",
    "]\n",
    "\n",
    "lost_ids = set(df_isoforms_sequences[\"uniprot\"].unique()).difference(\n",
    "    set(df_model_isoforms_sequences[\"uniprot\"].unique())\n",
    ")\n",
    "if lost_ids:\n",
    "    lost_ids = df_isoforms_sequences[df_isoforms_sequences[\"uniprot\"].isin(lost_ids)]\n",
    "    df_model_isoforms_sequences = pd.concat(\n",
    "        (df_model_isoforms_sequences, lost_ids[lost_ids[\"canonical\"]]), axis=0\n",
    "    )\n",
    "print()\n",
    "print(\n",
    "    df_model_isoforms_sequences[[\"canonical\", \"erythroid\", \"backup\", \"avoid\"]].sum(\n",
    "        axis=0\n",
    "    )\n",
    ")\n",
    "print(f\"Total: {len(df_model_isoforms_sequences)}\")\n",
    "\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences.loc[\n",
    "    :, [\"uniprot\", \"sequence.id\", \"sequence\"]\n",
    "].copy()\n",
    "df_sequence_data = (\n",
    "    df_model_mappings.merge(\n",
    "        df_model_isoforms_sequences, left_on=\"uniprot\", right_on=\"uniprot\"\n",
    "    )\n",
    "    .loc[:, [\"genes\", \"uniprot\", \"sequence.id\", \"sequence\"]]\n",
    "    .copy()\n",
    ")\n",
    "df_sequence_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "73e577b4-0524-458d-b169-84a4c38ab585",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins: 2\n",
      "Number of isoforms: 5\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein.id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Slc11a2</td>\n",
       "      <td>P49282</td>\n",
       "      <td>P49282-1</td>\n",
       "      <td>MVLDPKEKMPDDGASGDHGDSASLGAINPAYSNSSLPHSTGDSEEP...</td>\n",
       "      <td>Slc11a2_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Slc26a6</td>\n",
       "      <td>Q8CIW6</td>\n",
       "      <td>Q8CIW6-1</td>\n",
       "      <td>MGLPDGSDQGTHQTQALLSAAQEMELQRRDYHVERPLLNQEQLEDL...</td>\n",
       "      <td>Slc26a6_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Slc9a6</td>\n",
       "      <td>A1L3P4</td>\n",
       "      <td>A1L3P4</td>\n",
       "      <td>MAGARRGWRLAPVRRGVCGPRARPLMRPLWLLFAVSFFGWTGALDG...</td>\n",
       "      <td>Slc9a6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Cul4b</td>\n",
       "      <td>A2A432</td>\n",
       "      <td>A2A432-1</td>\n",
       "      <td>MSRSTRSKERRENDTDSEDNSSETSNQERRRCRQGPPRPPYPPLLP...</td>\n",
       "      <td>Cul4b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ptpn3</td>\n",
       "      <td>A2ALK8</td>\n",
       "      <td>A2ALK8</td>\n",
       "      <td>MTSRLRALGGRINNTRTSELPKEKTRSEVICSIRFLDGLVQTFKVN...</td>\n",
       "      <td>Ptpn3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>Minpp1</td>\n",
       "      <td>Q9Z2L6</td>\n",
       "      <td>Q9Z2L6</td>\n",
       "      <td>MLRGARSHLPASVAPAAVLAAALLSSFARCSLPGRGDPVASVLSPY...</td>\n",
       "      <td>Minpp1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>825</th>\n",
       "      <td>Pmm2</td>\n",
       "      <td>Q9Z2M7</td>\n",
       "      <td>Q9Z2M7</td>\n",
       "      <td>MATLCLFDMDGTLTAPRQKITEEMDGFLQKLRQKTKIGVVGGSDFE...</td>\n",
       "      <td>Pmm2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>Pck1</td>\n",
       "      <td>Q9Z2V4</td>\n",
       "      <td>Q9Z2V4</td>\n",
       "      <td>MPPQLHNGLDFSAKVIQGSLDSLPQAVRKFVEGNAQLCQPEYIHIC...</td>\n",
       "      <td>Pck1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>Dnpep</td>\n",
       "      <td>Q9Z2W0</td>\n",
       "      <td>Q9Z2W0</td>\n",
       "      <td>MAMNGRARKEAIQATARELLKFVNRSPSPFHVVAECRSRLLQAGFR...</td>\n",
       "      <td>Dnpep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>828</th>\n",
       "      <td>Slc22a4</td>\n",
       "      <td>Q9Z306</td>\n",
       "      <td>Q9Z306</td>\n",
       "      <td>MRDYDEVIAFLGEWGPFQRLIFFLLSASIIPNGFNGMSVVFLAGTP...</td>\n",
       "      <td>Slc22a4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genes uniprot sequence.id  \\\n",
       "0    Slc11a2  P49282    P49282-1   \n",
       "2    Slc26a6  Q8CIW6    Q8CIW6-1   \n",
       "5     Slc9a6  A1L3P4      A1L3P4   \n",
       "6      Cul4b  A2A432    A2A432-1   \n",
       "7      Ptpn3  A2ALK8      A2ALK8   \n",
       "..       ...     ...         ...   \n",
       "824   Minpp1  Q9Z2L6      Q9Z2L6   \n",
       "825     Pmm2  Q9Z2M7      Q9Z2M7   \n",
       "826     Pck1  Q9Z2V4      Q9Z2V4   \n",
       "827    Dnpep  Q9Z2W0      Q9Z2W0   \n",
       "828  Slc22a4  Q9Z306      Q9Z306   \n",
       "\n",
       "                                              sequence protein.id  \n",
       "0    MVLDPKEKMPDDGASGDHGDSASLGAINPAYSNSSLPHSTGDSEEP...  Slc11a2_1  \n",
       "2    MGLPDGSDQGTHQTQALLSAAQEMELQRRDYHVERPLLNQEQLEDL...  Slc26a6_1  \n",
       "5    MAGARRGWRLAPVRRGVCGPRARPLMRPLWLLFAVSFFGWTGALDG...     Slc9a6  \n",
       "6    MSRSTRSKERRENDTDSEDNSSETSNQERRRCRQGPPRPPYPPLLP...      Cul4b  \n",
       "7    MTSRLRALGGRINNTRTSELPKEKTRSEVICSIRFLDGLVQTFKVN...      Ptpn3  \n",
       "..                                                 ...        ...  \n",
       "824  MLRGARSHLPASVAPAAVLAAALLSSFARCSLPGRGDPVASVLSPY...     Minpp1  \n",
       "825  MATLCLFDMDGTLTAPRQKITEEMDGFLQKLRQKTKIGVVGGSDFE...       Pmm2  \n",
       "826  MPPQLHNGLDFSAKVIQGSLDSLPQAVRKFVEGNAQLCQPEYIHIC...       Pck1  \n",
       "827  MAMNGRARKEAIQATARELLKFVNRSPSPFHVVAECRSRLLQAGFR...      Dnpep  \n",
       "828  MRDYDEVIAFLGEWGPFQRLIFFLLSASIIPNGFNGMSVVFLAGTP...    Slc22a4  \n",
       "\n",
       "[826 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_key = \"uniprot\"\n",
    "protein_id_key = (\n",
    "    \"sequence.id.genes\"  # genes, uniprot, sequence.id, or sequence.id.genes are best,\n",
    ")\n",
    "unique_gene_to_protein_map = True\n",
    "isoform_transform = False\n",
    "df_sequence_data = create_sequence_table(\n",
    "    df_sequence_data=df_sequence_data,\n",
    "    mapping_key=mapping_key,\n",
    "    isoform_transform=isoform_transform,\n",
    ")\n",
    "ordered_isoform_ids = df_sequence_data[df_sequence_data[\"uniprot\"].duplicated(False)][\n",
    "    \"sequence.id\"\n",
    "]\n",
    "df_isoforms = df_sequence_data[\n",
    "    df_sequence_data[\"sequence.id\"].isin(ordered_isoform_ids)\n",
    "].copy()\n",
    "print(f\"Number of proteins: {len(df_isoforms[mapping_key].unique())}\")\n",
    "print(f\"Number of isoforms: {len(df_isoforms['sequence.id'].unique())}\")\n",
    "df_sequence_data = df_sequence_data.set_index(\"sequence.id\")\n",
    "df_sequence_data = pd.concat(\n",
    "    (\n",
    "        df_sequence_data.loc[ordered_isoform_ids],\n",
    "        df_sequence_data.loc[df_sequence_data.index.difference(ordered_isoform_ids)],\n",
    "    ),\n",
    "    axis=0,\n",
    ")\n",
    "df_sequence_data = df_sequence_data.reset_index(drop=False)\n",
    "df_sequence_data = df_sequence_data.loc[\n",
    "    :, [\"genes\", \"uniprot\", \"sequence.id\", \"sequence\"]\n",
    "].copy()\n",
    "# print(df_isoforms[mapping_key])\n",
    "if protein_id_key == \"sequence.id.genes\":\n",
    "    protein_id_key = \"protein.id\"\n",
    "    sequence_id_updates = df_model_mappings.set_index(\"uniprot\")[\"genes\"].to_dict()\n",
    "    df_sequence_data[\"protein.id\"] = df_sequence_data[\"sequence.id\"].apply(\n",
    "        lambda seq_id: \"_\".join(\n",
    "            [sequence_id_updates.get(x, x) for x in seq_id.split(\"-\")]\n",
    "        )\n",
    "    )\n",
    "    df_isoforms[\"protein.id\"] = df_isoforms[\"sequence.id\"].apply(\n",
    "        lambda seq_id: \"_\".join(\n",
    "            [sequence_id_updates.get(x, x) for x in seq_id.split(\"-\")]\n",
    "        )\n",
    "    )\n",
    "    ids_to_fix = df_sequence_data[\n",
    "        ~df_sequence_data[\"sequence.id\"].isin(df_isoforms[\"sequence.id\"])\n",
    "    ].index\n",
    "    df_sequence_data.loc[ids_to_fix, \"protein.id\"] = df_sequence_data.loc[\n",
    "        ids_to_fix, \"protein.id\"\n",
    "    ].apply(lambda x: x.split(\"_\")[0])\n",
    "\n",
    "# Use to remove duplicates\n",
    "if unique_gene_to_protein_map:\n",
    "    df_sequence_data = df_sequence_data.drop_duplicates(\n",
    "        subset=[\"uniprot\"],\n",
    "        keep=\"first\",\n",
    "    )\n",
    "    protein_id_key = \"genes\"\n",
    "\n",
    "df_sequence_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3800b409-4eef-4530-b09f-72499523595d",
   "metadata": {},
   "source": [
    "###  List all unique proteins, complexes, and enzymes\n",
    "#### Option 1: Initialize draft tables\n",
    "1. The draft tables are created and used to initialize the draft PC-model.\n",
    "    * The protein table can be used to initialize proteins and their molar weight ($\\textbf{d}$ vector).\n",
    "    * The complex table can be used to initialize complexes with their subunit stoichiometry ($\\textbf{C}$ matrix).\n",
    "        * All stoichiometric coefficients are initialized at a value of one.\n",
    "    * The enzyme table can be used to initialize enzymes with their effective rate constants ($\\textbf{K}_\\mathrm{eff}$ matrix).\n",
    "        * All $k_\\mathrm{eff}$ values are initialized at average rate constant of 65 $s^{-1}$  (or equivalently, 234000 $hr^{-1})$.\n",
    "\n",
    "2. The draft tables are made to be facilitate curation and data replacement. Therefore, the draft PC-model is exported with the draft tables. \n",
    "3. A refined PC-model can be created using the curated tables. \n",
    "\n",
    "#### Option 2: Load tables from files\n",
    "4. The formation of a draft model can be skipped if the curated tables already exist. They can be loaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "18560139-b484-4255-bffa-deed41fa153a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pcmodel_tables = {}\n",
    "replace_compartments = {\n",
    "    # Cytosol:extracellular --> plasma membrane\n",
    "    \"c\": \"c\",\n",
    "    \"ce\": \"pm\",\n",
    "    \"e\": \"e\",  # Most extracellular reactions that occur are due to proteins bound to the external side of them membrane.\n",
    "}\n",
    "\n",
    "# Convert all protein compartments to one compartment\n",
    "simplify_compartments = True\n",
    "prefix = True\n",
    "optional_columns = True\n",
    "map_human_to_organism = True\n",
    "\n",
    "# Enzyme values for new tables\n",
    "max_weight_fraction = 100\n",
    "enzyme_keff_base = DEFAULT_KEFF\n",
    "\n",
    "dict_of_id_keys = {\n",
    "    \"proteins\": protein_id_key,\n",
    "    \"complexes\": None,\n",
    "    \"enzymes\": \"reactions\",\n",
    "}\n",
    "\n",
    "# Provide filepaths to load a specific model\n",
    "model_filepaths = {\n",
    "    # \"proteins\": model_dirpath / f\"pcmodel_{model}_proteins.tsv\",\n",
    "    # \"complexes\": model_dirpath / f\"pcmodel_{model}_complexes.tsv\",\n",
    "    # \"enzymes\": model_dirpath / f\"pcmodel_{model}_enzymes.tsv\",\n",
    "    # \"complex_keffs\": model_dirpath / f\"pcmodel_{model}_complex_keffs.tsv\",\n",
    "    # \"enzyme_keffs\": model_dirpath / f\"pcmodel_{model}_enzyme_keffs.tsv\",\n",
    "    # \"constraints_proteins\": model_dirpath / f\"pcmodel_{model}_constraints_proteins.tsv\",\n",
    "    # \"constraints_reactions\": model_dirpath / f\"pcmodel_{model}_constraints_reactions.tsv\",\n",
    "    # \"constraints_additional\": model_dirpath / f\"pcmodel_{model}_constraints_additional.tsv\",\n",
    "}\n",
    "\n",
    "# Provide filepaths to general files\n",
    "filepaths = {\n",
    "    \"proteins\": overlay_dirpath / \"pcmodel_proteins.tsv\",\n",
    "    \"complexes\": overlay_dirpath / \"pcmodel_complexes.tsv\",\n",
    "    # \"enzymes\": overlay_dirpath / \"pcmodel_enzymes.tsv\",\n",
    "    # \"complex_keffs\": overlay_dirpath / \"pcmodel_complex_keffs.tsv\",\n",
    "    # \"enzyme_keffs\": overlay_dirpath / \"pcmodel_enzyme_keffs.tsv\",\n",
    "    # \"constraints_proteins\": overlay_dirpath / f\"pcmodel_constraints_proteins.tsv\",\n",
    "    # \"constraints_reactions\": overlay_dirpath / f\"pcmodel_constraints_reactions.tsv\",\n",
    "    # \"constraints_additional\": overlay_dirpath / f\"pcmodel_constraints_additional.tsv\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "792a4744-149f-4fe3-a274-bfb7c2cb97b1",
   "metadata": {},
   "source": [
    "##### Proteins\n",
    "###### Map protein homologs for non-human organism (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8c6b108f-17ff-4aad-a3a1-7e383e131bfb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genes/proteins removed due to inability to map: 11\n",
      "Mapping Summary\n",
      "genes      827\n",
      "protein    827\n",
      "uniprot    827\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "table_type = \"proteins\"\n",
    "if organism != \"Human\" and map_human_to_organism:\n",
    "    df_human_to_organism = pd.read_csv(\n",
    "        database_dirpath / f\"Human{organism}Mapping.tsv\", sep=\"\\t\", index_col=None\n",
    "    )\n",
    "    df_human = pd.read_csv(\n",
    "        str(filepaths[table_type]).replace(\"Mouse\", \"Human\"), sep=\"\\t\", index_col=None\n",
    "    )\n",
    "    gene_key = \"genes\"\n",
    "    org_gene_key = \"mgnc.symbol\"\n",
    "\n",
    "    # Correct for dashes in genes\n",
    "    df_human_to_organism[org_gene_key] = df_human_to_organism[org_gene_key].apply(\n",
    "        lambda item: item.replace(\"-\", \"_\")\n",
    "    )\n",
    "    df_organism = df_human.copy()\n",
    "\n",
    "    # Map genes\n",
    "    df = df_human_to_organism[[gene_key, org_gene_key]].copy()\n",
    "    df = df.groupby(df.columns[0], as_index=False).agg(lambda item: list(item))\n",
    "    mapping_dict = df.set_index(df.columns[0])[df.columns[-1]].to_dict()\n",
    "    df_organism[gene_key] = df_human[gene_key].apply(\n",
    "        lambda item: mapping_dict[item] if item in mapping_dict else f\"TO_REMOVE:{item}\"\n",
    "    )\n",
    "    # Map uniprot\n",
    "    df = df_human_to_organism[[\"uniprot\", f\"uniprot.{organism.lower()}\"]].copy()\n",
    "    df = df.groupby(df.columns[0], as_index=False).agg(lambda item: list(item))\n",
    "    mapping_dict = df.set_index(df.columns[0])[df.columns[-1]].to_dict()\n",
    "    df_organism[\"uniprot\"] = df_human[\"uniprot\"].apply(\n",
    "        lambda item: mapping_dict[item] if item in mapping_dict else pd.NA\n",
    "    )\n",
    "    df_organism = df_organism.explode([gene_key, \"uniprot\"])\n",
    "\n",
    "    # Map proteins using genes\n",
    "    df_organism[\"protein\"] = df_organism[gene_key].apply(\n",
    "        lambda item: (\n",
    "            DEFAULT_PREFIX_SUFFIX_VALUES[table_type][\"prefix.metabolite\"] + item\n",
    "            if \"TO_REMOVE\" not in item\n",
    "            else pd.NA\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Identify genes/proteins to remove from the model due to an inability to map them\n",
    "    removed = df_organism[gene_key][\n",
    "        df_organism[gene_key].str.startswith(\"TO_REMOVE:\")\n",
    "    ].apply(lambda item: item.replace(\"TO_REMOVE:\", \"\"))\n",
    "    df_organism.loc[removed.index, gene_key] = pd.NA\n",
    "    removed = removed.drop_duplicates().reset_index(drop=True)\n",
    "    print(f\"Number of genes/proteins removed due to inability to map: {len(removed)}\")\n",
    "\n",
    "    keys = [gene_key, strip_plural(table_type), \"uniprot\"]\n",
    "    df_organism = df_organism.dropna(\n",
    "        subset=[\"protein\", \"uniprot\"], axis=0\n",
    "    ).drop_duplicates()\n",
    "    df_organism = df_organism.sort_values(by=keys).reset_index(drop=True)\n",
    "    # Those that aren't unique are due to having more than one compartment\n",
    "    print(f\"Mapping Summary\\n{df_organism[keys].nunique()}\")\n",
    "    df_organism.to_csv(filepaths[table_type], sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec9766ff-41be-42fe-944a-2b14d4dbf6b1",
   "metadata": {},
   "source": [
    "###### Create protein table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "acb342b9-c941-47ab-b80f-5b11fec4fb24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded from main RBC-GEM file\n",
      "Discrepancies from organism mapping: 1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>compartment</th>\n",
       "      <th>proteins</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>protein_A4galt</th>\n",
       "      <td>A4galt</td>\n",
       "      <td>Q67BJ4</td>\n",
       "      <td>Q67BJ4</td>\n",
       "      <td>MGISCSHLEETMSKPPDCLLRMLRGTPRQRVFTFFIISFKFMFLIS...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_A4galt_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Aars1</th>\n",
       "      <td>Aars1</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "      <td>MDATLTAREIRERFINFFRRNEHTYVHSSATIPLDDPTLLFANAGM...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Aars1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Aasdhppt</th>\n",
       "      <td>Aasdhppt</td>\n",
       "      <td>Q9CQF6</td>\n",
       "      <td>Q9CQF6-1</td>\n",
       "      <td>MVFPAKRLCVVPSMEGVRWAFSCGTWLPSRAEWLLAMRSIQPEEKE...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Aasdhppt_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Abca1</th>\n",
       "      <td>Abca1</td>\n",
       "      <td>P41233</td>\n",
       "      <td>P41233</td>\n",
       "      <td>MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Abca1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Abca7</th>\n",
       "      <td>Abca7</td>\n",
       "      <td>Q91V24</td>\n",
       "      <td>Q91V24</td>\n",
       "      <td>MALGTQLMLLLWKNYTYRRRQPIQLLVELLWPLFLFFILVAVRHSH...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Abca7_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Yes1</th>\n",
       "      <td>Yes1</td>\n",
       "      <td>Q04736</td>\n",
       "      <td>Q04736</td>\n",
       "      <td>MGCIKSKENKSPAIKYTPENLTEPVSPSASHYGVEHATVAPTSSTK...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Yes1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc2</th>\n",
       "      <td>Zdhhc2</td>\n",
       "      <td>P59267</td>\n",
       "      <td>P59267</td>\n",
       "      <td>MAPSGSGGVRRRCRRVLYWIPVVFISLLLGWSYYAYAIQLCIVSME...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Zdhhc2_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc20</th>\n",
       "      <td>Zdhhc20</td>\n",
       "      <td>Q5Y5T1</td>\n",
       "      <td>Q5Y5T1-1</td>\n",
       "      <td>MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVSTISRTG...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Zdhhc20_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc3</th>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>Q8R173</td>\n",
       "      <td>Q8R173</td>\n",
       "      <td>MMLIPTHHFRDIERKPEYLQPEKCAPPPFPGPAGAMWFIRDGCGIA...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Zdhhc3_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc5</th>\n",
       "      <td>Zdhhc5</td>\n",
       "      <td>Q8VDZ4</td>\n",
       "      <td>Q8VDZ4-1</td>\n",
       "      <td>MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLNVSPA...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_Zdhhc5_pc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     genes uniprot sequence.id  \\\n",
       "protein                                          \n",
       "protein_A4galt      A4galt  Q67BJ4      Q67BJ4   \n",
       "protein_Aars1        Aars1  Q8BGQ7      Q8BGQ7   \n",
       "protein_Aasdhppt  Aasdhppt  Q9CQF6    Q9CQF6-1   \n",
       "protein_Abca1        Abca1  P41233      P41233   \n",
       "protein_Abca7        Abca7  Q91V24      Q91V24   \n",
       "...                    ...     ...         ...   \n",
       "protein_Yes1          Yes1  Q04736      Q04736   \n",
       "protein_Zdhhc2      Zdhhc2  P59267      P59267   \n",
       "protein_Zdhhc20    Zdhhc20  Q5Y5T1    Q5Y5T1-1   \n",
       "protein_Zdhhc3      Zdhhc3  Q8R173      Q8R173   \n",
       "protein_Zdhhc5      Zdhhc5  Q8VDZ4    Q8VDZ4-1   \n",
       "\n",
       "                                                           sequence  \\\n",
       "protein                                                               \n",
       "protein_A4galt    MGISCSHLEETMSKPPDCLLRMLRGTPRQRVFTFFIISFKFMFLIS...   \n",
       "protein_Aars1     MDATLTAREIRERFINFFRRNEHTYVHSSATIPLDDPTLLFANAGM...   \n",
       "protein_Aasdhppt  MVFPAKRLCVVPSMEGVRWAFSCGTWLPSRAEWLLAMRSIQPEEKE...   \n",
       "protein_Abca1     MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...   \n",
       "protein_Abca7     MALGTQLMLLLWKNYTYRRRQPIQLLVELLWPLFLFFILVAVRHSH...   \n",
       "...                                                             ...   \n",
       "protein_Yes1      MGCIKSKENKSPAIKYTPENLTEPVSPSASHYGVEHATVAPTSSTK...   \n",
       "protein_Zdhhc2    MAPSGSGGVRRRCRRVLYWIPVVFISLLLGWSYYAYAIQLCIVSME...   \n",
       "protein_Zdhhc20   MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVSTISRTG...   \n",
       "protein_Zdhhc3    MMLIPTHHFRDIERKPEYLQPEKCAPPPFPGPAGAMWFIRDGCGIA...   \n",
       "protein_Zdhhc5    MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLNVSPA...   \n",
       "\n",
       "                 compartment             proteins  \n",
       "protein                                            \n",
       "protein_A4galt            pc    protein_A4galt_pc  \n",
       "protein_Aars1             pc     protein_Aars1_pc  \n",
       "protein_Aasdhppt          pc  protein_Aasdhppt_pc  \n",
       "protein_Abca1             pc     protein_Abca1_pc  \n",
       "protein_Abca7             pc     protein_Abca7_pc  \n",
       "...                      ...                  ...  \n",
       "protein_Yes1              pc      protein_Yes1_pc  \n",
       "protein_Zdhhc2            pc    protein_Zdhhc2_pc  \n",
       "protein_Zdhhc20           pc   protein_Zdhhc20_pc  \n",
       "protein_Zdhhc3            pc    protein_Zdhhc3_pc  \n",
       "protein_Zdhhc5            pc    protein_Zdhhc5_pc  \n",
       "\n",
       "[826 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"proteins\"\n",
    "try:\n",
    "    # Try loading previously build model proteins\n",
    "    df_proteins = pd.read_csv(model_filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    print(\"Loaded from previously generated file\")\n",
    "except (FileNotFoundError, KeyError):\n",
    "    # Otherwise try using main RBC-GEM files to make model proteins\n",
    "    try:\n",
    "        df_proteins = pd.read_csv(filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    except (FileNotFoundError, KeyError):\n",
    "        # Otherwise, make from scratch\n",
    "        df_proteins = create_protein_table(\n",
    "            model,\n",
    "            df_sequence_data,\n",
    "            id_key=dict_of_id_keys.get(table_type),\n",
    "            prefix=prefix,\n",
    "            optional_columns=optional_columns,\n",
    "            annotation_columns=[\n",
    "                \"uniprot\",\n",
    "            ],\n",
    "            replace_compartments=replace_compartments,\n",
    "        )\n",
    "        print(\"Created new table\")\n",
    "        # Create column for identifiers if None exists, or if compartments were replaced\n",
    "        if not isoform_transform:\n",
    "            df_proteins[df_proteins[table_type].duplicated(False)]\n",
    "    else:\n",
    "        df_proteins = df_sequence_data.merge(\n",
    "            df_proteins[[\"uniprot\", \"compartment\"]],\n",
    "            left_on=\"uniprot\",\n",
    "            right_on=\"uniprot\",\n",
    "            how=\"left\",\n",
    "        )\n",
    "        df_proteins[\"protein\"] = df_proteins[protein_id_key].apply(\n",
    "            lambda x: f\"protein_{x}\"\n",
    "        )\n",
    "        df_proteins = df_proteins.drop(\"protein.id\", axis=1)\n",
    "        df_proteins = df_proteins[\n",
    "            df_proteins[\"genes\"].isin(model.genes.list_attr(\"id\"))\n",
    "        ].reset_index(drop=True)\n",
    "        print(\"Loaded from main RBC-GEM file\")\n",
    "\n",
    "if simplify_compartments:\n",
    "    df_proteins = df_proteins.groupby([\"genes\", \"protein\"]).agg(\n",
    "        lambda values: \";\".join(\n",
    "            [str(value) for value in list(values.dropna().unique())]\n",
    "        )\n",
    "    )\n",
    "    df_proteins[\"compartment\"] = DEFAULT_PROTEOME_COMPARTMENT\n",
    "    df_proteins = df_proteins.reset_index(drop=False)\n",
    "\n",
    "df_proteins[\"proteins\"] = df_proteins[[strip_plural(table_type), \"compartment\"]].apply(\n",
    "    lambda x: \"_\".join(x.values), axis=1\n",
    ")\n",
    "df_proteins = df_proteins.set_index(strip_plural(table_type))\n",
    "pcmodel_tables[table_type] = df_proteins.copy()\n",
    "\n",
    "if organism != \"Human\" and map_human_to_organism:\n",
    "    discrepancies = df_organism[~df_organism[\"genes\"].isin(df_proteins[\"genes\"])]\n",
    "    if not discrepancies.empty:\n",
    "        # A discrepancy may arise if a protein was deleted from UniProt but has not yet been recorded in the organism database.\n",
    "        # This variable can be used for manually checking\n",
    "        print(f\"Discrepancies from organism mapping: {len(discrepancies)}\")\n",
    "\n",
    "df_proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610036dd-56de-4d9b-8a5b-147d569624ed",
   "metadata": {},
   "source": [
    "##### Complexes\n",
    "###### Map protein homologs for non-human organism (optional)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c6231352-5d76-45de-95f3-110d18126cbe",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of identical complexes kept seperated : 0\n",
      "Number of identical complexes merged together: 1\n",
      "Number of complexes dropped due to mising proteins or inability to map: 23\n",
      "Mapping Summary\n",
      "complex     888\n",
      "subunits    874\n",
      "genes       874\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "table_type = \"complexes\"\n",
    "if organism != \"Human\" and map_human_to_organism:\n",
    "    df_human_to_organism = pd.read_csv(\n",
    "        database_dirpath / f\"Human{organism}Mapping.tsv\", sep=\"\\t\", index_col=None\n",
    "    )\n",
    "    df_human = pd.read_csv(\n",
    "        str(filepaths[table_type]).replace(\"Mouse\", \"Human\"), sep=\"\\t\", index_col=None\n",
    "    )\n",
    "    gene_key = \"genes\"\n",
    "    org_gene_key = \"mgnc.symbol\"\n",
    "\n",
    "    # Correct for dashes in genes\n",
    "    df_human_to_organism[org_gene_key] = df_human_to_organism[org_gene_key].apply(\n",
    "        lambda item: item.replace(\"-\", \"_\")\n",
    "    )\n",
    "    df_organism = df_human.copy()\n",
    "\n",
    "    # Map genes\n",
    "    df = df_human_to_organism[[gene_key, org_gene_key]].copy()\n",
    "    df = df.groupby(df.columns[0], as_index=False).agg(lambda item: list(item))\n",
    "    mapping_dict = df.set_index(df.columns[0])[df.columns[-1]].to_dict()\n",
    "    df_organism[gene_key] = df_human[gene_key].apply(\n",
    "        lambda item: [\n",
    "            set(mapping_dict[x]) if x in mapping_dict else f\"TO_REMOVE:{x}\"\n",
    "            for x in split_string(item)\n",
    "        ]\n",
    "    )\n",
    "    for idx, row in df_organism[gene_key].items():\n",
    "        if any([\"TO_REMOVE:\" in x for x in row]):\n",
    "            # Complexes that cannot be resolved due to a missing gene\n",
    "            row = [set([\"TO_REMOVE\"])]\n",
    "        else:\n",
    "            row = set([build_string(sorted(x)) for x in list(itertools.product(*row))])\n",
    "        df_organism.loc[idx, gene_key] = [row]\n",
    "    df_organism = df_organism.explode(\"genes\").explode(\"genes\").reset_index(drop=True)\n",
    "\n",
    "    # Map subunits to genes\n",
    "    df_organism[\"subunits\"] = df_organism[gene_key].apply(\n",
    "        lambda item: build_string(\n",
    "            [\n",
    "                DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.metabolite\"] + x\n",
    "                for x in split_string(item)\n",
    "            ]\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # Map complexes using genes\n",
    "    df = df_human_to_organism[[org_gene_key, gene_key]].copy()\n",
    "    df = df.groupby(df.columns[0], as_index=False).agg(lambda item: list(item))\n",
    "    mapping_dict = df.set_index(df.columns[0])[df.columns[-1]].to_dict()\n",
    "    for idx, row in df_organism.iterrows():\n",
    "        if \"TO_REMOVE\" not in row[\"genes\"]:\n",
    "            replacements = {\n",
    "                item: k for k in split_string(row[\"genes\"]) for item in mapping_dict[k]\n",
    "            }\n",
    "            new_id = row[\"complex\"].split(\"_\")[:2]\n",
    "            new_id += [\n",
    "                (\n",
    "                    replacements[item]\n",
    "                    if item in replacements\n",
    "                    else item.lower().capitalize()\n",
    "                )\n",
    "                for item in row[\"complex\"].split(\"_\")[2:]\n",
    "            ]\n",
    "            new_id = \"_\".join(new_id)\n",
    "            df_organism.loc[idx, \"complex\"] = new_id\n",
    "\n",
    "    # Identify genes/proteins to remove from the model due to an inability to map them\n",
    "    missing_proteins = df_organism[\n",
    "        df_organism[\"subunits\"].apply(\n",
    "            lambda x: len(set(x.split(\";\")).difference(df_proteins.index)) != 0\n",
    "        )\n",
    "    ]\n",
    "    removed = pd.concat(\n",
    "        (\n",
    "            df_organism[df_organism[\"genes\"] == \"TO_REMOVE\"][\"complex\"],\n",
    "            missing_proteins[\"complex\"],\n",
    "        )\n",
    "    )\n",
    "    df_organism.loc[removed.index, gene_key] = pd.NA\n",
    "    df_organism = (\n",
    "        df_organism.dropna(subset=[\"genes\"], axis=0).drop_duplicates().fillna(\"\")\n",
    "    )\n",
    "    removed = removed.drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "    if not df_organism[\n",
    "        df_organism.duplicated(subset=[\"complex\", \"compartment\"], keep=False)\n",
    "    ].empty:\n",
    "        duplicated_ids_w_diff_genes = []\n",
    "        # Duplicates may exist, extra steps to reconcile\n",
    "        # Join all duplicate entries where values differ\n",
    "        start = len(df_organism)\n",
    "        df_organism = df_organism.groupby(\n",
    "            [\"complex\", \"compartment\", \"subunits\", \"genes\"], as_index=False\n",
    "        ).agg(lambda x: x.unique()[0] if x.nunique() == 1 else \";\".join(x))\n",
    "        # Unite reactions and remove duplicates\n",
    "        df_organism[\"reactions\"] = df_organism[\"reactions\"].apply(\n",
    "            lambda x: build_string(x.split(\";\"))\n",
    "        )\n",
    "        # Any remaining complexes that are duplicated need to be seperated into different entries\n",
    "        complex_counts = df_organism.groupby([\"complex\", \"compartment\"])[\n",
    "            \"complex\"\n",
    "        ].value_counts()\n",
    "        for (duplicated_id, comp), count in complex_counts[complex_counts > 1].items():\n",
    "            duplicated_ids_w_diff_genes += [duplicated_id]\n",
    "            # Find all rows with duplicated IDs, get indices\n",
    "            for ind in df_organism[df_organism[\"complex\"] == duplicated_id].index:\n",
    "                # Use genes to replace duplicated labels as they are expected to be different.\n",
    "                df_organism.loc[ind, \"complex\"] = \"_\".join(\n",
    "                    duplicated_id.split(\"_\")[:2]\n",
    "                    + [df_organism.loc[ind, \"genes\"].replace(\";\", \"_\")]\n",
    "                )\n",
    "        print(\n",
    "            f\"Number of identical complexes kept seperated : {len(duplicated_ids_w_diff_genes)}\"\n",
    "        )\n",
    "        print(\n",
    "            f\"Number of identical complexes merged together: {start - len(df_organism)}\"\n",
    "        )\n",
    "\n",
    "    # Check that proteins exist, remove those that having missing proteins\n",
    "    print(\n",
    "        f\"Number of complexes dropped due to mising proteins or inability to map: {len(removed)}\"\n",
    "    )\n",
    "    keys = [\"complex\", \"subunits\", \"genes\"]\n",
    "    df_organism = df_organism.sort_values(by=keys).reset_index(drop=True)\n",
    "    # Those that aren't unique are due to having more than one compartment\n",
    "    print(f\"Mapping Summary\\n{df_organism[keys].nunique()}\")\n",
    "    df_organism.to_csv(filepaths[table_type], sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfb74d96-8899-46c0-a1ee-a5039cb1a479",
   "metadata": {},
   "source": [
    "###### Create complex table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7190d981-1b98-486b-9e06-8c9344d6fd63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded from main RBC-GEM file\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subunits</th>\n",
       "      <th>compartment</th>\n",
       "      <th>reactions</th>\n",
       "      <th>genes</th>\n",
       "      <th>coefficients</th>\n",
       "      <th>cofactors</th>\n",
       "      <th>notes</th>\n",
       "      <th>complexes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>complex</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_A4galt</th>\n",
       "      <td>protein_A4galt_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>A4GALTc</td>\n",
       "      <td>A4galt</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_A4galt_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Aars1</th>\n",
       "      <td>protein_Aars1_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ALATRS</td>\n",
       "      <td>Aars1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_Aars1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Aasdhppt</th>\n",
       "      <td>protein_Aasdhppt_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>AASDHPPT</td>\n",
       "      <td>Aasdhppt</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_Aasdhppt_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_HOMOTETRAMER_Abca1</th>\n",
       "      <td>protein_Abca1_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PCFLOPt;PSFLOPt;SMFLOPt</td>\n",
       "      <td>Abca1</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_HOMOTETRAMER_Abca1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Abca7</th>\n",
       "      <td>protein_Abca7_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PCFLOPt;PEFLOPt;PSFLOPt;SMFLOPt</td>\n",
       "      <td>Abca7</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_Abca7_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Yes1</th>\n",
       "      <td>protein_Yes1_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTYK</td>\n",
       "      <td>Yes1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_Yes1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc2</th>\n",
       "      <td>protein_Zdhhc2_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>Zdhhc2</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_Zdhhc2_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc20</th>\n",
       "      <td>protein_Zdhhc20_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>Zdhhc20</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_Zdhhc20_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc3</th>\n",
       "      <td>protein_Zdhhc3_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_Zdhhc3_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc5</th>\n",
       "      <td>protein_Zdhhc5_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>Zdhhc5</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_Zdhhc5_pc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>888 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    subunits compartment  \\\n",
       "complex                                                    \n",
       "cplx_MONOMER_A4galt        protein_A4galt_pc          pc   \n",
       "cplx_MONOMER_Aars1          protein_Aars1_pc          pc   \n",
       "cplx_MONOMER_Aasdhppt    protein_Aasdhppt_pc          pc   \n",
       "cplx_HOMOTETRAMER_Abca1     protein_Abca1_pc          pc   \n",
       "cplx_MONOMER_Abca7          protein_Abca7_pc          pc   \n",
       "...                                      ...         ...   \n",
       "cplx_MONOMER_Yes1            protein_Yes1_pc          pc   \n",
       "cplx_MONOMER_Zdhhc2        protein_Zdhhc2_pc          pc   \n",
       "cplx_MONOMER_Zdhhc20      protein_Zdhhc20_pc          pc   \n",
       "cplx_MONOMER_Zdhhc3        protein_Zdhhc3_pc          pc   \n",
       "cplx_MONOMER_Zdhhc5        protein_Zdhhc5_pc          pc   \n",
       "\n",
       "                                                         reactions     genes  \\\n",
       "complex                                                                        \n",
       "cplx_MONOMER_A4galt                                        A4GALTc    A4galt   \n",
       "cplx_MONOMER_Aars1                                          ALATRS     Aars1   \n",
       "cplx_MONOMER_Aasdhppt                                     AASDHPPT  Aasdhppt   \n",
       "cplx_HOMOTETRAMER_Abca1                    PCFLOPt;PSFLOPt;SMFLOPt     Abca1   \n",
       "cplx_MONOMER_Abca7                 PCFLOPt;PEFLOPt;PSFLOPt;SMFLOPt     Abca7   \n",
       "...                                                            ...       ...   \n",
       "cplx_MONOMER_Yes1                                           PROTYK      Yes1   \n",
       "cplx_MONOMER_Zdhhc2      PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0    Zdhhc2   \n",
       "cplx_MONOMER_Zdhhc20     PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0   Zdhhc20   \n",
       "cplx_MONOMER_Zdhhc3      PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0    Zdhhc3   \n",
       "cplx_MONOMER_Zdhhc5      PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0    Zdhhc5   \n",
       "\n",
       "                        coefficients cofactors            notes  \\\n",
       "complex                                                           \n",
       "cplx_MONOMER_A4galt                1            ASSUMED MONOMER   \n",
       "cplx_MONOMER_Aars1                 1                              \n",
       "cplx_MONOMER_Aasdhppt              1                              \n",
       "cplx_HOMOTETRAMER_Abca1            4                              \n",
       "cplx_MONOMER_Abca7                 1            ASSUMED MONOMER   \n",
       "...                              ...       ...              ...   \n",
       "cplx_MONOMER_Yes1                  1            ASSUMED MONOMER   \n",
       "cplx_MONOMER_Zdhhc2                1                              \n",
       "cplx_MONOMER_Zdhhc20               1            ASSUMED MONOMER   \n",
       "cplx_MONOMER_Zdhhc3                1                              \n",
       "cplx_MONOMER_Zdhhc5                1                              \n",
       "\n",
       "                                          complexes  \n",
       "complex                                              \n",
       "cplx_MONOMER_A4galt          cplx_MONOMER_A4galt_pc  \n",
       "cplx_MONOMER_Aars1            cplx_MONOMER_Aars1_pc  \n",
       "cplx_MONOMER_Aasdhppt      cplx_MONOMER_Aasdhppt_pc  \n",
       "cplx_HOMOTETRAMER_Abca1  cplx_HOMOTETRAMER_Abca1_pc  \n",
       "cplx_MONOMER_Abca7            cplx_MONOMER_Abca7_pc  \n",
       "...                                             ...  \n",
       "cplx_MONOMER_Yes1              cplx_MONOMER_Yes1_pc  \n",
       "cplx_MONOMER_Zdhhc2          cplx_MONOMER_Zdhhc2_pc  \n",
       "cplx_MONOMER_Zdhhc20        cplx_MONOMER_Zdhhc20_pc  \n",
       "cplx_MONOMER_Zdhhc3          cplx_MONOMER_Zdhhc3_pc  \n",
       "cplx_MONOMER_Zdhhc5          cplx_MONOMER_Zdhhc5_pc  \n",
       "\n",
       "[888 rows x 8 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"complexes\"\n",
    "try:\n",
    "    df_complexes = pd.read_csv(model_filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    print(\"Loaded from previously generated file\")\n",
    "except (FileNotFoundError, KeyError):\n",
    "    try:\n",
    "        df_complexes = pd.read_csv(filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    except (FileNotFoundError, KeyError):\n",
    "        genes_to_proteins = (\n",
    "            pcmodel_tables[\"proteins\"]\n",
    "            .groupby([\"genes\"], as_index=True)[\"proteins\"]\n",
    "            .agg(lambda x: build_string(list(x)))\n",
    "            .to_dict()\n",
    "        )\n",
    "        cofactor_genes = {}\n",
    "        # Create table\n",
    "        df_complexes = create_complex_table(\n",
    "            model,\n",
    "            genes_to_proteins=genes_to_proteins,\n",
    "            cofactor_genes=cofactor_genes,\n",
    "            id_key=dict_of_id_keys.get(table_type),\n",
    "            optional_columns=optional_columns,\n",
    "            annotation_columns=[\n",
    "                # \"uniprot\"\n",
    "            ],\n",
    "            replace_compartments=replace_compartments,\n",
    "        )\n",
    "        print(\"Created new table\")\n",
    "    else:\n",
    "        df_complexes = df_complexes[\n",
    "            df_complexes[\"genes\"].apply(\n",
    "                lambda genes: all(\n",
    "                    [model.genes.has_id(gene) for gene in genes.split(\";\")]\n",
    "                )\n",
    "            )\n",
    "        ]\n",
    "        df_complexes[\"reactions\"] = df_complexes[\"reactions\"].apply(\n",
    "            lambda reactions: \";\".join(\n",
    "                [r for r in reactions.split(\";\") if model.reactions.has_id(r)]\n",
    "            )\n",
    "        )\n",
    "        df_complexes = df_complexes[df_complexes[\"reactions\"] != \"\"]\n",
    "        df_complexes = df_complexes.loc[\n",
    "            :,\n",
    "            [\n",
    "                \"complex\",\n",
    "                \"subunits\",\n",
    "                \"compartment\",\n",
    "                \"reactions\",\n",
    "                \"genes\",\n",
    "                \"coefficients\",\n",
    "                \"cofactors\",\n",
    "                \"notes\",\n",
    "            ],\n",
    "        ]\n",
    "        print(\"Loaded from main RBC-GEM file\")\n",
    "\n",
    "    # Address isoform mapping to complexes\n",
    "    isoforms_map = defaultdict(list)\n",
    "    complex_name_update = defaultdict(list)\n",
    "    for x in df_proteins[df_proteins[\"genes\"].duplicated(False)].index:\n",
    "        isoforms_map[x.rsplit(\"_\", maxsplit=1)[0]].append(x)\n",
    "        complex_name_update[\n",
    "            x.rsplit(\"_\", maxsplit=1)[0].replace(\"protein_\", \"\")\n",
    "        ].append(x.replace(\"protein_\", \"\"))\n",
    "    df_isoforms_complexes = df_complexes[\n",
    "        df_complexes[\"subunits\"].apply(\n",
    "            lambda proteins: bool(set(isoforms_map).intersection(proteins.split(\";\")))\n",
    "        )\n",
    "    ]\n",
    "    df_updated_rows = []\n",
    "    for _, row in df_isoforms_complexes.iterrows():\n",
    "        complex_names = [\n",
    "            complex_name\n",
    "            for complex_name in itertools.product(\n",
    "                *[complex_name_update.get(c, [c]) for c in row[\"complex\"].split(\"_\")]\n",
    "            )\n",
    "        ]\n",
    "        combos = [\n",
    "            list(combo)\n",
    "            for combo in itertools.product(\n",
    "                *[\n",
    "                    isoforms_map.get(protein, [protein])\n",
    "                    for protein in row[\"subunits\"].split(\";\")\n",
    "                ]\n",
    "            )\n",
    "        ]\n",
    "\n",
    "        for complex_name, combo in zip(complex_names, combos):\n",
    "            new_row = row.to_dict()\n",
    "            new_row[\"complex\"] = \"_\".join(complex_name)\n",
    "            new_row[\"subunits\"] = \";\".join(combo)\n",
    "            df_updated_rows.append(new_row)\n",
    "\n",
    "    df_complexes = pd.concat(\n",
    "        (\n",
    "            df_complexes[~df_complexes.index.isin(df_isoforms_complexes.index)],\n",
    "            pd.DataFrame(df_updated_rows),\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "\n",
    "if simplify_compartments:\n",
    "    df_complexes = df_complexes.groupby([\"subunits\", \"complex\"]).agg(\n",
    "        lambda values: \";\".join(\n",
    "            [str(value) for value in list(values.dropna().unique())]\n",
    "        )\n",
    "    )\n",
    "    df_complexes[\"compartment\"] = DEFAULT_PROTEOME_COMPARTMENT\n",
    "    df_complexes = df_complexes.reset_index(drop=False)\n",
    "\n",
    "df_complexes[\"complexes\"] = df_complexes[\n",
    "    [strip_plural(table_type), \"compartment\"]\n",
    "].apply(lambda x: \"_\".join(x.values), axis=1)\n",
    "df_complexes[\"subunits\"] = df_complexes[[\"subunits\", \"compartment\"]].apply(\n",
    "    lambda values: \";\".join(\n",
    "        [\n",
    "            \"_\".join((x, values[\"compartment\"])) if not x.endswith(\"_pc\") else x\n",
    "            for x in values[\"subunits\"].split(\";\")\n",
    "        ]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "df_complexes = df_complexes.set_index(strip_plural(table_type))\n",
    "pcmodel_tables[table_type] = df_complexes.copy()\n",
    "if organism != \"Human\" and map_human_to_organism:\n",
    "    discrepancies = df_organism[~df_organism[\"genes\"].isin(df_complexes[\"genes\"])]\n",
    "    if not discrepancies.empty:\n",
    "        # A discrepancy may arise if a protein was deleted from UniProt but has not yet been recorded in the organism database.\n",
    "        # This variable can be used for manually checking\n",
    "        print(f\"Discrepancies from organism mapping: {len(discrepancies)}\")\n",
    "df_complexes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5388d72-79df-4444-886c-82ca4ff0d7ba",
   "metadata": {},
   "source": [
    "##### Enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3ec3f525-31d4-41a4-bd1e-7b0c6242f572",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Created new table\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complexes</th>\n",
       "      <th>compartment</th>\n",
       "      <th>reactions</th>\n",
       "      <th>enzymes</th>\n",
       "      <th>direction</th>\n",
       "      <th>enzyme_keff</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO3ARGH_fwd</th>\n",
       "      <td>cplx_COMPLEX_Agt_Anpep_pc;cplx_COMPLEX_Agt_Dpp...</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO3ARGH</td>\n",
       "      <td>enzyme_ANGIO3ARGH_fwd_pc</td>\n",
       "      <td>forward</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO3ARGH_rev</th>\n",
       "      <td>cplx_COMPLEX_Agt_Anpep_pc;cplx_COMPLEX_Agt_Dpp...</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO3ARGH</td>\n",
       "      <td>enzyme_ANGIO3ARGH_rev_pc</td>\n",
       "      <td>reverse</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2ASPH_fwd</th>\n",
       "      <td>cplx_COMPLEX_Agt_Dnpep_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO2ASPH</td>\n",
       "      <td>enzyme_ANGIO2ASPH_fwd_pc</td>\n",
       "      <td>forward</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2ASPH_rev</th>\n",
       "      <td>cplx_COMPLEX_Agt_Dnpep_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO2ASPH</td>\n",
       "      <td>enzyme_ANGIO2ASPH_rev_pc</td>\n",
       "      <td>reverse</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2DDP_fwd</th>\n",
       "      <td>cplx_COMPLEX_Agt_Dpp3_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO2DDP</td>\n",
       "      <td>enzyme_ANGIO2DDP_fwd_pc</td>\n",
       "      <td>forward</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_14_0_rev</th>\n",
       "      <td>cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0</td>\n",
       "      <td>enzyme_PROTCSAT_14_0_rev_pc</td>\n",
       "      <td>reverse</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_16_0_fwd</th>\n",
       "      <td>cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_16_0</td>\n",
       "      <td>enzyme_PROTCSAT_16_0_fwd_pc</td>\n",
       "      <td>forward</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_16_0_rev</th>\n",
       "      <td>cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_16_0</td>\n",
       "      <td>enzyme_PROTCSAT_16_0_rev_pc</td>\n",
       "      <td>reverse</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_18_0_fwd</th>\n",
       "      <td>cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_18_0</td>\n",
       "      <td>enzyme_PROTCSAT_18_0_fwd_pc</td>\n",
       "      <td>forward</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_18_0_rev</th>\n",
       "      <td>cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_18_0</td>\n",
       "      <td>enzyme_PROTCSAT_18_0_rev_pc</td>\n",
       "      <td>reverse</td>\n",
       "      <td>234000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4352 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  complexes  \\\n",
       "enzyme                                                                        \n",
       "enzyme_ANGIO3ARGH_fwd     cplx_COMPLEX_Agt_Anpep_pc;cplx_COMPLEX_Agt_Dpp...   \n",
       "enzyme_ANGIO3ARGH_rev     cplx_COMPLEX_Agt_Anpep_pc;cplx_COMPLEX_Agt_Dpp...   \n",
       "enzyme_ANGIO2ASPH_fwd                             cplx_COMPLEX_Agt_Dnpep_pc   \n",
       "enzyme_ANGIO2ASPH_rev                             cplx_COMPLEX_Agt_Dnpep_pc   \n",
       "enzyme_ANGIO2DDP_fwd                               cplx_COMPLEX_Agt_Dpp3_pc   \n",
       "...                                                                     ...   \n",
       "enzyme_PROTCSAT_14_0_rev  cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...   \n",
       "enzyme_PROTCSAT_16_0_fwd  cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...   \n",
       "enzyme_PROTCSAT_16_0_rev  cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...   \n",
       "enzyme_PROTCSAT_18_0_fwd  cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...   \n",
       "enzyme_PROTCSAT_18_0_rev  cplx_MONOMER_Zdhhc2_pc;cplx_MONOMER_Zdhhc20_pc...   \n",
       "\n",
       "                         compartment      reactions  \\\n",
       "enzyme                                                \n",
       "enzyme_ANGIO3ARGH_fwd             pc     ANGIO3ARGH   \n",
       "enzyme_ANGIO3ARGH_rev             pc     ANGIO3ARGH   \n",
       "enzyme_ANGIO2ASPH_fwd             pc     ANGIO2ASPH   \n",
       "enzyme_ANGIO2ASPH_rev             pc     ANGIO2ASPH   \n",
       "enzyme_ANGIO2DDP_fwd              pc      ANGIO2DDP   \n",
       "...                              ...            ...   \n",
       "enzyme_PROTCSAT_14_0_rev          pc  PROTCSAT_14_0   \n",
       "enzyme_PROTCSAT_16_0_fwd          pc  PROTCSAT_16_0   \n",
       "enzyme_PROTCSAT_16_0_rev          pc  PROTCSAT_16_0   \n",
       "enzyme_PROTCSAT_18_0_fwd          pc  PROTCSAT_18_0   \n",
       "enzyme_PROTCSAT_18_0_rev          pc  PROTCSAT_18_0   \n",
       "\n",
       "                                              enzymes direction enzyme_keff  \n",
       "enzyme                                                                       \n",
       "enzyme_ANGIO3ARGH_fwd        enzyme_ANGIO3ARGH_fwd_pc   forward      234000  \n",
       "enzyme_ANGIO3ARGH_rev        enzyme_ANGIO3ARGH_rev_pc   reverse      234000  \n",
       "enzyme_ANGIO2ASPH_fwd        enzyme_ANGIO2ASPH_fwd_pc   forward      234000  \n",
       "enzyme_ANGIO2ASPH_rev        enzyme_ANGIO2ASPH_rev_pc   reverse      234000  \n",
       "enzyme_ANGIO2DDP_fwd          enzyme_ANGIO2DDP_fwd_pc   forward      234000  \n",
       "...                                               ...       ...         ...  \n",
       "enzyme_PROTCSAT_14_0_rev  enzyme_PROTCSAT_14_0_rev_pc   reverse      234000  \n",
       "enzyme_PROTCSAT_16_0_fwd  enzyme_PROTCSAT_16_0_fwd_pc   forward      234000  \n",
       "enzyme_PROTCSAT_16_0_rev  enzyme_PROTCSAT_16_0_rev_pc   reverse      234000  \n",
       "enzyme_PROTCSAT_18_0_fwd  enzyme_PROTCSAT_18_0_fwd_pc   forward      234000  \n",
       "enzyme_PROTCSAT_18_0_rev  enzyme_PROTCSAT_18_0_rev_pc   reverse      234000  \n",
       "\n",
       "[4352 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"enzymes\"\n",
    "try:\n",
    "    df_enzymes = pd.read_csv(model_filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    print(\"Loaded from previously generated file\")\n",
    "except (FileNotFoundError, KeyError):\n",
    "    try:\n",
    "        df_enzymes = pd.read_csv(filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    except (FileNotFoundError, KeyError):\n",
    "        complexes_to_reactions = (\n",
    "            pcmodel_tables[\"complexes\"].set_index(\"complexes\")[\"reactions\"].to_dict()\n",
    "        )\n",
    "        df_enzymes = create_enzyme_table(\n",
    "            model,\n",
    "            complexes_to_reactions=complexes_to_reactions,\n",
    "            id_key=dict_of_id_keys.get(table_type),\n",
    "            optional_columns=optional_columns,\n",
    "            annotation_columns=[\n",
    "                # \"uniprot\"\n",
    "            ],\n",
    "            replace_compartments=replace_compartments,\n",
    "        )\n",
    "        if replace_compartments:\n",
    "            df_enzymes[\"compartment\"] = df_enzymes[\"compartment\"].replace(\n",
    "                replace_compartments\n",
    "            )\n",
    "        print(\"Created new table\")\n",
    "    else:\n",
    "        print(\"Loaded from main RBC-GEM file\")\n",
    "\n",
    "\n",
    "if simplify_compartments:\n",
    "    df_enzymes = df_enzymes.groupby([\"complexes\", \"enzyme\"]).agg(\n",
    "        lambda values: \";\".join(\n",
    "            [str(value) for value in list(values.dropna().unique())]\n",
    "        )\n",
    "    )\n",
    "    df_enzymes[\"compartment\"] = DEFAULT_PROTEOME_COMPARTMENT\n",
    "    df_enzymes = df_enzymes.reset_index(drop=False)\n",
    "\n",
    "df_enzymes[table_type] = df_enzymes[[strip_plural(table_type), \"compartment\"]].apply(\n",
    "    lambda x: \"_\".join(x.values), axis=1\n",
    ")\n",
    "df_enzymes = df_enzymes.set_index(strip_plural(table_type))\n",
    "pcmodel_tables[table_type] = df_enzymes.copy()\n",
    "df_enzymes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aee3dbe-cb7b-40a7-b3e6-56c9a238df06",
   "metadata": {},
   "source": [
    "## Create PC-model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bf7b1787-95e1-4848-a4cc-d4c79245c81c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of metabolites: 2157\n",
      "Number of proteins: 826\n",
      "Number of complexes: 888\n",
      "Number of enzymes: 6528\n",
      "Forward variable: 2176/6528\n",
      "Reverse variable: 2176/6528\n",
      "Summation variable : 2176/6528\n",
      "Number of budgets: 1\n",
      "\n",
      "Number of reactions: 3274\n",
      "Number of protein.dilutions: 826\n",
      "Number of complex.formation: 888\n",
      "Number of complex.dilutions: 888\n",
      "Number of enzyme.formation: 6572\n",
      "Number of enzyme.dilutions: 6528\n",
      "Forward variable: 2176/6528\n",
      "Reverse variable: 2176/6528\n",
      "Summation variable : 2176/6528\n",
      "Number of budget.dilutions: 1\n",
      "\n",
      "Number of non-zero complex rate constants: 6572\n",
      "Number of non-zero enzyme rate constants: 4352\n"
     ]
    }
   ],
   "source": [
    "protein_table = pcmodel_tables[\"proteins\"].reset_index(drop=False)\n",
    "complex_table = pcmodel_tables[\"complexes\"].reset_index(drop=False)\n",
    "enzyme_table = pcmodel_tables[\"enzymes\"].reset_index(drop=False)\n",
    "max_weight_fraction = 100\n",
    "\n",
    "pcmodel, final_pcmodel_tables = construct_pcmodel_from_tables(\n",
    "    model,\n",
    "    protein_table=protein_table,\n",
    "    complex_table=complex_table,\n",
    "    enzyme_table=enzyme_table,\n",
    "    max_weight_fraction=max_weight_fraction,\n",
    "    include_complex_dilutions=True,  # Relaxes constraints areound complexes. Recommend to start, can be set to zero later or removed entirely\n",
    "    irrev_rxn_complex_keff=0,  # Set as None to ignore, small number to keep in model, 0 to remove from complex-enzyme mapping\n",
    ")\n",
    "if simplify_compartments:\n",
    "    pcmodel.compartments = {DEFAULT_PROTEOME_COMPARTMENT: \"protein compartment\"}\n",
    "# Print summary\n",
    "for attr, subclass_dict in ATTR_SUBCLASS_DICT.items():\n",
    "    n = len(\n",
    "        getattr(pcmodel, attr).query(\n",
    "            lambda x: not isinstance(x, tuple(subclass_dict.values()))\n",
    "        )\n",
    "    )\n",
    "    print(f\"Number of {attr}: {n}\")\n",
    "    for key, subcls in subclass_dict.items():\n",
    "        obj_list = getattr(pcmodel, attr).query(lambda x: isinstance(x, subcls))\n",
    "        n = len(obj_list)\n",
    "        print(f\"Number of {key}: {n}\")\n",
    "        if subcls in (Enzyme, EnzymeDilution):\n",
    "            print(\n",
    "                f'Forward variable: {len(obj_list.query(lambda x: DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.forward\"] in x.id))}/{n}'\n",
    "            )\n",
    "            print(\n",
    "                f'Reverse variable: {len(obj_list.query(lambda x: DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.reverse\"] in x.id))}/{n}'\n",
    "            )\n",
    "            print(\n",
    "                f'Summation variable : {len(obj_list.query(lambda x: DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"] in x.id))}/{n}'\n",
    "            )\n",
    "    print()\n",
    "\n",
    "keff_table = final_pcmodel_tables[\"enzymes\"].copy()\n",
    "keff_table[\"direction\"] = keff_table[\"reactions\"].apply(\n",
    "    lambda rid: model.reactions.get_by_id(rid).reaction\n",
    ")\n",
    "keff_table[\"direction\"] = keff_table[\"direction\"].apply(\n",
    "    lambda x: x.replace(\"<=>\", \"-->\")\n",
    ")\n",
    "keff_table[\"direction\"] = keff_table[[\"enzyme\", \"direction\"]].apply(\n",
    "    lambda x: (\n",
    "        x[\"direction\"].replace(\"-->\", \"<--\")\n",
    "        if x[\"enzyme\"].endswith(\n",
    "            DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.reverse\"]\n",
    "        )\n",
    "        else x[\"direction\"]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "keff_table[\"complexes\"] = keff_table[\"complexes\"].apply(lambda x: x.split(\";\"))\n",
    "keff_table[\"complex_keff\"] = keff_table[\"complex_keff\"].apply(lambda x: x.split(\";\"))\n",
    "keff_table = keff_table.explode([\"complexes\", \"complex_keff\"])\n",
    "keff_table[\"complex\"] = keff_table[[\"complexes\", \"compartment\"]].apply(\n",
    "    lambda x: x[\"complexes\"].replace(f\"_{x['compartment']}\", \"\"), axis=1\n",
    ")\n",
    "keff_table = keff_table.groupby([\"enzyme\", \"complex\"], as_index=False).agg(\n",
    "    lambda x: list(x.unique())[0]\n",
    ")\n",
    "keff_table = keff_table.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"enzyme\",\n",
    "        \"enzyme_keff\",\n",
    "        \"complex\",\n",
    "        \"complex_keff\",\n",
    "        \"compartment\",\n",
    "        \"reactions\",\n",
    "        \"direction\",\n",
    "    ],\n",
    "]\n",
    "complex_keff_table = keff_table.drop(\"enzyme_keff\", axis=1).drop_duplicates()\n",
    "enzyme_keff_table = (\n",
    "    keff_table.groupby([\"enzyme\", \"enzyme_keff\"], as_index=False)[\n",
    "        [\"reactions\", \"direction\"]\n",
    "    ]\n",
    "    .agg(lambda x: list(x.unique())[0])\n",
    "    .drop_duplicates()\n",
    ")\n",
    "final_pcmodel_tables[\"complex_keffs\"] = complex_keff_table\n",
    "final_pcmodel_tables[\"enzyme_keffs\"] = enzyme_keff_table\n",
    "\n",
    "n_cplx_keff = len(\n",
    "    complex_keff_table[complex_keff_table[\"complex_keff\"].astype(float) != 0]\n",
    ")\n",
    "print(f\"Number of non-zero complex rate constants: {n_cplx_keff}\")\n",
    "\n",
    "n_enzyme_keff = len(\n",
    "    enzyme_keff_table[enzyme_keff_table[\"enzyme_keff\"].astype(float) != 0]\n",
    ")\n",
    "print(f\"Number of non-zero enzyme rate constants: {n_enzyme_keff}\")\n",
    "\n",
    "\n",
    "for table_type, df_table in final_pcmodel_tables.items():\n",
    "    df_table.to_csv(\n",
    "        model_dirpath / f\"pcmodel_{pcmodel}_{table_type}.tsv\", sep=\"\\t\", index=False\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e7554cd-dbf7-4286-a29e-81281e44f767",
   "metadata": {},
   "source": [
    "### Formulate additional protein constraints\n",
    "#### Address isoforms and compartments with additional constraints\n",
    "For isoforms and/or compartments, place an additional constraint such that the total sum of all isoforms does not exceed the measured concentraiton value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ae04a412-75e3-4ca8-b629-46f8fafcb243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# protein_table = pcmodel_tables[\"proteins\"]\n",
    "# mapping_key = \"uniprot\"\n",
    "# df_additional_constraints = protein_table[protein_table[mapping_key].duplicated(False)]\n",
    "# df_additional_constraints = df_additional_constraints.groupby(\n",
    "#     [\n",
    "#         \"genes\",\n",
    "#         mapping_key,\n",
    "#     ],\n",
    "#     as_index=False,\n",
    "# ).agg(lambda x: list(x))\n",
    "# if not df_additional_constraints.empty:\n",
    "#     if \"lower_bound\" in df_additional_constraints.columns:\n",
    "#         df_additional_constraints[\"lower_bound\"] = df_additional_constraints[\n",
    "#             \"lower_bound\"\n",
    "#         ].apply(min)\n",
    "#     if \"upper_bound\" in df_additional_constraints.columns:\n",
    "#         df_additional_constraints[\"upper_bound\"] = df_additional_constraints[\n",
    "#             \"upper_bound\"\n",
    "#         ].apply(max)\n",
    "\n",
    "# data = {}\n",
    "# for idx, row in df_additional_constraints.iterrows():\n",
    "#     # Technically, always one gene but refers to genes attribute\n",
    "#     genes = row[\"genes\"]\n",
    "#     uniprot = model.genes.get_by_id(genes).annotation.get(mapping_key, \"\")\n",
    "#     proteins = split_string(row.get(\"proteins\"))\n",
    "#     proteins = pcmodel.metabolites.get_by_any(proteins)\n",
    "#     is_compartment = len({p.compartment for p in proteins}) > 1\n",
    "#     is_isoform = (\n",
    "#         len(\n",
    "#             {\n",
    "#                 p.id.replace(f\"_{p.compartment}\", \"\").split(\n",
    "#                     \"_\",\n",
    "#                 )[-1]\n",
    "#                 for p in proteins\n",
    "#                 if p.id.replace(f\"_{p.compartment}\", \"\")\n",
    "#                 .split(\n",
    "#                     \"_\",\n",
    "#                 )[-1]\n",
    "#                 .isnumeric()\n",
    "#             }\n",
    "#         )\n",
    "#         > 1\n",
    "#     )\n",
    "#     if is_compartment and not is_isoform:\n",
    "#         default_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"constraints\"][\"prefix.compartent\"]\n",
    "#     elif is_isoform and not is_compartment:\n",
    "#         default_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"constraints\"][\"prefix.isoform\"]\n",
    "#     else:\n",
    "#         default_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"constraints\"][\"prefix.constraint\"]\n",
    "#     constraint_id = row.get(\"constraints\", f\"{default_prefix}{genes}\")\n",
    "#     lower_bound = float(row.get(\"lower_bound\")) if row.get(\"lower_bound\") else 0\n",
    "#     upper_bound = (\n",
    "#         float(row.get(\"upper_bound\"))\n",
    "#         if row.get(\"upper_bound\")\n",
    "#         else DEFAULT_CONCENTRATION_BOUND\n",
    "#     )\n",
    "#     protein_dilutions = [\n",
    "#         reaction\n",
    "#         for protein in proteins\n",
    "#         for reaction in list(protein.reactions)\n",
    "#         if reaction.id.endswith(protein.id)\n",
    "#     ]\n",
    "#     # \"ISOCONS\" is short for \"ISOFORM CONSTRAINT\"\n",
    "#     # \"COMPCONS\" is short for \"COMPARTMENT CONSTRAINT\"\n",
    "#     # \"CONS\" for general constraint\n",
    "#     data[idx] = {\n",
    "#         \"constraints\": constraint_id,\n",
    "#         \"genes\": genes,\n",
    "#         \"proteins\": build_string([p.id for p in proteins]),\n",
    "#         \"reactions\": build_string([p.id for p in protein_dilutions]),\n",
    "#         # Assume sum of isoforms is a constant, works well with proteomic measurements that do not distinguish\n",
    "#         \"coefficients\": \";\".join([str(1) for p in protein_dilutions]),\n",
    "#         \"lower_bound\": lower_bound,\n",
    "#         \"upper_bound\": upper_bound,\n",
    "#         \"unit\": \"nmol / gDW\",\n",
    "#         mapping_key: uniprot,\n",
    "#     }\n",
    "# df_additional_constraints = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "# df_additional_constraints.to_csv(\n",
    "#     model_dirpath / f\"pcmodel_{pcmodel.id}_constraints_proteins.tsv\", sep=\"\\t\", index=False\n",
    "# )\n",
    "# df_additional_constraints"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f7be1c-67b8-4ebb-92ee-bb53b312d30f",
   "metadata": {},
   "source": [
    "## Add additional protein constraints to model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "df3bb685-f825-4f2a-bc0d-66d50606d73c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try:\n",
    "#     df_additional_constraints = pd.read_csv(\n",
    "#         model_dirpath / f\"pcmodel_{pcmodel.id}_constraints_proteins.tsv\",\n",
    "#         sep=\"\\t\",\n",
    "#         index_col=None,\n",
    "#     )\n",
    "# except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "#     df_additional_constraints = pd.DataFrame()\n",
    "# else:\n",
    "#     if not df_additional_constraints.empty:\n",
    "#         for constraint_id, row in df_additional_constraints.set_index(\n",
    "#             \"constraints\"\n",
    "#         ).iterrows():\n",
    "#             reactions = pcmodel.reactions.get_by_any(row[\"reactions\"].split(\";\"))\n",
    "#             coefficients = row[\"coefficients\"].split(\";\")\n",
    "#             abundance = sum(\n",
    "#                 [\n",
    "#                     int(coeff) * reaction.flux_expression\n",
    "#                     for reaction, coeff in zip(reactions, coefficients)\n",
    "#                 ]\n",
    "#             )\n",
    "#             lower_bound = float(row.get(\"lower_bound\")) if row.get(\"lower_bound\") else 0\n",
    "#             upper_bound = (\n",
    "#                 float(row.get(\"upper_bound\"))\n",
    "#                 if row.get(\"upper_bound\")\n",
    "#                 else DEFAULT_CONCENTRATION_BOUND\n",
    "#             )\n",
    "#             if constraint_id in pcmodel.constraints:\n",
    "#                 # TODO warn\n",
    "#                 pcmodel.remove_cons_vars(pcmodel.constraints[constraint_id])\n",
    "#             additional_constraint = pcmodel.problem.Constraint(\n",
    "#                 abundance,\n",
    "#                 name=constraint_id,\n",
    "#                 lb=lower_bound,\n",
    "#                 ub=upper_bound,\n",
    "#             )\n",
    "#             pcmodel.add_cons_vars(additional_constraint)\n",
    "\n",
    "# df_additional_constraints"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c0ba53d-f212-4358-bf78-d410f3b39182",
   "metadata": {},
   "source": [
    "### Add other additional constraints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "733376b2-7369-4fea-8954-ac8ac428393e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# constraints_ratios_filepath = overlay_dirpath / f\"pcmodel_constraints_additional.tsv\"\n",
    "\n",
    "# df_constraints_additional = pd.read_csv(\n",
    "#     constraints_ratios_filepath,\n",
    "#     sep=\"\\t\",\n",
    "#     index_col=\"constraints\",\n",
    "# )\n",
    "\n",
    "# ratio_ids = set()\n",
    "# skipped_constraints = set()\n",
    "# not_found = set()\n",
    "# for constraint_id, row in df_constraints_additional.iterrows():\n",
    "#     subs_dict = {}\n",
    "#     lhs = parse_expr(row[\"lhs\"])\n",
    "#     rhs = parse_expr(row[\"rhs\"])\n",
    "\n",
    "#     csense = row[\"csense\"]\n",
    "#     lb=None if csense == \"<\" else 0\n",
    "#     ub=None if csense == \">\" else 0\n",
    "#     reactions = row[\"reactions\"].split(\";\")\n",
    "#     for reaction in reactions:\n",
    "#         try:\n",
    "#             reaction = model.reactions.get_by_id(reaction)\n",
    "#         except Exception:\n",
    "#             if reaction == str(rhs) or reaction == str(lhs):\n",
    "#                 skipped_constraints.add(constraint_id)\n",
    "#                 continue\n",
    "#             else:\n",
    "#                 not_found.add(reaction)\n",
    "#                 subs_dict[reaction] = 0\n",
    "#         else:\n",
    "#             subs_dict[reaction.id] = reaction.flux_expression\n",
    "#     if (str(rhs) == \"0\" or str(lhs) == \"0\") and len([r for r in reactions if r not in not_found]) <= 1:\n",
    "#         skipped_constraints.add(constraint_id)\n",
    "#     if constraint_id in skipped_constraints:\n",
    "#         continue\n",
    "#     abundance = lhs - rhs\n",
    "#     abundance = abundance.subs(subs_dict)\n",
    "#     if str(abundance) == \"0\":\n",
    "#         print(f\"{constraint_id} is always equal to 0, not including.\")\n",
    "#         skipped_constraints.add(constraint_id)\n",
    "#         continue\n",
    "#     try:\n",
    "#         constraint = model.constraints[constraint_id]\n",
    "#     except Exception:\n",
    "#         pass\n",
    "#     else:\n",
    "#         model.remove_cons_vars(constraint)\n",
    "#     constraint = model.problem.Constraint(\n",
    "#         abundance=abundance,\n",
    "#         name=constraint_id,\n",
    "#         lb=float(lb) if lb is not None else lb,\n",
    "#         ub=float(ub) if ub is not None else ub,\n",
    "#     )\n",
    "#     model.add_cons_vars(constraint)\n",
    "#     # Convert units\n",
    "#     if constraint.lb is not None:\n",
    "#         constraint.lb = convert_L_to_gDW(float(constraint.lb))\n",
    "#     if constraint.ub is not None:\n",
    "#         constraint.ub = convert_L_to_gDW(float(constraint.ub))\n",
    "#     df_constraints_additional.loc[constraint_id, \"reactions\"] = \";\".join([r for r in reactions if r not in not_found])\n",
    "#     print(constraint)\n",
    "#     df_constraints_additional.loc[constraint_id, \"lhs\"] = str(lhs)\n",
    "#     df_constraints_additional.loc[constraint_id, \"rhs\"] = str(rhs)\n",
    "\n",
    "\n",
    "# df_constraints_additional = df_constraints_additional.loc[~df_constraints_additional.index.isin(not_found.union(skipped_constraints))]\n",
    "# df_constraints_additional.to_csv(model_dirpath / f\"pcmodel_{pcmodel.id}_constraints_additional.tsv\", sep=\"\\t\", index=False)\n",
    "# df_constraints_additional"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9398f43f-208a-413a-850b-e78e37a0025f",
   "metadata": {},
   "source": [
    "### Annotate objects for protein constraints\n",
    "Annotating objects that are used for protein constraints will make subsequent analyses easier to perform.\n",
    "#### Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "befcc19f-35ac-4090-b384-c969bcfde46c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>uniprot.sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>proteins</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>protein_A4galt_pc</th>\n",
       "      <td>A4galt</td>\n",
       "      <td>Q67BJ4</td>\n",
       "      <td>MGISCSHLEETMSKPPDCLLRMLRGTPRQRVFTFFIISFKFMFLIS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Aars1_pc</th>\n",
       "      <td>Aars1</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "      <td>MDATLTAREIRERFINFFRRNEHTYVHSSATIPLDDPTLLFANAGM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Aasdhppt_pc</th>\n",
       "      <td>Aasdhppt</td>\n",
       "      <td>Q9CQF6</td>\n",
       "      <td>MVFPAKRLCVVPSMEGVRWAFSCGTWLPSRAEWLLAMRSIQPEEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Abca1_pc</th>\n",
       "      <td>Abca1</td>\n",
       "      <td>P41233</td>\n",
       "      <td>MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Abca7_pc</th>\n",
       "      <td>Abca7</td>\n",
       "      <td>Q91V24</td>\n",
       "      <td>MALGTQLMLLLWKNYTYRRRQPIQLLVELLWPLFLFFILVAVRHSH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Yes1_pc</th>\n",
       "      <td>Yes1</td>\n",
       "      <td>Q04736</td>\n",
       "      <td>MGCIKSKENKSPAIKYTPENLTEPVSPSASHYGVEHATVAPTSSTK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc2_pc</th>\n",
       "      <td>Zdhhc2</td>\n",
       "      <td>P59267</td>\n",
       "      <td>MAPSGSGGVRRRCRRVLYWIPVVFISLLLGWSYYAYAIQLCIVSME...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc20_pc</th>\n",
       "      <td>Zdhhc20</td>\n",
       "      <td>Q5Y5T1</td>\n",
       "      <td>MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVSTISRTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc3_pc</th>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>Q8R173</td>\n",
       "      <td>MMLIPTHHFRDIERKPEYLQPEKCAPPPFPGPAGAMWFIRDGCGIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_Zdhhc5_pc</th>\n",
       "      <td>Zdhhc5</td>\n",
       "      <td>Q8VDZ4</td>\n",
       "      <td>MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLNVSPA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>826 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        genes uniprot  \\\n",
       "proteins                                \n",
       "protein_A4galt_pc      A4galt  Q67BJ4   \n",
       "protein_Aars1_pc        Aars1  Q8BGQ7   \n",
       "protein_Aasdhppt_pc  Aasdhppt  Q9CQF6   \n",
       "protein_Abca1_pc        Abca1  P41233   \n",
       "protein_Abca7_pc        Abca7  Q91V24   \n",
       "...                       ...     ...   \n",
       "protein_Yes1_pc          Yes1  Q04736   \n",
       "protein_Zdhhc2_pc      Zdhhc2  P59267   \n",
       "protein_Zdhhc20_pc    Zdhhc20  Q5Y5T1   \n",
       "protein_Zdhhc3_pc      Zdhhc3  Q8R173   \n",
       "protein_Zdhhc5_pc      Zdhhc5  Q8VDZ4   \n",
       "\n",
       "                                                      uniprot.sequence  \n",
       "proteins                                                                \n",
       "protein_A4galt_pc    MGISCSHLEETMSKPPDCLLRMLRGTPRQRVFTFFIISFKFMFLIS...  \n",
       "protein_Aars1_pc     MDATLTAREIRERFINFFRRNEHTYVHSSATIPLDDPTLLFANAGM...  \n",
       "protein_Aasdhppt_pc  MVFPAKRLCVVPSMEGVRWAFSCGTWLPSRAEWLLAMRSIQPEEKE...  \n",
       "protein_Abca1_pc     MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...  \n",
       "protein_Abca7_pc     MALGTQLMLLLWKNYTYRRRQPIQLLVELLWPLFLFFILVAVRHSH...  \n",
       "...                                                                ...  \n",
       "protein_Yes1_pc      MGCIKSKENKSPAIKYTPENLTEPVSPSASHYGVEHATVAPTSSTK...  \n",
       "protein_Zdhhc2_pc    MAPSGSGGVRRRCRRVLYWIPVVFISLLLGWSYYAYAIQLCIVSME...  \n",
       "protein_Zdhhc20_pc   MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVSTISRTG...  \n",
       "protein_Zdhhc3_pc    MMLIPTHHFRDIERKPEYLQPEKCAPPPFPGPAGAMWFIRDGCGIA...  \n",
       "protein_Zdhhc5_pc    MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLNVSPA...  \n",
       "\n",
       "[826 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"proteins\"\n",
    "table_cols = [\"genes\", \"uniprot\", \"sequence\"]\n",
    "obj_type = Protein\n",
    "table = protein_table.set_index(table_type)[table_cols]\n",
    "\n",
    "# Rename sequence\n",
    "table = table.rename({\"sequence\": \"uniprot.sequence\"}, axis=1)\n",
    "\n",
    "\n",
    "annotation_mappings = {\n",
    "    table_key: table[table_key].to_dict() for table_key in table.columns\n",
    "}\n",
    "\n",
    "for met in pcmodel.metabolites.query(lambda x: isinstance(x, obj_type)):\n",
    "    annotation_dict = {\n",
    "        table_key: mapping_dict[met.id]\n",
    "        for table_key, mapping_dict in annotation_mappings.items()\n",
    "        if mapping_dict.get(met.id)\n",
    "    }\n",
    "    met.annotation.update(annotation_dict)\n",
    "    rxn = pcmodel.reactions.get_by_id(\n",
    "        f'{DEFAULT_PREFIX_SUFFIX_VALUES[table_type][\"prefix.dilution\"]}{met.id}'\n",
    "    )\n",
    "    rxn.annotation.update(\n",
    "        {\n",
    "            table_key: mapping_dict[met.id]\n",
    "            for table_key, mapping_dict in annotation_mappings.items()\n",
    "            if mapping_dict.get(met.id)\n",
    "        }\n",
    "    )\n",
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fccc456-65b7-4846-93d0-8b881b9d102f",
   "metadata": {},
   "source": [
    "#### Complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5dca20fc-e146-410a-adeb-75d762b52c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>complexes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_A4galt_pc</th>\n",
       "      <td>A4galt</td>\n",
       "      <td>Q67BJ4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Aars1_pc</th>\n",
       "      <td>Aars1</td>\n",
       "      <td>Q8BGQ7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Aasdhppt_pc</th>\n",
       "      <td>Aasdhppt</td>\n",
       "      <td>Q9CQF6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_HOMOTETRAMER_Abca1_pc</th>\n",
       "      <td>Abca1</td>\n",
       "      <td>P41233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Abca7_pc</th>\n",
       "      <td>Abca7</td>\n",
       "      <td>Q91V24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Yes1_pc</th>\n",
       "      <td>Yes1</td>\n",
       "      <td>Q04736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc2_pc</th>\n",
       "      <td>Zdhhc2</td>\n",
       "      <td>P59267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc20_pc</th>\n",
       "      <td>Zdhhc20</td>\n",
       "      <td>Q5Y5T1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc3_pc</th>\n",
       "      <td>Zdhhc3</td>\n",
       "      <td>Q8R173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_Zdhhc5_pc</th>\n",
       "      <td>Zdhhc5</td>\n",
       "      <td>Q8VDZ4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>888 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               genes uniprot\n",
       "complexes                                   \n",
       "cplx_MONOMER_A4galt_pc        A4galt  Q67BJ4\n",
       "cplx_MONOMER_Aars1_pc          Aars1  Q8BGQ7\n",
       "cplx_MONOMER_Aasdhppt_pc    Aasdhppt  Q9CQF6\n",
       "cplx_HOMOTETRAMER_Abca1_pc     Abca1  P41233\n",
       "cplx_MONOMER_Abca7_pc          Abca7  Q91V24\n",
       "...                              ...     ...\n",
       "cplx_MONOMER_Yes1_pc            Yes1  Q04736\n",
       "cplx_MONOMER_Zdhhc2_pc        Zdhhc2  P59267\n",
       "cplx_MONOMER_Zdhhc20_pc      Zdhhc20  Q5Y5T1\n",
       "cplx_MONOMER_Zdhhc3_pc        Zdhhc3  Q8R173\n",
       "cplx_MONOMER_Zdhhc5_pc        Zdhhc5  Q8VDZ4\n",
       "\n",
       "[888 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"complexes\"\n",
    "table_cols = [\"genes\"]\n",
    "obj_type = Complex\n",
    "table = complex_table.set_index(table_type)[table_cols]\n",
    "# Add uniprot to table\n",
    "table[\"uniprot\"] = table[\"genes\"].apply(\n",
    "    lambda x: build_string(\n",
    "        [y.annotation[\"uniprot\"] for y in pcmodel.genes.get_by_any(split_string(x))]\n",
    "    )\n",
    ")\n",
    "table\n",
    "\n",
    "\n",
    "annotation_mappings = {\n",
    "    table_key: table[table_key].to_dict() for table_key in table.columns\n",
    "}\n",
    "\n",
    "for met in pcmodel.metabolites.query(lambda x: isinstance(x, obj_type)):\n",
    "    annotation_dict = {\n",
    "        table_key: mapping_dict[met.id]\n",
    "        for table_key, mapping_dict in annotation_mappings.items()\n",
    "        if mapping_dict.get(met.id)\n",
    "    }\n",
    "    met.annotation.update(annotation_dict)\n",
    "    rxn = pcmodel.reactions.get_by_id(\n",
    "        f'{DEFAULT_PREFIX_SUFFIX_VALUES[table_type][\"prefix.dilution\"]}{met.id}'\n",
    "    )\n",
    "    rxn.annotation.update(\n",
    "        {\n",
    "            table_key: mapping_dict[met.id]\n",
    "            for table_key, mapping_dict in annotation_mappings.items()\n",
    "            if mapping_dict.get(met.id)\n",
    "        }\n",
    "    )\n",
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a6491fb-eff8-4dd3-b8c9-ac788bbd4bde",
   "metadata": {},
   "source": [
    "#### Enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "17932c29-74f1-430f-a746-6789bc30724f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzymes</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO3ARGH_fwd_pc</th>\n",
       "      <td>ANGIO3ARGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO3ARGH_rev_pc</th>\n",
       "      <td>ANGIO3ARGH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2ASPH_fwd_pc</th>\n",
       "      <td>ANGIO2ASPH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2ASPH_rev_pc</th>\n",
       "      <td>ANGIO2ASPH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2DDP_fwd_pc</th>\n",
       "      <td>ANGIO2DDP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_14_0_total_pc</th>\n",
       "      <td>PROTCSAT_14_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_16_0_total_pc</th>\n",
       "      <td>PROTCSAT_16_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_16_0_total_pc</th>\n",
       "      <td>PROTCSAT_16_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_18_0_total_pc</th>\n",
       "      <td>PROTCSAT_18_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_18_0_total_pc</th>\n",
       "      <td>PROTCSAT_18_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8704 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   reactions\n",
       "enzymes                                     \n",
       "enzyme_ANGIO3ARGH_fwd_pc          ANGIO3ARGH\n",
       "enzyme_ANGIO3ARGH_rev_pc          ANGIO3ARGH\n",
       "enzyme_ANGIO2ASPH_fwd_pc          ANGIO2ASPH\n",
       "enzyme_ANGIO2ASPH_rev_pc          ANGIO2ASPH\n",
       "enzyme_ANGIO2DDP_fwd_pc            ANGIO2DDP\n",
       "...                                      ...\n",
       "enzyme_PROTCSAT_14_0_total_pc  PROTCSAT_14_0\n",
       "enzyme_PROTCSAT_16_0_total_pc  PROTCSAT_16_0\n",
       "enzyme_PROTCSAT_16_0_total_pc  PROTCSAT_16_0\n",
       "enzyme_PROTCSAT_18_0_total_pc  PROTCSAT_18_0\n",
       "enzyme_PROTCSAT_18_0_total_pc  PROTCSAT_18_0\n",
       "\n",
       "[8704 rows x 1 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"enzymes\"\n",
    "table_cols = [\"reactions\"]\n",
    "obj_type = Enzyme\n",
    "table = enzyme_table.set_index(table_type)[table_cols]\n",
    "# Add enzyme totals to table\n",
    "table.index = pd.Index(\n",
    "    [\n",
    "        f'{x.rsplit(\"_\", maxsplit=2)[0]}{DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]}_{x.rsplit(\"_\", maxsplit=1)[-1]}'\n",
    "        for x in table.index\n",
    "    ],\n",
    "    name=table.index.name,\n",
    ")\n",
    "table = pd.concat((enzyme_table.set_index(table_type)[table_cols], table))\n",
    "\n",
    "\n",
    "annotation_mappings = {\n",
    "    table_key: table[table_key].to_dict() for table_key in table.columns\n",
    "}\n",
    "\n",
    "for met in pcmodel.metabolites.query(lambda x: isinstance(x, obj_type)):\n",
    "    annotation_dict = {\n",
    "        table_key: mapping_dict[met.id]\n",
    "        for table_key, mapping_dict in annotation_mappings.items()\n",
    "        if mapping_dict.get(met.id)\n",
    "    }\n",
    "    met.annotation.update(annotation_dict)\n",
    "    rxn = pcmodel.reactions.get_by_id(\n",
    "        f'{DEFAULT_PREFIX_SUFFIX_VALUES[table_type][\"prefix.dilution\"]}{met.id}'\n",
    "    )\n",
    "    rxn.annotation.update(\n",
    "        {\n",
    "            table_key: mapping_dict[met.id]\n",
    "            for table_key, mapping_dict in annotation_mappings.items()\n",
    "            if mapping_dict.get(met.id)\n",
    "        }\n",
    "    )\n",
    "table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75b5ed10-611d-44a1-b924-cf3badb5a39f",
   "metadata": {},
   "source": [
    "### Set budget constraints for low-abundance and high-abundance proteomes\n",
    "* RBCs are enucleated, terminally differentiated cells that are composed of 95% to 98% Hb by dry mass (mass of all the constituents of a cell in the absence of water)\n",
    "    * PMID: 13429433, PMID: 13999462, PMID: 21796773, **PMID: 34378368**\n",
    "* Therefore, remove hemoglobin from the low abundance proteome budget constraint and create a new constraint specific to hemoglobin abundance.\n",
    "* Assume 90-95% minimum of dry mass is hemoglobin, and up to 5-10% of dry mass are other proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "aab2f453-f4e1-430e-8f29-4a104c7b21c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BDL_budget_total: budget_total --> \n",
      "BDL_budget_proteome: budget_proteome --> budget_total\n",
      "BDL_budget_hemoglobin: budget_hemoglobin --> budget_total\n"
     ]
    }
   ],
   "source": [
    "# Split total budget into sectors, one for hemoglobin and one for low abundance proteome\n",
    "budget_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"budgets\"][\"prefix.dilution\"]\n",
    "budget_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"budgets\"][\"prefix.metabolite\"]\n",
    "# Get budget total\n",
    "budget_met_total = pcmodel.metabolites.get_by_id(f\"{budget_met_prefix}total\")\n",
    "budget_rxn_total = pcmodel.reactions.get_by_id(f\"{budget_rxn_prefix}{budget_met_total}\")\n",
    "\n",
    "# Create budget for low abundance proteins\n",
    "budget_rxn_lap = budget_rxn_total.copy()\n",
    "budget_met_lap = list(budget_rxn_lap.metabolites).pop()\n",
    "# Set new IDs and names for metabolites\n",
    "budget_met_lap.id = f\"{budget_met_prefix}proteome\"\n",
    "budget_met_lap.name = \"Budget constraint (Low abundance proteins)\"\n",
    "# Set new IDs and names for reactions\n",
    "budget_rxn_lap.id = f\"{budget_rxn_prefix}{budget_met_lap.id}\"\n",
    "budget_rxn_lap.name = \"Budget demand (Low abundance proteins)\"\n",
    "\n",
    "# Create budget for hemoglobin proteins\n",
    "budget_rxn_hbp = budget_rxn_total.copy()\n",
    "budget_met_hbp = list(budget_rxn_hbp.metabolites).pop()\n",
    "# Set new IDs and names for metabolites\n",
    "budget_met_hbp.id = f\"{budget_met_prefix}hemoglobin\"\n",
    "budget_met_hbp.name = \"Budget constraint (Hemoglobin proteins)\"\n",
    "# Set new IDs and names for reactions\n",
    "budget_rxn_hbp.id = f\"{budget_rxn_prefix}{budget_met_hbp.id}\"\n",
    "budget_rxn_hbp.name = \"Budget demand (Hemoglobin proteins)\"\n",
    "\n",
    "# Budget bounds for generic model\n",
    "budget_rxn_total.bounds = (0, 1000)\n",
    "budget_rxn_hbp.bounds = (900, 1000)\n",
    "budget_rxn_lap.bounds = (0, 100)\n",
    "pcmodel.add_reactions([budget_rxn_lap, budget_rxn_hbp])\n",
    "\n",
    "# Divide total budget into hemoglobin and low abundance\n",
    "budget_reations = pcmodel.reactions.query(lambda x: isinstance(x, BudgetDilution))\n",
    "for reaction in budget_met_total.reactions:\n",
    "    if reaction.id in budget_reations:\n",
    "        continue\n",
    "    # Determine if hemoglobin or low abundance\n",
    "    elif any(\n",
    "        [\n",
    "            f'{DEFAULT_PREFIX_SUFFIX_VALUES[\"proteins\"][\"prefix.metabolite\"]}{gid}'\n",
    "            in reaction.id\n",
    "            for gid in list(hemoglobin_proteins)\n",
    "        ]\n",
    "    ):\n",
    "        budget_met = budget_met_hbp\n",
    "    else:\n",
    "        budget_met = budget_met_lap\n",
    "    coeff = reaction.get_coefficient(budget_met_total)\n",
    "    reaction.add_metabolites(\n",
    "        {\n",
    "            # Add to sector\n",
    "            budget_met: coeff,\n",
    "            budget_met_total: -coeff,\n",
    "        }\n",
    "    )\n",
    "for budget_rxn in budget_reations:\n",
    "    coeff = -1 if budget_rxn_total.id == budget_rxn.id else 1\n",
    "    # Ensure budget total is reactant in its own reaction, otherwise is a product\n",
    "    budget_rxn.add_metabolites({budget_met_total: coeff}, combine=False)\n",
    "    print(budget_rxn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17b852f8-6444-40e7-bed3-27741a75842e",
   "metadata": {},
   "source": [
    "### Ensure model can be optimized for glucose uptake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "2548b619-2bf2-4923-96ca-b6136398e614",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "COt                        -92.762052\n",
       "DPGM                       -23.190513\n",
       "ENO                         23.190513\n",
       "EX_co_e                    -92.762052\n",
       "EX_h2o2_e                  -23.190513\n",
       "EX_k_e                     -92.762052\n",
       "EX_na1_e                   139.143078\n",
       "EX_pi_e                     46.381026\n",
       "EX_pyr_e                    23.190513\n",
       "FAt_hs_3_0                  23.190513\n",
       "H2O2t                      -23.190513\n",
       "HB23DPGB                   -23.190513\n",
       "HBCOB                       92.762052\n",
       "METOXR                      23.190513\n",
       "NaKt                        46.381026\n",
       "PGK                         23.190513\n",
       "PGM                         23.190513\n",
       "PIt                         46.381026\n",
       "PYK                         23.190513\n",
       "PYR_PPAtex                  23.190513\n",
       "SK_carboxyhb_c              92.762052\n",
       "SK_hb4_23dpg_c             -23.190513\n",
       "SK_k_c                      92.762052\n",
       "SK_na1_c                  -139.143078\n",
       "SK_protres_metSox__SL_c     23.190513\n",
       "SK_protres_met__L_c        -23.190513\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "objective_rxns = [\"NaKt\"]\n",
    "pcmodel.objective = sum(\n",
    "    [pcmodel.reactions.get_by_id(rid).flux_expression for rid in objective_rxns]\n",
    ")\n",
    "pcsol = pcmodel.optimize()\n",
    "pcsol.fluxes.loc[\n",
    "    [r.id for r in model.reactions if r.id in pcsol.fluxes[pcsol.fluxes != 0].index]\n",
    "].sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "376fde0a-9090-4c06-9849-937303d72538",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PROTDL_protein_Atp1a3_pc       236.192932\n",
       "PROTDL_protein_Atp1b3_pc       236.192932\n",
       "PROTDL_protein_Eno3_pc         172.091504\n",
       "PROTDL_protein_Hbb_bh1_pc    54535.984709\n",
       "PROTDL_protein_Hbb_y_pc         22.940217\n",
       "PROTDL_protein_Pgam1_pc        238.476145\n",
       "PROTDL_protein_Pgk1_pc          49.132894\n",
       "PROTDL_protein_Pkm_pc          676.117321\n",
       "PROTDL_protein_Slc16a7_pc      187.238568\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcsol.fluxes.loc[\n",
    "    [\n",
    "        r.id\n",
    "        for r in pcmodel.reactions.query(lambda x: isinstance(x, ProteinDilution))\n",
    "        if r.id in pcsol.fluxes[pcsol.fluxes != 0].index\n",
    "    ]\n",
    "].sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5f91db47-c712-4090-998c-d9e528c40975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENZDL_enzyme_DPGM_fwd_pc             99.104757\n",
       "ENZDL_enzyme_DPGM_total_pc           99.104757\n",
       "ENZDL_enzyme_ENO_rev_pc              99.104757\n",
       "ENZDL_enzyme_ENO_total_pc            99.104757\n",
       "ENZDL_enzyme_HB23DPGB_fwd_pc         99.104757\n",
       "ENZDL_enzyme_HB23DPGB_total_pc       99.104757\n",
       "ENZDL_enzyme_HBCOB_rev_pc           396.419027\n",
       "ENZDL_enzyme_HBCOB_total_pc         396.419027\n",
       "ENZDL_enzyme_NaKt_rev_pc            198.209513\n",
       "ENZDL_enzyme_NaKt_total_pc          198.209513\n",
       "ENZDL_enzyme_PGK_rev_pc              99.104757\n",
       "ENZDL_enzyme_PGK_total_pc            99.104757\n",
       "ENZDL_enzyme_PGM_rev_pc              99.104757\n",
       "ENZDL_enzyme_PGM_total_pc            99.104757\n",
       "ENZDL_enzyme_PYK_rev_pc              99.104757\n",
       "ENZDL_enzyme_PYK_total_pc            99.104757\n",
       "ENZDL_enzyme_PYR_PPAtex_rev_pc       99.104757\n",
       "ENZDL_enzyme_PYR_PPAtex_total_pc     99.104757\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcsol.fluxes.loc[\n",
    "    [\n",
    "        r.id\n",
    "        for r in pcmodel.reactions.query(lambda x: isinstance(x, EnzymeDilution))\n",
    "        if r.id in pcsol.fluxes[pcsol.fluxes != 0].index\n",
    "    ]\n",
    "].sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a01f53-12d7-464b-b633-762fb818f823",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d995d1dc-1355-4ace-b469-3067a9932791",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regular model\n",
    "write_cobra_model(model, filename=model_dirpath / f\"{model}.xml\")\n",
    "write_cobra_model(model, filename=model_dirpath / f\"{model}.json\")\n",
    "\n",
    "# Protein constrained  without curated keffs\n",
    "write_cobra_model(pcmodel, filename=model_dirpath / f\"{pcmodel}.xml\")\n",
    "write_cobra_model(pcmodel, filename=model_dirpath / f\"{pcmodel}.json\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a59cb7b4-83fe-4b44-8d15-0e3c4a71a4ed",
   "metadata": {},
   "source": [
    "### Update rate constants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "778dc096-2677-4dfd-aa7d-058cd8aed1cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# pcmodel_curated = load_overlay_model(filename=model_dirpath / f\"{pcmodel.id}.xml\")\n",
    "# pcmodel_curated.id += \"_keff_curated\"\n",
    "# df_complex_keffs = pd.read_csv(overlay_dirpath / \"pcmodel_complex_keffs.tsv\", sep=\"\\t\", index_col=None)\n",
    "# df_enzyme_keffs = pd.read_csv(overlay_dirpath / \"pcmodel_enzyme_keffs.tsv\", sep=\"\\t\", index_col=None)\n",
    "\n",
    "# cf = 1 / 1e6  # Conversion factor from nmol to mmol\n",
    "# if simplify_compartments:\n",
    "#     df_enzyme_keffs[\"compartment\"] = DEFAULT_PROTEOME_COMPARTMENT\n",
    "#     df_complex_keffs[\"compartment\"] = DEFAULT_PROTEOME_COMPARTMENT\n",
    "\n",
    "# df_enzyme_keffs[\"enzymes\"] = df_enzyme_keffs[[\"enzyme\", \"compartment\"]].apply(lambda x: \"_\".join(x.values), axis=1)\n",
    "# df_complex_keffs[\"enzymes\"] = df_complex_keffs[[\"enzyme\", \"compartment\"]].apply(lambda x: \"_\".join(x.values), axis=1)\n",
    "# df_complex_keffs[\"complexes\"] = df_complex_keffs[[\"complex\", \"compartment\"]].apply(lambda x: \"_\".join(x.values), axis=1)\n",
    "# df_complex_keffs = df_complex_keffs[df_complex_keffs[\"complexes\"].isin(\n",
    "#     pcmodel_curated.metabolites.query(\n",
    "#         lambda x: x.id.startswith(\"cplx_\")).list_attr(\"id\")\n",
    "# )]\n",
    "\n",
    "# df_complex_keffs[\"complex_keff\"] = df_complex_keffs[\"complex_keff\"].astype(float)\n",
    "# df_enzyme_keffs[\"enzyme_keff\"] = df_enzyme_keffs[\"enzyme_keff\"].astype(float)\n",
    "\n",
    "# df_enzyme_keffs = df_enzyme_keffs.drop_duplicates()\n",
    "# df_complex_keffs = df_complex_keffs.drop_duplicates()\n",
    "# for _, row in df_complex_keffs.iterrows():\n",
    "#     enz = row[\"enzymes\"]\n",
    "#     try:\n",
    "#         enz = pcmodel_curated.metabolites.get_by_id(enz)\n",
    "#     except KeyError:\n",
    "#         if enz.replace(DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.reverse\"], \"\") in pcmodel_curated.metabolites:\n",
    "#             enz_other_dir = pcmodel_curated.metabolites.get_by_id(enz.replace(DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.reverse\"], \"\"))\n",
    "#             missing_enz = enz_other_dir.copy()\n",
    "#             missing_enz.id = enz\n",
    "#             pcmodel_curated.add_metabolites([missing_enz])\n",
    "\n",
    "#             enzyme_keff = df_enzyme_keffs[df_enzyme_keffs[\"enzymes\"] == missing_enz.id][\"enzyme_keff\"].item()\n",
    "#             df = df_complex_keffs[df_complex_keffs[\"enzymes\"] == missing_enz.id].copy()\n",
    "#             for _, (enzyme, cplx, complex_keff) in df[[\"enzyme\", \"complexes\", \"complex_keff\"]].iterrows():\n",
    "#                 if complex_keff == 0 or enzyme_keff == 0:\n",
    "#                     continue\n",
    "\n",
    "#                 keff = float(complex_keff) / float(enzyme_keff)\n",
    "#                 formation_rxn = add_complex_formation_reaction(\n",
    "#                     pcmodel_curated,\n",
    "#                     missing_enz,\n",
    "#                     \"enzyme\",\n",
    "#                     coeff_map=f\"{cplx}({keff})\",\n",
    "#                 )\n",
    "#             if enzyme_keff != 0:\n",
    "#                 dilution_rxn = add_dilution_reaction(\n",
    "#                     pcmodel_curated,\n",
    "#                     missing_enz,\n",
    "#                     \"enzyme\",\n",
    "#                 )\n",
    "\n",
    "#                 for r in enz_other_dir.reactions:\n",
    "#                     if not r.id in model.reactions:\n",
    "#                         continue\n",
    "#                     sign = -1 if enz_other_dir in r.reactants else 1\n",
    "#                     pcmodel_curated.reactions.get_by_id(r.id).add_metabolites(\n",
    "#                         {missing_enz: sign * (1 / enzyme_keff / cf)}, combine=False\n",
    "#                     )\n",
    "#     else:\n",
    "#         enzyme_keff = df_enzyme_keffs[df_enzyme_keffs[\"enzymes\"] == enz.id][\"enzyme_keff\"].item()\n",
    "#         df = df_complex_keffs[df_complex_keffs[\"enzymes\"] == enz.id].copy()\n",
    "#         for _, (enzyme, cplx, complex_keff) in df[[\"enzyme\", \"complexes\", \"complex_keff\"]].iterrows():\n",
    "#             if complex_keff == 0 or enzyme_keff == 0:\n",
    "#                 continue\n",
    "#             keff = float(complex_keff) / float(enzyme_keff)\n",
    "#             try:\n",
    "#                 formation_rxn = pcmodel_curated.reactions.get_by_id(f\"ENZFM_{enzyme}_{cplx}\")\n",
    "#             except KeyError:\n",
    "#                 formation_rxn = add_complex_formation_reaction(\n",
    "#                     pcmodel_curated,\n",
    "#                     enz,\n",
    "#                     \"enzyme\",\n",
    "#                     coeff_map=f\"{cplx}({keff})\",\n",
    "#                 )\n",
    "#             else:\n",
    "#                 formation_rxn.add_metabolites({cplx: -keff}, combine=False)\n",
    "#         if enzyme_keff != 0:\n",
    "#             for r in enz.reactions:\n",
    "#                 if not r.id in model.reactions:\n",
    "#                     continue\n",
    "#                 sign = 1 if enz in r.products else -1\n",
    "#                 pcmodel_curated.reactions.get_by_id(r.id).add_metabolites(\n",
    "#                     {enz: sign * (1 / enzyme_keff / cf)}, combine=False\n",
    "#                 )\n",
    "\n",
    "# pcmodel_curated.remove_metabolites(pcmodel_curated.metabolites.query(lambda x: not x.reactions))\n",
    "# df_complex_keffs.to_csv(model_dirpath / f\"complex_keffs_{pcmodel_curated.id}.tsv\", sep=\"\\t\", index=False)\n",
    "# df_enzyme_keffs.to_csv(model_dirpath / f\"enzyme_keffs_{pcmodel_curated.id}.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# # Print summary\n",
    "# for attr, subclass_dict in ATTR_SUBCLASS_DICT.items():\n",
    "#     n = len(getattr(pcmodel_curated, attr).query(lambda x: not isinstance(x, tuple(subclass_dict.values()))))\n",
    "#     print(f\"Number of {attr}: {n}\")\n",
    "#     for key, subcls in subclass_dict.items():\n",
    "#         n = len(getattr(pcmodel_curated, attr).query(lambda x: isinstance(x, subcls)))\n",
    "#         print(f\"Number of {key}: {n}\")\n",
    "#     print()\n",
    "\n",
    "# # Print summary\n",
    "# for attr, subclass_dict in ATTR_SUBCLASS_DICT.items():\n",
    "#     n = len(getattr(pcmodel_curated, attr).query(lambda x: not isinstance(x, tuple(subclass_dict.values()))))\n",
    "#     print(f\"Number of {attr}: {n}\")\n",
    "#     for key, subcls in subclass_dict.items():\n",
    "#         n = len(getattr(pcmodel_curated, attr).query(lambda x: isinstance(x, subcls)))\n",
    "#         print(f\"Number of {key}: {n}\")\n",
    "#     print()\n",
    "\n",
    "\n",
    "# try:\n",
    "#     df_curated_complex_keffs = final_pcmodel_tables[\"complex_keffs\"].set_index(\"enzymes\")[[\"complexes\", \"complex_keff\"]].copy()\n",
    "# except KeyError:\n",
    "#     print(f\"Number of non-zero complex rate constants (curated): 0\")\n",
    "# else:\n",
    "#     df_curated_complex_keffs = df_curated_complex_keffs.explode([\"complexes\", \"complex_keff\"]).reset_index(drop=False).drop_duplicates()\n",
    "#     df_curated_complex_keffs = df_curated_complex_keffs[df_curated_complex_keffs[\"complex_keff\"].astype(float) != 0.]\n",
    "#     print(f\"Number of non-zero complex rate constants (curated): {len(df_curated_complex_keffs)}\")\n",
    "# finally:\n",
    "#     print(f\"Number of non-zero complex rate constants (total): {n_cplx_keff}\")\n",
    "# formation_rxn\n",
    "# try:\n",
    "#     df_curated_enzyme_keffs = final_pcmodel_tables[\"enzyme_keffs\"][[\"enzymes\", \"enzyme_keff\"]].copy()\n",
    "# except KeyError:\n",
    "#     print(f\"Number of non-zero enzyme rate constants (curated): 0\")\n",
    "# else:\n",
    "#     df_curated_enzyme_keffs = df_curated_enzyme_keffs[df_curated_enzyme_keffs[\"enzyme_keff\"].astype(float) != 0.]\n",
    "#     print(f\"Number of non-zero enzyme rate constants (curated): {len(df_curated_enzyme_keffs)}\")\n",
    "# finally:\n",
    "#     print(f\"Number of non-zero enzyme rate constants (total): {n_enzyme_keff}\")\n",
    "\n",
    "# write_cobra_model(pcmodel_curated, filename=model_dirpath / f\"{pcmodel_curated}.xml\")\n",
    "# write_cobra_model(pcmodel_curated, filename=model_dirpath / f\"{pcmodel_curated}.json\")\n",
    "# pcmodel_curated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "aa5c7c65-f808-4970-af4d-fc43f20ab8bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>22fd5338550</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3274</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>826</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x22fd5338550>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "ea753932-c514-4210-a683-993e04abddc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>2300a0b6490</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>10402</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>18979</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>826</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC at 0x2300a0b6490>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcmodel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56e962b7-843e-4f2f-872b-de81a71925df",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
