{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2bb106c3-5799-428c-8fda-1501fa4e0dd2",
   "metadata": {},
   "source": [
    "# Aggregate Interaction and association data into a Network\n",
    "* Known interactions are setup within a global network:\n",
    "    * Drug-Drug interactions (DDI)\n",
    "        * DrugBank\n",
    "        * Continuous Drug Combination Database (CDCDB)\n",
    "    * Drug-Protein interactions (DPI)\n",
    "        * DrugBank\n",
    "    * Drug-SNP interactions\n",
    "        * DrugBank\n",
    "    * Protein-SNP associations\n",
    "        * UniProt\n",
    "        * dbSNP\n",
    "    * Protein-Phenotype interactions (a.k.a genotype-phenotype)\n",
    "        * MIM\n",
    "\n",
    "For best results, run the following notebooks to ensure extracted data is up to date:\n",
    "* For UniProt data: [ExtractAnnotationsData_UniProt](../database/ExtractAnnotationsData_UniProt.ipynb)\n",
    "* For DrugBank data: [ExtractAnnotationsData_DrugBank](../database/ExtractAnnotationsData_DrugBank.ipynb)\n",
    "* For SNP data: [ExtractData_dbSNP](../database/ExtractData_dbSNP.ipynb)\n",
    "* For MIM data: [ExtractAnnotationsData_MIM](../database/ExtractAnnotationsData_MIM.ipynb)\n",
    "\n",
    "Once the basic interaction structure is created, various node and attributes can be extracted from the relevant sources \n",
    "It is expected that this notebook will grow over time and extract additional relevant information for nodes/edges to create a knowledge-graph setup.\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d4ea31-84e7-4227-995d-5a0ad4a272d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    GEM_NAME,\n",
    "    build_string,\n",
    "    get_annotation_df,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    split_string,\n",
    ")\n",
    "from rbc_gem_utils.database import CDCDB_DB_TAG, DRUGBANK_DB_TAG, MIM_DB_TAG\n",
    "from rbc_gem_utils.util import explode_column, strip_plural\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee1c1f4f-036d-414a-b2d8-73a6590aab9c",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffe73f45-9d86-4497-acff-f5e66d20abd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3514f3d0-9105-4fd4-bbca-a6947b307df5",
   "metadata": {},
   "outputs": [],
   "source": [
    "annotation_type = \"genes\"\n",
    "df_model_mappings = get_annotation_df(\n",
    "    getattr(model, annotation_type),\n",
    "    [\n",
    "        \"uniprot\",\n",
    "        \"hgnc.symbol\",\n",
    "        \"drugbank\",\n",
    "        \"dbsnp\",\n",
    "        \"mim\",\n",
    "        \"ec-code\",\n",
    "        \"orphanet\",\n",
    "        \"pubchem.compound\",\n",
    "    ],\n",
    ").rename({\"id\": annotation_type}, axis=1)\n",
    "for col in df_model_mappings.columns:\n",
    "    df = df_model_mappings[col].apply(split_string).explode(col).drop_duplicates()\n",
    "    print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9232ffe-b091-4ed5-9b1f-272731eeaa0d",
   "metadata": {},
   "source": [
    "### Setup node alias table and data dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9ae2481-bb7e-419e-89e9-58355a4b6b1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Edge data\n",
    "interaction_dfs = defaultdict(list)\n",
    "\n",
    "protein_node_id_type = \"uniprot\"\n",
    "df_network_node_aliases = df_model_mappings[\n",
    "    [protein_node_id_type, protein_node_id_type]\n",
    "].copy()\n",
    "df_network_node_aliases.columns = [\"node_id\", \"node_alias\"]\n",
    "df_network_node_aliases[\"node_source\"] = protein_node_id_type\n",
    "df_network_node_aliases[\"node_type\"] = \"gene/protein\"\n",
    "df_network_node_aliases[\"alias_type\"] = \"identifier\"\n",
    "for key, alias_type in zip([\"genes\", \"hgnc.symbol\"], [\"identifier\", \"symbol\"]):\n",
    "    df = df_model_mappings.loc[:, [protein_node_id_type, key]]\n",
    "    if key != \"genes\":\n",
    "        df[\"node_source\"] = key\n",
    "    else:\n",
    "        df[\"node_source\"] = GEM_NAME\n",
    "        df[key] = df[key].apply(lambda x: f\"G_{x}\")\n",
    "    df[\"node_type\"] = \"gene/protein\"\n",
    "    df[\"alias_type\"] = alias_type\n",
    "\n",
    "    df_network_node_aliases = pd.concat(\n",
    "        (\n",
    "            df_network_node_aliases,\n",
    "            df.rename(\n",
    "                {\n",
    "                    protein_node_id_type: \"node_id\",\n",
    "                    key: \"node_alias\",\n",
    "                },\n",
    "                axis=1,\n",
    "            ),\n",
    "        )\n",
    "    )\n",
    "node_types_keys = {\n",
    "    \"drug\": \"drugbank\",\n",
    "    \"snp\": \"dbsnp\",\n",
    "}\n",
    "for node_type, key in node_types_keys.items():\n",
    "    list_of_ids = sorted(\n",
    "        df_model_mappings[key].apply(split_string).explode(key).dropna().unique()\n",
    "    )\n",
    "    df = pd.DataFrame.from_dict(\n",
    "        {\n",
    "            \"node_id\": list_of_ids,\n",
    "            \"node_alias\": list_of_ids,\n",
    "            \"node_source\": [key] * len(list_of_ids),\n",
    "            \"node_type\": [node_type] * len(list_of_ids),\n",
    "            \"alias_type\": \"identifier\",\n",
    "        },\n",
    "        orient=\"columns\",\n",
    "    )\n",
    "    df_network_node_aliases = pd.concat(\n",
    "        (\n",
    "            df_network_node_aliases,\n",
    "            df,\n",
    "        )\n",
    "    )\n",
    "\n",
    "# Note that node aliases at this point may include nodes that never get connected. Nodes are pruned after edges are created\n",
    "node_types = (\n",
    "    df_network_node_aliases.groupby(\"node_type\")[\"node_alias\"]\n",
    "    .agg(lambda x: set(x.unique()))\n",
    "    .to_dict()\n",
    ")\n",
    "df_network_node_aliases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08c2bd46-369c-46d0-9f3b-35071a22a1a1",
   "metadata": {},
   "source": [
    "## Load interaction data\n",
    "#### Load drug interactions\n",
    "##### With other drugs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f85691b3-8776-467e-b5cc-4bddafd81b08",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interactions = pd.read_csv(\n",
    "    get_dirpath(\"database\", DRUGBANK_DB_TAG) / f\"Interactions_{DRUGBANK_DB_TAG}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ")\n",
    "# Add names to aliases\n",
    "df = pd.concat(\n",
    "    (\n",
    "        df_interactions.loc[:, [\"drugbank_A\", \"name_A\"]].rename(\n",
    "            {\"drugbank_A\": \"node_id\", \"name_A\": \"node_alias\"}, axis=1\n",
    "        ),\n",
    "        df_interactions.loc[:, [\"drugbank_B\", \"name_B\"]].rename(\n",
    "            {\"drugbank_B\": \"node_id\", \"name_B\": \"node_alias\"}, axis=1\n",
    "        ),\n",
    "    ),\n",
    "    axis=0,\n",
    ").drop_duplicates()\n",
    "df[\"node_source\"] = \"drugbank\"  # Is INN?\n",
    "df[\"node_type\"] = \"drug\"\n",
    "df[\"alias_type\"] = \"name\"\n",
    "\n",
    "df_network_node_aliases = (\n",
    "    pd.concat(\n",
    "        (\n",
    "            df_network_node_aliases,\n",
    "            df,\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "df_network_node_aliases\n",
    "\n",
    "# Add Drug Drug mappings\n",
    "edge_type = \"drug-drug\"\n",
    "rename_mapping = {\n",
    "    \"drugbank_A\": \"node_A\",\n",
    "    \"drugbank_B\": \"node_B\",\n",
    "}\n",
    "\n",
    "df_interactions = df_interactions.loc[:, list(rename_mapping)].rename(\n",
    "    rename_mapping, axis=1\n",
    ")\n",
    "df_interactions = df_interactions[df_interactions[\"node_A\"].isin(node_types[\"drug\"])]\n",
    "df_interactions = df_interactions[df_interactions[\"node_B\"].isin(node_types[\"drug\"])]\n",
    "df_interactions = df_interactions.drop_duplicates().reset_index(drop=True)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_subtype\"] = float(\"nan\")\n",
    "df_interactions[\"edge_source\"] = \"drugbank\"\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1af45b9-cd52-474f-8d08-9993b3e6efd6",
   "metadata": {},
   "source": [
    "##### With proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e603d4aa-8241-4616-8982-10d3333f60c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interactions = pd.read_csv(\n",
    "    get_dirpath(\"database\", DRUGBANK_DB_TAG) / f\"Proteins_{DRUGBANK_DB_TAG}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ")\n",
    "edge_type = \"drug-gene/protein\"\n",
    "# Add Drug Protein mappings\n",
    "rename_mapping = {\n",
    "    \"drugbank-id\": \"node_A\",\n",
    "    \"polypeptide.uniprot-id\": \"node_B\",\n",
    "    \"type\": \"edge_subtype\",\n",
    "}\n",
    "df_interactions = df_interactions.loc[:, list(rename_mapping)].rename(\n",
    "    rename_mapping, axis=1\n",
    ")\n",
    "df_interactions = df_interactions[df_interactions[\"node_A\"].isin(node_types[\"drug\"])]\n",
    "df_interactions = df_interactions[\n",
    "    df_interactions[\"node_B\"].isin(node_types[\"gene/protein\"])\n",
    "]\n",
    "df_interactions = df_interactions.drop_duplicates().reset_index(drop=True)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_source\"] = \"drugbank\"\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9b22e41-ebb5-4910-a4d5-1e281a1fc7cc",
   "metadata": {},
   "source": [
    "##### With SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82e5ced0-42a6-45ab-a297-87a484c68c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interactions = pd.read_csv(\n",
    "    get_dirpath(\"database\", DRUGBANK_DB_TAG) / f\"SNP_{DRUGBANK_DB_TAG}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ")\n",
    "edge_type = \"drug-snp\"\n",
    "df_interactions[\"edge_subtype\"] = df_interactions[\"adverse-reaction\"].notna()\n",
    "\n",
    "# Add Drug SNP mappings\n",
    "rename_mapping = {\n",
    "    \"drugbank-id\": \"node_A\",\n",
    "    \"rs-id\": \"node_B\",\n",
    "    \"edge_subtype\": \"edge_subtype\",\n",
    "}\n",
    "df_interactions = df_interactions.loc[:, list(rename_mapping)].rename(\n",
    "    rename_mapping, axis=1\n",
    ")\n",
    "df_interactions = df_interactions[df_interactions[\"node_A\"].isin(node_types[\"drug\"])]\n",
    "df_interactions = df_interactions[df_interactions[\"node_B\"].isin(node_types[\"snp\"])]\n",
    "df_interactions = df_interactions.drop_duplicates().reset_index(drop=True)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_source\"] = \"drugbank\"\n",
    "df_interactions[\"edge_subtype\"] = df_interactions[\"edge_subtype\"].apply(\n",
    "    lambda x: \"known-effect-adverse\" if x else \"known-effect\"\n",
    ")\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "defe98ae-82e0-4a4f-8619-07025fd75bc5",
   "metadata": {},
   "source": [
    "#### Load protein SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77c993e9-58b6-4fe8-b175-73efb2ff83f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interactions = pd.read_csv(\n",
    "    get_dirpath(\"database\") / f\"snps_{GEM_NAME}.tsv\", sep=\"\\t\", index_col=0, dtype=str\n",
    ")\n",
    "df_model_snps = (\n",
    "    df_model_mappings.set_index(\"uniprot\")[\"dbsnp\"]\n",
    "    .apply(split_string)\n",
    "    .explode()\n",
    "    .dropna()\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "\n",
    "edge_type = \"gene/protein-snp\"\n",
    "df_interactions = df_model_snps.merge(\n",
    "    df_interactions, left_on=\"dbsnp\", right_on=\"dbsnp\", how=\"inner\"\n",
    ")\n",
    "rename_mapping = {\"uniprot\": \"node_A\", \"dbsnp\": \"node_B\", \"snp_class\": \"edge_subtype\"}\n",
    "df_interactions = df_interactions.loc[:, list(rename_mapping)].rename(\n",
    "    rename_mapping, axis=1\n",
    ")\n",
    "df_interactions = df_interactions[\n",
    "    df_interactions[\"node_A\"].isin(node_types[\"gene/protein\"])\n",
    "]\n",
    "df_interactions = df_interactions[df_interactions[\"node_B\"].isin(node_types[\"snp\"])]\n",
    "df_interactions = df_interactions.drop_duplicates().reset_index(drop=True)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_source\"] = build_string([\"uniprot\", \"dbsnp\"])\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27b338ac-b55c-4738-8908-0e58a5f9c41d",
   "metadata": {},
   "source": [
    "### Load MIM genotypes/phenotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513bba3b-eba2-4d83-a3a6-29f377a45185",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interactions = pd.read_csv(\n",
    "    get_dirpath(\"database\", MIM_DB_TAG) / f\"{MIM_DB_TAG}_{GEM_NAME}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    "    dtype=str,\n",
    ")\n",
    "df_interactions = pd.merge(\n",
    "    df_model_mappings.set_index(\"genes\")[[protein_node_id_type]],\n",
    "    df_interactions,\n",
    "    left_index=True,\n",
    "    right_on=\"genes\",\n",
    ").reset_index(drop=True)\n",
    "df_interactions = df_interactions.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"genes\",\n",
    "        \"uniprot\",\n",
    "        \"mim\",\n",
    "        \"title\",\n",
    "        \"mim.phenotype\",\n",
    "        \"mim.phenotype.title\",\n",
    "        \"mim.phenotype.title.symbol\",\n",
    "    ],\n",
    "]\n",
    "# Add MIM aliases for gene/protein\n",
    "rename_mapping = {protein_node_id_type: \"node_id\", \"mim\": \"node_alias\"}\n",
    "df = (\n",
    "    df_interactions.loc[:, list(rename_mapping)]\n",
    "    .rename(rename_mapping, axis=1)\n",
    "    .dropna()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "df[\"node_alias\"] = df[\"node_alias\"].apply(lambda x: f\"mim:{x}\")\n",
    "df[\"node_source\"] = \"mim\"\n",
    "df[\"node_type\"] = \"gene/protein\"\n",
    "df[\"alias_type\"] = \"identifier\"\n",
    "df_network_node_aliases = (\n",
    "    pd.concat(\n",
    "        (\n",
    "            df_network_node_aliases,\n",
    "            df,\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "# Add MIM symbols as aliases for phenotypes\n",
    "rename_mapping = {\n",
    "    \"mim.phenotype\": \"node_id\",\n",
    "    \"mim.phenotype.title.symbol\": \"node_alias\",\n",
    "}\n",
    "df = (\n",
    "    df_interactions.loc[:, list(rename_mapping)]\n",
    "    .rename(rename_mapping, axis=1)\n",
    "    .dropna(subset=[\"node_id\"])\n",
    "    .drop_duplicates()\n",
    ")\n",
    "df[\"node_source\"] = \"mim\"\n",
    "df[\"node_type\"] = \"phenotype/disease\"\n",
    "df[\"alias_type\"] = \"symbol\"\n",
    "df_network_node_aliases = (\n",
    "    pd.concat(\n",
    "        (\n",
    "            df_network_node_aliases,\n",
    "            df.copy(),\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "# Also record the node ID alias\n",
    "df[\"node_alias\"] = df[\"node_id\"]\n",
    "df[\"alias_type\"] = \"identifier\"\n",
    "df_network_node_aliases = (\n",
    "    pd.concat(\n",
    "        (\n",
    "            df_network_node_aliases,\n",
    "            df.copy(),\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "# Add edge type\n",
    "edge_type = \"gene/protein-phenotype/disease\"\n",
    "rename_mapping = {\n",
    "    \"uniprot\": \"node_A\",\n",
    "    \"mim.phenotype\": \"node_B\",\n",
    "}\n",
    "df_interactions = df_interactions.loc[:, list(rename_mapping)].rename(\n",
    "    rename_mapping, axis=1\n",
    ")\n",
    "df_interactions = df_interactions.dropna().drop_duplicates().reset_index(drop=True)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_source\"] = \"mim\"\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8453d0c3-f70b-432c-a3ff-2613ebd12fe0",
   "metadata": {},
   "source": [
    "### Load effective drug combinations\n",
    "* If both PubChem and DrugBank are used, harmonization of data will be necessary\n",
    "* Combinations with more than two drug components are split into combination pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "180a78db-d785-4b93-a0da-bf4bf85d6a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_interactions = pd.read_csv(\n",
    "    get_dirpath(\"database\", CDCDB_DB_TAG) / f\"EffectiveCombos_{CDCDB_DB_TAG}.tsv\",\n",
    "    sep=\"\t\",\n",
    "    index_col=None,\n",
    "    dtype=str,\n",
    ")\n",
    "edge_type = \"drug-drug\"\n",
    "df_interactions = df_interactions[\n",
    "    df_interactions[\"node_A\"].apply(\n",
    "        lambda x: x.startswith(\"DB\") if isinstance(x, str) else False\n",
    "    )\n",
    "]\n",
    "df_interactions = df_interactions[\n",
    "    df_interactions[\"node_B\"].apply(\n",
    "        lambda x: x.startswith(\"DB\") if isinstance(x, str) else False\n",
    "    )\n",
    "]\n",
    "df_interactions = df_interactions.loc[\n",
    "    :, [\"node_A\", \"node_B\", \"edge_type\", \"edge_subtype\", \"edge_source\"]\n",
    "]\n",
    "df_interactions = df_interactions.loc[\n",
    "    df_interactions.loc[:, [\"node_A\", \"node_B\"]]\n",
    "    .apply(lambda x: build_string(sorted(x.values)), axis=1)\n",
    "    .drop_duplicates()\n",
    "    .index,\n",
    "    :,\n",
    "]\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab536401-6174-4591-a691-09586e845d96",
   "metadata": {},
   "source": [
    "### Load model as nodes/edges\n",
    "Reactions, metabolites, genes all included, pseudoreactions excluded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e72489b-2a12-4efa-9d6a-a640adf9097a",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_exclude = {\"Pseudoreactions\"}\n",
    "\n",
    "# Genes already have aliases previously\n",
    "gene_to_uniprot = df_model_mappings.set_index(\"genes\")[protein_node_id_type].to_dict()\n",
    "\n",
    "edge_type = \"gene/protein-reaction\"\n",
    "df_interactions = pd.DataFrame.from_dict(\n",
    "    {\n",
    "        gene_to_uniprot[gene.id]: build_string(\n",
    "            [\n",
    "                reaction.id\n",
    "                for reaction in list(gene.reactions)\n",
    "                if reaction.subsystem not in to_exclude\n",
    "            ]\n",
    "        )\n",
    "        for gene in model.genes\n",
    "    },\n",
    "    orient=\"index\",\n",
    "    columns=[\"node_B\"],\n",
    ")\n",
    "df_interactions.index.name = \"node_A\"\n",
    "df_interactions = explode_column(df_interactions, name=\"node_B\").reset_index(drop=False)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_source\"] = GEM_NAME\n",
    "# Prefix node for clarity\n",
    "df_interactions[\"node_B\"] = df_interactions[\"node_B\"].apply(lambda x: f\"R_{x}\")\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cee03b5-5677-4365-99d6-2e67d17929e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reactions and metabolites need aliases, can be added here.\n",
    "edge_type = \"reaction-metabolite\"\n",
    "df_interactions = pd.DataFrame.from_dict(\n",
    "    {\n",
    "        reaction.id: build_string([m.id for m in reaction.metabolites])\n",
    "        for reaction in model.reactions\n",
    "        if reaction.subsystem not in to_exclude\n",
    "    },\n",
    "    orient=\"index\",\n",
    "    columns=[\"node_B\"],\n",
    ")\n",
    "df_interactions.index.name = \"node_A\"\n",
    "df_interactions = explode_column(df_interactions, name=\"node_B\").reset_index(drop=False)\n",
    "df_interactions[\"edge_type\"] = edge_type\n",
    "df_interactions[\"edge_subtype\"] = df_interactions.apply(\n",
    "    lambda x: (\n",
    "        \"reactant\"\n",
    "        if model.metabolites.get_by_id(x[\"node_B\"])\n",
    "        in model.reactions.get_by_id(x[\"node_A\"]).reactants\n",
    "        else \"product\"\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "df_interactions[\"edge_source\"] = GEM_NAME\n",
    "\n",
    "# Prefix node for clarity\n",
    "df_interactions[\"node_A\"] = df_interactions[\"node_A\"].apply(lambda x: f\"R_{x}\")\n",
    "df_interactions[\"node_B\"] = df_interactions[\"node_B\"].apply(lambda x: f\"M_{x}\")\n",
    "interaction_dfs[edge_type] += [df_interactions.copy()]\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac71e3cb-7900-487b-a67e-6fe7315dcbab",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_types_and_aliases = {\n",
    "    \"metabolites\": [\"kegg.compound\", \"metatlas\"],\n",
    "    \"reactions\": [\"kegg.reaction\", \"metatlas\", \"ec-code\"],\n",
    "}\n",
    "\n",
    "for attribute_type, annotations_for_aliases in attribute_types_and_aliases.items():\n",
    "    if attribute_type == \"reactions\":\n",
    "        prefix = \"R_\"\n",
    "        obj_list = getattr(model, attribute_type).query(\n",
    "            lambda x: x.subsystem not in to_exclude\n",
    "        )\n",
    "    else:\n",
    "        prefix = \"M_\"\n",
    "        obj_list = getattr(model, attribute_type)\n",
    "    df = get_annotation_df(obj_list, annotations_for_aliases)\n",
    "    df[\"id\"] = df[\"id\"].apply(lambda x: f\"{prefix}{x}\")\n",
    "    df[\"node_id\"] = df[\"id\"]\n",
    "    df = df.rename({\"id\": f\"model.{attribute_type}\"}, axis=1)\n",
    "    # Add model metabolite names to aliases\n",
    "    attribute = \"name\"\n",
    "    df[f\"model.{attribute_type}.{attribute}\"] = df[f\"model.{attribute_type}\"].apply(\n",
    "        lambda x: getattr(obj_list.get_by_id(x[len(prefix) :]), attribute)\n",
    "    )\n",
    "    df = df.melt(id_vars=\"node_id\", var_name=\"node_source\", value_name=\"node_alias\")\n",
    "    df[\"alias_type\"] = \"identifier\"\n",
    "\n",
    "    # Clean up attribute alias types\n",
    "    index = df[df[\"node_source\"].str.startswith(f\"model.{attribute_type}.\")].index\n",
    "    df.loc[list(index), \"alias_type\"] = df.loc[list(index), \"node_source\"].apply(\n",
    "        lambda x: x.split(\".\", 2)[-1]\n",
    "    )\n",
    "    df.loc[list(index), \"node_source\"] = GEM_NAME\n",
    "    # Set node type\n",
    "    df[\"node_type\"] = strip_plural(attribute_type)\n",
    "\n",
    "    df_network_node_aliases = (\n",
    "        pd.concat(\n",
    "            (\n",
    "                df_network_node_aliases,\n",
    "                df,\n",
    "            ),\n",
    "            axis=0,\n",
    "        )\n",
    "        .drop_duplicates()\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "df_network_node_aliases"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8ddd967-b948-4590-86d1-300711ef6db2",
   "metadata": {},
   "source": [
    "### Aggregate data for nodes and edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ec4c891-e349-4d3a-80e1-3f46d2c1b3fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_network_edges = pd.DataFrame(columns=[\"node_A\", \"node_B\", \"edge_type\"])\n",
    "interaction_dfs_concat = {\n",
    "    key: pd.concat(df_list, axis=0) if len(df_list) > 1 else df_list[0]\n",
    "    for key, df_list in interaction_dfs.items()\n",
    "}\n",
    "interaction_dfs_final = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cbd94c3-be26-4fbb-b7a9-26e5411ba275",
   "metadata": {},
   "source": [
    "### Harmonize data\n",
    "#### Drug-Drug interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45f90188-a33c-4ee6-99cb-2332aed7f1f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_type = \"drug-drug\"\n",
    "df_interactions = interaction_dfs_concat[edge_type]\n",
    "df_interactions[\"nodes\"] = df_interactions[[\"node_A\", \"node_B\"]].apply(\n",
    "    lambda x: build_string(sorted(x)), axis=1\n",
    ")\n",
    "\n",
    "df_harmonized = df_interactions[df_interactions[\"nodes\"].duplicated(keep=False)]\n",
    "df_harmonized = df_harmonized.groupby(\"nodes\", as_index=False)[\n",
    "    [\"edge_type\", \"edge_subtype\", \"edge_source\"]\n",
    "].agg(lambda x: \";\".join(sorted(x.dropna().unique())))\n",
    "df_harmonized[\"node_A\"] = df_harmonized[\"nodes\"].apply(lambda x: x.split(\";\")[0])\n",
    "df_harmonized[\"node_B\"] = df_harmonized[\"nodes\"].apply(lambda x: x.split(\";\")[-1])\n",
    "\n",
    "df_interactions = pd.concat(\n",
    "    (df_interactions[~df_interactions[\"nodes\"].duplicated(keep=False)], df_harmonized),\n",
    "    axis=0,\n",
    ")\n",
    "df_interactions = df_interactions.loc[\n",
    "    :, [\"node_A\", \"node_B\", \"edge_type\", \"edge_subtype\", \"edge_source\"]\n",
    "]\n",
    "interaction_dfs_final[edge_type] = df_interactions\n",
    "df_interactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d9a9b17-6cd5-4b5a-99dd-a35b007866db",
   "metadata": {},
   "outputs": [],
   "source": [
    "for edge_type, df in interaction_dfs_concat.items():\n",
    "    if edge_type not in interaction_dfs_final:\n",
    "        interaction_dfs_final[edge_type] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de88516d-7f4c-4027-86b2-a9b2cd30a33e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_network_edges = pd.DataFrame(columns=[\"node_A\", \"node_B\", \"edge_type\"])\n",
    "for df in interaction_dfs_final.values():\n",
    "    df_network_edges = pd.concat((df_network_edges, df), axis=0)\n",
    "df_network_node_aliases = df_network_node_aliases.dropna(\n",
    "    subset=[\"node_alias\"]\n",
    ").reset_index(drop=True)\n",
    "df_network_nodes = (\n",
    "    df_network_node_aliases.loc[:, [\"node_id\", \"node_type\"]]\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "all_edge_node_ids = set(df_network_edges[\"node_A\"]).union(df_network_edges[\"node_B\"])\n",
    "df_isolated = (\n",
    "    df_network_node_aliases[~df_network_node_aliases[\"node_id\"].isin(all_edge_node_ids)]\n",
    "    .loc[:, [\"node_id\", \"node_type\"]]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "df_network_nodes = (\n",
    "    df_network_node_aliases[df_network_node_aliases[\"node_id\"].isin(all_edge_node_ids)]\n",
    "    .loc[:, [\"node_id\", \"node_type\"]]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "df_edge_summary = df_network_edges.set_index([\"node_A\", \"node_B\"])[\n",
    "    \"edge_type\"\n",
    "].value_counts()\n",
    "df_node_summary = df_network_nodes.set_index([\"node_id\"])[\"node_type\"].value_counts()\n",
    "df_isolated_summary = df_isolated.set_index([\"node_id\"])[\"node_type\"].value_counts()\n",
    "\n",
    "print(df_edge_summary, f\"\\n\\nTotal number of edges:\\t {df_edge_summary.sum()}\\n\")\n",
    "print(df_node_summary, f\"\\n\\nTotal number of nodes:\\t {df_node_summary.sum()}\\n\")\n",
    "print(\n",
    "    df_isolated_summary,\n",
    "    f\"\\n\\nTotal number of isolated nodes:\\t {df_isolated_summary.sum()}\\n\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6529a94-ad4d-49b2-8759-27990c8e6ebd",
   "metadata": {},
   "source": [
    "## Export data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4993e47e-f45f-42f6-a978-085008cac74f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_network_edges.to_csv(\n",
    "    get_dirpath(\"database\") / f\"NetworkEdges_{GEM_NAME}.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "df_network_node_aliases.to_csv(\n",
    "    get_dirpath(\"database\") / f\"NetworkNodes_{GEM_NAME}.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81f80eab-0790-49f5-adef-e9076c97100f",
   "metadata": {},
   "source": [
    "## Visualize global interaction network\n",
    "TODO find a better color scheme and additional representations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3643fe49-e65b-48f8-b5df-4109c8598aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "from operator import itemgetter\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172b19b1-d7e1-451c-801c-a141050a549f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_network_edges = pd.read_csv(\n",
    "    get_dirpath(\"database\") / f\"NetworkEdges_{GEM_NAME}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ")\n",
    "df_network_node_aliases = pd.read_csv(\n",
    "    get_dirpath(\"database\") / f\"NetworkNodes_{GEM_NAME}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ")\n",
    "\n",
    "df_network_node_aliases = df_network_node_aliases.dropna(\n",
    "    subset=[\"node_alias\"]\n",
    ").reset_index(drop=True)\n",
    "df_network_nodes = (\n",
    "    df_network_node_aliases.loc[:, [\"node_id\", \"node_type\"]]\n",
    "    .drop_duplicates()\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "\n",
    "all_edge_node_ids = set(df_network_edges[\"node_A\"]).union(df_network_edges[\"node_B\"])\n",
    "df_isolated = (\n",
    "    df_network_node_aliases[~df_network_node_aliases[\"node_id\"].isin(all_edge_node_ids)]\n",
    "    .loc[:, [\"node_id\", \"node_type\"]]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "df_network_nodes = (\n",
    "    df_network_node_aliases[df_network_node_aliases[\"node_id\"].isin(all_edge_node_ids)]\n",
    "    .loc[:, [\"node_id\", \"node_type\"]]\n",
    "    .drop_duplicates()\n",
    ")\n",
    "df_edge_summary = df_network_edges.set_index([\"node_A\", \"node_B\"])[\n",
    "    \"edge_type\"\n",
    "].value_counts()\n",
    "df_node_summary = df_network_nodes.set_index([\"node_id\"])[\"node_type\"].value_counts()\n",
    "df_isolated_summary = df_isolated.set_index([\"node_id\"])[\"node_type\"].value_counts()\n",
    "\n",
    "print(df_edge_summary, f\"\\n\\nTotal number of edges:\\t {df_edge_summary.sum()}\\n\")\n",
    "print(df_node_summary, f\"\\n\\nTotal number of nodes:\\t {df_node_summary.sum()}\\n\")\n",
    "print(\n",
    "    df_isolated_summary,\n",
    "    f\"\\n\\nTotal number of isolated nodes:\\t {df_isolated_summary.sum()}\\n\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e580ab47-f709-4039-b6c3-867924ead4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmap = mpl.colormaps[\"jet\"].resampled(len(df_node_summary.index))\n",
    "node_colors_dict = {\n",
    "    node_type: mpl.colors.to_hex(cmap(idx))\n",
    "    for idx, node_type in enumerate(df_node_summary.index)\n",
    "}\n",
    "\n",
    "cmap = mpl.colormaps[\"viridis\"].resampled(len(df_edge_summary.index))\n",
    "edge_colors_dict = {\n",
    "    node_type: mpl.colors.to_hex(cmap(idx))\n",
    "    for idx, node_type in enumerate(df_edge_summary.index)\n",
    "}\n",
    "# node_colors_dict = {\n",
    "#     \"drug\": mpl.colors.to_hex(\"xkcd:red\"),\n",
    "#     \"snp\":  mpl.colors.to_hex(\"xkcd:blue\"),\n",
    "#     \"gene/protein\":  mpl.colors.to_hex(\"xkcd:yellow\"),\n",
    "#     \"phenotype/disease\":  mpl.colors.to_hex(\"xkcd:pink\"),\n",
    "#     \"reaction\": mpl.colors.to_hex(\"xkcd:green\"),\n",
    "#     \"metabolite\": mpl.colors.to_hex(\"xkcd:purple\"),\n",
    "# }\n",
    "# edge_colors_dict = {\n",
    "#      \"drug-drug\": mpl.colors.to_hex(\"xkcd:red\"),\n",
    "#      \"drug-gene/protein\": mpl.colors.to_hex(\"xkcd:purple\"),\n",
    "#      \"drug-snp\": mpl.colors.to_hex(\"xkcd:orange\"),\n",
    "#      \"gene/protein-snp\": mpl.colors.to_hex(\"xkcd:green\"),\n",
    "#      \"gene/protein-phenotype/disease\": mpl.colors.to_hex(\"xkcd:magenta\"),\n",
    "#      \"gene/protein-reaction\": mpl.colors.to_hex(\"xkcd:light blue\"),\n",
    "#      \"reaction-metabolite\": mpl.colors.to_hex(\"xkcd:light blue\"),\n",
    "# }\n",
    "include_isolated = True\n",
    "fig, ax_nodes = plt.subplots(1, 1, figsize=(10, 5))\n",
    "sns.despine(fig)\n",
    "barsize = 0.5\n",
    "edgecolor = \"black\"\n",
    "edgewidth = 0.8\n",
    "\n",
    "\n",
    "df_nodes = (\n",
    "    pd.concat((df_node_summary, df_isolated_summary), axis=1).fillna(0).astype(int)\n",
    ")\n",
    "df_nodes.columns = [\"Connected\", \"Isolated\"]\n",
    "table_cell_text = []\n",
    "\n",
    "# include_isolated = False\n",
    "offset = np.zeros(len(df_nodes.index))\n",
    "for idx, (label, series) in enumerate(df_nodes.items()):\n",
    "    if idx == 1:\n",
    "        # Lighten the color for isolated nodes\n",
    "        colors = [\n",
    "            mpl.colors.to_hex(mpl.colors.to_rgba_array(node_colors_dict[k]) / 2 + 0.5)\n",
    "            for k in series.index\n",
    "        ]\n",
    "    else:\n",
    "        colors = [node_colors_dict[k] for k in series.index]\n",
    "    rects = ax_nodes.bar(\n",
    "        series.index,\n",
    "        series.values,\n",
    "        barsize,\n",
    "        bottom=offset,\n",
    "        color=colors,\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "    if not include_isolated and idx == 0:\n",
    "        break\n",
    "ax_nodes.set_xticks([])\n",
    "ax_nodes.set_xticklabels([])\n",
    "ax_nodes.set_ylabel(\"Number of nodes\")\n",
    "\n",
    "\n",
    "for node_type_target in df_nodes.index:\n",
    "    row_text = []\n",
    "    for node_type_source in df_nodes.index:\n",
    "        key = f\"{node_type_source}-{node_type_target}\"\n",
    "        try:\n",
    "            value = df_edge_summary[key]\n",
    "        except KeyError:\n",
    "            try:\n",
    "                value = df_edge_summary[f\"{node_type_target}-{node_type_source}\"]\n",
    "            except KeyError:\n",
    "                value = \"\"\n",
    "        row_text.append(str(value))\n",
    "    table_cell_text.append(row_text)\n",
    "table = ax_nodes.table(\n",
    "    table_cell_text,\n",
    "    rowLabels=df_nodes.index,\n",
    "    colLabels=df_nodes.index,\n",
    "    loc=\"bottom\",\n",
    "    cellLoc=\"center\",\n",
    "    rowLoc=\"center\",\n",
    "    colLoc=\"center\",\n",
    ")\n",
    "table.auto_set_font_size(False)\n",
    "table.set_fontsize(9)\n",
    "ax_nodes.annotate(\n",
    "    \"# of Interactions\",\n",
    "    xy=(-0.1, -0.03),\n",
    "    xycoords=ax_nodes.transAxes,\n",
    "    fontsize=9,\n",
    "    ha=\"center\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "fig;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9e2b8e9-f830-4ed1-99cd-52b488c60d39",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_nodes = df_network_nodes.copy()\n",
    "# df_edges = df_network_edges.copy()\n",
    "\n",
    "# df_nodes[\"color\"] = df_nodes[\"node_type\"].apply(lambda x: node_colors_dict[x])\n",
    "# df_edges[\"color\"] = df_edges[\"edge_type\"].apply(lambda x: edge_colors_dict[x])\n",
    "\n",
    "# fig, ax = plt.subplots(1, 1, figsize=(10, 10))\n",
    "# nx_graph = nx.from_pandas_edgelist(\n",
    "#     df=df_edges,\n",
    "#     source=\"node_A\",\n",
    "#     target=\"node_B\",\n",
    "#     create_using=nx.Graph,\n",
    "#     edge_attr=True,\n",
    "# )\n",
    "# nx_graph.add_nodes_from(\n",
    "#     (n, dict(d))\n",
    "#     for n, d in df_nodes.set_index(\"node_id\").iterrows()\n",
    "#     if n in nx_graph.nodes\n",
    "# )\n",
    "\n",
    "# node_colors = [attr_dict.get(\"color\", \"grey\") for (node_id, attr_dict) in nx_graph.nodes(data=True)]\n",
    "# edge_colors = [attr_dict.get(\"color\", \"grey\") for (x, y, attr_dict) in nx_graph.edges(data=True)]\n",
    "\n",
    "# pos = nx.circular_layout(nx_graph, scale=1, center=None, dim=2)\n",
    "# nx.draw_networkx_nodes(\n",
    "#     nx_graph,\n",
    "#     pos,\n",
    "#     ax=ax,\n",
    "#     # nodelist=None,\n",
    "#     # node_size=300,\n",
    "#     node_color=node_colors,\n",
    "#     # node_shape='o',\n",
    "#     # alpha=None,\n",
    "#     # cmap=None,\n",
    "#     # vmin=None,\n",
    "#     # vmax=None,\n",
    "#     # linewidths=None,\n",
    "#     # edgecolors=None,\n",
    "#     # label=None,\n",
    "#     # margins=None,\n",
    "# )\n",
    "# nx.draw_networkx_edges(\n",
    "#     nx_graph,\n",
    "#     pos,\n",
    "#     ax=ax,\n",
    "#     # edgelist=None,\n",
    "#     # width=1.0,\n",
    "#     edge_color=edge_colors,\n",
    "#     # style='solid',\n",
    "#     # alpha=None,\n",
    "#     # arrowstyle=None,\n",
    "#     # arrowsize=10,\n",
    "#     # edge_cmap=None,\n",
    "#     # edge_vmin=None,\n",
    "#     # edge_vmax=None,\n",
    "#     # arrows=None,\n",
    "#     # label=None,\n",
    "#     # node_size=300,\n",
    "#     # nodelist=None,\n",
    "#     # node_shape='o',\n",
    "#     # connectionstyle='arc3',\n",
    "#     # min_source_margin=0,\n",
    "#     # min_target_margin=0,\n",
    "# )\n",
    "# fig;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af31c364-3fcb-41bf-a0a2-3bcb762a4c0a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
