{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "13bddfd6-19f9-44de-9062-5462c031dfb5",
   "metadata": {},
   "source": [
    "# Extract data from dbSNP\n",
    "\n",
    "Utilize the [Bio.Entrez package](https://biopython.org/docs/latest/api/Bio.Entrez.html) to get data from [dbSNP](https://www.ncbi.nlm.nih.gov/snp/).\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "49fed61a-e9eb-47f9-b97e-8edeff408452",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.1\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.12.3\n",
      "bio                                   1.6.2\n",
      "cobra                                0.29.0\n",
      "depinfo                               2.2.0\n",
      "kaleido                               0.2.1\n",
      "matplotlib                            3.8.2\n",
      "memote                               0.17.0\n",
      "networkx                              3.2.1\n",
      "notebook                              7.0.7\n",
      "openpyxl                              3.1.2\n",
      "pandas                                2.2.0\n",
      "pre-commit                            3.6.0\n",
      "pyvis                                 0.3.2\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.31.0\n",
      "scipy                                1.12.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip        23.3.1\n",
      "setuptools 68.2.2\n",
      "wheel      0.41.2\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Darwin  22.6.0-x86_64\n",
      "CPython        3.11.7\n"
     ]
    }
   ],
   "source": [
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from xml.etree import ElementTree\n",
    "\n",
    "from rbc_gem_utils import (\n",
    "    ROOT_PATH,\n",
    "    DATABASE_PATH,\n",
    "    GEM_NAME,\n",
    "    read_rbc_model,\n",
    "    build_string,\n",
    "    split_string,\n",
    "    get_annotation_df,\n",
    "    show_versions,\n",
    ")\n",
    "from rbc_gem_utils.util import has_value_type, explode_column\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6da6a907-3339-4f93-868a-01622f9e6de6",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13d4753e-2d6d-44d9-bad0-7a477220b3a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>110ea3290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1971</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2798</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>656</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>75</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x110ea3290>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_rbc_model(filetype=\"xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1a26d35-de9c-4390-aca2-a6c8481eac4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes: 656\n",
      "dbsnp: 4308\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPIA</td>\n",
       "      <td>rs121918591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs145813597;rs1042079;rs930337;rs149975952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKR7A2</td>\n",
       "      <td>rs6670759;rs1043657;rs859208;rs859210;rs223120...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRM</td>\n",
       "      <td>rs1049932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>SLC9B2</td>\n",
       "      <td>rs7672710;rs7672707;rs2276976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>SMPD1</td>\n",
       "      <td>rs550365194;rs1050228;rs3838786;rs748589919;rs...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>653</th>\n",
       "      <td>SPHK1</td>\n",
       "      <td>rs346803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>rs3802829;rs121434390;rs146776939;rs121434391;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>655</th>\n",
       "      <td>TSPO2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>656 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes                                              dbsnp\n",
       "0       RPE                                                NaN\n",
       "1      RPIA                                        rs121918591\n",
       "2      SORD         rs145813597;rs1042079;rs930337;rs149975952\n",
       "3    AKR7A2  rs6670759;rs1043657;rs859208;rs859210;rs223120...\n",
       "4       SRM                                          rs1049932\n",
       "..      ...                                                ...\n",
       "651  SLC9B2                      rs7672710;rs7672707;rs2276976\n",
       "652   SMPD1  rs550365194;rs1050228;rs3838786;rs748589919;rs...\n",
       "653   SPHK1                                           rs346803\n",
       "654   TRPC6  rs3802829;rs121434390;rs146776939;rs121434391;...\n",
       "655   TSPO2                                                NaN\n",
       "\n",
       "[656 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_type = \"genes\"\n",
    "df_model_mappings = get_annotation_df(\n",
    "    getattr(model, annotation_type), [\"dbsnp\"]\n",
    ").rename({\"id\": annotation_type}, axis=1)\n",
    "for col in df_model_mappings.columns:\n",
    "    df = (\n",
    "        df_model_mappings[col]\n",
    "        .apply(lambda x: split_string(x))\n",
    "        .explode(col)\n",
    "        .drop_duplicates()\n",
    "    )\n",
    "    print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21905f9a-fb31-466a-84c5-cf9ee6681287",
   "metadata": {},
   "source": [
    "### Get IDs for query from model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ae646a47-2606-4d32-bb97-bafb1413a2ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPIA</td>\n",
       "      <td>rs121918591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs145813597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs1042079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs930337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs149975952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>rs777715086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>rs36111323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>rs771594597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>rs121434394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>654</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>rs121434395</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4308 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genes        dbsnp\n",
       "1     RPIA  rs121918591\n",
       "2     SORD  rs145813597\n",
       "2     SORD    rs1042079\n",
       "2     SORD     rs930337\n",
       "2     SORD  rs149975952\n",
       "..     ...          ...\n",
       "654  TRPC6  rs777715086\n",
       "654  TRPC6   rs36111323\n",
       "654  TRPC6  rs771594597\n",
       "654  TRPC6  rs121434394\n",
       "654  TRPC6  rs121434395\n",
       "\n",
       "[4308 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_snps = df_model_mappings.loc[:, [\"genes\", \"dbsnp\"]].copy()\n",
    "df_model_snps[\"dbsnp\"] = df_model_mappings[\"dbsnp\"].apply(split_string)\n",
    "df_model_snps = df_model_snps.explode(\"dbsnp\")\n",
    "df_model_snps = df_model_snps.dropna().drop_duplicates()\n",
    "rs_ids = df_model_snps[\"dbsnp\"].unique()\n",
    "df_model_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cdc8c330-2057-4c12-820d-4aa9cdf72b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_results_dbSNP(email, rs_ids):\n",
    "    Entrez.email = email\n",
    "    search = Entrez.efetch(db=\"snp\", retmode=\"text\", id=rs_ids)\n",
    "    all_lines = search.readlines()\n",
    "    search.close()\n",
    "    return all_lines"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e996f2bd-5293-4f6d-99b5-345a172f0840",
   "metadata": {},
   "source": [
    "### Get results from SNP as a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ef42f63-2692-4dfa-a53c-4a2cd679c3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "overwrite = True\n",
    "email = \"zhaiman@ucsd.edu\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2e253df0-8aee-48f6-87a2-6f9198e5ad18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching results for batch 1  (500/4308)\n",
      "Fetching results for batch 2  (1000/4308)\n",
      "Fetching results for batch 3  (1500/4308)\n",
      "Fetching results for batch 4  (2000/4308)\n",
      "Fetching results for batch 5  (2500/4308)\n",
      "Fetching results for batch 6  (3000/4308)\n",
      "Fetching results for batch 7  (3500/4308)\n",
      "Fetching results for batch 8  (4000/4308)\n",
      "Fetching results for batch 9  (4308/4308)\n",
      "\n",
      "Number of failed IDs: 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GLOBAL_SAMPLESIZE</th>\n",
       "      <th>CLINICAL_SIGNIFICANCE</th>\n",
       "      <th>ACC</th>\n",
       "      <th>CHR</th>\n",
       "      <th>HANDLE</th>\n",
       "      <th>SPDI</th>\n",
       "      <th>FXN_CLASS</th>\n",
       "      <th>VALIDATED</th>\n",
       "      <th>DOCSUM</th>\n",
       "      <th>...</th>\n",
       "      <th>SS</th>\n",
       "      <th>ALLELE</th>\n",
       "      <th>SNP_CLASS</th>\n",
       "      <th>CHRPOS</th>\n",
       "      <th>CHRPOS_PREV_ASSM</th>\n",
       "      <th>SNP_ID_SORT</th>\n",
       "      <th>CLINICAL_SORT</th>\n",
       "      <th>CHRPOS_SORT</th>\n",
       "      <th>MERGED_SORT</th>\n",
       "      <th>error</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbsnp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs121918591</th>\n",
       "      <td>121918591</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>TOPMED,ILLUMINA,PAGE_CC,OMIM-CURATED-RECORDS,G...</td>\n",
       "      <td>NC_000002.12:88729278:C:T</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000002.12:g.88729279C&gt;T,NC_000002.11:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>275518424,1958438289,2732810607,3021993357,365...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>2:88729279</td>\n",
       "      <td>2:89028797</td>\n",
       "      <td>0121918591</td>\n",
       "      <td>1</td>\n",
       "      <td>0088729279</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs145813597</th>\n",
       "      <td>145813597</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic-likely-pathogenic</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>AFFY,TOPMED,EVA_EXAC,HUGCELL_USP,ILLUMINA,1000...</td>\n",
       "      <td>NC_000015.10:45065302:C:A,NC_000015.10:4506530...</td>\n",
       "      <td>non_coding_transcript_variant,missense_variant...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45065303C&gt;A,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>342402496,464140601,491493966,491700043,991737...</td>\n",
       "      <td>H</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45065303</td>\n",
       "      <td>15:45357501</td>\n",
       "      <td>0145813597</td>\n",
       "      <td>1</td>\n",
       "      <td>0045065303</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1042079</th>\n",
       "      <td>1042079</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>EVA-GONL,CNG,BUSHMAN,SSMP,1000GENOMES,EXOME_CH...</td>\n",
       "      <td>NC_000015.10:45068981:A:G,NC_000015.10:4506898...</td>\n",
       "      <td>coding_sequence_variant,non_coding_transcript_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45068982A&gt;G,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1509008,4403349,76874733,95213358,156480828,15...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45068982</td>\n",
       "      <td>15:45361180</td>\n",
       "      <td>0001042079</td>\n",
       "      <td>1</td>\n",
       "      <td>0045068982</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs930337</th>\n",
       "      <td>930337</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSEMBL,CNG,SYSTEMSBIOZJU,SSMP,1000GENOMES,EXO...</td>\n",
       "      <td>NC_000015.10:45072335:A:C</td>\n",
       "      <td>coding_sequence_variant,non_coding_transcript_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45072336A&gt;C,NC_000015.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>1368774,2558685,3214527,5252928,21245444,24802...</td>\n",
       "      <td>M</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45072336</td>\n",
       "      <td>15:45364534</td>\n",
       "      <td>0000930337</td>\n",
       "      <td>1</td>\n",
       "      <td>0045072336</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs149975952</th>\n",
       "      <td>149975952</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>ILLUMINA,WEILL_CORNELL_DGM,1000GENOMES,EXOME_C...</td>\n",
       "      <td>NC_000015.10:45073419:G:A,NC_000015.10:4507341...</td>\n",
       "      <td>coding_sequence_variant,missense_variant,non_c...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45073420G&gt;A,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>342402507,489048604,491493972,1353273008,16918...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45073420</td>\n",
       "      <td>15:45365618</td>\n",
       "      <td>0149975952</td>\n",
       "      <td>1</td>\n",
       "      <td>0045073420</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs777715086</th>\n",
       "      <td>777715086</td>\n",
       "      <td>0</td>\n",
       "      <td>uncertain-significance</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>EVA_EXAC,TOPMED,ILLUMINA,GNOMAD,TOMMO_GENOMICS</td>\n",
       "      <td>NC_000011.10:101491604:C:T</td>\n",
       "      <td>missense_variant,coding_sequence_variant,intro...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.101491605C&gt;T,NC_000011.9:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1690602079,2739340408,3021369583,3651736504,37...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:101491605</td>\n",
       "      <td>11:101362336</td>\n",
       "      <td>0777715086</td>\n",
       "      <td>1</td>\n",
       "      <td>0101491605</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs36111323</th>\n",
       "      <td>36111323</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>GNOMAD,GRF,HAMMER_LAB,1000GENOMES,JMKIDD_LAB,S...</td>\n",
       "      <td>NC_000011.10:101489018:G:A,NC_000011.10:101489...</td>\n",
       "      <td>coding_sequence_variant,intron_variant,missens...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.101489019G&gt;A,NC_000011.10:...</td>\n",
       "      <td>...</td>\n",
       "      <td>38731112,52066061,97412791,110958554,142662462...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:101489019</td>\n",
       "      <td>11:101359750</td>\n",
       "      <td>0036111323</td>\n",
       "      <td>1</td>\n",
       "      <td>0101489019</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs771594597</th>\n",
       "      <td>771594597</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>GNOMAD,EVA_EXAC</td>\n",
       "      <td>NC_000011.10:101471252:A:G</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.101471253A&gt;G,NC_000011.9:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1690601835,2739340052</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:101471253</td>\n",
       "      <td>11:101341984</td>\n",
       "      <td>0771594597</td>\n",
       "      <td>0</td>\n",
       "      <td>0101471253</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121434394</th>\n",
       "      <td>121434394</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>OMIM-CURATED-RECORDS,ILLUMINA</td>\n",
       "      <td>NC_000011.10:101453067:G:A</td>\n",
       "      <td>genic_downstream_transcript_variant,missense_v...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.101453068G&gt;A,NC_000011.9:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>256301976,3725270345</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:101453068</td>\n",
       "      <td>11:101323799</td>\n",
       "      <td>0121434394</td>\n",
       "      <td>1</td>\n",
       "      <td>0101453068</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs121434395</th>\n",
       "      <td>121434395</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>OMIM-CURATED-RECORDS,ILLUMINA</td>\n",
       "      <td>NC_000011.10:101453061:C:T</td>\n",
       "      <td>genic_downstream_transcript_variant,missense_v...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.101453062C&gt;T,NC_000011.9:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>256301977,3725270344</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:101453062</td>\n",
       "      <td>11:101323793</td>\n",
       "      <td>0121434395</td>\n",
       "      <td>1</td>\n",
       "      <td>0101453062</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4306 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNP_ID GLOBAL_SAMPLESIZE         CLINICAL_SIGNIFICANCE  \\\n",
       "dbsnp                                                                    \n",
       "rs121918591  121918591                 0                    pathogenic   \n",
       "rs145813597  145813597                 0  pathogenic-likely-pathogenic   \n",
       "rs1042079      1042079                 0                        benign   \n",
       "rs930337        930337                 0                        benign   \n",
       "rs149975952  149975952                 0                        benign   \n",
       "...                ...               ...                           ...   \n",
       "rs777715086  777715086                 0        uncertain-significance   \n",
       "rs36111323    36111323                 0                        benign   \n",
       "rs771594597  771594597                 0                           NaN   \n",
       "rs121434394  121434394                 0                    pathogenic   \n",
       "rs121434395  121434395                 0                    pathogenic   \n",
       "\n",
       "                      ACC CHR  \\\n",
       "dbsnp                           \n",
       "rs121918591  NC_000002.12   2   \n",
       "rs145813597  NC_000015.10  15   \n",
       "rs1042079    NC_000015.10  15   \n",
       "rs930337     NC_000015.10  15   \n",
       "rs149975952  NC_000015.10  15   \n",
       "...                   ...  ..   \n",
       "rs777715086  NC_000011.10  11   \n",
       "rs36111323   NC_000011.10  11   \n",
       "rs771594597  NC_000011.10  11   \n",
       "rs121434394  NC_000011.10  11   \n",
       "rs121434395  NC_000011.10  11   \n",
       "\n",
       "                                                        HANDLE  \\\n",
       "dbsnp                                                            \n",
       "rs121918591  TOPMED,ILLUMINA,PAGE_CC,OMIM-CURATED-RECORDS,G...   \n",
       "rs145813597  AFFY,TOPMED,EVA_EXAC,HUGCELL_USP,ILLUMINA,1000...   \n",
       "rs1042079    EVA-GONL,CNG,BUSHMAN,SSMP,1000GENOMES,EXOME_CH...   \n",
       "rs930337     ENSEMBL,CNG,SYSTEMSBIOZJU,SSMP,1000GENOMES,EXO...   \n",
       "rs149975952  ILLUMINA,WEILL_CORNELL_DGM,1000GENOMES,EXOME_C...   \n",
       "...                                                        ...   \n",
       "rs777715086     EVA_EXAC,TOPMED,ILLUMINA,GNOMAD,TOMMO_GENOMICS   \n",
       "rs36111323   GNOMAD,GRF,HAMMER_LAB,1000GENOMES,JMKIDD_LAB,S...   \n",
       "rs771594597                                    GNOMAD,EVA_EXAC   \n",
       "rs121434394                      OMIM-CURATED-RECORDS,ILLUMINA   \n",
       "rs121434395                      OMIM-CURATED-RECORDS,ILLUMINA   \n",
       "\n",
       "                                                          SPDI  \\\n",
       "dbsnp                                                            \n",
       "rs121918591                          NC_000002.12:88729278:C:T   \n",
       "rs145813597  NC_000015.10:45065302:C:A,NC_000015.10:4506530...   \n",
       "rs1042079    NC_000015.10:45068981:A:G,NC_000015.10:4506898...   \n",
       "rs930337                             NC_000015.10:45072335:A:C   \n",
       "rs149975952  NC_000015.10:45073419:G:A,NC_000015.10:4507341...   \n",
       "...                                                        ...   \n",
       "rs777715086                         NC_000011.10:101491604:C:T   \n",
       "rs36111323   NC_000011.10:101489018:G:A,NC_000011.10:101489...   \n",
       "rs771594597                         NC_000011.10:101471252:A:G   \n",
       "rs121434394                         NC_000011.10:101453067:G:A   \n",
       "rs121434395                         NC_000011.10:101453061:C:T   \n",
       "\n",
       "                                                     FXN_CLASS  \\\n",
       "dbsnp                                                            \n",
       "rs121918591           coding_sequence_variant,missense_variant   \n",
       "rs145813597  non_coding_transcript_variant,missense_variant...   \n",
       "rs1042079    coding_sequence_variant,non_coding_transcript_...   \n",
       "rs930337     coding_sequence_variant,non_coding_transcript_...   \n",
       "rs149975952  coding_sequence_variant,missense_variant,non_c...   \n",
       "...                                                        ...   \n",
       "rs777715086  missense_variant,coding_sequence_variant,intro...   \n",
       "rs36111323   coding_sequence_variant,intron_variant,missens...   \n",
       "rs771594597           coding_sequence_variant,missense_variant   \n",
       "rs121434394  genic_downstream_transcript_variant,missense_v...   \n",
       "rs121434395  genic_downstream_transcript_variant,missense_v...   \n",
       "\n",
       "                                   VALIDATED  \\\n",
       "dbsnp                                          \n",
       "rs121918591  by-frequency,by-alfa,by-cluster   \n",
       "rs145813597  by-frequency,by-alfa,by-cluster   \n",
       "rs1042079    by-frequency,by-alfa,by-cluster   \n",
       "rs930337     by-frequency,by-alfa,by-cluster   \n",
       "rs149975952  by-frequency,by-alfa,by-cluster   \n",
       "...                                      ...   \n",
       "rs777715086  by-frequency,by-alfa,by-cluster   \n",
       "rs36111323   by-frequency,by-alfa,by-cluster   \n",
       "rs771594597  by-frequency,by-alfa,by-cluster   \n",
       "rs121434394  by-frequency,by-alfa,by-cluster   \n",
       "rs121434395  by-frequency,by-alfa,by-cluster   \n",
       "\n",
       "                                                        DOCSUM  ...  \\\n",
       "dbsnp                                                           ...   \n",
       "rs121918591  HGVS=NC_000002.12:g.88729279C>T,NC_000002.11:g...  ...   \n",
       "rs145813597  HGVS=NC_000015.10:g.45065303C>A,NC_000015.10:g...  ...   \n",
       "rs1042079    HGVS=NC_000015.10:g.45068982A>G,NC_000015.10:g...  ...   \n",
       "rs930337     HGVS=NC_000015.10:g.45072336A>C,NC_000015.9:g....  ...   \n",
       "rs149975952  HGVS=NC_000015.10:g.45073420G>A,NC_000015.10:g...  ...   \n",
       "...                                                        ...  ...   \n",
       "rs777715086  HGVS=NC_000011.10:g.101491605C>T,NC_000011.9:g...  ...   \n",
       "rs36111323   HGVS=NC_000011.10:g.101489019G>A,NC_000011.10:...  ...   \n",
       "rs771594597  HGVS=NC_000011.10:g.101471253A>G,NC_000011.9:g...  ...   \n",
       "rs121434394  HGVS=NC_000011.10:g.101453068G>A,NC_000011.9:g...  ...   \n",
       "rs121434395  HGVS=NC_000011.10:g.101453062C>T,NC_000011.9:g...  ...   \n",
       "\n",
       "                                                            SS ALLELE  \\\n",
       "dbsnp                                                                   \n",
       "rs121918591  275518424,1958438289,2732810607,3021993357,365...      Y   \n",
       "rs145813597  342402496,464140601,491493966,491700043,991737...      H   \n",
       "rs1042079    1509008,4403349,76874733,95213358,156480828,15...      D   \n",
       "rs930337     1368774,2558685,3214527,5252928,21245444,24802...      M   \n",
       "rs149975952  342402507,489048604,491493972,1353273008,16918...      D   \n",
       "...                                                        ...    ...   \n",
       "rs777715086  1690602079,2739340408,3021369583,3651736504,37...      Y   \n",
       "rs36111323   38731112,52066061,97412791,110958554,142662462...      D   \n",
       "rs771594597                              1690601835,2739340052      R   \n",
       "rs121434394                               256301976,3725270345      R   \n",
       "rs121434395                               256301977,3725270344      Y   \n",
       "\n",
       "            SNP_CLASS        CHRPOS CHRPOS_PREV_ASSM SNP_ID_SORT  \\\n",
       "dbsnp                                                              \n",
       "rs121918591       snv    2:88729279       2:89028797  0121918591   \n",
       "rs145813597       snv   15:45065303      15:45357501  0145813597   \n",
       "rs1042079         snv   15:45068982      15:45361180  0001042079   \n",
       "rs930337          snv   15:45072336      15:45364534  0000930337   \n",
       "rs149975952       snv   15:45073420      15:45365618  0149975952   \n",
       "...               ...           ...              ...         ...   \n",
       "rs777715086       snv  11:101491605     11:101362336  0777715086   \n",
       "rs36111323        snv  11:101489019     11:101359750  0036111323   \n",
       "rs771594597       snv  11:101471253     11:101341984  0771594597   \n",
       "rs121434394       snv  11:101453068     11:101323799  0121434394   \n",
       "rs121434395       snv  11:101453062     11:101323793  0121434395   \n",
       "\n",
       "            CLINICAL_SORT CHRPOS_SORT MERGED_SORT error  \n",
       "dbsnp                                                    \n",
       "rs121918591             1  0088729279           0   NaN  \n",
       "rs145813597             1  0045065303           0   NaN  \n",
       "rs1042079               1  0045068982           0   NaN  \n",
       "rs930337                1  0045072336           0   NaN  \n",
       "rs149975952             1  0045073420           0   NaN  \n",
       "...                   ...         ...         ...   ...  \n",
       "rs777715086             1  0101491605           0   NaN  \n",
       "rs36111323              1  0101489019           0   NaN  \n",
       "rs771594597             0  0101471253           0   NaN  \n",
       "rs121434394             1  0101453068           0   NaN  \n",
       "rs121434395             1  0101453062           0   NaN  \n",
       "\n",
       "[4306 rows x 25 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = defaultdict(dict)\n",
    "batch_size = 500\n",
    "for batch_idx, batch in enumerate(np.arange(0, len(rs_ids), batch_size), start=0):\n",
    "    query_ids = rs_ids[batch : batch + batch_size]\n",
    "    print(\n",
    "        f\"Fetching results for batch {batch_idx + 1}  ({batch + len(query_ids)}/{len(rs_ids)})\"\n",
    "    )\n",
    "    all_lines = fetch_results_dbSNP(email=email, rs_ids=query_ids)\n",
    "    for entry_idx, xml_str in enumerate(all_lines, start=batch_idx * batch_size):\n",
    "        entry = ElementTree.fromstring(xml_str)\n",
    "        data[rs_ids[entry_idx]].update(\n",
    "            {e.tag: e.text for e in entry if has_value_type(e)}\n",
    "        )\n",
    "\n",
    "# Make into DataFrame\n",
    "df_snps_all = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "failed_ids = list(df_snps_all[\"error\"].dropna().index)\n",
    "print(f\"\\nNumber of failed IDs: {len(failed_ids)}\")\n",
    "df_snps_all.index.name = \"dbsnp\"\n",
    "df_snps_all = df_snps_all[~df_snps_all[\"error\"].notna()]\n",
    "df_snps_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "901fb1a6-0667-4b1b-a90a-657395466b5b",
   "metadata": {},
   "source": [
    "### Parse and reformat results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "be8c543e-c172-4d1c-bd42-8bbdf88d8ff8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>handle</th>\n",
       "      <th>spdi</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>...</th>\n",
       "      <th>ss</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "      <th>sequence_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>rs334</td>\n",
       "      <td>334</td>\n",
       "      <td>protective;other;pathogenic;benign;likely-beni...</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...</td>\n",
       "      <td>NC_000011.10:5227001:T:A;NC_000011.10:5227001:...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>335;24811263;48419811;49850538;49850540;498505...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>11:5248232</td>\n",
       "      <td>0000000334</td>\n",
       "      <td>1</td>\n",
       "      <td>0005227002</td>\n",
       "      <td>0</td>\n",
       "      <td>[T:A, T:C, T:G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSTP1</td>\n",
       "      <td>rs1695</td>\n",
       "      <td>1695</td>\n",
       "      <td>benign;not-provided</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...</td>\n",
       "      <td>NC_000011.10:67585217:A:G;NC_000011.10:6758521...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>1720;5330;9110;1390210;1545478;3178003;5586573...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:67585218</td>\n",
       "      <td>11:67352689</td>\n",
       "      <td>0000001695</td>\n",
       "      <td>1</td>\n",
       "      <td>0067585218</td>\n",
       "      <td>0</td>\n",
       "      <td>[A:G, A:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FKBP2</td>\n",
       "      <td>rs4672</td>\n",
       "      <td>4672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...</td>\n",
       "      <td>NC_000011.10:64242406:G:A;NC_000011.10:6424240...</td>\n",
       "      <td>3_prime_UTR_variant;intron_variant;missense_va...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>1555019;4436250;20825221;24790172;39886556;752...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:64242407</td>\n",
       "      <td>11:64009879</td>\n",
       "      <td>0000004672</td>\n",
       "      <td>0</td>\n",
       "      <td>0064242407</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:A, G:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COMT</td>\n",
       "      <td>rs4680</td>\n",
       "      <td>4680</td>\n",
       "      <td>drug-response;benign</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...</td>\n",
       "      <td>NC_000022.11:19963747:G:A</td>\n",
       "      <td>2KB_upstream_variant;coding_sequence_variant;u...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>5363;10902;232568;458935;593353;869771;1058120...</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:19963748</td>\n",
       "      <td>22:19951271</td>\n",
       "      <td>0000004680</td>\n",
       "      <td>1</td>\n",
       "      <td>0019963748</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLO1</td>\n",
       "      <td>rs4746</td>\n",
       "      <td>4746</td>\n",
       "      <td>uncertain-significance</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>6</td>\n",
       "      <td>ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...</td>\n",
       "      <td>NC_000006.12:38682851:T:A;NC_000006.12:3868285...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>3863270;20275576;24369377;28505662;44730940;48...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>6:38682852</td>\n",
       "      <td>6:38650628</td>\n",
       "      <td>0000004746</td>\n",
       "      <td>1</td>\n",
       "      <td>0038682852</td>\n",
       "      <td>0</td>\n",
       "      <td>[T:A, T:G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4301</th>\n",
       "      <td>VPS4A</td>\n",
       "      <td>rs1965499910</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000016.10:69320767:A:G;NC_000016.10:6932076...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>16:69320768</td>\n",
       "      <td>16:69354671</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>1</td>\n",
       "      <td>0069320768</td>\n",
       "      <td>0</td>\n",
       "      <td>[A:G, A:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4302</th>\n",
       "      <td>FARSA</td>\n",
       "      <td>rs1971301248</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000019.10:12924491:G:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:12924492</td>\n",
       "      <td>19:13035306</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>1</td>\n",
       "      <td>0012924492</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4303</th>\n",
       "      <td>CNP</td>\n",
       "      <td>rs2050933471</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000017.11:41968308:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>17:41968309</td>\n",
       "      <td>17:40120327</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>1</td>\n",
       "      <td>0041968309</td>\n",
       "      <td>0</td>\n",
       "      <td>[C:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4304</th>\n",
       "      <td>PISD</td>\n",
       "      <td>rs2072505076</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>pathogenic-likely-pathogenic</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000022.11:31620658:C:T</td>\n",
       "      <td>coding_sequence_variant;intron_variant;missens...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:31620659</td>\n",
       "      <td>22:32016645</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>1</td>\n",
       "      <td>0031620659</td>\n",
       "      <td>0</td>\n",
       "      <td>[C:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4305</th>\n",
       "      <td>DNM2</td>\n",
       "      <td>rs2072577342</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>likely-pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>TOPMED</td>\n",
       "      <td>NC_000019.10:10812269:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa</td>\n",
       "      <td>...</td>\n",
       "      <td>5068054504</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:10812270</td>\n",
       "      <td>19:10922946</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>1</td>\n",
       "      <td>0010812270</td>\n",
       "      <td>0</td>\n",
       "      <td>[C:T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4306 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes         dbsnp      snp_id  \\\n",
       "0       HBB         rs334         334   \n",
       "1     GSTP1        rs1695        1695   \n",
       "2     FKBP2        rs4672        4672   \n",
       "3      COMT        rs4680        4680   \n",
       "4      GLO1        rs4746        4746   \n",
       "...     ...           ...         ...   \n",
       "4301  VPS4A  rs1965499910  1965499910   \n",
       "4302  FARSA  rs1971301248  1971301248   \n",
       "4303    CNP  rs2050933471  2050933471   \n",
       "4304   PISD  rs2072505076  2072505076   \n",
       "4305   DNM2  rs2072577342  2072577342   \n",
       "\n",
       "                                  clinical_significance           acc chr  \\\n",
       "0     protective;other;pathogenic;benign;likely-beni...  NC_000011.10  11   \n",
       "1                                   benign;not-provided  NC_000011.10  11   \n",
       "2                                                   NaN  NC_000011.10  11   \n",
       "3                                  drug-response;benign  NC_000022.11  22   \n",
       "4                                uncertain-significance  NC_000006.12   6   \n",
       "...                                                 ...           ...  ..   \n",
       "4301                                         pathogenic  NC_000016.10  16   \n",
       "4302                                         pathogenic  NC_000019.10  19   \n",
       "4303                                         pathogenic  NC_000017.11  17   \n",
       "4304                       pathogenic-likely-pathogenic  NC_000022.11  22   \n",
       "4305                                  likely-pathogenic  NC_000019.10  19   \n",
       "\n",
       "                                                 handle  \\\n",
       "0     JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...   \n",
       "1     CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...   \n",
       "2     JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...   \n",
       "3     GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...   \n",
       "4     ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...   \n",
       "...                                                 ...   \n",
       "4301                                               None   \n",
       "4302                                               None   \n",
       "4303                                               None   \n",
       "4304                                               None   \n",
       "4305                                             TOPMED   \n",
       "\n",
       "                                                   spdi  \\\n",
       "0     NC_000011.10:5227001:T:A;NC_000011.10:5227001:...   \n",
       "1     NC_000011.10:67585217:A:G;NC_000011.10:6758521...   \n",
       "2     NC_000011.10:64242406:G:A;NC_000011.10:6424240...   \n",
       "3                             NC_000022.11:19963747:G:A   \n",
       "4     NC_000006.12:38682851:T:A;NC_000006.12:3868285...   \n",
       "...                                                 ...   \n",
       "4301  NC_000016.10:69320767:A:G;NC_000016.10:6932076...   \n",
       "4302                          NC_000019.10:12924491:G:T   \n",
       "4303                          NC_000017.11:41968308:C:T   \n",
       "4304                          NC_000022.11:31620658:C:T   \n",
       "4305                          NC_000019.10:10812269:C:T   \n",
       "\n",
       "                                              fxn_class  \\\n",
       "0              missense_variant;coding_sequence_variant   \n",
       "1              missense_variant;coding_sequence_variant   \n",
       "2     3_prime_UTR_variant;intron_variant;missense_va...   \n",
       "3     2KB_upstream_variant;coding_sequence_variant;u...   \n",
       "4              coding_sequence_variant;missense_variant   \n",
       "...                                                 ...   \n",
       "4301           missense_variant;coding_sequence_variant   \n",
       "4302           coding_sequence_variant;missense_variant   \n",
       "4303           coding_sequence_variant;missense_variant   \n",
       "4304  coding_sequence_variant;intron_variant;missens...   \n",
       "4305           coding_sequence_variant;missense_variant   \n",
       "\n",
       "                            validated  ...  \\\n",
       "0     by-frequency;by-alfa;by-cluster  ...   \n",
       "1     by-frequency;by-alfa;by-cluster  ...   \n",
       "2     by-frequency;by-alfa;by-cluster  ...   \n",
       "3     by-frequency;by-alfa;by-cluster  ...   \n",
       "4     by-frequency;by-alfa;by-cluster  ...   \n",
       "...                               ...  ...   \n",
       "4301                              NaN  ...   \n",
       "4302                              NaN  ...   \n",
       "4303                              NaN  ...   \n",
       "4304                              NaN  ...   \n",
       "4305             by-frequency;by-alfa  ...   \n",
       "\n",
       "                                                     ss allele snp_class  \\\n",
       "0     335;24811263;48419811;49850538;49850540;498505...      N       snv   \n",
       "1     1720;5330;9110;1390210;1545478;3178003;5586573...      D       snv   \n",
       "2     1555019;4436250;20825221;24790172;39886556;752...      D       snv   \n",
       "3     5363;10902;232568;458935;593353;869771;1058120...      R       snv   \n",
       "4     3863270;20275576;24369377;28505662;44730940;48...      D       snv   \n",
       "...                                                 ...    ...       ...   \n",
       "4301                                                  0      D       snv   \n",
       "4302                                                  0      K       snv   \n",
       "4303                                                  0      Y       snv   \n",
       "4304                                                  0      Y       snv   \n",
       "4305                                         5068054504      Y       snv   \n",
       "\n",
       "           chrpos chrpos_prev_assm snp_id_sort clinical_sort chrpos_sort  \\\n",
       "0      11:5227002       11:5248232  0000000334             1  0005227002   \n",
       "1     11:67585218      11:67352689  0000001695             1  0067585218   \n",
       "2     11:64242407      11:64009879  0000004672             0  0064242407   \n",
       "3     22:19963748      22:19951271  0000004680             1  0019963748   \n",
       "4      6:38682852       6:38650628  0000004746             1  0038682852   \n",
       "...           ...              ...         ...           ...         ...   \n",
       "4301  16:69320768      16:69354671  1965499910             1  0069320768   \n",
       "4302  19:12924492      19:13035306  1971301248             1  0012924492   \n",
       "4303  17:41968309      17:40120327  2050933471             1  0041968309   \n",
       "4304  22:31620659      22:32016645  2072505076             1  0031620659   \n",
       "4305  19:10812270      19:10922946  2072577342             1  0010812270   \n",
       "\n",
       "     merged_sort  sequence_change  \n",
       "0              0  [T:A, T:C, T:G]  \n",
       "1              0       [A:G, A:T]  \n",
       "2              0       [G:A, G:T]  \n",
       "3              0            [G:A]  \n",
       "4              0       [T:A, T:G]  \n",
       "...          ...              ...  \n",
       "4301           0       [A:G, A:T]  \n",
       "4302           0            [G:T]  \n",
       "4303           0            [C:T]  \n",
       "4304           0            [C:T]  \n",
       "4305           0            [C:T]  \n",
       "\n",
       "[4306 rows x 26 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_snps = df_model_mappings.loc[:, [\"genes\", \"dbsnp\"]].copy()\n",
    "df_model_snps[\"dbsnp\"] = df_model_mappings[\"dbsnp\"].apply(split_string)\n",
    "df_model_snps = df_model_snps.explode(\"dbsnp\")\n",
    "df_model_snps = df_model_snps.dropna().drop_duplicates()\n",
    "rs_ids = df_model_snps[\"dbsnp\"].unique()\n",
    "df_model_snps\n",
    "\n",
    "rename_mapping = {\n",
    "    key: key.lower()\n",
    "    for key in df_snps_all.columns\n",
    "    if key not in {\"error\", \"GLOBAL_SAMPLESIZE\"}\n",
    "}\n",
    "\n",
    "df_model_snps = df_model_snps.merge(\n",
    "    df_snps_all.loc[:, list(rename_mapping)].rename(rename_mapping, axis=1),\n",
    "    left_on=\"dbsnp\",\n",
    "    right_on=\"dbsnp\",\n",
    "    how=\"inner\",\n",
    ")\n",
    "for col in df_model_snps.columns:\n",
    "    df_model_snps[col] = df_model_snps[col].str.split(\",\")\n",
    "    # Switch to default seperator (currently semicolon)\n",
    "    df_model_snps[col] = df_model_snps[col].apply(lambda x: build_string(x))\n",
    "df_model_snps = df_model_snps.sort_values([\"snp_id_sort\"]).reset_index(drop=True)\n",
    "df_model_snps[\"sequence_change\"] = (\n",
    "    df_model_snps[\"spdi\"]\n",
    "    .dropna()\n",
    "    .apply(lambda value: [s.split(\":\", 2)[-1] for s in split_string(value)])\n",
    ")\n",
    "df_model_snps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d00ca838-c94d-414b-9c4b-54f82369065e",
   "metadata": {},
   "source": [
    "### Export data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "99b623b9-e30d-4b1e-843a-7d2f2f057ec1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>handle</th>\n",
       "      <th>spdi</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>...</th>\n",
       "      <th>ss</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "      <th>sequence_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HBB</td>\n",
       "      <td>rs334</td>\n",
       "      <td>334</td>\n",
       "      <td>protective;other;pathogenic;benign;likely-beni...</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...</td>\n",
       "      <td>NC_000011.10:5227001:T:A;NC_000011.10:5227001:...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>335;24811263;48419811;49850538;49850540;498505...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>11:5248232</td>\n",
       "      <td>0000000334</td>\n",
       "      <td>1</td>\n",
       "      <td>0005227002</td>\n",
       "      <td>0</td>\n",
       "      <td>[T:A, T:C, T:G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSTP1</td>\n",
       "      <td>rs1695</td>\n",
       "      <td>1695</td>\n",
       "      <td>benign;not-provided</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...</td>\n",
       "      <td>NC_000011.10:67585217:A:G;NC_000011.10:6758521...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>1720;5330;9110;1390210;1545478;3178003;5586573...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:67585218</td>\n",
       "      <td>11:67352689</td>\n",
       "      <td>0000001695</td>\n",
       "      <td>1</td>\n",
       "      <td>0067585218</td>\n",
       "      <td>0</td>\n",
       "      <td>[A:G, A:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FKBP2</td>\n",
       "      <td>rs4672</td>\n",
       "      <td>4672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...</td>\n",
       "      <td>NC_000011.10:64242406:G:A;NC_000011.10:6424240...</td>\n",
       "      <td>3_prime_UTR_variant;intron_variant;missense_va...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>1555019;4436250;20825221;24790172;39886556;752...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:64242407</td>\n",
       "      <td>11:64009879</td>\n",
       "      <td>0000004672</td>\n",
       "      <td>0</td>\n",
       "      <td>0064242407</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:A, G:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COMT</td>\n",
       "      <td>rs4680</td>\n",
       "      <td>4680</td>\n",
       "      <td>drug-response;benign</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...</td>\n",
       "      <td>NC_000022.11:19963747:G:A</td>\n",
       "      <td>2KB_upstream_variant;coding_sequence_variant;u...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>5363;10902;232568;458935;593353;869771;1058120...</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:19963748</td>\n",
       "      <td>22:19951271</td>\n",
       "      <td>0000004680</td>\n",
       "      <td>1</td>\n",
       "      <td>0019963748</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:A]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GLO1</td>\n",
       "      <td>rs4746</td>\n",
       "      <td>4746</td>\n",
       "      <td>uncertain-significance</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>6</td>\n",
       "      <td>ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...</td>\n",
       "      <td>NC_000006.12:38682851:T:A;NC_000006.12:3868285...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>...</td>\n",
       "      <td>3863270;20275576;24369377;28505662;44730940;48...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>6:38682852</td>\n",
       "      <td>6:38650628</td>\n",
       "      <td>0000004746</td>\n",
       "      <td>1</td>\n",
       "      <td>0038682852</td>\n",
       "      <td>0</td>\n",
       "      <td>[T:A, T:G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4301</th>\n",
       "      <td>VPS4A</td>\n",
       "      <td>rs1965499910</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000016.10:69320767:A:G;NC_000016.10:6932076...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>16:69320768</td>\n",
       "      <td>16:69354671</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>1</td>\n",
       "      <td>0069320768</td>\n",
       "      <td>0</td>\n",
       "      <td>[A:G, A:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4302</th>\n",
       "      <td>FARSA</td>\n",
       "      <td>rs1971301248</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000019.10:12924491:G:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:12924492</td>\n",
       "      <td>19:13035306</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>1</td>\n",
       "      <td>0012924492</td>\n",
       "      <td>0</td>\n",
       "      <td>[G:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4303</th>\n",
       "      <td>CNP</td>\n",
       "      <td>rs2050933471</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000017.11:41968308:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>17:41968309</td>\n",
       "      <td>17:40120327</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>1</td>\n",
       "      <td>0041968309</td>\n",
       "      <td>0</td>\n",
       "      <td>[C:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4304</th>\n",
       "      <td>PISD</td>\n",
       "      <td>rs2072505076</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>pathogenic-likely-pathogenic</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000022.11:31620658:C:T</td>\n",
       "      <td>coding_sequence_variant;intron_variant;missens...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:31620659</td>\n",
       "      <td>22:32016645</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>1</td>\n",
       "      <td>0031620659</td>\n",
       "      <td>0</td>\n",
       "      <td>[C:T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4305</th>\n",
       "      <td>DNM2</td>\n",
       "      <td>rs2072577342</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>likely-pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>TOPMED</td>\n",
       "      <td>NC_000019.10:10812269:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa</td>\n",
       "      <td>...</td>\n",
       "      <td>5068054504</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:10812270</td>\n",
       "      <td>19:10922946</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>1</td>\n",
       "      <td>0010812270</td>\n",
       "      <td>0</td>\n",
       "      <td>[C:T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4306 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes         dbsnp      snp_id  \\\n",
       "0       HBB         rs334         334   \n",
       "1     GSTP1        rs1695        1695   \n",
       "2     FKBP2        rs4672        4672   \n",
       "3      COMT        rs4680        4680   \n",
       "4      GLO1        rs4746        4746   \n",
       "...     ...           ...         ...   \n",
       "4301  VPS4A  rs1965499910  1965499910   \n",
       "4302  FARSA  rs1971301248  1971301248   \n",
       "4303    CNP  rs2050933471  2050933471   \n",
       "4304   PISD  rs2072505076  2072505076   \n",
       "4305   DNM2  rs2072577342  2072577342   \n",
       "\n",
       "                                  clinical_significance           acc chr  \\\n",
       "0     protective;other;pathogenic;benign;likely-beni...  NC_000011.10  11   \n",
       "1                                   benign;not-provided  NC_000011.10  11   \n",
       "2                                                   NaN  NC_000011.10  11   \n",
       "3                                  drug-response;benign  NC_000022.11  22   \n",
       "4                                uncertain-significance  NC_000006.12   6   \n",
       "...                                                 ...           ...  ..   \n",
       "4301                                         pathogenic  NC_000016.10  16   \n",
       "4302                                         pathogenic  NC_000019.10  19   \n",
       "4303                                         pathogenic  NC_000017.11  17   \n",
       "4304                       pathogenic-likely-pathogenic  NC_000022.11  22   \n",
       "4305                                  likely-pathogenic  NC_000019.10  19   \n",
       "\n",
       "                                                 handle  \\\n",
       "0     JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...   \n",
       "1     CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...   \n",
       "2     JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...   \n",
       "3     GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...   \n",
       "4     ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...   \n",
       "...                                                 ...   \n",
       "4301                                               None   \n",
       "4302                                               None   \n",
       "4303                                               None   \n",
       "4304                                               None   \n",
       "4305                                             TOPMED   \n",
       "\n",
       "                                                   spdi  \\\n",
       "0     NC_000011.10:5227001:T:A;NC_000011.10:5227001:...   \n",
       "1     NC_000011.10:67585217:A:G;NC_000011.10:6758521...   \n",
       "2     NC_000011.10:64242406:G:A;NC_000011.10:6424240...   \n",
       "3                             NC_000022.11:19963747:G:A   \n",
       "4     NC_000006.12:38682851:T:A;NC_000006.12:3868285...   \n",
       "...                                                 ...   \n",
       "4301  NC_000016.10:69320767:A:G;NC_000016.10:6932076...   \n",
       "4302                          NC_000019.10:12924491:G:T   \n",
       "4303                          NC_000017.11:41968308:C:T   \n",
       "4304                          NC_000022.11:31620658:C:T   \n",
       "4305                          NC_000019.10:10812269:C:T   \n",
       "\n",
       "                                              fxn_class  \\\n",
       "0              missense_variant;coding_sequence_variant   \n",
       "1              missense_variant;coding_sequence_variant   \n",
       "2     3_prime_UTR_variant;intron_variant;missense_va...   \n",
       "3     2KB_upstream_variant;coding_sequence_variant;u...   \n",
       "4              coding_sequence_variant;missense_variant   \n",
       "...                                                 ...   \n",
       "4301           missense_variant;coding_sequence_variant   \n",
       "4302           coding_sequence_variant;missense_variant   \n",
       "4303           coding_sequence_variant;missense_variant   \n",
       "4304  coding_sequence_variant;intron_variant;missens...   \n",
       "4305           coding_sequence_variant;missense_variant   \n",
       "\n",
       "                            validated  ...  \\\n",
       "0     by-frequency;by-alfa;by-cluster  ...   \n",
       "1     by-frequency;by-alfa;by-cluster  ...   \n",
       "2     by-frequency;by-alfa;by-cluster  ...   \n",
       "3     by-frequency;by-alfa;by-cluster  ...   \n",
       "4     by-frequency;by-alfa;by-cluster  ...   \n",
       "...                               ...  ...   \n",
       "4301                              NaN  ...   \n",
       "4302                              NaN  ...   \n",
       "4303                              NaN  ...   \n",
       "4304                              NaN  ...   \n",
       "4305             by-frequency;by-alfa  ...   \n",
       "\n",
       "                                                     ss allele snp_class  \\\n",
       "0     335;24811263;48419811;49850538;49850540;498505...      N       snv   \n",
       "1     1720;5330;9110;1390210;1545478;3178003;5586573...      D       snv   \n",
       "2     1555019;4436250;20825221;24790172;39886556;752...      D       snv   \n",
       "3     5363;10902;232568;458935;593353;869771;1058120...      R       snv   \n",
       "4     3863270;20275576;24369377;28505662;44730940;48...      D       snv   \n",
       "...                                                 ...    ...       ...   \n",
       "4301                                                  0      D       snv   \n",
       "4302                                                  0      K       snv   \n",
       "4303                                                  0      Y       snv   \n",
       "4304                                                  0      Y       snv   \n",
       "4305                                         5068054504      Y       snv   \n",
       "\n",
       "           chrpos chrpos_prev_assm snp_id_sort clinical_sort chrpos_sort  \\\n",
       "0      11:5227002       11:5248232  0000000334             1  0005227002   \n",
       "1     11:67585218      11:67352689  0000001695             1  0067585218   \n",
       "2     11:64242407      11:64009879  0000004672             0  0064242407   \n",
       "3     22:19963748      22:19951271  0000004680             1  0019963748   \n",
       "4      6:38682852       6:38650628  0000004746             1  0038682852   \n",
       "...           ...              ...         ...           ...         ...   \n",
       "4301  16:69320768      16:69354671  1965499910             1  0069320768   \n",
       "4302  19:12924492      19:13035306  1971301248             1  0012924492   \n",
       "4303  17:41968309      17:40120327  2050933471             1  0041968309   \n",
       "4304  22:31620659      22:32016645  2072505076             1  0031620659   \n",
       "4305  19:10812270      19:10922946  2072577342             1  0010812270   \n",
       "\n",
       "     merged_sort  sequence_change  \n",
       "0              0  [T:A, T:C, T:G]  \n",
       "1              0       [A:G, A:T]  \n",
       "2              0       [G:A, G:T]  \n",
       "3              0            [G:A]  \n",
       "4              0       [T:A, T:G]  \n",
       "...          ...              ...  \n",
       "4301           0       [A:G, A:T]  \n",
       "4302           0            [G:T]  \n",
       "4303           0            [C:T]  \n",
       "4304           0            [C:T]  \n",
       "4305           0            [C:T]  \n",
       "\n",
       "[4306 rows x 26 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if overwrite:\n",
    "    df_model_snps.to_csv(f\"{ROOT_PATH}{DATABASE_PATH}/snps_{GEM_NAME}.tsv\", sep=\"\\t\")\n",
    "else:\n",
    "    df_model_snps.to_csv(f\"{ROOT_PATH}{INTERIM_PATH}/snps_{GEM_NAME}.tsv\", sep=\"\\t\")\n",
    "df_model_snps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d10bf208-340f-4827-b946-c91b8dba024b",
   "metadata": {},
   "source": [
    "### Visualize genetic landscape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82148420-c8cc-4de5-955c-fa249c677656",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import matplotlib.pyplot as plt\n",
    "# import matplotlib as mpl\n",
    "# import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef439002-06b2-44e1-9eda-e486735c7fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_snps_per_gene = (\n",
    "#     df_model_snps.groupby(\"genes\")[\"dbsnp\"]\n",
    "#     .agg(lambda x: x.nunique())\n",
    "#     .sort_values(ascending=False)\n",
    "# )\n",
    "# df_snp_fxn_class = (\n",
    "#     df_model_snps[\"fxn_class\"].apply(split_string).explode().value_counts()\n",
    "# )\n",
    "# df_snp_clinical_significance = (\n",
    "#     df_model_snps[\"clinical_significance\"].apply(split_string).explode().value_counts()\n",
    "# )\n",
    "# df_snp_allele = df_model_snps[\"allele\"].apply(split_string).explode().value_counts()\n",
    "\n",
    "# fig, axes = plt.subplots(3, 1, figsize=(5, 9))\n",
    "# (ax_allele, ax_fxn_class, ax_clinical_significance) = axes.flatten()\n",
    "# sns.despine(fig)\n",
    "# cmin = 0\n",
    "# cmax = 1\n",
    "# colors = mpl.colormaps[\"tab20\"](np.linspace(cmin, cmax, len(df_snp_allele.index)))\n",
    "\n",
    "# ax_allele.barh(\n",
    "#     df_snp_allele.index,\n",
    "#     df_snp_allele.values,\n",
    "#     color=colors,\n",
    "# )\n",
    "# ax_fxn_class.barh(\n",
    "#     df_snp_fxn_class.index,\n",
    "#     df_snp_fxn_class.values,\n",
    "#     color=colors,\n",
    "# )\n",
    "# ax_clinical_significance.barh(\n",
    "#     df_snp_clinical_significance.index,\n",
    "#     df_snp_clinical_significance.values,\n",
    "#     color=colors,\n",
    "# )\n",
    "# fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95fe9ed5-749f-4423-8b55-69a4c7f66c6b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
