{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "13bddfd6-19f9-44de-9062-5462c031dfb5",
   "metadata": {},
   "source": [
    "# Extract data from dbSNP\n",
    "\n",
    "Utilize the [Bio.Entrez package](https://biopython.org/docs/latest/api/Bio.Entrez.html) to get data from [dbSNP](https://www.ncbi.nlm.nih.gov/snp/).\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "49fed61a-e9eb-47f9-b97e-8edeff408452",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from xml.etree import ElementTree\n",
    "\n",
    "from rbc_gem_utils import (\n",
    "    ROOT_PATH,\n",
    "    DATABASE_PATH,\n",
    "    RESULTS_PATH,\n",
    "    GEM_NAME,\n",
    "    read_rbc_model,\n",
    "    build_string,\n",
    "    split_string,\n",
    "    get_annotation_df,\n",
    "    show_versions,\n",
    "    explode_column,\n",
    ")\n",
    "from rbc_gem_utils.util import has_value_type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6da6a907-3339-4f93-868a-01622f9e6de6",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13d4753e-2d6d-44d9-bad0-7a477220b3a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14d04d050</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1955</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2777</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>649</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>74</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14d04d050>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_rbc_model(filetype=\"xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a1a26d35-de9c-4390-aca2-a6c8481eac4a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes: 649\n",
      "dbsnp: 4108\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RPE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPIA</td>\n",
       "      <td>rs121918591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs145813597;rs1042079;rs930337;rs149975952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AKR7A2</td>\n",
       "      <td>rs6670759;rs1043657;rs859208;rs859210;rs223120...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRM</td>\n",
       "      <td>rs1049932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>644</th>\n",
       "      <td>SLC25A31</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>645</th>\n",
       "      <td>ENTPD1</td>\n",
       "      <td>rs1566229309;rs3793744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>SLC22A7</td>\n",
       "      <td>rs70953684;rs36040909;rs70953693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>SLC22A4</td>\n",
       "      <td>rs272893;rs4646201;rs1050152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>rs11541159;rs28915383;rs74315453;rs28940572;rs...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>649 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genes                                              dbsnp\n",
       "0         RPE                                                NaN\n",
       "1        RPIA                                        rs121918591\n",
       "2        SORD         rs145813597;rs1042079;rs930337;rs149975952\n",
       "3      AKR7A2  rs6670759;rs1043657;rs859208;rs859210;rs223120...\n",
       "4         SRM                                          rs1049932\n",
       "..        ...                                                ...\n",
       "644  SLC25A31                                                NaN\n",
       "645    ENTPD1                             rs1566229309;rs3793744\n",
       "646   SLC22A7                   rs70953684;rs36040909;rs70953693\n",
       "647   SLC22A4                       rs272893;rs4646201;rs1050152\n",
       "648    A4GALT  rs11541159;rs28915383;rs74315453;rs28940572;rs...\n",
       "\n",
       "[649 rows x 2 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_type = \"genes\"\n",
    "df_model_mappings = get_annotation_df(\n",
    "    getattr(model, annotation_type), [\"dbsnp\"]\n",
    ").rename({\"id\": annotation_type}, axis=1)\n",
    "for col in df_model_mappings.columns:\n",
    "    df = (\n",
    "        df_model_mappings[col]\n",
    "        .apply(lambda x: split_string(x))\n",
    "        .explode(col)\n",
    "        .drop_duplicates()\n",
    "    )\n",
    "    print(f\"{df.name}: {df.nunique()}\")\n",
    "\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21905f9a-fb31-466a-84c5-cf9ee6681287",
   "metadata": {},
   "source": [
    "### Get IDs for query from model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ae646a47-2606-4d32-bb97-bafb1413a2ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>dbsnp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RPIA</td>\n",
       "      <td>rs121918591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs145813597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs1042079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs930337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SORD</td>\n",
       "      <td>rs149975952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>rs28915383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>rs74315453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>rs28940572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>rs397514502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>rs28940571</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4108 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes        dbsnp\n",
       "1      RPIA  rs121918591\n",
       "2      SORD  rs145813597\n",
       "2      SORD    rs1042079\n",
       "2      SORD     rs930337\n",
       "2      SORD  rs149975952\n",
       "..      ...          ...\n",
       "648  A4GALT   rs28915383\n",
       "648  A4GALT   rs74315453\n",
       "648  A4GALT   rs28940572\n",
       "648  A4GALT  rs397514502\n",
       "648  A4GALT   rs28940571\n",
       "\n",
       "[4108 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_model_snps = df_model_mappings.loc[:, [\"genes\", \"dbsnp\"]].copy()\n",
    "df_model_snps[\"dbsnp\"] = df_model_mappings[\"dbsnp\"].apply(split_string)\n",
    "df_model_snps = df_model_snps.explode(\"dbsnp\")\n",
    "df_model_snps = df_model_snps.dropna().drop_duplicates()\n",
    "rs_ids = df_model_snps[\"dbsnp\"].unique()\n",
    "df_model_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cdc8c330-2057-4c12-820d-4aa9cdf72b9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_results_dbSNP(email, rs_ids):\n",
    "    Entrez.email = email\n",
    "    search = Entrez.efetch(db=\"snp\", retmode=\"text\", id=rs_ids)\n",
    "    all_lines = search.readlines()\n",
    "    search.close()\n",
    "    return all_lines"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e996f2bd-5293-4f6d-99b5-345a172f0840",
   "metadata": {},
   "source": [
    "### Get results from SNP as a DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9ef42f63-2692-4dfa-a53c-4a2cd679c3e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "overwrite = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e253df0-8aee-48f6-87a2-6f9198e5ad18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching results for batch 1  (500/4108)\n",
      "Fetching results for batch 2  (1000/4108)\n",
      "Fetching results for batch 3  (1500/4108)\n",
      "Fetching results for batch 4  (2000/4108)\n",
      "Fetching results for batch 5  (2500/4108)\n",
      "Fetching results for batch 6  (3000/4108)\n",
      "Fetching results for batch 7  (3500/4108)\n",
      "Fetching results for batch 8  (4000/4108)\n",
      "Fetching results for batch 9  (4108/4108)\n",
      "\n",
      "Number of failed IDs: 2\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP_ID</th>\n",
       "      <th>GLOBAL_SAMPLESIZE</th>\n",
       "      <th>CLINICAL_SIGNIFICANCE</th>\n",
       "      <th>ACC</th>\n",
       "      <th>CHR</th>\n",
       "      <th>HANDLE</th>\n",
       "      <th>SPDI</th>\n",
       "      <th>FXN_CLASS</th>\n",
       "      <th>VALIDATED</th>\n",
       "      <th>DOCSUM</th>\n",
       "      <th>...</th>\n",
       "      <th>SS</th>\n",
       "      <th>ALLELE</th>\n",
       "      <th>SNP_CLASS</th>\n",
       "      <th>CHRPOS</th>\n",
       "      <th>CHRPOS_PREV_ASSM</th>\n",
       "      <th>SNP_ID_SORT</th>\n",
       "      <th>CLINICAL_SORT</th>\n",
       "      <th>CHRPOS_SORT</th>\n",
       "      <th>MERGED_SORT</th>\n",
       "      <th>error</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dbsnp</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>rs121918591</th>\n",
       "      <td>121918591</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000002.12</td>\n",
       "      <td>2</td>\n",
       "      <td>TOPMED,ILLUMINA,PAGE_CC,OMIM-CURATED-RECORDS,G...</td>\n",
       "      <td>NC_000002.12:88729278:C:T</td>\n",
       "      <td>coding_sequence_variant,missense_variant</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000002.12:g.88729279C&gt;T,NC_000002.11:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>275518424,1958438289,2732810607,3021993357,365...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>2:88729279</td>\n",
       "      <td>2:89028797</td>\n",
       "      <td>0121918591</td>\n",
       "      <td>1</td>\n",
       "      <td>0088729279</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs145813597</th>\n",
       "      <td>145813597</td>\n",
       "      <td>0</td>\n",
       "      <td>pathogenic-likely-pathogenic</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>AFFY,TOPMED,EVA_EXAC,HUGCELL_USP,ILLUMINA,1000...</td>\n",
       "      <td>NC_000015.10:45065302:C:A,NC_000015.10:4506530...</td>\n",
       "      <td>non_coding_transcript_variant,missense_variant...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45065303C&gt;A,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>342402496,464140601,491493966,491700043,991737...</td>\n",
       "      <td>H</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45065303</td>\n",
       "      <td>15:45357501</td>\n",
       "      <td>0145813597</td>\n",
       "      <td>1</td>\n",
       "      <td>0045065303</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs1042079</th>\n",
       "      <td>1042079</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>EVA-GONL,CNG,BUSHMAN,SSMP,1000GENOMES,EXOME_CH...</td>\n",
       "      <td>NC_000015.10:45068981:A:G,NC_000015.10:4506898...</td>\n",
       "      <td>coding_sequence_variant,non_coding_transcript_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45068982A&gt;G,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1509008,4403349,76874733,95213358,156480828,15...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45068982</td>\n",
       "      <td>15:45361180</td>\n",
       "      <td>0001042079</td>\n",
       "      <td>1</td>\n",
       "      <td>0045068982</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs930337</th>\n",
       "      <td>930337</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>ENSEMBL,CNG,SYSTEMSBIOZJU,SSMP,1000GENOMES,EXO...</td>\n",
       "      <td>NC_000015.10:45072335:A:C</td>\n",
       "      <td>coding_sequence_variant,non_coding_transcript_...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45072336A&gt;C,NC_000015.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>1368774,2558685,3214527,5252928,21245444,24802...</td>\n",
       "      <td>M</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45072336</td>\n",
       "      <td>15:45364534</td>\n",
       "      <td>0000930337</td>\n",
       "      <td>1</td>\n",
       "      <td>0045072336</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs149975952</th>\n",
       "      <td>149975952</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000015.10</td>\n",
       "      <td>15</td>\n",
       "      <td>ILLUMINA,WEILL_CORNELL_DGM,1000GENOMES,EXOME_C...</td>\n",
       "      <td>NC_000015.10:45073419:G:A,NC_000015.10:4507341...</td>\n",
       "      <td>coding_sequence_variant,missense_variant,non_c...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000015.10:g.45073420G&gt;A,NC_000015.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>342402507,489048604,491493972,1353273008,16918...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>15:45073420</td>\n",
       "      <td>15:45365618</td>\n",
       "      <td>0149975952</td>\n",
       "      <td>1</td>\n",
       "      <td>0045073420</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs28915383</th>\n",
       "      <td>28915383</td>\n",
       "      <td>0</td>\n",
       "      <td>benign</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>SANFORD_IMAGENETICS,HUGCELL_USP,PGA-UW-FHCRC,E...</td>\n",
       "      <td>NC_000022.11:42693463:T:C</td>\n",
       "      <td>genic_downstream_transcript_variant,coding_seq...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.42693464T&gt;C,NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>35529216,160596039,173525433,204101017,2287004...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:42693464</td>\n",
       "      <td>22:43089470</td>\n",
       "      <td>0028915383</td>\n",
       "      <td>1</td>\n",
       "      <td>0042693464</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs74315453</th>\n",
       "      <td>74315453</td>\n",
       "      <td>0</td>\n",
       "      <td>affects</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>ILLUMINA,GNOMAD,OMICIA,PAGE_CC,EVA_EXAC,OMIM-C...</td>\n",
       "      <td>NC_000022.11:42693403:A:T</td>\n",
       "      <td>genic_downstream_transcript_variant,coding_seq...</td>\n",
       "      <td>by-frequency,by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.42693404A&gt;T,NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>179320298,254223349,1694384290,1959984485,2745...</td>\n",
       "      <td>W</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:42693404</td>\n",
       "      <td>22:43089410</td>\n",
       "      <td>0074315453</td>\n",
       "      <td>1</td>\n",
       "      <td>0042693404</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs28940572</th>\n",
       "      <td>28940572</td>\n",
       "      <td>0</td>\n",
       "      <td>affects,likely-pathogenic</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>ILLUMINA,GNOMAD,EVA_EXAC,OMICIA,OMIM-CURATED-R...</td>\n",
       "      <td>NC_000022.11:42693391:C:T</td>\n",
       "      <td>genic_downstream_transcript_variant,coding_seq...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.42693392C&gt;T,NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>169697184,254223366,1694384286,2745199187,3725...</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:42693392</td>\n",
       "      <td>22:43089398</td>\n",
       "      <td>0028940572</td>\n",
       "      <td>1</td>\n",
       "      <td>0042693392</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs397514502</th>\n",
       "      <td>397514502</td>\n",
       "      <td>0</td>\n",
       "      <td>affects,likely-pathogenic</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>OMIM-CURATED-RECORDS,ICHGLYCOCONJ,ILLUMINA,PAG...</td>\n",
       "      <td>NC_000022.11:42693320:G:C</td>\n",
       "      <td>coding_sequence_variant,missense_variant,genic...</td>\n",
       "      <td>by-frequency,by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.42693321G&gt;C,NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>836176862,1959984483,2137510514,3022191358,365...</td>\n",
       "      <td>S</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:42693321</td>\n",
       "      <td>22:43089327</td>\n",
       "      <td>0397514502</td>\n",
       "      <td>1</td>\n",
       "      <td>0042693321</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rs28940571</th>\n",
       "      <td>28940571</td>\n",
       "      <td>0</td>\n",
       "      <td>affects</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>ILLUMINA,GNOMAD,EVA,1000G_HIGH_COVERAGE,TOPMED...</td>\n",
       "      <td>NC_000022.11:42693199:G:A,NC_000022.11:4269319...</td>\n",
       "      <td>genic_downstream_transcript_variant,coding_seq...</td>\n",
       "      <td>by-frequency,by-alfa,by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.42693200G&gt;A,NC_000022.11:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>169696744,254223354,466206869,713629464,136735...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:42693200</td>\n",
       "      <td>22:43089206</td>\n",
       "      <td>0028940571</td>\n",
       "      <td>1</td>\n",
       "      <td>0042693200</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4106 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                SNP_ID GLOBAL_SAMPLESIZE         CLINICAL_SIGNIFICANCE  \\\n",
       "dbsnp                                                                    \n",
       "rs121918591  121918591                 0                    pathogenic   \n",
       "rs145813597  145813597                 0  pathogenic-likely-pathogenic   \n",
       "rs1042079      1042079                 0                        benign   \n",
       "rs930337        930337                 0                        benign   \n",
       "rs149975952  149975952                 0                        benign   \n",
       "...                ...               ...                           ...   \n",
       "rs28915383    28915383                 0                        benign   \n",
       "rs74315453    74315453                 0                       affects   \n",
       "rs28940572    28940572                 0     affects,likely-pathogenic   \n",
       "rs397514502  397514502                 0     affects,likely-pathogenic   \n",
       "rs28940571    28940571                 0                       affects   \n",
       "\n",
       "                      ACC CHR  \\\n",
       "dbsnp                           \n",
       "rs121918591  NC_000002.12   2   \n",
       "rs145813597  NC_000015.10  15   \n",
       "rs1042079    NC_000015.10  15   \n",
       "rs930337     NC_000015.10  15   \n",
       "rs149975952  NC_000015.10  15   \n",
       "...                   ...  ..   \n",
       "rs28915383   NC_000022.11  22   \n",
       "rs74315453   NC_000022.11  22   \n",
       "rs28940572   NC_000022.11  22   \n",
       "rs397514502  NC_000022.11  22   \n",
       "rs28940571   NC_000022.11  22   \n",
       "\n",
       "                                                        HANDLE  \\\n",
       "dbsnp                                                            \n",
       "rs121918591  TOPMED,ILLUMINA,PAGE_CC,OMIM-CURATED-RECORDS,G...   \n",
       "rs145813597  AFFY,TOPMED,EVA_EXAC,HUGCELL_USP,ILLUMINA,1000...   \n",
       "rs1042079    EVA-GONL,CNG,BUSHMAN,SSMP,1000GENOMES,EXOME_CH...   \n",
       "rs930337     ENSEMBL,CNG,SYSTEMSBIOZJU,SSMP,1000GENOMES,EXO...   \n",
       "rs149975952  ILLUMINA,WEILL_CORNELL_DGM,1000GENOMES,EXOME_C...   \n",
       "...                                                        ...   \n",
       "rs28915383   SANFORD_IMAGENETICS,HUGCELL_USP,PGA-UW-FHCRC,E...   \n",
       "rs74315453   ILLUMINA,GNOMAD,OMICIA,PAGE_CC,EVA_EXAC,OMIM-C...   \n",
       "rs28940572   ILLUMINA,GNOMAD,EVA_EXAC,OMICIA,OMIM-CURATED-R...   \n",
       "rs397514502  OMIM-CURATED-RECORDS,ICHGLYCOCONJ,ILLUMINA,PAG...   \n",
       "rs28940571   ILLUMINA,GNOMAD,EVA,1000G_HIGH_COVERAGE,TOPMED...   \n",
       "\n",
       "                                                          SPDI  \\\n",
       "dbsnp                                                            \n",
       "rs121918591                          NC_000002.12:88729278:C:T   \n",
       "rs145813597  NC_000015.10:45065302:C:A,NC_000015.10:4506530...   \n",
       "rs1042079    NC_000015.10:45068981:A:G,NC_000015.10:4506898...   \n",
       "rs930337                             NC_000015.10:45072335:A:C   \n",
       "rs149975952  NC_000015.10:45073419:G:A,NC_000015.10:4507341...   \n",
       "...                                                        ...   \n",
       "rs28915383                           NC_000022.11:42693463:T:C   \n",
       "rs74315453                           NC_000022.11:42693403:A:T   \n",
       "rs28940572                           NC_000022.11:42693391:C:T   \n",
       "rs397514502                          NC_000022.11:42693320:G:C   \n",
       "rs28940571   NC_000022.11:42693199:G:A,NC_000022.11:4269319...   \n",
       "\n",
       "                                                     FXN_CLASS  \\\n",
       "dbsnp                                                            \n",
       "rs121918591           coding_sequence_variant,missense_variant   \n",
       "rs145813597  non_coding_transcript_variant,missense_variant...   \n",
       "rs1042079    coding_sequence_variant,non_coding_transcript_...   \n",
       "rs930337     coding_sequence_variant,non_coding_transcript_...   \n",
       "rs149975952  coding_sequence_variant,missense_variant,non_c...   \n",
       "...                                                        ...   \n",
       "rs28915383   genic_downstream_transcript_variant,coding_seq...   \n",
       "rs74315453   genic_downstream_transcript_variant,coding_seq...   \n",
       "rs28940572   genic_downstream_transcript_variant,coding_seq...   \n",
       "rs397514502  coding_sequence_variant,missense_variant,genic...   \n",
       "rs28940571   genic_downstream_transcript_variant,coding_seq...   \n",
       "\n",
       "                                   VALIDATED  \\\n",
       "dbsnp                                          \n",
       "rs121918591  by-frequency,by-alfa,by-cluster   \n",
       "rs145813597  by-frequency,by-alfa,by-cluster   \n",
       "rs1042079    by-frequency,by-alfa,by-cluster   \n",
       "rs930337     by-frequency,by-alfa,by-cluster   \n",
       "rs149975952  by-frequency,by-alfa,by-cluster   \n",
       "...                                      ...   \n",
       "rs28915383   by-frequency,by-alfa,by-cluster   \n",
       "rs74315453           by-frequency,by-cluster   \n",
       "rs28940572   by-frequency,by-alfa,by-cluster   \n",
       "rs397514502          by-frequency,by-cluster   \n",
       "rs28940571   by-frequency,by-alfa,by-cluster   \n",
       "\n",
       "                                                        DOCSUM  ...  \\\n",
       "dbsnp                                                           ...   \n",
       "rs121918591  HGVS=NC_000002.12:g.88729279C>T,NC_000002.11:g...  ...   \n",
       "rs145813597  HGVS=NC_000015.10:g.45065303C>A,NC_000015.10:g...  ...   \n",
       "rs1042079    HGVS=NC_000015.10:g.45068982A>G,NC_000015.10:g...  ...   \n",
       "rs930337     HGVS=NC_000015.10:g.45072336A>C,NC_000015.9:g....  ...   \n",
       "rs149975952  HGVS=NC_000015.10:g.45073420G>A,NC_000015.10:g...  ...   \n",
       "...                                                        ...  ...   \n",
       "rs28915383   HGVS=NC_000022.11:g.42693464T>C,NC_000022.10:g...  ...   \n",
       "rs74315453   HGVS=NC_000022.11:g.42693404A>T,NC_000022.10:g...  ...   \n",
       "rs28940572   HGVS=NC_000022.11:g.42693392C>T,NC_000022.10:g...  ...   \n",
       "rs397514502  HGVS=NC_000022.11:g.42693321G>C,NC_000022.10:g...  ...   \n",
       "rs28940571   HGVS=NC_000022.11:g.42693200G>A,NC_000022.11:g...  ...   \n",
       "\n",
       "                                                            SS ALLELE  \\\n",
       "dbsnp                                                                   \n",
       "rs121918591  275518424,1958438289,2732810607,3021993357,365...      Y   \n",
       "rs145813597  342402496,464140601,491493966,491700043,991737...      H   \n",
       "rs1042079    1509008,4403349,76874733,95213358,156480828,15...      D   \n",
       "rs930337     1368774,2558685,3214527,5252928,21245444,24802...      M   \n",
       "rs149975952  342402507,489048604,491493972,1353273008,16918...      D   \n",
       "...                                                        ...    ...   \n",
       "rs28915383   35529216,160596039,173525433,204101017,2287004...      Y   \n",
       "rs74315453   179320298,254223349,1694384290,1959984485,2745...      W   \n",
       "rs28940572   169697184,254223366,1694384286,2745199187,3725...      Y   \n",
       "rs397514502  836176862,1959984483,2137510514,3022191358,365...      S   \n",
       "rs28940571   169696744,254223354,466206869,713629464,136735...      D   \n",
       "\n",
       "            SNP_CLASS       CHRPOS CHRPOS_PREV_ASSM SNP_ID_SORT CLINICAL_SORT  \\\n",
       "dbsnp                                                                           \n",
       "rs121918591       snv   2:88729279       2:89028797  0121918591             1   \n",
       "rs145813597       snv  15:45065303      15:45357501  0145813597             1   \n",
       "rs1042079         snv  15:45068982      15:45361180  0001042079             1   \n",
       "rs930337          snv  15:45072336      15:45364534  0000930337             1   \n",
       "rs149975952       snv  15:45073420      15:45365618  0149975952             1   \n",
       "...               ...          ...              ...         ...           ...   \n",
       "rs28915383        snv  22:42693464      22:43089470  0028915383             1   \n",
       "rs74315453        snv  22:42693404      22:43089410  0074315453             1   \n",
       "rs28940572        snv  22:42693392      22:43089398  0028940572             1   \n",
       "rs397514502       snv  22:42693321      22:43089327  0397514502             1   \n",
       "rs28940571        snv  22:42693200      22:43089206  0028940571             1   \n",
       "\n",
       "            CHRPOS_SORT MERGED_SORT error  \n",
       "dbsnp                                      \n",
       "rs121918591  0088729279           0   NaN  \n",
       "rs145813597  0045065303           0   NaN  \n",
       "rs1042079    0045068982           0   NaN  \n",
       "rs930337     0045072336           0   NaN  \n",
       "rs149975952  0045073420           0   NaN  \n",
       "...                 ...         ...   ...  \n",
       "rs28915383   0042693464           0   NaN  \n",
       "rs74315453   0042693404           0   NaN  \n",
       "rs28940572   0042693392           0   NaN  \n",
       "rs397514502  0042693321           0   NaN  \n",
       "rs28940571   0042693200           0   NaN  \n",
       "\n",
       "[4106 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "email = \"zhaiman@ucsd.edu\"\n",
    "data = defaultdict(dict)\n",
    "batch_size = 500\n",
    "for batch_idx, batch in enumerate(np.arange(0, len(rs_ids), batch_size), start=0):\n",
    "    query_ids = rs_ids[batch : batch + batch_size]\n",
    "    print(\n",
    "        f\"Fetching results for batch {batch_idx + 1}  ({batch + len(query_ids)}/{len(rs_ids)})\"\n",
    "    )\n",
    "    all_lines = fetch_results_dbSNP(email=email, rs_ids=query_ids)\n",
    "    for entry_idx, xml_str in enumerate(all_lines, start=batch_idx * batch_size):\n",
    "        entry = ElementTree.fromstring(xml_str)\n",
    "        data[rs_ids[entry_idx]].update(\n",
    "            {e.tag: e.text for e in entry if has_value_type(e)}\n",
    "        )\n",
    "\n",
    "# Make into DataFrame\n",
    "df_snps_all = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "failed_ids = list(df_snps_all[\"error\"].dropna().index)\n",
    "print(f\"\\nNumber of failed IDs: {len(failed_ids)}\")\n",
    "df_snps_all.index.name = \"dbsnp\"\n",
    "df_snps_all = df_snps_all[~df_snps_all[\"error\"].notna()]\n",
    "df_snps_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "901fb1a6-0667-4b1b-a90a-657395466b5b",
   "metadata": {},
   "source": [
    "### Parse and reformat results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d76f9758-4964-4910-925c-6c9163095ef0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbsnp</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>handle</th>\n",
       "      <th>spdi</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>docsum</th>\n",
       "      <th>...</th>\n",
       "      <th>ss</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "      <th>sequence_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs334</td>\n",
       "      <td>334</td>\n",
       "      <td>protective;other;pathogenic;benign;likely-beni...</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...</td>\n",
       "      <td>NC_000011.10:5227001:T:A;NC_000011.10:5227001:...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.5227002T&gt;A;NC_000011.10:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>335;24811263;48419811;49850538;49850540;498505...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>11:5248232</td>\n",
       "      <td>0000000334</td>\n",
       "      <td>1</td>\n",
       "      <td>0005227002</td>\n",
       "      <td>0</td>\n",
       "      <td>T:A;T:C;T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1695</td>\n",
       "      <td>1695</td>\n",
       "      <td>benign;not-provided</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...</td>\n",
       "      <td>NC_000011.10:67585217:A:G;NC_000011.10:6758521...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.67585218A&gt;G;NC_000011.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1720;5330;9110;1390210;1545478;3178003;5586573...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:67585218</td>\n",
       "      <td>11:67352689</td>\n",
       "      <td>0000001695</td>\n",
       "      <td>1</td>\n",
       "      <td>0067585218</td>\n",
       "      <td>0</td>\n",
       "      <td>A:G;A:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4672</td>\n",
       "      <td>4672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...</td>\n",
       "      <td>NC_000011.10:64242406:G:A;NC_000011.10:6424240...</td>\n",
       "      <td>3_prime_UTR_variant;intron_variant;missense_va...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.64242407G&gt;A;NC_000011.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1555019;4436250;20825221;24790172;39886556;752...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:64242407</td>\n",
       "      <td>11:64009879</td>\n",
       "      <td>0000004672</td>\n",
       "      <td>0</td>\n",
       "      <td>0064242407</td>\n",
       "      <td>0</td>\n",
       "      <td>G:A;G:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4680</td>\n",
       "      <td>4680</td>\n",
       "      <td>drug-response;benign</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...</td>\n",
       "      <td>NC_000022.11:19963747:G:A</td>\n",
       "      <td>2KB_upstream_variant;coding_sequence_variant;u...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.19963748G&gt;A;NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>5363;10902;232568;458935;593353;869771;1058120...</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:19963748</td>\n",
       "      <td>22:19951271</td>\n",
       "      <td>0000004680</td>\n",
       "      <td>1</td>\n",
       "      <td>0019963748</td>\n",
       "      <td>0</td>\n",
       "      <td>G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4746</td>\n",
       "      <td>4746</td>\n",
       "      <td>uncertain-significance</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>6</td>\n",
       "      <td>ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...</td>\n",
       "      <td>NC_000006.12:38682851:T:A;NC_000006.12:3868285...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000006.12:g.38682852T&gt;A;NC_000006.12:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>3863270;20275576;24369377;28505662;44730940;48...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>6:38682852</td>\n",
       "      <td>6:38650628</td>\n",
       "      <td>0000004746</td>\n",
       "      <td>1</td>\n",
       "      <td>0038682852</td>\n",
       "      <td>0</td>\n",
       "      <td>T:A;T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4099</th>\n",
       "      <td>rs1965499910</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000016.10:69320767:A:G;NC_000016.10:6932076...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000016.10:g.69320768A&gt;G;NC_000016.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>16:69320768</td>\n",
       "      <td>16:69354671</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>1</td>\n",
       "      <td>0069320768</td>\n",
       "      <td>0</td>\n",
       "      <td>A:G;A:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4100</th>\n",
       "      <td>rs1971301248</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000019.10:12924491:G:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000019.10:g.12924492G&gt;T;NC_000019.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:12924492</td>\n",
       "      <td>19:13035306</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>1</td>\n",
       "      <td>0012924492</td>\n",
       "      <td>0</td>\n",
       "      <td>G:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4101</th>\n",
       "      <td>rs2050933471</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000017.11:41968308:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000017.11:g.41968309C&gt;T;NC_000017.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>17:41968309</td>\n",
       "      <td>17:40120327</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>1</td>\n",
       "      <td>0041968309</td>\n",
       "      <td>0</td>\n",
       "      <td>C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4102</th>\n",
       "      <td>rs2072505076</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>pathogenic-likely-pathogenic</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000022.11:31620658:C:T</td>\n",
       "      <td>coding_sequence_variant;intron_variant;missens...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000022.11:g.31620659C&gt;T;NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:31620659</td>\n",
       "      <td>22:32016645</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>1</td>\n",
       "      <td>0031620659</td>\n",
       "      <td>0</td>\n",
       "      <td>C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4103</th>\n",
       "      <td>rs2072577342</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>likely-pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>TOPMED</td>\n",
       "      <td>NC_000019.10:10812269:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa</td>\n",
       "      <td>HGVS=NC_000019.10:g.10812270C&gt;T;NC_000019.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>5068054504</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:10812270</td>\n",
       "      <td>19:10922946</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>1</td>\n",
       "      <td>0010812270</td>\n",
       "      <td>0</td>\n",
       "      <td>C:T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4104 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             dbsnp      snp_id  \\\n",
       "0            rs334         334   \n",
       "1           rs1695        1695   \n",
       "2           rs4672        4672   \n",
       "3           rs4680        4680   \n",
       "4           rs4746        4746   \n",
       "...            ...         ...   \n",
       "4099  rs1965499910  1965499910   \n",
       "4100  rs1971301248  1971301248   \n",
       "4101  rs2050933471  2050933471   \n",
       "4102  rs2072505076  2072505076   \n",
       "4103  rs2072577342  2072577342   \n",
       "\n",
       "                                  clinical_significance           acc chr  \\\n",
       "0     protective;other;pathogenic;benign;likely-beni...  NC_000011.10  11   \n",
       "1                                   benign;not-provided  NC_000011.10  11   \n",
       "2                                                   NaN  NC_000011.10  11   \n",
       "3                                  drug-response;benign  NC_000022.11  22   \n",
       "4                                uncertain-significance  NC_000006.12   6   \n",
       "...                                                 ...           ...  ..   \n",
       "4099                                         pathogenic  NC_000016.10  16   \n",
       "4100                                         pathogenic  NC_000019.10  19   \n",
       "4101                                         pathogenic  NC_000017.11  17   \n",
       "4102                       pathogenic-likely-pathogenic  NC_000022.11  22   \n",
       "4103                                  likely-pathogenic  NC_000019.10  19   \n",
       "\n",
       "                                                 handle  \\\n",
       "0     JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...   \n",
       "1     CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...   \n",
       "2     JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...   \n",
       "3     GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...   \n",
       "4     ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...   \n",
       "...                                                 ...   \n",
       "4099                                               None   \n",
       "4100                                               None   \n",
       "4101                                               None   \n",
       "4102                                               None   \n",
       "4103                                             TOPMED   \n",
       "\n",
       "                                                   spdi  \\\n",
       "0     NC_000011.10:5227001:T:A;NC_000011.10:5227001:...   \n",
       "1     NC_000011.10:67585217:A:G;NC_000011.10:6758521...   \n",
       "2     NC_000011.10:64242406:G:A;NC_000011.10:6424240...   \n",
       "3                             NC_000022.11:19963747:G:A   \n",
       "4     NC_000006.12:38682851:T:A;NC_000006.12:3868285...   \n",
       "...                                                 ...   \n",
       "4099  NC_000016.10:69320767:A:G;NC_000016.10:6932076...   \n",
       "4100                          NC_000019.10:12924491:G:T   \n",
       "4101                          NC_000017.11:41968308:C:T   \n",
       "4102                          NC_000022.11:31620658:C:T   \n",
       "4103                          NC_000019.10:10812269:C:T   \n",
       "\n",
       "                                              fxn_class  \\\n",
       "0              missense_variant;coding_sequence_variant   \n",
       "1              missense_variant;coding_sequence_variant   \n",
       "2     3_prime_UTR_variant;intron_variant;missense_va...   \n",
       "3     2KB_upstream_variant;coding_sequence_variant;u...   \n",
       "4              coding_sequence_variant;missense_variant   \n",
       "...                                                 ...   \n",
       "4099           missense_variant;coding_sequence_variant   \n",
       "4100           coding_sequence_variant;missense_variant   \n",
       "4101           coding_sequence_variant;missense_variant   \n",
       "4102  coding_sequence_variant;intron_variant;missens...   \n",
       "4103           coding_sequence_variant;missense_variant   \n",
       "\n",
       "                            validated  \\\n",
       "0     by-frequency;by-alfa;by-cluster   \n",
       "1     by-frequency;by-alfa;by-cluster   \n",
       "2     by-frequency;by-alfa;by-cluster   \n",
       "3     by-frequency;by-alfa;by-cluster   \n",
       "4     by-frequency;by-alfa;by-cluster   \n",
       "...                               ...   \n",
       "4099                              NaN   \n",
       "4100                              NaN   \n",
       "4101                              NaN   \n",
       "4102                              NaN   \n",
       "4103             by-frequency;by-alfa   \n",
       "\n",
       "                                                 docsum  ...  \\\n",
       "0     HGVS=NC_000011.10:g.5227002T>A;NC_000011.10:g....  ...   \n",
       "1     HGVS=NC_000011.10:g.67585218A>G;NC_000011.10:g...  ...   \n",
       "2     HGVS=NC_000011.10:g.64242407G>A;NC_000011.10:g...  ...   \n",
       "3     HGVS=NC_000022.11:g.19963748G>A;NC_000022.10:g...  ...   \n",
       "4     HGVS=NC_000006.12:g.38682852T>A;NC_000006.12:g...  ...   \n",
       "...                                                 ...  ...   \n",
       "4099  HGVS=NC_000016.10:g.69320768A>G;NC_000016.10:g...  ...   \n",
       "4100  HGVS=NC_000019.10:g.12924492G>T;NC_000019.9:g....  ...   \n",
       "4101  HGVS=NC_000017.11:g.41968309C>T;NC_000017.10:g...  ...   \n",
       "4102  HGVS=NC_000022.11:g.31620659C>T;NC_000022.10:g...  ...   \n",
       "4103  HGVS=NC_000019.10:g.10812270C>T;NC_000019.9:g....  ...   \n",
       "\n",
       "                                                     ss allele snp_class  \\\n",
       "0     335;24811263;48419811;49850538;49850540;498505...      N       snv   \n",
       "1     1720;5330;9110;1390210;1545478;3178003;5586573...      D       snv   \n",
       "2     1555019;4436250;20825221;24790172;39886556;752...      D       snv   \n",
       "3     5363;10902;232568;458935;593353;869771;1058120...      R       snv   \n",
       "4     3863270;20275576;24369377;28505662;44730940;48...      D       snv   \n",
       "...                                                 ...    ...       ...   \n",
       "4099                                                  0      D       snv   \n",
       "4100                                                  0      K       snv   \n",
       "4101                                                  0      Y       snv   \n",
       "4102                                                  0      Y       snv   \n",
       "4103                                         5068054504      Y       snv   \n",
       "\n",
       "           chrpos chrpos_prev_assm snp_id_sort clinical_sort chrpos_sort  \\\n",
       "0      11:5227002       11:5248232  0000000334             1  0005227002   \n",
       "1     11:67585218      11:67352689  0000001695             1  0067585218   \n",
       "2     11:64242407      11:64009879  0000004672             0  0064242407   \n",
       "3     22:19963748      22:19951271  0000004680             1  0019963748   \n",
       "4      6:38682852       6:38650628  0000004746             1  0038682852   \n",
       "...           ...              ...         ...           ...         ...   \n",
       "4099  16:69320768      16:69354671  1965499910             1  0069320768   \n",
       "4100  19:12924492      19:13035306  1971301248             1  0012924492   \n",
       "4101  17:41968309      17:40120327  2050933471             1  0041968309   \n",
       "4102  22:31620659      22:32016645  2072505076             1  0031620659   \n",
       "4103  19:10812270      19:10922946  2072577342             1  0010812270   \n",
       "\n",
       "     merged_sort sequence_change  \n",
       "0              0     T:A;T:C;T:G  \n",
       "1              0         A:G;A:T  \n",
       "2              0         G:A;G:T  \n",
       "3              0             G:A  \n",
       "4              0         T:A;T:G  \n",
       "...          ...             ...  \n",
       "4099           0         A:G;A:T  \n",
       "4100           0             G:T  \n",
       "4101           0             C:T  \n",
       "4102           0             C:T  \n",
       "4103           0             C:T  \n",
       "\n",
       "[4104 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rename_mapping = {\n",
    "    key: key.lower()\n",
    "    for key in df_snps_all.columns\n",
    "    if key not in {\"error\", \"GLOBAL_SAMPLESIZE\"}\n",
    "}\n",
    "\n",
    "df_model_snps = (\n",
    "    df_snps_all.loc[:, list(rename_mapping)]\n",
    "    .rename(rename_mapping, axis=1)\n",
    "    .reset_index(drop=False)\n",
    ")\n",
    "df_model_snps = df_model_snps.dropna(subset=[\"allele\"], axis=0)\n",
    "for col in df_model_snps.columns:\n",
    "    df_model_snps[col] = df_model_snps[col].str.split(\",\")\n",
    "    # Switch to default seperator (currently semicolon)\n",
    "    df_model_snps[col] = df_model_snps[col].apply(lambda x: build_string(x))\n",
    "df_model_snps = df_model_snps.sort_values([\"snp_id_sort\"]).reset_index(drop=True)\n",
    "df_model_snps[\"sequence_change\"] = df_model_snps[\"spdi\"].apply(\n",
    "    lambda value: build_string([s.split(\":\", 2)[-1] for s in split_string(value)])\n",
    ")\n",
    "df_model_snps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d00ca838-c94d-414b-9c4b-54f82369065e",
   "metadata": {},
   "source": [
    "### Export data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "99b623b9-e30d-4b1e-843a-7d2f2f057ec1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbsnp</th>\n",
       "      <th>snp_id</th>\n",
       "      <th>clinical_significance</th>\n",
       "      <th>acc</th>\n",
       "      <th>chr</th>\n",
       "      <th>handle</th>\n",
       "      <th>spdi</th>\n",
       "      <th>fxn_class</th>\n",
       "      <th>validated</th>\n",
       "      <th>docsum</th>\n",
       "      <th>...</th>\n",
       "      <th>ss</th>\n",
       "      <th>allele</th>\n",
       "      <th>snp_class</th>\n",
       "      <th>chrpos</th>\n",
       "      <th>chrpos_prev_assm</th>\n",
       "      <th>snp_id_sort</th>\n",
       "      <th>clinical_sort</th>\n",
       "      <th>chrpos_sort</th>\n",
       "      <th>merged_sort</th>\n",
       "      <th>sequence_change</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs334</td>\n",
       "      <td>334</td>\n",
       "      <td>protective;other;pathogenic;benign;likely-beni...</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...</td>\n",
       "      <td>NC_000011.10:5227001:T:A;NC_000011.10:5227001:...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.5227002T&gt;A;NC_000011.10:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>335;24811263;48419811;49850538;49850540;498505...</td>\n",
       "      <td>N</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:5227002</td>\n",
       "      <td>11:5248232</td>\n",
       "      <td>0000000334</td>\n",
       "      <td>1</td>\n",
       "      <td>0005227002</td>\n",
       "      <td>0</td>\n",
       "      <td>T:A;T:C;T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs1695</td>\n",
       "      <td>1695</td>\n",
       "      <td>benign;not-provided</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...</td>\n",
       "      <td>NC_000011.10:67585217:A:G;NC_000011.10:6758521...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.67585218A&gt;G;NC_000011.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1720;5330;9110;1390210;1545478;3178003;5586573...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:67585218</td>\n",
       "      <td>11:67352689</td>\n",
       "      <td>0000001695</td>\n",
       "      <td>1</td>\n",
       "      <td>0067585218</td>\n",
       "      <td>0</td>\n",
       "      <td>A:G;A:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs4672</td>\n",
       "      <td>4672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NC_000011.10</td>\n",
       "      <td>11</td>\n",
       "      <td>JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...</td>\n",
       "      <td>NC_000011.10:64242406:G:A;NC_000011.10:6424240...</td>\n",
       "      <td>3_prime_UTR_variant;intron_variant;missense_va...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000011.10:g.64242407G&gt;A;NC_000011.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>1555019;4436250;20825221;24790172;39886556;752...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>11:64242407</td>\n",
       "      <td>11:64009879</td>\n",
       "      <td>0000004672</td>\n",
       "      <td>0</td>\n",
       "      <td>0064242407</td>\n",
       "      <td>0</td>\n",
       "      <td>G:A;G:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs4680</td>\n",
       "      <td>4680</td>\n",
       "      <td>drug-response;benign</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...</td>\n",
       "      <td>NC_000022.11:19963747:G:A</td>\n",
       "      <td>2KB_upstream_variant;coding_sequence_variant;u...</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000022.11:g.19963748G&gt;A;NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>5363;10902;232568;458935;593353;869771;1058120...</td>\n",
       "      <td>R</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:19963748</td>\n",
       "      <td>22:19951271</td>\n",
       "      <td>0000004680</td>\n",
       "      <td>1</td>\n",
       "      <td>0019963748</td>\n",
       "      <td>0</td>\n",
       "      <td>G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4746</td>\n",
       "      <td>4746</td>\n",
       "      <td>uncertain-significance</td>\n",
       "      <td>NC_000006.12</td>\n",
       "      <td>6</td>\n",
       "      <td>ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...</td>\n",
       "      <td>NC_000006.12:38682851:T:A;NC_000006.12:3868285...</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa;by-cluster</td>\n",
       "      <td>HGVS=NC_000006.12:g.38682852T&gt;A;NC_000006.12:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>3863270;20275576;24369377;28505662;44730940;48...</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>6:38682852</td>\n",
       "      <td>6:38650628</td>\n",
       "      <td>0000004746</td>\n",
       "      <td>1</td>\n",
       "      <td>0038682852</td>\n",
       "      <td>0</td>\n",
       "      <td>T:A;T:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4099</th>\n",
       "      <td>rs1965499910</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000016.10</td>\n",
       "      <td>16</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000016.10:69320767:A:G;NC_000016.10:6932076...</td>\n",
       "      <td>missense_variant;coding_sequence_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000016.10:g.69320768A&gt;G;NC_000016.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>D</td>\n",
       "      <td>snv</td>\n",
       "      <td>16:69320768</td>\n",
       "      <td>16:69354671</td>\n",
       "      <td>1965499910</td>\n",
       "      <td>1</td>\n",
       "      <td>0069320768</td>\n",
       "      <td>0</td>\n",
       "      <td>A:G;A:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4100</th>\n",
       "      <td>rs1971301248</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000019.10:12924491:G:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000019.10:g.12924492G&gt;T;NC_000019.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>K</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:12924492</td>\n",
       "      <td>19:13035306</td>\n",
       "      <td>1971301248</td>\n",
       "      <td>1</td>\n",
       "      <td>0012924492</td>\n",
       "      <td>0</td>\n",
       "      <td>G:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4101</th>\n",
       "      <td>rs2050933471</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>pathogenic</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000017.11:41968308:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000017.11:g.41968309C&gt;T;NC_000017.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>17:41968309</td>\n",
       "      <td>17:40120327</td>\n",
       "      <td>2050933471</td>\n",
       "      <td>1</td>\n",
       "      <td>0041968309</td>\n",
       "      <td>0</td>\n",
       "      <td>C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4102</th>\n",
       "      <td>rs2072505076</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>pathogenic-likely-pathogenic</td>\n",
       "      <td>NC_000022.11</td>\n",
       "      <td>22</td>\n",
       "      <td>None</td>\n",
       "      <td>NC_000022.11:31620658:C:T</td>\n",
       "      <td>coding_sequence_variant;intron_variant;missens...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HGVS=NC_000022.11:g.31620659C&gt;T;NC_000022.10:g...</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>22:31620659</td>\n",
       "      <td>22:32016645</td>\n",
       "      <td>2072505076</td>\n",
       "      <td>1</td>\n",
       "      <td>0031620659</td>\n",
       "      <td>0</td>\n",
       "      <td>C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4103</th>\n",
       "      <td>rs2072577342</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>likely-pathogenic</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>TOPMED</td>\n",
       "      <td>NC_000019.10:10812269:C:T</td>\n",
       "      <td>coding_sequence_variant;missense_variant</td>\n",
       "      <td>by-frequency;by-alfa</td>\n",
       "      <td>HGVS=NC_000019.10:g.10812270C&gt;T;NC_000019.9:g....</td>\n",
       "      <td>...</td>\n",
       "      <td>5068054504</td>\n",
       "      <td>Y</td>\n",
       "      <td>snv</td>\n",
       "      <td>19:10812270</td>\n",
       "      <td>19:10922946</td>\n",
       "      <td>2072577342</td>\n",
       "      <td>1</td>\n",
       "      <td>0010812270</td>\n",
       "      <td>0</td>\n",
       "      <td>C:T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4104 rows Ã— 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             dbsnp      snp_id  \\\n",
       "0            rs334         334   \n",
       "1           rs1695        1695   \n",
       "2           rs4672        4672   \n",
       "3           rs4680        4680   \n",
       "4           rs4746        4746   \n",
       "...            ...         ...   \n",
       "4099  rs1965499910  1965499910   \n",
       "4100  rs1971301248  1971301248   \n",
       "4101  rs2050933471  2050933471   \n",
       "4102  rs2072505076  2072505076   \n",
       "4103  rs2072577342  2072577342   \n",
       "\n",
       "                                  clinical_significance           acc chr  \\\n",
       "0     protective;other;pathogenic;benign;likely-beni...  NC_000011.10  11   \n",
       "1                                   benign;not-provided  NC_000011.10  11   \n",
       "2                                                   NaN  NC_000011.10  11   \n",
       "3                                  drug-response;benign  NC_000022.11  22   \n",
       "4                                uncertain-significance  NC_000006.12   6   \n",
       "...                                                 ...           ...  ..   \n",
       "4099                                         pathogenic  NC_000016.10  16   \n",
       "4100                                         pathogenic  NC_000019.10  19   \n",
       "4101                                         pathogenic  NC_000017.11  17   \n",
       "4102                       pathogenic-likely-pathogenic  NC_000022.11  22   \n",
       "4103                                  likely-pathogenic  NC_000019.10  19   \n",
       "\n",
       "                                                 handle  \\\n",
       "0     JMKIDD_LAB;FSA-LAB;MILLER_NIDDK;KHV_HUMAN_GENO...   \n",
       "1     CSHL;BCM-HGSC-SUB;1000GENOMES;EVA_SVP;MGC_GENO...   \n",
       "2     JMKIDD_LAB;FSA-LAB;KHV_HUMAN_GENOMES;CLINSEQ_S...   \n",
       "3     GRF;YSAMUELS;EVA_SAMSUNG_MC;MGC_GENOME_DIFF;AB...   \n",
       "4     ACPOP;TRAN_CS_UWATERLOO;TISHKOFF;SI_EXO;EVA_UK...   \n",
       "...                                                 ...   \n",
       "4099                                               None   \n",
       "4100                                               None   \n",
       "4101                                               None   \n",
       "4102                                               None   \n",
       "4103                                             TOPMED   \n",
       "\n",
       "                                                   spdi  \\\n",
       "0     NC_000011.10:5227001:T:A;NC_000011.10:5227001:...   \n",
       "1     NC_000011.10:67585217:A:G;NC_000011.10:6758521...   \n",
       "2     NC_000011.10:64242406:G:A;NC_000011.10:6424240...   \n",
       "3                             NC_000022.11:19963747:G:A   \n",
       "4     NC_000006.12:38682851:T:A;NC_000006.12:3868285...   \n",
       "...                                                 ...   \n",
       "4099  NC_000016.10:69320767:A:G;NC_000016.10:6932076...   \n",
       "4100                          NC_000019.10:12924491:G:T   \n",
       "4101                          NC_000017.11:41968308:C:T   \n",
       "4102                          NC_000022.11:31620658:C:T   \n",
       "4103                          NC_000019.10:10812269:C:T   \n",
       "\n",
       "                                              fxn_class  \\\n",
       "0              missense_variant;coding_sequence_variant   \n",
       "1              missense_variant;coding_sequence_variant   \n",
       "2     3_prime_UTR_variant;intron_variant;missense_va...   \n",
       "3     2KB_upstream_variant;coding_sequence_variant;u...   \n",
       "4              coding_sequence_variant;missense_variant   \n",
       "...                                                 ...   \n",
       "4099           missense_variant;coding_sequence_variant   \n",
       "4100           coding_sequence_variant;missense_variant   \n",
       "4101           coding_sequence_variant;missense_variant   \n",
       "4102  coding_sequence_variant;intron_variant;missens...   \n",
       "4103           coding_sequence_variant;missense_variant   \n",
       "\n",
       "                            validated  \\\n",
       "0     by-frequency;by-alfa;by-cluster   \n",
       "1     by-frequency;by-alfa;by-cluster   \n",
       "2     by-frequency;by-alfa;by-cluster   \n",
       "3     by-frequency;by-alfa;by-cluster   \n",
       "4     by-frequency;by-alfa;by-cluster   \n",
       "...                               ...   \n",
       "4099                              NaN   \n",
       "4100                              NaN   \n",
       "4101                              NaN   \n",
       "4102                              NaN   \n",
       "4103             by-frequency;by-alfa   \n",
       "\n",
       "                                                 docsum  ...  \\\n",
       "0     HGVS=NC_000011.10:g.5227002T>A;NC_000011.10:g....  ...   \n",
       "1     HGVS=NC_000011.10:g.67585218A>G;NC_000011.10:g...  ...   \n",
       "2     HGVS=NC_000011.10:g.64242407G>A;NC_000011.10:g...  ...   \n",
       "3     HGVS=NC_000022.11:g.19963748G>A;NC_000022.10:g...  ...   \n",
       "4     HGVS=NC_000006.12:g.38682852T>A;NC_000006.12:g...  ...   \n",
       "...                                                 ...  ...   \n",
       "4099  HGVS=NC_000016.10:g.69320768A>G;NC_000016.10:g...  ...   \n",
       "4100  HGVS=NC_000019.10:g.12924492G>T;NC_000019.9:g....  ...   \n",
       "4101  HGVS=NC_000017.11:g.41968309C>T;NC_000017.10:g...  ...   \n",
       "4102  HGVS=NC_000022.11:g.31620659C>T;NC_000022.10:g...  ...   \n",
       "4103  HGVS=NC_000019.10:g.10812270C>T;NC_000019.9:g....  ...   \n",
       "\n",
       "                                                     ss allele snp_class  \\\n",
       "0     335;24811263;48419811;49850538;49850540;498505...      N       snv   \n",
       "1     1720;5330;9110;1390210;1545478;3178003;5586573...      D       snv   \n",
       "2     1555019;4436250;20825221;24790172;39886556;752...      D       snv   \n",
       "3     5363;10902;232568;458935;593353;869771;1058120...      R       snv   \n",
       "4     3863270;20275576;24369377;28505662;44730940;48...      D       snv   \n",
       "...                                                 ...    ...       ...   \n",
       "4099                                                  0      D       snv   \n",
       "4100                                                  0      K       snv   \n",
       "4101                                                  0      Y       snv   \n",
       "4102                                                  0      Y       snv   \n",
       "4103                                         5068054504      Y       snv   \n",
       "\n",
       "           chrpos chrpos_prev_assm snp_id_sort clinical_sort chrpos_sort  \\\n",
       "0      11:5227002       11:5248232  0000000334             1  0005227002   \n",
       "1     11:67585218      11:67352689  0000001695             1  0067585218   \n",
       "2     11:64242407      11:64009879  0000004672             0  0064242407   \n",
       "3     22:19963748      22:19951271  0000004680             1  0019963748   \n",
       "4      6:38682852       6:38650628  0000004746             1  0038682852   \n",
       "...           ...              ...         ...           ...         ...   \n",
       "4099  16:69320768      16:69354671  1965499910             1  0069320768   \n",
       "4100  19:12924492      19:13035306  1971301248             1  0012924492   \n",
       "4101  17:41968309      17:40120327  2050933471             1  0041968309   \n",
       "4102  22:31620659      22:32016645  2072505076             1  0031620659   \n",
       "4103  19:10812270      19:10922946  2072577342             1  0010812270   \n",
       "\n",
       "     merged_sort sequence_change  \n",
       "0              0     T:A;T:C;T:G  \n",
       "1              0         A:G;A:T  \n",
       "2              0         G:A;G:T  \n",
       "3              0             G:A  \n",
       "4              0         T:A;T:G  \n",
       "...          ...             ...  \n",
       "4099           0         A:G;A:T  \n",
       "4100           0             G:T  \n",
       "4101           0             C:T  \n",
       "4102           0             C:T  \n",
       "4103           0             C:T  \n",
       "\n",
       "[4104 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if overwrite:\n",
    "    df_model_snps.to_csv(f\"{ROOT_PATH}{DATABASE_PATH}/snps_{GEM_NAME}.tsv\", sep=\"\\t\")\n",
    "else:\n",
    "    df_model_snps.to_csv(f\"{ROOT_PATH}{INTERIM_PATH}/snps_{GEM_NAME}.tsv\", sep=\"\\t\")\n",
    "df_model_snps"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
