{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4738b5e2-1da1-4599-a435-5cb8a50fcbff",
   "metadata": {},
   "source": [
    "# Extract identifiers from HumanGEM\n",
    "\n",
    "Use HumanGEM (1.18.0) to get the MetabolicAtlas identifiers for reactions and metabolites.\n",
    "Will need for annotation purposes and identifier tracking in later scripts.\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a861a22e-f7b9-48ae-921b-edf672891ad2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.1\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "cobra       0.29.0\n",
      "depinfo      2.2.0\n",
      "memote      0.16.1\n",
      "notebook     7.0.6\n",
      "scipy       1.11.4\n",
      "simplejson missing\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip        23.3.1\n",
      "setuptools 68.2.2\n",
      "wheel      0.41.2\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Darwin  22.6.0-x86_64\n",
      "CPython        3.12.0\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION, \n",
    "    REPO_PATH, \n",
    "    show_versions,\n",
    "    read_rbc_model, write_rbc_model, \n",
    "    read_cobra_model,\n",
    ")\n",
    "\n",
    "from rbc_gem_utils.util import build_string\n",
    "# Display versions of last time notebook ran and worked\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1bc7831-abf7-4ec4-8618-78cd0d0b2ed2",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2c17fa71-5e41-46fa-8a5b-e745e334567a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f433ff-e33a-4f26-9712-eba496eca52c",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 0.1.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dceb1ce4-c33f-4ec4-b4c0-46e73f126cef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2024-11-28\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iAB_RBC_283</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>1466f6ea0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>342</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>469</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>346</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>41</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iAB_RBC_283 at 0x1466f6ea0>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_rbc_model(filetype=\"xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cac469ef-abd2-4e92-ba94-e77cceb5d357",
   "metadata": {},
   "source": [
    "## Load HumanGEM model\n",
    "### Version: 1.18.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b79aab33-634e-4eed-901a-4dbe83a4b44f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>HumanGEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>146ed4cb0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>8456</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>12995</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>2889</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>148</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*MAR13082 - 1.0*MAR13082_reverse_11d67</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytosol, Extracellular, Lysosome, Endoplasmic reticulum, Mitochondria, Peroxisome, Golgi apparatus, Nucleus, Inner mitochondria</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model HumanGEM at 0x146ed4cb0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HumanGEM = read_cobra_model(f\"{REPO_PATH}/data/raw/Human-GEM.xml\")\n",
    "HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "562e9311-c7a9-43bf-a053-8754cbf0e369",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[' Arachidonic acid metabolism',\n",
       " 'Acyl-CoA hydrolysis',\n",
       " 'Acylglycerides metabolism',\n",
       " 'Alanine, aspartate and glutamate metabolism',\n",
       " 'Alkaloids biosynthesis',\n",
       " 'Amino sugar and nucleotide sugar metabolism',\n",
       " 'Aminoacyl-tRNA biosynthesis',\n",
       " 'Androgen metabolism',\n",
       " 'Arachidonic acid metabolism',\n",
       " 'Arginine and proline metabolism',\n",
       " 'Artificial reactions',\n",
       " 'Ascorbate and aldarate metabolism',\n",
       " 'Beta oxidation of branched-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of di-unsaturated fatty acids (n-6) (mitochondrial)',\n",
       " 'Beta oxidation of di-unsaturated fatty acids (n-6) (peroxisomal)',\n",
       " 'Beta oxidation of even-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of even-chain fatty acids (peroxisomal)',\n",
       " 'Beta oxidation of odd-chain fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of odd-chain fatty acids (peroxisomal)',\n",
       " 'Beta oxidation of phytanic acid (peroxisomal)',\n",
       " 'Beta oxidation of poly-unsaturated fatty acids (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-7) (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-7) (peroxisomal)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-9) (mitochondrial)',\n",
       " 'Beta oxidation of unsaturated fatty acids (n-9) (peroxisomal)',\n",
       " 'Beta-alanine metabolism',\n",
       " 'Bile acid biosynthesis',\n",
       " 'Bile acid recycling',\n",
       " 'Biopterin metabolism',\n",
       " 'Biotin metabolism',\n",
       " 'Blood group biosynthesis',\n",
       " 'Butanoate metabolism',\n",
       " 'C5-branched dibasic acid metabolism',\n",
       " 'Carnitine shuttle (cytosolic)',\n",
       " 'Carnitine shuttle (endoplasmic reticular)',\n",
       " 'Carnitine shuttle (mitochondrial)',\n",
       " 'Carnitine shuttle (peroxisomal)',\n",
       " 'Cholesterol biosynthesis 1 (Bloch pathway)',\n",
       " 'Cholesterol biosynthesis 2',\n",
       " 'Cholesterol biosynthesis 3 (Kandustch-Russell pathway)',\n",
       " 'Cholesterol metabolism',\n",
       " 'Chondroitin / heparan sulfate biosynthesis',\n",
       " 'Chondroitin sulfate degradation',\n",
       " 'Cysteine and methionine metabolism',\n",
       " 'Dietary fiber binding',\n",
       " 'Drug metabolism',\n",
       " 'Eicosanoid metabolism',\n",
       " 'Estrogen metabolism',\n",
       " 'Ether lipid metabolism',\n",
       " 'Exchange/demand reactions',\n",
       " 'Fatty acid activation (cytosolic)',\n",
       " 'Fatty acid activation (endoplasmic reticular)',\n",
       " 'Fatty acid biosynthesis',\n",
       " 'Fatty acid biosynthesis (even-chain)',\n",
       " 'Fatty acid biosynthesis (odd-chain)',\n",
       " 'Fatty acid biosynthesis (unsaturated)',\n",
       " 'Fatty acid degradation',\n",
       " 'Fatty acid desaturation (even-chain)',\n",
       " 'Fatty acid desaturation (odd-chain)',\n",
       " 'Fatty acid elongation (even-chain)',\n",
       " 'Fatty acid elongation (odd-chain)',\n",
       " 'Fatty acid metabolism',\n",
       " 'Fatty acid oxidation',\n",
       " 'Folate metabolism',\n",
       " 'Formation and hydrolysis of cholesterol esters',\n",
       " 'Fructose and mannose metabolism',\n",
       " 'Galactose metabolism',\n",
       " 'Glucocorticoid biosynthesis',\n",
       " 'Glutathione metabolism',\n",
       " 'Glycerolipid metabolism',\n",
       " 'Glycerophospholipid metabolism',\n",
       " 'Glycine, serine and threonine metabolism',\n",
       " 'Glycolysis / Gluconeogenesis',\n",
       " 'Glycosphingolipid biosynthesis-ganglio series',\n",
       " 'Glycosphingolipid biosynthesis-globo series',\n",
       " 'Glycosphingolipid biosynthesis-lacto and neolacto series',\n",
       " 'Glycosphingolipid metabolism',\n",
       " 'Glycosylphosphatidylinositol (GPI)-anchor biosynthesis',\n",
       " 'Glyoxylate and dicarboxylate metabolism',\n",
       " 'Heme degradation',\n",
       " 'Heparan sulfate degradation',\n",
       " 'Hippurate metabolism',\n",
       " 'Histidine metabolism',\n",
       " 'Inositol phosphate metabolism',\n",
       " 'Insect hormone biosynthesis',\n",
       " 'Isolated',\n",
       " 'Keratan sulfate biosynthesis',\n",
       " 'Keratan sulfate degradation',\n",
       " 'Leukotriene metabolism',\n",
       " 'Linoleate metabolism',\n",
       " 'Lipoic acid metabolism',\n",
       " 'Lysine metabolism',\n",
       " 'Metabolism of other amino acids',\n",
       " 'Miscellaneous',\n",
       " 'N-glycan metabolism',\n",
       " 'Nicotinate and nicotinamide metabolism',\n",
       " 'Nucleotide metabolism',\n",
       " 'O-glycan metabolism',\n",
       " 'Omega-3 fatty acid metabolism',\n",
       " 'Omega-6 fatty acid metabolism',\n",
       " 'Oxidative phosphorylation',\n",
       " 'Pantothenate and CoA metabolism',\n",
       " 'Pentose Phosphate Pathway',\n",
       " 'Pentose and glucuronate interconversions',\n",
       " 'Pentose phosphate pathway',\n",
       " 'Peptide metabolism',\n",
       " 'Phenylalanine metabolism',\n",
       " 'Phenylalanine, tyrosine and tryptophan biosynthesis',\n",
       " 'Phosphatidylinositol phosphate metabolism',\n",
       " 'Pool reactions',\n",
       " 'Porphyrin metabolism',\n",
       " 'Propanoate metabolism',\n",
       " 'Prostaglandin biosynthesis',\n",
       " 'Protein assembly',\n",
       " 'Protein degradation',\n",
       " 'Protein modification',\n",
       " 'Purine metabolism',\n",
       " 'Pyrimidine metabolism',\n",
       " 'Pyruvate metabolism',\n",
       " 'ROS detoxification',\n",
       " 'Retinol metabolism',\n",
       " 'Riboflavin metabolism',\n",
       " 'Serotonin and melatonin biosynthesis',\n",
       " 'Sphingolipid metabolism',\n",
       " 'Starch and sucrose metabolism',\n",
       " 'Steroid metabolism',\n",
       " 'Sulfur metabolism',\n",
       " 'Terpenoid backbone biosynthesis',\n",
       " 'Thiamine metabolism',\n",
       " 'Toluene degradation',\n",
       " 'Transport reactions',\n",
       " 'Triacylglycerol synthesis',\n",
       " 'Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism',\n",
       " 'Tryptophan metabolism',\n",
       " 'Tyrosine metabolism',\n",
       " 'Ubiquinone and other terpenoid-quinone biosynthesis',\n",
       " 'Ubiquinone synthesis',\n",
       " 'Urea cycle',\n",
       " 'Valine, leucine, and isoleucine metabolism',\n",
       " 'Vitamin A metabolism',\n",
       " 'Vitamin B12 metabolism',\n",
       " 'Vitamin B2 metabolism',\n",
       " 'Vitamin B6 metabolism',\n",
       " 'Vitamin C metabolism',\n",
       " 'Vitamin D metabolism',\n",
       " 'Vitamin E metabolism',\n",
       " 'Xenobiotics metabolism',\n",
       " 'octane oxidation']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HumanGEM.groups.list_attr(\"name\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8e268bc2-586d-42cd-ba11-228e2982ddd3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "4-aminobutanal <=> 1-pyrroline + H2O\n",
      "{'c'}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for g in HumanGEM.groups.query(lambda x: x.name.lower() == 'Urea cycle'.lower()):\n",
    "    for m in g.members:\n",
    "        if any([c not in {\"e\", \"c\"} for c in m.compartments]):\n",
    "            continue\n",
    "        print(m.name)\n",
    "        print(m.build_reaction_string(use_metabolite_names=True))\n",
    "        print(m.compartments)\n",
    "        print()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "903d7dcb-aae7-4a04-95a0-01a3940b1168",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>BILIRBU</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>Bilirubin UDP-glucuronosyltransferase</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x1474ff9e0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>bilirub_c + 2.0 h_c + udpglcur_c --> bilglcur_c + udp_c</p>\n",
       "                    <p style='text-align:right'>Bilirubin cytosol + 2.0 H+ + UDP-D-glucuronate --> Bilirubin monoglucuronide + UDP C9H11N2O12P2</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>Ugt1A1_AT1 or Ugt1A4_AT1</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction BILIRBU at 0x1474ff9e0>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.BILIRBU"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4544fdd4-791f-42c7-b153-c0cfc0d86bd8",
   "metadata": {},
   "source": [
    "## Map model BiGG ID reactions to MetabolicAtlas via HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4188a4dc-1a6b-4a22-9d92-69f75fe058d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions in RBC-GEM (excluding boundaries): 469\n",
      "Number of reactions that could be found: 276\n",
      "\n",
      "Check CAATPS for false mappings\n",
      "Check CAT for false mappings\n",
      "Check CHLPCTD for false mappings\n",
      "Check CHOLK for false mappings\n",
      "Check CHOLt4 for false mappings\n",
      "Check COt for false mappings\n",
      "Check CYStec for false mappings\n",
      "Check CYTK1 for false mappings\n",
      "Check EX_ac_e for false mappings\n",
      "Check EX_adrnl_e for false mappings\n",
      "Check EX_ala__L_e for false mappings\n",
      "Check EX_arg__L_e for false mappings\n",
      "Check EX_chol_e for false mappings\n",
      "Check EX_cl_e for false mappings\n",
      "Check EX_dopa_e for false mappings\n",
      "Check EX_fe2_e for false mappings\n",
      "Check EX_gal_e for false mappings\n",
      "Check EX_gam_e for false mappings\n",
      "Check EX_glc__D_e for false mappings\n",
      "Check EX_gln__L_e for false mappings\n",
      "Check EX_h2o_e for false mappings\n",
      "Check EX_h_e for false mappings\n",
      "Check EX_hco3_e for false mappings\n",
      "Check EX_k_e for false mappings\n",
      "Check EX_lnlc_e for false mappings\n",
      "Check EX_met__L_e for false mappings\n",
      "Check EX_nac_e for false mappings\n",
      "Check EX_nh4_e for false mappings\n",
      "Check EX_nrpphr_e for false mappings\n",
      "Check EX_ocdcea_e for false mappings\n",
      "Check EX_phe__L_e for false mappings\n",
      "Check EX_pyr_e for false mappings\n",
      "Check EX_ribflv_e for false mappings\n",
      "Check EX_thmmp_e for false mappings\n",
      "Check FBA for false mappings\n",
      "Check FBP26 for false mappings\n",
      "Check FUM for false mappings\n",
      "Check G6PDH2r for false mappings\n",
      "Check GALKr for false mappings\n",
      "Check GAMt1r for false mappings\n",
      "Check GAPD for false mappings\n",
      "Check GLNt4 for false mappings\n",
      "Check MEPIVESSte for false mappings\n",
      "Check NADK for false mappings\n",
      "Check NH4t3r for false mappings\n",
      "Check PPPGO for false mappings\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bigg</th>\n",
       "      <th>metatlas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3MOXTYRESSte</td>\n",
       "      <td>MAR11306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4PYRDX</td>\n",
       "      <td>MAR08103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5AOPt2</td>\n",
       "      <td>MAR11307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACALDt</td>\n",
       "      <td>MAR04948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGAM2E</td>\n",
       "      <td>MAR04527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>TMDPPK</td>\n",
       "      <td>MAR04206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>UNK3</td>\n",
       "      <td>MAR05391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>UPPDC1</td>\n",
       "      <td>MAR04750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>XYLK</td>\n",
       "      <td>MAR04595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>XYLTD_D</td>\n",
       "      <td>MAR04593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>322 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             bigg  metatlas\n",
       "0    3MOXTYRESSte  MAR11306\n",
       "1          4PYRDX  MAR08103\n",
       "2          5AOPt2  MAR11307\n",
       "3          ACALDt  MAR04948\n",
       "4         ACGAM2E  MAR04527\n",
       "..            ...       ...\n",
       "317        TMDPPK  MAR04206\n",
       "318          UNK3  MAR05391\n",
       "319        UPPDC1  MAR04750\n",
       "320          XYLK  MAR04595\n",
       "321       XYLTD_D  MAR04593\n",
       "\n",
       "[322 rows x 2 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_reaction_ids = {reaction.id for reaction in model.reactions}\n",
    "\n",
    "bigg_reactions = set()\n",
    "human_gem_mapping = {}\n",
    "for reaction in HumanGEM.reactions:\n",
    "    bigg_ids = reaction.annotation.get(\"bigg.reaction\")\n",
    "    if not bigg_ids:\n",
    "        continue\n",
    "    if isinstance(bigg_ids, str):\n",
    "        bigg_ids = [bigg_ids]\n",
    "    bigg_ids = set(bigg_ids)\n",
    "    bigg_reactions.update(bigg_ids)\n",
    "    for bigg_id in bigg_ids:\n",
    "        if bigg_id in human_gem_mapping:\n",
    "            human_gem_mapping[bigg_id] += [reaction.id]\n",
    "        else:\n",
    "            human_gem_mapping[bigg_id] = [reaction.id]\n",
    "\n",
    "\n",
    "print(f\"Number of reactions in RBC-GEM (excluding boundaries): {len(model_reaction_ids)}\")\n",
    "\n",
    "intersection = sorted(bigg_reactions.intersection(model_reaction_ids))\n",
    "print(f\"Number of reactions that could be found: {len(intersection)}\")\n",
    "print()\n",
    "id_mapping_dict = {}\n",
    "for rbc_reaction in intersection:\n",
    "    reaction_ids = human_gem_mapping.get(rbc_reaction, [])\n",
    "    if len(reaction_ids) > 1:\n",
    "        print(f\"Check {rbc_reaction} for false mappings\")\n",
    "    id_mapping_dict[rbc_reaction] = build_string(reaction_ids)\n",
    "\n",
    "manual_updates_and_corrections = {\n",
    "    # Selected reactions below needed corrections.\n",
    "    \"CAATPS\": \"MAR07629\",\n",
    "    \"CAT\": \"MAR03980\",\n",
    "    \"CHLPCTD\": \"MAR00638\",\n",
    "    \"CHOLK\": \"MAR00636\",\n",
    "    \"CHOLt4\": \"MAR07734\",\n",
    "    \"COt\": \"MAR07798\",\n",
    "    \"CYStec\": \"MAR05084\",\n",
    "    \"CYTK1\": \"MAR04024\",\n",
    "    \"EX_ac_e\": \"MAR09086\",\n",
    "    \"EX_adrnl_e\": \"MAR09095\",\n",
    "    \"EX_ala__L_e\": \"MAR09061\",\n",
    "    \"EX_arg__L_e\": \"MAR09066\",\n",
    "    \"EX_chol_e\": \"MAR09083\",\n",
    "    \"EX_cl_e\": \"MAR09150\",\n",
    "    \"EX_dopa_e\": \"MAR09092\",\n",
    "    \"EX_fe2_e\": \"MAR09076\",\n",
    "    \"EX_gal_e\": \"MAR09140\",\n",
    "    \"EX_gam_e\": \"MAR09168\",\n",
    "    \"EX_glc__D_e\": \"MAR09034\",\n",
    "    \"EX_gln__L_e\": \"MAR09063\",\n",
    "    \"EX_h2o_e\": \"MAR09047\",\n",
    "    \"EX_h_e\": \"MAR09079\",\n",
    "    \"EX_hco3_e\": \"MAR09078\",\n",
    "    \"EX_k_e\": \"MAR09081\",\n",
    "    \"EX_lnlc_e\": \"MAR09035\",\n",
    "    \"EX_met__L_e\": \"MAR09042\",\n",
    "    \"EX_nac_e\": \"MAR09142\",\n",
    "    \"EX_nh4_e\": \"MAR11420\",\n",
    "    \"EX_nrpphr_e\": \"MAR09093\",\n",
    "    \"EX_ocdcea_e\": \"MAR00650\",\n",
    "    \"EX_phe__L_e\": \"MAR09043\",\n",
    "    \"EX_pyr_e\": \"MAR09133\",\n",
    "    \"EX_ribflv_e\": \"MAR09143\",\n",
    "    \"EX_thmmp_e\": \"MAR09105\",\n",
    "    \"FBA\": \"MAR04375\",\n",
    "    \"FBP26\": \"MAR04706\",\n",
    "    \"FUM\": \"MAR04408\",\n",
    "    \"G6PDH2r\": \"MAR08971\",\n",
    "    \"GALKr\": \"MAR04130\",\n",
    "    \"GAMt1r\": \"MAR04996\",\n",
    "    \"GAPD\": \"MAR04373\",\n",
    "    \"GLNt4\": \"MAR05308\",\n",
    "    \"MEPIVESSte\": \"MAR08922\",\n",
    "    \"NADK\": \"MAR04269\",\n",
    "    \"NH4t3r\": \"MAR01534\",\n",
    "    \"PPPGO\": \"MAR11316\",\n",
    "    # Selected reactions needed manaul updates\n",
    "    \"ACP1_FMN\": \"MAR06507\",\n",
    "    \"ADRNLtu\": \"MAR09192\",\n",
    "    \"ARD\": \"MAR05389\",\n",
    "    \"ARGN\": \"MAR03816\",\n",
    "    \"BILIRBU\": \"MAR11321\",\n",
    "    \"C160CPT2rbc\": \"MAR02626\",\n",
    "    \"C181CPT2rbc\": \"MAR11310\",\n",
    "    \"CHLP\": \"MAR08424\",\n",
    "    \"DGULND\": \"MAR08353\",\n",
    "    \"DHAAt1r\": \"MAR08846\",\n",
    "    \"DOPAMT\": \"MAR06763\",\n",
    "    \"DPGM\": \"MAR04371\",\n",
    "    \"DPGase\": \"MAR04372\",\n",
    "    \"ENOPH\": \"MAR05387\",\n",
    "    \"ETHAt\": \"MAR07896\",\n",
    "    \"GALOR\": \"MAR08766\",\n",
    "    \"GALT\": \"MAR08767\",\n",
    "    \"GLCt1\": \"MAR05029\",\n",
    "    \"GMPR\": \"MAR04419\",\n",
    "    \"GPDDA1\": \"MAR00635\",\n",
    "    \"GULND\": \"MAR06537\",\n",
    "    \"HCO3_CLt\": \"MAR06525\",\n",
    "    \"LEUKTRA4t\": \"MAR06254\",\n",
    "    \"LEUKTRB4t\": \"MAR06255\",\n",
    "    \"LNLCCPT2rbc\": \"MAR02742\",\n",
    "    \"LTA4H\": \"MAR01080\",\n",
    "    \"MDRPD\": \"MAR05386\",\n",
    "    \"MI1345PP\": \"MAR06565\",\n",
    "    \"MI145PK\": \"MAR06563\",\n",
    "    \"MI145PP\": \"MAR06560\",\n",
    "    \"MTRI\": \"MAR05385\",\n",
    "    \"NADPN\": \"MAR07627\",\n",
    "    \"NMNHYD\": \"MAR04264\",\n",
    "    \"ORNDC\": \"MAR04212\",\n",
    "    \"RNMK\": \"MAR04265\",\n",
    "    \"SALMCOM\": \"MAR06750\",\n",
    "    \"SALMCOM2\": \"MAR06746\",\n",
    "    \"SBTD_D2\": \"MAR04315\",\n",
    "    \"SBTR\": \"MAR04316\",\n",
    "    \"SPMDtex2\": \"MAR04994\",\n",
    "    \"TDP\": \"MAR04208\",\n",
    "    \"THMTP\": \"MAR04207\",\n",
    "    \"TMDPK\": \"MAR04204\",\n",
    "    \"TMDPPK\": \"MAR04206\",\n",
    "    \"UNK3\": \"MAR05391\",\n",
    "    \"UPPDC1\": \"MAR04750\",\n",
    "    \"XYLK\": \"MAR04595\",\n",
    "    \"XYLTD_D\": \"MAR04593\",\n",
    "}\n",
    "\n",
    "id_mapping_dict.update(manual_updates_and_corrections)\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"bigg\", \"metatlas\"]\n",
    "# id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{REPO_PATH}/data/interim/BiGGMetAtlasReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c123ae55-daa2-4ec8-a8a5-041cb8965a29",
   "metadata": {},
   "source": [
    "## Map model BiGG ID metabolites to MetabolicAtlas via HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0163b9ac-86d5-4f92-adeb-573f68a82832",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of metabolites in RBC-GEM (excluding compartments): 267\n",
      "Number of metabolites that could be found: 205\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bigg</th>\n",
       "      <th>metatlas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13dpg</td>\n",
       "      <td>MAM00247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23dpg</td>\n",
       "      <td>MAM00569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2kmb</td>\n",
       "      <td>MAM01016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2pg</td>\n",
       "      <td>MAM00674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35cgmp</td>\n",
       "      <td>MAM01433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>xmp</td>\n",
       "      <td>MAM03150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>xu5p__D</td>\n",
       "      <td>MAM01761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>xylt</td>\n",
       "      <td>MAM03155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>xylu__D</td>\n",
       "      <td>MAM01759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>xylu__L</td>\n",
       "      <td>MAM02425</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        bigg  metatlas\n",
       "0      13dpg  MAM00247\n",
       "1      23dpg  MAM00569\n",
       "2       2kmb  MAM01016\n",
       "3        2pg  MAM00674\n",
       "4     35cgmp  MAM01433\n",
       "..       ...       ...\n",
       "200      xmp  MAM03150\n",
       "201  xu5p__D  MAM01761\n",
       "202     xylt  MAM03155\n",
       "203  xylu__D  MAM01759\n",
       "204  xylu__L  MAM02425\n",
       "\n",
       "[205 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_metabolite_ids = {metabolite.id.replace(f\"_{metabolite.compartment}\", \"\") for metabolite in model.metabolites}\n",
    "bigg_metabolites = set()\n",
    "human_gem_mapping = {}\n",
    "for metabolite in HumanGEM.metabolites:\n",
    "    bigg_ids = metabolite.annotation.get(\"bigg.metabolite\")\n",
    "    if not bigg_ids:\n",
    "        continue\n",
    "    if isinstance(bigg_ids, str):\n",
    "        bigg_ids = [bigg_ids]\n",
    "    bigg_ids = set(bigg_ids)\n",
    "    bigg_metabolites.update(bigg_ids)\n",
    "    for bigg_id in bigg_ids:\n",
    "        if bigg_id in human_gem_mapping:\n",
    "            human_gem_mapping[bigg_id] += [metabolite.id[:-1]]\n",
    "        else:\n",
    "            human_gem_mapping[bigg_id] = [metabolite.id[:-1]]\n",
    "\n",
    "\n",
    "print(f\"Number of metabolites in RBC-GEM (excluding compartments): {len(model_metabolite_ids)}\")\n",
    "\n",
    "intersection = sorted(bigg_metabolites.intersection(model_metabolite_ids))\n",
    "print(f\"Number of metabolites that could be found: {len(intersection)}\")\n",
    "print()\n",
    "\n",
    "id_mapping_dict = {}\n",
    "for rbc_metabolite in intersection:\n",
    "    metabolite_ids = set(human_gem_mapping.get(rbc_metabolite, []))\n",
    "    if len(metabolite_ids) > 1:\n",
    "        print(f'\"{rbc_metabolite}\": \"{build_string(metabolite_ids)}\",')\n",
    "    id_mapping_dict[rbc_metabolite] = build_string(metabolite_ids)\n",
    "\n",
    "manual_updates_and_corrections = {\n",
    "    # Selected reactions below needed corrections.\n",
    " \n",
    "}\n",
    "\n",
    "id_mapping_dict.update(manual_updates_and_corrections)\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"bigg\", \"metatlas\"]\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{REPO_PATH}/data/interim/BiGGMetAtlasMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2944c812-56b9-4051-a483-543d9cd73135",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
