{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "56f56243-7f70-407b-a216-5cee91ed7dc8",
   "metadata": {},
   "source": [
    "# Model Connectivity Comparison Overview\n",
    "The purpose of this notebook is to compare the current version of the RBC-GEM to earlier versions.\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d043a88b-e734-40e9-aa5c-b0def780e477",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "import seaborn as sns\n",
    "from cobra.manipulation import rename_genes\n",
    "from rbc_gem_utils import (\n",
    "    CURATION_PATH,\n",
    "    GEM_NAME,\n",
    "    GEM_URL,\n",
    "    MODEL_PATH,\n",
    "    ROOT_PATH,\n",
    "    build_string,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    ")\n",
    "from rbc_gem_utils.util import ensure_iterable\n",
    "from rbc_gem_utils.visualization import cmap_map\n",
    "\n",
    "# Display versions of last time notebook ran and worked\n",
    "show_versions()\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c373aa30-79e3-4844-b43f-bcb79301b907",
   "metadata": {},
   "source": [
    "## Load RBC-GEM models\n",
    "### Current model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71900e02-745e-43d4-b656-8a98766d8652",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "current_model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "current_model.metabolites.sort()\n",
    "current_model.genes.sort()\n",
    "current_model.reactions.sort()\n",
    "current_model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df766764-8c11-47bd-9762-616e4e666203",
   "metadata": {},
   "source": [
    "### Earlier models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f4eafa1-ff76-4244-9a35-8996d55847db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Probably a better way to do this instead of erroring out.\n",
    "download = True\n",
    "version = \"0.2.0\"\n",
    "ftype = \"xml\"\n",
    "\n",
    "filepath = get_dirpath(use_temp=\"interim\") / f\"{GEM_NAME}_{version}.{ftype}\"\n",
    "\n",
    "if download:\n",
    "    response = requests.get(f\"{GEM_URL}/{version}/{MODEL_PATH}/{GEM_NAME}.{ftype}\")\n",
    "    response.raise_for_status()\n",
    "\n",
    "    # Write file\n",
    "    # Is there a better way of checking whether binary file?\n",
    "    if not response.encoding:\n",
    "        with open(filepath, \"wb\") as file:\n",
    "            file.write(response.content)\n",
    "    else:\n",
    "        with open(filepath, \"w\") as file:\n",
    "            file.write(response.text)\n",
    "\n",
    "prev_model_020 = read_cobra_model(filepath)\n",
    "prev_model_020.metabolites.sort()\n",
    "prev_model_020.genes.sort()\n",
    "prev_model_020.reactions.sort()\n",
    "prev_model_020.groups.sort()\n",
    "# For comparison\n",
    "group = prev_model_020.groups.get_by_id(\"Transport reactions\")\n",
    "group.id = \"Transport, extracellular\"\n",
    "\n",
    "\n",
    "# For comparison rename genes and remove splice isoforms\n",
    "rename_genes(\n",
    "    prev_model_020,\n",
    "    {g.id: g.annotation.get(\"hgnc.symbol\") for g in prev_model_020.genes},\n",
    ")\n",
    "prev_model_020.repair()\n",
    "print(f\"Genes: {len(set([x.id for x in prev_model_020.genes]))}\")\n",
    "print(f\"Metabolites (all): {len(set([x.id for x in prev_model_020.metabolites]))}\")\n",
    "print(\n",
    "    \"Metabolites (unique): {}\".format(\n",
    "        len({x.id.replace(f\"_{x.compartment}\", \"\") for x in prev_model_020.metabolites})\n",
    "    )\n",
    ")\n",
    "print(\n",
    "    f\"Reactions: {len(set([x.id for x in prev_model_020.reactions.query(lambda x: not x.boundary)]))}\"\n",
    ")\n",
    "prev_model_020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b59b8f4-f907-437d-8e09-715c3716ed74",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(prev_model_020.reactions.query(lambda x: x.subsystem != \"Pseudoreactions\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0814b51-1a0e-40d4-88c9-39915d4a3267",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_list = [prev_model_020, current_model]\n",
    "model_data = defaultdict(dict)\n",
    "\n",
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\"\n",
    "\n",
    "results_path = Path(\n",
    "    get_dirpath(use_temp=\"processed\") / current_model.id / \"ConnectivityComparison\"\n",
    ")\n",
    "results_path.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1971732-15f2-4469-a24e-63d445ef659c",
   "metadata": {},
   "source": [
    "### Compare connectivities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "537ba740-51c9-49f7-ac11-079b218be9c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_connectivities(\n",
    "    model,\n",
    "    attribute_type=\"metabolites\",\n",
    "    compartments=None,\n",
    "    exclude=None,\n",
    "    sum_compartments=False,\n",
    "    aggregate_ranking=False,\n",
    "    include_boundary=False,\n",
    "):\n",
    "    \"\"\"Compute the metabolite connectivities of the model.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    model : Model\n",
    "        The model object\n",
    "    attribute_type : {\"metabolites\", \"genes\"}\n",
    "        The type of object.\n",
    "    compartments : list, None\n",
    "        Compartments that should be accounted for. If ``None``, all metabolite compartments are utilized.\n",
    "    exclude : iterable\n",
    "        Contains identifiers for metabolites that should be excluded from calculations.\n",
    "        Useful for excluding currency metabolites (e.g., hydrogen, water) from calculations.\n",
    "    sum_compartments : bool\n",
    "        Whether to treat the metabolite as one entity and sum metabolite connectivities across compartments.\n",
    "        If True, the connectivity is determined without regard to the species compartment.\n",
    "        If False, the connectivity is determined seperately for species in different compartments.\n",
    "    aggregate_ranking : bool\n",
    "        If True, metabolites that have the same connectivity value are grouped together.\n",
    "        If False, metabolites are kept seperately.\n",
    "    include_boundary : bool\n",
    "        If True, pseudoreactions are included in the connectivity count.\n",
    "        If False, metabolites are kept seperately.\n",
    "    \"\"\"\n",
    "    if not include_boundary:\n",
    "        # Copy model and remove boundaries\n",
    "        model = model.copy()\n",
    "        model.remove_groups([model.groups.get_by_id(\"Pseudoreactions\")])\n",
    "        model.remove_reactions(model.boundary)\n",
    "    species = getattr(model, attribute_type)\n",
    "    # Get compartments\n",
    "    if compartments:\n",
    "        compartments = ensure_iterable(compartments)\n",
    "    else:\n",
    "        compartments = sorted(model.compartments)\n",
    "\n",
    "    # Exclude metabolites\n",
    "    if exclude is None:\n",
    "        exclude = set()\n",
    "    else:\n",
    "        exclude = {str(x) for x in ensure_iterable(exclude)}\n",
    "\n",
    "    if attribute_type == \"metabolites\":\n",
    "        species = species.query(\n",
    "            lambda x: (\n",
    "                (\n",
    "                    not x.id.replace(f\"_{x.compartment}\", \"\") in exclude\n",
    "                    and not x.id in exclude\n",
    "                )\n",
    "                if sum_compartments\n",
    "                else (not x.id in exclude)\n",
    "            )\n",
    "        )\n",
    "        species = species.query(lambda x: x.compartment in set(compartments))\n",
    "    else:\n",
    "        species = species.query(lambda x: not x.id in exclude)\n",
    "\n",
    "    # Compute absolute connectivities\n",
    "    absolute_connectivities = defaultdict(int)\n",
    "    absolute_connectivities = pd.Series(\n",
    "        {specie.id: len(specie.reactions) for specie in species}, name=\"absolute\"\n",
    "    ).sort_values(ascending=False)\n",
    "    absolute_connectivities.index.name = attribute_type\n",
    "\n",
    "    if sum_compartments and attribute_type == \"metabolites\":\n",
    "        # Aggregate metabolites across compartments and sum\n",
    "        absolute_connectivities = absolute_connectivities.rename(\n",
    "            {\n",
    "                sid: sid.replace(f\"_{species.get_by_id(sid).compartment}\", \"\")\n",
    "                for sid in absolute_connectivities.index\n",
    "            },\n",
    "            axis=0,\n",
    "        )\n",
    "        absolute_connectivities = absolute_connectivities.groupby(\n",
    "            absolute_connectivities.index\n",
    "        ).sum()\n",
    "\n",
    "    # Compute relative connectivities\n",
    "    relative_connectivities = absolute_connectivities / absolute_connectivities.sum()\n",
    "    relative_connectivities.name = \"relative\"\n",
    "    # Create DataFrame\n",
    "    df_connectivity = pd.merge(\n",
    "        absolute_connectivities,\n",
    "        relative_connectivities,\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "    )\n",
    "    df_connectivity = df_connectivity.reset_index(drop=False)\n",
    "    df_connectivity = df_connectivity.sort_values(\n",
    "        [\"absolute\", attribute_type], ascending=False\n",
    "    )\n",
    "\n",
    "    if aggregate_ranking:\n",
    "        # Aggregate metabolites that share the same ranking\n",
    "        df_connectivity = (\n",
    "            df_connectivity.groupby([\"absolute\", \"relative\"], as_index=False)\n",
    "            .agg(lambda x: set(x))\n",
    "            .sort_index(ascending=False)\n",
    "        )\n",
    "    # Reset index, reorganize columns, and return\n",
    "    df_connectivity = df_connectivity.reset_index(drop=True).loc[\n",
    "        :, [attribute_type, \"absolute\", \"relative\"]\n",
    "    ]\n",
    "    return df_connectivity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c915ba7f-2309-4e0a-876d-6df310f41222",
   "metadata": {},
   "source": [
    "#### Calculate connectivities without excluding any metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cad5922-e177-4bd2-b5b1-22306c7e7bcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_display = 15\n",
    "sum_compartments = True\n",
    "aggregate_ranking = False\n",
    "include_boundary = False\n",
    "attribute_type = \"metabolites\"\n",
    "exclude_list = []\n",
    "\n",
    "for model in model_list:\n",
    "    compartments = [\"e\"]\n",
    "    df_connectivity = compute_connectivities(\n",
    "        model,\n",
    "        attribute_type=attribute_type,\n",
    "        compartments=compartments,\n",
    "        exclude=exclude_list,\n",
    "        sum_compartments=sum_compartments,\n",
    "        aggregate_ranking=aggregate_ranking,\n",
    "    )\n",
    "    df_connectivity[attribute_type] = df_connectivity[attribute_type].apply(\n",
    "        lambda x: build_string(x, sep=\", \")\n",
    "    )\n",
    "    df_connectivity = df_connectivity.set_index(\"metabolites\")\n",
    "    df_connectivity.columns = pd.MultiIndex.from_product(\n",
    "        [[model.compartments[\"e\"]], df_connectivity.columns]\n",
    "    )\n",
    "    df_connectivity_e = df_connectivity.copy()\n",
    "\n",
    "    compartments = [\"c\"]\n",
    "    df_connectivity = compute_connectivities(\n",
    "        model,\n",
    "        attribute_type=attribute_type,\n",
    "        compartments=compartments,\n",
    "        exclude=exclude_list,\n",
    "        sum_compartments=sum_compartments,\n",
    "        aggregate_ranking=aggregate_ranking,\n",
    "    )\n",
    "    df_connectivity[attribute_type] = df_connectivity[attribute_type].apply(\n",
    "        lambda x: build_string(x, sep=\", \")\n",
    "    )\n",
    "    # table_data[attribute_type] = df_connectivity\n",
    "    df_connectivity = df_connectivity.set_index(\"metabolites\")\n",
    "    df_connectivity.columns = pd.MultiIndex.from_product(\n",
    "        [[model.compartments[\"c\"]], df_connectivity.columns]\n",
    "    )\n",
    "    df_connectivity_c = df_connectivity.copy()\n",
    "\n",
    "    df_metabolite_connectivity_table = pd.merge(\n",
    "        df_connectivity_c,\n",
    "        df_connectivity_e,\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "        how=\"outer\",\n",
    "    )\n",
    "\n",
    "    compartments = None\n",
    "    df_connectivity = compute_connectivities(\n",
    "        model,\n",
    "        attribute_type=attribute_type,\n",
    "        compartments=compartments,\n",
    "        exclude=exclude_list,\n",
    "        sum_compartments=sum_compartments,\n",
    "        aggregate_ranking=aggregate_ranking,\n",
    "    )\n",
    "    df_connectivity[attribute_type] = df_connectivity[attribute_type].apply(\n",
    "        lambda x: build_string(x, sep=\", \")\n",
    "    )\n",
    "    # table_data[attribute_type] = df_connectivity\n",
    "    df_connectivity = df_connectivity.set_index(\"metabolites\")\n",
    "    df_connectivity.columns = pd.MultiIndex.from_product(\n",
    "        [[\"total\"], df_connectivity.columns]\n",
    "    )\n",
    "    df_connectivity_all = df_connectivity.copy()\n",
    "\n",
    "    df_metabolite_connectivity_table = (\n",
    "        pd.merge(\n",
    "            df_connectivity_all,\n",
    "            df_metabolite_connectivity_table,\n",
    "            left_index=True,\n",
    "            right_index=True,\n",
    "            how=\"outer\",\n",
    "        )\n",
    "        .fillna(0)\n",
    "        .sort_values(by=(\"total\", \"absolute\"), ascending=False)\n",
    "    )\n",
    "    for col in df_metabolite_connectivity_table.columns:\n",
    "        if col[1] == \"absolute\":\n",
    "            df_metabolite_connectivity_table[col] = df_metabolite_connectivity_table[\n",
    "                col\n",
    "            ].astype(int)\n",
    "        else:\n",
    "            df_metabolite_connectivity_table[col] = df_metabolite_connectivity_table[\n",
    "                col\n",
    "            ].astype(float)\n",
    "\n",
    "    model_data[model.id][\n",
    "        f\"{attribute_type}-con-all\"\n",
    "    ] = df_metabolite_connectivity_table.copy()\n",
    "    print(model.id)\n",
    "    print(df_metabolite_connectivity_table.head(num_display))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3c5bfa3-984d-466b-b973-a70a4426d14c",
   "metadata": {},
   "source": [
    "##### Visualize comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0777b8-7827-4ffb-8f8a-9746215fb0ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_connectivity = model_data[prev_model_020.id][f\"{attribute_type}-con-all\"][\n",
    "    \"total\"\n",
    "].reset_index(drop=False)\n",
    "df_connectivity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32982b92-4882-4903-972c-b8cb458d004a",
   "metadata": {},
   "outputs": [],
   "source": [
    "key = \"absolute\"\n",
    "scale_points = False\n",
    "\n",
    "fig, ax = plt.subplots(1, 1, figsize=(4, 4))\n",
    "sns.despine(fig)\n",
    "\n",
    "color_list = [\"xkcd:light blue\", \"xkcd:light red\"]\n",
    "for idx, (model, color) in enumerate(zip(model_list, color_list)):\n",
    "    if idx == len(model_list) - 1:\n",
    "        refline_color = \"black\"\n",
    "    else:\n",
    "        refline_color = \"xkcd:grey\"\n",
    "    df_connectivity = model_data[model.id][f\"{attribute_type}-con-all\"][\n",
    "        \"total\"\n",
    "    ].reset_index(drop=False)\n",
    "\n",
    "    ax.scatter(\n",
    "        df_connectivity[key].index + 1,  # Log plots do not do well with index of 0.\n",
    "        df_connectivity[key].values,\n",
    "        marker=\"o\",\n",
    "        edgecolors=\"black\",\n",
    "        lw=0.5,\n",
    "        # linestyle=\"\",\n",
    "        color=color,\n",
    "        s=df_connectivity[\"relative\"].values * 10000 if scale_points else 100,\n",
    "        zorder=2,\n",
    "        label=model.id.replace(\"_\", \"-\"),\n",
    "    )\n",
    "    # A reference line, linking the minimum and maximum connectivities\n",
    "    ax.loglog(\n",
    "        [df_connectivity[key].max(), df_connectivity[key].min()],\n",
    "        [df_connectivity[key].min(), df_connectivity[key].max()],\n",
    "        zorder=1,\n",
    "        color=refline_color,\n",
    "        linestyle=\":\",\n",
    "    )\n",
    "ax.set_xscale(\"log\")\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_title(f\"Metabolite Connectivity\", fontdict={\"size\": 20}, y=1.1)\n",
    "ax.set_xlabel(\"Rank\", fontdict={\"size\": 16})\n",
    "ax.set_ylabel(\"Number of associated reactions\", fontdict={\"size\": 16})\n",
    "ax.set_xlim(\n",
    "    (0.9, df_connectivity[key].max() * 2)\n",
    ")  # Add some padding for figure clarity\n",
    "ax.set_ylim(\n",
    "    (0.9, df_connectivity[key].max() * 2)\n",
    ")  # Add some padding for figure clarity\n",
    "ax.set_xticks([1e0, 1e1, 1e2, 1e3])\n",
    "ax.set_yticks([1e0, 1e1, 1e2, 1e3])\n",
    "ax.xaxis.set_tick_params(labelsize=14)\n",
    "ax.yaxis.set_tick_params(labelsize=14)\n",
    "ax.legend(\n",
    "    frameon=False,\n",
    "    bbox_to_anchor=(0.3, 0.75),\n",
    "    loc=\"lower left\",\n",
    "    labelspacing=0.25,\n",
    "    handlelength=0.25,\n",
    "    fontsize=18,\n",
    ")\n",
    "\n",
    "mets_to_labels = {\n",
    "    # Name, xscalar, yscalar\n",
    "    \"h\": (\"H$^+$\", 1, 2.5),\n",
    "    \"na1\": (\"Na$^+$\", 0.75, 7),\n",
    "    \"atp\": (\"ATP\", 1.25, 6),\n",
    "    \"amp\": (\"AMP\", 2.5, 4),\n",
    "    \"nadp\": (\"NADP(H)\", 4.5, 10),\n",
    "    \"nad\": (\"NAD(H)\", 9, 9),\n",
    "    \"coa\": (\"CoA\", 1.8, 3),\n",
    "    \"o2\": (\"O$_2$\", 12, 4),\n",
    "    \"ala__L\": (\"ALA\", 25, 3.2),\n",
    "    \"ser__L\": (\"SER\", 23, 1.6),\n",
    "    \"cys__L\": (\"CYS\", 22, 0.8),\n",
    "    \"gln__L\": (\"GLN\", 21.5, 0.4),\n",
    "    \"gly\": (\"GLY\", 20, 0.2),\n",
    "}\n",
    "for i, row in df_connectivity[\n",
    "    df_connectivity[\"metabolites\"].isin(list(mets_to_labels))\n",
    "].iterrows():\n",
    "    x = i + 1\n",
    "    met, y = row[[\"metabolites\", \"absolute\"]].values\n",
    "    name, xscalar, yscalar = mets_to_labels[met]\n",
    "    ax.annotate(\n",
    "        xy=(x, y),\n",
    "        xytext=(x * xscalar, y * yscalar),\n",
    "        text=name,\n",
    "        xycoords=\"data\",\n",
    "        fontsize=14,\n",
    "        arrowprops=dict(arrowstyle=\"->\"),\n",
    "    )\n",
    "\n",
    "\n",
    "ax.legend(\n",
    "    frameon=False,\n",
    "    bbox_to_anchor=(-0.02, -0.05),\n",
    "    loc=\"lower left\",\n",
    "    labelspacing=0.25,\n",
    "    handlelength=0.25,\n",
    "    fontsize=14,\n",
    ")\n",
    "fig.tight_layout()\n",
    "if save_figures:\n",
    "    fig.savefig(\n",
    "        results_path\n",
    "        / f\"{attribute_type}_connectivity_{model.id}_{version}.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        dpi=None if imagetype != \"png\" else 600,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8deb19b-8fde-47bc-a2e0-35ecbcda4291",
   "metadata": {},
   "source": [
    "### Comparison of subsystems\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9641325b-f058-4cf7-9d12-032f8ae7189b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pathways = pd.read_csv(\n",
    "    ROOT_PATH / CURATION_PATH / \"subsystems.tsv\", sep=\"\\t\", index_col=None, dtype=str\n",
    ")\n",
    "df_pathways = df_pathways.fillna(\"\")\n",
    "# Group \"Metabolism of other amino acids\" with amino acids rather than treat as \"other\"\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].replace(\n",
    "    \"Metabolism of other amino acids\", \"Amino acid metabolism\"\n",
    ")\n",
    "df_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a2fd64-be06-41fc-a226-1579ddbde8fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Categories that should be excluded from the figure\n",
    "categories_to_exclude = {\"Pseudoreactions\", \"Model total\"}\n",
    "# Main categories in figure and abbreviations, all unmapped categories are mapped to \"Other\"\n",
    "categories_to_keep = {\n",
    "    \"Amino acid metabolism\": mpl.cm.spring,\n",
    "    \"Carbohydrate metabolism\": mpl.cm.Greens,\n",
    "    \"Lipid metabolism\": mpl.cm.Blues,\n",
    "    \"Metabolism of cofactors and vitamins\": mpl.cm.summer,\n",
    "    \"Nucleotide metabolism\": mpl.cm.winter,\n",
    "    \"Reactive species\": mpl.cm.Reds,\n",
    "    \"Transport reactions\": mpl.cm.Purples,\n",
    "    \"Other\": mpl.cm.gray_r,\n",
    "}\n",
    "use_abbrevs = True\n",
    "abbrevs = {\n",
    "    \"Amino acid metabolism\": \"A\",\n",
    "    \"Carbohydrate metabolism\": \"C\",\n",
    "    \"Lipid metabolism\": \"L\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"V\",\n",
    "    \"Nucleotide metabolism\": \"N\",\n",
    "    \"Reactive species\": \"R\",\n",
    "    \"Transport reactions\": \"T\",\n",
    "    \"Other\": \"O\",\n",
    "}\n",
    "barsize = 0.8\n",
    "fontsize = 12\n",
    "cmax = 0.8\n",
    "cmin = 0.3\n",
    "\n",
    "\n",
    "colormaps_normal = dict(\n",
    "    zip(\n",
    "        categories_to_keep,\n",
    "        [\n",
    "            cmap_map(lambda x: x * 1, categories_to_keep[k])(cmax)\n",
    "            for k in categories_to_keep\n",
    "        ],\n",
    "    )\n",
    ")\n",
    "colormaps_light = dict(\n",
    "    zip(\n",
    "        categories_to_keep,\n",
    "        [\n",
    "            cmap_map(lambda x: x / 2 + 0.5, categories_to_keep[k])(cmax)\n",
    "            for k in categories_to_keep\n",
    "        ],\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9837285e-3d35-4af3-ab59-b4cef9b75d95",
   "metadata": {},
   "outputs": [],
   "source": [
    "categories = [\"Removed\", \"Unchanged\", \"Added\"]\n",
    "colors_dict = dict(\n",
    "    zip(categories, [\"xkcd:light blue\", \"xkcd:purple\", \"xkcd:light red\"])\n",
    ")\n",
    "attribute_types = [\"reactions\", \"metabolites\", \"genes\"]\n",
    "edgecolor = \"black\"\n",
    "linewidth = 0.5\n",
    "data = defaultdict(dict)\n",
    "for attribute_type in attribute_types:\n",
    "    try:\n",
    "        df_removed = pd.read_csv(\n",
    "            f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_removed.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "    if attribute_type == \"reactions\":\n",
    "        obj_lists = [\n",
    "            getattr(model, attribute_type).query(lambda x: not x.boundary)\n",
    "            for model in model_list\n",
    "        ]\n",
    "    else:\n",
    "        obj_lists = [getattr(model, attribute_type) for model in model_list]\n",
    "\n",
    "    shared = obj_lists[0].query(lambda x: x.id not in df_removed[attribute_type].values)\n",
    "    removed = df_removed[attribute_type].values\n",
    "    new = obj_lists[1].query(lambda x: x.id not in set(shared.list_attr(\"id\")))\n",
    "    data[attribute_type].update(\n",
    "        dict(zip(categories, list(map(len, [removed, shared, new]))))\n",
    "    )\n",
    "\n",
    "\n",
    "df_data = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "df_data.index = df_data.index.str.capitalize()\n",
    "\n",
    "fig, ax_general = plt.subplots(figsize=(6, 2))\n",
    "total = df_data.cumsum(axis=1)\n",
    "offsets = df_data[categories[0]].values\n",
    "labels = [s.capitalize() for s in attribute_types]\n",
    "tick_pos = [0.05, 0.11, 0.17]\n",
    "for name, column in df_data.items():\n",
    "    ax_general.barh(\n",
    "        tick_pos,\n",
    "        column.values,\n",
    "        height=0.045,\n",
    "        left=total[name] - column.values - offsets,\n",
    "        color=colors_dict[name],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=linewidth,\n",
    "    )\n",
    "    ax_general.set_yticks(tick_pos, labels=attribute_types)\n",
    "    ax_general.set_yticklabels([s.capitalize() for s in attribute_types])\n",
    "\n",
    "# Remove spines\n",
    "ax_general.spines[\"right\"].set_visible(False)\n",
    "ax_general.spines[\"top\"].set_visible(False)\n",
    "ax_general.spines[\"left\"].set_visible(False)\n",
    "ax_general.axvline(0, linestyle=\":\", color=\"black\", alpha=0.8, linewidth=linewidth)\n",
    "xticks = list(np.linspace(0, 2400, 7))\n",
    "ax_general.set_xticks(xticks)\n",
    "ax_general.xaxis.set_major_formatter(lambda x, pos: int(abs(x)))\n",
    "ax_general.yaxis.set_tick_params(left=False, labelsize=\"xx-large\")\n",
    "ax_general.xaxis.set_tick_params(labelsize=\"x-large\")\n",
    "legend = ax_general.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(\n",
    "            label=label, facecolor=color, edgecolor=edgecolor, linewidth=linewidth\n",
    "        )\n",
    "        for label, color in colors_dict.items()\n",
    "    ],\n",
    "    frameon=False,\n",
    "    edgecolor=\"black\",\n",
    "    ncol=3,\n",
    "    loc=\"lower center\",\n",
    "    bbox_to_anchor=(0.4, 1, 0, 0),\n",
    "    fontsize=\"xx-large\",\n",
    "    handlelength=1,\n",
    ")\n",
    "\n",
    "fig.tight_layout()\n",
    "if save_figures:\n",
    "    fig.savefig(\n",
    "        results_path / f\"{attribute_type}_GeneralComparision.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        dpi=None if imagetype != \"png\" else 600,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eecdd25d-44ed-4a90-8a1a-ec81900f63e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\n",
    "    f\"Number of unique genes in {prev_model_020.id}:\\t\\t{len([x.id for x in prev_model_020.genes])}\\n\"\n",
    "    f\"Number of unique metabolites in {prev_model_020.id}:\\t{len(set([x.id.replace(f'_{x.compartment}', '') for x in prev_model_020.metabolites]))}\\n\"\n",
    "    f\"Number of biochemical reactions in {prev_model_020.id}:\\t{len(prev_model_020.reactions.query(lambda x: x.subsystem != 'Pseudoreactions'))}\\n\"\n",
    ")\n",
    "\n",
    "print(\n",
    "    f\"Number of unique genes in {current_model.id}:\\t\\t{len([x.id for x in current_model.genes])}\\n\"\n",
    "    f\"Number of unique metabolites in {current_model.id}:\\t{len(set([x.id.replace(f'_{x.compartment}', '') for x in current_model.metabolites]))}\\n\"\n",
    "    f\"Number of biochemical reactions in {current_model.id}:\\t{len(current_model.reactions.query(lambda x: x.subsystem != 'Pseudoreactions'))}\\n\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0c5c04b-3bcf-4b31-b6f0-e2c51c76bac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, model in enumerate(model_list):\n",
    "    data = defaultdict(dict)\n",
    "    all_reactions = set()\n",
    "    df_cat_subsystems = df_pathways.groupby(\"category\")[\"name\"].agg(lambda x: list(x))\n",
    "    for category, subsystem_list in df_cat_subsystems.items():\n",
    "        if category not in categories_to_keep and category not in categories_to_exclude:\n",
    "            category = \"Other\"\n",
    "\n",
    "        if category not in data:\n",
    "            data[category] = {\"Reactions\": set(), \"Metabolites\": set(), \"Genes\": set()}\n",
    "        # print([x for x in subsystem_list if not x in model.groups])\n",
    "        subsystem_list = [x for x in subsystem_list if x in model.groups]\n",
    "        reactions = [\n",
    "            reaction\n",
    "            for group in model.groups.get_by_any(subsystem_list)\n",
    "            for reaction in group.members\n",
    "        ]\n",
    "        for reaction in reactions:\n",
    "            all_reactions.update([reaction])\n",
    "            data[category][\"Reactions\"].update([reaction])\n",
    "            data[category][\"Metabolites\"].update(reaction.metabolites)\n",
    "            data[category][\"Genes\"].update(reaction.genes)\n",
    "\n",
    "    data[\"Model total\"] = {\n",
    "        \"Reactions\": model.reactions,\n",
    "        \"Metabolites\": model.metabolites,\n",
    "        \"Genes\": model.genes,\n",
    "    }\n",
    "    df_data = pd.DataFrame(\n",
    "        {\n",
    "            key: {k: len(x) for k, x in value_dict.items()}\n",
    "            for key, value_dict in data.items()\n",
    "        }\n",
    "    ).T.sort_index()\n",
    "    df_data = df_data.loc[df_data.index.difference(categories_to_exclude)]\n",
    "\n",
    "    model_data[model.id][f\"{attribute_type}-cat\"] = df_data\n",
    "    print(model.id, \"\\n\", df_data, \"\\n\")\n",
    "df_comparison = pd.merge(\n",
    "    model_data[model_list[0].id][f\"{attribute_type}-cat\"],\n",
    "    model_data[model_list[1].id][f\"{attribute_type}-cat\"],\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    suffixes=(f\"_{model_list[0].id}\", f\"_{model_list[1].id}\"),\n",
    ")\n",
    "df_comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3a21b75-449f-481a-8218-2e192e173522",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(list(model_data[current_model.id].keys()))\n",
    "if save_figures:\n",
    "    with pd.ExcelWriter(results_path / \"ModelComparison.xlsx\") as writer:\n",
    "        for model_name, dict_of_dfs in model_data.items():\n",
    "            for sheet_name, df in dict_of_dfs.items():\n",
    "                model_name = model_name.replace(\"_\", \"-\")\n",
    "                df.to_excel(writer, sheet_name=f\"{model_name}-{sheet_name}\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cd6067d-c70b-4c90-9ad6-c27282100129",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
