{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a1e69c41-9c8e-4c93-962b-a7981b422a94",
   "metadata": {},
   "source": [
    "# RBC-GEM 1.0.0 Updates - Initial\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b3b792e3-d4e1-4305-8aaf-1bbd08260daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from cobra.core import Reaction, Gene, Metabolite, Group\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    ROOT_PATH,\n",
    "    CURATION_PATH,\n",
    "    read_rbc_model,\n",
    "    write_rbc_model,\n",
    ")\n",
    "from rbc_gem_utils.qc import standardardize_metabolite_formulas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caf18f40-c50a-4b26-b48e-2e0db76eff68",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "778339d1-01af-4afa-a6bf-3a78c0fc2db9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>glpk</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: glpk\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c976553a-d69a-4981-a906-a9effaea0f12",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 0.3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0e4d876-7d91-483d-9f12-ca8021151cd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14939f8d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>292</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>348</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>283</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14939f8d0>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_rbc_model(filetype=\"yml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6f76c4-304f-4c19-a322-59f7cc3885e9",
   "metadata": {},
   "source": [
    "### Update model based on curation and proteomic evidence\n",
    "* Add metabolites first, then genes and lastly, reactions\n",
    "* Add metabolites with mass and charge balanced formulas (ChemAxon)\n",
    "* After adding to model, connect to MetabolicAtlas where possible and run annotation scripts.\n",
    "* Evidence tables also contain current model objects, adjust stoichiometry if necessary for mass balancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6c927168-df6b-4dfe-8470-86b5352e49fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_updated = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f3a3d99-8ba3-4b1b-9df1-3a28d7e42491",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a8fc008-7d03-43ad-8f02-067775df9e5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>name</th>\n",
       "      <th>formula</th>\n",
       "      <th>charge</th>\n",
       "      <th>compartment</th>\n",
       "      <th>metabolomic evidence (#studies)</th>\n",
       "      <th>metabolomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10fthf_c</td>\n",
       "      <td>10-formyl-tetrahydrofolate</td>\n",
       "      <td>C20H21N7O7</td>\n",
       "      <td>-2</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12ppd__R_c</td>\n",
       "      <td>(R)-propane-1,2-diol</td>\n",
       "      <td>C3H8O2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12ppd__S_c</td>\n",
       "      <td>(S)-propane-1,2-diol</td>\n",
       "      <td>C3H8O2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13dampp_c</td>\n",
       "      <td>1,3-diaminopropane</td>\n",
       "      <td>C3H12N2</td>\n",
       "      <td>2</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13dpg_c</td>\n",
       "      <td>3-Phospho-D-glyceroyl phosphate</td>\n",
       "      <td>C3H4O10P2</td>\n",
       "      <td>-4</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1986</th>\n",
       "      <td>galacglcgalacglcgal14acglcgalgluside_hs_c</td>\n",
       "      <td>iso-nLc8Cer</td>\n",
       "      <td>C73H125N4O43R</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1987</th>\n",
       "      <td>fucfucgalacglcgalacglcgal14acglcgalgluside_hs_c</td>\n",
       "      <td>G00079</td>\n",
       "      <td>C85H145N4O51R</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1988</th>\n",
       "      <td>fucgalacglcgalacglcgal14acglcgalgluside_hs_c</td>\n",
       "      <td>monofucosyllactoisooctaosylceramide</td>\n",
       "      <td>C79H135N4O47R</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1989</th>\n",
       "      <td>protein_Nwadpribarg__L_e</td>\n",
       "      <td>Nomega-(ADP-D-ribosyl)-L-arginyl-[protein]</td>\n",
       "      <td>C21H32N9O14P2</td>\n",
       "      <td>-1</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1990</th>\n",
       "      <td>protein_arg__L_e</td>\n",
       "      <td>L-arginyl-[protein]</td>\n",
       "      <td>C6H13N4OR</td>\n",
       "      <td>1</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1967 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          metabolites  \\\n",
       "0                                            10fthf_c   \n",
       "1                                          12ppd__R_c   \n",
       "2                                          12ppd__S_c   \n",
       "3                                           13dampp_c   \n",
       "4                                             13dpg_c   \n",
       "...                                               ...   \n",
       "1986        galacglcgalacglcgal14acglcgalgluside_hs_c   \n",
       "1987  fucfucgalacglcgalacglcgal14acglcgalgluside_hs_c   \n",
       "1988     fucgalacglcgalacglcgal14acglcgalgluside_hs_c   \n",
       "1989                         protein_Nwadpribarg__L_e   \n",
       "1990                                 protein_arg__L_e   \n",
       "\n",
       "                                            name        formula charge  \\\n",
       "0                     10-formyl-tetrahydrofolate     C20H21N7O7     -2   \n",
       "1                           (R)-propane-1,2-diol         C3H8O2      0   \n",
       "2                           (S)-propane-1,2-diol         C3H8O2      0   \n",
       "3                             1,3-diaminopropane        C3H12N2      2   \n",
       "4                3-Phospho-D-glyceroyl phosphate      C3H4O10P2     -4   \n",
       "...                                          ...            ...    ...   \n",
       "1986                                 iso-nLc8Cer  C73H125N4O43R      0   \n",
       "1987                                      G00079  C85H145N4O51R      0   \n",
       "1988         monofucosyllactoisooctaosylceramide  C79H135N4O47R      0   \n",
       "1989  Nomega-(ADP-D-ribosyl)-L-arginyl-[protein]  C21H32N9O14P2     -1   \n",
       "1990                        L-arginyl-[protein]       C6H13N4OR      1   \n",
       "\n",
       "     compartment metabolomic evidence (#studies)  \\\n",
       "0              c                                   \n",
       "1              c                                   \n",
       "2              c                                   \n",
       "3              c                                   \n",
       "4              c                                   \n",
       "...          ...                             ...   \n",
       "1986           c                                   \n",
       "1987           c                                   \n",
       "1988           c                                   \n",
       "1989           e                                   \n",
       "1990           e                                   \n",
       "\n",
       "     metabolomic evidence (pubmed) references notes  \n",
       "0                                                    \n",
       "1                                                    \n",
       "2                                                    \n",
       "3                                                    \n",
       "4                                                    \n",
       "...                            ...        ...   ...  \n",
       "1986                                                 \n",
       "1987                                                 \n",
       "1988                                                 \n",
       "1989                                                 \n",
       "1990                                                 \n",
       "\n",
       "[1967 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "\n",
    "df_evidence = pd.read_csv(\n",
    "    f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_evidence.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "for idx, row in df_evidence.iterrows():\n",
    "    mid, name, formula, charge, compartment = row[\n",
    "        [\"metabolites\", \"name\", \"formula\", \"charge\", \"compartment\"]\n",
    "    ]\n",
    "    if not model.metabolites.has_id(mid):\n",
    "        # Add reaction to model\n",
    "        model.add_metabolites([Metabolite(mid)])\n",
    "\n",
    "    metabolite = model.metabolites.get_by_id(mid)\n",
    "    metabolite.name = name\n",
    "    metabolite.formula = formula\n",
    "    metabolite.charge = int(charge)\n",
    "    metabolite.compartment = compartment\n",
    "\n",
    "met_formulas = standardardize_metabolite_formulas(\n",
    "    dict(zip(model.metabolites.list_attr(\"id\"), model.metabolites.list_attr(\"formula\")))\n",
    ")\n",
    "for mid, formula in met_formulas.items():\n",
    "    model.metabolites.get_by_id(mid).formula = formula\n",
    "\n",
    "dataframes_updated[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57490574-bcd3-43b3-b91c-3601e0e180c8",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9f50928d-4096-4bb7-b43e-b5bb2e22ce2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>proteomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAT</td>\n",
       "      <td>P04040</td>\n",
       "      <td>847</td>\n",
       "      <td>CAT</td>\n",
       "      <td>21</td>\n",
       "      <td>pubmed:16861337;pubmed:18399644;pubmed:1849451...</td>\n",
       "      <td>pubmed:16748387;pubmed:901548</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>P04406</td>\n",
       "      <td>2597</td>\n",
       "      <td>GAPDH</td>\n",
       "      <td>21</td>\n",
       "      <td>pubmed:16861337;pubmed:18399644;pubmed:1849451...</td>\n",
       "      <td>pubmed:19430704</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HSPA8</td>\n",
       "      <td>P11142</td>\n",
       "      <td>3312</td>\n",
       "      <td>HSPA8</td>\n",
       "      <td>20</td>\n",
       "      <td>pubmed:16861337;pubmed:18399644;pubmed:1849451...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RAB5C</td>\n",
       "      <td>P51148</td>\n",
       "      <td>5878</td>\n",
       "      <td>RAB5C</td>\n",
       "      <td>20</td>\n",
       "      <td>pubmed:16861337;pubmed:18494517;pubmed:1861456...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACLY</td>\n",
       "      <td>P53396</td>\n",
       "      <td>47</td>\n",
       "      <td>ACLY</td>\n",
       "      <td>19</td>\n",
       "      <td>pubmed:16861337;pubmed:18399644;pubmed:1861456...</td>\n",
       "      <td>pubmed:27813142</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>SCD</td>\n",
       "      <td>O00767</td>\n",
       "      <td>6319</td>\n",
       "      <td>SCD</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:33904180</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>RPS6KA4</td>\n",
       "      <td>O75676</td>\n",
       "      <td>8986</td>\n",
       "      <td>RPS6KA4</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:26611568</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>650</th>\n",
       "      <td>GCNT2</td>\n",
       "      <td>Q8N0V5</td>\n",
       "      <td>2651</td>\n",
       "      <td>GCNT2</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:29662110</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>GRIA1</td>\n",
       "      <td>P42261</td>\n",
       "      <td>2890</td>\n",
       "      <td>GRIA1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:19320779</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>652</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Q9NPC4</td>\n",
       "      <td>53947</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:12888565;pubmed:29662110;pubmed:34360826</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>653 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genes uniprot ncbigene hgnc.symbol proteomic evidence (#studies)  \\\n",
       "0        CAT  P04040      847         CAT                            21   \n",
       "1      GAPDH  P04406     2597       GAPDH                            21   \n",
       "2      HSPA8  P11142     3312       HSPA8                            20   \n",
       "3      RAB5C  P51148     5878       RAB5C                            20   \n",
       "4       ACLY  P53396       47        ACLY                            19   \n",
       "..       ...     ...      ...         ...                           ...   \n",
       "648      SCD  O00767     6319         SCD                             0   \n",
       "649  RPS6KA4  O75676     8986     RPS6KA4                             0   \n",
       "650    GCNT2  Q8N0V5     2651       GCNT2                             0   \n",
       "651    GRIA1  P42261     2890       GRIA1                             0   \n",
       "652   A4GALT  Q9NPC4    53947      A4GALT                             0   \n",
       "\n",
       "                           proteomic evidence (pubmed)  \\\n",
       "0    pubmed:16861337;pubmed:18399644;pubmed:1849451...   \n",
       "1    pubmed:16861337;pubmed:18399644;pubmed:1849451...   \n",
       "2    pubmed:16861337;pubmed:18399644;pubmed:1849451...   \n",
       "3    pubmed:16861337;pubmed:18494517;pubmed:1861456...   \n",
       "4    pubmed:16861337;pubmed:18399644;pubmed:1861456...   \n",
       "..                                                 ...   \n",
       "648                                                      \n",
       "649                                                      \n",
       "650                                                      \n",
       "651                                                      \n",
       "652                                                      \n",
       "\n",
       "                                          references notes  \n",
       "0                      pubmed:16748387;pubmed:901548        \n",
       "1                                    pubmed:19430704        \n",
       "2                                                           \n",
       "3                                                           \n",
       "4                                    pubmed:27813142        \n",
       "..                                               ...   ...  \n",
       "648                                  pubmed:33904180        \n",
       "649                                  pubmed:26611568        \n",
       "650                                  pubmed:29662110        \n",
       "651                                  pubmed:19320779        \n",
       "652  pubmed:12888565;pubmed:29662110;pubmed:34360826        \n",
       "\n",
       "[653 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "\n",
    "df_evidence = pd.read_csv(\n",
    "    f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_evidence.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "\n",
    "for idx, row in df_evidence.iterrows():\n",
    "    # Between these three annotation fields, should be likely that the rest can get extracted from UniProt.\n",
    "    gid = row[\"genes\"]\n",
    "    if not model.genes.has_id(gid):\n",
    "        gene = Gene(gid)\n",
    "        model.genes.extend([gene])\n",
    "    gene = model.genes.get_by_id(gid)\n",
    "    for key in [\"uniprot\", \"ncbigene\", \"hgnc.symbol\"]:\n",
    "        item = row[key]\n",
    "        if item:\n",
    "            gene.annotation[key] = item\n",
    "\n",
    "dataframes_updated[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b179ad6-3125-47c8-8fc9-d6f82952fb3c",
   "metadata": {},
   "source": [
    "#### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "12401ee3-4a2e-45c5-94c8-81af0003b3c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>reaction</th>\n",
       "      <th>gene reaction rule</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>spontaneous</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FACOAE_10_0</td>\n",
       "      <td>Acyl CoA thioesterase (CoA 10:0)</td>\n",
       "      <td>FAcoa_hs_10_0_c + h2o_c --&gt; FA_hs_10_0_c + coa...</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>Acyl-CoA hydrolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>ACOT7 (11)</td>\n",
       "      <td>pubmed:10578051;pubmed:5042454</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FACOAE_11_0</td>\n",
       "      <td>Acyl CoA thioesterase (CoA 11:0)</td>\n",
       "      <td>FAcoa_hs_11_0_c + h2o_c --&gt; FA_hs_11_0_c + coa...</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>Acyl-CoA hydrolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>ACOT7 (11)</td>\n",
       "      <td>pubmed:10578051;pubmed:5042454</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FACOAE_12_0</td>\n",
       "      <td>Acyl CoA thioesterase (CoA 12:0)</td>\n",
       "      <td>FAcoa_hs_12_0_c + h2o_c --&gt; FA_hs_12_0_c + coa...</td>\n",
       "      <td>ACOT7 or MBLAC2</td>\n",
       "      <td>Acyl-CoA hydrolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>ACOT7 (11);MBLAC2 (9)</td>\n",
       "      <td>pubmed:10578051;pubmed:5042454</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FACOAE_13_0</td>\n",
       "      <td>Acyl CoA thioesterase (CoA 13:0)</td>\n",
       "      <td>FAcoa_hs_13_0_c + h2o_c --&gt; FA_hs_13_0_c + coa...</td>\n",
       "      <td>ACOT7 or MBLAC2</td>\n",
       "      <td>Acyl-CoA hydrolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>ACOT7 (11);MBLAC2 (9)</td>\n",
       "      <td>pubmed:10578051;pubmed:5042454</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FACOAE_14_0</td>\n",
       "      <td>Acyl CoA thioesterase (CoA 14:0)</td>\n",
       "      <td>FAcoa_hs_14_0_c + h2o_c --&gt; FA_hs_14_0_c + coa...</td>\n",
       "      <td>ACOT7 or MBLAC2</td>\n",
       "      <td>Acyl-CoA hydrolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>ACOT7 (11);MBLAC2 (9)</td>\n",
       "      <td>pubmed:10578051;pubmed:5042454</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2376</th>\n",
       "      <td>PYDXPP</td>\n",
       "      <td>Pyridoxal 5-phosphate phosphatase</td>\n",
       "      <td>h2o_c + pydx5p_c --&gt; pi_c + pydx_c</td>\n",
       "      <td>PDXP</td>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>PDXP (16)</td>\n",
       "      <td>pubmed:1322411;pubmed:14522954;pubmed:7448195;...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2377</th>\n",
       "      <td>AVITE1RASCBR1</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via L-as...</td>\n",
       "      <td>ascb__L_c + avite1r_c &lt;=&gt; avite1_c + mdhdascb_...</td>\n",
       "      <td></td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:18243141;pubmed:36643550;pubmed:8388377...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>AVITE1RCYTB5R</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via L-as...</td>\n",
       "      <td>avite1r_c + focytb5_c + h_c --&gt; avite1_c + fic...</td>\n",
       "      <td>CYB5R3 or (CYB5R3 and CYB5B) or (CYB5R3 and CY...</td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>CYB5A (8);CYB5B (8);CYB5R3 (19)</td>\n",
       "      <td>pubmed:8388377</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379</th>\n",
       "      <td>AVITE1RQ10H2R</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via ubiq...</td>\n",
       "      <td>avite1r_c + q10h2_c --&gt; avite1_c + q10h_c</td>\n",
       "      <td></td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:8388377;pubmed:8434935</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380</th>\n",
       "      <td>AVITE1RQ10HR</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via ubiq...</td>\n",
       "      <td>avite1r_c + q10h_c --&gt; avite1_c + q10_c</td>\n",
       "      <td></td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:8388377;pubmed:8434935</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2381 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          reactions                                               name  \\\n",
       "0       FACOAE_10_0                   Acyl CoA thioesterase (CoA 10:0)   \n",
       "1       FACOAE_11_0                   Acyl CoA thioesterase (CoA 11:0)   \n",
       "2       FACOAE_12_0                   Acyl CoA thioesterase (CoA 12:0)   \n",
       "3       FACOAE_13_0                   Acyl CoA thioesterase (CoA 13:0)   \n",
       "4       FACOAE_14_0                   Acyl CoA thioesterase (CoA 14:0)   \n",
       "...             ...                                                ...   \n",
       "2376         PYDXPP                  Pyridoxal 5-phosphate phosphatase   \n",
       "2377  AVITE1RASCBR1  Reduction of alpha-tocopherol radical via L-as...   \n",
       "2378  AVITE1RCYTB5R  Reduction of alpha-tocopherol radical via L-as...   \n",
       "2379  AVITE1RQ10H2R  Reduction of alpha-tocopherol radical via ubiq...   \n",
       "2380   AVITE1RQ10HR  Reduction of alpha-tocopherol radical via ubiq...   \n",
       "\n",
       "                                               reaction  \\\n",
       "0     FAcoa_hs_10_0_c + h2o_c --> FA_hs_10_0_c + coa...   \n",
       "1     FAcoa_hs_11_0_c + h2o_c --> FA_hs_11_0_c + coa...   \n",
       "2     FAcoa_hs_12_0_c + h2o_c --> FA_hs_12_0_c + coa...   \n",
       "3     FAcoa_hs_13_0_c + h2o_c --> FA_hs_13_0_c + coa...   \n",
       "4     FAcoa_hs_14_0_c + h2o_c --> FA_hs_14_0_c + coa...   \n",
       "...                                                 ...   \n",
       "2376                 h2o_c + pydx5p_c --> pi_c + pydx_c   \n",
       "2377  ascb__L_c + avite1r_c <=> avite1_c + mdhdascb_...   \n",
       "2378  avite1r_c + focytb5_c + h_c --> avite1_c + fic...   \n",
       "2379          avite1r_c + q10h2_c --> avite1_c + q10h_c   \n",
       "2380            avite1r_c + q10h_c --> avite1_c + q10_c   \n",
       "\n",
       "                                     gene reaction rule  \\\n",
       "0                                                 ACOT7   \n",
       "1                                                 ACOT7   \n",
       "2                                       ACOT7 or MBLAC2   \n",
       "3                                       ACOT7 or MBLAC2   \n",
       "4                                       ACOT7 or MBLAC2   \n",
       "...                                                 ...   \n",
       "2376                                               PDXP   \n",
       "2377                                                      \n",
       "2378  CYB5R3 or (CYB5R3 and CYB5B) or (CYB5R3 and CY...   \n",
       "2379                                                      \n",
       "2380                                                      \n",
       "\n",
       "                  subsystem spontaneous    proteomic evidence (#studies)  \\\n",
       "0       Acyl-CoA hydrolysis           0                       ACOT7 (11)   \n",
       "1       Acyl-CoA hydrolysis           0                       ACOT7 (11)   \n",
       "2       Acyl-CoA hydrolysis           0            ACOT7 (11);MBLAC2 (9)   \n",
       "3       Acyl-CoA hydrolysis           0            ACOT7 (11);MBLAC2 (9)   \n",
       "4       Acyl-CoA hydrolysis           0            ACOT7 (11);MBLAC2 (9)   \n",
       "...                     ...         ...                              ...   \n",
       "2376  Vitamin B6 metabolism           0                        PDXP (16)   \n",
       "2377   Vitamin E metabolism           1                                    \n",
       "2378   Vitamin E metabolism           0  CYB5A (8);CYB5B (8);CYB5R3 (19)   \n",
       "2379   Vitamin E metabolism           1                                    \n",
       "2380   Vitamin E metabolism           1                                    \n",
       "\n",
       "                                             references notes  \n",
       "0                        pubmed:10578051;pubmed:5042454        \n",
       "1                        pubmed:10578051;pubmed:5042454        \n",
       "2                        pubmed:10578051;pubmed:5042454        \n",
       "3                        pubmed:10578051;pubmed:5042454        \n",
       "4                        pubmed:10578051;pubmed:5042454        \n",
       "...                                                 ...   ...  \n",
       "2376  pubmed:1322411;pubmed:14522954;pubmed:7448195;...        \n",
       "2377  pubmed:18243141;pubmed:36643550;pubmed:8388377...        \n",
       "2378                                     pubmed:8388377        \n",
       "2379                      pubmed:8388377;pubmed:8434935        \n",
       "2380                      pubmed:8388377;pubmed:8434935        \n",
       "\n",
       "[2381 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "\n",
    "df_evidence = pd.read_csv(\n",
    "    f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_evidence.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "for idx, row in df_evidence.iterrows():\n",
    "    rid, name, reaction_str, gpr, subsystem = row[\n",
    "        [\"reactions\", \"name\", \"reaction\", \"gene reaction rule\", \"subsystem\"]\n",
    "    ]\n",
    "    if not model.reactions.has_id(rid):\n",
    "        # Add reaction to model\n",
    "        model.add_reactions([Reaction(rid)])\n",
    "\n",
    "    reaction = model.reactions.get_by_id(rid)\n",
    "    reaction.build_reaction_from_string(reaction_str)\n",
    "    reaction.name = name\n",
    "    reaction.gene_reaction_rule = gpr\n",
    "    reaction.subsystem = subsystem\n",
    "\n",
    "dataframes_updated[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1185f5fb-9f60-4c8d-8c61-2480f0e7a90c",
   "metadata": {},
   "source": [
    "### Remove from model\n",
    "Remove reactions first, then genes, and lastly metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a407a3bd-ebd0-4c72-bd98-66cb0eb14eec",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_removed = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56d67737-ef34-4a18-adff-5d9c4b584320",
   "metadata": {},
   "source": [
    "#### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dcb9e282-0d2f-4dc2-b23a-e04dda2a4de8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3MOXTYRESSte</td>\n",
       "      <td>3-Methoxytyramine secretion via secretory vesi...</td>\n",
       "      <td>Replaced by simple transport reaction (3MOXTYR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALAt4</td>\n",
       "      <td>Alanine-Sodium symporter</td>\n",
       "      <td>Replaced by reactions involving the amino acid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BANDMT</td>\n",
       "      <td>Band membrane protein-methyltransferase</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BILGLCURt</td>\n",
       "      <td>Bilirubin monoglucuronide transport via bicarb...</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BILIRBU</td>\n",
       "      <td>Bilirubin UDP-glucuronosyltransferase</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>THMMPtrbc</td>\n",
       "      <td>Thiamine monophosphate transport (passive - re...</td>\n",
       "      <td>Replaced by reactions involving reduced folate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>TMDPPK</td>\n",
       "      <td>Thiamine diphosphate kinase</td>\n",
       "      <td>Replaced by ADK7. Adenylate kinase isozyme 1 h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176</th>\n",
       "      <td>UDPGD</td>\n",
       "      <td>UDPglucose 6-dehydrogenase</td>\n",
       "      <td>No proteomic or activity evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>UDPGNP</td>\n",
       "      <td>UDP-glucuronate glucuronohydrolase</td>\n",
       "      <td>No proteomic or activity evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>SBTR</td>\n",
       "      <td>D sorbitol reductase</td>\n",
       "      <td>Replaced by SBTRa. KEGG indicates alpha glucos...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        reactions                                               name  \\\n",
       "0    3MOXTYRESSte  3-Methoxytyramine secretion via secretory vesi...   \n",
       "1           ALAt4                           Alanine-Sodium symporter   \n",
       "2          BANDMT            Band membrane protein-methyltransferase   \n",
       "3       BILGLCURt  Bilirubin monoglucuronide transport via bicarb...   \n",
       "4         BILIRBU              Bilirubin UDP-glucuronosyltransferase   \n",
       "..            ...                                                ...   \n",
       "174     THMMPtrbc  Thiamine monophosphate transport (passive - re...   \n",
       "175        TMDPPK                        Thiamine diphosphate kinase   \n",
       "176         UDPGD                         UDPglucose 6-dehydrogenase   \n",
       "177        UDPGNP                 UDP-glucuronate glucuronohydrolase   \n",
       "178          SBTR                               D sorbitol reductase   \n",
       "\n",
       "                                                 notes  \n",
       "0    Replaced by simple transport reaction (3MOXTYR...  \n",
       "1    Replaced by reactions involving the amino acid...  \n",
       "2    Replaced by generic methylation reactions. Ban...  \n",
       "3    Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "4    Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "..                                                 ...  \n",
       "174  Replaced by reactions involving reduced folate...  \n",
       "175  Replaced by ADK7. Adenylate kinase isozyme 1 h...  \n",
       "176                  No proteomic or activity evidence  \n",
       "177                  No proteomic or activity evidence  \n",
       "178  Replaced by SBTRa. KEGG indicates alpha glucos...  \n",
       "\n",
       "[175 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=0,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for reaction in df_removed[attribute_type]:\n",
    "    try:\n",
    "        reaction = getattr(model, attribute_type).get_by_id(reaction)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(reaction)\n",
    "\n",
    "model.remove_reactions(to_remove)\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8d0eca-5c50-4679-bf22-fb43535ccccf",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7a0a09ff-04bf-4f3f-8db6-00da29e45e77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADCY1</td>\n",
       "      <td>Q08828</td>\n",
       "      <td>ADCY1</td>\n",
       "      <td>107</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADCY3</td>\n",
       "      <td>O60266</td>\n",
       "      <td>ADCY3</td>\n",
       "      <td>109</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADCY4</td>\n",
       "      <td>Q8NFM4</td>\n",
       "      <td>ADCY4</td>\n",
       "      <td>196883</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADCY5</td>\n",
       "      <td>O95622</td>\n",
       "      <td>ADCY5</td>\n",
       "      <td>111</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADCY6</td>\n",
       "      <td>O43306</td>\n",
       "      <td>ADCY6</td>\n",
       "      <td>112</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>TAT</td>\n",
       "      <td>P17735</td>\n",
       "      <td>TAT</td>\n",
       "      <td>6898</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>UGDH</td>\n",
       "      <td>O60701</td>\n",
       "      <td>UGDH</td>\n",
       "      <td>7358</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>P22309</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>54658</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>P22310</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>54657</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>ZNF384</td>\n",
       "      <td>Q8TF68</td>\n",
       "      <td>ZNF384</td>\n",
       "      <td>171017</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes uniprot hgnc.symbol ncbigene  \\\n",
       "0     ADCY1  Q08828       ADCY1      107   \n",
       "1     ADCY3  O60266       ADCY3      109   \n",
       "2     ADCY4  Q8NFM4       ADCY4   196883   \n",
       "3     ADCY5  O95622       ADCY5      111   \n",
       "4     ADCY6  O43306       ADCY6      112   \n",
       "..      ...     ...         ...      ...   \n",
       "119     TAT  P17735         TAT     6898   \n",
       "120    UGDH  O60701        UGDH     7358   \n",
       "121  UGT1A1  P22309      UGT1A1    54658   \n",
       "122  UGT1A4  P22310      UGT1A4    54657   \n",
       "123  ZNF384  Q8TF68      ZNF384   171017   \n",
       "\n",
       "                                                 notes  \n",
       "0    no proteomic evidence. evidence indicates othe...  \n",
       "1    no proteomic evidence. evidence indicates othe...  \n",
       "2    no proteomic evidence. evidence indicates othe...  \n",
       "3    no proteomic evidence. evidence indicates othe...  \n",
       "4    no proteomic evidence. evidence indicates othe...  \n",
       "..                                                 ...  \n",
       "119  no proteomic evidence. evidence indicates othe...  \n",
       "120                             no proteomic evidence.  \n",
       "121                             no proteomic evidence.  \n",
       "122                             no proteomic evidence.  \n",
       "123  no proteomic evidence. evidence indicates othe...  \n",
       "\n",
       "[114 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=0,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for gene in df_removed[attribute_type]:\n",
    "    try:\n",
    "        gene = model.genes.get_by_id(gene)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(gene)\n",
    "model.genes -= to_remove\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "\n",
    "for gene in model.genes:\n",
    "    if gene.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {gene}\")\n",
    "\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad923fa-b442-4667-93f5-971a90cb1318",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e987854a-c0db-4e4b-9a70-aa251551490e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>band_c</td>\n",
       "      <td>Band membrane protein (universal, erythrocyte ...</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bandmt_c</td>\n",
       "      <td>Band membrane protein (methylated, universal, ...</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bilglcur_c</td>\n",
       "      <td>Bilirubin monoglucuronide</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bilglcur_e</td>\n",
       "      <td>Bilirubin monoglucuronide</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cdpdag_hs_16_0_16_0_c</td>\n",
       "      <td>CDP diacylglycerol (homo sapiens, C16:0, C16:0)</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>pe_hs_18_9Z_18_9Z12Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:1,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>pe_hs_18_9Z12Z_16_0_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:2,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>pe_hs_18_9Z12Z_18_9Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:2,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>pppg9_c</td>\n",
       "      <td>protoporphyrinogen IX</td>\n",
       "      <td>Protoporphyrinogen oxidase is a mitochondrial ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>udpglcur_c</td>\n",
       "      <td>UDP-D-glucuronate</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               metabolites                                               name  \\\n",
       "0                   band_c  Band membrane protein (universal, erythrocyte ...   \n",
       "1                 bandmt_c  Band membrane protein (methylated, universal, ...   \n",
       "2               bilglcur_c                          Bilirubin monoglucuronide   \n",
       "3               bilglcur_e                          Bilirubin monoglucuronide   \n",
       "4    cdpdag_hs_16_0_16_0_c    CDP diacylglycerol (homo sapiens, C16:0, C16:0)   \n",
       "..                     ...                                                ...   \n",
       "65  pe_hs_18_9Z_18_9Z12Z_c  Phosphatidylethanolamine (homo sapiens, C18:1,...   \n",
       "66   pe_hs_18_9Z12Z_16_0_c  Phosphatidylethanolamine (homo sapiens, C18:2,...   \n",
       "67  pe_hs_18_9Z12Z_18_9Z_c  Phosphatidylethanolamine (homo sapiens, C18:2,...   \n",
       "68                 pppg9_c                              protoporphyrinogen IX   \n",
       "69              udpglcur_c                                  UDP-D-glucuronate   \n",
       "\n",
       "                                                notes  \n",
       "0   Replaced by generic methylation reactions. Ban...  \n",
       "1   Replaced by generic methylation reactions. Ban...  \n",
       "2   Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "3   Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "4   Lipids are pooled, replaced with pooled versio...  \n",
       "..                                                ...  \n",
       "65  Lipids are pooled, replaced with pooled versio...  \n",
       "66  Lipids are pooled, replaced with pooled versio...  \n",
       "67  Lipids are pooled, replaced with pooled versio...  \n",
       "68  Protoporphyrinogen oxidase is a mitochondrial ...  \n",
       "69  Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "\n",
       "[70 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        f\"{ROOT_PATH}{CURATION_PATH}/{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=0,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for metabolite in df_removed[attribute_type]:\n",
    "    try:\n",
    "        metabolite = model.metabolites.get_by_id(metabolite)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    model.remove_metabolites([metabolite])\n",
    "model.remove_metabolites(to_remove)\n",
    "\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {metabolite}\")\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbd78766-e03a-4fc2-8215-32d696fa4b1c",
   "metadata": {},
   "source": [
    "#### Add exchanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eb647fa6-94e1-4d85-8353-2f6eef374b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in model.metabolites.query(lambda x: x.compartment == \"e\"):\n",
    "    try:\n",
    "        model.add_boundary(met, type=\"exchange\")\n",
    "    except ValueError:\n",
    "        pass\n",
    "for reaction in model.boundary:\n",
    "    reaction.subsystem = \"Pseudoreactions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b93b32-2bc9-44b1-b896-32460fa3c53c",
   "metadata": {},
   "source": [
    "#### Add pooled lipid reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6babdc32-7ac2-4132-8ba1-97418e4d299b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions(\n",
    "    [\n",
    "        Reaction(\n",
    "            \"POOL_FACOA\",\n",
    "            name=\"Pooling reaction Acyl-CoA (CoA)\",\n",
    "            subsystem=\"Pseudoreactions\",\n",
    "            lower_bound=COBRA_CONFIGURATION.lower_bound,\n",
    "            upper_bound=COBRA_CONFIGURATION.upper_bound,\n",
    "        ),\n",
    "        Reaction(\n",
    "            \"POOL_FA\",\n",
    "            name=\"Pooling reaction fatty acids (FA)\",\n",
    "            subsystem=\"Pseudoreactions\",\n",
    "            lower_bound=COBRA_CONFIGURATION.lower_bound,\n",
    "            upper_bound=COBRA_CONFIGURATION.upper_bound,\n",
    "        ),\n",
    "    ]\n",
    ")\n",
    "model.reactions.get_by_id(\"POOL_FACOA\").add_metabolites(\n",
    "    {\n",
    "        \"FAcoa_hs_c\": -1,\n",
    "        \"FAcoa_hs_12_0_c\": 0.0004,  # lauroyl-CoA\n",
    "        \"FAcoa_hs_13_0_c\": 0.0004,  # tridecanoyl-CoA\n",
    "        \"FAcoa_hs_14_0_c\": 0.0133,  # myristoyl-CoA\n",
    "        \"FAcoa_hs_14_5Z_c\": 0.0004,  # (5Z)-tetradecanoyl-CoA\n",
    "        \"FAcoa_hs_14_7Z_c\": 0.0004,  # (7Z)-tetradecenoyl-CoA\n",
    "        \"FAcoa_hs_14_9Z_c\": 0.0004,  # (9Z)-tetradecenoyl-CoA\n",
    "        \"FAcoa_hs_15_0_c\": 0.0004,  # Pentadecanoyl-CoA\n",
    "        \"FAcoa_hs_16_0_c\": 0.2220,  # Palmitoyl-CoA\n",
    "        \"FAcoa_hs_16_7Z_c\": 0.0004,  # (7Z)-palmitoleoyl-CoA\n",
    "        \"FAcoa_hs_16_9Z_c\": 0.0219,  # Palmitoleoyl-CoA\n",
    "        \"FAcoa_hs_17_0_c\": 0.0004,  # margaric acid\n",
    "        \"FAcoa_hs_17_10Z_c\": 0.0004,  #  (10Z)-heptadecylenic acid\n",
    "        \"FAcoa_hs_17_9Z_c\": 0.0004,  #  (9Z)-heptadecylenic acid\n",
    "        \"FAcoa_hs_18_0_c\": 0.1498,  # stearoyl-CoA\n",
    "        \"FAcoa_hs_18_9Z_c\": 0.1545,  # oleoyl-CoA\n",
    "        \"FAcoa_hs_18_11Z_c\": 0.0250,  # cis-vaccenoyl-CoA\n",
    "        \"FAcoa_hs_18_13Z_c\": 0.0004,  # (13Z)-octadecenoyl-CoA\n",
    "        \"FAcoa_hs_18_6Z9Z12Z15Z_c\": 0.0025,  # stearoyl-CoA\n",
    "        \"FAcoa_hs_18_6Z9Z12Z_c\": 0.0029,  # gamma-linolenoyl-CoA\n",
    "        \"FAcoa_hs_18_6Z9Z_c\": 0.0004,  # (6Z,9Z)-octadecadienoyl-CoA\n",
    "        \"FAcoa_hs_18_7Z_c\": 0.0004,  # (7Z)-octadecenoyl-CoA\n",
    "        \"FAcoa_hs_18_9E_c\": 0.0004,  # (9E)-octadecenoyl-CoA\n",
    "        \"FAcoa_hs_18_9Z12Z_c\": 0.1915,  # Linoleoyl-CoA\n",
    "        \"FAcoa_hs_18_9Z12Z15Z_c\": 0.0084,  # alpha-linolenoyl-CoA\n",
    "        \"FAcoa_hs_19_0_c\": 0.0004,  # nonadecanoyl-CoA\n",
    "        \"FAcoa_hs_20_0_c\": 0.0004,  # Arachidoyl-CoA /eicosanoyl-CoA\n",
    "        \"FAcoa_hs_20_11Z14Z17Z_c\": 0.0215,  # (11Z,14Z,17Z)-eicosatrienoyl-CoA\n",
    "        \"FAcoa_hs_20_11Z14Z_c\": 0.0004,  # (11Z,14Z)-eicosadienoyl-CoA\n",
    "        \"FAcoa_hs_20_11Z_c\": 0.0004,  # (11Z)-eicosenoyl-CoA\n",
    "        \"FAcoa_hs_20_13Z_c\": 0.0004,  # (13Z)-eicosenoyl-CoA\n",
    "        \"FAcoa_hs_20_5Z8Z11Z14Z17Z_c\": 0.0116,  # eicosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_20_5Z8Z11Z14Z_c\": 0.1083,  # arachidonoyl-CoA\n",
    "        \"FAcoa_hs_20_5Z8Z11Z_c\": 0.0004,  # (5Z,8Z,11Z)-eicosatrienoyl-CoA\n",
    "        \"FAcoa_hs_20_8Z11Z14Z17Z_c\": 0.0115,  # (8Z,11Z,14Z,17Z)-eicosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_20_8Z11Z14Z_c\": 0.0215,  #  dihomo-gamma-linoleoyl-CoA\n",
    "        \"FAcoa_hs_20_8Z11Z_c\": 0.0004,  # 8,11-eicosadienoyl-CoA\n",
    "        \"FAcoa_hs_20_9Z_c\": 0.0004,  # 9-eicosenoyl-CoA\n",
    "        \"FAcoa_hs_21_0_c\": 0.0004,  # henicosanoyl-CoA\n",
    "        \"FAcoa_hs_22_0_c\": 0.0004,  # docosanoyl-CoA\n",
    "        \"FAcoa_hs_22_10Z13Z16Z19Z_c\": 0.0004,  # 10,13,16,19-docosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_22_10Z13Z16Z_c\": 0.0004,  # 10,13,16-docosatriynoyl-CoA\n",
    "        \"FAcoa_hs_22_11Z_c\": 0.0004,  # (11Z)-docosenoyl-CoA\n",
    "        \"FAcoa_hs_22_13Z16Z19Z_c\": 0.0004,  # 13,16,19-docosatrienoyl-CoA\n",
    "        \"FAcoa_hs_22_13Z16Z_c\": 0.0004,  # (13Z,16Z)-docosadienoyl-CoA\n",
    "        \"FAcoa_hs_22_13Z_c\": 0.0004,  # (13Z)-docosenoyl-CoA\n",
    "        \"FAcoa_hs_22_4Z7Z10Z13Z16Z19Z_c\": 0.0278,  # DHA\n",
    "        \"FAcoa_hs_22_4Z7Z10Z13Z16Z_c\": 0.0038,  #  (4Z,7Z,10Z,13Z,16Z)-docosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_22_7Z10Z13Z16Z19Z_c\": 0.0059,  # (7Z,10Z,13Z,16Z,19Z)-docosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_22_7Z10Z13Z16Z_c\": 0.0014,  #  (7Z,10Z,13Z,16Z)-docosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_23_0_c\": 0.0004,  # tricosanoyl-CoA\n",
    "        \"FAcoa_hs_24_0_c\": 0.0004,  # tetracosanoyl-CoA\n",
    "        \"FAcoa_hs_24_12Z15Z18Z21Z_c\": 0.0004,  # 12,15,18,21-tetracosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_24_15Z_c\": 0.0004,  # (15Z)-tetracosenoyl-CoA\n",
    "        \"FAcoa_hs_24_6Z9Z12Z15Z18Z21Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z,21Z)-tetracosahexaenoyl-CoA\n",
    "        \"FAcoa_hs_24_6Z9Z12Z15Z18Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z)-tetracosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_24_9Z12Z15Z18Z21Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z,21Z)-tetracosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_24_9Z12Z15Z18Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z)-tetracosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_26_0_c\": 0.0004,  # hexacosanoyl-CoA\n",
    "        \"FAcoa_hs_26_17Z_c\": 0.0004,  #  hexacosenoyl-CoA\n",
    "    }\n",
    ")\n",
    "model.reactions.get_by_id(\"POOL_FA\").add_metabolites(\n",
    "    {\n",
    "        \"FA_hs_c\": -1,\n",
    "        \"FA_hs_12_0_c\": 0.0004,  # lauric acid\n",
    "        \"FA_hs_13_0_c\": 0.0004,  # tridecylic acid\n",
    "        \"FA_hs_14_0_c\": 0.0133,  # myristic acid\n",
    "        \"FA_hs_14_5Z_c\": 0.0004,  # physeteric acid\n",
    "        \"FA_hs_14_7Z_c\": 0.0004,  # 7Z-tetradecenoic acid\n",
    "        \"FA_hs_14_9Z_c\": 0.0004,  # 9Z-tetradecenoic acid\n",
    "        \"FA_hs_15_0_c\": 0.0004,  # pentadecylic acid\n",
    "        \"FA_hs_16_0_c\": 0.2220,  # Palmitic acid\n",
    "        \"FA_hs_16_7Z_c\": 0.0004,  # 7-palmitoleic acid\n",
    "        \"FA_hs_16_9Z_c\": 0.0219,  # Palmitoleic acid\n",
    "        \"FA_hs_17_0_c\": 0.0004,  # margaric acid\n",
    "        \"FA_hs_17_10Z_c\": 0.0004,  #  10-heptadecylenic acid\n",
    "        \"FA_hs_17_9Z_c\": 0.0004,  #  9-heptadecylenic acid\n",
    "        \"FA_hs_18_0_c\": 0.1498,  # stearate acid\n",
    "        \"FA_hs_18_9Z_c\": 0.1545,  # oleic acid\n",
    "        \"FA_hs_18_11Z_c\": 0.0250,  # cis-vaccenic acid\n",
    "        \"FA_hs_18_13Z_c\": 0.0004,  # (13Z)-octadecenoic acid\n",
    "        \"FA_hs_18_6Z9Z12Z15Z_c\": 0.0025,  # stearidonic acid\n",
    "        \"FA_hs_18_6Z9Z12Z_c\": 0.0029,  # gamma-linolenic acid\n",
    "        \"FA_hs_18_6Z9Z_c\": 0.0004,  # (6Z,9Z)-octadecadienoic acid\n",
    "        \"FA_hs_18_7Z_c\": 0.0004,  # (7Z)-octadecenoic acid\n",
    "        \"FA_hs_18_9E_c\": 0.0004,  # elaidate acid\n",
    "        \"FA_hs_18_9Z12Z_c\": 0.1915,  # Linoleic acid\n",
    "        \"FA_hs_18_9Z12Z15Z_c\": 0.0084,  # alpha-linolenic acid\n",
    "        \"FA_hs_19_0_c\": 0.0004,  # nonadecylic acid\n",
    "        \"FA_hs_20_0_c\": 0.0004,  # Arachidic acid /eicosanoic acid\n",
    "        \"FA_hs_20_11Z14Z17Z_c\": 0.0215,  # (11Z,14Z,17Z)-eicosatrienoic acid\n",
    "        \"FA_hs_20_11Z14Z_c\": 0.0004,  # (11Z,14Z)-eicosadienoic acid\n",
    "        \"FA_hs_20_11Z_c\": 0.0004,  # cis-gondoic acid\n",
    "        \"FA_hs_20_13Z_c\": 0.0004,  # (13Z)-eicosenoic acid\n",
    "        \"FA_hs_20_5Z8Z11Z14Z17Z_c\": 0.0116,  # EPA\n",
    "        \"FA_hs_20_5Z8Z11Z14Z_c\": 0.1083,  # arachidonoic acid\n",
    "        \"FA_hs_20_5Z8Z11Z_c\": 0.0004,  # mead acid\n",
    "        \"FA_hs_20_8Z11Z14Z17Z_c\": 0.0115,  # omega-3-arachidonic acid\n",
    "        \"FA_hs_20_8Z11Z14Z_c\": 0.0215,  #  dihomo-gamma-linoleic acid\n",
    "        \"FA_hs_20_8Z11Z_c\": 0.0004,  # 8,11-eicosadienoic acid\n",
    "        \"FA_hs_20_9Z_c\": 0.0004,  # 9-eicosenoic acid\n",
    "        \"FA_hs_21_0_c\": 0.0004,  # henicosanoic acid\n",
    "        \"FA_hs_22_0_c\": 0.0004,  # behenic acid\n",
    "        \"FA_hs_22_10Z13Z16Z19Z_c\": 0.0004,  # 10,13,16,19-docosatetraenoic acid\n",
    "        \"FA_hs_22_10Z13Z16Z_c\": 0.0004,  # 10,13,16-docosatriynoic acid\n",
    "        \"FA_hs_22_11Z_c\": 0.0004,  # cis-cetoleic acid\n",
    "        \"FA_hs_22_13Z16Z19Z_c\": 0.0004,  # 13,16,19-docosatrienoic acid\n",
    "        \"FA_hs_22_13Z16Z_c\": 0.0004,  # (13Z,16Z)-docosadienoic acid\n",
    "        \"FA_hs_22_13Z_c\": 0.0004,  # cis-erucic acid\n",
    "        \"FA_hs_22_4Z7Z10Z13Z16Z19Z_c\": 0.0278,  # DHA\n",
    "        \"FA_hs_22_4Z7Z10Z13Z16Z_c\": 0.0038,  # (4Z,7Z,10Z,13Z,16Z)-DPA\n",
    "        \"FA_hs_22_7Z10Z13Z16Z19Z_c\": 0.0059,  # DPA\n",
    "        \"FA_hs_22_7Z10Z13Z16Z_c\": 0.0014,  # adrenic acid\n",
    "        \"FA_hs_23_0_c\": 0.0004,  # tricosanoic acid\n",
    "        \"FA_hs_24_0_c\": 0.0004,  # lignocerate\n",
    "        \"FA_hs_24_12Z15Z18Z21Z_c\": 0.0004,  # 12,15,18,21-tetracosatetraenoic acid\n",
    "        \"FA_hs_24_15Z_c\": 0.0004,  # nervonic acid\n",
    "        \"FA_hs_24_6Z9Z12Z15Z18Z21Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z,21Z)-THA\n",
    "        \"FA_hs_24_6Z9Z12Z15Z18Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z)-TPA\n",
    "        \"FA_hs_24_9Z12Z15Z18Z21Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z,21Z)-TPA\n",
    "        \"FA_hs_24_9Z12Z15Z18Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z)-TTA\n",
    "        \"FA_hs_26_0_c\": 0.0004,  # cerotic acid\n",
    "        \"FA_hs_26_17Z_c\": 0.0004,  # ximenic acid\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "666c9e41-7b07-4f0b-a5a0-518b4d2fe503",
   "metadata": {},
   "source": [
    "#### Update annotation mappings\n",
    "Map to identifiers of database initially to be able to extract annotations from those databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8692a4fd-db0b-4467-a2fb-9e27e7f36103",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metabolite_mappings = pd.read_csv(\n",
    "    f\"{ROOT_PATH}{CURATION_PATH}/metabolites_InitialMappings_1.0.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=str,\n",
    "    index_col=0,\n",
    ").fillna(\"\")\n",
    "df_metabolite_mappings = df_metabolite_mappings.set_index(\"metabolites\")\n",
    "for metabolite, row in df_metabolite_mappings.iterrows():\n",
    "    metabolite = model.metabolites.get_by_id(metabolite)\n",
    "    for key in df_metabolite_mappings.columns:\n",
    "        item = row[key]\n",
    "        if item:\n",
    "            metabolite.annotation[key] = item\n",
    "\n",
    "\n",
    "df_reaction_mappings = pd.read_csv(\n",
    "    f\"{ROOT_PATH}{CURATION_PATH}/reactions_InitialMappings_1.0.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=str,\n",
    "    index_col=0,\n",
    ").fillna(\"\")\n",
    "df_reaction_mappings = df_reaction_mappings.set_index(\"reactions\")\n",
    "for reaction, row in df_reaction_mappings.iterrows():\n",
    "    reaction = model.reactions.get_by_id(reaction)\n",
    "    for key in df_reaction_mappings.columns:\n",
    "        item = row[key]\n",
    "        if item:\n",
    "            reaction.annotation[key] = item"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76cf33ba-798a-4513-a799-877457ca2f8c",
   "metadata": {},
   "source": [
    "#### Reset subsystem groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d5cddb5d-8226-4086-85f8-9ed52e035bbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.remove_groups(model.groups)\n",
    "for subsystem in sorted(set(model.reactions.list_attr(\"subsystem\"))):\n",
    "    reaction_list = model.reactions.query(lambda x: x.subsystem == subsystem)\n",
    "    if subsystem not in model.groups:\n",
    "        group = Group(id=subsystem, name=subsystem, members=reaction_list)\n",
    "        model.add_groups([group])\n",
    "    else:\n",
    "        group = model.groups.get_by_id(subsystem).add_members(reaction_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaeae0fb-1f5b-4375-8510-f0f28c79f905",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d5f9e94e-1123-46d6-8352-ba96d2af7682",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14939f8d0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1967</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2790</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>653</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>74</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14939f8d0>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_rbc_model(model, filetype=\"xml\")\n",
    "model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
