{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a1e69c41-9c8e-4c93-962b-a7981b422a94",
   "metadata": {},
   "source": [
    "# RBC-GEM 1.0.0 Updates\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3b792e3-d4e1-4305-8aaf-1bbd08260daa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from cobra.core import Gene, Group, Metabolite, Reaction\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    GEM_NAME,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    write_cobra_model,\n",
    ")\n",
    "from rbc_gem_utils.qc import standardardize_metabolite_formulas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caf18f40-c50a-4b26-b48e-2e0db76eff68",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "778339d1-01af-4afa-a6bf-3a78c0fc2db9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c976553a-d69a-4981-a906-a9effaea0f12",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 0.3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b0e4d876-7d91-483d-9f12-ca8021151cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>150044610</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x150044610>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd6f76c4-304f-4c19-a322-59f7cc3885e9",
   "metadata": {},
   "source": [
    "### Update model based on curation and proteomic evidence\n",
    "* Add metabolites first, then genes and lastly, reactions\n",
    "* Add metabolites with mass and charge balanced formulas (ChemAxon)\n",
    "* After adding to model, connect to MetabolicAtlas where possible and run annotation scripts.\n",
    "* Evidence tables also contain current model objects, adjust stoichiometry if necessary for mass balancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6c927168-df6b-4dfe-8470-86b5352e49fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_updated = {}\n",
    "overwrite = False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f3a3d99-8ba3-4b1b-9df1-3a28d7e42491",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9a8fc008-7d03-43ad-8f02-067775df9e5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>metabolite</th>\n",
       "      <th>name</th>\n",
       "      <th>formula</th>\n",
       "      <th>charge</th>\n",
       "      <th>compartment</th>\n",
       "      <th>metabolomic evidence (#studies)</th>\n",
       "      <th>metabolomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10fthf_c</td>\n",
       "      <td>10fthf</td>\n",
       "      <td>10-formyl-tetrahydrofolate</td>\n",
       "      <td>C20H21N7O7</td>\n",
       "      <td>-2</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12ppd__R_c</td>\n",
       "      <td>12ppd__R</td>\n",
       "      <td>(R)-propane-1,2-diol</td>\n",
       "      <td>C3H8O2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12ppd__S_c</td>\n",
       "      <td>12ppd__S</td>\n",
       "      <td>(S)-propane-1,2-diol</td>\n",
       "      <td>C3H8O2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13dampp_c</td>\n",
       "      <td>13dampp</td>\n",
       "      <td>1,3-diaminopropane</td>\n",
       "      <td>C3H12N2</td>\n",
       "      <td>2</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13dpg_c</td>\n",
       "      <td>13dpg</td>\n",
       "      <td>3-Phospho-D-glyceroyl phosphate</td>\n",
       "      <td>C3H4O10P2</td>\n",
       "      <td>-4</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2152</th>\n",
       "      <td>zn2_c</td>\n",
       "      <td>zn2</td>\n",
       "      <td>Zinc (II)</td>\n",
       "      <td>Zn</td>\n",
       "      <td>2</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153</th>\n",
       "      <td>zn2_e</td>\n",
       "      <td>zn2</td>\n",
       "      <td>Zinc (II)</td>\n",
       "      <td>Zn</td>\n",
       "      <td>2</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154</th>\n",
       "      <td>ztp_c</td>\n",
       "      <td>ztp</td>\n",
       "      <td>5-amino-4-imidazolecarboxamide riboside 5'-tri...</td>\n",
       "      <td>C9H13N4O14P3</td>\n",
       "      <td>-4</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155</th>\n",
       "      <td>zymst_c</td>\n",
       "      <td>zymst</td>\n",
       "      <td>zymosterol</td>\n",
       "      <td>C27H44O</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156</th>\n",
       "      <td>zymstnl_c</td>\n",
       "      <td>zymstnl</td>\n",
       "      <td>zymostenol</td>\n",
       "      <td>C27H46O</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2157 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     metabolites metabolite  \\\n",
       "0       10fthf_c     10fthf   \n",
       "1     12ppd__R_c   12ppd__R   \n",
       "2     12ppd__S_c   12ppd__S   \n",
       "3      13dampp_c    13dampp   \n",
       "4        13dpg_c      13dpg   \n",
       "...          ...        ...   \n",
       "2152       zn2_c        zn2   \n",
       "2153       zn2_e        zn2   \n",
       "2154       ztp_c        ztp   \n",
       "2155     zymst_c      zymst   \n",
       "2156   zymstnl_c    zymstnl   \n",
       "\n",
       "                                                   name       formula charge  \\\n",
       "0                            10-formyl-tetrahydrofolate    C20H21N7O7     -2   \n",
       "1                                  (R)-propane-1,2-diol        C3H8O2      0   \n",
       "2                                  (S)-propane-1,2-diol        C3H8O2      0   \n",
       "3                                    1,3-diaminopropane       C3H12N2      2   \n",
       "4                       3-Phospho-D-glyceroyl phosphate     C3H4O10P2     -4   \n",
       "...                                                 ...           ...    ...   \n",
       "2152                                          Zinc (II)            Zn      2   \n",
       "2153                                          Zinc (II)            Zn      2   \n",
       "2154  5-amino-4-imidazolecarboxamide riboside 5'-tri...  C9H13N4O14P3     -4   \n",
       "2155                                         zymosterol       C27H44O      0   \n",
       "2156                                         zymostenol       C27H46O      0   \n",
       "\n",
       "     compartment metabolomic evidence (#studies)  \\\n",
       "0              c                                   \n",
       "1              c                                   \n",
       "2              c                                   \n",
       "3              c                                   \n",
       "4              c                                   \n",
       "...          ...                             ...   \n",
       "2152           c                                   \n",
       "2153           e                                   \n",
       "2154           c                                   \n",
       "2155           c                                   \n",
       "2156           c                                   \n",
       "\n",
       "     metabolomic evidence (pubmed) references notes  \n",
       "0                                                    \n",
       "1                                                    \n",
       "2                                                    \n",
       "3                                                    \n",
       "4                                                    \n",
       "...                            ...        ...   ...  \n",
       "2152                                                 \n",
       "2153                                                 \n",
       "2154                                                 \n",
       "2155                                                 \n",
       "2156                                                 \n",
       "\n",
       "[2157 rows x 10 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "\n",
    "df_evidence = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "for idx, row in df_evidence.iterrows():\n",
    "    mid, name, formula, charge, compartment = row[\n",
    "        [\"metabolites\", \"name\", \"formula\", \"charge\", \"compartment\"]\n",
    "    ]\n",
    "    if not model.metabolites.has_id(mid):\n",
    "        # Add reaction to model\n",
    "        model.add_metabolites([Metabolite(mid)])\n",
    "\n",
    "    metabolite = model.metabolites.get_by_id(mid)\n",
    "    metabolite.name = name\n",
    "    metabolite.formula = formula\n",
    "    metabolite.charge = int(charge)\n",
    "    metabolite.compartment = compartment\n",
    "\n",
    "met_formulas = standardardize_metabolite_formulas(\n",
    "    dict(zip(model.metabolites.list_attr(\"id\"), model.metabolites.list_attr(\"formula\")))\n",
    ")\n",
    "for mid, formula in met_formulas.items():\n",
    "    model.metabolites.get_by_id(mid).formula = formula\n",
    "\n",
    "dataframes_updated[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57490574-bcd3-43b3-b91c-3601e0e180c8",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9f50928d-4096-4bb7-b43e-b5bb2e22ce2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>proteomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>P04406</td>\n",
       "      <td>2597</td>\n",
       "      <td>GAPDH</td>\n",
       "      <td>28</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1686133...</td>\n",
       "      <td>pubmed:19430704</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HSPA8</td>\n",
       "      <td>P11142</td>\n",
       "      <td>3312</td>\n",
       "      <td>HSPA8</td>\n",
       "      <td>26</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1686133...</td>\n",
       "      <td>pubmed:33832207</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CA1</td>\n",
       "      <td>P00915</td>\n",
       "      <td>759</td>\n",
       "      <td>CA1</td>\n",
       "      <td>25</td>\n",
       "      <td>pubmed:14963112;pubmed:16861337;pubmed:1849451...</td>\n",
       "      <td>pubmed:10090333;pubmed:4958988;pubmed:4975695</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>P00352</td>\n",
       "      <td>216</td>\n",
       "      <td>ALDH1A1</td>\n",
       "      <td>24</td>\n",
       "      <td>pubmed:16861337;pubmed:18399644;pubmed:1849451...</td>\n",
       "      <td>pubmed:17175089;pubmed:224930;pubmed:2776714;p...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BLVRB</td>\n",
       "      <td>P30043</td>\n",
       "      <td>645</td>\n",
       "      <td>BLVRB</td>\n",
       "      <td>24</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1861456...</td>\n",
       "      <td>pubmed:10858451;pubmed:34598;pubmed:8117274</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>SMPD3</td>\n",
       "      <td>Q9NY59</td>\n",
       "      <td>55512</td>\n",
       "      <td>SMPD3</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:22824271;pubmed:25075126</td>\n",
       "      <td>pubmed:25075126;pubmed:22824271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>ST3GAL1</td>\n",
       "      <td>Q11201</td>\n",
       "      <td>6482</td>\n",
       "      <td>ST3GAL1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:3084191;pubmed:5003667</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>ST3GAL2</td>\n",
       "      <td>Q16842</td>\n",
       "      <td>6483</td>\n",
       "      <td>ST3GAL2</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:3084191;pubmed:5003667</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>TRPC6</td>\n",
       "      <td>Q9Y210</td>\n",
       "      <td>7225</td>\n",
       "      <td>TRPC6</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:18209485</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819</th>\n",
       "      <td>TSPO2</td>\n",
       "      <td>Q5TGU0</td>\n",
       "      <td>222642</td>\n",
       "      <td>TSPO2</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:27641616;pubmed:30061676;pubmed:31989647</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>820 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genes uniprot ncbigene hgnc.symbol proteomic evidence (#studies)  \\\n",
       "0      GAPDH  P04406     2597       GAPDH                            28   \n",
       "1      HSPA8  P11142     3312       HSPA8                            26   \n",
       "2        CA1  P00915      759         CA1                            25   \n",
       "3    ALDH1A1  P00352      216     ALDH1A1                            24   \n",
       "4      BLVRB  P30043      645       BLVRB                            24   \n",
       "..       ...     ...      ...         ...                           ...   \n",
       "815    SMPD3  Q9NY59    55512       SMPD3                             0   \n",
       "816  ST3GAL1  Q11201     6482     ST3GAL1                             0   \n",
       "817  ST3GAL2  Q16842     6483     ST3GAL2                             0   \n",
       "818    TRPC6  Q9Y210     7225       TRPC6                             0   \n",
       "819    TSPO2  Q5TGU0   222642       TSPO2                             0   \n",
       "\n",
       "                           proteomic evidence (pubmed)  \\\n",
       "0    pubmed:12362340;pubmed:14963112;pubmed:1686133...   \n",
       "1    pubmed:12362340;pubmed:14963112;pubmed:1686133...   \n",
       "2    pubmed:14963112;pubmed:16861337;pubmed:1849451...   \n",
       "3    pubmed:16861337;pubmed:18399644;pubmed:1849451...   \n",
       "4    pubmed:12362340;pubmed:14963112;pubmed:1861456...   \n",
       "..                                                 ...   \n",
       "815                                                      \n",
       "816                                                      \n",
       "817                                                      \n",
       "818                                                      \n",
       "819                                                      \n",
       "\n",
       "                                            references  \\\n",
       "0                                      pubmed:19430704   \n",
       "1                                      pubmed:33832207   \n",
       "2        pubmed:10090333;pubmed:4958988;pubmed:4975695   \n",
       "3    pubmed:17175089;pubmed:224930;pubmed:2776714;p...   \n",
       "4          pubmed:10858451;pubmed:34598;pubmed:8117274   \n",
       "..                                                 ...   \n",
       "815                    pubmed:22824271;pubmed:25075126   \n",
       "816                      pubmed:3084191;pubmed:5003667   \n",
       "817                      pubmed:3084191;pubmed:5003667   \n",
       "818                                    pubmed:18209485   \n",
       "819    pubmed:27641616;pubmed:30061676;pubmed:31989647   \n",
       "\n",
       "                               notes  \n",
       "0                                     \n",
       "1                                     \n",
       "2                                     \n",
       "3                                     \n",
       "4                                     \n",
       "..                               ...  \n",
       "815  pubmed:25075126;pubmed:22824271  \n",
       "816                                   \n",
       "817                                   \n",
       "818                                   \n",
       "819                                   \n",
       "\n",
       "[820 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "\n",
    "df_evidence = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "\n",
    "for idx, row in df_evidence.iterrows():\n",
    "    # Between these three annotation fields, should be likely that the rest can get extracted from UniProt.\n",
    "    gid = row[\"genes\"]\n",
    "    if not model.genes.has_id(gid):\n",
    "        gene = Gene(gid)\n",
    "        model.genes.extend([gene])\n",
    "    gene = model.genes.get_by_id(gid)\n",
    "    for key in [\"uniprot\", \"ncbigene\", \"hgnc.symbol\"]:\n",
    "        item = row[key]\n",
    "        if item:\n",
    "            gene.annotation[key] = item\n",
    "\n",
    "dataframes_updated[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b179ad6-3125-47c8-8fc9-d6f82952fb3c",
   "metadata": {},
   "source": [
    "#### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "12401ee3-4a2e-45c5-94c8-81af0003b3c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>reaction</th>\n",
       "      <th>gene reaction rule</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>spontaneous</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5FLUDPK</td>\n",
       "      <td>Nucleoside-diphosphate kinase (ATP:5-fluorouri...</td>\n",
       "      <td>5fludp_c + atp_c --&gt; 5flutp_c + adp_c</td>\n",
       "      <td>NME1 or (NME1 and NME2) or NME2 or (NME1 and N...</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>NME1(17);NME2(12);NME3(1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5FLUMPK</td>\n",
       "      <td>Uridylate kinase (ATP:5-fluorouridine 5-monoph...</td>\n",
       "      <td>5flump_c + atp_c --&gt; 5fludp_c + adp_c</td>\n",
       "      <td>CMPK1</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>CMPK1(13)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5FLURAPT</td>\n",
       "      <td>5-fluorouracil phosphoribosyltransferase</td>\n",
       "      <td>5flura_c + prpp_c --&gt; 5flump_c + ppi_c</td>\n",
       "      <td>UMPS</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>UMPS(7)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D5FURADy</td>\n",
       "      <td>5,6-dihydro-5-fluorouracil:NADP + 5-oxidoreduc...</td>\n",
       "      <td>5flura_c + h_c + nadph_c --&gt; 56dh5flura_c + na...</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>DPYD(1)</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FACOAE_10_0</td>\n",
       "      <td>Acyl CoA thioesterase (CoA 10:0)</td>\n",
       "      <td>FAcoa_hs_10_0_c + h2o_c --&gt; FA_hs_10_0_c + coa...</td>\n",
       "      <td>ACOT7</td>\n",
       "      <td>Acyl-CoA hydrolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>ACOT7(14)</td>\n",
       "      <td>pubmed:10578051;pubmed:5042454</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2718</th>\n",
       "      <td>PYDXPPe</td>\n",
       "      <td>Pyridoxal 5-phosphate phosphatase</td>\n",
       "      <td>h2o_e + pydx5p_e --&gt; pi_e + pydx_e</td>\n",
       "      <td>ALPL</td>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:1322411;pubmed:14522954;pubmed:7448195;...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2719</th>\n",
       "      <td>AVITE1RASCBR1</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via L-as...</td>\n",
       "      <td>ascb__L_c + avite1r_c &lt;=&gt; avite1_c + mdhdascb_...</td>\n",
       "      <td></td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:18243141;pubmed:36643550;pubmed:8388377...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2720</th>\n",
       "      <td>AVITE1RCYTB5R</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via L-as...</td>\n",
       "      <td>avite1r_c + focytb5_c + h_c --&gt; avite1_c + fic...</td>\n",
       "      <td>(CYB5R3 and CYB5B) or (CYB5R3 and CYB5A)</td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>CYB5A(13);CYB5B(8);CYB5R3(20)</td>\n",
       "      <td>pubmed:8388377</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2721</th>\n",
       "      <td>AVITE1RQ10H2R</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via ubiq...</td>\n",
       "      <td>avite1r_c + q10h2_c --&gt; avite1_c + q10h_c</td>\n",
       "      <td></td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:8388377;pubmed:8434935</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2722</th>\n",
       "      <td>AVITE1RQ10HR</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via ubiq...</td>\n",
       "      <td>avite1r_c + q10h_c --&gt; avite1_c + q10_c</td>\n",
       "      <td></td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:8388377;pubmed:8434935</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2723 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          reactions                                               name  \\\n",
       "0           5FLUDPK  Nucleoside-diphosphate kinase (ATP:5-fluorouri...   \n",
       "1           5FLUMPK  Uridylate kinase (ATP:5-fluorouridine 5-monoph...   \n",
       "2          5FLURAPT           5-fluorouracil phosphoribosyltransferase   \n",
       "3          D5FURADy  5,6-dihydro-5-fluorouracil:NADP + 5-oxidoreduc...   \n",
       "4       FACOAE_10_0                   Acyl CoA thioesterase (CoA 10:0)   \n",
       "...             ...                                                ...   \n",
       "2718        PYDXPPe                  Pyridoxal 5-phosphate phosphatase   \n",
       "2719  AVITE1RASCBR1  Reduction of alpha-tocopherol radical via L-as...   \n",
       "2720  AVITE1RCYTB5R  Reduction of alpha-tocopherol radical via L-as...   \n",
       "2721  AVITE1RQ10H2R  Reduction of alpha-tocopherol radical via ubiq...   \n",
       "2722   AVITE1RQ10HR  Reduction of alpha-tocopherol radical via ubiq...   \n",
       "\n",
       "                                               reaction  \\\n",
       "0                 5fludp_c + atp_c --> 5flutp_c + adp_c   \n",
       "1                 5flump_c + atp_c --> 5fludp_c + adp_c   \n",
       "2                5flura_c + prpp_c --> 5flump_c + ppi_c   \n",
       "3     5flura_c + h_c + nadph_c --> 56dh5flura_c + na...   \n",
       "4     FAcoa_hs_10_0_c + h2o_c --> FA_hs_10_0_c + coa...   \n",
       "...                                                 ...   \n",
       "2718                 h2o_e + pydx5p_e --> pi_e + pydx_e   \n",
       "2719  ascb__L_c + avite1r_c <=> avite1_c + mdhdascb_...   \n",
       "2720  avite1r_c + focytb5_c + h_c --> avite1_c + fic...   \n",
       "2721          avite1r_c + q10h2_c --> avite1_c + q10h_c   \n",
       "2722            avite1r_c + q10h_c --> avite1_c + q10_c   \n",
       "\n",
       "                                     gene reaction rule  \\\n",
       "0     NME1 or (NME1 and NME2) or NME2 or (NME1 and N...   \n",
       "1                                                 CMPK1   \n",
       "2                                                  UMPS   \n",
       "3                                                  DPYD   \n",
       "4                                                 ACOT7   \n",
       "...                                                 ...   \n",
       "2718                                               ALPL   \n",
       "2719                                                      \n",
       "2720           (CYB5R3 and CYB5B) or (CYB5R3 and CYB5A)   \n",
       "2721                                                      \n",
       "2722                                                      \n",
       "\n",
       "                      subsystem spontaneous  proteomic evidence (#studies)  \\\n",
       "0     5-fluorouracil metabolism           0      NME1(17);NME2(12);NME3(1)   \n",
       "1     5-fluorouracil metabolism           0                      CMPK1(13)   \n",
       "2     5-fluorouracil metabolism           0                        UMPS(7)   \n",
       "3     5-fluorouracil metabolism           0                        DPYD(1)   \n",
       "4           Acyl-CoA hydrolysis           0                      ACOT7(14)   \n",
       "...                         ...         ...                            ...   \n",
       "2718      Vitamin B6 metabolism           0                                  \n",
       "2719       Vitamin E metabolism           1                                  \n",
       "2720       Vitamin E metabolism           0  CYB5A(13);CYB5B(8);CYB5R3(20)   \n",
       "2721       Vitamin E metabolism           1                                  \n",
       "2722       Vitamin E metabolism           1                                  \n",
       "\n",
       "                                             references notes  \n",
       "0                                                              \n",
       "1                                                              \n",
       "2                                                              \n",
       "3                                                              \n",
       "4                        pubmed:10578051;pubmed:5042454        \n",
       "...                                                 ...   ...  \n",
       "2718  pubmed:1322411;pubmed:14522954;pubmed:7448195;...        \n",
       "2719  pubmed:18243141;pubmed:36643550;pubmed:8388377...        \n",
       "2720                                     pubmed:8388377        \n",
       "2721                      pubmed:8388377;pubmed:8434935        \n",
       "2722                      pubmed:8388377;pubmed:8434935        \n",
       "\n",
       "[2723 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "\n",
    "df_evidence = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "for idx, row in df_evidence.iterrows():\n",
    "    rid, name, reaction_str, gpr, subsystem = row[\n",
    "        [\"reactions\", \"name\", \"reaction\", \"gene reaction rule\", \"subsystem\"]\n",
    "    ]\n",
    "    if not model.reactions.has_id(rid):\n",
    "        # Add reaction to model\n",
    "        model.add_reactions([Reaction(rid)])\n",
    "\n",
    "    reaction = model.reactions.get_by_id(rid)\n",
    "    reaction.build_reaction_from_string(reaction_str)\n",
    "    reaction.name = name\n",
    "    reaction.gene_reaction_rule = gpr\n",
    "    reaction.subsystem = subsystem\n",
    "\n",
    "dataframes_updated[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1185f5fb-9f60-4c8d-8c61-2480f0e7a90c",
   "metadata": {},
   "source": [
    "### Remove from model\n",
    "Remove reactions first, then genes, and lastly metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a407a3bd-ebd0-4c72-bd98-66cb0eb14eec",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_removed = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56d67737-ef34-4a18-adff-5d9c4b584320",
   "metadata": {},
   "source": [
    "#### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "dcb9e282-0d2f-4dc2-b23a-e04dda2a4de8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3MOXTYRESSte</td>\n",
       "      <td>3-Methoxytyramine secretion via secretory vesi...</td>\n",
       "      <td>Replaced by simple transport reaction (3MOXTYR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALAt4</td>\n",
       "      <td>Alanine-Sodium symporter</td>\n",
       "      <td>Replaced by reactions involving the amino acid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BANDMT</td>\n",
       "      <td>Band membrane protein-methyltransferase</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BILGLCURt</td>\n",
       "      <td>Bilirubin monoglucuronide transport via bicarb...</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BILIRBU</td>\n",
       "      <td>Bilirubin UDP-glucuronosyltransferase</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>SK_pe_hs_18_9Z_18_9Z_c</td>\n",
       "      <td>Sink pe hs 18 1 18 1(c)</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>THMMPtrbc</td>\n",
       "      <td>Thiamine monophosphate transport (passive - re...</td>\n",
       "      <td>Replaced by reactions involving reduced folate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>TMDPPK</td>\n",
       "      <td>Thiamine diphosphate kinase</td>\n",
       "      <td>Replaced by ADK7. Adenylate kinase isozyme 1 h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>UDPGD</td>\n",
       "      <td>UDPglucose 6-dehydrogenase</td>\n",
       "      <td>No proteomic or activity evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>UDPGNP</td>\n",
       "      <td>UDP-glucuronate glucuronohydrolase</td>\n",
       "      <td>No proteomic or activity evidence</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  reactions  \\\n",
       "0              3MOXTYRESSte   \n",
       "1                     ALAt4   \n",
       "2                    BANDMT   \n",
       "3                 BILGLCURt   \n",
       "4                   BILIRBU   \n",
       "..                      ...   \n",
       "171  SK_pe_hs_18_9Z_18_9Z_c   \n",
       "172               THMMPtrbc   \n",
       "173                  TMDPPK   \n",
       "174                   UDPGD   \n",
       "175                  UDPGNP   \n",
       "\n",
       "                                                  name  \\\n",
       "0    3-Methoxytyramine secretion via secretory vesi...   \n",
       "1                             Alanine-Sodium symporter   \n",
       "2              Band membrane protein-methyltransferase   \n",
       "3    Bilirubin monoglucuronide transport via bicarb...   \n",
       "4                Bilirubin UDP-glucuronosyltransferase   \n",
       "..                                                 ...   \n",
       "171                            Sink pe hs 18 1 18 1(c)   \n",
       "172  Thiamine monophosphate transport (passive - re...   \n",
       "173                        Thiamine diphosphate kinase   \n",
       "174                         UDPglucose 6-dehydrogenase   \n",
       "175                 UDP-glucuronate glucuronohydrolase   \n",
       "\n",
       "                                                 notes  \n",
       "0    Replaced by simple transport reaction (3MOXTYR...  \n",
       "1    Replaced by reactions involving the amino acid...  \n",
       "2    Replaced by generic methylation reactions. Ban...  \n",
       "3    Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "4    Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "..                                                 ...  \n",
       "171  Lipids are pooled, replaced with pooled versio...  \n",
       "172  Replaced by reactions involving reduced folate...  \n",
       "173  Replaced by ADK7. Adenylate kinase isozyme 1 h...  \n",
       "174                  No proteomic or activity evidence  \n",
       "175                  No proteomic or activity evidence  \n",
       "\n",
       "[176 rows x 3 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for reaction in df_removed[attribute_type]:\n",
    "    try:\n",
    "        reaction = getattr(model, attribute_type).get_by_id(reaction)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(reaction)\n",
    "\n",
    "model.remove_reactions(to_remove)\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]\n",
    "\n",
    "# Clean up removal file before archial\n",
    "df_removed = df_removed.sort_values(attribute_type, ascending=True).reset_index(\n",
    "    drop=True\n",
    ")\n",
    "df_removed.to_csv(\n",
    "    get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "df_removed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d8d0eca-5c50-4679-bf22-fb43535ccccf",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7a0a09ff-04bf-4f3f-8db6-00da29e45e77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ADCY1</td>\n",
       "      <td>Q08828</td>\n",
       "      <td>ADCY1</td>\n",
       "      <td>107</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADCY3</td>\n",
       "      <td>O60266</td>\n",
       "      <td>ADCY3</td>\n",
       "      <td>109</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADCY4</td>\n",
       "      <td>Q8NFM4</td>\n",
       "      <td>ADCY4</td>\n",
       "      <td>196883</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADCY5</td>\n",
       "      <td>O95622</td>\n",
       "      <td>ADCY5</td>\n",
       "      <td>111</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADCY6</td>\n",
       "      <td>O43306</td>\n",
       "      <td>ADCY6</td>\n",
       "      <td>112</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>TAT</td>\n",
       "      <td>P17735</td>\n",
       "      <td>TAT</td>\n",
       "      <td>6898</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>UGDH</td>\n",
       "      <td>O60701</td>\n",
       "      <td>UGDH</td>\n",
       "      <td>7358</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>P22309</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>54658</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>P22310</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>54657</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>ZNF384</td>\n",
       "      <td>Q8TF68</td>\n",
       "      <td>ZNF384</td>\n",
       "      <td>171017</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes uniprot hgnc.symbol ncbigene  \\\n",
       "0     ADCY1  Q08828       ADCY1      107   \n",
       "1     ADCY3  O60266       ADCY3      109   \n",
       "2     ADCY4  Q8NFM4       ADCY4   196883   \n",
       "3     ADCY5  O95622       ADCY5      111   \n",
       "4     ADCY6  O43306       ADCY6      112   \n",
       "..      ...     ...         ...      ...   \n",
       "108     TAT  P17735         TAT     6898   \n",
       "109    UGDH  O60701        UGDH     7358   \n",
       "110  UGT1A1  P22309      UGT1A1    54658   \n",
       "111  UGT1A4  P22310      UGT1A4    54657   \n",
       "112  ZNF384  Q8TF68      ZNF384   171017   \n",
       "\n",
       "                                                 notes  \n",
       "0    no proteomic evidence. evidence indicates othe...  \n",
       "1    no proteomic evidence. evidence indicates othe...  \n",
       "2    no proteomic evidence. evidence indicates othe...  \n",
       "3    no proteomic evidence. evidence indicates othe...  \n",
       "4    no proteomic evidence. evidence indicates othe...  \n",
       "..                                                 ...  \n",
       "108  no proteomic evidence. evidence indicates othe...  \n",
       "109                             no proteomic evidence.  \n",
       "110                             no proteomic evidence.  \n",
       "111                             no proteomic evidence.  \n",
       "112  no proteomic evidence. evidence indicates othe...  \n",
       "\n",
       "[113 rows x 5 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for gene in df_removed[attribute_type]:\n",
    "    try:\n",
    "        gene = model.genes.get_by_id(gene)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(gene)\n",
    "model.genes -= to_remove\n",
    "\n",
    "\n",
    "# Clean up removal file before archial\n",
    "df_removed = df_removed.sort_values(attribute_type, ascending=True).reset_index(\n",
    "    drop=True\n",
    ")\n",
    "df_removed.to_csv(\n",
    "    get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\", sep=\"\\t\", index=False\n",
    ")\n",
    "df_removed\n",
    "\n",
    "\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "\n",
    "for gene in model.genes:\n",
    "    if gene.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {gene}\")\n",
    "\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ad923fa-b442-4667-93f5-971a90cb1318",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e987854a-c0db-4e4b-9a70-aa251551490e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>band_c</td>\n",
       "      <td>Band membrane protein (universal, erythrocyte ...</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bandmt_c</td>\n",
       "      <td>Band membrane protein (methylated, universal, ...</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bilglcur_c</td>\n",
       "      <td>Bilirubin monoglucuronide</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bilglcur_e</td>\n",
       "      <td>Bilirubin monoglucuronide</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cdpdag_hs_16_0_16_0_c</td>\n",
       "      <td>CDP diacylglycerol (homo sapiens, C16:0, C16:0)</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>pe_hs_18_9Z_18_9Z12Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:1,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>pe_hs_18_9Z12Z_16_0_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:2,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>pe_hs_18_9Z12Z_18_9Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:2,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>pppg9_c</td>\n",
       "      <td>protoporphyrinogen IX</td>\n",
       "      <td>Protoporphyrinogen oxidase is a mitochondrial ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>udpglcur_c</td>\n",
       "      <td>UDP-D-glucuronate</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               metabolites                                               name  \\\n",
       "0                   band_c  Band membrane protein (universal, erythrocyte ...   \n",
       "1                 bandmt_c  Band membrane protein (methylated, universal, ...   \n",
       "2               bilglcur_c                          Bilirubin monoglucuronide   \n",
       "3               bilglcur_e                          Bilirubin monoglucuronide   \n",
       "4    cdpdag_hs_16_0_16_0_c    CDP diacylglycerol (homo sapiens, C16:0, C16:0)   \n",
       "..                     ...                                                ...   \n",
       "65  pe_hs_18_9Z_18_9Z12Z_c  Phosphatidylethanolamine (homo sapiens, C18:1,...   \n",
       "66   pe_hs_18_9Z12Z_16_0_c  Phosphatidylethanolamine (homo sapiens, C18:2,...   \n",
       "67  pe_hs_18_9Z12Z_18_9Z_c  Phosphatidylethanolamine (homo sapiens, C18:2,...   \n",
       "68                 pppg9_c                              protoporphyrinogen IX   \n",
       "69              udpglcur_c                                  UDP-D-glucuronate   \n",
       "\n",
       "                                                notes  \n",
       "0   Replaced by generic methylation reactions. Ban...  \n",
       "1   Replaced by generic methylation reactions. Ban...  \n",
       "2   Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "3   Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "4   Lipids are pooled, replaced with pooled versio...  \n",
       "..                                                ...  \n",
       "65  Lipids are pooled, replaced with pooled versio...  \n",
       "66  Lipids are pooled, replaced with pooled versio...  \n",
       "67  Lipids are pooled, replaced with pooled versio...  \n",
       "68  Protoporphyrinogen oxidase is a mitochondrial ...  \n",
       "69  Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "\n",
       "[70 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for metabolite in df_removed[attribute_type]:\n",
    "    try:\n",
    "        metabolite = model.metabolites.get_by_id(metabolite)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    model.remove_metabolites([metabolite])\n",
    "model.remove_metabolites(to_remove)\n",
    "\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {metabolite}\")\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbd78766-e03a-4fc2-8215-32d696fa4b1c",
   "metadata": {},
   "source": [
    "#### Add exchanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "eb647fa6-94e1-4d85-8353-2f6eef374b1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in model.metabolites.query(lambda x: x.compartment == \"e\"):\n",
    "    try:\n",
    "        model.add_boundary(met, type=\"exchange\")\n",
    "    except ValueError:\n",
    "        pass\n",
    "for reaction in model.boundary:\n",
    "    reaction.subsystem = \"Pseudoreactions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b93b32-2bc9-44b1-b896-32460fa3c53c",
   "metadata": {},
   "source": [
    "#### Add pooled lipid reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6babdc32-7ac2-4132-8ba1-97418e4d299b",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.add_reactions(\n",
    "    [\n",
    "        Reaction(\n",
    "            \"POOL_FACOA\",\n",
    "            name=\"Pooling reaction Acyl-CoA (CoA)\",\n",
    "            subsystem=\"Pseudoreactions\",\n",
    "            lower_bound=COBRA_CONFIGURATION.lower_bound,\n",
    "            upper_bound=COBRA_CONFIGURATION.upper_bound,\n",
    "        ),\n",
    "        Reaction(\n",
    "            \"POOL_FA\",\n",
    "            name=\"Pooling reaction fatty acids (FA)\",\n",
    "            subsystem=\"Pseudoreactions\",\n",
    "            lower_bound=COBRA_CONFIGURATION.lower_bound,\n",
    "            upper_bound=COBRA_CONFIGURATION.upper_bound,\n",
    "        ),\n",
    "    ]\n",
    ")\n",
    "model.reactions.get_by_id(\"POOL_FACOA\").add_metabolites(\n",
    "    {\n",
    "        \"FAcoa_hs_c\": -1,\n",
    "        \"FAcoa_hs_12_0_c\": 0.0004,  # lauroyl-CoA\n",
    "        \"FAcoa_hs_13_0_c\": 0.0004,  # tridecanoyl-CoA\n",
    "        \"FAcoa_hs_14_0_c\": 0.0133,  # myristoyl-CoA\n",
    "        \"FAcoa_hs_14_5Z_c\": 0.0004,  # (5Z)-tetradecanoyl-CoA\n",
    "        \"FAcoa_hs_14_7Z_c\": 0.0004,  # (7Z)-tetradecenoyl-CoA\n",
    "        \"FAcoa_hs_14_9Z_c\": 0.0004,  # (9Z)-tetradecenoyl-CoA\n",
    "        \"FAcoa_hs_15_0_c\": 0.0004,  # Pentadecanoyl-CoA\n",
    "        \"FAcoa_hs_16_0_c\": 0.2220,  # Palmitoyl-CoA\n",
    "        \"FAcoa_hs_16_7Z_c\": 0.0004,  # (7Z)-palmitoleoyl-CoA\n",
    "        \"FAcoa_hs_16_9Z_c\": 0.0219,  # Palmitoleoyl-CoA\n",
    "        \"FAcoa_hs_17_0_c\": 0.0004,  # margaric acid\n",
    "        \"FAcoa_hs_17_10Z_c\": 0.0004,  #  (10Z)-heptadecylenic acid\n",
    "        \"FAcoa_hs_17_9Z_c\": 0.0004,  #  (9Z)-heptadecylenic acid\n",
    "        \"FAcoa_hs_18_0_c\": 0.1498,  # stearoyl-CoA\n",
    "        \"FAcoa_hs_18_9Z_c\": 0.1545,  # oleoyl-CoA\n",
    "        \"FAcoa_hs_18_11Z_c\": 0.0250,  # cis-vaccenoyl-CoA\n",
    "        \"FAcoa_hs_18_13Z_c\": 0.0004,  # (13Z)-octadecenoyl-CoA\n",
    "        \"FAcoa_hs_18_6Z9Z12Z15Z_c\": 0.0025,  # stearoyl-CoA\n",
    "        \"FAcoa_hs_18_6Z9Z12Z_c\": 0.0029,  # gamma-linolenoyl-CoA\n",
    "        \"FAcoa_hs_18_6Z9Z_c\": 0.0004,  # (6Z,9Z)-octadecadienoyl-CoA\n",
    "        \"FAcoa_hs_18_7Z_c\": 0.0004,  # (7Z)-octadecenoyl-CoA\n",
    "        \"FAcoa_hs_18_9E_c\": 0.0004,  # (9E)-octadecenoyl-CoA\n",
    "        \"FAcoa_hs_18_9Z12Z_c\": 0.1915,  # Linoleoyl-CoA\n",
    "        \"FAcoa_hs_18_9Z12Z15Z_c\": 0.0084,  # alpha-linolenoyl-CoA\n",
    "        \"FAcoa_hs_19_0_c\": 0.0004,  # nonadecanoyl-CoA\n",
    "        \"FAcoa_hs_20_0_c\": 0.0004,  # Arachidoyl-CoA /eicosanoyl-CoA\n",
    "        \"FAcoa_hs_20_11Z14Z17Z_c\": 0.0215,  # (11Z,14Z,17Z)-eicosatrienoyl-CoA\n",
    "        \"FAcoa_hs_20_11Z14Z_c\": 0.0004,  # (11Z,14Z)-eicosadienoyl-CoA\n",
    "        \"FAcoa_hs_20_11Z_c\": 0.0004,  # (11Z)-eicosenoyl-CoA\n",
    "        \"FAcoa_hs_20_13Z_c\": 0.0004,  # (13Z)-eicosenoyl-CoA\n",
    "        \"FAcoa_hs_20_5Z8Z11Z14Z17Z_c\": 0.0116,  # eicosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_20_5Z8Z11Z14Z_c\": 0.1083,  # arachidonoyl-CoA\n",
    "        \"FAcoa_hs_20_5Z8Z11Z_c\": 0.0004,  # (5Z,8Z,11Z)-eicosatrienoyl-CoA\n",
    "        \"FAcoa_hs_20_8Z11Z14Z17Z_c\": 0.0115,  # (8Z,11Z,14Z,17Z)-eicosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_20_8Z11Z14Z_c\": 0.0215,  #  dihomo-gamma-linoleoyl-CoA\n",
    "        \"FAcoa_hs_20_8Z11Z_c\": 0.0004,  # 8,11-eicosadienoyl-CoA\n",
    "        \"FAcoa_hs_20_9Z_c\": 0.0004,  # 9-eicosenoyl-CoA\n",
    "        \"FAcoa_hs_21_0_c\": 0.0004,  # henicosanoyl-CoA\n",
    "        \"FAcoa_hs_22_0_c\": 0.0004,  # docosanoyl-CoA\n",
    "        \"FAcoa_hs_22_10Z13Z16Z19Z_c\": 0.0004,  # 10,13,16,19-docosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_22_10Z13Z16Z_c\": 0.0004,  # 10,13,16-docosatriynoyl-CoA\n",
    "        \"FAcoa_hs_22_11Z_c\": 0.0004,  # (11Z)-docosenoyl-CoA\n",
    "        \"FAcoa_hs_22_13Z16Z19Z_c\": 0.0004,  # 13,16,19-docosatrienoyl-CoA\n",
    "        \"FAcoa_hs_22_13Z16Z_c\": 0.0004,  # (13Z,16Z)-docosadienoyl-CoA\n",
    "        \"FAcoa_hs_22_13Z_c\": 0.0004,  # (13Z)-docosenoyl-CoA\n",
    "        \"FAcoa_hs_22_4Z7Z10Z13Z16Z19Z_c\": 0.0278,  # DHA\n",
    "        \"FAcoa_hs_22_4Z7Z10Z13Z16Z_c\": 0.0038,  #  (4Z,7Z,10Z,13Z,16Z)-docosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_22_7Z10Z13Z16Z19Z_c\": 0.0059,  # (7Z,10Z,13Z,16Z,19Z)-docosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_22_7Z10Z13Z16Z_c\": 0.0014,  #  (7Z,10Z,13Z,16Z)-docosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_23_0_c\": 0.0004,  # tricosanoyl-CoA\n",
    "        \"FAcoa_hs_24_0_c\": 0.0004,  # tetracosanoyl-CoA\n",
    "        \"FAcoa_hs_24_12Z15Z18Z21Z_c\": 0.0004,  # 12,15,18,21-tetracosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_24_15Z_c\": 0.0004,  # (15Z)-tetracosenoyl-CoA\n",
    "        \"FAcoa_hs_24_6Z9Z12Z15Z18Z21Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z,21Z)-tetracosahexaenoyl-CoA\n",
    "        \"FAcoa_hs_24_6Z9Z12Z15Z18Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z)-tetracosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_24_9Z12Z15Z18Z21Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z,21Z)-tetracosapentaenoyl-CoA\n",
    "        \"FAcoa_hs_24_9Z12Z15Z18Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z)-tetracosatetraenoyl-CoA\n",
    "        \"FAcoa_hs_26_0_c\": 0.0004,  # hexacosanoyl-CoA\n",
    "        \"FAcoa_hs_26_17Z_c\": 0.0004,  #  hexacosenoyl-CoA\n",
    "    }\n",
    ")\n",
    "model.reactions.get_by_id(\"POOL_FA\").add_metabolites(\n",
    "    {\n",
    "        \"FA_hs_c\": -1,\n",
    "        \"FA_hs_12_0_c\": 0.0004,  # lauric acid\n",
    "        \"FA_hs_13_0_c\": 0.0004,  # tridecylic acid\n",
    "        \"FA_hs_14_0_c\": 0.0133,  # myristic acid\n",
    "        \"FA_hs_14_5Z_c\": 0.0004,  # physeteric acid\n",
    "        \"FA_hs_14_7Z_c\": 0.0004,  # 7Z-tetradecenoic acid\n",
    "        \"FA_hs_14_9Z_c\": 0.0004,  # 9Z-tetradecenoic acid\n",
    "        \"FA_hs_15_0_c\": 0.0004,  # pentadecylic acid\n",
    "        \"FA_hs_16_0_c\": 0.2220,  # Palmitic acid\n",
    "        \"FA_hs_16_7Z_c\": 0.0004,  # 7-palmitoleic acid\n",
    "        \"FA_hs_16_9Z_c\": 0.0219,  # Palmitoleic acid\n",
    "        \"FA_hs_17_0_c\": 0.0004,  # margaric acid\n",
    "        \"FA_hs_17_10Z_c\": 0.0004,  #  10-heptadecylenic acid\n",
    "        \"FA_hs_17_9Z_c\": 0.0004,  #  9-heptadecylenic acid\n",
    "        \"FA_hs_18_0_c\": 0.1498,  # stearate acid\n",
    "        \"FA_hs_18_9Z_c\": 0.1545,  # oleic acid\n",
    "        \"FA_hs_18_11Z_c\": 0.0250,  # cis-vaccenic acid\n",
    "        \"FA_hs_18_13Z_c\": 0.0004,  # (13Z)-octadecenoic acid\n",
    "        \"FA_hs_18_6Z9Z12Z15Z_c\": 0.0025,  # stearidonic acid\n",
    "        \"FA_hs_18_6Z9Z12Z_c\": 0.0029,  # gamma-linolenic acid\n",
    "        \"FA_hs_18_6Z9Z_c\": 0.0004,  # (6Z,9Z)-octadecadienoic acid\n",
    "        \"FA_hs_18_7Z_c\": 0.0004,  # (7Z)-octadecenoic acid\n",
    "        \"FA_hs_18_9E_c\": 0.0004,  # elaidate acid\n",
    "        \"FA_hs_18_9Z12Z_c\": 0.1915,  # Linoleic acid\n",
    "        \"FA_hs_18_9Z12Z15Z_c\": 0.0084,  # alpha-linolenic acid\n",
    "        \"FA_hs_19_0_c\": 0.0004,  # nonadecylic acid\n",
    "        \"FA_hs_20_0_c\": 0.0004,  # Arachidic acid /eicosanoic acid\n",
    "        \"FA_hs_20_11Z14Z17Z_c\": 0.0215,  # (11Z,14Z,17Z)-eicosatrienoic acid\n",
    "        \"FA_hs_20_11Z14Z_c\": 0.0004,  # (11Z,14Z)-eicosadienoic acid\n",
    "        \"FA_hs_20_11Z_c\": 0.0004,  # cis-gondoic acid\n",
    "        \"FA_hs_20_13Z_c\": 0.0004,  # (13Z)-eicosenoic acid\n",
    "        \"FA_hs_20_5Z8Z11Z14Z17Z_c\": 0.0116,  # EPA\n",
    "        \"FA_hs_20_5Z8Z11Z14Z_c\": 0.1083,  # arachidonoic acid\n",
    "        \"FA_hs_20_5Z8Z11Z_c\": 0.0004,  # mead acid\n",
    "        \"FA_hs_20_8Z11Z14Z17Z_c\": 0.0115,  # omega-3-arachidonic acid\n",
    "        \"FA_hs_20_8Z11Z14Z_c\": 0.0215,  #  dihomo-gamma-linoleic acid\n",
    "        \"FA_hs_20_8Z11Z_c\": 0.0004,  # 8,11-eicosadienoic acid\n",
    "        \"FA_hs_20_9Z_c\": 0.0004,  # 9-eicosenoic acid\n",
    "        \"FA_hs_21_0_c\": 0.0004,  # henicosanoic acid\n",
    "        \"FA_hs_22_0_c\": 0.0004,  # behenic acid\n",
    "        \"FA_hs_22_10Z13Z16Z19Z_c\": 0.0004,  # 10,13,16,19-docosatetraenoic acid\n",
    "        \"FA_hs_22_10Z13Z16Z_c\": 0.0004,  # 10,13,16-docosatriynoic acid\n",
    "        \"FA_hs_22_11Z_c\": 0.0004,  # cis-cetoleic acid\n",
    "        \"FA_hs_22_13Z16Z19Z_c\": 0.0004,  # 13,16,19-docosatrienoic acid\n",
    "        \"FA_hs_22_13Z16Z_c\": 0.0004,  # (13Z,16Z)-docosadienoic acid\n",
    "        \"FA_hs_22_13Z_c\": 0.0004,  # cis-erucic acid\n",
    "        \"FA_hs_22_4Z7Z10Z13Z16Z19Z_c\": 0.0278,  # DHA\n",
    "        \"FA_hs_22_4Z7Z10Z13Z16Z_c\": 0.0038,  # (4Z,7Z,10Z,13Z,16Z)-DPA\n",
    "        \"FA_hs_22_7Z10Z13Z16Z19Z_c\": 0.0059,  # DPA\n",
    "        \"FA_hs_22_7Z10Z13Z16Z_c\": 0.0014,  # adrenic acid\n",
    "        \"FA_hs_23_0_c\": 0.0004,  # tricosanoic acid\n",
    "        \"FA_hs_24_0_c\": 0.0004,  # lignocerate\n",
    "        \"FA_hs_24_12Z15Z18Z21Z_c\": 0.0004,  # 12,15,18,21-tetracosatetraenoic acid\n",
    "        \"FA_hs_24_15Z_c\": 0.0004,  # nervonic acid\n",
    "        \"FA_hs_24_6Z9Z12Z15Z18Z21Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z,21Z)-THA\n",
    "        \"FA_hs_24_6Z9Z12Z15Z18Z_c\": 0.0004,  # (6Z,9Z,12Z,15Z,18Z)-TPA\n",
    "        \"FA_hs_24_9Z12Z15Z18Z21Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z,21Z)-TPA\n",
    "        \"FA_hs_24_9Z12Z15Z18Z_c\": 0.0004,  # (9Z,12Z,15Z,18Z)-TTA\n",
    "        \"FA_hs_26_0_c\": 0.0004,  # cerotic acid\n",
    "        \"FA_hs_26_17Z_c\": 0.0004,  # ximenic acid\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "666c9e41-7b07-4f0b-a5a0-518b4d2fe503",
   "metadata": {},
   "source": [
    "#### Update annotation mappings\n",
    "Map to identifiers of database initially to be able to extract annotations from those databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8692a4fd-db0b-4467-a2fb-9e27e7f36103",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metabolite_mappings = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / \"metabolites_InitialMappings_1.0.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=str,\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "df_metabolite_mappings = df_metabolite_mappings.set_index(\"metabolites\")\n",
    "for metabolite, row in df_metabolite_mappings.iterrows():\n",
    "    metabolite = model.metabolites.get_by_id(metabolite)\n",
    "    for key in df_metabolite_mappings.columns:\n",
    "        item = row[key]\n",
    "        if item:\n",
    "            metabolite.annotation[key] = item\n",
    "# Clean up mapping file\n",
    "df_metabolite_mappings = df_metabolite_mappings.sort_index().reset_index(drop=False)\n",
    "df_metabolite_mappings.to_csv(\n",
    "    get_dirpath(\"curation\") / \"metabolites_InitialMappings_1.0.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "\n",
    "df_reaction_mappings = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / \"reactions_InitialMappings_1.0.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    dtype=str,\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "df_reaction_mappings = df_reaction_mappings.set_index(\"reactions\")\n",
    "for reaction, row in df_reaction_mappings.iterrows():\n",
    "    reaction = model.reactions.get_by_id(reaction)\n",
    "    for key in df_reaction_mappings.columns:\n",
    "        item = row[key]\n",
    "        if item:\n",
    "            reaction.annotation[key] = item\n",
    "\n",
    "# Clean up mapping file\n",
    "df_reaction_mappings = df_reaction_mappings.sort_index().reset_index(drop=False)\n",
    "df_reaction_mappings.to_csv(\n",
    "    get_dirpath(\"curation\") / \"reactions_InitialMappings_1.0.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76cf33ba-798a-4513-a799-877457ca2f8c",
   "metadata": {},
   "source": [
    "#### Reset subsystem groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d5cddb5d-8226-4086-85f8-9ed52e035bbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "model.remove_groups(model.groups)\n",
    "for subsystem in sorted(set(model.reactions.list_attr(\"subsystem\"))):\n",
    "    reaction_list = model.reactions.query(lambda x: x.subsystem == subsystem)\n",
    "    if subsystem not in model.groups:\n",
    "        group = Group(id=subsystem, name=subsystem, members=reaction_list)\n",
    "        model.add_groups([group])\n",
    "    else:\n",
    "        group = model.groups.get_by_id(subsystem).add_members(reaction_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7b21d3f-3444-499c-83ea-faef86cd24fa",
   "metadata": {},
   "source": [
    "### Check mass balancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3f69d937-06b5-489b-bee2-af836bc25668",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6LTHPI: 6lthp_c --> h_c + sppt_c\n",
      "{'charge': 2}\n",
      "\n",
      "METHBCYTBR: 2.0 focytb5_c + methb_c --> 2.0 ficytb5_c + hb_c\n",
      "{'charge': 1.0}\n",
      "\n",
      "METHBFMNR: fmnh2_c + methb_c --> fmn_c + 3.0 h_c + hb_c\n",
      "{'charge': 1.0}\n",
      "\n",
      "POOL_FACOA: FAcoa_hs_c <=> 0.0004 FAcoa_hs_12_0_c + 0.0004 FAcoa_hs_13_0_c + 0.0133 FAcoa_hs_14_0_c + 0.0004 FAcoa_hs_14_5Z_c + 0.0004 FAcoa_hs_14_7Z_c + 0.0004 FAcoa_hs_14_9Z_c + 0.0004 FAcoa_hs_15_0_c + 0.222 FAcoa_hs_16_0_c + 0.0004 FAcoa_hs_16_7Z_c + 0.0219 FAcoa_hs_16_9Z_c + 0.0004 FAcoa_hs_17_0_c + 0.0004 FAcoa_hs_17_10Z_c + 0.0004 FAcoa_hs_17_9Z_c + 0.1498 FAcoa_hs_18_0_c + 0.025 FAcoa_hs_18_11Z_c + 0.0004 FAcoa_hs_18_13Z_c + 0.0025 FAcoa_hs_18_6Z9Z12Z15Z_c + 0.0029 FAcoa_hs_18_6Z9Z12Z_c + 0.0004 FAcoa_hs_18_6Z9Z_c + 0.0004 FAcoa_hs_18_7Z_c + 0.0004 FAcoa_hs_18_9E_c + 0.0084 FAcoa_hs_18_9Z12Z15Z_c + 0.1915 FAcoa_hs_18_9Z12Z_c + 0.1545 FAcoa_hs_18_9Z_c + 0.0004 FAcoa_hs_19_0_c + 0.0004 FAcoa_hs_20_0_c + 0.0215 FAcoa_hs_20_11Z14Z17Z_c + 0.0004 FAcoa_hs_20_11Z14Z_c + 0.0004 FAcoa_hs_20_11Z_c + 0.0004 FAcoa_hs_20_13Z_c + 0.0116 FAcoa_hs_20_5Z8Z11Z14Z17Z_c + 0.1083 FAcoa_hs_20_5Z8Z11Z14Z_c + 0.0004 FAcoa_hs_20_5Z8Z11Z_c + 0.0115 FAcoa_hs_20_8Z11Z14Z17Z_c + 0.0215 FAcoa_hs_20_8Z11Z14Z_c + 0.0004 FAcoa_hs_20_8Z11Z_c + 0.0004 FAcoa_hs_20_9Z_c + 0.0004 FAcoa_hs_21_0_c + 0.0004 FAcoa_hs_22_0_c + 0.0004 FAcoa_hs_22_10Z13Z16Z19Z_c + 0.0004 FAcoa_hs_22_10Z13Z16Z_c + 0.0004 FAcoa_hs_22_11Z_c + 0.0004 FAcoa_hs_22_13Z16Z19Z_c + 0.0004 FAcoa_hs_22_13Z16Z_c + 0.0004 FAcoa_hs_22_13Z_c + 0.0278 FAcoa_hs_22_4Z7Z10Z13Z16Z19Z_c + 0.0038 FAcoa_hs_22_4Z7Z10Z13Z16Z_c + 0.0059 FAcoa_hs_22_7Z10Z13Z16Z19Z_c + 0.0014 FAcoa_hs_22_7Z10Z13Z16Z_c + 0.0004 FAcoa_hs_23_0_c + 0.0004 FAcoa_hs_24_0_c + 0.0004 FAcoa_hs_24_12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_15Z_c + 0.0004 FAcoa_hs_24_6Z9Z12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_6Z9Z12Z15Z18Z_c + 0.0004 FAcoa_hs_24_9Z12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_9Z12Z15Z18Z_c + 0.0004 FAcoa_hs_26_0_c + 0.0004 FAcoa_hs_26_17Z_c\n",
      "{'charge': -0.08439999999999875, 'C': 17.81750000000001, 'H': 33.302200000000006, 'N': 0.14769999999999742, 'O': 0.3586999999999909, 'P': 0.06329999999999877, 'R': -1, 'S': 0.02109999999999969}\n",
      "\n",
      "POOL_FA: FA_hs_c <=> 0.0004 FA_hs_12_0_c + 0.0004 FA_hs_13_0_c + 0.0133 FA_hs_14_0_c + 0.0004 FA_hs_14_5Z_c + 0.0004 FA_hs_14_7Z_c + 0.0004 FA_hs_14_9Z_c + 0.0004 FA_hs_15_0_c + 0.222 FA_hs_16_0_c + 0.0004 FA_hs_16_7Z_c + 0.0219 FA_hs_16_9Z_c + 0.0004 FA_hs_17_0_c + 0.0004 FA_hs_17_10Z_c + 0.0004 FA_hs_17_9Z_c + 0.1498 FA_hs_18_0_c + 0.025 FA_hs_18_11Z_c + 0.0004 FA_hs_18_13Z_c + 0.0025 FA_hs_18_6Z9Z12Z15Z_c + 0.0029 FA_hs_18_6Z9Z12Z_c + 0.0004 FA_hs_18_6Z9Z_c + 0.0004 FA_hs_18_7Z_c + 0.0004 FA_hs_18_9E_c + 0.0084 FA_hs_18_9Z12Z15Z_c + 0.1915 FA_hs_18_9Z12Z_c + 0.1545 FA_hs_18_9Z_c + 0.0004 FA_hs_19_0_c + 0.0004 FA_hs_20_0_c + 0.0215 FA_hs_20_11Z14Z17Z_c + 0.0004 FA_hs_20_11Z14Z_c + 0.0004 FA_hs_20_11Z_c + 0.0004 FA_hs_20_13Z_c + 0.0116 FA_hs_20_5Z8Z11Z14Z17Z_c + 0.1083 FA_hs_20_5Z8Z11Z14Z_c + 0.0004 FA_hs_20_5Z8Z11Z_c + 0.0115 FA_hs_20_8Z11Z14Z17Z_c + 0.0215 FA_hs_20_8Z11Z14Z_c + 0.0004 FA_hs_20_8Z11Z_c + 0.0004 FA_hs_20_9Z_c + 0.0004 FA_hs_21_0_c + 0.0004 FA_hs_22_0_c + 0.0004 FA_hs_22_10Z13Z16Z19Z_c + 0.0004 FA_hs_22_10Z13Z16Z_c + 0.0004 FA_hs_22_11Z_c + 0.0004 FA_hs_22_13Z16Z19Z_c + 0.0004 FA_hs_22_13Z16Z_c + 0.0004 FA_hs_22_13Z_c + 0.0278 FA_hs_22_4Z7Z10Z13Z16Z19Z_c + 0.0038 FA_hs_22_4Z7Z10Z13Z16Z_c + 0.0059 FA_hs_22_7Z10Z13Z16Z19Z_c + 0.0014 FA_hs_22_7Z10Z13Z16Z_c + 0.0004 FA_hs_23_0_c + 0.0004 FA_hs_24_0_c + 0.0004 FA_hs_24_12Z15Z18Z21Z_c + 0.0004 FA_hs_24_15Z_c + 0.0004 FA_hs_24_6Z9Z12Z15Z18Z21Z_c + 0.0004 FA_hs_24_6Z9Z12Z15Z18Z_c + 0.0004 FA_hs_24_9Z12Z15Z18Z21Z_c + 0.0004 FA_hs_24_9Z12Z15Z18Z_c + 0.0004 FA_hs_26_0_c + 0.0004 FA_hs_26_17Z_c\n",
      "{'charge': -0.02109999999999969, 'C': 17.374399999999987, 'O': 0.04219999999999938, 'R': -1, 'H': 32.6481}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for reaction in model.reactions:\n",
    "    if reaction.boundary:\n",
    "        continue\n",
    "    if reaction.check_mass_balance():\n",
    "        print(reaction)\n",
    "        print(reaction.check_mass_balance())\n",
    "        print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eaeae0fb-1f5b-4375-8510-f0f28c79f905",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d5f9e94e-1123-46d6-8352-ba96d2af7682",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>148d7cfd0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1984</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>2823</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>693</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>76</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x148d7cfd0>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_cobra_model(\n",
    "    filename=get_dirpath(\"model\", use_temp=\"interim\" if not overwrite else None)\n",
    "    / f\"{model.id}.xml\"\n",
    ")\n",
    "write_cobra_model(\n",
    "    filename=get_dirpath(\"model\", use_temp=\"interim\" if not overwrite else None)\n",
    "    / f\"{model.id}.json\"\n",
    ")\n",
    "model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
