{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ba9c3d2c-72da-416c-8612-8a3f91eee20a",
   "metadata": {},
   "source": [
    "# RBC-GEM 1.3.0 Updates\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "194c1f88-6d9b-40d5-b468-86da5bc66f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from warnings import warn\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from cobra.core import Gene, Metabolite, Reaction\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    GEM_NAME,\n",
    "    build_string,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    split_string,\n",
    "    write_cobra_model,\n",
    ")\n",
    "from rbc_gem_utils.qc import (\n",
    "    reset_reaction_bounds,\n",
    "    reset_subsystem_groups,\n",
    "    standardardize_metabolite_formulas,\n",
    ")\n",
    "from rbc_gem_utils.util import strip_plural"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "945a5ce1-932d-4622-a823-c54b76b48d99",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c9d3e47-f2f1-44e6-9333-4adec35ef021",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "907271c3-244f-4dbe-8539-bfd8adcb4fb3",
   "metadata": {},
   "source": [
    "## Set notebook options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c935e77-d22e-48bc-a9a9-8983392a649c",
   "metadata": {},
   "outputs": [],
   "source": [
    "overwrite = False\n",
    "update_version = \"1.3.0\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f967b496-7302-49a9-a474-45b13dc70980",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 1.2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "787d69b8-f885-4cda-9598-505f2b72b8c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Set parameter LicenseID to value 2664189\n",
      "Academic license - for non-commercial use only - expires 2026-05-12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>11d3fc610</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x11d3fc610>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.yml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a60f5b7-60f2-4300-94ce-77f2b0951345",
   "metadata": {},
   "source": [
    "## Load Dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d081c9f6-3475-4a71-bc23-7352e08b4076",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_types = [\"metabolites\", \"genes\", \"reactions\"]\n",
    "retired_shorthands = [\"met\", \"gene\", \"rxn\"]\n",
    "\n",
    "dataframes_updated = {}\n",
    "dataframes_removed = {}\n",
    "\n",
    "# For export\n",
    "dataframes_evidence = {}\n",
    "dataframes_removals = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c1dd6fa-b301-462c-9bb2-21ce46f82a46",
   "metadata": {},
   "source": [
    "### Dataframes for updates and removals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "912e5f90-331c-49ee-a2cd-5e2f4fca3026",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_updated = dict(\n",
    "    (\n",
    "        attribute_type, \n",
    "        pd.read_csv(\n",
    "            get_dirpath(\"curation\") / f\"{attribute_type}_updated_{update_version}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "        )\n",
    "    )\n",
    "    if (get_dirpath(\"curation\") / f\"{attribute_type}_updated_{update_version}.tsv\").exists()\n",
    "    else (\n",
    "        attribute_type, \n",
    "        pd.DataFrame()\n",
    "    )\n",
    "    for attribute_type in attribute_types\n",
    ")\n",
    "\n",
    "dataframes_removed = dict(\n",
    "    (\n",
    "        attribute_type, \n",
    "        pd.read_csv(\n",
    "            get_dirpath(\"curation\") / f\"{attribute_type}_removed_{update_version}.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "\n",
    "        )\n",
    "    )\n",
    "    if (get_dirpath(\"curation\") / f\"{attribute_type}_removed_{update_version}.tsv\").exists()\n",
    "    else (\n",
    "        attribute_type, \n",
    "        pd.DataFrame()\n",
    "    )\n",
    "    for attribute_type in attribute_types\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef70f39b-d83b-47fe-8893-088b6d95429a",
   "metadata": {},
   "source": [
    "#### Determine IDs to update"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e5978532-a50e-4a53-9626-911199d561d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'metabolites': {}, 'genes': {}, 'reactions': {'DIHOMEt1213e': 'DIHOME1213te'}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapping_dicts = dict(\n",
    "    (attribute_type, {})\n",
    "    if dataframes_updated[attribute_type].empty or dataframes_updated[attribute_type][[attribute_type, \"retired\"]].dropna().empty\n",
    "    else (\n",
    "        attribute_type,\n",
    "        dataframes_updated[attribute_type][[attribute_type, \"retired\"]].dropna().set_index(\"retired\")[attribute_type].to_dict()\n",
    "    )\n",
    "    for attribute_type in attribute_types\n",
    ")\n",
    "id_mapping_dicts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d774ce5e-5cc7-4bb1-bf3f-68fe2e92047b",
   "metadata": {},
   "source": [
    "#### Load omic evidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4aefc10e-82d5-4a0c-ab52-23758fc41278",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>proteomic evidence (pubmed)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P04406</th>\n",
       "      <td>28</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1686133...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P11142</th>\n",
       "      <td>26</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1686133...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q00013</th>\n",
       "      <td>25</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1686133...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00915</th>\n",
       "      <td>25</td>\n",
       "      <td>pubmed:14963112;pubmed:16861337;pubmed:1849451...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P30043</th>\n",
       "      <td>24</td>\n",
       "      <td>pubmed:12362340;pubmed:14963112;pubmed:1861456...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9Y5Z0</th>\n",
       "      <td>1</td>\n",
       "      <td>pubmed:31552303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P55011</th>\n",
       "      <td>1</td>\n",
       "      <td>pubmed:31552303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53701</th>\n",
       "      <td>1</td>\n",
       "      <td>pubmed:37942280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53801</th>\n",
       "      <td>1</td>\n",
       "      <td>pubmed:37942280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P53805</th>\n",
       "      <td>1</td>\n",
       "      <td>pubmed:38964323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4696 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        proteomic evidence (#studies)  \\\n",
       "P04406                             28   \n",
       "P11142                             26   \n",
       "Q00013                             25   \n",
       "P00915                             25   \n",
       "P30043                             24   \n",
       "...                               ...   \n",
       "Q9Y5Z0                              1   \n",
       "P55011                              1   \n",
       "P53701                              1   \n",
       "P53801                              1   \n",
       "P53805                              1   \n",
       "\n",
       "                              proteomic evidence (pubmed)  \n",
       "P04406  pubmed:12362340;pubmed:14963112;pubmed:1686133...  \n",
       "P11142  pubmed:12362340;pubmed:14963112;pubmed:1686133...  \n",
       "Q00013  pubmed:12362340;pubmed:14963112;pubmed:1686133...  \n",
       "P00915  pubmed:14963112;pubmed:16861337;pubmed:1849451...  \n",
       "P30043  pubmed:12362340;pubmed:14963112;pubmed:1861456...  \n",
       "...                                                   ...  \n",
       "Q9Y5Z0                                    pubmed:31552303  \n",
       "P55011                                    pubmed:31552303  \n",
       "P53701                                    pubmed:37942280  \n",
       "P53801                                    pubmed:37942280  \n",
       "P53805                                    pubmed:38964323  \n",
       "\n",
       "[4696 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_proteomic_evidence = pd.read_csv(\n",
    "    get_dirpath(\"proteomics\", use_temp=\"external\") / \"proteomic_evidence_table.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "df_proteomic_evidence = pd.concat(\n",
    "    (\n",
    "        df_proteomic_evidence.sum(axis=1),\n",
    "        pd.DataFrame.from_dict(\n",
    "            {\n",
    "                uniprot_id: build_string(\n",
    "                    sorted(\n",
    "                        [\n",
    "                            f\"{pubmed}\"\n",
    "                            for pubmed, is_detected in value_dict.items()\n",
    "                            if bool(is_detected)\n",
    "                        ]\n",
    "                    )\n",
    "                )\n",
    "                for uniprot_id, value_dict in df_proteomic_evidence.T.to_dict().items()\n",
    "            },\n",
    "            orient=\"index\",\n",
    "        ),\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "df_proteomic_evidence.columns = [\n",
    "    \"proteomic evidence (#studies)\",\n",
    "    \"proteomic evidence (pubmed)\",\n",
    "]\n",
    "df_proteomic_evidence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a1c443b-d26f-41e6-a674-30a1c91900db",
   "metadata": {},
   "source": [
    "## Apply updates\n",
    "Update order: \n",
    "1. Metabolites\n",
    "2. Genes\n",
    "3. Reactions\n",
    "### Deprecate identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "faa09151-f2d0-47fd-b95b-51a731d20580",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No identifiers to deprecate/update for metabolites.\n",
      "No identifiers to deprecate/update for genes.\n"
     ]
    }
   ],
   "source": [
    "for attribute_type, col_key in zip(attribute_types, retired_shorthands):\n",
    "    id_mapping_dict = id_mapping_dicts[attribute_type]\n",
    "    if not id_mapping_dict:\n",
    "        print(f\"No identifiers to deprecate/update for {attribute_type}.\")\n",
    "        continue\n",
    "    id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "    id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "    id_mapping_df.columns = [f\"{col_key}Retired\", f\"{col_key}s\"]\n",
    "    # Reverse column irder for mapping\n",
    "    id_mapping_df = id_mapping_df.loc[:, list(id_mapping_df.columns[::-1])]\n",
    "\n",
    "    # Get DataFrame for previous deprecated IDs\n",
    "    previous_id_mapping_df = pd.read_csv(\n",
    "        get_dirpath(\"deprecatedIdentifiers\") / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "    )\n",
    "    \n",
    "    for idx, row in id_mapping_df.iterrows():\n",
    "        new_id, retiring = row[list(id_mapping_df.columns)]\n",
    "        previously_retired = previous_id_mapping_df[previous_id_mapping_df[f\"{col_key}s\"] == retiring]\n",
    "        retired_set_of_ids = {retiring}\n",
    "        if not previously_retired.empty:\n",
    "            # Get all previously retired IDs\n",
    "            retired_set_of_ids.update(previously_retired[f\"{col_key}Retired\"].apply(split_string).item())\n",
    "            # Pulling the ID out of retirement\n",
    "            if new_id in retired_set_of_ids:\n",
    "                retired_set_of_ids.remove(new_id)\n",
    "            retired_set_of_ids.add(retiring)\n",
    "        id_mapping_df.loc[idx, f\"{col_key}Retired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "     # Replace ID\n",
    "    for old, new in id_mapping_dict.items():\n",
    "        try:\n",
    "            obj = getattr(model, attribute_type).get_by_id(old)\n",
    "        except KeyError:\n",
    "            print(f\"Could not map {old} to new ID.\")\n",
    "        else:\n",
    "            try:\n",
    "                obj.id = id_mapping_dict[obj.id]\n",
    "            except ValueError as e:\n",
    "                warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n",
    "                getattr(model, f\"remove_{attribute_type}\")([obj])\n",
    "    # Repair model \n",
    "    model.repair(rebuild_index=True, rebuild_relationships=True)\n",
    "    # Add to DataFrane\n",
    "    id_mapping_df = pd.concat((id_mapping_df, previous_id_mapping_df), axis=0)\n",
    "    id_mapping_df = (\n",
    "        id_mapping_df.drop_duplicates()\n",
    "        .sort_values(f\"{col_key}s\", ascending=True)\n",
    "        .reset_index(drop=True)\n",
    "    )\n",
    "    get_dirpath(\"deprecatedIdentifiers\", use_temp=\"interim\" if not overwrite else None).mkdir(exist_ok=True, parents=True)\n",
    "    id_mapping_df.to_csv(\n",
    "        get_dirpath(\"deprecatedIdentifiers\", use_temp=\"interim\" if not overwrite else None) / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index=False,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b32955-a0d5-45c6-81e4-183760ce44d0",
   "metadata": {},
   "source": [
    "#### Metabolites\n",
    "* Removed odd symbol characters from names:\n",
    "    * polyadprib1, polyadprib2\n",
    "* Compartment corrections:\n",
    "    * e217obglcur_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aec9d8bb-9581-4278-a2db-f24a7c4425c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>metabolite</th>\n",
       "      <th>name</th>\n",
       "      <th>formula</th>\n",
       "      <th>charge</th>\n",
       "      <th>compartment</th>\n",
       "      <th>metabolomic evidence (#studies)</th>\n",
       "      <th>metabolomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1065</th>\n",
       "      <td>e217obglcur_c</td>\n",
       "      <td>e217obglcur_c</td>\n",
       "      <td>17beta-estradiol 17-O-(beta-D-glucuronate)</td>\n",
       "      <td>C24H31O8</td>\n",
       "      <td>-1</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1754</th>\n",
       "      <td>polyadprib1_c</td>\n",
       "      <td>polyadprib1</td>\n",
       "      <td>[(1--&gt;2)-ADP-alpha-D-ribose](n-1)</td>\n",
       "      <td>C15H21N5O14P2X</td>\n",
       "      <td>-2</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1755</th>\n",
       "      <td>polyadprib2_c</td>\n",
       "      <td>polyadprib2</td>\n",
       "      <td>[(1--&gt;2)-ADP-alpha-D-ribose](n)</td>\n",
       "      <td>C30H40N10O27P4X</td>\n",
       "      <td>-4</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        metabolites     metabolite  \\\n",
       "1065  e217obglcur_c  e217obglcur_c   \n",
       "1754  polyadprib1_c    polyadprib1   \n",
       "1755  polyadprib2_c    polyadprib2   \n",
       "\n",
       "                                            name          formula  charge  \\\n",
       "1065  17beta-estradiol 17-O-(beta-D-glucuronate)         C24H31O8      -1   \n",
       "1754           [(1-->2)-ADP-alpha-D-ribose](n-1)   C15H21N5O14P2X      -2   \n",
       "1755             [(1-->2)-ADP-alpha-D-ribose](n)  C30H40N10O27P4X      -4   \n",
       "\n",
       "     compartment  metabolomic evidence (#studies)  \\\n",
       "1065           c                             <NA>   \n",
       "1754           c                             <NA>   \n",
       "1755           c                             <NA>   \n",
       "\n",
       "      metabolomic evidence (pubmed)  references  notes  \n",
       "1065                           <NA>        <NA>   <NA>  \n",
       "1754                           <NA>        <NA>   <NA>  \n",
       "1755                           <NA>        <NA>   <NA>  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "attr_cols = [\"name\", \"formula\", \"charge\", \"compartment\"]\n",
    "obj_type = Metabolite\n",
    "\n",
    "df_updated = dataframes_updated[attribute_type]\n",
    "if not df_updated.empty:\n",
    "    df_updated = df_updated.drop(\"retired\", axis=1).set_index(attribute_type).convert_dtypes()\n",
    "    for idx, row in df_updated.iterrows():\n",
    "        obj_id = idx\n",
    "        # Deprecated IDs should be updated at this point, only refers to new metabolites\n",
    "        try:\n",
    "            has_id = getattr(model, attribute_type).has_id(obj_id)\n",
    "            if not has_id:\n",
    "                if attribute_type == \"genes\":\n",
    "                    # No 'add_genes' method in cobrapy\n",
    "                    add_method = getattr(model, f\"{attribute_type}\").extend\n",
    "                else:\n",
    "                    add_method = getattr(model, f\"add_{attribute_type}\")\n",
    "                add_method([obj_type(obj_id)])\n",
    "        except ValueError as e:\n",
    "            print(f\"Error with {obj_id}\")\n",
    "            raise e\n",
    "\n",
    "        obj = getattr(model, attribute_type).get_by_id(obj_id)\n",
    "        for attr, value in zip(attr_cols, row.convert_dtypes()[attr_cols].fillna(\"\")):\n",
    "            if value:\n",
    "                attr = attr.replace(\" \", \"_\")\n",
    "                setattr(obj, attr, value)\n",
    "        annotations_dict = {\n",
    "            k: v for k, v in row.convert_dtypes()[~row.index.isin([attribute_type] + attr_cols)].to_dict().items()\n",
    "            if not \"evidence\" in k and (v and v != float(\"nan\"))\n",
    "        }\n",
    "        try:\n",
    "            notes_dict = {\"notes\": annotations_dict.pop(\"notes\")}\n",
    "        except KeyError:\n",
    "            notes_dict = obj.notes\n",
    "        obj.annotation.update(annotations_dict)\n",
    "        obj.notes.update(notes_dict)\n",
    "\n",
    "     # Update evidence\n",
    "    try:\n",
    "        df_previous_evidence = pd.read_csv(\n",
    "            get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "    \n",
    "    # Replace previous IDs with new ones\n",
    "    df_previous_evidence[attribute_type] = df_previous_evidence[attribute_type].replace(\n",
    "        id_mapping_dicts[attribute_type]\n",
    "    )\n",
    "    # Replace non-empty values with previous ones\n",
    "    df_previous_values = df_previous_evidence[df_previous_evidence[attribute_type].isin(df_updated.index)].set_index(attribute_type).convert_dtypes()\n",
    "    cols_to_update = list(df_previous_evidence.columns[~df_previous_evidence.columns.isin(attr_cols + [attribute_type, strip_plural(attribute_type)])])\n",
    "    # Evidence can be problematic due to data coercion \n",
    "    cols_to_update = [c for c in cols_to_update if \"evidence\" not in c]\n",
    "\n",
    "    # Set indices to align\n",
    "    df_updated.loc[list(df_updated.index), cols_to_update] = df_updated.loc[list(df_updated.index), cols_to_update].fillna(df_previous_values.loc[list(df_updated.index), cols_to_update], method=None, axis=1)\n",
    "    # Reset index of updated DataFrame\n",
    "    df_updated = df_updated.reset_index(drop=False)\n",
    "\n",
    "    # Add updated entries to DataFrame\n",
    "    df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0).replace(\"nan\", str).convert_dtypes()\n",
    "    df_evidence = (\n",
    "        df_evidence.drop_duplicates(subset=[attribute_type], keep=\"first\")\n",
    "        .sort_values(attribute_type, ascending=True)\n",
    "        .reset_index(drop=True)\n",
    "    )[df_previous_evidence.columns]\n",
    "    dataframes_evidence[attribute_type] = df_evidence\n",
    "df_evidence[df_evidence[attribute_type].isin(df_updated[attribute_type])]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eaa1ed6-3934-4014-85dc-c4d3a7175cf7",
   "metadata": {},
   "source": [
    "##### Standardize metabolite formulas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "385822c4-5487-44cc-a8b6-6f7389e5882f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>metabolite</th>\n",
       "      <th>name</th>\n",
       "      <th>formula</th>\n",
       "      <th>charge</th>\n",
       "      <th>compartment</th>\n",
       "      <th>metabolomic evidence (#studies)</th>\n",
       "      <th>metabolomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10fthf_c</td>\n",
       "      <td>10fthf</td>\n",
       "      <td>10-formyl-tetrahydrofolate</td>\n",
       "      <td>C20H21N7O7</td>\n",
       "      <td>-2</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12ppd__R_c</td>\n",
       "      <td>12ppd__R</td>\n",
       "      <td>(R)-propane-1,2-diol</td>\n",
       "      <td>C3H8O2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12ppd__S_c</td>\n",
       "      <td>12ppd__S</td>\n",
       "      <td>(S)-propane-1,2-diol</td>\n",
       "      <td>C3H8O2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13dampp_c</td>\n",
       "      <td>13dampp</td>\n",
       "      <td>1,3-diaminopropane</td>\n",
       "      <td>C3H12N2</td>\n",
       "      <td>2</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13dpg_c</td>\n",
       "      <td>13dpg</td>\n",
       "      <td>3-Phospho-D-glyceroyl phosphate</td>\n",
       "      <td>C3H4O10P2</td>\n",
       "      <td>-4</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2152</th>\n",
       "      <td>zn2_c</td>\n",
       "      <td>zn2</td>\n",
       "      <td>Zinc (II)</td>\n",
       "      <td>Zn</td>\n",
       "      <td>2</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2153</th>\n",
       "      <td>zn2_e</td>\n",
       "      <td>zn2</td>\n",
       "      <td>Zinc (II)</td>\n",
       "      <td>Zn</td>\n",
       "      <td>2</td>\n",
       "      <td>e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154</th>\n",
       "      <td>ztp_c</td>\n",
       "      <td>ztp</td>\n",
       "      <td>5-amino-4-imidazolecarboxamide riboside 5'-tri...</td>\n",
       "      <td>C9H13N4O14P3</td>\n",
       "      <td>-4</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2155</th>\n",
       "      <td>zymst_c</td>\n",
       "      <td>zymst</td>\n",
       "      <td>zymosterol</td>\n",
       "      <td>C27H44O</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2156</th>\n",
       "      <td>zymstnl_c</td>\n",
       "      <td>zymstnl</td>\n",
       "      <td>zymostenol</td>\n",
       "      <td>C27H46O</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2157 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     metabolites metabolite  \\\n",
       "0       10fthf_c     10fthf   \n",
       "1     12ppd__R_c   12ppd__R   \n",
       "2     12ppd__S_c   12ppd__S   \n",
       "3      13dampp_c    13dampp   \n",
       "4        13dpg_c      13dpg   \n",
       "...          ...        ...   \n",
       "2152       zn2_c        zn2   \n",
       "2153       zn2_e        zn2   \n",
       "2154       ztp_c        ztp   \n",
       "2155     zymst_c      zymst   \n",
       "2156   zymstnl_c    zymstnl   \n",
       "\n",
       "                                                   name       formula  charge  \\\n",
       "0                            10-formyl-tetrahydrofolate    C20H21N7O7      -2   \n",
       "1                                  (R)-propane-1,2-diol        C3H8O2       0   \n",
       "2                                  (S)-propane-1,2-diol        C3H8O2       0   \n",
       "3                                    1,3-diaminopropane       C3H12N2       2   \n",
       "4                       3-Phospho-D-glyceroyl phosphate     C3H4O10P2      -4   \n",
       "...                                                 ...           ...     ...   \n",
       "2152                                          Zinc (II)            Zn       2   \n",
       "2153                                          Zinc (II)            Zn       2   \n",
       "2154  5-amino-4-imidazolecarboxamide riboside 5'-tri...  C9H13N4O14P3      -4   \n",
       "2155                                         zymosterol       C27H44O       0   \n",
       "2156                                         zymostenol       C27H46O       0   \n",
       "\n",
       "     compartment  metabolomic evidence (#studies)  \\\n",
       "0              c                             <NA>   \n",
       "1              c                             <NA>   \n",
       "2              c                             <NA>   \n",
       "3              c                             <NA>   \n",
       "4              c                             <NA>   \n",
       "...          ...                              ...   \n",
       "2152           c                             <NA>   \n",
       "2153           e                             <NA>   \n",
       "2154           c                             <NA>   \n",
       "2155           c                             <NA>   \n",
       "2156           c                             <NA>   \n",
       "\n",
       "      metabolomic evidence (pubmed)  references  notes  \n",
       "0                              <NA>        <NA>   <NA>  \n",
       "1                              <NA>        <NA>   <NA>  \n",
       "2                              <NA>        <NA>   <NA>  \n",
       "3                              <NA>        <NA>   <NA>  \n",
       "4                              <NA>        <NA>   <NA>  \n",
       "...                             ...         ...    ...  \n",
       "2152                           <NA>        <NA>   <NA>  \n",
       "2153                           <NA>        <NA>   <NA>  \n",
       "2154                           <NA>        <NA>   <NA>  \n",
       "2155                           <NA>        <NA>   <NA>  \n",
       "2156                           <NA>        <NA>   <NA>  \n",
       "\n",
       "[2157 rows x 10 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "df_evidence = dataframes_evidence[attribute_type].set_index(attribute_type)\n",
    "met_formulas = standardardize_metabolite_formulas(\n",
    "    dict(zip(model.metabolites.list_attr(\"id\"), model.metabolites.list_attr(\"formula\")))\n",
    ")\n",
    "for mid, formula in met_formulas.items():\n",
    "    model.metabolites.get_by_id(mid).formula = formula\n",
    "    df_evidence.loc[mid, \"formula\"] = formula\n",
    "dataframes_evidence[attribute_type] = df_evidence.reset_index(drop=False)\n",
    "dataframes_evidence[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "142935ef-76b4-4ccb-8ba6-2a3f999c2f2b",
   "metadata": {},
   "source": [
    "#### Genes\n",
    "* Added the lost pubmed evidence for the following:\n",
    "    * SLCO1A2, SLCO2B1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b188598d-3b9d-4c40-9ecc-45752ba68d85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>name</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>proteomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>729</th>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>SO1A2</td>\n",
       "      <td>P46721</td>\n",
       "      <td>6579</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>730</th>\n",
       "      <td>SLCO2B1</td>\n",
       "      <td>SO2B1</td>\n",
       "      <td>O94956</td>\n",
       "      <td>11309</td>\n",
       "      <td>SLCO2B1</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       genes   name uniprot ncbigene hgnc.symbol  \\\n",
       "729  SLCO1A2  SO1A2  P46721     6579     SLCO1A2   \n",
       "730  SLCO2B1  SO2B1  O94956    11309     SLCO2B1   \n",
       "\n",
       "     proteomic evidence (#studies) proteomic evidence (pubmed)  \\\n",
       "729                           <NA>                        <NA>   \n",
       "730                           <NA>                        <NA>   \n",
       "\n",
       "          references notes  \n",
       "729  pubmed:27504015  <NA>  \n",
       "730  pubmed:27504015  <NA>  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "attr_cols = []\n",
    "obj_type = Gene\n",
    "\n",
    "df_updated = dataframes_updated[attribute_type]\n",
    "if not df_updated.empty:\n",
    "    df_updated = df_updated.drop(\"retired\", axis=1).set_index(attribute_type).convert_dtypes()\n",
    "    for idx, row in df_updated.iterrows():\n",
    "        obj_id = idx\n",
    "        # Deprecated IDs should be updated at this point, only refers to new metabolites\n",
    "        try:\n",
    "            has_id = getattr(model, attribute_type).has_id(obj_id)\n",
    "            if not has_id:\n",
    "                if attribute_type == \"genes\":\n",
    "                    # No 'add_genes' method in cobrapy\n",
    "                    add_method = getattr(model, f\"{attribute_type}\").extend\n",
    "                else:\n",
    "                    add_method = getattr(model, f\"add_{attribute_type}\")\n",
    "                add_method([obj_type(obj_id)])\n",
    "        except ValueError as e:\n",
    "            print(f\"Error with {obj_id}\")\n",
    "            raise e\n",
    "\n",
    "        obj = getattr(model, attribute_type).get_by_id(obj_id)\n",
    "        for attr, value in zip(attr_cols, row.convert_dtypes()[attr_cols].fillna(\"\")):\n",
    "            if value:\n",
    "                attr = attr.replace(\" \", \"_\")\n",
    "                setattr(obj, attr, value)\n",
    "        annotations_dict = {\n",
    "            k: v for k, v in row.convert_dtypes()[~row.index.isin([attribute_type] + attr_cols)].to_dict().items()\n",
    "            if not \"evidence\" in k and (v and v != float(\"nan\"))\n",
    "        }\n",
    "        try:\n",
    "            notes_dict = {\"notes\": annotations_dict.pop(\"notes\")}\n",
    "        except KeyError:\n",
    "            notes_dict = obj.notes\n",
    "        obj.annotation.update(annotations_dict)\n",
    "        obj.notes.update(notes_dict)\n",
    "\n",
    "     # Update evidence\n",
    "    try:\n",
    "        df_previous_evidence = pd.read_csv(\n",
    "            get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "    \n",
    "    # Replace previous IDs with new ones\n",
    "    df_previous_evidence[attribute_type] = df_previous_evidence[attribute_type].replace(\n",
    "        id_mapping_dicts[attribute_type]\n",
    "    )\n",
    "    # Replace non-empty values with previous ones\n",
    "    df_previous_values = df_previous_evidence[df_previous_evidence[attribute_type].isin(df_updated.index)].set_index(attribute_type).convert_dtypes()\n",
    "    cols_to_update = list(df_previous_evidence.columns[~df_previous_evidence.columns.isin(attr_cols + [attribute_type, strip_plural(attribute_type)])])\n",
    "    # Evidence can be problematic due to data coercion \n",
    "    cols_to_update = [c for c in cols_to_update if \"evidence\" not in c]\n",
    "\n",
    "    # Set indices to align\n",
    "    df_updated.loc[list(df_updated.index), cols_to_update] = df_updated.loc[list(df_updated.index), cols_to_update].fillna(df_previous_values.loc[list(df_updated.index), cols_to_update], method=None, axis=1)\n",
    "    # Reset index of updated DataFrame\n",
    "    df_updated = df_updated.reset_index(drop=False)\n",
    "\n",
    "    # Add updated entries to DataFrame\n",
    "    df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0).replace(\"nan\", str).convert_dtypes()\n",
    "    df_evidence = (\n",
    "        df_evidence.drop_duplicates(subset=[attribute_type], keep=\"first\")\n",
    "        .sort_values(attribute_type, ascending=True)\n",
    "        .reset_index(drop=True)\n",
    "    )[df_previous_evidence.columns]\n",
    "    dataframes_evidence[attribute_type] = df_evidence\n",
    "df_evidence[df_evidence[attribute_type].isin(df_updated[attribute_type])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f4b663bd-ec10-4b25-8185-774ddebad38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"genes\"\n",
    "df_evidence = dataframes_evidence[attribute_type].set_index(attribute_type)\n",
    "gene_names = df_evidence[\"name\"].to_dict()\n",
    "df_evidence = df_evidence.reset_index(drop=False)\n",
    "for gid, gname in gene_names.items():\n",
    "    model.genes.get_by_id(gid).name = gname"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ed7ece2-0414-4a4a-ae01-c54bda877529",
   "metadata": {},
   "source": [
    "##### Annotate evidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dc0d17ef-8853-4e7a-a055-97773294e767",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_evidence = dataframes_evidence[attribute_type]\n",
    "# columns_ordered = list(df_evidence.columns)\n",
    "# df_evidence = df_evidence.set_index(\"uniprot\")\n",
    "# df_omic_evidence = df_proteomic_evidence[df_proteomic_evidence.index.isin(df_evidence.index)]\n",
    "# df_evidence.loc[df_omic_evidence.index, df_omic_evidence.columns] = df_omic_evidence\n",
    "\n",
    "# df_no_evidence  = df_evidence[~df_evidence.index.isin(df_proteomic_evidence.index)][df_omic_evidence.columns]\n",
    "# df_no_evidence.loc[:, df_proteomic_evidence.columns[0]] = 0 \n",
    "# df_no_evidence.loc[:, df_proteomic_evidence.columns[1]] = pd.NA\n",
    "# df_evidence.loc[df_no_evidence.index, df_no_evidence.columns] = df_no_evidence\n",
    "\n",
    "# dataframes_evidence[attribute_type] = df_evidence.reset_index(drop=False).loc[:, columns_ordered]\n",
    "# dataframes_evidence[attribute_type]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3b1b3476-7c28-4164-a425-87ea22f92085",
   "metadata": {},
   "source": [
    "#### Reactions\n",
    "* Added the lost pubmed evidence for the following genes/proteins:\n",
    "    * CHOLATEt, CPPP3te, DGCHOLte, DHEASte, E217BGLCRte, ESTRONESte, GCHOLAte, GDCHOLAte, GUDCHOLAte, PGD2te, PGE2te, PRGNStec, TCHOLAte, TDCHOLAte, TDECHOLAte, TETIODTHYt2, TRIIODTHYt2, TUDCHOLAte\n",
    "    * Based on the protein evidence and UniProt database\n",
    "* Removed odd symbol characters from names:\n",
    "\t*  ACMAH\n",
    "* Updated subsystems for the following:\n",
    "    * PAFH, PAFS: Glycerophopsholipid metabolism --> Ether lipid metabolism"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a25730a7-5ab0-4827-8e18-18ffb1b3f4aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>reaction</th>\n",
       "      <th>gene reaction rule</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>spontaneous</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>ACMAH</td>\n",
       "      <td>N-acetylmethionyl L-alanine hydrolase</td>\n",
       "      <td>acmetala_c + h2o_c --&gt; acmet__L_c + ala__L_c</td>\n",
       "      <td>APEH</td>\n",
       "      <td>Peptide metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:10719179;pubmed:7275011</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>CHOLATEt</td>\n",
       "      <td>Transport of cholate</td>\n",
       "      <td>cholate_e &lt;=&gt; cholate_c</td>\n",
       "      <td>SLCO1A2 or SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>515</th>\n",
       "      <td>CPPP3te</td>\n",
       "      <td>Transport of coproporphyrin III</td>\n",
       "      <td>cppp3_e &lt;=&gt; cppp3_c</td>\n",
       "      <td>SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>762</th>\n",
       "      <td>DGCHOLte</td>\n",
       "      <td>Transport of glycochenodeoxycholate</td>\n",
       "      <td>dgchol_e &lt;=&gt; dgchol_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>DHEASte</td>\n",
       "      <td>Transport of dehydroepiandrosterone 3-sulfate</td>\n",
       "      <td>dheas_e &lt;=&gt; dheas_c</td>\n",
       "      <td>SLCO1A2 or SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>DIHOME1213te</td>\n",
       "      <td>Transport of 12(13)-DiHOME via diffusion</td>\n",
       "      <td>dihome1213_c --&gt; dihome1213_e</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:15294349;pubmed:17164127;pubmed:1765199...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>E217BGLCRte</td>\n",
       "      <td>Transport of 17beta-estradiol 17-O-(beta-D-glu...</td>\n",
       "      <td>e217obglcur_e &lt;=&gt; e217obglcur_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>ESTRONESte</td>\n",
       "      <td>Transport of estrone 3-sulfate</td>\n",
       "      <td>estrones_e &lt;=&gt; estrones_c</td>\n",
       "      <td>SLCO1A2 or SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1354</th>\n",
       "      <td>GCHOLAte</td>\n",
       "      <td>Transport of glycocholate</td>\n",
       "      <td>gchola_e &lt;=&gt; gchola_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1356</th>\n",
       "      <td>GDCHOLAte</td>\n",
       "      <td>Transport of glycodeoxycholate</td>\n",
       "      <td>gdchac_e &lt;=&gt; gdchac_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1497</th>\n",
       "      <td>GUDCHOLAte</td>\n",
       "      <td>Transport of glycoursodeoxycholate</td>\n",
       "      <td>gudchola_e &lt;=&gt; gudchola_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2096</th>\n",
       "      <td>PAFH</td>\n",
       "      <td>1-Alkyl-2-acetyl-sn-glycero-3-phosphocholine (...</td>\n",
       "      <td>h2o_c + paf_hs_c --&gt; ac_c + h_c + lpco_hs_c</td>\n",
       "      <td>(PAFAH1B1 and PAFAH1B3) or (PAFAH1B1 and PAFAH...</td>\n",
       "      <td>Ether lipid metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>PAFAH1B1(19);PAFAH1B2(16);PAFAH1B3(20)</td>\n",
       "      <td>pubmed:21844189;pubmed:8440681</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2097</th>\n",
       "      <td>PAFS</td>\n",
       "      <td>Lysoplas:Acetyl-CoA (LPC O) 2-O-acetyltransferase</td>\n",
       "      <td>accoa_c + lpco_hs_c --&gt; coa_c + paf_hs_c</td>\n",
       "      <td>LPCAT3 or LPCAT1</td>\n",
       "      <td>Ether lipid metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>LPCAT1(2);LPCAT3(11)</td>\n",
       "      <td>pubmed:18156367;pubmed:18285344</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2104</th>\n",
       "      <td>PCFLOPt</td>\n",
       "      <td>Phosphatidylcholine (PC) floppase</td>\n",
       "      <td>atp_c + h2o_c + pc_hs_c --&gt; adp_c + h_c + pc_h...</td>\n",
       "      <td>ABCB4 or ABCA7 or ABCB1 or ABCA1 or ABCC1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>ABCA7(7);ABCB1(1);ABCB4(1)</td>\n",
       "      <td>pubmed:10947968;pubmed:28134466;pubmed:9651491...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2136</th>\n",
       "      <td>PGD2te</td>\n",
       "      <td>Transport of prostaglandin D2</td>\n",
       "      <td>prostgd2_e &lt;=&gt; prostgd2_c</td>\n",
       "      <td>SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2149</th>\n",
       "      <td>PGE2te</td>\n",
       "      <td>Transport of prostaglandin E2</td>\n",
       "      <td>prostge2_e &lt;=&gt; prostge2_c</td>\n",
       "      <td>SLCO1A2 or SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2266</th>\n",
       "      <td>PRGNStec</td>\n",
       "      <td>Transport of pregnenolone sulfate</td>\n",
       "      <td>prgnlones_e &lt;=&gt; prgnlones_c</td>\n",
       "      <td>SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2487</th>\n",
       "      <td>SMFLOPt</td>\n",
       "      <td>Sphingomyelin floppase</td>\n",
       "      <td>atp_c + h2o_c + sphmyln_hs_c --&gt; adp_c + h_c +...</td>\n",
       "      <td>ABCB4 or ABCA7 or ABCB1 or ABCA1 or ABCC1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>ABCA7(7);ABCB1(1);ABCB4(1)</td>\n",
       "      <td>pubmed:10947968;pubmed:9778357;pubmed:14578045</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2539</th>\n",
       "      <td>TCHOLAte</td>\n",
       "      <td>Transport of taurocholate</td>\n",
       "      <td>tchola_e &lt;=&gt; tchola_c</td>\n",
       "      <td>SLCO1A2 or SLCO2B1</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2541</th>\n",
       "      <td>TDCHOLAte</td>\n",
       "      <td>Transport of taurochenodeoxycholate</td>\n",
       "      <td>tdchola_e &lt;=&gt; tdchola_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2542</th>\n",
       "      <td>TDECHOLAte</td>\n",
       "      <td>Transport of taurodeoxycholate</td>\n",
       "      <td>tdechola_e &lt;=&gt; tdechola_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2544</th>\n",
       "      <td>TETIODTHYt2</td>\n",
       "      <td>3,3',5,5'-Tetraiodo-L-thyronine transport</td>\n",
       "      <td>tetiodthy__L_c &lt;=&gt; tetiodthy__L_e</td>\n",
       "      <td>(SLC7A5 and SLC3A2) or SLC16A10 or SLCO1A2 or ...</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:34925607;pubmed:4479710;pubmed:7915653;...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2627</th>\n",
       "      <td>TRIIODTHYt2</td>\n",
       "      <td>3,3,3'-Triiodo-L-thyronine (T3) transport via ...</td>\n",
       "      <td>triiodthy__L_c &lt;=&gt; triiodthy__L_e</td>\n",
       "      <td>(SLC7A5 and SLC3A2) or SLC16A10 or SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:34925607;pubmed:4479710;pubmed:7915653;...</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2642</th>\n",
       "      <td>TUDCHOLAte</td>\n",
       "      <td>Transport of tauroursodeoxycholate</td>\n",
       "      <td>tudchola_e &lt;=&gt; tudchola_c</td>\n",
       "      <td>SLCO1A2</td>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>pubmed:27504015</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         reactions                                               name  \\\n",
       "216          ACMAH              N-acetylmethionyl L-alanine hydrolase   \n",
       "493       CHOLATEt                               Transport of cholate   \n",
       "515        CPPP3te                    Transport of coproporphyrin III   \n",
       "762       DGCHOLte                Transport of glycochenodeoxycholate   \n",
       "775        DHEASte      Transport of dehydroepiandrosterone 3-sulfate   \n",
       "798   DIHOME1213te           Transport of 12(13)-DiHOME via diffusion   \n",
       "881    E217BGLCRte  Transport of 17beta-estradiol 17-O-(beta-D-glu...   \n",
       "957     ESTRONESte                     Transport of estrone 3-sulfate   \n",
       "1354      GCHOLAte                          Transport of glycocholate   \n",
       "1356     GDCHOLAte                     Transport of glycodeoxycholate   \n",
       "1497    GUDCHOLAte                 Transport of glycoursodeoxycholate   \n",
       "2096          PAFH  1-Alkyl-2-acetyl-sn-glycero-3-phosphocholine (...   \n",
       "2097          PAFS  Lysoplas:Acetyl-CoA (LPC O) 2-O-acetyltransferase   \n",
       "2104       PCFLOPt                  Phosphatidylcholine (PC) floppase   \n",
       "2136        PGD2te                      Transport of prostaglandin D2   \n",
       "2149        PGE2te                      Transport of prostaglandin E2   \n",
       "2266      PRGNStec                  Transport of pregnenolone sulfate   \n",
       "2487       SMFLOPt                             Sphingomyelin floppase   \n",
       "2539      TCHOLAte                          Transport of taurocholate   \n",
       "2541     TDCHOLAte                Transport of taurochenodeoxycholate   \n",
       "2542    TDECHOLAte                     Transport of taurodeoxycholate   \n",
       "2544   TETIODTHYt2          3,3',5,5'-Tetraiodo-L-thyronine transport   \n",
       "2627   TRIIODTHYt2  3,3,3'-Triiodo-L-thyronine (T3) transport via ...   \n",
       "2642    TUDCHOLAte                 Transport of tauroursodeoxycholate   \n",
       "\n",
       "                                               reaction  \\\n",
       "216        acmetala_c + h2o_c --> acmet__L_c + ala__L_c   \n",
       "493                             cholate_e <=> cholate_c   \n",
       "515                                 cppp3_e <=> cppp3_c   \n",
       "762                               dgchol_e <=> dgchol_c   \n",
       "775                                 dheas_e <=> dheas_c   \n",
       "798                       dihome1213_c --> dihome1213_e   \n",
       "881                     e217obglcur_e <=> e217obglcur_c   \n",
       "957                           estrones_e <=> estrones_c   \n",
       "1354                              gchola_e <=> gchola_c   \n",
       "1356                              gdchac_e <=> gdchac_c   \n",
       "1497                          gudchola_e <=> gudchola_c   \n",
       "2096        h2o_c + paf_hs_c --> ac_c + h_c + lpco_hs_c   \n",
       "2097           accoa_c + lpco_hs_c --> coa_c + paf_hs_c   \n",
       "2104  atp_c + h2o_c + pc_hs_c --> adp_c + h_c + pc_h...   \n",
       "2136                          prostgd2_e <=> prostgd2_c   \n",
       "2149                          prostge2_e <=> prostge2_c   \n",
       "2266                        prgnlones_e <=> prgnlones_c   \n",
       "2487  atp_c + h2o_c + sphmyln_hs_c --> adp_c + h_c +...   \n",
       "2539                              tchola_e <=> tchola_c   \n",
       "2541                            tdchola_e <=> tdchola_c   \n",
       "2542                          tdechola_e <=> tdechola_c   \n",
       "2544                  tetiodthy__L_c <=> tetiodthy__L_e   \n",
       "2627                  triiodthy__L_c <=> triiodthy__L_e   \n",
       "2642                          tudchola_e <=> tudchola_c   \n",
       "\n",
       "                                     gene reaction rule  \\\n",
       "216                                                APEH   \n",
       "493                                  SLCO1A2 or SLCO2B1   \n",
       "515                                             SLCO2B1   \n",
       "762                                             SLCO1A2   \n",
       "775                                  SLCO1A2 or SLCO2B1   \n",
       "798                                                <NA>   \n",
       "881                                             SLCO1A2   \n",
       "957                                  SLCO1A2 or SLCO2B1   \n",
       "1354                                            SLCO1A2   \n",
       "1356                                            SLCO1A2   \n",
       "1497                                            SLCO1A2   \n",
       "2096  (PAFAH1B1 and PAFAH1B3) or (PAFAH1B1 and PAFAH...   \n",
       "2097                                   LPCAT3 or LPCAT1   \n",
       "2104          ABCB4 or ABCA7 or ABCB1 or ABCA1 or ABCC1   \n",
       "2136                                            SLCO2B1   \n",
       "2149                                 SLCO1A2 or SLCO2B1   \n",
       "2266                                            SLCO2B1   \n",
       "2487          ABCB4 or ABCA7 or ABCB1 or ABCA1 or ABCC1   \n",
       "2539                                 SLCO1A2 or SLCO2B1   \n",
       "2541                                            SLCO1A2   \n",
       "2542                                            SLCO1A2   \n",
       "2544  (SLC7A5 and SLC3A2) or SLC16A10 or SLCO1A2 or ...   \n",
       "2627         (SLC7A5 and SLC3A2) or SLC16A10 or SLCO1A2   \n",
       "2642                                            SLCO1A2   \n",
       "\n",
       "                     subsystem  spontaneous  \\\n",
       "216         Peptide metabolism            0   \n",
       "493   Transport, extracellular            0   \n",
       "515   Transport, extracellular            0   \n",
       "762   Transport, extracellular            0   \n",
       "775   Transport, extracellular            0   \n",
       "798   Transport, extracellular            0   \n",
       "881   Transport, extracellular            0   \n",
       "957   Transport, extracellular            0   \n",
       "1354  Transport, extracellular            0   \n",
       "1356  Transport, extracellular            0   \n",
       "1497  Transport, extracellular            0   \n",
       "2096    Ether lipid metabolism            0   \n",
       "2097    Ether lipid metabolism            0   \n",
       "2104  Transport, extracellular            0   \n",
       "2136  Transport, extracellular            0   \n",
       "2149  Transport, extracellular            0   \n",
       "2266  Transport, extracellular            0   \n",
       "2487  Transport, extracellular            0   \n",
       "2539  Transport, extracellular            0   \n",
       "2541  Transport, extracellular            0   \n",
       "2542  Transport, extracellular            0   \n",
       "2544  Transport, extracellular            0   \n",
       "2627  Transport, extracellular            0   \n",
       "2642  Transport, extracellular            0   \n",
       "\n",
       "               proteomic evidence (#studies)  \\\n",
       "216                                     <NA>   \n",
       "493                                     <NA>   \n",
       "515                                     <NA>   \n",
       "762                                     <NA>   \n",
       "775                                     <NA>   \n",
       "798                                     <NA>   \n",
       "881                                     <NA>   \n",
       "957                                     <NA>   \n",
       "1354                                    <NA>   \n",
       "1356                                    <NA>   \n",
       "1497                                    <NA>   \n",
       "2096  PAFAH1B1(19);PAFAH1B2(16);PAFAH1B3(20)   \n",
       "2097                    LPCAT1(2);LPCAT3(11)   \n",
       "2104              ABCA7(7);ABCB1(1);ABCB4(1)   \n",
       "2136                                    <NA>   \n",
       "2149                                    <NA>   \n",
       "2266                                    <NA>   \n",
       "2487              ABCA7(7);ABCB1(1);ABCB4(1)   \n",
       "2539                                    <NA>   \n",
       "2541                                    <NA>   \n",
       "2542                                    <NA>   \n",
       "2544                                    <NA>   \n",
       "2627                                    <NA>   \n",
       "2642                                    <NA>   \n",
       "\n",
       "                                             references notes  \n",
       "216                      pubmed:10719179;pubmed:7275011  <NA>  \n",
       "493                                     pubmed:27504015  <NA>  \n",
       "515                                     pubmed:27504015  <NA>  \n",
       "762                                     pubmed:27504015  <NA>  \n",
       "775                                     pubmed:27504015  <NA>  \n",
       "798   pubmed:15294349;pubmed:17164127;pubmed:1765199...  <NA>  \n",
       "881                                     pubmed:27504015  <NA>  \n",
       "957                                     pubmed:27504015  <NA>  \n",
       "1354                                    pubmed:27504015  <NA>  \n",
       "1356                                    pubmed:27504015  <NA>  \n",
       "1497                                    pubmed:27504015  <NA>  \n",
       "2096                     pubmed:21844189;pubmed:8440681  <NA>  \n",
       "2097                    pubmed:18156367;pubmed:18285344  <NA>  \n",
       "2104  pubmed:10947968;pubmed:28134466;pubmed:9651491...  <NA>  \n",
       "2136                                    pubmed:27504015  <NA>  \n",
       "2149                                    pubmed:27504015  <NA>  \n",
       "2266                                    pubmed:27504015  <NA>  \n",
       "2487     pubmed:10947968;pubmed:9778357;pubmed:14578045  <NA>  \n",
       "2539                                    pubmed:27504015  <NA>  \n",
       "2541                                    pubmed:27504015  <NA>  \n",
       "2542                                    pubmed:27504015  <NA>  \n",
       "2544  pubmed:34925607;pubmed:4479710;pubmed:7915653;...  <NA>  \n",
       "2627  pubmed:34925607;pubmed:4479710;pubmed:7915653;...  <NA>  \n",
       "2642                                    pubmed:27504015  <NA>  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "attr_cols = [\"name\", \"reaction\", \"gene reaction rule\", \"subsystem\"]\n",
    "obj_type = Reaction\n",
    "\n",
    "df_updated = dataframes_updated[attribute_type]\n",
    "if not df_updated.empty:\n",
    "    df_updated = df_updated.drop(\"retired\", axis=1).set_index(attribute_type).convert_dtypes()\n",
    "    for idx, row in df_updated.iterrows():\n",
    "        obj_id = idx\n",
    "        # Deprecated IDs should be updated at this point, only refers to new metabolites\n",
    "        try:\n",
    "            has_id = getattr(model, attribute_type).has_id(obj_id)\n",
    "            if not has_id:\n",
    "                if attribute_type == \"genes\":\n",
    "                    # No 'add_genes' method in cobrapy\n",
    "                    add_method = getattr(model, f\"{attribute_type}\").extend\n",
    "                else:\n",
    "                    add_method = getattr(model, f\"add_{attribute_type}\")\n",
    "                add_method([obj_type(obj_id)])\n",
    "        except ValueError as e:\n",
    "            print(f\"Error with {obj_id}\")\n",
    "            raise e\n",
    "\n",
    "        obj = getattr(model, attribute_type).get_by_id(obj_id)\n",
    "        for attr, value in zip(attr_cols, row.convert_dtypes()[attr_cols].fillna(\"\")):\n",
    "            if value:\n",
    "                attr = attr.replace(\" \", \"_\")\n",
    "                setattr(obj, attr, value)\n",
    "        annotations_dict = {\n",
    "            k: v for k, v in row.convert_dtypes()[~row.index.isin([attribute_type] + attr_cols)].to_dict().items()\n",
    "            if not \"evidence\" in k and (v and v != float(\"nan\"))\n",
    "        }\n",
    "        try:\n",
    "            notes_dict = {\"notes\": annotations_dict.pop(\"notes\")}\n",
    "        except KeyError:\n",
    "            notes_dict = obj.notes\n",
    "        obj.annotation.update(annotations_dict)\n",
    "        obj.notes.update(notes_dict)\n",
    "\n",
    "     # Update evidence\n",
    "    try:\n",
    "        df_previous_evidence = pd.read_csv(\n",
    "            get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "    \n",
    "    # Replace previous IDs with new ones\n",
    "    df_previous_evidence[attribute_type] = df_previous_evidence[attribute_type].replace(\n",
    "        id_mapping_dicts[attribute_type]\n",
    "    )\n",
    "    # Replace non-empty values with previous ones\n",
    "    df_previous_values = df_previous_evidence[df_previous_evidence[attribute_type].isin(df_updated.index)].set_index(attribute_type).convert_dtypes()\n",
    "    cols_to_update = list(df_previous_evidence.columns[~df_previous_evidence.columns.isin(attr_cols + [attribute_type, strip_plural(attribute_type)])])\n",
    "    # Evidence can be problematic due to data coercion \n",
    "    cols_to_update = [c for c in cols_to_update if \"evidence\" not in c]\n",
    "\n",
    "    # Set indices to align\n",
    "    df_updated.loc[list(df_updated.index), cols_to_update] = df_updated.loc[list(df_updated.index), cols_to_update].fillna(df_previous_values.loc[list(df_updated.index), cols_to_update], method=None, axis=1)\n",
    "    # Reset index of updated DataFrame\n",
    "    df_updated = df_updated.reset_index(drop=False)\n",
    "\n",
    "    # Add updated entries to DataFrame\n",
    "    df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0).replace(\"nan\", str).convert_dtypes()\n",
    "    df_evidence = (\n",
    "        df_evidence.drop_duplicates(subset=[attribute_type], keep=\"first\")\n",
    "        .sort_values(attribute_type, ascending=True)\n",
    "        .reset_index(drop=True)\n",
    "    )[df_previous_evidence.columns]\n",
    "    dataframes_evidence[attribute_type] = df_evidence\n",
    "df_evidence[df_evidence[attribute_type].isin(df_updated[attribute_type])]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "64d45450-d19e-4e09-b76b-8ad6471b8153",
   "metadata": {},
   "source": [
    "## Refine model through removing items\n",
    "Removal order: \n",
    "1. Reactions\n",
    "2. Genes\n",
    "3. Metabolites\n",
    "#### Reactions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7d434a25-de3f-4e6b-98be-bb902e3092eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# attribute_type = \"reactions\"\n",
    "# try:\n",
    "#     df_removal = pd.read_csv(\n",
    "#         get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "#         sep=\"\\t\",\n",
    "#         index_col=None,\n",
    "#     )\n",
    "# except FileNotFoundError:\n",
    "#     df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "# to_remove = []\n",
    "# for reaction in df_removed[attribute_type]:\n",
    "#     try:\n",
    "#         reaction = getattr(model, attribute_type).get_by_id(reaction)\n",
    "#     except KeyError:\n",
    "#         continue\n",
    "#     to_remove.append(reaction)\n",
    "\n",
    "# model.remove_reactions(to_remove)\n",
    "# # Clean up removal file before archival\n",
    "# df_removed = (\n",
    "#     df_removed.drop_duplicates(subset=[attribute_type])\n",
    "#     .sort_values(attribute_type, ascending=True)\n",
    "#     .reset_index(drop=True)\n",
    "# )\n",
    "# dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "#     ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "#         df_removed[attribute_type].values\n",
    "#     )\n",
    "# ]\n",
    "# dataframes_removed[attribute_type] = df_removed\n",
    "# dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7279505a-5e9f-44f3-b4ba-33e9b957d6e2",
   "metadata": {},
   "source": [
    "#### Genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8987ef03-a8fb-433e-8396-99a6d9bd6d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# attribute_type = \"genes\"\n",
    "# try:\n",
    "#     df_removed = pd.read_csv(\n",
    "#         get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "#         sep=\"\\t\",\n",
    "#         index_col=None,\n",
    "#         dtype=str,\n",
    "#     )\n",
    "# except FileNotFoundError:\n",
    "#     df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "# to_remove = []\n",
    "# for gene in df_removed[attribute_type]:\n",
    "#     try:\n",
    "#         gene = model.genes.get_by_id(gene)\n",
    "#     except KeyError:\n",
    "#         continue\n",
    "#     to_remove.append(gene)\n",
    "# model.genes -= to_remove\n",
    "\n",
    "# # Clean up removal file before archival\n",
    "# df_removed = (\n",
    "#     df_removed.drop_duplicates(subset=[attribute_type])\n",
    "#     .sort_values(attribute_type, ascending=True)\n",
    "#     .reset_index(drop=True)\n",
    "# )\n",
    "# for gene in model.genes:\n",
    "#     if gene.reactions:\n",
    "#         continue\n",
    "#     print(f\"Orphaned: {gene}\")\n",
    "\n",
    "# dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "#     ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "#         df_removed[attribute_type].values\n",
    "#     )\n",
    "# ]\n",
    "# dataframes_removed[attribute_type] = df_removed\n",
    "# dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab7e51b-9730-439a-a0ce-271b65d72dcd",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "8d9b992c-9399-45aa-921f-c415db9e0178",
   "metadata": {},
   "outputs": [],
   "source": [
    "# attribute_type = \"metabolites\"\n",
    "# try:\n",
    "#     df_removed = pd.read_csv(\n",
    "#         get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "#         sep=\"\\t\",\n",
    "#         index_col=None,\n",
    "#         dtype=str,\n",
    "#     )\n",
    "# except FileNotFoundError:\n",
    "#     df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "# to_remove = []\n",
    "# for metabolite in df_removed[attribute_type]:\n",
    "#     try:\n",
    "#         metabolite = model.metabolites.get_by_id(metabolite)\n",
    "#     except KeyError:\n",
    "#         continue\n",
    "#     model.remove_metabolites([metabolite])\n",
    "# model.remove_metabolites(to_remove)\n",
    "\n",
    "# # Clean up removal file before archival\n",
    "# df_removed = (\n",
    "#     df_removed.drop_duplicates(subset=[attribute_type])\n",
    "#     .sort_values(attribute_type, ascending=True)\n",
    "#     .reset_index(drop=True)\n",
    "# )\n",
    "# for metabolite in model.metabolites:\n",
    "#     if metabolite.reactions:\n",
    "#         continue\n",
    "#     print(f\"Orphaned: {metabolite}\")\n",
    "\n",
    "# dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "#     ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "#         df_removed[attribute_type].values\n",
    "#     )\n",
    "# ]\n",
    "# dataframes_removed[attribute_type] = df_removed\n",
    "# dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4cfe3007-e6fe-4fdd-97a7-4f764d13ae57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_dirpath(\"curation\", use_temp=\"interim\" if not overwrite else None).mkdir(exist_ok=True, parents=True)\n",
    "# for dtype, dataframe_dict in zip([\"evidence\", \"removals\"], [dataframes_evidence, dataframes_removals]):\n",
    "#     for attribute_type in attribute_types:\n",
    "#         df = dataframe_dict.get(attribute_type)\n",
    "#         if df is not None and not df.empty:\n",
    "#             df.to_csv(\n",
    "#                 get_dirpath(\"curation\", use_temp=\"interim\" if not overwrite else None) / f\"{attribute_type}_{dtype}.tsv\",\n",
    "#                 sep=\"\\t\",\n",
    "#                 index=False,\n",
    "#             )\n",
    "#             print(f\"Saving {dtype} for {attribute_type}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "886da88c-00c0-4638-974f-2766333f4d7b",
   "metadata": {},
   "source": [
    "### Ensure all metabolites, genes, and reactions exist\n",
    "If removed (e.g., a duplicate), will show up in missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c56939d1-4a0c-4f68-b200-b84d7da345d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of missing metabolites: 0\n",
      "Number of missing genes: 0\n",
      "Number of missing reactions: 0\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'metabolites': set(), 'genes': set(), 'reactions': set()}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_sets = {}\n",
    "for attribute_type in attribute_types:\n",
    "    obj_list = getattr(model, attribute_type)\n",
    "    if attribute_type == \"reactions\":\n",
    "        obj_list = obj_list.query(lambda x: not x.subsystem == \"Pseudoreactions\")    \n",
    "    id_set = set(obj_list.list_attr(\"id\"))\n",
    "    missing_sets[attribute_type] = id_set.symmetric_difference(dataframes_evidence[attribute_type][attribute_type].values)\n",
    "    print(f\"Number of missing {attribute_type}: {len(missing_sets[attribute_type])}\")\n",
    "missing_sets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc83e690-0829-4236-9161-165a400539cb",
   "metadata": {},
   "source": [
    "### Check for extra metabolites, genes, and reactions without any associations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "834b542d-2669-43ab-b212-4d875f54558c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No extra metabolites.\n",
      "No extra genes.\n",
      "No extra reactions.\n"
     ]
    }
   ],
   "source": [
    "for attribute_type in attribute_types:\n",
    "    if attribute_type == \"reactions\":\n",
    "        query_function = lambda x: not len(x.metabolites)\n",
    "    else:\n",
    "        query_function = lambda x: not len(x.reactions)\n",
    "    extras_list = getattr(model, attribute_type).query(query_function)\n",
    "    if extras_list:\n",
    "        print(\"\\n\".join((\"\", attribute_type.capitalize(),len(attribute_type) * \"=\", \"\")))\n",
    "        for item in extras_list:\n",
    "            print(item)\n",
    "    else:\n",
    "        print(f\"No extra {attribute_type}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac24451-ec72-4565-8cb8-8fc57c6ee2d9",
   "metadata": {},
   "source": [
    "#### Add boundary reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b424c4c7-dd7d-4473-b9e4-f0b69b9278b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO create from a list\n",
    "boundaries = {\n",
    "    # All exchange boundary reactions added\n",
    "    \"exchange\": model.metabolites.query(lambda x: x.compartment == \"e\").list_attr(\"id\"),\n",
    "    # Intracellular demands, only used for accumulation is allowed for a compound\n",
    "    \"demand\": [],\n",
    "    # Intracellular sinks, only used for when a source is needed for a compound\n",
    "    \"sink\": [\n",
    "        # Globin/Hemoglobin\n",
    "        \"oxyhb_c\",\n",
    "        \"hb4_23dpg_c\",\n",
    "        \"hb_hco2_c\",\n",
    "        \"globin_c\",\n",
    "        \"hbsno_c\",\n",
    "        \"carboxyhb_c\",\n",
    "        \"cclglobin_c\",\n",
    "        \"hemedegprods_c\",\n",
    "        \"hba1c_c\",\n",
    "        # Amino acids\n",
    "        # AA protein Residues\n",
    "        \"protres_arg__L_c\",\n",
    "        \"protres_asn__L_c\",\n",
    "        \"protres_asp__L_c\",\n",
    "        \"protres_cys__L_c\",\n",
    "        \"protres_gln__L_c\",\n",
    "        \"protres_glu__L_c\",\n",
    "        \"protres_his__L_c\",\n",
    "        \"protres_lys__L_c\",\n",
    "        \"protres_met__L_c\",\n",
    "        \"protres_ser__L_c\",\n",
    "        \"protres_thr__L_c\",\n",
    "        \"protres_tyr__L_c\",\n",
    "        \"protres_tyr__L_c\",\n",
    "        # Other AA residues\n",
    "        \"protres_asp__D_c\",\n",
    "        \"protres_isoasp__L_c\",\n",
    "        \"protres_isoasp__D_c\",\n",
    "        # Phosphorylated residues\n",
    "        \"protres_Nproshispi_c\",\n",
    "        \"protres_Ntelehispi_c\",\n",
    "        \"protres_serpi_c\",\n",
    "        \"protres_thrpi_c\",\n",
    "        \"protres_tyrpi_c\",\n",
    "        # Acetylated residues\n",
    "        \"protres_aclys__L_c\",\n",
    "        # Glycosylated residues\n",
    "        \"protres_ser3oacgam_c\",\n",
    "        \"protres_thr3oacgam_c\",\n",
    "        \"protres_serTAg_c\",\n",
    "        \"protres_thrTAg_c\",\n",
    "        # Glycated residues\n",
    "        \"protres_frulys_c\",\n",
    "        \"protres_rbllys_c\",\n",
    "        # Methylated residues\n",
    "        \"protres_admarg__L_c\",\n",
    "        \"protres_sdmarg__L_c\",\n",
    "        # Lipidated residues\n",
    "        \"protres_ttdcacys_c\",\n",
    "        \"protres_hxdcacys_c\",\n",
    "        \"protres_ocdcacys_c\",\n",
    "        # Oxidized residues\n",
    "        \"protres_metSox__SL_c\",\n",
    "        # Nitrosylated residues\n",
    "        \"protres_snocys__L_c\",\n",
    "        # ADP-ribosylated residues\n",
    "        \"protres_oadpribser_c\",\n",
    "        \"protres_sadpribcys_c\",\n",
    "        # Amine\n",
    "        \"protres_gludpam_c\",\n",
    "        \"protres_gluhista_c\",\n",
    "        \"protres_glunpphr_c\",\n",
    "        \"protres_glusrtn_c\",\n",
    "        \"protres_glu5meo__L_c\",\n",
    "        \"protres_lysglu_protres_c\",\n",
    "        # Ubiquitin\n",
    "        \"polyubb_c\",\n",
    "        \"ubiquitin_c\",\n",
    "        \"accprot_monoubiqlys_c\",\n",
    "        \"accprot_ubiqlys_c\",\n",
    "        \"accprot_lys__L_c\",\n",
    "        \"cullin_lys__L_c\",\n",
    "        \"cullin_nedd8lys_c\",\n",
    "        \"nedd8_c\",\n",
    "        # Small ions\n",
    "        \"na1_c\",\n",
    "        \"k_c\",\n",
    "        \"ca2_c\",\n",
    "        \"hno_c\",\n",
    "        \"co3r_c\",\n",
    "        # Vitamin E\n",
    "        \"avite1_c\",\n",
    "        \"avite1qn_c\",\n",
    "        # 'Redoxins'\n",
    "        \"prdx2crd_c\",\n",
    "        \"prdx2cso3_c\",\n",
    "        \"grdx2crd_c\",\n",
    "        \"grdx2cox_c\",\n",
    "        # Phospholipids\n",
    "        \"pc_hs_c\",\n",
    "        \"pco_hs_c\",\n",
    "        \"pcp_hs_c\",\n",
    "        \"pe_hs_c\",\n",
    "        \"pep_hs_c\",\n",
    "        \"ps_hs_c\",\n",
    "        \"paf_hs_c\",\n",
    "        \"pail_hs_c\",\n",
    "        \"pail345p_hs_c\",\n",
    "        \"pail34p_hs_c\",\n",
    "        \"pail35p_hs_c\",\n",
    "        \"pail3p_hs_c\",\n",
    "        \"pail45p_hs_c\",\n",
    "        \"pail4p_hs_c\",\n",
    "        \"pail5p_hs_c\",\n",
    "        \"sphmyln_hs_c\",\n",
    "\n",
    "        # CoA\n",
    "        # 'FAcoa_10_DC_c',\n",
    "        # 'FAcoa_12_DC_c',\n",
    "        # 'FAcoa_16_DC_c',\n",
    "        # 'FAcoa_4_DC_c',\n",
    "        # 'FAcoa_5_2EDC_c',\n",
    "        # 'FAcoa_5_DC_c',\n",
    "        # 'FAcoa_6_DC_c',\n",
    "        # 'FAcoa_7_DC_c',\n",
    "        # 'FAcoa_8_DC_c',\n",
    "        # 'FAcoa_5_3M3OH__S_c',\n",
    "        # 'FAcoa_hs_10_3OH__S_c',\n",
    "        # 'FAcoa_hs_12_3OH__S_c',\n",
    "        # 'FAcoa_hs_14_3OH__S_c',\n",
    "        # 'FAcoa_hs_14_5E8Z3OH__S_c',\n",
    "        # 'FAcoa_hs_14_7Z3OH__S_c',\n",
    "        # 'FAcoa_hs_16_3OH__R_c',\n",
    "        # 'FAcoa_hs_16_3OH__S_c',\n",
    "        # 'FAcoa_hs_16_7E10Z3OH__S_c',\n",
    "        # 'FAcoa_hs_16_9Z3OH__S_c',\n",
    "        # 'FAcoa_hs_17_3OH__R_c',\n",
    "        # 'FAcoa_hs_18_3OH__R_c',\n",
    "        # 'FAcoa_hs_18_3OH__S_c',\n",
    "        # 'FAcoa_hs_18_9Z12Z3OH__S_c',\n",
    "        # 'FAcoa_hs_18_9Z3OH__S_c',\n",
    "        # 'FAcoa_hs_19_3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_13Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_9Z3OH__R_c',\n",
    "        # 'FAcoa_hs_21_3OH__R_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_3OH__R_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_23_3OH__R_c',\n",
    "        # 'FAcoa_hs_24_12Z15Z18Z21Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_15Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_3OH__R_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z21Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z3OH__R_c',\n",
    "        # 'FAcoa_hs_26_17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_26_3OH__R_c',\n",
    "        # 'FAcoa_hs_3_3OH__S_c',\n",
    "        # 'FAcoa_hs_4_3OH__R_c',\n",
    "        # 'FAcoa_hs_6_3OH__S_c',\n",
    "        # 'FAcoa_hs_7_3OH__S_c',\n",
    "        # 'FAcoa_hs_8_3OH__S_c',\n",
    "        # 'FAcoa_hs_9_3OH__S_c',\n",
    "        # 'FAcoa_5_2E2M_c',\n",
    "        # 'FAcoa_5_2M_c',\n",
    "        # 'FAcoa_hs_10_2E6Z_c',\n",
    "        # 'FAcoa_hs_10_2E_c',\n",
    "        # 'FAcoa_hs_12_2E_c',\n",
    "        # 'FAcoa_hs_14_2E_c',\n",
    "        # 'FAcoa_hs_16_2E_c',\n",
    "        # 'FAcoa_hs_17_2E_c',\n",
    "        # 'FAcoa_hs_18_2E_c',\n",
    "        # 'FAcoa_hs_19_2E_c',\n",
    "        # 'FAcoa_hs_20_2E11Z14Z17Z_c',\n",
    "        # 'FAcoa_hs_20_2E11Z14Z_c',\n",
    "        # 'FAcoa_hs_20_2E11Z_c',\n",
    "        # 'FAcoa_hs_20_2E13Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z14Z17Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z14Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z_c',\n",
    "        # 'FAcoa_hs_20_2E9Z_c',\n",
    "        # 'FAcoa_hs_20_2E_c',\n",
    "        # 'FAcoa_hs_21_2E_c',\n",
    "        # 'FAcoa_hs_22_2E10Z13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E10Z13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E11Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z_c',\n",
    "        # 'FAcoa_hs_22_2E7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E7Z10Z13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E_c',\n",
    "        # 'FAcoa_hs_23_2E_c',\n",
    "        # 'FAcoa_hs_24_2E12Z15Z18Z21Z_c',\n",
    "        # 'FAcoa_hs_24_2E15Z_c',\n",
    "        # 'FAcoa_hs_24_2E9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcoa_hs_24_2E9Z12Z15Z18Z_c',\n",
    "        # 'FAcoa_hs_24_2E_c',\n",
    "        # 'FAcoa_hs_26_2E17Z_c',\n",
    "        # 'FAcoa_hs_26_2E_c',\n",
    "        # 'FAcoa_hs_3_2E_c',\n",
    "        # 'FAcoa_hs_4_2E_c',\n",
    "        # 'FAcoa_hs_6_2E_c',\n",
    "        # 'FAcoa_hs_8_2E_c',\n",
    "        # 'FAcoa_hs_14_5E8Z_c',\n",
    "        # 'FAcoa_4_2M_c',\n",
    "        # 'FAcoa_hs_16_3O_c',\n",
    "        # 'FAcoa_hs_17_3O_c',\n",
    "        # 'FAcoa_hs_18_3O_c',\n",
    "        # 'FAcoa_hs_19_3O_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z17Z3O_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z3O_c',\n",
    "        # 'FAcoa_hs_20_11Z3O_c',\n",
    "        # 'FAcoa_hs_20_13Z3O_c',\n",
    "        # 'FAcoa_hs_20_3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z17Z3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z3O_c',\n",
    "        # 'FAcoa_hs_20_9Z3O_c',\n",
    "        # 'FAcoa_hs_21_3O_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_22_11Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z3O_c',\n",
    "        # 'FAcoa_hs_22_3O_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_23_3O_c',\n",
    "        # 'FAcoa_hs_24_12Z15Z18Z21Z3O_c',\n",
    "        # 'FAcoa_hs_24_15Z3O_c',\n",
    "        # 'FAcoa_hs_24_3O_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z21Z3O_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z3O_c',\n",
    "        # 'FAcoa_hs_26_17Z3O_c',\n",
    "        # 'FAcoa_hs_26_3O_c',\n",
    "        # 'dmnoncoa_c',\n",
    "        # 'dmhptcoa_c',\n",
    "        # # Carnitine\n",
    "        # 'FAcrn_10_DC_c',\n",
    "        # 'FAcrn_12_DC_c',\n",
    "        # 'FAcrn_16_DC_c',\n",
    "        # 'FAcrn_4_2M_c',\n",
    "        # 'FAcrn_4_DC_c',\n",
    "        # 'FAcrn_5_2E2M_c',\n",
    "        # 'FAcrn_5_2EDC_c',\n",
    "        # 'FAcrn_5_2M_c',\n",
    "        # 'FAcrn_5_3M3OH__S_c',\n",
    "        # 'FAcrn_5_DC_c',\n",
    "        # 'FAcrn_6_DC_c',\n",
    "        # 'FAcrn_7_DC_c',\n",
    "        # 'FAcrn_8_DC_c',\n",
    "        # 'FAcrn_hs_10_0_c',\n",
    "        # 'FAcrn_hs_10_2E6Z_c',\n",
    "        # 'FAcrn_hs_10_2E_c',\n",
    "        # 'FAcrn_hs_10_3OH__S_c',\n",
    "        # 'FAcrn_hs_11_0_c',\n",
    "        # 'FAcrn_hs_12_0_c',\n",
    "        # 'FAcrn_hs_12_2E_c',\n",
    "        # 'FAcrn_hs_12_3OH__S_c',\n",
    "        # 'FAcrn_hs_13_0_c',\n",
    "        # 'FAcrn_hs_14_0_c',\n",
    "        # 'FAcrn_hs_14_2E_c',\n",
    "        # 'FAcrn_hs_14_3OH__S_c',\n",
    "        # 'FAcrn_hs_14_5E8Z3OH__S_c',\n",
    "        # 'FAcrn_hs_14_5E8Z_c',\n",
    "        # 'FAcrn_hs_14_5Z_c',\n",
    "        # 'FAcrn_hs_14_7Z3OH__S_c',\n",
    "        # 'FAcrn_hs_14_7Z_c',\n",
    "        # 'FAcrn_hs_14_9Z_c',\n",
    "        # 'FAcrn_hs_15_0_c',\n",
    "        # 'FAcrn_hs_16_0_c',\n",
    "        # 'FAcrn_hs_16_2E_c',\n",
    "        # 'FAcrn_hs_16_3OH__S_c',\n",
    "        # 'FAcrn_hs_16_7E10Z3OH__S_c',\n",
    "        # 'FAcrn_hs_16_7Z_c',\n",
    "        # 'FAcrn_hs_16_9Z3OH__S_c',\n",
    "        # 'FAcrn_hs_16_9Z_c',\n",
    "        # 'FAcrn_hs_17_0_c',\n",
    "        # 'FAcrn_hs_17_10Z_c',\n",
    "        # 'FAcrn_hs_17_9Z_c',\n",
    "        # 'FAcrn_hs_18_0_c',\n",
    "        # 'FAcrn_hs_18_11Z_c',\n",
    "        # 'FAcrn_hs_18_13Z_c',\n",
    "        # 'FAcrn_hs_18_2E_c',\n",
    "        # 'FAcrn_hs_18_3OH__S_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z12Z15Z_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z12Z_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z_c',\n",
    "        # 'FAcrn_hs_18_7Z_c',\n",
    "        # 'FAcrn_hs_18_9E_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z15Z_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z3OH__S_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z_c',\n",
    "        # 'FAcrn_hs_18_9Z3OH__S_c',\n",
    "        # 'FAcrn_hs_18_9Z_c',\n",
    "        # 'FAcrn_hs_19_0_c',\n",
    "        # 'FAcrn_hs_20_0_c',\n",
    "        # 'FAcrn_hs_20_11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_11Z_c',\n",
    "        # 'FAcrn_hs_20_13Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z_c',\n",
    "        # 'FAcrn_hs_20_9Z_c',\n",
    "        # 'FAcrn_hs_21_0_c',\n",
    "        # 'FAcrn_hs_22_0_c',\n",
    "        # 'FAcrn_hs_22_10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_11Z_c',\n",
    "        # 'FAcrn_hs_22_13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_13Z_c',\n",
    "        # 'FAcrn_hs_22_4Z7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_4Z7Z10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_7Z10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_23_0_c',\n",
    "        # 'FAcrn_hs_24_0_c',\n",
    "        # 'FAcrn_hs_24_12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_15Z_c',\n",
    "        # 'FAcrn_hs_24_6Z9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_6Z9Z12Z15Z18Z_c',\n",
    "        # 'FAcrn_hs_24_9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_9Z12Z15Z18Z_c',\n",
    "        # 'FAcrn_hs_26_0_c',\n",
    "        # 'FAcrn_hs_26_17Z_c',\n",
    "        # 'FAcrn_hs_3_0_c',\n",
    "        # 'FAcrn_hs_3_2E_c',\n",
    "        # 'FAcrn_hs_3_3OH__S_c',\n",
    "        # 'FAcrn_hs_4_0_c',\n",
    "        # 'FAcrn_hs_4_2E_c',\n",
    "        # 'FAcrn_hs_4_3OH__R_c',\n",
    "        # 'FAcrn_hs_5_0_c',\n",
    "        # 'FAcrn_hs_6_0_c',\n",
    "        # 'FAcrn_hs_6_2E_c',\n",
    "        # 'FAcrn_hs_6_3OH__S_c',\n",
    "        # 'FAcrn_hs_7_0_c',\n",
    "        # 'FAcrn_hs_7_3OH__S_c',\n",
    "        # 'FAcrn_hs_8_0_c',\n",
    "        # 'FAcrn_hs_8_2E_c',\n",
    "        # 'FAcrn_hs_8_3OH__S_c',\n",
    "        # 'FAcrn_hs_9_0_c',\n",
    "        # 'FAcrn_hs_9_3OH__S_c',\n",
    "        # 'acrn_c',\n",
    "        # 'dmhptcrn_c',\n",
    "        # 'dmnoncrn_c',\n",
    "        # 'malcrn_c',\n",
    "        # tRNA\n",
    "        # 'trnaala_c',\n",
    "        # 'trnaarg_c',\n",
    "        # 'trnaasn_c',\n",
    "        # 'trnaasp_c',\n",
    "        # 'trnacys_c',\n",
    "        # 'trnagln_c',\n",
    "        # 'trnaglu_c',\n",
    "        # 'trnagly_c',\n",
    "        # 'trnahis_c',\n",
    "        # 'trnaile_c',\n",
    "        # 'trnaleu_c',\n",
    "        # 'trnalys_c',\n",
    "        # 'trnamet_c',\n",
    "        # 'trnaphe_c',\n",
    "        # 'trnapro_c',\n",
    "        # 'trnaser_c',\n",
    "        # 'trnathr_c',\n",
    "        # 'trnatrp_c',\n",
    "        # 'trnatyr_c',\n",
    "        # 'trnaval_c',\n",
    "        # 'alatrna_c',\n",
    "        # 'argtrna_c',\n",
    "        # 'asntrna_c',\n",
    "        # 'asptrna_c',\n",
    "        # 'cystrna_c',\n",
    "        # 'glntrna_c',\n",
    "        # 'glutrna_c',\n",
    "        # 'glytrna_c',\n",
    "        # 'histrna_c',\n",
    "        # 'iletrna_c',\n",
    "        # 'leutrna_c',\n",
    "        # 'lystrna_c',\n",
    "        # 'mettrna_c',\n",
    "        # 'phetrna_c',\n",
    "        # 'protrna_c',\n",
    "        # 'sertrna_c',\n",
    "        # 'thrtrna_c',\n",
    "        # 'trptrna_c',\n",
    "        # 'tyrtrna_c',\n",
    "        # 'valtrna_c',\n",
    "        # Sugar\n",
    "        # '2ddglcn_c',\n",
    "        # '3dfru_c',\n",
    "        # Nucleotides\n",
    "        # \"23camp_c\",\n",
    "        # \"23ccmp_c\",\n",
    "        # \"23cgmp_c\",\n",
    "        # \"23cump_c\",\n",
    "        # \"dctp_c\",\n",
    "        # \"dgtp_c\",\n",
    "        # \"datp_c\",\n",
    "        # \"dttp_c\",\n",
    "        # 'dutp_c',\n",
    "        # \"psump_c\",\n",
    "        # \"psi_c\",\n",
    "        # Sourced from somewhere/\n",
    "        # Drains to somewhere/accumulates\n",
    "        # 'so2gth_c',\n",
    "        # \"dh15kprostge1_c\",\n",
    "        # \"dh15kprostge2_c\",\n",
    "        # \"dh15kprostge3_c\",\n",
    "        # \"dh15kprostgf1_c\",\n",
    "        # \"dh15kprostgf2_c\",\n",
    "        # \"dh15kprostgf3_c\",\n",
    "        # 'polyadprib2_c',\n",
    "        # 'polyadprib1_c',\n",
    "    ],\n",
    "}\n",
    "model.remove_reactions(model.reactions.query(lambda x: x.boundary))\n",
    "default_closed = []\n",
    "for btype, met_list in boundaries.items():\n",
    "    for met in met_list:\n",
    "        met = model.metabolites.get_by_id(met)\n",
    "        try:\n",
    "            reaction = model.add_boundary(met, type=btype)\n",
    "        except ValueError:\n",
    "            rid = {\n",
    "                \"exchange\": f\"EX_{met}\",\n",
    "                \"demand\": f\"DM_{met}\",\n",
    "                \"sink\": f\"SK_{met}\",\n",
    "            }[btype]\n",
    "            reaction = model.reactions.get_by_id(rid)\n",
    "            reaction.name = f\"{met.name} {btype}\"\n",
    "\n",
    "        if met in default_closed:\n",
    "            reaction.lower_bound = 0\n",
    "\n",
    "for reaction in model.boundary:\n",
    "    reaction.subsystem = \"Pseudoreactions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff186a49-66a1-444c-92dc-0f1eca0464d9",
   "metadata": {},
   "source": [
    "#### Reset subsystem groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5037b248-ee81-442e-aa81-dd8acd25380f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>11d3fc610</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x11d3fc610>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reset_subsystem_groups(model)\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ba5b6-68bb-4160-bcfa-58fe424d8025",
   "metadata": {},
   "source": [
    "### Check mass balancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d47506ac-f31a-4730-93dc-95fef5cf1986",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6LTHPI: 6lthp_c --> h_c + sppt_c\n",
      "{'charge': 2.0}\n",
      "\n",
      "METHBCYTBR: 2.0 focytb5_c + methb_c --> 2.0 ficytb5_c + hb_c\n",
      "{'charge': 1.0}\n",
      "\n",
      "METHBFMNR: fmnh2_c + methb_c --> fmn_c + 3.0 h_c + hb_c\n",
      "{'charge': 1.0}\n",
      "\n",
      "POOL_FACOA: FAcoa_hs_c <=> 0.0004 FAcoa_hs_12_0_c + 0.0004 FAcoa_hs_13_0_c + 0.0133 FAcoa_hs_14_0_c + 0.0004 FAcoa_hs_14_5Z_c + 0.0004 FAcoa_hs_14_7Z_c + 0.0004 FAcoa_hs_14_9Z_c + 0.0004 FAcoa_hs_15_0_c + 0.222 FAcoa_hs_16_0_c + 0.0004 FAcoa_hs_16_7Z_c + 0.0219 FAcoa_hs_16_9Z_c + 0.0004 FAcoa_hs_17_0_c + 0.0004 FAcoa_hs_17_10Z_c + 0.0004 FAcoa_hs_17_9Z_c + 0.1498 FAcoa_hs_18_0_c + 0.025 FAcoa_hs_18_11Z_c + 0.0004 FAcoa_hs_18_13Z_c + 0.0025 FAcoa_hs_18_6Z9Z12Z15Z_c + 0.0029 FAcoa_hs_18_6Z9Z12Z_c + 0.0004 FAcoa_hs_18_6Z9Z_c + 0.0004 FAcoa_hs_18_7Z_c + 0.0004 FAcoa_hs_18_9E_c + 0.0084 FAcoa_hs_18_9Z12Z15Z_c + 0.1915 FAcoa_hs_18_9Z12Z_c + 0.1545 FAcoa_hs_18_9Z_c + 0.0004 FAcoa_hs_19_0_c + 0.0004 FAcoa_hs_20_0_c + 0.0215 FAcoa_hs_20_11Z14Z17Z_c + 0.0004 FAcoa_hs_20_11Z14Z_c + 0.0004 FAcoa_hs_20_11Z_c + 0.0004 FAcoa_hs_20_13Z_c + 0.0116 FAcoa_hs_20_5Z8Z11Z14Z17Z_c + 0.1083 FAcoa_hs_20_5Z8Z11Z14Z_c + 0.0004 FAcoa_hs_20_5Z8Z11Z_c + 0.0115 FAcoa_hs_20_8Z11Z14Z17Z_c + 0.0215 FAcoa_hs_20_8Z11Z14Z_c + 0.0004 FAcoa_hs_20_8Z11Z_c + 0.0004 FAcoa_hs_20_9Z_c + 0.0004 FAcoa_hs_21_0_c + 0.0004 FAcoa_hs_22_0_c + 0.0004 FAcoa_hs_22_10Z13Z16Z19Z_c + 0.0004 FAcoa_hs_22_10Z13Z16Z_c + 0.0004 FAcoa_hs_22_11Z_c + 0.0004 FAcoa_hs_22_13Z16Z19Z_c + 0.0004 FAcoa_hs_22_13Z16Z_c + 0.0004 FAcoa_hs_22_13Z_c + 0.0278 FAcoa_hs_22_4Z7Z10Z13Z16Z19Z_c + 0.0038 FAcoa_hs_22_4Z7Z10Z13Z16Z_c + 0.0059 FAcoa_hs_22_7Z10Z13Z16Z19Z_c + 0.0014 FAcoa_hs_22_7Z10Z13Z16Z_c + 0.0004 FAcoa_hs_23_0_c + 0.0004 FAcoa_hs_24_0_c + 0.0004 FAcoa_hs_24_12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_15Z_c + 0.0004 FAcoa_hs_24_6Z9Z12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_6Z9Z12Z15Z18Z_c + 0.0004 FAcoa_hs_24_9Z12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_9Z12Z15Z18Z_c + 0.0004 FAcoa_hs_26_0_c + 0.0004 FAcoa_hs_26_17Z_c\n",
      "{'charge': -0.08439999999999515, 'C': 17.817500000000024, 'H': 33.30219999999997, 'N': 0.14769999999999328, 'O': 0.3586999999999705, 'P': 0.0632999999999968, 'S': 0.021099999999998786, 'R': -1.0}\n",
      "\n",
      "POOL_FA: FA_hs_c <=> 0.0004 FA_hs_12_0_c + 0.0004 FA_hs_13_0_c + 0.0133 FA_hs_14_0_c + 0.0004 FA_hs_14_5Z_c + 0.0004 FA_hs_14_7Z_c + 0.0004 FA_hs_14_9Z_c + 0.0004 FA_hs_15_0_c + 0.222 FA_hs_16_0_c + 0.0004 FA_hs_16_7Z_c + 0.0219 FA_hs_16_9Z_c + 0.0004 FA_hs_17_0_c + 0.0004 FA_hs_17_10Z_c + 0.0004 FA_hs_17_9Z_c + 0.1498 FA_hs_18_0_c + 0.025 FA_hs_18_11Z_c + 0.0004 FA_hs_18_13Z_c + 0.0025 FA_hs_18_6Z9Z12Z15Z_c + 0.0029 FA_hs_18_6Z9Z12Z_c + 0.0004 FA_hs_18_6Z9Z_c + 0.0004 FA_hs_18_7Z_c + 0.0004 FA_hs_18_9E_c + 0.0084 FA_hs_18_9Z12Z15Z_c + 0.1915 FA_hs_18_9Z12Z_c + 0.1545 FA_hs_18_9Z_c + 0.0004 FA_hs_19_0_c + 0.0004 FA_hs_20_0_c + 0.0215 FA_hs_20_11Z14Z17Z_c + 0.0004 FA_hs_20_11Z14Z_c + 0.0004 FA_hs_20_11Z_c + 0.0004 FA_hs_20_13Z_c + 0.0116 FA_hs_20_5Z8Z11Z14Z17Z_c + 0.1083 FA_hs_20_5Z8Z11Z14Z_c + 0.0004 FA_hs_20_5Z8Z11Z_c + 0.0115 FA_hs_20_8Z11Z14Z17Z_c + 0.0215 FA_hs_20_8Z11Z14Z_c + 0.0004 FA_hs_20_8Z11Z_c + 0.0004 FA_hs_20_9Z_c + 0.0004 FA_hs_21_0_c + 0.0004 FA_hs_22_0_c + 0.0004 FA_hs_22_10Z13Z16Z19Z_c + 0.0004 FA_hs_22_10Z13Z16Z_c + 0.0004 FA_hs_22_11Z_c + 0.0004 FA_hs_22_13Z16Z19Z_c + 0.0004 FA_hs_22_13Z16Z_c + 0.0004 FA_hs_22_13Z_c + 0.0278 FA_hs_22_4Z7Z10Z13Z16Z19Z_c + 0.0038 FA_hs_22_4Z7Z10Z13Z16Z_c + 0.0059 FA_hs_22_7Z10Z13Z16Z19Z_c + 0.0014 FA_hs_22_7Z10Z13Z16Z_c + 0.0004 FA_hs_23_0_c + 0.0004 FA_hs_24_0_c + 0.0004 FA_hs_24_12Z15Z18Z21Z_c + 0.0004 FA_hs_24_15Z_c + 0.0004 FA_hs_24_6Z9Z12Z15Z18Z21Z_c + 0.0004 FA_hs_24_6Z9Z12Z15Z18Z_c + 0.0004 FA_hs_24_9Z12Z15Z18Z21Z_c + 0.0004 FA_hs_24_9Z12Z15Z18Z_c + 0.0004 FA_hs_26_0_c + 0.0004 FA_hs_26_17Z_c\n",
      "{'charge': -0.021099999999998786, 'C': 17.374399999999994, 'H': 32.6481, 'O': 0.04219999999999757, 'R': -1.0}\n",
      "\n",
      "SNOHBNO: nitrosylhb_c --> h_c + hbsno_c\n",
      "{'charge': 1.0}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for reaction in model.reactions:\n",
    "    if reaction.boundary:\n",
    "        continue\n",
    "    try:\n",
    "        if reaction.check_mass_balance():\n",
    "            print(reaction)\n",
    "            print(reaction.check_mass_balance())\n",
    "            print()\n",
    "    except:\n",
    "        print({m.id: m.charge for m in reaction.metabolites})\n",
    "        raise "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18d61d2b-a4a0-4ac8-84c6-68bd7b421790",
   "metadata": {},
   "source": [
    "### Set bounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dc710bb4-325e-4db2-bd30-c3595b60052e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reset_reaction_bounds(model)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cad316c-ec10-461b-9c38-4e1bdeed2db5",
   "metadata": {},
   "source": [
    "### Ensure correct types before export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ea115b87-ffa6-4773-8207-08701a8d03b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SBML will not export charges correctly if they are float\n",
    "for metabolite in model.metabolites:\n",
    "    metabolite.charge = int(metabolite.charge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77e6dc2-ed20-455b-aadd-8a03db5666f1",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "742aeb63-f41b-4280-8fc0-9a79e9410e3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>11d3fc610</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3290</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x11d3fc610>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_model_dirpath = get_dirpath(\"model\", use_temp=\"interim\" if not overwrite else None)\n",
    "new_model_dirpath.mkdir(exist_ok=True, parents=True)\n",
    "\n",
    "write_cobra_model(model=model, filename=new_model_dirpath / f\"{model.id.replace('_', '-')}.xml\")\n",
    "write_cobra_model(model=model, filename=new_model_dirpath / f\"{model.id.replace('_', '-')}.json\")\n",
    "model\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "978d93a7-23dc-47ee-b9b9-f330fdfa8f05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes: 820\n",
      "Metabolites (all): 2157\n",
      "Metabolites (unique): 1684\n",
      "Reactions: 2725\n"
     ]
    }
   ],
   "source": [
    "print(f\"Genes: {len(set([x.id for x in model.genes]))}\")\n",
    "print(f\"Metabolites (all): {len(set([x.id for x in model.metabolites]))}\")\n",
    "nmets_unique = len({x.id.replace(f\"_{x.compartment}\", \"\") for x in model.metabolites})\n",
    "print(f\"Metabolites (unique): {nmets_unique}\")\n",
    "print(\n",
    "    f\"Reactions: {len(set([x.id for x in model.reactions.query(lambda x: not x.boundary)]))}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b097e5-fcad-4928-a6ff-8b44a9d82aa1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
