{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ba9c3d2c-72da-416c-8612-8a3f91eee20a",
   "metadata": {},
   "source": [
    "# RBC-GEM 1.2.0 Updates\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "194c1f88-6d9b-40d5-b468-86da5bc66f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from warnings import warn\n",
    "\n",
    "import pandas as pd\n",
    "from cobra.core import Gene, Metabolite, Reaction\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    build_string,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    split_string,\n",
    "    write_cobra_model,\n",
    ")\n",
    "from rbc_gem_utils.qc import (\n",
    "    reset_reaction_bounds,\n",
    "    reset_subsystem_groups,\n",
    "    standardardize_metabolite_formulas,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "945a5ce1-932d-4622-a823-c54b76b48d99",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0c9d3e47-f2f1-44e6-9333-4adec35ef021",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f967b496-7302-49a9-a474-45b13dc70980",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 1.1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "787d69b8-f885-4cda-9598-505f2b72b8c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n"
     ]
    }
   ],
   "source": [
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.yml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a1c443b-d26f-41e6-a674-30a1c91900db",
   "metadata": {},
   "source": [
    "### Deprecate identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "91e9978b-8421-45e9-8260-35b86b855dfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dicts = {\n",
    "    \"metabolites\": {\n",
    "        \"PPP2CA_leu__L_c\": \"pp2a_leu__L_c\",\n",
    "        \"PPP2CA_leumeo_c\": \"pp2a_leumeo_c\",\n",
    "        # Protein ID updates required for protein constrained modeling compatibility\n",
    "        \"protein_aclys__L_c\": \"protres_aclys__L_c\",\n",
    "        \"protein_admarg__L_c\": \"protres_admarg__L_c\",\n",
    "        \"protein_arg__L_c\": \"protres_arg__L_c\",\n",
    "        \"protein_arg__L_e\": \"protres_arg__L_e\",\n",
    "        \"protein_asn__L_c\": \"protres_asn__L_c\",\n",
    "        \"protein_asp__D_c\": \"protres_asp__D_c\",\n",
    "        \"protein_asp__L_c\": \"protres_asp__L_c\",\n",
    "        \"protein_aspmeo__D_c\": \"protres_aspmeo__D_c\",\n",
    "        \"protein_cys__L_c\": \"protres_cys__L_c\",\n",
    "        \"protein_frulys_c\": \"protres_frulys_c\",\n",
    "        \"protein_frulys3p_c\": \"protres_frulys3p_c\",\n",
    "        \"protein_frulys6p_c\": \"protres_frulys6p_c\",\n",
    "        \"protein_gln__L_c\": \"protres_gln__L_c\",\n",
    "        \"protein_glu__L_c\": \"protres_glu__L_c\",\n",
    "        \"protein_glu5meo__L_c\": \"protres_glu5meo__L_c\",\n",
    "        \"protein_gludpam_c\": \"protres_gludpam_c\",\n",
    "        \"protein_gluhista_c\": \"protres_gluhista_c\",\n",
    "        \"protein_glunpphr_c\": \"protres_glunpphr_c\",\n",
    "        \"protein_glusrtn_c\": \"protres_glusrtn_c\",\n",
    "        \"protein_his__L_c\": \"protres_his__L_c\",\n",
    "        \"protein_hxdcacys_c\": \"protres_hxdcacys_c\",\n",
    "        \"protein_isoasp__D_c\": \"protres_isoasp__D_c\",\n",
    "        \"protein_isoasp__L_c\": \"protres_isoasp__L_c\",\n",
    "        \"protein_isoaspmeo__L_c\": \"protres_isoaspmeo__L_c\",\n",
    "        \"protein_lacarg_c\": \"protres_lacarg_c\",\n",
    "        \"protein_laccys_c\": \"protres_laccys_c\",\n",
    "        \"protein_laclys_c\": \"protres_laclys_c\",\n",
    "        \"protein_lys__L_c\": \"protres_lys__L_c\",\n",
    "        \"protein_lysglu_protein_c\": \"protres_lysglu_protres_c\",\n",
    "        \"protein_mmarg__L_c\": \"protres_mmarg__L_c\",\n",
    "        \"protein_Nproshispi_c\": \"protres_Nproshispi_c\",\n",
    "        \"protein_Ntelehispi_c\": \"protres_Ntelehispi_c\",\n",
    "        \"protein_Nwadpribarg__L_e\": \"protres_Nwadpribarg__L_e\",\n",
    "        \"protein_oadpribser_c\": \"protres_oadpribser_c\",\n",
    "        \"protein_ocdcacys_c\": \"protres_ocdcacys_c\",\n",
    "        \"protein_pepproW0_c\": \"protres_pepproW0_c\",\n",
    "        \"protein_pepproW180_c\": \"protres_pepproW180_c\",\n",
    "        \"protein_rbllys_c\": \"protres_rbllys_c\",\n",
    "        \"protein_rbllys3p_c\": \"protres_rbllys3p_c\",\n",
    "        \"protein_rbllys5p_c\": \"protres_rbllys5p_c\",\n",
    "        \"protein_sadpribcys_c\": \"protres_sadpribcys_c\",\n",
    "        \"protein_sdmarg__L_c\": \"protres_sdmarg__L_c\",\n",
    "        \"protein_ser__L_c\": \"protres_ser__L_c\",\n",
    "        \"protein_ser3oacgam_c\": \"protres_ser3oacgam_c\",\n",
    "        \"protein_serpi_c\": \"protres_serpi_c\",\n",
    "        \"protein_serTAg_c\": \"protres_serTAg_c\",\n",
    "        \"protein_serTnAg_c\": \"protres_serTnAg_c\",\n",
    "        \"protein_sucim__D_c\": \"protres_sucim__D_c\",\n",
    "        \"protein_sucim__L_c\": \"protres_sucim__L_c\",\n",
    "        \"protein_thr__L_c\": \"protres_thr__L_c\",\n",
    "        \"protein_thr3oacgam_c\": \"protres_thr3oacgam_c\",\n",
    "        \"protein_thrpi_c\": \"protres_thrpi_c\",\n",
    "        \"protein_thrTAg_c\": \"protres_thrTAg_c\",\n",
    "        \"protein_thrTnAg_c\": \"protres_thrTnAg_c\",\n",
    "        \"protein_ttdcacys_c\": \"protres_ttdcacys_c\",\n",
    "        \"protein_tyr__L_c\": \"protres_tyr__L_c\",\n",
    "        \"protein_tyrpi_c\": \"protres_tyrpi_c\",\n",
    "    },\n",
    "    \"genes\": {},\n",
    "    \"reactions\": {\n",
    "        \"ClPItex\": \"Cl_PItex\",\n",
    "        \"NO22Natex\": \"NO2_2Natex\",\n",
    "        \"CYSL\": \"CYSLNET\",\n",
    "        \"CYSTCYSL\": \"CYSTL1\",\n",
    "        \"SELCYSTL\": \"SELCYSTL1\",\n",
    "        \"SELMETHL\": \"SELMETHL1\",\n",
    "        \"LCMTPPP2CA\": \"LCMTPP2A\",\n",
    "        \"PPMEPPP2CA\": \"PPMEPP2A\",\n",
    "        \"GALt1r\": \"GAL_Dt\",\n",
    "        \"EX_protein_Nwadpribarg__L_e\": \"EX_protres_Nwadpribarg__L_e\",\n",
    "        \"EX_protein_arg__L_e\": \"EX_protres_arg__L_e\",\n",
    "        \"IMRBVPT\": \"RBVIMPT\",\n",
    "        # Cleanup of nucleotide IDs --> NTD(D)(N)MP\n",
    "        \"NTD1\": \"NTDDUMP\",\n",
    "        \"NTD2\": \"NTDUMP\",\n",
    "        \"NTD3\": \"NTDDCMP\",\n",
    "        \"NTD4\": \"NTDCMP\",\n",
    "        \"NTD5\": \"NTDDTMP\",\n",
    "        \"NTD6\": \"NTDDAMP\",\n",
    "        \"NTD7\": \"NTDAMP\",\n",
    "        \"NTD8\": \"NTDDGMP\",\n",
    "        \"NTD9\": \"NTDGMP\",\n",
    "        \"NTD10\": \"NTDXMP\",\n",
    "        \"NTD11\": \"NTDIMP\",\n",
    "        \"NTD12\": \"NTDDIMP\",\n",
    "        \"NTD15\": \"NTDPSUMP\",\n",
    "        \"2NTDA\": \"NTD2AMP\",\n",
    "        \"2NTDC\": \"NTD2CMP\",\n",
    "        \"2NTDG\": \"NTD2GMP\",\n",
    "        \"2NTDU\": \"NTD2UMP\",\n",
    "        \"NTDAe\": \"NTDAMPe\",\n",
    "        \"NDPAe\": \"NDPADPe\",\n",
    "        \"NTDADPe\": \"NDPADPe\",\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "473c1a5d-5b44-4028-affe-9b059fa50c3f",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fc76e7fa-8a1e-4f6e-bb74-4d4745e7f189",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not map PPP2CA_leu__L_c to new ID.\n",
      "Could not map PPP2CA_leumeo_c to new ID.\n",
      "Could not map protein_aclys__L_c to new ID.\n",
      "Could not map protein_admarg__L_c to new ID.\n",
      "Could not map protein_arg__L_c to new ID.\n",
      "Could not map protein_arg__L_e to new ID.\n",
      "Could not map protein_asn__L_c to new ID.\n",
      "Could not map protein_asp__D_c to new ID.\n",
      "Could not map protein_asp__L_c to new ID.\n",
      "Could not map protein_aspmeo__D_c to new ID.\n",
      "Could not map protein_cys__L_c to new ID.\n",
      "Could not map protein_frulys_c to new ID.\n",
      "Could not map protein_frulys3p_c to new ID.\n",
      "Could not map protein_frulys6p_c to new ID.\n",
      "Could not map protein_gln__L_c to new ID.\n",
      "Could not map protein_glu__L_c to new ID.\n",
      "Could not map protein_glu5meo__L_c to new ID.\n",
      "Could not map protein_gludpam_c to new ID.\n",
      "Could not map protein_gluhista_c to new ID.\n",
      "Could not map protein_glunpphr_c to new ID.\n",
      "Could not map protein_glusrtn_c to new ID.\n",
      "Could not map protein_his__L_c to new ID.\n",
      "Could not map protein_hxdcacys_c to new ID.\n",
      "Could not map protein_isoasp__D_c to new ID.\n",
      "Could not map protein_isoasp__L_c to new ID.\n",
      "Could not map protein_isoaspmeo__L_c to new ID.\n",
      "Could not map protein_lacarg_c to new ID.\n",
      "Could not map protein_laccys_c to new ID.\n",
      "Could not map protein_laclys_c to new ID.\n",
      "Could not map protein_lys__L_c to new ID.\n",
      "Could not map protein_lysglu_protein_c to new ID.\n",
      "Could not map protein_mmarg__L_c to new ID.\n",
      "Could not map protein_Nproshispi_c to new ID.\n",
      "Could not map protein_Ntelehispi_c to new ID.\n",
      "Could not map protein_Nwadpribarg__L_e to new ID.\n",
      "Could not map protein_oadpribser_c to new ID.\n",
      "Could not map protein_ocdcacys_c to new ID.\n",
      "Could not map protein_pepproW0_c to new ID.\n",
      "Could not map protein_pepproW180_c to new ID.\n",
      "Could not map protein_rbllys_c to new ID.\n",
      "Could not map protein_rbllys3p_c to new ID.\n",
      "Could not map protein_rbllys5p_c to new ID.\n",
      "Could not map protein_sadpribcys_c to new ID.\n",
      "Could not map protein_sdmarg__L_c to new ID.\n",
      "Could not map protein_ser__L_c to new ID.\n",
      "Could not map protein_ser3oacgam_c to new ID.\n",
      "Could not map protein_serpi_c to new ID.\n",
      "Could not map protein_serTAg_c to new ID.\n",
      "Could not map protein_serTnAg_c to new ID.\n",
      "Could not map protein_sucim__D_c to new ID.\n",
      "Could not map protein_sucim__L_c to new ID.\n",
      "Could not map protein_thr__L_c to new ID.\n",
      "Could not map protein_thr3oacgam_c to new ID.\n",
      "Could not map protein_thrpi_c to new ID.\n",
      "Could not map protein_thrTAg_c to new ID.\n",
      "Could not map protein_thrTnAg_c to new ID.\n",
      "Could not map protein_ttdcacys_c to new ID.\n",
      "Could not map protein_tyr__L_c to new ID.\n",
      "Could not map protein_tyrpi_c to new ID.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mets</th>\n",
       "      <th>metRetired</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>35camp</td>\n",
       "      <td>camp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FA_hs_16_0</td>\n",
       "      <td>hdca</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FA_hs_18_9Z</td>\n",
       "      <td>ocdcea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FA_hs_18_9Z12Z</td>\n",
       "      <td>lnlc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FAcoa_hs_16_0</td>\n",
       "      <td>pmtcoa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>protres_ttdcacys_c</td>\n",
       "      <td>protein_ttdcacys_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>protres_tyr__L_c</td>\n",
       "      <td>protein_tyr__L_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>protres_tyrpi_c</td>\n",
       "      <td>protein_tyrpi_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>slgth__R</td>\n",
       "      <td>lgt__S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>udpglc__D</td>\n",
       "      <td>udpg</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>138 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   mets          metRetired\n",
       "0                35camp                camp\n",
       "1            FA_hs_16_0                hdca\n",
       "2           FA_hs_18_9Z              ocdcea\n",
       "3        FA_hs_18_9Z12Z                lnlc\n",
       "4         FAcoa_hs_16_0              pmtcoa\n",
       "..                  ...                 ...\n",
       "133  protres_ttdcacys_c  protein_ttdcacys_c\n",
       "134    protres_tyr__L_c    protein_tyr__L_c\n",
       "135     protres_tyrpi_c     protein_tyrpi_c\n",
       "136            slgth__R              lgt__S\n",
       "137           udpglc__D                udpg\n",
       "\n",
       "[138 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "id_mapping_dict = id_mapping_dicts[attribute_type]\n",
    "\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"metRetired\", \"mets\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    get_dirpath(\"deprecatedIdentifiers\")\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"mets\", \"metRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[\n",
    "        previous_id_mapping_df[\"mets\"] == retiring\n",
    "    ]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        retired_set_of_ids.update(\n",
    "            previously_retired[\"metRetired\"].apply(split_string).item()\n",
    "        )\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"metRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "for old, new in id_mapping_dict.items():\n",
    "    try:\n",
    "        metabolite = model.metabolites.get_by_id(old)\n",
    "    except KeyError:\n",
    "        print(f\"Could not map {old} to new ID.\")\n",
    "    else:\n",
    "        try:\n",
    "            metabolite.id = id_mapping_dict[metabolite.id]\n",
    "        except ValueError as e:\n",
    "            warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n",
    "            model.remove_metabolites([metabolite])\n",
    "\n",
    "model.repair()\n",
    "\n",
    "id_mapping_df = pd.concat((id_mapping_df, previous_id_mapping_df), axis=0)\n",
    "id_mapping_df = (\n",
    "    id_mapping_df.drop_duplicates()\n",
    "    .sort_values(\"mets\", ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "id_mapping_df.to_csv(\n",
    "    get_dirpath(\"deprecatedIdentifiers\")\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e053b43e-9c15-4c91-8032-68bf5f5ca30b",
   "metadata": {},
   "source": [
    "#### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d3ce4c3a-e273-491d-a597-f7847cd19479",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        rxns rxnRetired\n",
      "119  NTDADPe      NDPAe\n",
      "Could not map ClPItex to new ID.\n",
      "Could not map NO22Natex to new ID.\n",
      "Could not map CYSL to new ID.\n",
      "Could not map CYSTCYSL to new ID.\n",
      "Could not map SELCYSTL to new ID.\n",
      "Could not map SELMETHL to new ID.\n",
      "Could not map LCMTPPP2CA to new ID.\n",
      "Could not map PPMEPPP2CA to new ID.\n",
      "Could not map EX_protein_Nwadpribarg__L_e to new ID.\n",
      "Could not map EX_protein_arg__L_e to new ID.\n",
      "Could not map IMRBVPT to new ID.\n",
      "Could not map NTD1 to new ID.\n",
      "Could not map NTD2 to new ID.\n",
      "Could not map NTD3 to new ID.\n",
      "Could not map NTD4 to new ID.\n",
      "Could not map NTD5 to new ID.\n",
      "Could not map NTD6 to new ID.\n",
      "Could not map NTD7 to new ID.\n",
      "Could not map NTD8 to new ID.\n",
      "Could not map NTD9 to new ID.\n",
      "Could not map NTD10 to new ID.\n",
      "Could not map NTD11 to new ID.\n",
      "Could not map NTD12 to new ID.\n",
      "Could not map NTD15 to new ID.\n",
      "Could not map 2NTDA to new ID.\n",
      "Could not map 2NTDC to new ID.\n",
      "Could not map 2NTDG to new ID.\n",
      "Could not map 2NTDU to new ID.\n",
      "Could not map NTDAe to new ID.\n",
      "Could not map NDPAe to new ID.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/ipykernel_3116/3331829275.py:43: UserWarning: The model already contains a reaction with the id: GAL_Dt, this is expected if removing/renaming a duplicate\n",
      "  warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n",
      "/var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/ipykernel_3116/3331829275.py:43: UserWarning: The model already contains a reaction with the id: NDPADPe, this is expected if removing/renaming a duplicate\n",
      "  warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rxns</th>\n",
       "      <th>rxnRetired</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4PYRDXABCte</td>\n",
       "      <td>4PYRDX</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACt2</td>\n",
       "      <td>ACt2r</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADRNLt</td>\n",
       "      <td>ADRNLtu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AHC</td>\n",
       "      <td>AHCi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ARDFE2</td>\n",
       "      <td>ARD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216</th>\n",
       "      <td>UMPK1</td>\n",
       "      <td>UMPK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>UPPG3DC</td>\n",
       "      <td>UPPDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218</th>\n",
       "      <td>UPPG3S</td>\n",
       "      <td>UPP3S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>XYLTD_Dx</td>\n",
       "      <td>XYLTD_D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>XYLUR_Ly</td>\n",
       "      <td>XYLUR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>221 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            rxns rxnRetired\n",
       "0    4PYRDXABCte     4PYRDX\n",
       "1           ACt2      ACt2r\n",
       "2         ADRNLt    ADRNLtu\n",
       "3            AHC       AHCi\n",
       "4         ARDFE2        ARD\n",
       "..           ...        ...\n",
       "216        UMPK1       UMPK\n",
       "217      UPPG3DC     UPPDC1\n",
       "218       UPPG3S      UPP3S\n",
       "219     XYLTD_Dx    XYLTD_D\n",
       "220     XYLUR_Ly      XYLUR\n",
       "\n",
       "[221 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "id_mapping_dict = id_mapping_dicts[attribute_type]\n",
    "\n",
    "\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"rxnRetired\", \"rxns\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    get_dirpath(\"deprecatedIdentifiers\")\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"rxns\", \"rxnRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[\n",
    "        previous_id_mapping_df[\"rxns\"] == retiring\n",
    "    ]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        print(previously_retired)\n",
    "        retired_set_of_ids.update(\n",
    "            previously_retired[\"rxnRetired\"].apply(split_string).item()\n",
    "        )\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"rxnRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "for old, new in id_mapping_dict.items():\n",
    "    try:\n",
    "        reaction = model.reactions.get_by_id(old)\n",
    "    except KeyError:\n",
    "        print(f\"Could not map {old} to new ID.\")\n",
    "    else:\n",
    "        try:\n",
    "            reaction.id = id_mapping_dict[reaction.id]\n",
    "        except ValueError as e:\n",
    "            warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n",
    "            model.remove_reactions([reaction])\n",
    "\n",
    "model.repair()\n",
    "\n",
    "id_mapping_df = pd.concat((id_mapping_df, previous_id_mapping_df), axis=0)\n",
    "id_mapping_df = (\n",
    "    id_mapping_df.drop_duplicates()\n",
    "    .sort_values(\"rxns\", ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "id_mapping_df.to_csv(\n",
    "    get_dirpath(\"deprecatedIdentifiers\")\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "624e20c3-db6d-4ec6-96fc-9600cc998e1f",
   "metadata": {},
   "source": [
    "### Update model based on curation and proteomic evidence\n",
    "* Add metabolites first, then genes and lastly, reactions\n",
    "* Add metabolites with mass and charge balanced formulas (ChemAxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "03deaf1e-18e7-4ce7-ae1b-597650a70cd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_updated = {}\n",
    "dataframes_evidence = {}\n",
    "dataframes_removed = {}\n",
    "overwrite = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b32955-a0d5-45c6-81e4-183760ce44d0",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e073dc66-8abb-48e3-9c70-7dc447806c2d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>metabolite</th>\n",
       "      <th>name</th>\n",
       "      <th>formula</th>\n",
       "      <th>charge</th>\n",
       "      <th>compartment</th>\n",
       "      <th>metabolomic evidence (#studies)</th>\n",
       "      <th>metabolomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2amac_c</td>\n",
       "      <td>2amac</td>\n",
       "      <td>2-aminoprop-2-enoic acid</td>\n",
       "      <td>C3H5NO2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2imppa_c</td>\n",
       "      <td>2imppa</td>\n",
       "      <td>2-Iminopropanoate</td>\n",
       "      <td>C3H5NO2</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2mop_c</td>\n",
       "      <td>2mop</td>\n",
       "      <td>2-methyl-3-oxopropanoate</td>\n",
       "      <td>C4H5O3</td>\n",
       "      <td>-1</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>34dhmald_c</td>\n",
       "      <td>34dhmald</td>\n",
       "      <td>3,4-dihydroxymandelaldehyde</td>\n",
       "      <td>C8H8O4</td>\n",
       "      <td>0</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>34dhoxmand_c</td>\n",
       "      <td>34dhoxmand</td>\n",
       "      <td>3,4-dihydroxymandelate</td>\n",
       "      <td>C8H7O5</td>\n",
       "      <td>-1</td>\n",
       "      <td>c</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219</th>\n",
       "      <td>3mob_e</td>\n",
       "      <td>3mob</td>\n",
       "      <td>3-methyl-2-oxobutyrate</td>\n",
       "      <td>C5H7O3</td>\n",
       "      <td>-1</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>220</th>\n",
       "      <td>crm_hs_e</td>\n",
       "      <td>crm_hs</td>\n",
       "      <td>Ceramide (Cer d18:2)</td>\n",
       "      <td>C19H36NO3R</td>\n",
       "      <td>0</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>221</th>\n",
       "      <td>alltn_e</td>\n",
       "      <td>alltn</td>\n",
       "      <td>Allantoin</td>\n",
       "      <td>C4H6N4O3</td>\n",
       "      <td>0</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>3dfru_e</td>\n",
       "      <td>3dfru</td>\n",
       "      <td>3-deoxyfructose</td>\n",
       "      <td>C6H12O5</td>\n",
       "      <td>0</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>mtthiogmp_e</td>\n",
       "      <td>mtthiogmp</td>\n",
       "      <td>6-methylthioguanosine monophosphate</td>\n",
       "      <td>C11H14N5O7PS</td>\n",
       "      <td>-2</td>\n",
       "      <td>e</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>224 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      metabolites  metabolite                                 name  \\\n",
       "0         2amac_c       2amac             2-aminoprop-2-enoic acid   \n",
       "1        2imppa_c      2imppa                    2-Iminopropanoate   \n",
       "2          2mop_c        2mop             2-methyl-3-oxopropanoate   \n",
       "3      34dhmald_c    34dhmald          3,4-dihydroxymandelaldehyde   \n",
       "4    34dhoxmand_c  34dhoxmand               3,4-dihydroxymandelate   \n",
       "..            ...         ...                                  ...   \n",
       "219        3mob_e        3mob               3-methyl-2-oxobutyrate   \n",
       "220      crm_hs_e      crm_hs                 Ceramide (Cer d18:2)   \n",
       "221       alltn_e       alltn                            Allantoin   \n",
       "222       3dfru_e       3dfru                      3-deoxyfructose   \n",
       "223   mtthiogmp_e   mtthiogmp  6-methylthioguanosine monophosphate   \n",
       "\n",
       "          formula  charge compartment metabolomic evidence (#studies)  \\\n",
       "0         C3H5NO2       0           c                                   \n",
       "1         C3H5NO2       0           c                                   \n",
       "2          C4H5O3      -1           c                                   \n",
       "3          C8H8O4       0           c                                   \n",
       "4          C8H7O5      -1           c                                   \n",
       "..            ...     ...         ...                             ...   \n",
       "219        C5H7O3      -1           e                                   \n",
       "220    C19H36NO3R       0           e                                   \n",
       "221      C4H6N4O3       0           e                                   \n",
       "222       C6H12O5       0           e                                   \n",
       "223  C11H14N5O7PS      -2           e                                   \n",
       "\n",
       "    metabolomic evidence (pubmed) references notes  \n",
       "0                                                   \n",
       "1                                                   \n",
       "2                                                   \n",
       "3                                                   \n",
       "4                                                   \n",
       "..                            ...        ...   ...  \n",
       "219                                                 \n",
       "220                                                 \n",
       "221                                                 \n",
       "222                                                 \n",
       "223                                                 \n",
       "\n",
       "[224 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "\n",
    "df_updated = pd.read_csv(\n",
    "    get_dirpath(\"curated\") / f\"{attribute_type}_updated_1.2.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "\n",
    "attr_cols = [\"metabolites\", \"name\", \"formula\", \"charge\", \"compartment\"]\n",
    "for idx, row in df_updated.iterrows():\n",
    "    mid, name, formula, charge, compartment = row[attr_cols]\n",
    "    if not model.metabolites.has_id(mid):\n",
    "        # Add reaction to model\n",
    "        model.add_metabolites([Metabolite(mid)])\n",
    "        # print(mid)\n",
    "\n",
    "    metabolite = model.metabolites.get_by_id(mid)\n",
    "    metabolite.name = name\n",
    "    metabolite.formula = formula\n",
    "    metabolite.charge = int(charge)\n",
    "    metabolite.compartment = compartment\n",
    "\n",
    "    annotations_dict = row[~row.index.isin(attr_cols)].to_dict()\n",
    "    notes = annotations_dict.pop(\"notes\")\n",
    "    if notes:\n",
    "        metabolite.notes.update({\"notes\": str(notes)})\n",
    "    annotations_dict = {\n",
    "        k: v\n",
    "        for k, v in annotations_dict.items()\n",
    "        if not k.startswith(\"metabolomic\") and v\n",
    "    }\n",
    "    metabolite.annotation.update(annotations_dict)\n",
    "\n",
    "met_formulas = standardardize_metabolite_formulas(\n",
    "    dict(zip(model.metabolites.list_attr(\"id\"), model.metabolites.list_attr(\"formula\")))\n",
    ")\n",
    "for mid, formula in met_formulas.items():\n",
    "    model.metabolites.get_by_id(mid).formula = formula\n",
    "\n",
    "\n",
    "dataframes_updated[attribute_type] = df_updated\n",
    "try:\n",
    "    df_previous_evidence = pd.read_csv(\n",
    "        get_dirpath(\"curated\") / f\"{attribute_type}_evidence.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0)\n",
    "df_evidence[attribute_type] = df_evidence[attribute_type].replace(\n",
    "    id_mapping_dicts[attribute_type]\n",
    ")\n",
    "df_evidence = (\n",
    "    df_evidence.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")[df_previous_evidence.columns]\n",
    "\n",
    "dataframes_evidence[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "142935ef-76b4-4ccb-8ba6-2a3f999c2f2b",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "852b2f5f-4195-464b-b880-63ada6769a84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>proteomic evidence (pubmed)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACER2</td>\n",
       "      <td>Q5QJU3</td>\n",
       "      <td>340485</td>\n",
       "      <td>ACER2</td>\n",
       "      <td>0.0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:20207939;pubmed:35886939;pubmed:2940161...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACER3</td>\n",
       "      <td>Q9NUN7</td>\n",
       "      <td>55331</td>\n",
       "      <td>ACER3</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>pubmed:20207939;pubmed:35886939;pubmed:29401619</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACSS2</td>\n",
       "      <td>Q9NR19</td>\n",
       "      <td>55902</td>\n",
       "      <td>ACSS2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>pubmed:30327373;pubmed:38260479</td>\n",
       "      <td>pubmed:4379089;pubmed:6115779;pubmed:38513237</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTA1</td>\n",
       "      <td>P68133</td>\n",
       "      <td>58</td>\n",
       "      <td>ACTA1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>pubmed:16861337;pubmed:18614565;pubmed:2777734...</td>\n",
       "      <td>pubmed:284418;pubmed:6682109;pubmed:25717184;p...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>P60709</td>\n",
       "      <td>60</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>19.0</td>\n",
       "      <td>pubmed:12362340;pubmed:16861337;pubmed:1839964...</td>\n",
       "      <td>pubmed:284418;pubmed:6682109;pubmed:25717184;p...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>VCPIP1</td>\n",
       "      <td>Q96JH7</td>\n",
       "      <td>80124</td>\n",
       "      <td>VCPIP1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>VPS4A</td>\n",
       "      <td>Q9UN37</td>\n",
       "      <td>27183</td>\n",
       "      <td>VPS4A</td>\n",
       "      <td>11.0</td>\n",
       "      <td>pubmed:18494517;pubmed:18614565;pubmed:2826317...</td>\n",
       "      <td>pubmed:33186543</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>VPS4B</td>\n",
       "      <td>O75351</td>\n",
       "      <td>9525</td>\n",
       "      <td>VPS4B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>WDR77</td>\n",
       "      <td>Q9BQA1</td>\n",
       "      <td>79084</td>\n",
       "      <td>WDR77</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>YES1</td>\n",
       "      <td>P07947</td>\n",
       "      <td>7525</td>\n",
       "      <td>YES1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>155 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes uniprot ncbigene hgnc.symbol proteomic evidence (#studies)  \\\n",
       "0     ACER2  Q5QJU3   340485       ACER2                           0.0   \n",
       "1     ACER3  Q9NUN7    55331       ACER3                                 \n",
       "2     ACSS2  Q9NR19    55902       ACSS2                           2.0   \n",
       "3     ACTA1  P68133       58       ACTA1                           8.0   \n",
       "4      ACTB  P60709       60        ACTB                          19.0   \n",
       "..      ...     ...      ...         ...                           ...   \n",
       "150  VCPIP1  Q96JH7    80124      VCPIP1                                 \n",
       "151   VPS4A  Q9UN37    27183       VPS4A                          11.0   \n",
       "152   VPS4B  O75351     9525       VPS4B                                 \n",
       "153   WDR77  Q9BQA1    79084       WDR77                                 \n",
       "154    YES1  P07947     7525        YES1                                 \n",
       "\n",
       "                           proteomic evidence (pubmed)  \\\n",
       "0                                                        \n",
       "1                                                        \n",
       "2                      pubmed:30327373;pubmed:38260479   \n",
       "3    pubmed:16861337;pubmed:18614565;pubmed:2777734...   \n",
       "4    pubmed:12362340;pubmed:16861337;pubmed:1839964...   \n",
       "..                                                 ...   \n",
       "150                                                      \n",
       "151  pubmed:18494517;pubmed:18614565;pubmed:2826317...   \n",
       "152                                                      \n",
       "153                                                      \n",
       "154                                                      \n",
       "\n",
       "                                            references notes  \n",
       "0    pubmed:20207939;pubmed:35886939;pubmed:2940161...        \n",
       "1      pubmed:20207939;pubmed:35886939;pubmed:29401619        \n",
       "2        pubmed:4379089;pubmed:6115779;pubmed:38513237        \n",
       "3    pubmed:284418;pubmed:6682109;pubmed:25717184;p...        \n",
       "4    pubmed:284418;pubmed:6682109;pubmed:25717184;p...        \n",
       "..                                                 ...   ...  \n",
       "150                                                           \n",
       "151                                    pubmed:33186543        \n",
       "152                                                           \n",
       "153                                                           \n",
       "154                                                           \n",
       "\n",
       "[155 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "\n",
    "df_updated = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / f\"{attribute_type}_updated_1.2.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "\n",
    "attr_cols = [\"genes\"]\n",
    "for idx, row in df_updated.iterrows():\n",
    "    # Between these three annotation fields, should be likely that the rest can get extracted from UniProt.\n",
    "    gid = row[attr_cols[0]]\n",
    "    if not model.genes.has_id(gid):\n",
    "        gene = Gene(gid)\n",
    "        model.genes.extend([gene])\n",
    "        # print(gene.id)\n",
    "    gene = model.genes.get_by_id(gid)\n",
    "    annotations_dict = row[~row.index.isin(attr_cols)].to_dict()\n",
    "    annotations_dict = {\n",
    "        k: v for k, v in annotations_dict.items() if not k.startswith(\"proteomic\") and v\n",
    "    }\n",
    "    gene.annotation.update(annotations_dict)\n",
    "\n",
    "\n",
    "dataframes_updated[attribute_type] = df_updated\n",
    "try:\n",
    "    df_previous_evidence = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_evidence.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "\n",
    "df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0)\n",
    "df_evidence[attribute_type] = df_evidence[attribute_type].replace(\n",
    "    id_mapping_dicts[attribute_type]\n",
    ")\n",
    "df_evidence = (\n",
    "    df_evidence.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")[df_previous_evidence.columns]\n",
    "\n",
    "dataframes_evidence[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3b1b3476-7c28-4164-a425-87ea22f92085",
   "metadata": {},
   "source": [
    "#### Reactions\n",
    "* Addition of hemoglobin glycation\n",
    "* All intracellular reactions converted to use/consume NH4, which is dominant at pH 7.25. Transport and protonation reactions for NH3 kept.\n",
    "* Add NEDD8 neddylation and fix ubiquitin reaction GPRs accordingly\n",
    "* Reversibility updates for phosphoribosyltransferase reactions\n",
    "* GPR updates according to Complex Portal and proteomic data\n",
    "    * Add PRMT5 methylsome complex\n",
    "    * Add PI3K class III complex proteins\n",
    "    * Multiple Ubiquination complexes\n",
    "    * Multiple VCP complexes and other ATPases\n",
    "* Add IDH2 and MDH2 to GPRs\n",
    "* Add SLC22A4 and SLC22A16 GPRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff19985d-2a62-4283-8da0-6b8177927b92",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>reaction</th>\n",
       "      <th>gene reaction rule</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>spontaneous</th>\n",
       "      <th>proteomic evidence (#studies)</th>\n",
       "      <th>references</th>\n",
       "      <th>notes</th>\n",
       "      <th>metatlas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5FLUDPK</td>\n",
       "      <td>Nucleoside-diphosphate kinase (ATP:5-fluorouri...</td>\n",
       "      <td>atp_c + 5fludp_c --&gt; adp_c + 5flutp_c</td>\n",
       "      <td>NME1 or (NME1 and NME2) or NME2 or (NME1 and N...</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5FLUMPK</td>\n",
       "      <td>Uridylate kinase (ATP:5-fluorouridine 5-monoph...</td>\n",
       "      <td>atp_c + 5flump_c --&gt; adp_c + 5fludp_c</td>\n",
       "      <td>CMPK1</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5FLURAPT</td>\n",
       "      <td>5-fluorouracil phosphoribosyltransferase</td>\n",
       "      <td>5flura_c + prpp_c --&gt; 5flump_c + ppi_c</td>\n",
       "      <td>UMPS</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D5FURADy</td>\n",
       "      <td>5,6-dihydro-5-fluorouracil:NADP + 5-oxidoreduc...</td>\n",
       "      <td>5flura_c + nadph_c + h_c --&gt; 56dh5flura_c + na...</td>\n",
       "      <td>DPYD</td>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L2HGDHx</td>\n",
       "      <td>L-2-hydroxyglutarate dehydrogenase</td>\n",
       "      <td>akg_c + h_c + nadh_c --&gt; 2hglut__L_c + nad_c</td>\n",
       "      <td>LDHB or LDHA or PHGDH or MDH1 or MDH2 or (LDHA...</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>LDHA (22);LDHB (22);MDH1 (16);PHGDH (2)</td>\n",
       "      <td>pubmed:27813142;pubmed:31473074</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534</th>\n",
       "      <td>UBIQE1T</td>\n",
       "      <td>[E1 ubiquitin-activating enzyme]-S-ubiquitinyl...</td>\n",
       "      <td>accprot_lys__L_c + ubiqSe1act_cys__L_c --&gt; acc...</td>\n",
       "      <td>UBE2O or CDC34</td>\n",
       "      <td>Ubiquitin mediated proteolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>CDC34 (5);UBE2O (16)</td>\n",
       "      <td>pubmed:15095787;pubmed:21508250;pubmed:3456453...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535</th>\n",
       "      <td>UBIQE2CONJT</td>\n",
       "      <td>[E2 ubiquitin-conjugating enzyme]-S-ubiquitiny...</td>\n",
       "      <td>accprot_lys__L_c + ubiqSe2conj_cys__L_c --&gt; ac...</td>\n",
       "      <td>UBR4 or HUWE1 or (UBAC1 and RNF123) or (RBX1 a...</td>\n",
       "      <td>Ubiquitin mediated proteolysis</td>\n",
       "      <td>0</td>\n",
       "      <td>CAND1 (18);CUL1 (18);HUWE1 (16);RBX1 (11);RNF1...</td>\n",
       "      <td>pubmed:15095787;pubmed:21508250;pubmed:3456453...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <td>MMALNDHx</td>\n",
       "      <td>(S)-methylmalonate semialdehyde:NAD + oxidored...</td>\n",
       "      <td>h2o_c + nad_c + 2mop_c --&gt; 2.0 h_c + nadh_c + ...</td>\n",
       "      <td>ALDH9A1 or ALDH3A2</td>\n",
       "      <td>Valine, leucine, and isoleucine metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>MAR04797</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>537</th>\n",
       "      <td>PYDXPPe</td>\n",
       "      <td>Pyridoxal 5-phosphate phosphatase</td>\n",
       "      <td>h2o_e + pydx5p_e --&gt; pi_e + pydx_e</td>\n",
       "      <td>ALPL</td>\n",
       "      <td>Vitamin B6 metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>pubmed:1322411;pubmed:14522954;pubmed:7448195;...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>538</th>\n",
       "      <td>AVITE1RCYTB5R</td>\n",
       "      <td>Reduction of alpha-tocopherol radical via L-as...</td>\n",
       "      <td>avite1r_c + focytb5_c + h_c --&gt; avite1_c + fic...</td>\n",
       "      <td>(CYB5R3 and CYB5B) or (CYB5R3 and CYB5A)</td>\n",
       "      <td>Vitamin E metabolism</td>\n",
       "      <td>0</td>\n",
       "      <td>CYB5A (10);CYB5B (9);CYB5R3 (21)</td>\n",
       "      <td>pubmed:8388377</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>539 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reactions                                               name  \\\n",
       "0          5FLUDPK  Nucleoside-diphosphate kinase (ATP:5-fluorouri...   \n",
       "1          5FLUMPK  Uridylate kinase (ATP:5-fluorouridine 5-monoph...   \n",
       "2         5FLURAPT           5-fluorouracil phosphoribosyltransferase   \n",
       "3         D5FURADy  5,6-dihydro-5-fluorouracil:NADP + 5-oxidoreduc...   \n",
       "4          L2HGDHx                 L-2-hydroxyglutarate dehydrogenase   \n",
       "..             ...                                                ...   \n",
       "534        UBIQE1T  [E1 ubiquitin-activating enzyme]-S-ubiquitinyl...   \n",
       "535    UBIQE2CONJT  [E2 ubiquitin-conjugating enzyme]-S-ubiquitiny...   \n",
       "536       MMALNDHx  (S)-methylmalonate semialdehyde:NAD + oxidored...   \n",
       "537        PYDXPPe                  Pyridoxal 5-phosphate phosphatase   \n",
       "538  AVITE1RCYTB5R  Reduction of alpha-tocopherol radical via L-as...   \n",
       "\n",
       "                                              reaction  \\\n",
       "0                atp_c + 5fludp_c --> adp_c + 5flutp_c   \n",
       "1                atp_c + 5flump_c --> adp_c + 5fludp_c   \n",
       "2               5flura_c + prpp_c --> 5flump_c + ppi_c   \n",
       "3    5flura_c + nadph_c + h_c --> 56dh5flura_c + na...   \n",
       "4         akg_c + h_c + nadh_c --> 2hglut__L_c + nad_c   \n",
       "..                                                 ...   \n",
       "534  accprot_lys__L_c + ubiqSe1act_cys__L_c --> acc...   \n",
       "535  accprot_lys__L_c + ubiqSe2conj_cys__L_c --> ac...   \n",
       "536  h2o_c + nad_c + 2mop_c --> 2.0 h_c + nadh_c + ...   \n",
       "537                 h2o_e + pydx5p_e --> pi_e + pydx_e   \n",
       "538  avite1r_c + focytb5_c + h_c --> avite1_c + fic...   \n",
       "\n",
       "                                    gene reaction rule  \\\n",
       "0    NME1 or (NME1 and NME2) or NME2 or (NME1 and N...   \n",
       "1                                                CMPK1   \n",
       "2                                                 UMPS   \n",
       "3                                                 DPYD   \n",
       "4    LDHB or LDHA or PHGDH or MDH1 or MDH2 or (LDHA...   \n",
       "..                                                 ...   \n",
       "534                                     UBE2O or CDC34   \n",
       "535  UBR4 or HUWE1 or (UBAC1 and RNF123) or (RBX1 a...   \n",
       "536                                 ALDH9A1 or ALDH3A2   \n",
       "537                                               ALPL   \n",
       "538           (CYB5R3 and CYB5B) or (CYB5R3 and CYB5A)   \n",
       "\n",
       "                                       subsystem  spontaneous  \\\n",
       "0                      5-fluorouracil metabolism            0   \n",
       "1                      5-fluorouracil metabolism            0   \n",
       "2                      5-fluorouracil metabolism            0   \n",
       "3                      5-fluorouracil metabolism            0   \n",
       "4    Alanine, aspartate and glutamate metabolism            0   \n",
       "..                                           ...          ...   \n",
       "534               Ubiquitin mediated proteolysis            0   \n",
       "535               Ubiquitin mediated proteolysis            0   \n",
       "536   Valine, leucine, and isoleucine metabolism            0   \n",
       "537                        Vitamin B6 metabolism            0   \n",
       "538                         Vitamin E metabolism            0   \n",
       "\n",
       "                         proteomic evidence (#studies)  \\\n",
       "0                                                        \n",
       "1                                                        \n",
       "2                                                        \n",
       "3                                                        \n",
       "4              LDHA (22);LDHB (22);MDH1 (16);PHGDH (2)   \n",
       "..                                                 ...   \n",
       "534                               CDC34 (5);UBE2O (16)   \n",
       "535  CAND1 (18);CUL1 (18);HUWE1 (16);RBX1 (11);RNF1...   \n",
       "536                                                      \n",
       "537                                                      \n",
       "538                   CYB5A (10);CYB5B (9);CYB5R3 (21)   \n",
       "\n",
       "                                            references notes  metatlas  \n",
       "0                                                                       \n",
       "1                                                                       \n",
       "2                                                                       \n",
       "3                                                                       \n",
       "4                      pubmed:27813142;pubmed:31473074                  \n",
       "..                                                 ...   ...       ...  \n",
       "534  pubmed:15095787;pubmed:21508250;pubmed:3456453...                  \n",
       "535  pubmed:15095787;pubmed:21508250;pubmed:3456453...                  \n",
       "536                                                           MAR04797  \n",
       "537  pubmed:1322411;pubmed:14522954;pubmed:7448195;...                  \n",
       "538                                     pubmed:8388377                  \n",
       "\n",
       "[539 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "\n",
    "df_updated = pd.read_csv(\n",
    "    get_dirpath(\"curated\") / f\"{attribute_type}_updated_1.2.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "attr_cols = [\"reactions\", \"name\", \"reaction\", \"gene reaction rule\", \"subsystem\"]\n",
    "for idx, row in df_updated.iterrows():\n",
    "    rid, name, reaction_str, gpr, subsystem = row[attr_cols]\n",
    "    try:\n",
    "        if not model.reactions.has_id(rid):\n",
    "            # Add reaction to model\n",
    "            model.add_reactions([Reaction(rid)])\n",
    "    except ValueError as e:\n",
    "        print(idx, rid, name)\n",
    "        raise e\n",
    "\n",
    "    reaction = model.reactions.get_by_id(rid)\n",
    "    reaction.build_reaction_from_string(reaction_str)\n",
    "    reaction.name = name\n",
    "    reaction.gene_reaction_rule = gpr\n",
    "    reaction.subsystem = subsystem\n",
    "\n",
    "    annotations_dict = row[~row.index.isin(attr_cols)].to_dict()\n",
    "    notes = annotations_dict.pop(\"notes\")\n",
    "    if notes:\n",
    "        reaction.notes.update({\"notes\": str(notes)})\n",
    "    if \"proteomic evidence (#studies)\" in annotations_dict:\n",
    "        annotations_dict.pop(\"proteomic evidence (#studies)\")\n",
    "    annotations_dict = {\n",
    "        k: v for k, v in annotations_dict.items() if not k.startswith(\"proteomic\") and v\n",
    "    }\n",
    "    reaction.annotation.update(annotations_dict)\n",
    "\n",
    "\n",
    "dataframes_updated[attribute_type] = df_updated\n",
    "try:\n",
    "    df_previous_evidence = pd.read_csv(\n",
    "        get_dirpath(\"curated\") / f\"{attribute_type}_evidence.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0)\n",
    "df_evidence[attribute_type] = df_evidence[attribute_type].replace(\n",
    "    id_mapping_dicts[attribute_type]\n",
    ")\n",
    "df_evidence = (\n",
    "    df_evidence.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(\n",
    "        [\"subsystem\", \"reactions\"],\n",
    "        ascending=[True, True],\n",
    "    )\n",
    "    .reset_index(drop=True)\n",
    ")[df_previous_evidence.columns]\n",
    "\n",
    "dataframes_evidence[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64d45450-d19e-4e09-b76b-8ad6471b8153",
   "metadata": {},
   "source": [
    "### Refine model through removing items\n",
    "#### Reactions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d434a25-de3f-4e6b-98be-bb902e3092eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3MOXTYRESSte</td>\n",
       "      <td>3-Methoxytyramine secretion via secretory vesi...</td>\n",
       "      <td>Replaced by simple transport reaction (3MOXTYR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALAt4</td>\n",
       "      <td>Alanine-Sodium symporter</td>\n",
       "      <td>Replaced by reactions involving the amino acid...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BANDMT</td>\n",
       "      <td>Band membrane protein-methyltransferase</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BILGLCURt</td>\n",
       "      <td>Bilirubin monoglucuronide transport via bicarb...</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BILIRBU</td>\n",
       "      <td>Bilirubin UDP-glucuronosyltransferase</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>171</th>\n",
       "      <td>SK_pe_hs_18_9Z_18_9Z_c</td>\n",
       "      <td>Sink pe hs 18 1 18 1(c)</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172</th>\n",
       "      <td>THMMPtrbc</td>\n",
       "      <td>Thiamine monophosphate transport (passive - re...</td>\n",
       "      <td>Replaced by reactions involving reduced folate...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173</th>\n",
       "      <td>TMDPPK</td>\n",
       "      <td>Thiamine diphosphate kinase</td>\n",
       "      <td>Replaced by ADK7. Adenylate kinase isozyme 1 h...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>UDPGD</td>\n",
       "      <td>UDPglucose 6-dehydrogenase</td>\n",
       "      <td>No proteomic or activity evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>UDPGNP</td>\n",
       "      <td>UDP-glucuronate glucuronohydrolase</td>\n",
       "      <td>No proteomic or activity evidence</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  reactions  \\\n",
       "0              3MOXTYRESSte   \n",
       "1                     ALAt4   \n",
       "2                    BANDMT   \n",
       "3                 BILGLCURt   \n",
       "4                   BILIRBU   \n",
       "..                      ...   \n",
       "171  SK_pe_hs_18_9Z_18_9Z_c   \n",
       "172               THMMPtrbc   \n",
       "173                  TMDPPK   \n",
       "174                   UDPGD   \n",
       "175                  UDPGNP   \n",
       "\n",
       "                                                  name  \\\n",
       "0    3-Methoxytyramine secretion via secretory vesi...   \n",
       "1                             Alanine-Sodium symporter   \n",
       "2              Band membrane protein-methyltransferase   \n",
       "3    Bilirubin monoglucuronide transport via bicarb...   \n",
       "4                Bilirubin UDP-glucuronosyltransferase   \n",
       "..                                                 ...   \n",
       "171                            Sink pe hs 18 1 18 1(c)   \n",
       "172  Thiamine monophosphate transport (passive - re...   \n",
       "173                        Thiamine diphosphate kinase   \n",
       "174                         UDPglucose 6-dehydrogenase   \n",
       "175                 UDP-glucuronate glucuronohydrolase   \n",
       "\n",
       "                                                 notes  \n",
       "0    Replaced by simple transport reaction (3MOXTYR...  \n",
       "1    Replaced by reactions involving the amino acid...  \n",
       "2    Replaced by generic methylation reactions. Ban...  \n",
       "3    Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "4    Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "..                                                 ...  \n",
       "171  Lipids are pooled, replaced with pooled versio...  \n",
       "172  Replaced by reactions involving reduced folate...  \n",
       "173  Replaced by ADK7. Adenylate kinase isozyme 1 h...  \n",
       "174                  No proteomic or activity evidence  \n",
       "175                  No proteomic or activity evidence  \n",
       "\n",
       "[176 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"reactions\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for reaction in df_removed[attribute_type]:\n",
    "    try:\n",
    "        reaction = getattr(model, attribute_type).get_by_id(reaction)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(reaction)\n",
    "\n",
    "model.remove_reactions(to_remove)\n",
    "# Clean up removal file before archival\n",
    "df_removed = (\n",
    "    df_removed.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "    ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "        df_removed[attribute_type].values\n",
    "    )\n",
    "]\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7279505a-5e9f-44f3-b4ba-33e9b957d6e2",
   "metadata": {},
   "source": [
    "#### Genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8987ef03-a8fb-433e-8396-99a6d9bd6d1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>ncbigene</th>\n",
       "      <th>hgnc.symbol</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCC10</td>\n",
       "      <td>Q5T3U5</td>\n",
       "      <td>89845</td>\n",
       "      <td>ABCC10</td>\n",
       "      <td>evidence indicates other enzymes are present i...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ADCY1</td>\n",
       "      <td>Q08828</td>\n",
       "      <td>107</td>\n",
       "      <td>ADCY1</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ADCY3</td>\n",
       "      <td>O60266</td>\n",
       "      <td>109</td>\n",
       "      <td>ADCY3</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ADCY4</td>\n",
       "      <td>Q8NFM4</td>\n",
       "      <td>196883</td>\n",
       "      <td>ADCY4</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ADCY5</td>\n",
       "      <td>O95622</td>\n",
       "      <td>111</td>\n",
       "      <td>ADCY5</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>TAT</td>\n",
       "      <td>P17735</td>\n",
       "      <td>6898</td>\n",
       "      <td>TAT</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>UGDH</td>\n",
       "      <td>O60701</td>\n",
       "      <td>7358</td>\n",
       "      <td>UGDH</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>P22309</td>\n",
       "      <td>54658</td>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>P22310</td>\n",
       "      <td>54657</td>\n",
       "      <td>UGT1A4</td>\n",
       "      <td>no proteomic evidence.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>ZNF384</td>\n",
       "      <td>Q8TF68</td>\n",
       "      <td>171017</td>\n",
       "      <td>ZNF384</td>\n",
       "      <td>no proteomic evidence. evidence indicates othe...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      genes uniprot ncbigene hgnc.symbol  \\\n",
       "0    ABCC10  Q5T3U5    89845      ABCC10   \n",
       "1     ADCY1  Q08828      107       ADCY1   \n",
       "2     ADCY3  O60266      109       ADCY3   \n",
       "3     ADCY4  Q8NFM4   196883       ADCY4   \n",
       "4     ADCY5  O95622      111       ADCY5   \n",
       "..      ...     ...      ...         ...   \n",
       "110     TAT  P17735     6898         TAT   \n",
       "111    UGDH  O60701     7358        UGDH   \n",
       "112  UGT1A1  P22309    54658      UGT1A1   \n",
       "113  UGT1A4  P22310    54657      UGT1A4   \n",
       "114  ZNF384  Q8TF68   171017      ZNF384   \n",
       "\n",
       "                                                 notes  \n",
       "0    evidence indicates other enzymes are present i...  \n",
       "1    no proteomic evidence. evidence indicates othe...  \n",
       "2    no proteomic evidence. evidence indicates othe...  \n",
       "3    no proteomic evidence. evidence indicates othe...  \n",
       "4    no proteomic evidence. evidence indicates othe...  \n",
       "..                                                 ...  \n",
       "110  no proteomic evidence. evidence indicates othe...  \n",
       "111                             no proteomic evidence.  \n",
       "112                             no proteomic evidence.  \n",
       "113                             no proteomic evidence.  \n",
       "114  no proteomic evidence. evidence indicates othe...  \n",
       "\n",
       "[115 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"genes\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for gene in df_removed[attribute_type]:\n",
    "    try:\n",
    "        gene = model.genes.get_by_id(gene)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(gene)\n",
    "model.genes -= to_remove\n",
    "\n",
    "# Clean up removal file before archival\n",
    "df_removed = (\n",
    "    df_removed.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "for gene in model.genes:\n",
    "    if gene.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {gene}\")\n",
    "\n",
    "dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "    ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "        df_removed[attribute_type].values\n",
    "    )\n",
    "]\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab7e51b-9730-439a-a0ce-271b65d72dcd",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d9b992c-9399-45aa-921f-c415db9e0178",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metabolites</th>\n",
       "      <th>name</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>band_c</td>\n",
       "      <td>Band membrane protein (universal, erythrocyte ...</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bandmt_c</td>\n",
       "      <td>Band membrane protein (methylated, universal, ...</td>\n",
       "      <td>Replaced by generic methylation reactions. Ban...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bilglcur_c</td>\n",
       "      <td>Bilirubin monoglucuronide</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>bilglcur_e</td>\n",
       "      <td>Bilirubin monoglucuronide</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cdpdag_hs_16_0_16_0_c</td>\n",
       "      <td>CDP diacylglycerol (homo sapiens, C16:0, C16:0)</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>pe_hs_18_9Z12Z_18_9Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:2,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>pe_hs_18_9Z_18_9Z12Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:1,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>pe_hs_18_9Z_18_9Z_c</td>\n",
       "      <td>Phosphatidylethanolamine (homo sapiens, C18:1,...</td>\n",
       "      <td>Lipids are pooled, replaced with pooled versio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>pppg9_c</td>\n",
       "      <td>protoporphyrinogen IX</td>\n",
       "      <td>Protoporphyrinogen oxidase is a mitochondrial ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>udpglcur_c</td>\n",
       "      <td>UDP-D-glucuronate</td>\n",
       "      <td>Heme oxygenase occurs in macrophages. Furtherm...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               metabolites                                               name  \\\n",
       "0                   band_c  Band membrane protein (universal, erythrocyte ...   \n",
       "1                 bandmt_c  Band membrane protein (methylated, universal, ...   \n",
       "2               bilglcur_c                          Bilirubin monoglucuronide   \n",
       "3               bilglcur_e                          Bilirubin monoglucuronide   \n",
       "4    cdpdag_hs_16_0_16_0_c    CDP diacylglycerol (homo sapiens, C16:0, C16:0)   \n",
       "..                     ...                                                ...   \n",
       "65  pe_hs_18_9Z12Z_18_9Z_c  Phosphatidylethanolamine (homo sapiens, C18:2,...   \n",
       "66  pe_hs_18_9Z_18_9Z12Z_c  Phosphatidylethanolamine (homo sapiens, C18:1,...   \n",
       "67     pe_hs_18_9Z_18_9Z_c  Phosphatidylethanolamine (homo sapiens, C18:1,...   \n",
       "68                 pppg9_c                              protoporphyrinogen IX   \n",
       "69              udpglcur_c                                  UDP-D-glucuronate   \n",
       "\n",
       "                                                notes  \n",
       "0   Replaced by generic methylation reactions. Ban...  \n",
       "1   Replaced by generic methylation reactions. Ban...  \n",
       "2   Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "3   Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "4   Lipids are pooled, replaced with pooled versio...  \n",
       "..                                                ...  \n",
       "65  Lipids are pooled, replaced with pooled versio...  \n",
       "66  Lipids are pooled, replaced with pooled versio...  \n",
       "67  Lipids are pooled, replaced with pooled versio...  \n",
       "68  Protoporphyrinogen oxidase is a mitochondrial ...  \n",
       "69  Heme oxygenase occurs in macrophages. Furtherm...  \n",
       "\n",
       "[70 rows x 3 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        get_dirpath(\"curation\") / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for metabolite in df_removed[attribute_type]:\n",
    "    try:\n",
    "        metabolite = model.metabolites.get_by_id(metabolite)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    model.remove_metabolites([metabolite])\n",
    "model.remove_metabolites(to_remove)\n",
    "\n",
    "# Clean up removal file before archival\n",
    "df_removed = (\n",
    "    df_removed.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {metabolite}\")\n",
    "\n",
    "dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "    ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "        df_removed[attribute_type].values\n",
    "    )\n",
    "]\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4cfe3007-e6fe-4fdd-97a7-4f764d13ae57",
   "metadata": {},
   "outputs": [],
   "source": [
    "for attribute_type in [\"reactions\", \"genes\", \"metabolites\"]:\n",
    "    df_evidence = dataframes_evidence[attribute_type]\n",
    "    df_removed = dataframes_removed[attribute_type]\n",
    "    if overwrite:\n",
    "        df_evidence.to_csv(\n",
    "            get_dirpath(\"curated\") / f\"{attribute_type}_evidence.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "        )\n",
    "        df_removed.to_csv(\n",
    "            get_dirpath(\"curated\") / f\"{attribute_type}_removed.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "886da88c-00c0-4638-974f-2766333f4d7b",
   "metadata": {},
   "source": [
    "### Ensure all metabolites, genes, and reactions exist\n",
    "If removed (e.g., a duplicate), will show up in missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c56939d1-4a0c-4f68-b200-b84d7da345d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "0\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "missing_metabolites = set(model.metabolites.list_attr(\"id\")).symmetric_difference(\n",
    "    dataframes_evidence[\"metabolites\"][\"metabolites\"].values\n",
    ")\n",
    "\n",
    "missing_genes = set(model.genes.list_attr(\"id\")).symmetric_difference(\n",
    "    dataframes_evidence[\"genes\"][\"genes\"].values\n",
    ")\n",
    "\n",
    "missing_reactions = set(\n",
    "    model.reactions.query(lambda x: not x.subsystem == \"Pseudoreactions\").list_attr(\n",
    "        \"id\"\n",
    "    )\n",
    ").symmetric_difference(dataframes_evidence[\"reactions\"][\"reactions\"].values)\n",
    "\n",
    "print(len(missing_genes))\n",
    "print(len(missing_metabolites))\n",
    "print(len(missing_reactions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "2f237f06-e5c8-49d2-ae2f-c2f9a67c8fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in missing_metabolites:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "382efe39-c1a3-43b8-a129-853ad91020b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in missing_genes:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2dd0a570-1026-4b40-8e95-1b270900a61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in missing_reactions:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc83e690-0829-4236-9161-165a400539cb",
   "metadata": {},
   "source": [
    "### Check for extra metabolites, genes, and reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "834b542d-2669-43ab-b212-4d875f54558c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in model.metabolites.query(lambda x: not len(x.reactions)):\n",
    "    print(f\"{met.id}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4bc9de67-ade2-46d6-b873-7f392e1c2e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in model.genes.query(lambda x: not len(x.reactions)):\n",
    "    print(gene.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f764f34d-4465-44df-b973-9ea11b0fb93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for reaction in model.reactions.query(lambda x: not len(x.metabolites)):\n",
    "    print(reaction.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a0c06ba-7a50-4dfc-962c-fa8044219372",
   "metadata": {},
   "source": [
    "### Cleanup model values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "285cb1ab-9039-447a-a567-3bcf51bdc0fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for mid, row in dataframes_evidence[\"metabolites\"].set_index(\"metabolites\").iterrows():\n",
    "    if gid in missing_metabolites:\n",
    "        continue\n",
    "    metabolite = model.metabolites.get_by_id(mid)\n",
    "    metabolite.name = row[\"name\"]\n",
    "    metabolite.formula = row[\"formula\"]\n",
    "    metabolite.charge = int(row[\"charge\"])\n",
    "    metabolite.compartment = row[\"compartment\"]\n",
    "    metabolite.annotation[\"metabolite\"] = metabolite.id.replace(\n",
    "        f\"_{metabolite.compartment}\", \"\"\n",
    "    )\n",
    "\n",
    "for rid, row in dataframes_evidence[\"reactions\"].set_index(\"reactions\").iterrows():\n",
    "    if rid in missing_reactions:\n",
    "        continue\n",
    "    reaction = model.reactions.get_by_id(rid)\n",
    "    reaction.name = row[\"name\"]\n",
    "\n",
    "for gid, row in dataframes_evidence[\"genes\"].set_index(\"genes\").iterrows():\n",
    "    if gid in missing_genes:\n",
    "        continue\n",
    "    gene = model.genes.get_by_id(gid)\n",
    "    gene.name = row[\"hgnc.symbol\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac24451-ec72-4565-8cb8-8fc57c6ee2d9",
   "metadata": {},
   "source": [
    "#### Add boundary reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b424c4c7-dd7d-4473-b9e4-f0b69b9278b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "boundaries = {\n",
    "    # All exchange boundary reactions added\n",
    "    \"exchange\": model.metabolites.query(lambda x: x.compartment == \"e\").list_attr(\"id\"),\n",
    "    # Intracellular demands, only used for accumulation is allowed for a compound\n",
    "    \"demand\": [],\n",
    "    # Intracellular sinks, only used for when a source is needed for a compound\n",
    "    \"sink\": [\n",
    "        # Globin/Hemoglobin\n",
    "        \"oxyhb_c\",\n",
    "        \"hb4_23dpg_c\",\n",
    "        \"hb_hco2_c\",\n",
    "        \"globin_c\",\n",
    "        \"hbsno_c\",\n",
    "        \"carboxyhb_c\",\n",
    "        \"cclglobin_c\",\n",
    "        \"hemedegprods_c\",\n",
    "        \"hba1c_c\",\n",
    "        # Amino acids\n",
    "        # AA protein Residues\n",
    "        \"protres_arg__L_c\",\n",
    "        \"protres_asn__L_c\",\n",
    "        \"protres_asp__L_c\",\n",
    "        \"protres_cys__L_c\",\n",
    "        \"protres_gln__L_c\",\n",
    "        \"protres_glu__L_c\",\n",
    "        \"protres_his__L_c\",\n",
    "        \"protres_lys__L_c\",\n",
    "        \"protres_met__L_c\",\n",
    "        \"protres_ser__L_c\",\n",
    "        \"protres_thr__L_c\",\n",
    "        \"protres_tyr__L_c\",\n",
    "        \"protres_tyr__L_c\",\n",
    "        # Other AA residues\n",
    "        \"protres_asp__D_c\",\n",
    "        \"protres_isoasp__L_c\",\n",
    "        \"protres_isoasp__D_c\",\n",
    "        # Phosphorylated residues\n",
    "        \"protres_Nproshispi_c\",\n",
    "        \"protres_Ntelehispi_c\",\n",
    "        \"protres_serpi_c\",\n",
    "        \"protres_thrpi_c\",\n",
    "        \"protres_tyrpi_c\",\n",
    "        # Acetylated residues\n",
    "        \"protres_aclys__L_c\",\n",
    "        # Glycosylated residues\n",
    "        \"protres_ser3oacgam_c\",\n",
    "        \"protres_thr3oacgam_c\",\n",
    "        \"protres_serTAg_c\",\n",
    "        \"protres_thrTAg_c\",\n",
    "        # Glycated residues\n",
    "        \"protres_frulys_c\",\n",
    "        \"protres_rbllys_c\",\n",
    "        # Methylated residues\n",
    "        \"protres_admarg__L_c\",\n",
    "        \"protres_sdmarg__L_c\",\n",
    "        # Lipidated residues\n",
    "        \"protres_ttdcacys_c\",\n",
    "        \"protres_hxdcacys_c\",\n",
    "        \"protres_ocdcacys_c\",\n",
    "        # Oxidized residues\n",
    "        \"protres_metSox__SL_c\",\n",
    "        # Nitrosylated residues\n",
    "        \"protres_snocys__L_c\",\n",
    "        # ADP-ribosylated residues\n",
    "        \"protres_oadpribser_c\",\n",
    "        \"protres_sadpribcys_c\",\n",
    "        # Amine\n",
    "        \"protres_gludpam_c\",\n",
    "        \"protres_gluhista_c\",\n",
    "        \"protres_glunpphr_c\",\n",
    "        \"protres_glusrtn_c\",\n",
    "        \"protres_glu5meo__L_c\",\n",
    "        \"protres_lysglu_protres_c\",\n",
    "        # Ubiquitin\n",
    "        \"polyubb_c\",\n",
    "        \"ubiquitin_c\",\n",
    "        \"accprot_monoubiqlys_c\",\n",
    "        \"accprot_ubiqlys_c\",\n",
    "        \"accprot_lys__L_c\",\n",
    "        \"cullin_lys__L_c\",\n",
    "        \"cullin_nedd8lys_c\",\n",
    "        \"nedd8_c\",\n",
    "        # Small ions\n",
    "        \"na1_c\",\n",
    "        \"k_c\",\n",
    "        \"ca2_c\",\n",
    "        \"hno_c\",\n",
    "        \"co3r_c\",\n",
    "        # Vitamin E\n",
    "        \"avite1_c\",\n",
    "        \"avite1qn_c\",\n",
    "        # 'Redoxins'\n",
    "        \"prdx2crd_c\",\n",
    "        \"prdx2cso3_c\",\n",
    "        \"grdx2crd_c\",\n",
    "        \"grdx2cox_c\",\n",
    "        # CoA\n",
    "        # 'FAcoa_10_DC_c',\n",
    "        # 'FAcoa_12_DC_c',\n",
    "        # 'FAcoa_16_DC_c',\n",
    "        # 'FAcoa_4_DC_c',\n",
    "        # 'FAcoa_5_2EDC_c',\n",
    "        # 'FAcoa_5_DC_c',\n",
    "        # 'FAcoa_6_DC_c',\n",
    "        # 'FAcoa_7_DC_c',\n",
    "        # 'FAcoa_8_DC_c',\n",
    "        # 'FAcoa_5_3M3OH__S_c',\n",
    "        # 'FAcoa_hs_10_3OH__S_c',\n",
    "        # 'FAcoa_hs_12_3OH__S_c',\n",
    "        # 'FAcoa_hs_14_3OH__S_c',\n",
    "        # 'FAcoa_hs_14_5E8Z3OH__S_c',\n",
    "        # 'FAcoa_hs_14_7Z3OH__S_c',\n",
    "        # 'FAcoa_hs_16_3OH__R_c',\n",
    "        # 'FAcoa_hs_16_3OH__S_c',\n",
    "        # 'FAcoa_hs_16_7E10Z3OH__S_c',\n",
    "        # 'FAcoa_hs_16_9Z3OH__S_c',\n",
    "        # 'FAcoa_hs_17_3OH__R_c',\n",
    "        # 'FAcoa_hs_18_3OH__R_c',\n",
    "        # 'FAcoa_hs_18_3OH__S_c',\n",
    "        # 'FAcoa_hs_18_9Z12Z3OH__S_c',\n",
    "        # 'FAcoa_hs_18_9Z3OH__S_c',\n",
    "        # 'FAcoa_hs_19_3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_13Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_9Z3OH__R_c',\n",
    "        # 'FAcoa_hs_21_3OH__R_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_3OH__R_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_23_3OH__R_c',\n",
    "        # 'FAcoa_hs_24_12Z15Z18Z21Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_15Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_3OH__R_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z21Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z3OH__R_c',\n",
    "        # 'FAcoa_hs_26_17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_26_3OH__R_c',\n",
    "        # 'FAcoa_hs_3_3OH__S_c',\n",
    "        # 'FAcoa_hs_4_3OH__R_c',\n",
    "        # 'FAcoa_hs_6_3OH__S_c',\n",
    "        # 'FAcoa_hs_7_3OH__S_c',\n",
    "        # 'FAcoa_hs_8_3OH__S_c',\n",
    "        # 'FAcoa_hs_9_3OH__S_c',\n",
    "        # 'FAcoa_5_2E2M_c',\n",
    "        # 'FAcoa_5_2M_c',\n",
    "        # 'FAcoa_hs_10_2E6Z_c',\n",
    "        # 'FAcoa_hs_10_2E_c',\n",
    "        # 'FAcoa_hs_12_2E_c',\n",
    "        # 'FAcoa_hs_14_2E_c',\n",
    "        # 'FAcoa_hs_16_2E_c',\n",
    "        # 'FAcoa_hs_17_2E_c',\n",
    "        # 'FAcoa_hs_18_2E_c',\n",
    "        # 'FAcoa_hs_19_2E_c',\n",
    "        # 'FAcoa_hs_20_2E11Z14Z17Z_c',\n",
    "        # 'FAcoa_hs_20_2E11Z14Z_c',\n",
    "        # 'FAcoa_hs_20_2E11Z_c',\n",
    "        # 'FAcoa_hs_20_2E13Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z14Z17Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z14Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z_c',\n",
    "        # 'FAcoa_hs_20_2E9Z_c',\n",
    "        # 'FAcoa_hs_20_2E_c',\n",
    "        # 'FAcoa_hs_21_2E_c',\n",
    "        # 'FAcoa_hs_22_2E10Z13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E10Z13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E11Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z_c',\n",
    "        # 'FAcoa_hs_22_2E7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E7Z10Z13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E_c',\n",
    "        # 'FAcoa_hs_23_2E_c',\n",
    "        # 'FAcoa_hs_24_2E12Z15Z18Z21Z_c',\n",
    "        # 'FAcoa_hs_24_2E15Z_c',\n",
    "        # 'FAcoa_hs_24_2E9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcoa_hs_24_2E9Z12Z15Z18Z_c',\n",
    "        # 'FAcoa_hs_24_2E_c',\n",
    "        # 'FAcoa_hs_26_2E17Z_c',\n",
    "        # 'FAcoa_hs_26_2E_c',\n",
    "        # 'FAcoa_hs_3_2E_c',\n",
    "        # 'FAcoa_hs_4_2E_c',\n",
    "        # 'FAcoa_hs_6_2E_c',\n",
    "        # 'FAcoa_hs_8_2E_c',\n",
    "        # 'FAcoa_hs_14_5E8Z_c',\n",
    "        # 'FAcoa_4_2M_c',\n",
    "        # 'FAcoa_hs_16_3O_c',\n",
    "        # 'FAcoa_hs_17_3O_c',\n",
    "        # 'FAcoa_hs_18_3O_c',\n",
    "        # 'FAcoa_hs_19_3O_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z17Z3O_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z3O_c',\n",
    "        # 'FAcoa_hs_20_11Z3O_c',\n",
    "        # 'FAcoa_hs_20_13Z3O_c',\n",
    "        # 'FAcoa_hs_20_3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z17Z3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z3O_c',\n",
    "        # 'FAcoa_hs_20_9Z3O_c',\n",
    "        # 'FAcoa_hs_21_3O_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_22_11Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z3O_c',\n",
    "        # 'FAcoa_hs_22_3O_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_23_3O_c',\n",
    "        # 'FAcoa_hs_24_12Z15Z18Z21Z3O_c',\n",
    "        # 'FAcoa_hs_24_15Z3O_c',\n",
    "        # 'FAcoa_hs_24_3O_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z21Z3O_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z3O_c',\n",
    "        # 'FAcoa_hs_26_17Z3O_c',\n",
    "        # 'FAcoa_hs_26_3O_c',\n",
    "        # 'dmnoncoa_c',\n",
    "        # 'dmhptcoa_c',\n",
    "        # # Carnitine\n",
    "        # 'FAcrn_10_DC_c',\n",
    "        # 'FAcrn_12_DC_c',\n",
    "        # 'FAcrn_16_DC_c',\n",
    "        # 'FAcrn_4_2M_c',\n",
    "        # 'FAcrn_4_DC_c',\n",
    "        # 'FAcrn_5_2E2M_c',\n",
    "        # 'FAcrn_5_2EDC_c',\n",
    "        # 'FAcrn_5_2M_c',\n",
    "        # 'FAcrn_5_3M3OH__S_c',\n",
    "        # 'FAcrn_5_DC_c',\n",
    "        # 'FAcrn_6_DC_c',\n",
    "        # 'FAcrn_7_DC_c',\n",
    "        # 'FAcrn_8_DC_c',\n",
    "        # 'FAcrn_hs_10_0_c',\n",
    "        # 'FAcrn_hs_10_2E6Z_c',\n",
    "        # 'FAcrn_hs_10_2E_c',\n",
    "        # 'FAcrn_hs_10_3OH__S_c',\n",
    "        # 'FAcrn_hs_11_0_c',\n",
    "        # 'FAcrn_hs_12_0_c',\n",
    "        # 'FAcrn_hs_12_2E_c',\n",
    "        # 'FAcrn_hs_12_3OH__S_c',\n",
    "        # 'FAcrn_hs_13_0_c',\n",
    "        # 'FAcrn_hs_14_0_c',\n",
    "        # 'FAcrn_hs_14_2E_c',\n",
    "        # 'FAcrn_hs_14_3OH__S_c',\n",
    "        # 'FAcrn_hs_14_5E8Z3OH__S_c',\n",
    "        # 'FAcrn_hs_14_5E8Z_c',\n",
    "        # 'FAcrn_hs_14_5Z_c',\n",
    "        # 'FAcrn_hs_14_7Z3OH__S_c',\n",
    "        # 'FAcrn_hs_14_7Z_c',\n",
    "        # 'FAcrn_hs_14_9Z_c',\n",
    "        # 'FAcrn_hs_15_0_c',\n",
    "        # 'FAcrn_hs_16_0_c',\n",
    "        # 'FAcrn_hs_16_2E_c',\n",
    "        # 'FAcrn_hs_16_3OH__S_c',\n",
    "        # 'FAcrn_hs_16_7E10Z3OH__S_c',\n",
    "        # 'FAcrn_hs_16_7Z_c',\n",
    "        # 'FAcrn_hs_16_9Z3OH__S_c',\n",
    "        # 'FAcrn_hs_16_9Z_c',\n",
    "        # 'FAcrn_hs_17_0_c',\n",
    "        # 'FAcrn_hs_17_10Z_c',\n",
    "        # 'FAcrn_hs_17_9Z_c',\n",
    "        # 'FAcrn_hs_18_0_c',\n",
    "        # 'FAcrn_hs_18_11Z_c',\n",
    "        # 'FAcrn_hs_18_13Z_c',\n",
    "        # 'FAcrn_hs_18_2E_c',\n",
    "        # 'FAcrn_hs_18_3OH__S_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z12Z15Z_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z12Z_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z_c',\n",
    "        # 'FAcrn_hs_18_7Z_c',\n",
    "        # 'FAcrn_hs_18_9E_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z15Z_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z3OH__S_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z_c',\n",
    "        # 'FAcrn_hs_18_9Z3OH__S_c',\n",
    "        # 'FAcrn_hs_18_9Z_c',\n",
    "        # 'FAcrn_hs_19_0_c',\n",
    "        # 'FAcrn_hs_20_0_c',\n",
    "        # 'FAcrn_hs_20_11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_11Z_c',\n",
    "        # 'FAcrn_hs_20_13Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z_c',\n",
    "        # 'FAcrn_hs_20_9Z_c',\n",
    "        # 'FAcrn_hs_21_0_c',\n",
    "        # 'FAcrn_hs_22_0_c',\n",
    "        # 'FAcrn_hs_22_10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_11Z_c',\n",
    "        # 'FAcrn_hs_22_13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_13Z_c',\n",
    "        # 'FAcrn_hs_22_4Z7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_4Z7Z10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_7Z10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_23_0_c',\n",
    "        # 'FAcrn_hs_24_0_c',\n",
    "        # 'FAcrn_hs_24_12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_15Z_c',\n",
    "        # 'FAcrn_hs_24_6Z9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_6Z9Z12Z15Z18Z_c',\n",
    "        # 'FAcrn_hs_24_9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_9Z12Z15Z18Z_c',\n",
    "        # 'FAcrn_hs_26_0_c',\n",
    "        # 'FAcrn_hs_26_17Z_c',\n",
    "        # 'FAcrn_hs_3_0_c',\n",
    "        # 'FAcrn_hs_3_2E_c',\n",
    "        # 'FAcrn_hs_3_3OH__S_c',\n",
    "        # 'FAcrn_hs_4_0_c',\n",
    "        # 'FAcrn_hs_4_2E_c',\n",
    "        # 'FAcrn_hs_4_3OH__R_c',\n",
    "        # 'FAcrn_hs_5_0_c',\n",
    "        # 'FAcrn_hs_6_0_c',\n",
    "        # 'FAcrn_hs_6_2E_c',\n",
    "        # 'FAcrn_hs_6_3OH__S_c',\n",
    "        # 'FAcrn_hs_7_0_c',\n",
    "        # 'FAcrn_hs_7_3OH__S_c',\n",
    "        # 'FAcrn_hs_8_0_c',\n",
    "        # 'FAcrn_hs_8_2E_c',\n",
    "        # 'FAcrn_hs_8_3OH__S_c',\n",
    "        # 'FAcrn_hs_9_0_c',\n",
    "        # 'FAcrn_hs_9_3OH__S_c',\n",
    "        # 'acrn_c',\n",
    "        # 'dmhptcrn_c',\n",
    "        # 'dmnoncrn_c',\n",
    "        # 'malcrn_c',\n",
    "        # tRNA\n",
    "        # 'trnaala_c',\n",
    "        # 'trnaarg_c',\n",
    "        # 'trnaasn_c',\n",
    "        # 'trnaasp_c',\n",
    "        # 'trnacys_c',\n",
    "        # 'trnagln_c',\n",
    "        # 'trnaglu_c',\n",
    "        # 'trnagly_c',\n",
    "        # 'trnahis_c',\n",
    "        # 'trnaile_c',\n",
    "        # 'trnaleu_c',\n",
    "        # 'trnalys_c',\n",
    "        # 'trnamet_c',\n",
    "        # 'trnaphe_c',\n",
    "        # 'trnapro_c',\n",
    "        # 'trnaser_c',\n",
    "        # 'trnathr_c',\n",
    "        # 'trnatrp_c',\n",
    "        # 'trnatyr_c',\n",
    "        # 'trnaval_c',\n",
    "        # 'alatrna_c',\n",
    "        # 'argtrna_c',\n",
    "        # 'asntrna_c',\n",
    "        # 'asptrna_c',\n",
    "        # 'cystrna_c',\n",
    "        # 'glntrna_c',\n",
    "        # 'glutrna_c',\n",
    "        # 'glytrna_c',\n",
    "        # 'histrna_c',\n",
    "        # 'iletrna_c',\n",
    "        # 'leutrna_c',\n",
    "        # 'lystrna_c',\n",
    "        # 'mettrna_c',\n",
    "        # 'phetrna_c',\n",
    "        # 'protrna_c',\n",
    "        # 'sertrna_c',\n",
    "        # 'thrtrna_c',\n",
    "        # 'trptrna_c',\n",
    "        # 'tyrtrna_c',\n",
    "        # 'valtrna_c',\n",
    "        # Sugar\n",
    "        # '2ddglcn_c',\n",
    "        # '3dfru_c',\n",
    "        # Nucleotides\n",
    "        # \"23camp_c\",\n",
    "        # \"23ccmp_c\",\n",
    "        # \"23cgmp_c\",\n",
    "        # \"23cump_c\",\n",
    "        # \"dctp_c\",\n",
    "        # \"dgtp_c\",\n",
    "        # \"datp_c\",\n",
    "        # \"dttp_c\",\n",
    "        # 'dutp_c',\n",
    "        # \"psump_c\",\n",
    "        # \"psi_c\",\n",
    "        # Sourced from somewhere/\n",
    "        # Drains to somewhere/accumulates\n",
    "        # 'so2gth_c',\n",
    "        # \"dh15kprostge1_c\",\n",
    "        # \"dh15kprostge2_c\",\n",
    "        # \"dh15kprostge3_c\",\n",
    "        # \"dh15kprostgf1_c\",\n",
    "        # \"dh15kprostgf2_c\",\n",
    "        # \"dh15kprostgf3_c\",\n",
    "        # 'polyadprib2_c',\n",
    "        # 'polyadprib1_c',\n",
    "    ],\n",
    "}\n",
    "default_closed = []\n",
    "for btype, met_list in boundaries.items():\n",
    "    for met in met_list:\n",
    "        met = model.metabolites.get_by_id(met)\n",
    "        try:\n",
    "            reaction = model.add_boundary(met, type=btype)\n",
    "        except ValueError:\n",
    "            rid = {\n",
    "                \"exchange\": f\"EX_{met}\",\n",
    "                \"demand\": f\"DM_{met}\",\n",
    "                \"sink\": f\"SK_{met}\",\n",
    "            }[btype]\n",
    "            reaction = model.reactions.get_by_id(rid)\n",
    "            reaction.name = f\"{met.name} {btype}\"\n",
    "\n",
    "        if met in default_closed:\n",
    "            reaction.lower_bound = 0\n",
    "\n",
    "for reaction in model.boundary:\n",
    "    reaction.subsystem = \"Pseudoreactions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff186a49-66a1-444c-92dc-0f1eca0464d9",
   "metadata": {},
   "source": [
    "#### Reset subsystem groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5037b248-ee81-442e-aa81-dd8acd25380f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14a880c50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14a880c50>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reset_subsystem_groups(model)\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ba5b6-68bb-4160-bcfa-58fe424d8025",
   "metadata": {},
   "source": [
    "### Check mass balancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d47506ac-f31a-4730-93dc-95fef5cf1986",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6LTHPI: 6lthp_c --> h_c + sppt_c\n",
      "{'charge': 2.0}\n",
      "\n",
      "METHBCYTBR: 2.0 focytb5_c + methb_c --> 2.0 ficytb5_c + hb_c\n",
      "{'charge': 1.0}\n",
      "\n",
      "METHBFMNR: fmnh2_c + methb_c --> fmn_c + 3.0 h_c + hb_c\n",
      "{'charge': 1.0}\n",
      "\n",
      "POOL_FACOA: FAcoa_hs_c <=> 0.0004 FAcoa_hs_12_0_c + 0.0004 FAcoa_hs_13_0_c + 0.0133 FAcoa_hs_14_0_c + 0.0004 FAcoa_hs_14_5Z_c + 0.0004 FAcoa_hs_14_7Z_c + 0.0004 FAcoa_hs_14_9Z_c + 0.0004 FAcoa_hs_15_0_c + 0.222 FAcoa_hs_16_0_c + 0.0004 FAcoa_hs_16_7Z_c + 0.0219 FAcoa_hs_16_9Z_c + 0.0004 FAcoa_hs_17_0_c + 0.0004 FAcoa_hs_17_10Z_c + 0.0004 FAcoa_hs_17_9Z_c + 0.1498 FAcoa_hs_18_0_c + 0.025 FAcoa_hs_18_11Z_c + 0.0004 FAcoa_hs_18_13Z_c + 0.0025 FAcoa_hs_18_6Z9Z12Z15Z_c + 0.0029 FAcoa_hs_18_6Z9Z12Z_c + 0.0004 FAcoa_hs_18_6Z9Z_c + 0.0004 FAcoa_hs_18_7Z_c + 0.0004 FAcoa_hs_18_9E_c + 0.0084 FAcoa_hs_18_9Z12Z15Z_c + 0.1915 FAcoa_hs_18_9Z12Z_c + 0.1545 FAcoa_hs_18_9Z_c + 0.0004 FAcoa_hs_19_0_c + 0.0004 FAcoa_hs_20_0_c + 0.0215 FAcoa_hs_20_11Z14Z17Z_c + 0.0004 FAcoa_hs_20_11Z14Z_c + 0.0004 FAcoa_hs_20_11Z_c + 0.0004 FAcoa_hs_20_13Z_c + 0.0116 FAcoa_hs_20_5Z8Z11Z14Z17Z_c + 0.1083 FAcoa_hs_20_5Z8Z11Z14Z_c + 0.0004 FAcoa_hs_20_5Z8Z11Z_c + 0.0115 FAcoa_hs_20_8Z11Z14Z17Z_c + 0.0215 FAcoa_hs_20_8Z11Z14Z_c + 0.0004 FAcoa_hs_20_8Z11Z_c + 0.0004 FAcoa_hs_20_9Z_c + 0.0004 FAcoa_hs_21_0_c + 0.0004 FAcoa_hs_22_0_c + 0.0004 FAcoa_hs_22_10Z13Z16Z19Z_c + 0.0004 FAcoa_hs_22_10Z13Z16Z_c + 0.0004 FAcoa_hs_22_11Z_c + 0.0004 FAcoa_hs_22_13Z16Z19Z_c + 0.0004 FAcoa_hs_22_13Z16Z_c + 0.0004 FAcoa_hs_22_13Z_c + 0.0278 FAcoa_hs_22_4Z7Z10Z13Z16Z19Z_c + 0.0038 FAcoa_hs_22_4Z7Z10Z13Z16Z_c + 0.0059 FAcoa_hs_22_7Z10Z13Z16Z19Z_c + 0.0014 FAcoa_hs_22_7Z10Z13Z16Z_c + 0.0004 FAcoa_hs_23_0_c + 0.0004 FAcoa_hs_24_0_c + 0.0004 FAcoa_hs_24_12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_15Z_c + 0.0004 FAcoa_hs_24_6Z9Z12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_6Z9Z12Z15Z18Z_c + 0.0004 FAcoa_hs_24_9Z12Z15Z18Z21Z_c + 0.0004 FAcoa_hs_24_9Z12Z15Z18Z_c + 0.0004 FAcoa_hs_26_0_c + 0.0004 FAcoa_hs_26_17Z_c\n",
      "{'charge': -0.08439999999999515, 'C': 17.817500000000024, 'H': 33.30219999999997, 'N': 0.14769999999999328, 'O': 0.3586999999999705, 'P': 0.0632999999999968, 'S': 0.021099999999998786, 'R': -1.0}\n",
      "\n",
      "POOL_FA: FA_hs_c <=> 0.0004 FA_hs_12_0_c + 0.0004 FA_hs_13_0_c + 0.0133 FA_hs_14_0_c + 0.0004 FA_hs_14_5Z_c + 0.0004 FA_hs_14_7Z_c + 0.0004 FA_hs_14_9Z_c + 0.0004 FA_hs_15_0_c + 0.222 FA_hs_16_0_c + 0.0004 FA_hs_16_7Z_c + 0.0219 FA_hs_16_9Z_c + 0.0004 FA_hs_17_0_c + 0.0004 FA_hs_17_10Z_c + 0.0004 FA_hs_17_9Z_c + 0.1498 FA_hs_18_0_c + 0.025 FA_hs_18_11Z_c + 0.0004 FA_hs_18_13Z_c + 0.0025 FA_hs_18_6Z9Z12Z15Z_c + 0.0029 FA_hs_18_6Z9Z12Z_c + 0.0004 FA_hs_18_6Z9Z_c + 0.0004 FA_hs_18_7Z_c + 0.0004 FA_hs_18_9E_c + 0.0084 FA_hs_18_9Z12Z15Z_c + 0.1915 FA_hs_18_9Z12Z_c + 0.1545 FA_hs_18_9Z_c + 0.0004 FA_hs_19_0_c + 0.0004 FA_hs_20_0_c + 0.0215 FA_hs_20_11Z14Z17Z_c + 0.0004 FA_hs_20_11Z14Z_c + 0.0004 FA_hs_20_11Z_c + 0.0004 FA_hs_20_13Z_c + 0.0116 FA_hs_20_5Z8Z11Z14Z17Z_c + 0.1083 FA_hs_20_5Z8Z11Z14Z_c + 0.0004 FA_hs_20_5Z8Z11Z_c + 0.0115 FA_hs_20_8Z11Z14Z17Z_c + 0.0215 FA_hs_20_8Z11Z14Z_c + 0.0004 FA_hs_20_8Z11Z_c + 0.0004 FA_hs_20_9Z_c + 0.0004 FA_hs_21_0_c + 0.0004 FA_hs_22_0_c + 0.0004 FA_hs_22_10Z13Z16Z19Z_c + 0.0004 FA_hs_22_10Z13Z16Z_c + 0.0004 FA_hs_22_11Z_c + 0.0004 FA_hs_22_13Z16Z19Z_c + 0.0004 FA_hs_22_13Z16Z_c + 0.0004 FA_hs_22_13Z_c + 0.0278 FA_hs_22_4Z7Z10Z13Z16Z19Z_c + 0.0038 FA_hs_22_4Z7Z10Z13Z16Z_c + 0.0059 FA_hs_22_7Z10Z13Z16Z19Z_c + 0.0014 FA_hs_22_7Z10Z13Z16Z_c + 0.0004 FA_hs_23_0_c + 0.0004 FA_hs_24_0_c + 0.0004 FA_hs_24_12Z15Z18Z21Z_c + 0.0004 FA_hs_24_15Z_c + 0.0004 FA_hs_24_6Z9Z12Z15Z18Z21Z_c + 0.0004 FA_hs_24_6Z9Z12Z15Z18Z_c + 0.0004 FA_hs_24_9Z12Z15Z18Z21Z_c + 0.0004 FA_hs_24_9Z12Z15Z18Z_c + 0.0004 FA_hs_26_0_c + 0.0004 FA_hs_26_17Z_c\n",
      "{'charge': -0.021099999999998786, 'C': 17.374399999999994, 'H': 32.6481, 'O': 0.04219999999999757, 'R': -1.0}\n",
      "\n",
      "SNOHBNO: nitrosylhb_c --> h_c + hbsno_c\n",
      "{'charge': 1}\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for reaction in model.reactions:\n",
    "    if reaction.boundary:\n",
    "        continue\n",
    "    try:\n",
    "        if reaction.check_mass_balance():\n",
    "            print(reaction)\n",
    "            print(reaction.check_mass_balance())\n",
    "            print()\n",
    "    except:\n",
    "        print({m.id: m.charge for m in reaction.metabolites})\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18d61d2b-a4a0-4ac8-84c6-68bd7b421790",
   "metadata": {},
   "source": [
    "### Set bounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "dc710bb4-325e-4db2-bd30-c3595b60052e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reset_reaction_bounds(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "fa717ddf-cb70-4e15-839d-80e245c113dd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp9oocremi.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmplm7ug58m.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpwid_7oxg.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp7wueqtt3.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpkx_hx7e3.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp46zpu650.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpckxoc6yz.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpmkhz9t35.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpk415hbeh.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpb6j6cne0.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp1rqundb2.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmphr_lnksj.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp8kmyl9yb.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp2b9_3nog.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpdh4aog0m.lp\n",
      "Reading time = 0.05 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpxg7flbbu.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpjuj3jq44.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpla4g0mfb.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpvyjyw1gc.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpzxlxnp6k.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpb3ef0jyz.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp1otv_mkl.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp7ig7804d.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpf11x2n7j.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpra2kma17.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmp3efei25x.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpowwwhldi.lp\n",
      "Reading time = 0.03 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpezw82wd3.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmphgfolpl9.lp\n",
      "Reading time = 0.02 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n",
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "Read LP format model from file /var/folders/5t/hk8m3g6d1jn25x5rssjgsrmm0000gn/T/tmpmfka3ko2.lp\n",
      "Reading time = 0.04 seconds\n",
      ": 2158 rows, 6551 columns, 24877 nonzeros\n"
     ]
    }
   ],
   "source": [
    "from cobra.flux_analysis.variability import find_blocked_reactions\n",
    "\n",
    "blocked_reactions = find_blocked_reactions(model, open_exchanges=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "92ef012b-a0b7-4d18-b054-f262db4b7b55",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of blocked reactions: 695\n",
      "Number of metabolites with one associated reaction 487\n",
      "\n",
      "15KPGE1Ry: 15kprostge1_c + h_c + nadph_c --> dh15kprostge1_c + nadp_c\n",
      "15KPGE2Ry: 15kprostge2_c + h_c + nadph_c --> dh15kprostge2_c + nadp_c\n",
      "15KPGE3Ry: 15kprostge3_c + h_c + nadph_c --> dh15kprostge3_c + nadp_c\n",
      "15KPGF1Ry: 15kprostgf1_c + h_c + nadph_c --> dh15kprostgf1_c + nadp_c\n",
      "15KPGF2Ry: 15kprostgf2_c + h_c + nadph_c --> dh15kprostgf2_c + nadp_c\n",
      "15KPGF3Ry: 15kprostgf3_c + h_c + nadph_c --> dh15kprostgf3_c + nadp_c\n",
      "23CN3PA: 23camp_c + h2o_c --> 2amp_c + h_c\n",
      "23CN3PC: 23ccmp_c + h2o_c --> 2cmp_c + h_c\n",
      "23CN3PG: 23cgmp_c + h2o_c --> 2gmp_c + h_c\n",
      "23CN3PU: 23cump_c + h2o_c --> 2ump_c + h_c\n",
      "2DCYTt: 2dcytd_c <=> 2dcytd_e\n",
      "2H3MV: 3mop_c + h_c + nadh_c <=> 2h3mv_c + nad_c\n",
      "2HIV: 3mob_c + h_c + nadh_c <=> 2hiv_c + nad_c\n",
      "34HPLDH: 34hpp_c + h_c + nadh_c --> 34hpl_c + nad_c\n",
      "34HPPT: 34hpp_c --> 2h34hppr_c\n",
      "3DGDH: 3dglc_c + h2o_c + nad_c --> 2ddglcn_c + 2.0 h_c + nadh_c\n",
      "3HKYNAKGAT: akg_c + hLkynr_c --> 42a3hp24db_c + glu__L_c + h_c\n",
      "3MOBt2: 3mob_c + h_c <=> 3mob_e + h_e\n",
      "3OHAACPDH1: bhbACP_c --> buteACP_2E_c + h2o_c\n",
      "3OHAACPDH2: 3ohhxaACP_c --> h2o_c + hxeaACP_2E_c\n",
      "3OHAACPDH3: 3ohocaACP_c --> h2o_c + oceaACP_2E_c\n",
      "3OHAACPDH4: 3ohdcaACP_c --> dceaACP_2E_c + h2o_c\n",
      "3OHAACPDH5: 3ohddcaACP_c --> ddceaACP_2E_c + h2o_c\n",
      "3OHAACPDH6: 3ohttdcaACP_c --> h2o_c + ttdceaACP_2E_c\n",
      "3OHAACPDH7: 3ohhxdcaACP_c --> h2o_c + hxdceaACP_2E_c\n",
      "3OHAACPDH8: 3ohocdcaACP_c --> h2o_c + ocdceaACP_2E_c\n",
      "3OXAACPR1y: aacACP_c + h_c + nadph_c --> bhbACP_c + nadp_c\n",
      "3OXAACPR2y: 3oxhxaACP_c + h_c + nadph_c --> 3ohhxaACP_c + nadp_c\n",
      "3OXAACPR3y: 3oxocaACP_c + h_c + nadph_c --> 3ohocaACP_c + nadp_c\n",
      "3OXAACPR4y: 3oxdcaACP_c + h_c + nadph_c --> 3ohdcaACP_c + nadp_c\n",
      "3OXAACPR5y: 3oxddcaACP_c + h_c + nadph_c --> 3ohddcaACP_c + nadp_c\n",
      "3OXAACPR6y: 3oxttdcaACP_c + h_c + nadph_c --> 3ohttdcaACP_c + nadp_c\n",
      "3OXAACPR7y: 3oxhxdcaACP_c + h_c + nadph_c --> 3ohhxdcaACP_c + nadp_c\n",
      "3OXAACPR8y: 3oxocdcaACP_c + h_c + nadph_c --> 3ohocdcaACP_c + nadp_c\n",
      "3SALAOX: 2.0 3sala__L_c + o2_c --> 2.0 Lcyst_c\n",
      "4H2HXEAL2EDHx: 4h2hxeal_2E_c + h2o_c + nad_c --> 4h2hxea_2E_c + 2.0 h_c + nadh_c\n",
      "4H2NNEAL2EDHx: 4h2noneal_2E_c + h2o_c + nad_c --> 4h2nonea_2E_c + 2.0 h_c + nadh_c\n",
      "4HOXPACDORx: 4hoxpacd_c + h2o_c + nad_c --> 4hphac_c + 2.0 h_c + nadh_c\n",
      "4HOXPACDORy: 4hoxpacd_c + h2o_c + nadp_c --> 4hphac_c + 2.0 h_c + nadph_c\n",
      "4MOPt2: 4mop_c + h_c <=> 4mop_e + h_e\n",
      "4PCYSK: 4pcys__R_c + atp_c --> 4ppcys__R_c + adp_c + h_c\n",
      "4PERP: 4per_c + h2o_c --> erthn__D_c + pi_c\n",
      "4PYRNDK: 4pyrdp_c + atp_c --> 4pyrtp_c + adp_c\n",
      "4PYRNMAT: 4pyrmp_c + atp_c + h_c --> 4pyrnad_c + ppi_c\n",
      "4PYRNMK: 4pyrmp_c + atp_c --> 4pyrdp_c + adp_c\n",
      "4PYRNt: 4pyrn_c <=> 4pyrn_e\n",
      "5ADHTRy: 5adtststerone_c + h_c + nadph_c <=> 5aand3a17bdiol_c + nadp_c\n",
      "5FLUDPK: 5fludp_c + atp_c --> 5flutp_c + adp_c\n",
      "5FLUMPK: 5flump_c + atp_c --> 5fludp_c + adp_c\n",
      "5FLURAPT: 5flura_c + prpp_c --> 5flump_c + ppi_c\n",
      "5HOXINDACTOX: 5hoxindact_c + h2o_c + nad_c --> 5hoxindoa_c + 2.0 h_c + nadh_c\n",
      "5INS7P5PP: 5ppins5p_c + h2o_c --> h_c + ins6p_c + pi_c\n",
      "5PPINSP5P1K: 5ppins5p_c + atp_c --> 15ppins4p_c + adp_c\n",
      "A4GALTc: galgluside_hs_c + udpgal_c --> h_c + thcrm_hs_c + udp_c\n",
      "ABO1A: fucgalacglcgalgluside_hs_c + udpacgal_c --> acgalfucgalacglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO1B: fucgalacglcgalgluside_hs_c + udpgal_c --> galfucgalacglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO2A: fuc12gal14acglcgalgluside_hs_c + udpacgal_c --> acgalfuc12gal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO2B: fuc12gal14acglcgalgluside_hs_c + udpgal_c --> galfuc12gal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO3A: fucgalacgalfuc12gal14acglcgalgluside_hs_c + udpacgal_c --> acgalfucgalacgalfuc12gal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO3AB: fucgalacgalfucgalacglcgal14acglcgalgluside_hs_c + udpacgal_c --> acgalfucgalacgalfucgalacglcgal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO4A: globoH_hs_c + udpacgal_c --> globoA_hs_c + h_c + udp_c\n",
      "ABO4B: globoH_hs_c + udpgal_c --> globoB_hs_c + h_c + udp_c\n",
      "ABO5A: fucgalacglcgal14acglcgalgluside_hs_c + udpacgal_c --> acgalfucgalacglcgal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "ABO5B: fucgalacglcgal14acglcgalgluside_hs_c + udpgal_c --> galfucgalacglcgal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "AC4OACNAMDC: ac4oacnam_c + h2o_c --> ac_c + acnam_c + h_c\n",
      "AC9OACNAMDC: ac9oacnam_c + h2o_c --> ac_c + acnam_c + h_c\n",
      "ACACT1: aacoa_c + coa_c --> 2.0 accoa_c\n",
      "ACGAMNT2: galacglcgal14acglcgalgluside_hs_c + uacgam_c --> acglcgalacglcgal14acglcgalgluside_hs_c + h_c + udp_c\n",
      "ACHt: ach_c --> ach_e\n",
      "ACMAH: acmetala_c + h2o_c --> acmet__L_c + ala__L_c\n",
      "ACMDAC: acmet__L_c + h2o_c --> ac_c + met__L_c\n",
      "ACOATA: ACP_c + accoa_c --> acACP_c + coa_c\n",
      "ACYP2: actp_c + h2o_c --> ac_c + h_c + pi_c\n",
      "ADPGLCDP: adpglc_c + h2o_c --> amp_c + g1p_c + 2.0 h_c\n",
      "ADPMNDP: adpman_c + h2o_c --> amp_c + 2.0 h_c + man1p_c\n",
      "ADPRDACAC1: 1acadprib_c + h2o_c --> ac_c + adprib_c + h_c\n",
      "AG1tec: ag1_e --> ag1_c\n",
      "AIRC: air_c + co2_c <=> 5aizc_c + h_c\n",
      "ALATRS: ala__L_c + atp_c + trnaala_c --> alatrna_c + amp_c + ppi_c\n",
      "ALDD19x: h2o_c + nad_c + pacald_c --> 2.0 h_c + nadh_c + pac_c\n",
      "ALDD19y: h2o_c + nadp_c + pacald_c --> 2.0 h_c + nadph_c + pac_c\n",
      "ALDD20x: h2o_c + id3acald_c + nad_c --> 2.0 h_c + ind3ac_c + nadh_c\n",
      "ALLTN: alltn_c + h2o_c <=> alltt_c + h_c\n",
      "AMPAMD: ampamd_c + h2o_c --> amp_c + nh4_c\n",
      "ANDRSTRNRy: andrstrn_c + h_c + nadph_c <=> 5aand3a17bdiol_c + nadp_c\n",
      "ANGIO1THOP: angiotensin1_c + h2o_c --> angiotensin17_c + phehisleu_c\n",
      "ANGIO2ASPH: angiotensin2_c + h2o_c --> angiotensin3_c + asp__L_c + h_c\n",
      "ANGIO2DDP: angiotensin2_c + h2o_c --> angiotensin4_c + asparg_c\n",
      "ANGIO3ARGH: angiotensin3_c + h2o_c + h_c --> angiotensin4_c + arg__L_c\n",
      "ANGIO4DDP: angiotensin4_c + h2o_c --> angiotensin58_c + valtyr_c\n",
      "ANGIO58PEP: angiotensin58_c + h2o_c --> angiotensin57_c + phe__L_c\n",
      "AP5AH1: ap5a_c + h2o_c --> adp_c + atp_c + 2.0 h_c\n",
      "AP6AH1: ap6a_c + h2o_c --> 2.0 atp_c + 2.0 h_c\n",
      "AP6AH2: ap6a_c + h2o_c --> adp_c + 2.0 h_c + p4a_c\n",
      "APDPP: alapro_c + h2o_c --> ala__L_c + pro__L_c\n",
      "ARGSL: argsuc_c <=> arg__L_c + fum_c\n",
      "ARGTRS: arg__L_c + atp_c + trnaarg_c --> amp_c + argtrna_c + ppi_c\n",
      "ASNTRS: asn__L_c + atp_c + trnaasn_c --> amp_c + asntrna_c + ppi_c\n",
      "ASPCT: asp__L_c + cbp_c <=> cbasp_c + h_c + pi_c\n",
      "ASPTRS: asp__L_c + atp_c + trnaasp_c --> amp_c + asptrna_c + ppi_c\n",
      "A_MANASE: 2.0 h2o_c + m2mn_c --> 2.0 man_c + mn_c\n",
      "B3GALNT1c: thcrm_hs_c + udpacgal_c --> gbside_hs_c + h_c + udp_c\n",
      "BILIREDx: biliverd_c + h_c + nadh_c --> bilirub_c + nad_c\n",
      "BILIREDy: biliverd_c + h_c + nadph_c --> bilirub_c + nadp_c\n",
      "BILRUBABCte: atp_c + bilirub_c + h2o_c --> adp_c + bilirub_e + h_c + pi_c\n",
      "BKAACPSYN1: acACP_c + h_c + malACP_c --> ACP_c + aacACP_c + co2_c\n",
      "BKAACPSYN2: butACP_c + h_c + malACP_c --> 3oxhxaACP_c + ACP_c + co2_c\n",
      "BKAACPSYN3: h_c + hxaACP_c + malACP_c --> 3oxocaACP_c + ACP_c + co2_c\n",
      "BKAACPSYN4: h_c + malACP_c + ocaACP_c --> 3oxdcaACP_c + ACP_c + co2_c\n",
      "BKAACPSYN5: dcaACP_c + h_c + malACP_c --> 3oxddcaACP_c + ACP_c + co2_c\n",
      "BKAACPSYN6: ddcaACP_c + h_c + malACP_c --> 3oxttdcaACP_c + ACP_c + co2_c\n",
      "BKAACPSYN7: h_c + malACP_c + ttdcaACP_c --> 3oxhxdcaACP_c + ACP_c + co2_c\n",
      "BKAACPSYN8: h_c + hxdcaACP_c + malACP_c --> 3oxocdcaACP_c + ACP_c + co2_c\n",
      "BPNT2: h2o_c + paps_c --> aps_c + pi_c\n",
      "BZALDHx: bzal_c + h2o_c + nad_c --> bz_c + 2.0 h_c + nadh_c\n",
      "CADPRBPH: cadprbp_c + h2o_c --> adprbp_c\n",
      "CHLSTD78I: lthstrl_c --> zymstnl_c\n",
      "CHLSTI1: zymst_c --> chlstol_c\n",
      "CMPACNAGA1: cmpacna_c + ga1_hs_c --> cmp_c + gm1b_hs_c + h_c\n",
      "CMPACNAGALBSD: cmpacna_c + galgbside_hs_c --> acnagalgbside_hs_c + cmp_c + h_c\n",
      "CMPACNAGD1B: cmpacna_c + gd1b_hs_c --> cmp_c + gt1b_hs_c + h_c\n",
      "CMPACNAGM1: cmpacna_c + gm1_hs_c --> cmp_c + gd1a_hs_c + h_c\n",
      "CRNAT_10_0: FAcoa_hs_10_0_c + crn_c <=> FAcrn_hs_10_0_c + coa_c\n",
      "CRNAT_10_2E: FAcoa_hs_10_2E_c + crn_c <=> FAcrn_hs_10_2E_c + coa_c\n",
      "CRNAT_10_2E6Z: FAcoa_hs_10_2E6Z_c + crn_c <=> FAcrn_hs_10_2E6Z_c + coa_c\n",
      "CRNAT_10_3OH_S: FAcoa_hs_10_3OH__S_c + crn_c <=> FAcrn_hs_10_3OH__S_c + coa_c\n",
      "CRNAT_10_DC: FAcoa_10_DC_c + crn_c <=> FAcrn_10_DC_c + coa_c\n",
      "CRNAT_11_0: FAcoa_hs_11_0_c + crn_c <=> FAcrn_hs_11_0_c + coa_c\n",
      "CRNAT_11_4M8M_c: crn_c + dmnoncoa_c <=> coa_c + dmnoncrn_c\n",
      "CRNAT_12_0: FAcoa_hs_12_0_c + crn_c <=> FAcrn_hs_12_0_c + coa_c\n",
      "CRNAT_12_2E: FAcoa_hs_12_2E_c + crn_c <=> FAcrn_hs_12_2E_c + coa_c\n",
      "CRNAT_12_3OH_S: FAcoa_hs_12_3OH__S_c + crn_c <=> FAcrn_hs_12_3OH__S_c + coa_c\n",
      "CRNAT_12_DC: FAcoa_12_DC_c + crn_c <=> FAcrn_12_DC_c + coa_c\n",
      "CRNAT_13_0: FAcoa_hs_13_0_c + crn_c <=> FAcrn_hs_13_0_c + coa_c\n",
      "CRNAT_14_0: FAcoa_hs_14_0_c + crn_c <=> FAcrn_hs_14_0_c + coa_c\n",
      "CRNAT_14_2E: FAcoa_hs_14_2E_c + crn_c <=> FAcrn_hs_14_2E_c + coa_c\n",
      "CRNAT_14_3OH_S: FAcoa_hs_14_3OH__S_c + crn_c <=> FAcrn_hs_14_3OH__S_c + coa_c\n",
      "CRNAT_14_5E8Z: FAcoa_hs_14_5E8Z_c + crn_c <=> FAcrn_hs_14_5E8Z_c + coa_c\n",
      "CRNAT_14_5E8Z3OH_S: FAcoa_hs_14_5E8Z3OH__S_c + crn_c <=> FAcrn_hs_14_5E8Z3OH__S_c + coa_c\n",
      "CRNAT_14_5Z: FAcoa_hs_14_5Z_c + crn_c <=> FAcrn_hs_14_5Z_c + coa_c\n",
      "CRNAT_14_7Z: FAcoa_hs_14_7Z_c + crn_c <=> FAcrn_hs_14_7Z_c + coa_c\n",
      "CRNAT_14_7Z3OH_S: FAcoa_hs_14_7Z3OH__S_c + crn_c <=> FAcrn_hs_14_7Z3OH__S_c + coa_c\n",
      "CRNAT_14_9Z: FAcoa_hs_14_9Z_c + crn_c <=> FAcrn_hs_14_9Z_c + coa_c\n",
      "CRNAT_15_0: FAcoa_hs_15_0_c + crn_c <=> FAcrn_hs_15_0_c + coa_c\n",
      "CRNAT_16_0: FAcoa_hs_16_0_c + crn_c <=> FAcrn_hs_16_0_c + coa_c\n",
      "CRNAT_16_2E: FAcoa_hs_16_2E_c + crn_c <=> FAcrn_hs_16_2E_c + coa_c\n",
      "CRNAT_16_3OH_S: FAcoa_hs_16_3OH__S_c + crn_c <=> FAcrn_hs_16_3OH__S_c + coa_c\n",
      "CRNAT_16_7E10Z3OH_S: FAcoa_hs_16_7E10Z3OH__S_c + crn_c <=> FAcrn_hs_16_7E10Z3OH__S_c + coa_c\n",
      "CRNAT_16_7Z: FAcoa_hs_16_7Z_c + crn_c <=> FAcrn_hs_16_7Z_c + coa_c\n",
      "CRNAT_16_9Z: FAcoa_hs_16_9Z_c + crn_c <=> FAcrn_hs_16_9Z_c + coa_c\n",
      "CRNAT_16_9Z3OH_S: FAcoa_hs_16_9Z3OH__S_c + crn_c <=> FAcrn_hs_16_9Z3OH__S_c + coa_c\n",
      "CRNAT_16_DC: FAcoa_16_DC_c + crn_c <=> FAcrn_16_DC_c + coa_c\n",
      "CRNAT_17_0: FAcoa_hs_17_0_c + crn_c <=> FAcrn_hs_17_0_c + coa_c\n",
      "CRNAT_17_10Z: FAcoa_hs_17_10Z_c + crn_c <=> FAcrn_hs_17_10Z_c + coa_c\n",
      "CRNAT_17_9Z: FAcoa_hs_17_9Z_c + crn_c <=> FAcrn_hs_17_9Z_c + coa_c\n",
      "CRNAT_18_0: FAcoa_hs_18_0_c + crn_c <=> FAcrn_hs_18_0_c + coa_c\n",
      "CRNAT_18_11Z: FAcoa_hs_18_11Z_c + crn_c <=> FAcrn_hs_18_11Z_c + coa_c\n",
      "CRNAT_18_13Z: FAcoa_hs_18_13Z_c + crn_c <=> FAcrn_hs_18_13Z_c + coa_c\n",
      "CRNAT_18_2E: FAcoa_hs_18_2E_c + crn_c <=> FAcrn_hs_18_2E_c + coa_c\n",
      "CRNAT_18_3OH_S: FAcoa_hs_18_3OH__S_c + crn_c <=> FAcrn_hs_18_3OH__S_c + coa_c\n",
      "CRNAT_18_6Z9Z: FAcoa_hs_18_6Z9Z_c + crn_c <=> FAcrn_hs_18_6Z9Z_c + coa_c\n",
      "CRNAT_18_6Z9Z12Z: FAcoa_hs_18_6Z9Z12Z_c + crn_c <=> FAcrn_hs_18_6Z9Z12Z_c + coa_c\n",
      "CRNAT_18_6Z9Z12Z15Z: FAcoa_hs_18_6Z9Z12Z15Z_c + crn_c <=> FAcrn_hs_18_6Z9Z12Z15Z_c + coa_c\n",
      "CRNAT_18_7Z: FAcoa_hs_18_7Z_c + crn_c <=> FAcrn_hs_18_7Z_c + coa_c\n",
      "CRNAT_18_9E: FAcoa_hs_18_9E_c + crn_c <=> FAcrn_hs_18_9E_c + coa_c\n",
      "CRNAT_18_9Z: FAcoa_hs_18_9Z_c + crn_c <=> FAcrn_hs_18_9Z_c + coa_c\n",
      "CRNAT_18_9Z12Z: FAcoa_hs_18_9Z12Z_c + crn_c <=> FAcrn_hs_18_9Z12Z_c + coa_c\n",
      "CRNAT_18_9Z12Z15Z: FAcoa_hs_18_9Z12Z15Z_c + crn_c <=> FAcrn_hs_18_9Z12Z15Z_c + coa_c\n",
      "CRNAT_18_9Z12Z3OH_S: FAcoa_hs_18_9Z12Z3OH__S_c + crn_c <=> FAcrn_hs_18_9Z12Z3OH__S_c + coa_c\n",
      "CRNAT_18_9Z3OH_S: FAcoa_hs_18_9Z3OH__S_c + crn_c <=> FAcrn_hs_18_9Z3OH__S_c + coa_c\n",
      "CRNAT_19_0: FAcoa_hs_19_0_c + crn_c <=> FAcrn_hs_19_0_c + coa_c\n",
      "CRNAT_20_0: FAcoa_hs_20_0_c + crn_c <=> FAcrn_hs_20_0_c + coa_c\n",
      "CRNAT_20_11Z: FAcoa_hs_20_11Z_c + crn_c <=> FAcrn_hs_20_11Z_c + coa_c\n",
      "CRNAT_20_11Z14Z: FAcoa_hs_20_11Z14Z_c + crn_c <=> FAcrn_hs_20_11Z14Z_c + coa_c\n",
      "CRNAT_20_11Z14Z17Z: FAcoa_hs_20_11Z14Z17Z_c + crn_c <=> FAcrn_hs_20_11Z14Z17Z_c + coa_c\n",
      "CRNAT_20_13Z: FAcoa_hs_20_13Z_c + crn_c <=> FAcrn_hs_20_13Z_c + coa_c\n",
      "CRNAT_20_5Z8Z11Z: FAcoa_hs_20_5Z8Z11Z_c + crn_c <=> FAcrn_hs_20_5Z8Z11Z_c + coa_c\n",
      "CRNAT_20_5Z8Z11Z14Z: FAcoa_hs_20_5Z8Z11Z14Z_c + crn_c <=> FAcrn_hs_20_5Z8Z11Z14Z_c + coa_c\n",
      "CRNAT_20_5Z8Z11Z14Z17Z: FAcoa_hs_20_5Z8Z11Z14Z17Z_c + crn_c <=> FAcrn_hs_20_5Z8Z11Z14Z17Z_c + coa_c\n",
      "CRNAT_20_8Z11Z: FAcoa_hs_20_8Z11Z_c + crn_c <=> FAcrn_hs_20_8Z11Z_c + coa_c\n",
      "CRNAT_20_8Z11Z14Z: FAcoa_hs_20_8Z11Z14Z_c + crn_c <=> FAcrn_hs_20_8Z11Z14Z_c + coa_c\n",
      "CRNAT_20_8Z11Z14Z17Z: FAcoa_hs_20_8Z11Z14Z17Z_c + crn_c <=> FAcrn_hs_20_8Z11Z14Z17Z_c + coa_c\n",
      "CRNAT_20_9Z: FAcoa_hs_20_9Z_c + crn_c <=> FAcrn_hs_20_9Z_c + coa_c\n",
      "CRNAT_21_0: FAcoa_hs_21_0_c + crn_c <=> FAcrn_hs_21_0_c + coa_c\n",
      "CRNAT_22_0: FAcoa_hs_22_0_c + crn_c <=> FAcrn_hs_22_0_c + coa_c\n",
      "CRNAT_22_10Z13Z16Z: FAcoa_hs_22_10Z13Z16Z_c + crn_c <=> FAcrn_hs_22_10Z13Z16Z_c + coa_c\n",
      "CRNAT_22_10Z13Z16Z19Z: FAcoa_hs_22_10Z13Z16Z19Z_c + crn_c <=> FAcrn_hs_22_10Z13Z16Z19Z_c + coa_c\n",
      "CRNAT_22_11Z: FAcoa_hs_22_11Z_c + crn_c <=> FAcrn_hs_22_11Z_c + coa_c\n",
      "CRNAT_22_13Z: FAcoa_hs_22_13Z_c + crn_c <=> FAcrn_hs_22_13Z_c + coa_c\n",
      "CRNAT_22_13Z16Z: FAcoa_hs_22_13Z16Z_c + crn_c <=> FAcrn_hs_22_13Z16Z_c + coa_c\n",
      "CRNAT_22_13Z16Z19Z: FAcoa_hs_22_13Z16Z19Z_c + crn_c <=> FAcrn_hs_22_13Z16Z19Z_c + coa_c\n",
      "CRNAT_22_4Z7Z10Z13Z16Z: FAcoa_hs_22_4Z7Z10Z13Z16Z_c + crn_c <=> FAcrn_hs_22_4Z7Z10Z13Z16Z_c + coa_c\n",
      "CRNAT_22_4Z7Z10Z13Z16Z19Z: FAcoa_hs_22_4Z7Z10Z13Z16Z19Z_c + crn_c <=> FAcrn_hs_22_4Z7Z10Z13Z16Z19Z_c + coa_c\n",
      "CRNAT_22_7Z10Z13Z16Z: FAcoa_hs_22_7Z10Z13Z16Z_c + crn_c <=> FAcrn_hs_22_7Z10Z13Z16Z_c + coa_c\n",
      "CRNAT_22_7Z10Z13Z16Z19Z: FAcoa_hs_22_7Z10Z13Z16Z19Z_c + crn_c <=> FAcrn_hs_22_7Z10Z13Z16Z19Z_c + coa_c\n",
      "CRNAT_23_0: FAcoa_hs_23_0_c + crn_c <=> FAcrn_hs_23_0_c + coa_c\n",
      "CRNAT_24_0: FAcoa_hs_24_0_c + crn_c <=> FAcrn_hs_24_0_c + coa_c\n",
      "CRNAT_24_12Z15Z18Z21Z: FAcoa_hs_24_12Z15Z18Z21Z_c + crn_c <=> FAcrn_hs_24_12Z15Z18Z21Z_c + coa_c\n",
      "CRNAT_24_15Z: FAcoa_hs_24_15Z_c + crn_c <=> FAcrn_hs_24_15Z_c + coa_c\n",
      "CRNAT_24_6Z9Z12Z15Z18Z: FAcoa_hs_24_6Z9Z12Z15Z18Z_c + crn_c <=> FAcrn_hs_24_6Z9Z12Z15Z18Z_c + coa_c\n",
      "CRNAT_24_6Z9Z12Z15Z18Z21Z: FAcoa_hs_24_6Z9Z12Z15Z18Z21Z_c + crn_c <=> FAcrn_hs_24_6Z9Z12Z15Z18Z21Z_c + coa_c\n",
      "CRNAT_24_9Z12Z15Z18Z: FAcoa_hs_24_9Z12Z15Z18Z_c + crn_c <=> FAcrn_hs_24_9Z12Z15Z18Z_c + coa_c\n",
      "CRNAT_24_9Z12Z15Z18Z21Z: FAcoa_hs_24_9Z12Z15Z18Z21Z_c + crn_c <=> FAcrn_hs_24_9Z12Z15Z18Z21Z_c + coa_c\n",
      "CRNAT_26_0: FAcoa_hs_26_0_c + crn_c <=> FAcrn_hs_26_0_c + coa_c\n",
      "CRNAT_26_17Z: FAcoa_hs_26_17Z_c + crn_c <=> FAcrn_hs_26_17Z_c + coa_c\n",
      "CRNAT_2_0: accoa_c + crn_c <=> acrn_c + coa_c\n",
      "CRNAT_3_0: crn_c + ppcoa_c <=> FAcrn_hs_3_0_c + coa_c\n",
      "CRNAT_3_2E: FAcoa_hs_3_2E_c + crn_c <=> FAcrn_hs_3_2E_c + coa_c\n",
      "CRNAT_3_3OH_S: FAcoa_hs_3_3OH__S_c + crn_c <=> FAcrn_hs_3_3OH__S_c + coa_c\n",
      "CRNAT_3_DC: crn_c + malcoa_c <=> coa_c + malcrn_c\n",
      "CRNAT_4_0: btcoa_c + crn_c <=> FAcrn_hs_4_0_c + coa_c\n",
      "CRNAT_4_2E: FAcoa_hs_4_2E_c + crn_c <=> FAcrn_hs_4_2E_c + coa_c\n",
      "CRNAT_4_2M: FAcoa_4_2M_c + crn_c <=> FAcrn_4_2M_c + coa_c\n",
      "CRNAT_4_3OH_R: FAcoa_hs_4_3OH__R_c + crn_c <=> FAcrn_hs_4_3OH__R_c + coa_c\n",
      "CRNAT_4_DC: FAcoa_4_DC_c + crn_c <=> FAcrn_4_DC_c + coa_c\n",
      "CRNAT_5_0: FAcoa_hs_5_0_c + crn_c <=> FAcrn_hs_5_0_c + coa_c\n",
      "CRNAT_5_2EDC: FAcoa_5_2EDC_c + crn_c <=> FAcrn_5_2EDC_c + coa_c\n",
      "CRNAT_5_2M: FAcoa_5_2M_c + crn_c <=> FAcrn_5_2M_c + coa_c\n",
      "CRNAT_5_2M2E: FAcoa_5_2E2M_c + crn_c <=> FAcrn_5_2E2M_c + coa_c\n",
      "CRNAT_5_3M3OH__S: FAcoa_5_3M3OH__S_c + crn_c <=> FAcrn_5_3M3OH__S_c + coa_c\n",
      "CRNAT_5_DC: FAcoa_5_DC_c + crn_c <=> FAcrn_5_DC_c + coa_c\n",
      "CRNAT_6_0: FAcoa_hs_6_0_c + crn_c <=> FAcrn_hs_6_0_c + coa_c\n",
      "CRNAT_6_2E: FAcoa_hs_6_2E_c + crn_c <=> FAcrn_hs_6_2E_c + coa_c\n",
      "CRNAT_6_3OH_S: FAcoa_hs_6_3OH__S_c + crn_c <=> FAcrn_hs_6_3OH__S_c + coa_c\n",
      "CRNAT_6_DC: FAcoa_6_DC_c + crn_c <=> FAcrn_6_DC_c + coa_c\n",
      "CRNAT_7_0: FAcoa_hs_7_0_c + crn_c <=> FAcrn_hs_7_0_c + coa_c\n",
      "CRNAT_7_3OH_S: FAcoa_hs_7_3OH__S_c + crn_c <=> FAcrn_hs_7_3OH__S_c + coa_c\n",
      "CRNAT_7_DC: FAcoa_7_DC_c + crn_c <=> FAcrn_7_DC_c + coa_c\n",
      "CRNAT_8_0: FAcoa_hs_8_0_c + crn_c <=> FAcrn_hs_8_0_c + coa_c\n",
      "CRNAT_8_2E: FAcoa_hs_8_2E_c + crn_c <=> FAcrn_hs_8_2E_c + coa_c\n",
      "CRNAT_8_3OH_S: FAcoa_hs_8_3OH__S_c + crn_c <=> FAcrn_hs_8_3OH__S_c + coa_c\n",
      "CRNAT_8_DC: FAcoa_8_DC_c + crn_c <=> FAcrn_8_DC_c + coa_c\n",
      "CRNAT_9_0: FAcoa_hs_9_0_c + crn_c <=> FAcrn_hs_9_0_c + coa_c\n",
      "CRNAT_9_2M6M_c: crn_c + dmhptcoa_c <=> coa_c + dmhptcrn_c\n",
      "CRNAT_9_3OH_S: FAcoa_hs_9_3OH__S_c + crn_c <=> FAcrn_hs_9_3OH__S_c + coa_c\n",
      "CRTSLRy: crtsl_c + h_c + nadph_c --> 20bdhcrtsl_c + nadp_c\n",
      "CRTSLt: crtsl_c <=> crtsl_e\n",
      "CRTSTRNRy: crtstrn_c + h_c + nadph_c --> 20bdhcrtstrn_c + nadp_c\n",
      "CRTSTRNt: crtstrn_c <=> crtstrn_e\n",
      "CSND: csn_c + h2o_c + h_c --> nh4_c + ura_c\n",
      "CYSAMO: cysam_c + o2_c --> h_c + hyptaur_c\n",
      "CYSIL2: cysi__L_c --> 2amac_c + thcys_c\n",
      "CYSTRS: atp_c + cys__L_c + trnacys_c --> amp_c + cystrna_c + ppi_c\n",
      "DAGOACAH: dagoac_hs_c + h2o_c --> ac_c + h_c + mago_hs_c\n",
      "DASCBH: dhdascb__L_c + h2o_c --> 23doguln_c\n",
      "DCMPCYTDPT: cytd_c + dcmp_c --> 2dcytd_c + cmp_c\n",
      "DCMPNH: dcmp_c + h2o_c <=> 2dr5p_c + csn_c\n",
      "DCMPURIPT: dcmp_c + uri_c --> 2dcytd_c + ump_c\n",
      "DCT: dpchrm__L_c <=> 56dhica_c\n",
      "DGK1: atp_c + dgmp_c --> adp_c + dgdp_c\n",
      "DGK2: datp_c + dgmp_c --> dadp_c + dgdp_c\n",
      "DHFRy: dhf_c + h_c + nadph_c --> nadp_c + thf_c\n",
      "DHORTS: dhor__S_c + h2o_c <=> cbasp_c + h_c\n",
      "DIIODTHYt2: diiodthy__L_c <=> diiodthy__L_e\n",
      "DLEUtec: leu__D_c <=> leu__D_e\n",
      "DMATT: dmpp_c + ipdp_c --> grdp_c + ppi_c\n",
      "DNADPL: dnadp_c <=> cadprbp_c + h_c + nac_c\n",
      "DOGULND1: 23doguln_c + h2o_c --> co2_c + lyxnt__L_c\n",
      "DOGULND2: 23doguln_c + h2o_c --> co2_c + xylnt__L_c\n",
      "DOGULNO1: 23doguln_c + h2o2_c --> 2.0 h_c + oxa_c + thrnt__L_c\n",
      "DOGULNO2: 23doguln_c + h2o_c --> eryth__L_c + h_c + oxa_c\n",
      "DOPACHRMDC_D: dpchrm__D_c + h_c --> 56dhi_c + co2_c\n",
      "DOPADC_D: dopa__D_c + h_c --> co2_c + dpam__D_c\n",
      "DPMVD: 5dpmev_c + atp_c --> adp_c + co2_c + ipdp_c + pi_c\n",
      "DRDPP: asparg_c + h2o_c --> arg__L_c + asp__L_c\n",
      "DTMPK1: atp_c + dtmp_c --> adp_c + dtdp_c\n",
      "DTMPK2: datp_c + dtmp_c --> dadp_c + dtdp_c\n",
      "DUMPK1: atp_c + dump_c --> adp_c + dudp_c\n",
      "DUMPK2: datp_c + dump_c --> dadp_c + dudp_c\n",
      "E4PD: e4p_c + nad_c + pi_c --> e14bp_c + 2.0 h_c + nadh_c\n",
      "EACPR1y: buteACP_2E_c + h_c + nadph_c --> butACP_c + nadp_c\n",
      "EACPR2y: h_c + hxeaACP_2E_c + nadph_c --> hxaACP_c + nadp_c\n",
      "EACPR3y: h_c + nadph_c + oceaACP_2E_c --> nadp_c + ocaACP_c\n",
      "EACPR4y: dceaACP_2E_c + h_c + nadph_c --> dcaACP_c + nadp_c\n",
      "EACPR5y: ddceaACP_2E_c + h_c + nadph_c --> ddcaACP_c + nadp_c\n",
      "EACPR6y: h_c + nadph_c + ttdceaACP_2E_c --> nadp_c + ttdcaACP_c\n",
      "EACPR7y: h_c + hxdceaACP_2E_c + nadph_c --> hxdcaACP_c + nadp_c\n",
      "EACPR8y: h_c + nadph_c + ocdceaACP_2E_c --> nadp_c + ocdcaACP_c\n",
      "EHGLAT: akg_c + e4hglu_c --> 4h2oglt_c + glu__L_c\n",
      "ERGOTHtuNa: ergoth_e + na1_e --> ergoth_c + na1_c\n",
      "ESTRONESte: estrones_e <=> estrones_c\n",
      "EX_23cgamp_e: 23cgamp_e <=> \n",
      "EX_2dcytd_e: 2dcytd_e <=> \n",
      "EX_3mob_e: 3mob_e <=> \n",
      "EX_4mop_e: 4mop_e <=> \n",
      "EX_4pyrn_e: 4pyrn_e <=> \n",
      "EX_5mthf_e: 5mthf_e <=> \n",
      "EX_FA_hs_10_0_e: FA_hs_10_0_e <=> \n",
      "EX_FA_hs_11_0_e: FA_hs_11_0_e <=> \n",
      "EX_L2aadp_e: L2aadp_e <=> \n",
      "EX_admarg__L_e: admarg__L_e <=> \n",
      "EX_ag1_e: ag1_e <=> \n",
      "EX_bilirub_e: bilirub_e <=> \n",
      "EX_but_e: but_e <=> \n",
      "EX_cd2_e: cd2_e <=> \n",
      "EX_cholate_e: cholate_e <=> \n",
      "EX_chsterol_e: chsterol_e <=> \n",
      "EX_cobalt2_e: cobalt2_e <=> \n",
      "EX_crn_e: crn_e <=> \n",
      "EX_crtsl_e: crtsl_e <=> \n",
      "EX_crtstrn_e: crtstrn_e <=> \n",
      "EX_dgchol_e: dgchol_e <=> \n",
      "EX_dheas_e: dheas_e <=> \n",
      "EX_diiodthy__L_e: diiodthy__L_e <=> \n",
      "EX_ergoth_e: ergoth_e <=> \n",
      "EX_estrones_e: estrones_e <=> \n",
      "EX_gchola_e: gchola_e <=> \n",
      "EX_gdchac_e: gdchac_e <=> \n",
      "EX_glcn_e: glcn_e <=> \n",
      "EX_gudchola_e: gudchola_e <=> \n",
      "EX_hLkynr_e: hLkynr_e <=> \n",
      "EX_his__L_e: his__L_e <=> \n",
      "EX_hpta_e: hpta_e <=> \n",
      "EX_hxa_e: hxa_e <=> \n",
      "EX_i_e: i_e <=> \n",
      "EX_ile__L_e: ile__L_e <=> \n",
      "EX_im4ac_e: im4ac_e <=> \n",
      "EX_inds_e: inds_e <=> \n",
      "EX_kynr__L_e: kynr__L_e <=> \n",
      "EX_leu__D_e: leu__D_e <=> \n",
      "EX_leu__L_e: leu__L_e <=> \n",
      "EX_li1_e: li1_e <=> \n",
      "EX_lys__L_e: lys__L_e <=> \n",
      "EX_mg2_e: mg2_e <=> \n",
      "EX_mma_e: mma_e <=> \n",
      "EX_mmarg__L_e: mmarg__L_e <=> \n",
      "EX_mn2_e: mn2_e <=> \n",
      "EX_mobd_e: mobd_e <=> \n",
      "EX_nona_e: nona_e <=> \n",
      "EX_octa_e: octa_e <=> \n",
      "EX_pheme_e: pheme_e <=> \n",
      "EX_ppa_e: ppa_e <=> \n",
      "EX_ppp9_e: ppp9_e <=> \n",
      "EX_prgnlone_e: prgnlone_e <=> \n",
      "EX_prgnlones_e: prgnlones_e <=> \n",
      "EX_pta_e: pta_e <=> \n",
      "EX_retinol_e: retinol_e <=> \n",
      "EX_ribavirin_e: ribavirin_e <=> \n",
      "EX_ribflv_e: ribflv_e <=> \n",
      "EX_tchola_e: tchola_e <=> \n",
      "EX_tdchola_e: tdchola_e <=> \n",
      "EX_tdechola_e: tdechola_e <=> \n",
      "EX_tetiodthy__L_e: tetiodthy__L_e <=> \n",
      "EX_triiodthy__L_e: triiodthy__L_e <=> \n",
      "EX_trp__L_e: trp__L_e <=> \n",
      "EX_tudchola_e: tudchola_e <=> \n",
      "EX_val__L_e: val__L_e <=> \n",
      "EX_zn2_e: zn2_e <=> \n",
      "F1PGT: fuc1p__L_c + gtp_c + h_c <=> gdpfuc_c + ppi_c\n",
      "FA120ACPH: ddcaACP_c + h2o_c --> ACP_c + FA_hs_12_0_c + h_c\n",
      "FA140ACPH: h2o_c + ttdcaACP_c --> ACP_c + FA_hs_14_0_c + h_c\n",
      "FA160ACPH: h2o_c + hxdcaACP_c --> ACP_c + FA_hs_16_0_c + h_c\n",
      "FA180ACPH: h2o_c + ocdcaACP_c --> ACP_c + FA_hs_18_0_c + h_c\n",
      "FACOA2ERy_16_2E: FAcoa_hs_16_2E_c + h_c + nadph_c --> FAcoa_hs_16_0_c + nadp_c\n",
      "FACOA2ERy_17_2E: FAcoa_hs_17_2E_c + h_c + nadph_c --> FAcoa_hs_17_0_c + nadp_c\n",
      "FACOA2ERy_18_2E: FAcoa_hs_18_2E_c + h_c + nadph_c --> FAcoa_hs_18_0_c + nadp_c\n",
      "FACOA2ERy_19_2E: FAcoa_hs_19_2E_c + h_c + nadph_c --> FAcoa_hs_19_0_c + nadp_c\n",
      "FACOA2ERy_20_2E: FAcoa_hs_20_2E_c + h_c + nadph_c --> FAcoa_hs_20_0_c + nadp_c\n",
      "FACOA2ERy_20_2E11Z: FAcoa_hs_20_2E11Z_c + h_c + nadph_c --> FAcoa_hs_20_11Z_c + nadp_c\n",
      "FACOA2ERy_20_2E11Z14Z: FAcoa_hs_20_2E11Z14Z_c + h_c + nadph_c --> FAcoa_hs_20_11Z14Z_c + nadp_c\n",
      "FACOA2ERy_20_2E11Z14Z17Z: FAcoa_hs_20_2E11Z14Z17Z_c + h_c + nadph_c --> FAcoa_hs_20_11Z14Z17Z_c + nadp_c\n",
      "FACOA2ERy_20_2E13Z: FAcoa_hs_20_2E13Z_c + h_c + nadph_c --> FAcoa_hs_20_13Z_c + nadp_c\n",
      "FACOA2ERy_20_2E5Z8Z11Z14Z: FAcoa_hs_20_2E8Z11Z14Z17Z_c + h_c + nadph_c --> FAcoa_hs_20_8Z11Z14Z17Z_c + nadp_c\n",
      "FACOA2ERy_20_2E8Z11Z: FAcoa_hs_20_2E8Z11Z_c + h_c + nadph_c --> FAcoa_hs_20_8Z11Z_c + nadp_c\n",
      "FACOA2ERy_20_2E8Z11Z14Z: FAcoa_hs_20_2E8Z11Z14Z_c + h_c + nadph_c --> FAcoa_hs_20_8Z11Z14Z_c + nadp_c\n",
      "FACOA2ERy_20_2E9Z: FAcoa_hs_20_2E9Z_c + h_c + nadph_c --> FAcoa_hs_20_9Z_c + nadp_c\n",
      "FACOA2ERy_21_2E: FAcoa_hs_21_2E_c + h_c + nadph_c --> FAcoa_hs_21_0_c + nadp_c\n",
      "FACOA2ERy_22_2E: FAcoa_hs_22_2E_c + h_c + nadph_c --> FAcoa_hs_22_0_c + nadp_c\n",
      "FACOA2ERy_22_2E10Z13Z16Z: FAcoa_hs_22_2E10Z13Z16Z_c + h_c + nadph_c --> FAcoa_hs_22_10Z13Z16Z_c + nadp_c\n",
      "FACOA2ERy_22_2E10Z13Z16Z19Z: FAcoa_hs_22_2E10Z13Z16Z19Z_c + h_c + nadph_c --> FAcoa_hs_22_10Z13Z16Z19Z_c + nadp_c\n",
      "FACOA2ERy_22_2E11Z: FAcoa_hs_22_2E11Z_c + h_c + nadph_c --> FAcoa_hs_22_11Z_c + nadp_c\n",
      "FACOA2ERy_22_2E13Z: FAcoa_hs_22_2E13Z_c + h_c + nadph_c --> FAcoa_hs_22_13Z_c + nadp_c\n",
      "FACOA2ERy_22_2E13Z16Z: FAcoa_hs_22_2E13Z16Z_c + h_c + nadph_c --> FAcoa_hs_22_13Z16Z_c + nadp_c\n",
      "FACOA2ERy_22_2E13Z16Z19Z: FAcoa_hs_22_2E13Z16Z19Z_c + h_c + nadph_c --> FAcoa_hs_22_13Z16Z19Z_c + nadp_c\n",
      "FACOA2ERy_22_2E7Z10Z13Z16Z: FAcoa_hs_22_2E7Z10Z13Z16Z_c + h_c + nadph_c --> FAcoa_hs_22_7Z10Z13Z16Z_c + nadp_c\n",
      "FACOA2ERy_22_2E7Z10Z13Z16Z19Z: FAcoa_hs_22_2E7Z10Z13Z16Z19Z_c + h_c + nadph_c --> FAcoa_hs_22_7Z10Z13Z16Z19Z_c + nadp_c\n",
      "FACOA2ERy_23_2E: FAcoa_hs_23_2E_c + h_c + nadph_c --> FAcoa_hs_23_0_c + nadp_c\n",
      "FACOA2ERy_24_2E: FAcoa_hs_24_2E_c + h_c + nadph_c --> FAcoa_hs_24_0_c + nadp_c\n",
      "FACOA2ERy_24_2E12Z15Z18Z21Z: FAcoa_hs_24_2E12Z15Z18Z21Z_c + h_c + nadph_c --> FAcoa_hs_24_12Z15Z18Z21Z_c + nadp_c\n",
      "FACOA2ERy_24_2E15Z: FAcoa_hs_24_2E15Z_c + h_c + nadph_c --> FAcoa_hs_24_15Z_c + nadp_c\n",
      "FACOA2ERy_24_2E9Z12Z15Z18Z: FAcoa_hs_24_2E9Z12Z15Z18Z_c + h_c + nadph_c --> FAcoa_hs_24_9Z12Z15Z18Z_c + nadp_c\n",
      "FACOA2ERy_24_2E9Z12Z15Z18Z21Z: FAcoa_hs_24_2E9Z12Z15Z18Z21Z_c + h_c + nadph_c --> FAcoa_hs_24_9Z12Z15Z18Z21Z_c + nadp_c\n",
      "FACOA2ERy_26_2E: FAcoa_hs_26_2E_c + h_c + nadph_c --> FAcoa_hs_26_0_c + nadp_c\n",
      "FACOA2ERy_26_2E17Z: FAcoa_hs_26_2E17Z_c + h_c + nadph_c --> FAcoa_hs_26_17Z_c + nadp_c\n",
      "FACOA3OHDH_16_3OH: FAcoa_hs_16_3OH__R_c --> FAcoa_hs_16_2E_c + h2o_c\n",
      "FACOA3OHDH_17_3OH: FAcoa_hs_17_3OH__R_c --> FAcoa_hs_17_2E_c + h2o_c\n",
      "FACOA3OHDH_18_3OH: FAcoa_hs_18_3OH__R_c --> FAcoa_hs_18_2E_c + h2o_c\n",
      "FACOA3OHDH_19_3OH: FAcoa_hs_19_3OH__R_c --> FAcoa_hs_19_2E_c + h2o_c\n",
      "FACOA3OHDH_20_11Z14Z17Z3OH: FAcoa_hs_20_11Z14Z17Z3OH__R_c --> FAcoa_hs_20_2E11Z14Z17Z_c + h2o_c\n",
      "FACOA3OHDH_20_11Z14Z3OH: FAcoa_hs_20_11Z14Z3OH__R_c --> FAcoa_hs_20_2E11Z14Z_c + h2o_c\n",
      "FACOA3OHDH_20_11Z3OH: FAcoa_hs_20_11Z3OH__R_c --> FAcoa_hs_20_2E11Z_c + h2o_c\n",
      "FACOA3OHDH_20_13Z3OH: FAcoa_hs_20_13Z3OH__R_c --> FAcoa_hs_20_2E13Z_c + h2o_c\n",
      "FACOA3OHDH_20_3OH: FAcoa_hs_20_3OH__R_c --> FAcoa_hs_20_2E_c + h2o_c\n",
      "FACOA3OHDH_20_5Z8Z11Z14Z3OH: FAcoa_hs_20_8Z11Z14Z17Z3OH__R_c --> FAcoa_hs_20_2E8Z11Z14Z17Z_c + h2o_c\n",
      "FACOA3OHDH_20_8Z11Z14Z3OH: FAcoa_hs_20_8Z11Z14Z3OH__R_c --> FAcoa_hs_20_2E8Z11Z14Z_c + h2o_c\n",
      "FACOA3OHDH_20_8Z11Z3OH: FAcoa_hs_20_8Z11Z3OH__R_c --> FAcoa_hs_20_2E8Z11Z_c + h2o_c\n",
      "FACOA3OHDH_20_9Z3OH: FAcoa_hs_20_9Z3OH__R_c --> FAcoa_hs_20_2E9Z_c + h2o_c\n",
      "FACOA3OHDH_21_3OH: FAcoa_hs_21_3OH__R_c --> FAcoa_hs_21_2E_c + h2o_c\n",
      "FACOA3OHDH_22_10Z13Z16Z19Z3OH: FAcoa_hs_22_10Z13Z16Z19Z3OH__R_c --> FAcoa_hs_22_2E10Z13Z16Z19Z_c + h2o_c\n",
      "FACOA3OHDH_22_10Z13Z16Z3OH: FAcoa_hs_22_10Z13Z16Z3OH__R_c --> FAcoa_hs_22_2E10Z13Z16Z_c + h2o_c\n",
      "FACOA3OHDH_22_11Z3OH: FAcoa_hs_22_11Z3OH__R_c --> FAcoa_hs_22_2E11Z_c + h2o_c\n",
      "FACOA3OHDH_22_13Z16Z19Z3OH: FAcoa_hs_22_13Z16Z19Z3OH__R_c --> FAcoa_hs_22_2E13Z16Z19Z_c + h2o_c\n",
      "FACOA3OHDH_22_13Z16Z3OH: FAcoa_hs_22_13Z16Z3OH__R_c --> FAcoa_hs_22_2E13Z16Z_c + h2o_c\n",
      "FACOA3OHDH_22_13Z3OH: FAcoa_hs_22_13Z3OH__R_c --> FAcoa_hs_22_2E13Z_c + h2o_c\n",
      "FACOA3OHDH_22_3OH: FAcoa_hs_22_3OH__R_c --> FAcoa_hs_22_2E_c + h2o_c\n",
      "FACOA3OHDH_22_7Z10Z13Z16Z19Z3OH: FAcoa_hs_22_7Z10Z13Z16Z19Z3OH__R_c --> FAcoa_hs_22_2E7Z10Z13Z16Z19Z_c + h2o_c\n",
      "FACOA3OHDH_22_7Z10Z13Z16Z3OH: FAcoa_hs_22_7Z10Z13Z16Z3OH__R_c --> FAcoa_hs_22_2E7Z10Z13Z16Z_c + h2o_c\n",
      "FACOA3OHDH_23_3OH: FAcoa_hs_23_3OH__R_c --> FAcoa_hs_23_2E_c + h2o_c\n",
      "FACOA3OHDH_24_12Z15Z18Z21Z3OH: FAcoa_hs_24_12Z15Z18Z21Z3OH__R_c --> FAcoa_hs_24_2E12Z15Z18Z21Z_c + h2o_c\n",
      "FACOA3OHDH_24_15Z3OH: FAcoa_hs_24_15Z3OH__R_c --> FAcoa_hs_24_2E15Z_c + h2o_c\n",
      "FACOA3OHDH_24_3OH: FAcoa_hs_24_3OH__R_c --> FAcoa_hs_24_2E_c + h2o_c\n",
      "FACOA3OHDH_24_9Z12Z15Z18Z21Z3OH: FAcoa_hs_24_9Z12Z15Z18Z21Z3OH__R_c --> FAcoa_hs_24_2E9Z12Z15Z18Z21Z_c + h2o_c\n",
      "FACOA3OHDH_24_9Z12Z15Z18Z3OH: FAcoa_hs_24_9Z12Z15Z18Z3OH__R_c --> FAcoa_hs_24_2E9Z12Z15Z18Z_c + h2o_c\n",
      "FACOA3OHDH_26_17Z3OH: FAcoa_hs_26_17Z3OH__R_c --> FAcoa_hs_26_2E17Z_c + h2o_c\n",
      "FACOA3OHDH_26_3OH: FAcoa_hs_26_3OH__R_c --> FAcoa_hs_26_2E_c + h2o_c\n",
      "FACOA3ORy_16_3O: FAcoa_hs_16_3O_c + h_c + nadph_c --> FAcoa_hs_16_3OH__R_c + nadp_c\n",
      "FACOA3ORy_17_3O: FAcoa_hs_17_3O_c + h_c + nadph_c --> FAcoa_hs_17_3OH__R_c + nadp_c\n",
      "FACOA3ORy_18_3O: FAcoa_hs_18_3O_c + h_c + nadph_c --> FAcoa_hs_18_3OH__R_c + nadp_c\n",
      "FACOA3ORy_19_3O: FAcoa_hs_19_3O_c + h_c + nadph_c --> FAcoa_hs_19_3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_11Z14Z17Z3O: FAcoa_hs_20_11Z14Z17Z3O_c + h_c + nadph_c --> FAcoa_hs_20_11Z14Z17Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_11Z14Z3O: FAcoa_hs_20_11Z14Z3O_c + h_c + nadph_c --> FAcoa_hs_20_11Z14Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_11Z3O: FAcoa_hs_20_11Z3O_c + h_c + nadph_c --> FAcoa_hs_20_11Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_13Z3O: FAcoa_hs_20_13Z3O_c + h_c + nadph_c --> FAcoa_hs_20_13Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_3O: FAcoa_hs_20_3O_c + h_c + nadph_c --> FAcoa_hs_20_3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_5Z8Z11Z14Z3O: FAcoa_hs_20_8Z11Z14Z17Z3O_c + h_c + nadph_c --> FAcoa_hs_20_8Z11Z14Z17Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_8Z11Z14Z3O: FAcoa_hs_20_8Z11Z14Z3O_c + h_c + nadph_c --> FAcoa_hs_20_8Z11Z14Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_8Z11Z3O: FAcoa_hs_20_8Z11Z3O_c + h_c + nadph_c --> FAcoa_hs_20_8Z11Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_20_9Z3O: FAcoa_hs_20_9Z3O_c + h_c + nadph_c --> FAcoa_hs_20_9Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_21_3O: FAcoa_hs_21_3O_c + h_c + nadph_c --> FAcoa_hs_21_3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_10Z13Z16Z19Z3O: FAcoa_hs_22_10Z13Z16Z19Z3O_c + h_c + nadph_c --> FAcoa_hs_22_10Z13Z16Z19Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_10Z13Z16Z3O: FAcoa_hs_22_10Z13Z16Z3O_c + h_c + nadph_c --> FAcoa_hs_22_10Z13Z16Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_11Z3O: FAcoa_hs_22_11Z3O_c + h_c + nadph_c --> FAcoa_hs_22_11Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_13Z16Z19Z3O: FAcoa_hs_22_13Z16Z19Z3O_c + h_c + nadph_c --> FAcoa_hs_22_13Z16Z19Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_13Z16Z3O: FAcoa_hs_22_13Z16Z3O_c + h_c + nadph_c --> FAcoa_hs_22_13Z16Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_13Z3O: FAcoa_hs_22_13Z3O_c + h_c + nadph_c --> FAcoa_hs_22_13Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_3O: FAcoa_hs_22_3O_c + h_c + nadph_c --> FAcoa_hs_22_3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_7Z10Z13Z16Z19Z3O: FAcoa_hs_22_7Z10Z13Z16Z19Z3O_c + h_c + nadph_c --> FAcoa_hs_22_7Z10Z13Z16Z19Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_22_7Z10Z13Z16Z3O: FAcoa_hs_22_7Z10Z13Z16Z3O_c + h_c + nadph_c --> FAcoa_hs_22_7Z10Z13Z16Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_23_3O: FAcoa_hs_23_3O_c + h_c + nadph_c --> FAcoa_hs_23_3OH__R_c + nadp_c\n",
      "FACOA3ORy_24_12Z15Z18Z21Z3O: FAcoa_hs_24_12Z15Z18Z21Z3O_c + h_c + nadph_c --> FAcoa_hs_24_12Z15Z18Z21Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_24_15Z3O: FAcoa_hs_24_15Z3O_c + h_c + nadph_c --> FAcoa_hs_24_15Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_24_3O: FAcoa_hs_24_3O_c + h_c + nadph_c --> FAcoa_hs_24_3OH__R_c + nadp_c\n",
      "FACOA3ORy_24_9Z12Z15Z18Z21Z3O: FAcoa_hs_24_9Z12Z15Z18Z21Z3O_c + h_c + nadph_c --> FAcoa_hs_24_9Z12Z15Z18Z21Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_24_9Z12Z15Z18Z3O: FAcoa_hs_24_9Z12Z15Z18Z3O_c + h_c + nadph_c --> FAcoa_hs_24_9Z12Z15Z18Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_26_17Z3O: FAcoa_hs_26_17Z3O_c + h_c + nadph_c --> FAcoa_hs_26_17Z3OH__R_c + nadp_c\n",
      "FACOA3ORy_26_3O: FAcoa_hs_26_3O_c + h_c + nadph_c --> FAcoa_hs_26_3OH__R_c + nadp_c\n",
      "FADRy: fad_c + 2.0 h_c + nadph_c --> fadh2_c + nadp_c\n",
      "FAWOXx_10_10OH: FA_hs_10_10OH_c + nad_c --> FA_hs_10_10O_c + h_c + nadh_c\n",
      "FAWOXx_10_10OXO: FA_hs_10_10O_c + h2o_c + nad_c --> FA_hs_10_DC_c + 2.0 h_c + nadh_c\n",
      "FAWOXx_12_12OH: FA_hs_12_12OH_c + nad_c --> FA_hs_12_12O_c + h_c + nadh_c\n",
      "FAWOXx_12_12OXO: FA_hs_12_12O_c + h2o_c + nad_c --> FA_hs_12_DC_c + 2.0 h_c + nadh_c\n",
      "FAWOXx_14_14OH: FA_hs_14_14OH_c + nad_c --> FA_hs_14_14O_c + h_c + nadh_c\n",
      "FAWOXx_14_14OXO: FA_hs_14_14O_c + h2o_c + nad_c --> FA_hs_14_DC_c + 2.0 h_c + nadh_c\n",
      "FAWOXx_16_016OH: FA_hs_16_16OH_c + nad_c --> FA_hs_16_16O_c + h_c + nadh_c\n",
      "FAWOXx_16_016OXO: FA_hs_16_16O_c + h2o_c + nad_c --> FA_hs_16_DC_c + 2.0 h_c + nadh_c\n",
      "FAWOXx_20_5Z8Z11Z14Z20OH: FA_hs_20_5Z8Z11Z14Z20OH_c + nad_c --> FA_hs_20_5Z8Z11Z14Z20O_c + h_c + nadh_c\n",
      "FAWOXx_20_5Z8Z11Z14Z20OXO: FA_hs_20_5Z8Z11Z14Z20O_c + h2o_c + nad_c --> FA_hs_20_5Z8Z11Z14Z20DC_c + 2.0 h_c + nadh_c\n",
      "FAWOXx_22_22OH: FA_hs_22_22OH_c + nad_c --> FA_hs_22_22O_c + h_c + nadh_c\n",
      "FAWOXx_22_22OXO: FA_hs_22_22O_c + h2o_c + nad_c --> FA_hs_22_DC_c + 2.0 h_c + nadh_c\n",
      "FAt_hs_10_0: FA_hs_10_0_e <=> FA_hs_10_0_c\n",
      "FAt_hs_11_0: FA_hs_11_0_e <=> FA_hs_11_0_c\n",
      "FAt_hs_5_0: pta_e <=> pta_c\n",
      "FAt_hs_6_0: hxa_e <=> hxa_c\n",
      "FAt_hs_7_0: hpta_e <=> hpta_c\n",
      "FAt_hs_8_0: octa_e <=> octa_c\n",
      "FAt_hs_9_0: nona_e <=> nona_c\n",
      "FBA3: s17bp_c <=> dhap_c + e4p_c\n",
      "FBA6: o18bp_c <=> dhap_c + r5p_c\n",
      "FE2PHEMEABCte: atp_c + h2o_c + pheme_c --> adp_c + h_c + pheme_e + pi_c\n",
      "FE2PHEMEOX: h2o_c + o2_c + pheme_c --> h_c + hematin_c + o2s_c\n",
      "FISC: 4.0 fe2_c + 6.0 h2o_c + o2_c --> 4.0 fe3ooh_c + 8.0 h_c\n",
      "FK: atp_c + fuc__L_c --> adp_c + fuc1p__L_c + h_c\n",
      "FKYNH: Lfmkynr_c + h2o_c --> for_c + h_c + kynr__L_c\n",
      "FLDPP: h2o_c + pheleu_c --> leu__L_c + phe__L_c\n",
      "FLVRy: 2.0 h_c + nadph_c + ribflv_c --> nadp_c + rblvrd_c\n",
      "FPDPP: h2o_c + phepro_c --> phe__L_c + pro__L_c\n",
      "FRNSL2E6EDH: frnsl2e6e_c + h2o_c + nad_c --> frnsn2e6e_c + 2.0 h_c + nadh_c\n",
      "FRU3K: atp_c + fru_c --> adp_c + f3p_c + h_c\n",
      "FUTGAL2A: galacgalfuc12gal14acglcgalgluside_hs_c + gdpfuc_c --> fucgalacgalfuc12gal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTGLOBO1: galgbside_hs_c + gdpfuc_c --> gdp_c + globoH_hs_c + h_c\n",
      "FUTINLC8CER1: galacglcgalacglcgal14acglcgalgluside_hs_c + gdpfuc_c --> fucgalacglcgalacglcgal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTINLC8CER2: fucgalacglcgalacglcgal14acglcgalgluside_hs_c + gdpfuc_c --> fucfucgalacglcgalacglcgal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTLC4CER: galacglcgalgluside_hs_c + gdpfuc_c --> fucgalacglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTLC5CER: galgalacglcgalgluside_hs_c + gdpfuc_c --> galfucgalacglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTNLC5CER: galgal14acglcgalgluside_hs_c + gdpfuc_c --> galfuc12gal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTNLC6CER: galacglcgal14acglcgalgluside_hs_c + gdpfuc_c --> fucgalacglcgal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTNLC7CER: galacgalfucgalacglcgal14acglcgalgluside_hs_c + gdpfuc_c --> fucgalacgalfucgalacglcgal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FUTPGBSIDE: gdpfuc_c + paragbside_hs_c --> fuc12gal14acglcgalgluside_hs_c + gdp_c + h_c\n",
      "FYDPP: h2o_c + phetyr_c --> phe__L_c + tyr__L_c\n",
      "G5SAD: glu5sa__L_c <=> 1pyr5c__L_c + h2o_c + h_c\n",
      "GARFT: 10fthf_c + gar_c <=> fgam_c + h_c + thf_c\n",
      "GFDPP: glyphe_c + h2o_c --> gly_c + phe__L_c\n",
      "GFUCSx: gdpddman_c + h_c + nadh_c --> gdpfuc_c + nad_c\n",
      "GFUCSy: gdpddman_c + h_c + nadph_c --> gdpfuc_c + nadp_c\n",
      "GGDPP: glygly_c + h2o_c --> 2.0 gly_c\n",
      "GHBDHy: h_c + nadph_c + sucsal_c <=> 4ohbut_c + nadp_c\n",
      "GHMT3: 3htmelys_c + h_c --> 4tmeabut_c + gly_c\n",
      "GLACOx: glac_c + 2.0 h2o_c + nad_c --> glcr_c + 3.0 h_c + nadh_c\n",
      "GLCNte: glcn_c --> glcn_e\n",
      "GLDPP: glyleu_c + h2o_c --> gly_c + leu__L_c\n",
      "GLNTRS: atp_c + gln__L_c + trnagln_c --> amp_c + glntrna_c + ppi_c\n",
      "GLUTRS: atp_c + glu__L_c + trnaglu_c --> amp_c + glutrna_c + ppi_c\n",
      "GLYTRS: atp_c + gly_c + trnagly_c --> amp_c + glytrna_c + ppi_c\n",
      "GMAND: gdpmann_c --> gdpddman_c + h2o_c\n",
      "GNK: atp_c + glcn_c --> 6pgc_c + adp_c + h_c\n",
      "GP4AH1: gp4g_c + h2o_c --> gmp_c + gtp_c + 2.0 h_c\n",
      "GPDDA6: g3pnae_hs_c + h2o_c --> glyc3p_c + h_c + nae_hs_c\n",
      "GPDPP: glypro_c + h2o_c --> gly_c + pro__L_c\n",
      "GRTT: grdp_c + ipdp_c --> frdp_c + ppi_c\n",
      "GTHNHOHD: gthnhoh_c --> gthnoh2_c\n",
      "GTHNOH2HYD: gthnoh2_c + h2o_c --> nh4_c + so2gth_c\n",
      "GTHNOH2RD: gthnhoh_c + gthrd_c --> gthox_c + ham_c\n",
      "GTHNORDx: gthno_c + h_c + nadh_c --> gthnhoh_c + nad_c\n",
      "GTHNORDy: gthno_c + h_c + nadph_c --> gthnhoh_c + nadp_c\n",
      "GWDPP: glytrp_c + h2o_c --> gly_c + trp__L_c\n",
      "HEXGA2: ga2_hs_c + h2o_c --> acgal_c + galgluside_hs_c\n",
      "HEXGBSIDE: gbside_hs_c + h2o_c --> acgal_c + thcrm_hs_c\n",
      "HEXGM2: gm2_hs_c + h2o_c --> acgal_c + gm3_hs_c\n",
      "HISD: his__L_c --> nh4_c + urcan_c\n",
      "HISTRS: atp_c + his__L_c + trnahis_c --> amp_c + histrna_c + ppi_c\n",
      "HMOBD_Cltex: cl_c + h_e + mobd_e <=> cl_e + h_c + mobd_c\n",
      "HPTALDHx: h2o_c + hptal_c + nad_c --> 2.0 h_c + hpta_c + nadh_c\n",
      "HPTEAL2EDHx: h2o_c + hpteal_2E_c + nad_c --> 2.0 h_c + hptea_2E_c + nadh_c\n",
      "HXALDHx: h2o_c + hxal_c + nad_c --> 2.0 h_c + hxa_c + nadh_c\n",
      "HXEAL2EDHx: h2o_c + hxeal_2E_c + nad_c --> 2.0 h_c + hxea_2E_c + nadh_c\n",
      "ILETRS: atp_c + ile__L_c + trnaile_c --> amp_c + iletrna_c + ppi_c\n",
      "IM4ACte: im4ac_c <=> im4ac_e\n",
      "IMACTD: h2o_c + im4act_c + nad_c --> 2.0 h_c + im4ac_c + nadh_c\n",
      "INS6P1K: atp_c + ins6p_c --> 1ppins5p_c + adp_c\n",
      "INSULINGTHOR: 2.0 gthrd_c + insulinox_c --> gthox_c + insulinrd_c\n",
      "It: i_c <=> i_e\n",
      "KLDPP: h2o_c + lysleu_c --> leu__L_c + lys__L_c\n",
      "KYNAKGAT: akg_c + kynr__L_c --> 4aphdob_c + glu__L_c\n",
      "KYNATESYN: 4aphdob_c --> h2o_c + kynate_c\n",
      "L2AADPt: L2aadp_e --> L2aadp_c\n",
      "L2HGDHx: akg_c + h_c + nadh_c --> 2hglut__L_c + nad_c\n",
      "L4HGSA: 4hglusa__L_c <=> 1p3h5c__L_c + h2o_c + h_c\n",
      "LADPP: h2o_c + leuala_c --> ala__L_c + leu__L_c\n",
      "LCYSTAT: Lcyst_c + akg_c <=> 3spyr_c + glu__L_c\n",
      "LEUTRS: atp_c + leu__L_c + trnaleu_c --> amp_c + leutrna_c + ppi_c\n",
      "LGDPP: h2o_c + leugly_c --> gly_c + leu__L_c\n",
      "LGFTPAP: h2o_c + leuglyphe_c --> glyphe_c + leu__L_c\n",
      "LGGTPAP: h2o_c + leuglygly_c --> glygly_c + leu__L_c\n",
      "LKYNRtr: kynr__L_c <=> kynr__L_e\n",
      "LLDPP: h2o_c + leuleu_c --> 2.0 leu__L_c\n",
      "LLLTPAP: h2o_c + leuleuleu_c --> leu__L_c + leuleu_c\n",
      "LNAPELPLD: h2o_c + lnape_hs_c --> h_c + lpa_hs_c + nae_hs_c\n",
      "LNSTLS: Ssq23epx_c --> lanost_c\n",
      "LPCOLPLD: h2o_c + lpco_hs_c --> chol_c + h_c + lpao_hs_c\n",
      "LPDPP: h2o_c + leupro_c --> leu__L_c + pro__L_c\n",
      "LYSTRS: atp_c + lys__L_c + trnalys_c --> amp_c + lystrna_c + ppi_c\n",
      "M16BPS1: 13dpg_c + man1p_c <=> 3pg_c + h_c + man16bp_c\n",
      "MAGOK: atp_c + mago_hs_c --> adp_c + h_c + lpao_hs_c\n",
      "MALDHx: h2o_c + mald_c + nad_c --> 2.0 h_c + msa_c + nadh_c\n",
      "MAN1PT: gtp_c + h_c + man1p_c <=> gdpmann_c + ppi_c\n",
      "MCLOR: h_c + mercppyr_c + nadh_c --> mercplac_c + nad_c\n",
      "MCOATA: ACP_c + malcoa_c --> coa_c + malACP_c\n",
      "METMSPYRAT: 2kmb_c + smscys__L_c <=> met__L_c + metselpyr_c\n",
      "METSELNEORy: h_c + methselne_c + nadph_c --> h2o_c + methsel_c + nadp_c\n",
      "METSELNIORy: 2.0 h_c + methselni_c + nadph_c --> h2o_c + methselne_c + nadp_c\n",
      "METTRS: atp_c + met__L_c + trnamet_c --> amp_c + mettrna_c + ppi_c\n",
      "MI13456P3PP: h2o_c + mi13456p_c --> mi1456p_c + pi_c\n",
      "MI1345P5PP: h2o_c + mi1345p_c --> mi134p_c + pi_c\n",
      "MI1456P6PP: h2o_c + mi1456p_c --> mi145p_c + pi_c\n",
      "MMALNDHx: 2mop_c + h2o_c + nad_c --> 2.0 h_c + mmaln_c + nadh_c\n",
      "MMAt2: mma_c <=> mma_e\n",
      "MNDN2ERy: h_c + mndn_c + nadph_c --> mndiol_c + nadp_c\n",
      "MPDPP: h2o_c + metpro_c --> met__L_c + pro__L_c\n",
      "MPMOBDT: h_c + mobd_c + mptamp_c --> amp_c + h2o_c + moco_c\n",
      "MPTAT: atp_c + h_c + mpt_c --> mptamp_c + ppi_c\n",
      "N1ASPMDOX: N1aspmd_c + h2o_c + o2_c --> aappal_c + h2o2_c + ptrc_c\n",
      "N1SPRMOX: N1sprm_c + h2o_c + o2_c --> aappal_c + h2o2_c + spmd_c\n",
      "NABTNO: h2o_c + n4abutn_c + nad_c --> 4aabutn_c + 2.0 h_c + nadh_c\n",
      "NADN2: anad_c + h2o_c --> adprib_c + h_c + ncam_c\n",
      "NADPL: nadp_c <=> cadprbp_c + h_c + ncam_c\n",
      "NADPN: h2o_c + nadp_c --> adprbp_c + h_c + ncam_c\n",
      "NAPEPLD: h2o_c + nape_hs_c --> h_c + nae_hs_c + pa_hs_c\n",
      "NDPK10: atp_c + didp_c --> adp_c + ditp_c\n",
      "NDPK4: atp_c + dtdp_c --> adp_c + dttp_c\n",
      "NDPK5: atp_c + dgdp_c --> adp_c + dgtp_c\n",
      "NDPK6: atp_c + dudp_c --> adp_c + dutp_c\n",
      "NEU24: gm1_hs_c + h2o_c --> acnam_c + ga1_hs_c\n",
      "NEU25: gm2_hs_c + h2o_c --> acnam_c + ga2_hs_c\n",
      "NEU31: gm3_hs_c + h2o_c --> acnam_c + galgluside_hs_c\n",
      "NEU32: gd1a_hs_c + h2o_c --> acnam_c + gm1_hs_c\n",
      "NEU33: gd1b_hs_c + h2o_c --> acnam_c + gm1_hs_c\n",
      "NEU34: gt1b_hs_c + h2o_c --> acnam_c + gd1b_hs_c\n",
      "NEU36: gq1b_hs_c + h2o_c --> acnam_c + gt1b_hs_c\n",
      "NNDPR: 2.0 h_c + prpp_c + quln_c --> co2_c + nicrnt_c + ppi_c\n",
      "NONALDHx: h2o_c + nad_c + nonal_c --> 2.0 h_c + nadh_c + nona_c\n",
      "NONEAL2EDHx: h2o_c + nad_c + noneal_2E_c --> 2.0 h_c + nadh_c + nonea_2E_c\n",
      "NTD2AMP: 2amp_c + h2o_c --> adn_c + pi_c\n",
      "NTD2CMP: 2cmp_c + h2o_c --> cytd_c + pi_c\n",
      "NTD2GMP: 2gmp_c + h2o_c --> gsn_c + pi_c\n",
      "NTD2UMP: 2ump_c + h2o_c --> pi_c + uri_c\n",
      "NTDDCMP: dcmp_c + h2o_c --> 2dcytd_c + pi_c\n",
      "NTDOMP: h2o_c + orot5p_c --> orotdn_c + pi_c\n",
      "NTDPSUMP: h2o_c + psump_c --> pi_c + psuri_c\n",
      "NTDSAMP: h2o_c + succamp_c --> pi_c + succadn_c\n",
      "NTPPDI: ditp_c + h2o_c --> dimp_c + h_c + ppi_c\n",
      "NTPPX: h2o_c + xtp_c --> h_c + ppi_c + xmp_c\n",
      "OCTALDHx: h2o_c + nad_c + octal_c --> 2.0 h_c + nadh_c + octa_c\n",
      "OCTEAL2EDHx: h2o_c + nad_c + octeal_2E_c --> 2.0 h_c + nadh_c + octea_2E_c\n",
      "P4AH: h2o_c + p4a_c --> atp_c + h_c + pi_c\n",
      "PEPTDPROI: protres_pepproW0_c <=> protres_pepproW180_c\n",
      "PFDPP: h2o_c + prophe_c --> phe__L_c + pro__L_c\n",
      "PGD21GH: h2o_c + prostgd21glyc_c --> glyc_c + h_c + prostgd2_c\n",
      "PGDF21GH: h2o_c + prostgf21glyc_c --> glyc_c + h_c + prostgf2_c\n",
      "PGE1DHy: nadp_c + prostge1_c <=> 15kprostge1_c + h_c + nadph_c\n",
      "PGE21GH: h2o_c + prostge21glyc_c --> glyc_c + h_c + prostge2_c\n",
      "PGE2DHy: nadp_c + prostge2_c <=> 15kprostge2_c + h_c + nadph_c\n",
      "PGE3DHy: nadp_c + prostge3_c <=> 15kprostge3_c + h_c + nadph_c\n",
      "PGF1DHy: nadp_c + prostgf1_c <=> 15kprostgf1_c + h_c + nadph_c\n",
      "PGF2DHy: nadp_c + prostgf2_c <=> 15kprostgf2_c + h_c + nadph_c\n",
      "PGF3DHy: nadp_c + prostgf3_c <=> 15kprostgf3_c + h_c + nadph_c\n",
      "PHEMEt: pheme_c --> pheme_e\n",
      "PHETRS: atp_c + phe__L_c + trnaphe_c --> amp_c + phetrna_c + ppi_c\n",
      "PHLAC: h_c + nadh_c + phpyr_c --> nad_c + phlac_c\n",
      "PHPYRT: phpyr_c --> 2h3ppr_c\n",
      "PI35P4K: atp_c + pail35p_hs_c --> adp_c + h_c + pail345p_hs_c\n",
      "PI3P4K: atp_c + pail3p_hs_c --> adp_c + h_c + pail34p_hs_c\n",
      "PLDPP: h2o_c + proleu_c --> leu__L_c + pro__L_c\n",
      "PMANM: man1p_c <=> man6p_c\n",
      "POLYADPRH: h2o_c + polyadprib2_c --> adprib_c + polyadprib1_c\n",
      "PPALDHx: h2o_c + nad_c + ppal_c --> 2.0 h_c + nadh_c + ppa_c\n",
      "PPDPP: h2o_c + propro_c --> 2.0 pro__L_c\n",
      "PPGPPDP: h2o_c + ppgpp_c --> gdp_c + h_c + ppi_c\n",
      "PPP9ABCte: atp_c + h2o_c + ppp9_c --> adp_c + h_c + pi_c + ppp9_e\n",
      "PRAGS: atp_c + gly_c + pram_c --> adp_c + gar_c + h_c + pi_c\n",
      "PRAIS: atp_c + fpram_c --> adp_c + air_c + 2.0 h_c + pi_c\n",
      "PRASCS: 5aizc_c + asp__L_c + atp_c --> adp_c + h_c + pi_c + saicar_c\n",
      "PRFGS: atp_c + fgam_c + gln__L_c + h2o_c --> adp_c + fpram_c + glu__L_c + h_c + pi_c\n",
      "PRGNABCte: atp_c + h2o_c + prgnlone_c --> adp_c + h_c + pi_c + prgnlone_e\n",
      "PROTKACTPAC: actp_c + protres_lys__L_c --> h_c + pi_c + protres_aclys__L_c\n",
      "PROTKRBLK: atp_c + protres_rbllys_c --> adp_c + h_c + protres_rbllys3p_c\n",
      "PROTKRU3PDG: h2o_c + protres_rbllys3p_c --> 3dxyl_c + pi_c + protres_lys__L_c\n",
      "PROTKSUCT: FAcoa_4_DC_c + protres_lys__L_c <=> coa_c + h_c + protres_succlys_c\n",
      "PROTRS: atp_c + pro__L_c + trnapro_c --> amp_c + ppi_c + protrna_c\n",
      "PRPPSZ: aicar_c + h_c + prpp_c --> r5p_c + ztp_c\n",
      "PYRRLHYD: 4abutn_c <=> h2o_c + pyrrl_c\n",
      "QMPNH: h2o_c + qmp_c --> que_c + r5p_c\n",
      "QUILSYN: cmusa_c --> h2o_c + h_c + quln_c\n",
      "R5PCT: ctp_c + h_c + r5p_c --> cdprib_c + ppi_c\n",
      "RADH1x: h2o_c + nad_c + retinal_c --> 2.0 h_c + nadh_c + retn_c\n",
      "RADH2x: h2o_c + nad_c + retinal_9Z_c --> 2.0 h_c + nadh_c + retn_9Z_c\n",
      "RADH3x: h2o_c + nad_c + retinal_11Z_c --> 2.0 h_c + nadh_c + retn_11Z_c\n",
      "RADH4x: h2o_c + nad_c + retinal_13Z_c --> 2.0 h_c + nadh_c + retn_13Z_c\n",
      "RBTCT: ctp_c + h_c + rbt5p_c --> cdprbt_c + ppi_c\n",
      "RBTDG2_Dx: h_c + nadh_c + rbl__D_c --> nad_c + rbt_c\n",
      "RBVDA: h2o_c + ribavirin_c --> nh4_c + rbvcbxlt_c\n",
      "RBVPNP: pi_c + ribavirin_c <=> r1p_c + triaz3crbxam_c\n",
      "RBVt: ribavirin_c <=> ribavirin_e\n",
      "RDH1y: h_c + nadph_c + retinal_c <=> nadp_c + retinol_c\n",
      "RDH2y: h_c + nadph_c + retinal_9Z_c <=> nadp_c + retinol_9Z_c\n",
      "RDH3y: h_c + nadph_c + retinal_11Z_c <=> nadp_c + retinol_11Z_c\n",
      "RDH4y: h_c + nadph_c + retinal_13Z_c <=> nadp_c + retinol_13Z_c\n",
      "RETt: retinol_c <=> retinol_e\n",
      "RPPTPAP: argpropro_c + h2o_c --> arg__L_c + propro_c\n",
      "RU5PCT: ctp_c + h_c + ru5p__D_c --> cdpru5p__D_c + ppi_c\n",
      "SEAHCYSHYD: h2o_c + selahcys__L_c --> adn_c + selhcys__L_c\n",
      "SELCYSTL1: selcyst__L_c --> 2abuten_c + selcys__L_c\n",
      "SELCYSTL2: selcyst__L_c --> 2amac_c + selhcys__L_c\n",
      "SELMETAT: atp_c + h2o_c + selmet__L_c --> pi_c + ppi_c + selamet__L_c\n",
      "SELMETHL1: selmet__L_c --> 2abuten_c + methsel_c\n",
      "SERTRS: atp_c + ser__L_c + trnaser_c --> amp_c + ppi_c + sertrna_c\n",
      "SHPK: atp_c + sdhplse_c --> adp_c + h_c + s7p_c\n",
      "SLDx: 3spyr_c + h_c + nadh_c <=> nad_c + slac__L_c\n",
      "SMSCYSL2: smscys__L_c --> 2amac_c + methsel_c\n",
      "SNOCOARD: h_c + nadph_c + snocoa_c --> nadp_c + snoh2coa_c\n",
      "SPMDACT: accoa_c + spmd_c --> N1aspmd_c + coa_c + h_c\n",
      "SPRMACT: accoa_c + sprm_c --> N1sprm_c + coa_c + h_c\n",
      "SUCACETO: h_c + sucacetat_c --> co2_c + sucaceto_c\n",
      "TDECHOLAte: tdechola_e <=> tdechola_c\n",
      "TETIODTHYt2: tetiodthy__L_c <=> tetiodthy__L_e\n",
      "THGDPK: atp_c + thiogdp_c --> adp_c + thiogtp_c\n",
      "THGMPK: atp_c + thiogmp_c --> adp_c + thiogdp_c\n",
      "THIOGSNPNP: pi_c + thiogsn_c --> r1p_c + thiogua_c\n",
      "THIOGUAPRT: prpp_c + thiogua_c --> ppi_c + thiogmp_c\n",
      "THRTLD_Lx: nad_c + thrtl__L_c <=> eryth__L_c + h_c + nadh_c\n",
      "THRTRS: atp_c + thr__L_c + trnathr_c --> amp_c + ppi_c + thrtrna_c\n",
      "TMABADH: 4tmeabut_c + h2o_c + nad_c --> 4tmeabutn_c + 2.0 h_c + nadh_c\n",
      "TRIIODTHYt2: triiodthy__L_c <=> triiodthy__L_e\n",
      "TRPHLKYNR: hLkynr_e + trp__L_c <=> hLkynr_c + trp__L_e\n",
      "TRPTRS: atp_c + trnatrp_c + trp__L_c --> amp_c + ppi_c + trptrna_c\n",
      "TYRTRS: atp_c + trnatyr_c + tyr__L_c --> amp_c + ppi_c + tyrtrna_c\n",
      "UAGALDP: acgal1p_c + h_c + utp_c --> ppi_c + udpacgal_c\n",
      "UREAHYDL: urea_c <=> cynamd_c + h2o_c\n",
      "VALTRS: atp_c + trnaval_c + val__L_c --> amp_c + ppi_c + valtrna_c\n",
      "VLDPP: h2o_c + valleu_c --> leu__L_c + val__L_c\n",
      "VPDPP: h2o_c + valpro_c --> pro__L_c + val__L_c\n",
      "VYDPP: h2o_c + valtyr_c --> tyr__L_c + val__L_c\n",
      "XNTHRNSYN: 42a3hp24db_c + h_c --> h2o_c + xanthurn_c\n",
      "XOLEST2EH: h2o_c + xolest2_hs_c --> FA_hs_c + chsterol_c + h_c\n",
      "YYDPP: h2o_c + tyrtyr_c --> 2.0 tyr__L_c\n",
      "YYYTPAP: h2o_c + tyrtyrtyr_c --> tyr__L_c + tyrtyr_c\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of blocked reactions: {len(blocked_reactions)}\")\n",
    "mets_one_reaction = model.metabolites.query(\n",
    "    lambda x: len(x.reactions) == 1 and not any([r for r in x.reactions if r.boundary])\n",
    ")\n",
    "mets_one_reaction = sorted(mets_one_reaction.list_attr(\"id\"))\n",
    "print(f\"Number of metabolites with one associated reaction {len(mets_one_reaction)}\")\n",
    "print()\n",
    "for reaction in model.reactions.get_by_any(sorted(blocked_reactions)):\n",
    "    # Expected blocked\n",
    "    # if any([\n",
    "    #     x in reaction.id\n",
    "    #     for x in [\"3OHAA\", \"BKAACP\", 'EACPR1', 'CRNAT', 'FACOA3OHD', '3OXAACPR1', \"ACPH\", 'FACOA2ERy', '3OXAACPR', 'FACOA3ORy']\n",
    "    # ]):\n",
    "    #     continue\n",
    "    print(reaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cad316c-ec10-461b-9c38-4e1bdeed2db5",
   "metadata": {},
   "source": [
    "### Ensure correct types before export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "ea115b87-ffa6-4773-8207-08701a8d03b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SBML will not export charges correctly if they are float\n",
    "for metabolite in model.metabolites:\n",
    "    metabolite.charge = int(metabolite.charge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77e6dc2-ed20-455b-aadd-8a03db5666f1",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "742aeb63-f41b-4280-8fc0-9a79e9410e3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14a880c50</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14a880c50>"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_cobra_model(\n",
    "    filename=get_dirpath(\"model\", use_temp=\"interim\" if not overwrite else None)\n",
    "    / f\"{model.id}.xml\"\n",
    ")\n",
    "write_cobra_model(\n",
    "    filename=get_dirpath(\"model\", use_temp=\"interim\" if not overwrite else None)\n",
    "    / f\"{model.id}.json\"\n",
    ")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "978d93a7-23dc-47ee-b9b9-f330fdfa8f05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes: 820\n",
      "Metabolites (all): 2157\n",
      "Metabolites (unique): 1685\n",
      "Reactions: 2725\n"
     ]
    }
   ],
   "source": [
    "print(f\"Genes: {len(set([x.id for x in model.genes]))}\")\n",
    "print(f\"Metabolites (all): {len(set([x.id for x in model.metabolites]))}\")\n",
    "nmets_unique = len({x.id.replace(f\"_{x.compartment}\", \"\") for x in model.metabolites})\n",
    "print(f\"Metabolites (unique): {nmets_unique}\")\n",
    "print(\n",
    "    f\"Reactions: {len(set([x.id for x in model.reactions.query(lambda x: not x.boundary)]))}\"\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
