{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc9114c6-b2d1-48a1-8ba3-6686a1ea2449",
   "metadata": {},
   "source": [
    "# Compute statistically significant fluxes between groups - REDS Recall\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2d694e76-31c0-4aba-add6-a07a00a5d04c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.3\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.13.4\n",
      "bio                                   1.8.0\n",
      "cobra                                0.29.1\n",
      "depinfo                               2.2.0\n",
      "gurobipy                             12.0.3\n",
      "matplotlib                           3.10.3\n",
      "matplotlib-venn                       1.1.2\n",
      "memote                               0.17.0\n",
      "networkx                                3.5\n",
      "notebook                              7.4.4\n",
      "openpyxl                              3.1.5\n",
      "pandas                                2.3.1\n",
      "pre-commit                            4.2.0\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.32.4\n",
      "scikit-learn                          1.7.0\n",
      "scipy                                1.16.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip          25.1\n",
      "setuptools 78.1.1\n",
      "wheel      0.45.1\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Windows 10-AMD64\n",
      "CPython  3.11.13\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import textwrap\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from itertools import combinations\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    ensure_iterable,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    ")\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    DEFAULT_PREFIX_SUFFIX_VALUES,\n",
    "    DEFAULT_PROTEOME_COMPARTMENT,\n",
    "    EnzymeDilution,\n",
    "    add_relaxation_budget,\n",
    "    load_overlay_model,\n",
    "    plot_correlations,\n",
    ")\n",
    "from rbc_gem_utils.visualization import cmap_map\n",
    "from scipy.stats import false_discovery_control, kruskal, mannwhitneyu, spearmanr\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "384c972a-9f2f-4b88-a65c-017d1a6561d6",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "418ddc9c-b48e-400a-b5a1-e4e6bfe73bd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>100000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1e-08\n",
       "upper_bound: 100000000.0\n",
       "processes: 127\n",
       "cache_directory: C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "# Set bound defaults much larger to prevent model loading issues\n",
    "COBRA_CONFIGURATION.bounds = (-1e-8, 1e8)\n",
    "COBRA_CONFIGURATION.tolerance = 1e-9\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c3720a-d637-4115-8960-f48671ee725a",
   "metadata": {},
   "source": [
    "### Define organism, model, and dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5b08faa-240f-44ee-9d16-ddbdc556852f",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = \"Human\"\n",
    "model_id = \"RBC_GEM\"\n",
    "dataset_name = \"REDSRecall\"\n",
    "grouped_data_key = \"Sample\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0ccb7e-c784-4b63-bfe9-7ebdce0e24b5",
   "metadata": {},
   "source": [
    "### Set variables for sample identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6435bfd2-9c1f-4f09-8a7a-3d0a0efba741",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For sample IDs\n",
    "sample_key = \"SAMPLE ID\"\n",
    "donor_key = \"PUBLIC RECALL DONOR ID\"\n",
    "time_key = \"DAY\"\n",
    "timepoints = [\"D10\", \"D23\", \"D42\"]\n",
    "genotypes = [\"G6PD_V68M\", \"ATP11C_V972M\"]\n",
    "donor_re = re.compile(rf\"(?P<donor>S(?P<num>\\d+))\")\n",
    "time_re = re.compile(rf\"(?P<time>{'|'.join(timepoints)})\")\n",
    "genotype_re = re.compile(rf\"(?P<genotype>({'|'.join(genotypes)}))\")\n",
    "\n",
    "operations = \"|\".join([x.capitalize() for x in [\"mean\", \"median\"]])\n",
    "\n",
    "operation_re = re.compile(r\"(?P<op>\" + operations + r\")\\_(?P<group>\\w+)\")\n",
    "sample_id_re = re.compile(\n",
    "    r\"(?!\" + operations + r\")\" + donor_re.pattern + r\"\\_\" + time_re.pattern\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e478c85d-b0ed-466c-b6d3-e8dead52f018",
   "metadata": {},
   "source": [
    "### Set computation options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52e59494-b23d-4775-8863-1baeb1f46194",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"xml\"  # In our experience, SBML/XML loads faster, but will take up to 4x more space uncompressed as compared to JSON\n",
    "run_computations = True  # Keep off to use previously computed results\n",
    "overwrite = False  # Whether to allow overwriting of previous simulation results\n",
    "verbose = True\n",
    "\n",
    "# Objective reactions\n",
    "objective_reactions = [\"NaKt\"]\n",
    "# Reactions that must have the capability to carry flux, sort for consistency\n",
    "required_flux_reactions = []  # Add reactions to this list\n",
    "required_flux_reactions = sorted(set(objective_reactions + required_flux_reactions))\n",
    "\n",
    "\n",
    "enzyme_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.dilution\"]\n",
    "enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.metabolite\"]\n",
    "enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]\n",
    "comp_suffix = f\"_{DEFAULT_PROTEOME_COMPARTMENT}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5b2f1cf-43bb-4e48-8e56-031b2f14202a",
   "metadata": {},
   "source": [
    "#### Set prefixes/suffixes to expect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f6c451c4-4f3c-4db0-991b-fd56b743f587",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'DEFAULT_PREFIX_SUFFIX_VALUES' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[1]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m enzyme_rxn_prefix = \u001b[43mDEFAULT_PREFIX_SUFFIX_VALUES\u001b[49m[\u001b[33m\"\u001b[39m\u001b[33menzymes\u001b[39m\u001b[33m\"\u001b[39m][\u001b[33m\"\u001b[39m\u001b[33mprefix.dilution\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m      2\u001b[39m enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\u001b[33m\"\u001b[39m\u001b[33menzymes\u001b[39m\u001b[33m\"\u001b[39m][\u001b[33m\"\u001b[39m\u001b[33mprefix.metabolite\u001b[39m\u001b[33m\"\u001b[39m]\n\u001b[32m      3\u001b[39m enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\u001b[33m\"\u001b[39m\u001b[33menzymes\u001b[39m\u001b[33m\"\u001b[39m][\u001b[33m\"\u001b[39m\u001b[33msuffix.total\u001b[39m\u001b[33m\"\u001b[39m]\n",
      "\u001b[31mNameError\u001b[39m: name 'DEFAULT_PREFIX_SUFFIX_VALUES' is not defined"
     ]
    }
   ],
   "source": [
    "enzyme_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.dilution\"]\n",
    "enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.metabolite\"]\n",
    "enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]\n",
    "comp_suffix = f\"_{DEFAULT_PROTEOME_COMPARTMENT}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "295bb5e1-67d6-4436-af4a-0f34befaa111",
   "metadata": {},
   "source": [
    "### Set figure options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "06ad3312-69a9-4bd9-92c4-a2d467bed985",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc2d514-cb81-4968-95b6-95e758cd3e73",
   "metadata": {},
   "source": [
    "### Set paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ccd462b-0c65-4b33-b745-6a4e74996f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set paths\n",
    "processed_data_dirpath = get_dirpath(use_temp=\"processed\") / organism / dataset_name\n",
    "overlay_dirpath = get_dirpath(\"analysis\") / \"OVERLAY\" / organism\n",
    "model_dirpath = overlay_dirpath / model_id\n",
    "results_dirpath = (\n",
    "    get_dirpath(use_temp=\"processed\")\n",
    "    / model_id\n",
    "    / \"OVERLAY\"\n",
    "    / organism\n",
    "    / dataset_name\n",
    "    / grouped_data_key\n",
    ")\n",
    "pcfva_results_dirpath = (\n",
    "    results_dirpath\n",
    "    / \"pcFVA\"\n",
    "    / \"_\".join((\"REQ\", *required_flux_reactions))\n",
    "    / \"_\".join((\"OBJ\", *objective_reactions))\n",
    ")\n",
    "\n",
    "# Objective reaction does not matter since correlations are computed\n",
    "# based on min and max fluxes and abundance, which are obtained when optimum is 0.\n",
    "corr_results_dirpath = results_dirpath / \"correlations\"\n",
    "# Ensure directory  exists\n",
    "corr_results_dirpath.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "806b8d0f-8ca2-48e4-9da1-7db464f01407",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "541f7b2c-ad0d-4404-899b-fa05a0bd7fa1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Set parameter LicenseID to value 2664191\n",
      "Academic license - for non-commercial use only - expires 2026-05-12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>25671cdca10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>10411</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>19619</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC at 0x25671cdca10>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_cobra_model(filename=model_dirpath / f\"{model_id}.xml\")\n",
    "pcmodel = load_overlay_model(filename=model_dirpath / f\"{model_id}_PC.xml\")\n",
    "\n",
    "# Add relaxation budget to initial PC model to get names of relaxation reactions\n",
    "add_relaxation_budget(pcmodel, 0, verbose=False)\n",
    "pcmodel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49cf0307-bf28-4f10-8784-143d62c669a7",
   "metadata": {},
   "source": [
    "## Load pcFVA generated results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "537fccd8-52b6-44c9-83a0-af5c62b7fd2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755171</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755172</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755173</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755174</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54755175</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>54755176 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   model        reactions  optimum  min  max\n",
       "0         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.00  0.0  0.0\n",
       "1         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.50  0.0  0.0\n",
       "2         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.90  0.0  0.0\n",
       "3         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.99  0.0  0.0\n",
       "4         RBC_GEM_PC_Mean_ATP11C_V972M_0        15KPGE1Ry     0.00  0.0  0.0\n",
       "...                                  ...              ...      ...  ...  ...\n",
       "54755171             RBC_GEM_PC_S650_D42  ZN2_HCO3_SELNIt     0.99  0.0  0.0\n",
       "54755172             RBC_GEM_PC_S650_D42             ZN2t     0.00  0.0  0.0\n",
       "54755173             RBC_GEM_PC_S650_D42             ZN2t     0.50  0.0  0.0\n",
       "54755174             RBC_GEM_PC_S650_D42             ZN2t     0.90  0.0  0.0\n",
       "54755175             RBC_GEM_PC_S650_D42             ZN2t     0.99  0.0  0.0\n",
       "\n",
       "[54755176 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test to see if results were recently generated in this run, otherwise load DataFrame of generated results\n",
    "df_pcfva_all = pd.read_csv(\n",
    "    pcfva_results_dirpath / f\"{pcmodel.id}_All_FVAsols.zip\",\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "df_pcfva_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63a4524-c95b-4980-a95f-5ba109e75025",
   "metadata": {},
   "source": [
    "## Create DataFrame for calculations and visualizations\n",
    "### Get maximum reaction fluxes and associated abundance values\n",
    "#### Get maximum reaction fluxes and ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3efaec1d-6db8-4b96-975b-67303d06d9ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [min, max, range]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxns = model.reactions.query(lambda x: len(x.genes))\n",
    "df_max_flux_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "df_max_flux_per_model = df_max_flux_per_model.groupby(\n",
    "    [\"model\", \"reactions\", \"optimum\"]\n",
    ")[[\"min\", \"max\"]].agg(\n",
    "    {\n",
    "        \"min\": \"min\",  # Minimum reaction flux per model\n",
    "        \"max\": \"max\",  # Maximum reaction flux per model\n",
    "    }\n",
    ")\n",
    "# Address issues possibly caused by floating point precision, ideally a value that prevents any negative ranges\n",
    "df_max_flux_per_model.loc[\n",
    "    df_max_flux_per_model[\"max\"] < df_max_flux_per_model[\"min\"], [\"max\", \"min\"]\n",
    "] = [0, 0]\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_flux_per_model[\"max\"] = df_max_flux_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_flux_per_model[\"min\"] = df_max_flux_per_model[\"min\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_flux_per_model[\"range\"] = (\n",
    "    df_max_flux_per_model[\"max\"] - df_max_flux_per_model[\"min\"]\n",
    ")\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_flux_per_model[df_max_flux_per_model[\"range\"] < 1e-9]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e000d2d8-7a7c-4b02-b088-867e04ce7dbe",
   "metadata": {},
   "source": [
    "#### Get maximum \"enzyme\" abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4d6feeea-cfe4-4da9-99db-47f2729bd4e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [abundance]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxns = pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, EnzymeDilution)\n",
    "    and x.id.endswith(f\"{enzyme_met_suffix_total}{comp_suffix}\")\n",
    ").list_attr(\"id\")\n",
    "df_max_abundance_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "# Rename dilution reactions to match\n",
    "reaction_enzyme_map = {\n",
    "    enzyme_rid: enzyme_rid.replace(\n",
    "        f\"{enzyme_rxn_prefix}{enzyme_met_prefix}\", \"\"\n",
    "    ).replace(f\"{enzyme_met_suffix_total}{comp_suffix}\", \"\")\n",
    "    for enzyme_rid in df_max_abundance_per_model[\"reactions\"]\n",
    "}\n",
    "df_max_abundance_per_model[\"reactions\"] = df_max_abundance_per_model[\n",
    "    \"reactions\"\n",
    "].replace(reaction_enzyme_map)\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.groupby(\n",
    "    [\"model\", \"reactions\", \"optimum\"]\n",
    ")[[\"max\"]].max()\n",
    "# Address issues possibly caused by floating point precision, atol is ideally a value that prevents any negative ranges\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if x < 0 else x\n",
    ")\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.rename(\n",
    "    {\"max\": \"abundance\"}, axis=1\n",
    ")\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_abundance_per_model[(df_max_abundance_per_model < 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95269b45-4f16-4d09-95c5-af4734b23d82",
   "metadata": {},
   "source": [
    "#### Merge DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "60602d4b-89ba-493b-ae29-55a35c669a31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249083</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249084</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249085</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249086</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249087</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25249088 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   model        reactions  optimum  min  max  \\\n",
       "0         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.00  0.0  0.0   \n",
       "1         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.50  0.0  0.0   \n",
       "2         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.90  0.0  0.0   \n",
       "3         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.99  0.0  0.0   \n",
       "4         RBC_GEM_PC_Mean_ATP11C_V972M_0        15KPGE1Ry     0.00  0.0  0.0   \n",
       "...                                  ...              ...      ...  ...  ...   \n",
       "25249083             RBC_GEM_PC_S650_D42  ZN2_HCO3_SELNIt     0.99  0.0  0.0   \n",
       "25249084             RBC_GEM_PC_S650_D42             ZN2t     0.00  0.0  0.0   \n",
       "25249085             RBC_GEM_PC_S650_D42             ZN2t     0.50  0.0  0.0   \n",
       "25249086             RBC_GEM_PC_S650_D42             ZN2t     0.90  0.0  0.0   \n",
       "25249087             RBC_GEM_PC_S650_D42             ZN2t     0.99  0.0  0.0   \n",
       "\n",
       "          range  abundance  \n",
       "0           0.0        NaN  \n",
       "1           0.0        NaN  \n",
       "2           0.0        NaN  \n",
       "3           0.0        NaN  \n",
       "4           0.0   0.005077  \n",
       "...         ...        ...  \n",
       "25249083    0.0   0.000000  \n",
       "25249084    0.0   0.000000  \n",
       "25249085    0.0   0.000000  \n",
       "25249086    0.0   0.000000  \n",
       "25249087    0.0   0.000000  \n",
       "\n",
       "[25249088 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_all = pd.merge(\n",
    "    df_max_flux_per_model,\n",
    "    df_max_abundance_per_model,\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    ")\n",
    "df_data_all = df_data_all.reset_index(drop=False)\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3094239-cea4-4899-8cba-84b0f3969ba3",
   "metadata": {},
   "source": [
    "### Identify donor, timepoints, and genotypes for results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "04139aab-4c0c-4845-a65f-912ac235da32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G6PD_V68M\n",
      "0    621\n",
      "1      7\n",
      "2     11\n",
      "Name: count, dtype: Int64\n",
      "\n",
      "ATP11C_V972M\n",
      "0    625\n",
      "1     12\n",
      "2      2\n",
      "Name: count, dtype: Int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df_metadata = pd.read_csv(\n",
    "    processed_data_dirpath / f\"{dataset_name}_Metadata.csv\",\n",
    "    index_col=[sample_key],\n",
    ").convert_dtypes()\n",
    "df_metadata = df_metadata.loc[:, genotypes]\n",
    "\n",
    "df = df_metadata.reset_index(drop=False)\n",
    "df[sample_key] = df[sample_key].str.split(\"_\", expand=True).iloc[:, 0]\n",
    "df = df.drop_duplicates().set_index(sample_key)\n",
    "for col, series in df.items():\n",
    "    print(series.value_counts().sort_index())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "788ad86d-3278-43c9-9e41-27b9f358440d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>sample</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>G6PD_V68M</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mean_ATP11C_V972M_0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mean_ATP11C_V972M_0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mean_ATP11C_V972M_0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mean_ATP11C_V972M_0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_Mean_ATP11C_V972M_0</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.005077</td>\n",
       "      <td>Mean_ATP11C_V972M_0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249083</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249084</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249085</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249086</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25249087</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25249088 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   model        reactions  optimum  min  max  \\\n",
       "0         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.00  0.0  0.0   \n",
       "1         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.50  0.0  0.0   \n",
       "2         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.90  0.0  0.0   \n",
       "3         RBC_GEM_PC_Mean_ATP11C_V972M_0        13DAMPPOX     0.99  0.0  0.0   \n",
       "4         RBC_GEM_PC_Mean_ATP11C_V972M_0        15KPGE1Ry     0.00  0.0  0.0   \n",
       "...                                  ...              ...      ...  ...  ...   \n",
       "25249083             RBC_GEM_PC_S650_D42  ZN2_HCO3_SELNIt     0.99  0.0  0.0   \n",
       "25249084             RBC_GEM_PC_S650_D42             ZN2t     0.00  0.0  0.0   \n",
       "25249085             RBC_GEM_PC_S650_D42             ZN2t     0.50  0.0  0.0   \n",
       "25249086             RBC_GEM_PC_S650_D42             ZN2t     0.90  0.0  0.0   \n",
       "25249087             RBC_GEM_PC_S650_D42             ZN2t     0.99  0.0  0.0   \n",
       "\n",
       "          range  abundance               sample donor  time  G6PD_V68M  \\\n",
       "0           0.0        NaN  Mean_ATP11C_V972M_0  <NA>  <NA>       <NA>   \n",
       "1           0.0        NaN  Mean_ATP11C_V972M_0  <NA>  <NA>       <NA>   \n",
       "2           0.0        NaN  Mean_ATP11C_V972M_0  <NA>  <NA>       <NA>   \n",
       "3           0.0        NaN  Mean_ATP11C_V972M_0  <NA>  <NA>       <NA>   \n",
       "4           0.0   0.005077  Mean_ATP11C_V972M_0  <NA>  <NA>       <NA>   \n",
       "...         ...        ...                  ...   ...   ...        ...   \n",
       "25249083    0.0   0.000000             S650_D42  S650   D42          0   \n",
       "25249084    0.0   0.000000             S650_D42  S650   D42          0   \n",
       "25249085    0.0   0.000000             S650_D42  S650   D42          0   \n",
       "25249086    0.0   0.000000             S650_D42  S650   D42          0   \n",
       "25249087    0.0   0.000000             S650_D42  S650   D42          0   \n",
       "\n",
       "          ATP11C_V972M  \n",
       "0                    0  \n",
       "1                    0  \n",
       "2                    0  \n",
       "3                    0  \n",
       "4                    0  \n",
       "...                ...  \n",
       "25249083             0  \n",
       "25249084             0  \n",
       "25249085             0  \n",
       "25249086             0  \n",
       "25249087             0  \n",
       "\n",
       "[25249088 rows x 12 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_columns = [\"sample\", \"donor\", \"time\", \"genotype\"]\n",
    "for key, search_re in zip(\n",
    "    metadata_columns, [sample_id_re, donor_re, time_re, genotype_re]\n",
    "):\n",
    "    if key == \"sample\":\n",
    "        df_data_all[key] = df_data_all[\"model\"].apply(\n",
    "            lambda x: x.replace(f\"{pcmodel.id}_\", \"\")\n",
    "        )\n",
    "    else:\n",
    "        df_data_all[key] = df_data_all[\"model\"].apply(\n",
    "            lambda x: search_re.search(x).group(1) if search_re.search(x) else pd.NA\n",
    "        )\n",
    "\n",
    "df_data_all = df_data_all.merge(\n",
    "    df_metadata, left_on=\"sample\", right_index=True, how=\"left\"\n",
    ")\n",
    "\n",
    "# Add genotypes alleles\n",
    "for genotype in genotypes:\n",
    "    df = df_data_all[df_data_all[\"genotype\"] == genotype][\"sample\"]\n",
    "    df = df.str.rsplit(\"_\", n=1, expand=True)\n",
    "    if df.empty:\n",
    "        continue\n",
    "    df = df.iloc[:, -1]\n",
    "    df_data_all.loc[df.index, genotype] = df.values\n",
    "df_data_all = df_data_all.drop(\"genotype\", axis=1)\n",
    "metadata_columns = metadata_columns[:-1] + genotypes\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3902ff15-bf6a-43a9-ad82-d81742231410",
   "metadata": {},
   "source": [
    "## Compute statistically significant results between groups\n",
    "### Remove models based on data operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "76ce7d4c-834d-4b2f-b705-059839846fea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>sample</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>G6PD_V68M</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_S001_D10</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S001_D10</td>\n",
       "      <td>S001</td>\n",
       "      <td>D10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_S001_D10</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S001_D10</td>\n",
       "      <td>S001</td>\n",
       "      <td>D10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_S001_D10</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S001_D10</td>\n",
       "      <td>S001</td>\n",
       "      <td>D10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_S001_D10</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S001_D10</td>\n",
       "      <td>S001</td>\n",
       "      <td>D10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_S001_D10</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S001_D10</td>\n",
       "      <td>S001</td>\n",
       "      <td>D10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25013355</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25013356</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25013357</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25013358</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25013359</th>\n",
       "      <td>RBC_GEM_PC_S650_D42</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>S650_D42</td>\n",
       "      <td>S650</td>\n",
       "      <td>D42</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25013360 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        model        reactions  optimum  min  max  range  \\\n",
       "0         RBC_GEM_PC_S001_D10        13DAMPPOX     0.00  0.0  0.0    0.0   \n",
       "1         RBC_GEM_PC_S001_D10        13DAMPPOX     0.50  0.0  0.0    0.0   \n",
       "2         RBC_GEM_PC_S001_D10        13DAMPPOX     0.90  0.0  0.0    0.0   \n",
       "3         RBC_GEM_PC_S001_D10        13DAMPPOX     0.99  0.0  0.0    0.0   \n",
       "4         RBC_GEM_PC_S001_D10        15KPGE1Ry     0.00  0.0  0.0    0.0   \n",
       "...                       ...              ...      ...  ...  ...    ...   \n",
       "25013355  RBC_GEM_PC_S650_D42  ZN2_HCO3_SELNIt     0.99  0.0  0.0    0.0   \n",
       "25013356  RBC_GEM_PC_S650_D42             ZN2t     0.00  0.0  0.0    0.0   \n",
       "25013357  RBC_GEM_PC_S650_D42             ZN2t     0.50  0.0  0.0    0.0   \n",
       "25013358  RBC_GEM_PC_S650_D42             ZN2t     0.90  0.0  0.0    0.0   \n",
       "25013359  RBC_GEM_PC_S650_D42             ZN2t     0.99  0.0  0.0    0.0   \n",
       "\n",
       "          abundance    sample donor time  G6PD_V68M  ATP11C_V972M  \n",
       "0               NaN  S001_D10  S001  D10          0             0  \n",
       "1               NaN  S001_D10  S001  D10          0             0  \n",
       "2               NaN  S001_D10  S001  D10          0             0  \n",
       "3               NaN  S001_D10  S001  D10          0             0  \n",
       "4               0.0  S001_D10  S001  D10          0             0  \n",
       "...             ...       ...   ...  ...        ...           ...  \n",
       "25013355        0.0  S650_D42  S650  D42          0             0  \n",
       "25013356        0.0  S650_D42  S650  D42          0             0  \n",
       "25013357        0.0  S650_D42  S650  D42          0             0  \n",
       "25013358        0.0  S650_D42  S650  D42          0             0  \n",
       "25013359        0.0  S650_D42  S650  D42          0             0  \n",
       "\n",
       "[25013360 rows x 12 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_samples = df_data_all[\n",
    "    [not bool(operation_re.search(x)) for x in df_data_all[\"model\"]]\n",
    "].reset_index(drop=True)\n",
    "df_data_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98aa2ba8-3d0e-4e90-b79d-f46cf5940a2a",
   "metadata": {},
   "source": [
    "### Combine sample results based on donor genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e7e4e29e-14bc-4b7e-9642-df0ec2cf7b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_key = \"donor\"\n",
    "groupby_operation = \"mean\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c16ba1d1-7210-4800-897a-02f54455e3ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reactions</th>\n",
       "      <th>donor</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>G6PD_V68M</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>S001</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>S001</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>S001</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>S001</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>S002</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8512395</th>\n",
       "      <td>ZN2t</td>\n",
       "      <td>S649</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8512396</th>\n",
       "      <td>ZN2t</td>\n",
       "      <td>S650</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8512397</th>\n",
       "      <td>ZN2t</td>\n",
       "      <td>S650</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8512398</th>\n",
       "      <td>ZN2t</td>\n",
       "      <td>S650</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8512399</th>\n",
       "      <td>ZN2t</td>\n",
       "      <td>S650</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8355248 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         reactions donor  optimum  min  max  range  abundance  G6PD_V68M  \\\n",
       "0        13DAMPPOX  S001     0.00  0.0  0.0    0.0        NaN          0   \n",
       "1        13DAMPPOX  S001     0.50  0.0  0.0    0.0        NaN          0   \n",
       "2        13DAMPPOX  S001     0.90  0.0  0.0    0.0        NaN          0   \n",
       "3        13DAMPPOX  S001     0.99  0.0  0.0    0.0        NaN          0   \n",
       "4        13DAMPPOX  S002     0.00  0.0  0.0    0.0        NaN          0   \n",
       "...            ...   ...      ...  ...  ...    ...        ...        ...   \n",
       "8512395       ZN2t  S649     0.99  0.0  0.0    0.0        0.0          0   \n",
       "8512396       ZN2t  S650     0.00  0.0  0.0    0.0        0.0          0   \n",
       "8512397       ZN2t  S650     0.50  0.0  0.0    0.0        0.0          0   \n",
       "8512398       ZN2t  S650     0.90  0.0  0.0    0.0        0.0          0   \n",
       "8512399       ZN2t  S650     0.99  0.0  0.0    0.0        0.0          0   \n",
       "\n",
       "         ATP11C_V972M  \n",
       "0                   0  \n",
       "1                   0  \n",
       "2                   0  \n",
       "3                   0  \n",
       "4                   0  \n",
       "...               ...  \n",
       "8512395             0  \n",
       "8512396             0  \n",
       "8512397             0  \n",
       "8512398             0  \n",
       "8512399             0  \n",
       "\n",
       "[8355248 rows x 9 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if groupby_operation == \"mean\":\n",
    "    df_data_for_analyses = df_data_samples.groupby(\n",
    "        [\"reactions\", id_key, \"optimum\"], as_index=False\n",
    "    )[[\"min\", \"max\", \"range\", \"abundance\"] + genotypes].mean()\n",
    "elif groupby_operation == \"median\":\n",
    "\n",
    "    df_data_for_analyses = df_data_samples.groupby(\n",
    "        [\"reactions\", id_key, \"optimum\"], as_index=False\n",
    "    )[[\"min\", \"max\", \"range\", \"abundance\"] + genotypes].median()\n",
    "else:\n",
    "    raise ValueError(f\"Unrecognized operation to perform: '{groupby_operation}'\")\n",
    "df_data_for_analyses = df_data_for_analyses.dropna(subset=genotypes, axis=0, how=\"any\")\n",
    "df_data_for_analyses[genotypes] = df_data_for_analyses[genotypes].astype(int)\n",
    "\n",
    "df_metadata = (\n",
    "    df_data_for_analyses[[id_key] + genotypes].drop_duplicates().set_index(id_key)\n",
    ")\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7f4389-594f-4c01-8743-94cbef5ffe54",
   "metadata": {},
   "source": [
    "### Create groups of models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cf91c097-af01-425f-a72e-9c3a5df6450b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_key = \"ALL\"\n",
    "model_groups = {all_key: list(df_data_for_analyses[id_key].unique())}\n",
    "\n",
    "\n",
    "def create_group_of_models(df, id_key, groupby, verbose=False):\n",
    "    grouped = df.groupby(groupby)[id_key].agg(lambda x: list(x.unique()))\n",
    "    grouped = {\n",
    "        \"_\".join([str(x) for x in ensure_iterable(k)]): v\n",
    "        for k, v in grouped.to_dict().items()\n",
    "    }\n",
    "    if verbose:\n",
    "        max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "        for group_name, model_list in grouped.items():\n",
    "            spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "            print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "    return grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c290b56e-0285-4f7a-84b1-0c6376dc545d",
   "metadata": {},
   "source": [
    "#### Based on genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bd8b83f1-abda-4ade-b5bd-098b6885dd13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G6PD_V68M_0:\t620 samples\n",
      "G6PD_V68M_1:\t7 samples\n",
      "G6PD_V68M_2:\t11 samples\n",
      "\n",
      "ATP11C_V972M_0:\t624 samples\n",
      "ATP11C_V972M_1:\t12 samples\n",
      "ATP11C_V972M_2:\t2 samples\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for genotype in genotypes:\n",
    "    grouped = create_group_of_models(\n",
    "        df_data_for_analyses, id_key=id_key, groupby=genotype, verbose=False\n",
    "    )\n",
    "    grouped = {f\"{genotype}_{str(k)}\": v for k, v in grouped.items()}\n",
    "    if verbose:\n",
    "        max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "        for group_name, model_list in grouped.items():\n",
    "            spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "            print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "    model_groups.update(grouped)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f798fa7e-2a9a-4664-afbe-b30fd94b1df0",
   "metadata": {},
   "source": [
    "### View groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a1152679-d912-4768-802c-5d365e05c481",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Possible groups for analyses\n",
      "============================\n",
      "ALL:           \t638 samples\n",
      "G6PD_V68M_0:   \t620 samples\n",
      "G6PD_V68M_1:   \t7 samples\n",
      "G6PD_V68M_2:   \t11 samples\n",
      "ATP11C_V972M_0:\t624 samples\n",
      "ATP11C_V972M_1:\t12 samples\n",
      "ATP11C_V972M_2:\t2 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>G6PD_V68M</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th>donor</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">13DAMPPOX</th>\n",
       "      <th>S001</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S001</th>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S002</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">ZN2t</th>\n",
       "      <th>S649</th>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S650</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S650</th>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S650</th>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S650</th>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8355248 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 optimum  min  max  range  abundance  G6PD_V68M  ATP11C_V972M\n",
       "reactions donor                                                              \n",
       "13DAMPPOX S001      0.00  0.0  0.0    0.0        NaN          0             0\n",
       "          S001      0.50  0.0  0.0    0.0        NaN          0             0\n",
       "          S001      0.90  0.0  0.0    0.0        NaN          0             0\n",
       "          S001      0.99  0.0  0.0    0.0        NaN          0             0\n",
       "          S002      0.00  0.0  0.0    0.0        NaN          0             0\n",
       "...                  ...  ...  ...    ...        ...        ...           ...\n",
       "ZN2t      S649      0.99  0.0  0.0    0.0        0.0          0             0\n",
       "          S650      0.00  0.0  0.0    0.0        0.0          0             0\n",
       "          S650      0.50  0.0  0.0    0.0        0.0          0             0\n",
       "          S650      0.90  0.0  0.0    0.0        0.0          0             0\n",
       "          S650      0.99  0.0  0.0    0.0        0.0          0             0\n",
       "\n",
       "[8355248 rows x 7 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Possible groups for analyses\\n============================\")\n",
    "max_name_len = max([len(group_name) for group_name in list(model_groups)])\n",
    "for group_name, model_list in model_groups.items():\n",
    "    spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "    print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "\n",
    "df_data_for_analyses = df_data_for_analyses.set_index([\"reactions\", id_key])\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2684b9-37c0-462b-98a1-5d20012b4e93",
   "metadata": {},
   "source": [
    "#### Ensure groups exist and setup directory structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e89c9396-815f-450e-a20c-d58be95ac66d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected directory structure\n",
      "============================\n",
      "ALL\n",
      "└── ATP11C_V972M\n",
      "└── G6PD_V68M\n"
     ]
    }
   ],
   "source": [
    "groups_dict = defaultdict(dict)\n",
    "item_list = genotypes\n",
    "groups_dict[all_key].update({item: {} for item in item_list})\n",
    "\n",
    "header = \"Expected directory structure\"\n",
    "print(\"\\n\".join((header, \"=\" * len(header), all_key)))\n",
    "for idx, (group_name, subgroups) in enumerate(sorted(groups_dict[all_key].items())):\n",
    "    print(\"\\u2514\\u2500\\u2500\" + f\" {group_name}\")\n",
    "\n",
    "group_results_dirpath_dict = {all_key: corr_results_dirpath}\n",
    "for group_name, subgroups in groups_dict[all_key].items():\n",
    "    group_results_dirpath_dict[group_name] = (\n",
    "        group_results_dirpath_dict[all_key] / group_name\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8fc2e6-5be8-4ac8-98ad-92d69335d316",
   "metadata": {},
   "source": [
    "#### Load subsystems and metabolic categories to enrich results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f7fff7e1-e706-4abf-adc9-6ee5a72bfd5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subsystem</th>\n",
       "      <th>category</th>\n",
       "      <th>kegg.pathway.name</th>\n",
       "      <th>kegg.pathway</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>hsa00250</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>hsa00330</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>hsa00270</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>hsa00260</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>hsa00340</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>Other</td>\n",
       "      <td>Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>has00970</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Representative subsystem for all transport rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>hsa00983</td>\n",
       "      <td>Subnetwork of KEGG pathway \"Drug metabolism - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Azathioprine and 6-mercaptopurine metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>hsa00983</td>\n",
       "      <td>Subnetwork of KEGG pathway \"Drug metabolism - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Ribavirin metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Curated category, from reactome R-HSA-9755088,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subsystem               category  \\\n",
       "0    Alanine, aspartate and glutamate metabolism  Amino acid metabolism   \n",
       "1                Arginine and proline metabolism  Amino acid metabolism   \n",
       "2             Cysteine and methionine metabolism  Amino acid metabolism   \n",
       "3       Glycine, serine and threonine metabolism  Amino acid metabolism   \n",
       "4                           Histidine metabolism  Amino acid metabolism   \n",
       "..                                           ...                    ...   \n",
       "73                   Aminoacyl-tRNA biosynthesis                  Other   \n",
       "74                      Transport, extracellular    Transport reactions   \n",
       "75                     5-fluorouracil metabolism                  Other   \n",
       "76  Azathioprine and 6-mercaptopurine metabolism                  Other   \n",
       "77                          Ribavirin metabolism                  Other   \n",
       "\n",
       "                              kegg.pathway.name kegg.pathway  \\\n",
       "0   Alanine, aspartate and glutamate metabolism     hsa00250   \n",
       "1               Arginine and proline metabolism     hsa00330   \n",
       "2            Cysteine and methionine metabolism     hsa00270   \n",
       "3      Glycine, serine and threonine metabolism     hsa00260   \n",
       "4                          Histidine metabolism     hsa00340   \n",
       "..                                          ...          ...   \n",
       "73                  Aminoacyl-tRNA biosynthesis     has00970   \n",
       "74                                                             \n",
       "75              Drug metabolism - other enzymes     hsa00983   \n",
       "76              Drug metabolism - other enzymes     hsa00983   \n",
       "77                                                             \n",
       "\n",
       "                                                notes  \n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "..                                                ...  \n",
       "73                                                     \n",
       "74  Representative subsystem for all transport rea...  \n",
       "75  Subnetwork of KEGG pathway \"Drug metabolism - ...  \n",
       "76  Subnetwork of KEGG pathway \"Drug metabolism - ...  \n",
       "77  Curated category, from reactome R-HSA-9755088,...  \n",
       "\n",
       "[77 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystems_to_exclude = {\"Pseudoreactions\"}\n",
    "use_abbrevs = True\n",
    "abbreviations = {\n",
    "    \"Amino acid metabolism\": \"A\",\n",
    "    \"Carbohydrate metabolism\": \"C\",\n",
    "    \"Lipid metabolism\": \"L\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"V\",\n",
    "    \"Nucleotide metabolism\": \"N\",\n",
    "    \"Reactive species\": \"R\",\n",
    "    \"Transport reactions\": \"T\",\n",
    "    \"Other\": \"O\",\n",
    "}\n",
    "categories_to_keep = list(abbreviations)\n",
    "\n",
    "df_pathways = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / \"subsystems.tsv\", sep=\"\\t\", dtype=str\n",
    ").fillna(\"\")\n",
    "\n",
    "# Rename \"name\" to subsystem to match reaction attribute\n",
    "df_pathways = df_pathways.rename({\"name\": \"subsystem\"}, axis=1)\n",
    "# Group \"Metabolism of other amino acids\" with amino acids rather than treat as \"other\"\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].replace(\n",
    "    \"Metabolism of other amino acids\", \"Amino acid metabolism\"\n",
    ")\n",
    "\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].apply(\n",
    "    lambda x: (\"Other\" if x not in categories_to_keep else x)\n",
    ")\n",
    "df_pathways = df_pathways[~df_pathways[\"subsystem\"].isin(subsystems_to_exclude)].copy()\n",
    "subsystem_to_category_dict = df_pathways.set_index(\"subsystem\")[\"category\"].to_dict()\n",
    "df_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1281f9fb-3124-4f6b-95c6-ccc02b3c6733",
   "metadata": {},
   "source": [
    "## Compute significant results between groups\n",
    "#### Compare all subgroups at once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f5bffd48-9cbd-41c7-9994-7128a2878d68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Groups to compare\n",
      "=================\n",
      "('G6PD_V68M_0', 'G6PD_V68M_1', 'G6PD_V68M_2')\n",
      "('G6PD_V68M_0', 'G6PD_V68M_1')\n",
      "('G6PD_V68M_0', 'G6PD_V68M_2')\n",
      "('G6PD_V68M_1', 'G6PD_V68M_2')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>G6PD_V68M</th>\n",
       "      <th>ATP11C_V972M</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th>donor</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13DAMPPOX</th>\n",
       "      <th>S001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE1Ry</th>\n",
       "      <th>S001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE2Ry</th>\n",
       "      <th>S001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE3Ry</th>\n",
       "      <th>S001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGF1Ry</th>\n",
       "      <th>S001</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YYYTPAP</th>\n",
       "      <th>S605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.312651</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2Htex2</th>\n",
       "      <th>S605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.028201</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2_2HCO3t</th>\n",
       "      <th>S605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2_HCO3_SELNIt</th>\n",
       "      <th>S605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2t</th>\n",
       "      <th>S605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2088812 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       min  max  range  abundance  G6PD_V68M  ATP11C_V972M\n",
       "reactions       donor                                                     \n",
       "13DAMPPOX       S001   0.0  0.0    0.0        NaN          0             0\n",
       "15KPGE1Ry       S001   0.0  0.0    0.0   0.000000          0             0\n",
       "15KPGE2Ry       S001   0.0  0.0    0.0   0.000000          0             0\n",
       "15KPGE3Ry       S001   0.0  0.0    0.0   0.000000          0             0\n",
       "15KPGF1Ry       S001   0.0  0.0    0.0   0.000000          0             0\n",
       "...                    ...  ...    ...        ...        ...           ...\n",
       "YYYTPAP         S605   0.0  0.0    0.0   0.312651          2             0\n",
       "ZN2Htex2        S605   0.0  0.0    0.0   0.028201          2             0\n",
       "ZN2_2HCO3t      S605   0.0  0.0    0.0   0.000000          2             0\n",
       "ZN2_HCO3_SELNIt S605   0.0  0.0    0.0   0.000000          2             0\n",
       "ZN2t            S605   0.0  0.0    0.0   0.000000          2             0\n",
       "\n",
       "[2088812 rows x 6 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_name = \"G6PD_V68M\"\n",
    "optimum = 0\n",
    "value_to_compare = \"range\"\n",
    "group_timepoints_by = None\n",
    "compare_pairwise = True\n",
    "compare_all_groups = True\n",
    "ordered_group_to_compare = [f\"{group_name}_{alleles}\" for alleles in [0, 1, 2]]\n",
    "all_samples_for_comparison = [\n",
    "    value for g in ordered_group_to_compare for value in np.array(model_groups[g])\n",
    "]\n",
    "df_data_for_correlations = df_data_for_analyses.loc[\n",
    "    pd.IndexSlice[:, all_samples_for_comparison], :\n",
    "]\n",
    "df_data_for_correlations = df_data_for_correlations[\n",
    "    df_data_for_correlations[\"optimum\"] == optimum\n",
    "].drop(\"optimum\", axis=1)\n",
    "\n",
    "print(\"Groups to compare\\n=================\")\n",
    "pairwise_group_combos = []\n",
    "if compare_all_groups:\n",
    "    print(tuple(ordered_group_to_compare))\n",
    "if compare_pairwise:\n",
    "    pairwise_group_combos += list(combinations(ordered_group_to_compare, 2))\n",
    "    for group in pairwise_group_combos:\n",
    "        print(group)\n",
    "df_data_for_correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bba2921-371b-4f2a-83ac-52eab0b6ae13",
   "metadata": {},
   "source": [
    "### Kruskal Wallis H-test (3 or more groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2e98727-a180-40bf-a3bc-0795a3c4ae10",
   "metadata": {},
   "outputs": [],
   "source": [
    "ordered_group_to_compare = [\"G6PD_V68M_0\", \"G6PD_V68M_1\", \"G6PD_V68M_2\"]\n",
    "results_dict = defaultdict(dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba5742b8-10fe-42da-a85f-2df06d9dfc45",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if len(ordered_group_to_compare) > 2:\n",
    "    for rid in df_data_for_correlations.index.get_level_values(\"reactions\").unique():\n",
    "        df_data_rxn = df_data_for_correlations.loc[rid]\n",
    "        df_data_rxn_opt_value = df_data_rxn[value_to_compare].copy()\n",
    "        data_arrays = {\n",
    "            group_name: df_data_rxn_opt_value.loc[model_groups[group_name]].values\n",
    "            for group_name in ordered_group_to_compare\n",
    "        }\n",
    "        values = list(data_arrays.values())\n",
    "        unique_values = set(\n",
    "            [v for value_list in values for v in value_list if not np.isnan(v)]\n",
    "        )\n",
    "        if len(unique_values) <= 1:\n",
    "            # Skip variables that do not have any differences\n",
    "            results_dict[tuple(ordered_group_to_compare)][rid] = dict(\n",
    "                zip([\"statistic\", \"pvalue\"], [pd.NA, pd.NA])\n",
    "            )\n",
    "        else:\n",
    "            result = kruskal(*values, nan_policy=\"omit\")\n",
    "            results_dict[tuple(ordered_group_to_compare)][rid] = {\n",
    "                attr: getattr(result, attr) for attr in [\"statistic\", \"pvalue\"]\n",
    "            }\n",
    "\n",
    "dataframes = {\n",
    "    key: pd.DataFrame.from_dict(values, orient=\"index\")\n",
    "    for key, values in results_dict.items()\n",
    "}\n",
    "print(f\"Number of different comparisons made: {len(dataframes)}\")\n",
    "print(\"Groups compared\\n===============\")\n",
    "for key in list(dataframes):\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d05f036-101b-4e2f-9301-e71c65e81f67",
   "metadata": {},
   "source": [
    "### Mann Whiteney U test (2 groups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8136fe70-1abd-4b17-a351-c5ca42a969be",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(ordered_group_to_compare) == 2 or compare_pairwise:\n",
    "    for rid in df_data_for_correlations.index.get_level_values(\"reactions\").unique():\n",
    "        df_data_rxn = df_data_for_correlations.loc[rid]\n",
    "        df_data_rxn_opt_value = df_data_rxn[value_to_compare].copy()\n",
    "        data_arrays = {\n",
    "            group_name: df_data_rxn_opt_value.loc[model_groups[group_name]].values\n",
    "            for group_name in ordered_group_to_compare\n",
    "        }\n",
    "        combos = (\n",
    "            [tuple(ordered_group_to_compare)]\n",
    "            if not pairwise_group_combos\n",
    "            else pairwise_group_combos\n",
    "        )\n",
    "        for combo in combos:\n",
    "            values = [data_arrays[group] for group in combo]\n",
    "            unique_values = set(\n",
    "                [v for value_list in values for v in value_list if not np.isnan(v)]\n",
    "            )\n",
    "            if len(unique_values) <= 1:\n",
    "                # Skip variables that do not have any differences\n",
    "                result = dict(zip([\"statistic\", \"pvalue\"], [pd.NA, pd.NA]))\n",
    "            else:\n",
    "                result = mannwhitneyu(*values, nan_policy=\"omit\")\n",
    "                result = {\n",
    "                    attr: getattr(result, attr) for attr in [\"statistic\", \"pvalue\"]\n",
    "                }\n",
    "            results_dict[combo][rid] = result\n",
    "dataframes = {\n",
    "    key: pd.DataFrame.from_dict(values, orient=\"index\")\n",
    "    for key, values in results_dict.items()\n",
    "}\n",
    "print(f\"Number of different comparisons made: {len(dataframes)}\")\n",
    "print(\"Groups compared\\n===============\")\n",
    "for key in list(dataframes):\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a16a3b4-ff30-4ede-a251-5ef488ed13ab",
   "metadata": {},
   "source": [
    "### Determine significance using p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0864a7a5-d2c0-4b80-addf-830f9f36a424",
   "metadata": {},
   "outputs": [],
   "source": [
    "pvalue_sig = 0.0002\n",
    "enzyme_reactions_only = False\n",
    "include_boundary_reactions = False\n",
    "sort_by_subsystem = False\n",
    "standardize_by = \"mean\"\n",
    "use_group_means = False\n",
    "fdr_method = None\n",
    "\n",
    "significant_dataframes = {}\n",
    "for met in model.metabolites.query(lambda x: x.compartment == \"e\"):\n",
    "    met.name += \" (extracellular)\"\n",
    "metadata_columns = [\n",
    "    \"name\",\n",
    "    \"stoichiometry\",\n",
    "    \"proteins\",\n",
    "    \"pvalue\" if not fdr_method else \"adj_pvalue\",\n",
    "    \"subsystem\",\n",
    "    \"category\",\n",
    "]\n",
    "for key, df in dataframes.items():\n",
    "    df = df.dropna().copy()\n",
    "    df[\"pvalue\"] = df[\"pvalue\"].astype(float)\n",
    "    if fdr_method is not None and not fdr_method in {\"bon\", \"bh\", \"by\"}:\n",
    "        raise ValueError(f\"Unrecognized FDR correction method : {fdr_method}\")\n",
    "    elif fdr_method == \"bon\":\n",
    "        pvalue_key = \"adj_pvalue\"\n",
    "        df[pvalue_key] = df[\"pvalue\"] * len(df[\"pvalue\"])\n",
    "    elif fdr_method in {\"bh\", \"by\"}:\n",
    "        pvalue_key = \"adj_pvalue\"\n",
    "        df[pvalue_key] = false_discovery_control(\n",
    "            df[\"pvalue\"].astype(float), method=fdr_method\n",
    "        )\n",
    "    else:\n",
    "        pvalue_key = \"pvalue\"\n",
    "\n",
    "    df = df[df[pvalue_key] <= pvalue_sig].drop(\"statistic\", axis=1)\n",
    "    if enzyme_reactions_only:\n",
    "        df_pivot = df_data_for_correlations.loc[\n",
    "            df.index, [\"abundance\", value_to_compare]\n",
    "        ].dropna(subset=\"abundance\")\n",
    "        df_pivot = df_pivot.drop(\"abundance\", axis=1)\n",
    "    else:\n",
    "        df_pivot = df_data_for_correlations.loc[df.index, value_to_compare]\n",
    "    if not include_boundary_reactions:\n",
    "        df_pivot = df_pivot[\n",
    "            ~df_pivot.index.isin(\n",
    "                model.reactions.query(lambda x: x.boundary).list_attr(\"id\"),\n",
    "                level=\"reactions\",\n",
    "            )\n",
    "        ]\n",
    "    df_pivot = df_pivot.reset_index(drop=False)\n",
    "    df_pivot = df_pivot.pivot(\n",
    "        columns=id_key, index=\"reactions\", values=value_to_compare\n",
    "    )\n",
    "    df = pd.merge(df, df_pivot, left_index=True, right_index=True).sort_values(\n",
    "        pvalue_key\n",
    "    )\n",
    "    df.index.name = \"reactions\"\n",
    "    df = df.reset_index(drop=False).set_index([\"reactions\", pvalue_key]).T\n",
    "    if df.empty:\n",
    "        df = pd.DataFrame([], columns=metadata_columns)\n",
    "    else:\n",
    "        df = pd.concat(\n",
    "            [\n",
    "                # Sort index by donor number and subgroup while concatenating\n",
    "                df.loc[model_groups[g]].sort_index(\n",
    "                    key=lambda x: [\n",
    "                        int(donor_re.search(v).group(\"num\")) for v in x.values\n",
    "                    ]\n",
    "                )\n",
    "                for g in key\n",
    "            ],\n",
    "            axis=0,\n",
    "        )\n",
    "\n",
    "        df = df.T.reset_index(drop=False)\n",
    "        # Enrich results\n",
    "        df[\"name\"] = [\n",
    "            r.name.replace(\",\", \"\")\n",
    "            for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "        ]\n",
    "        df[\"stoichiometry\"] = [\n",
    "            r.build_reaction_string(use_metabolite_names=True)\n",
    "            for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "        ]\n",
    "        df[\"subsystem\"] = [\n",
    "            r.subsystem\n",
    "            for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "        ]\n",
    "        df[\"category\"] = df[\"subsystem\"].replace(subsystem_to_category_dict)\n",
    "        df[\"proteins\"] = [\n",
    "            \";\".join(sorted([g.id for g in r.genes]))\n",
    "            for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "        ]\n",
    "        # Replace commas to prevent issues with CSV export\n",
    "        df[\"subsystem\"] = df[\"subsystem\"].apply(lambda x: x.replace(\",\", \"\"))\n",
    "        df[\"category\"] = df[\"category\"].apply(lambda x: x.replace(\",\", \"\"))\n",
    "        df[pvalue_key] = df[pvalue_key].apply(lambda x: round(x, 5))\n",
    "\n",
    "        df = df.set_index(\"reactions\")\n",
    "        if sort_by_subsystem:\n",
    "            df = df.sort_values(by=[\"category\", \"subsystem\", \"proteins\"])\n",
    "\n",
    "        df_meta = df.loc[:, metadata_columns].copy()\n",
    "        df_data = df.loc[:, ~df.columns.isin(df_meta.columns)].copy()\n",
    "        if use_group_means:\n",
    "            df_data = pd.concat(\n",
    "                [df_data.loc[:, model_groups[g]].mean(axis=1) for g in key], axis=1\n",
    "            )\n",
    "            df_data.columns = list(key)\n",
    "\n",
    "        if standardize_by == \"mean\":\n",
    "            df_data = (\n",
    "                df_data.sub(df_data.mean(axis=1), axis=0)\n",
    "                .div(df_data.std(axis=1), axis=0)\n",
    "                .dropna(how=\"all\", axis=0)\n",
    "            )\n",
    "        elif standardize_by == \"median\":\n",
    "            df_data = (\n",
    "                (df_data.T - df_data.median(axis=1))\n",
    "                / (df_data.quantile(q=0.75, axis=1) - df_data.quantile(q=0.25, axis=1))\n",
    "            ).T\n",
    "        else:\n",
    "            pass\n",
    "        # Put dataframes back together for custom reordering\n",
    "        df = df_data.merge(df_meta, left_index=True, right_index=True)\n",
    "    significant_dataframes[key] = df\n",
    "    print(key)\n",
    "    print(f\"Min & Max values: ({df_data.min().min():.4f}, {df_data.max().max():.4f})\")\n",
    "    print()\n",
    "key = tuple(ordered_group_to_compare)\n",
    "df = significant_dataframes[key]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53b82606-e181-4c1d-8ce4-db067ae04c53",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = significant_dataframes[key].copy()\n",
    "# df = df.loc[:, ~df.columns.isin([\"proteins\", \"pvalue\", \"subsystem\", \"category\"])].T.merge(df_metadata[[\"G6PD_V68M\"]], left_index=True, right_index=True).reset_index(drop=False)\n",
    "# df = df.set_index([\"index\", \"G6PD_V68M\"]).groupby(level=1).mean().T\n",
    "# print(df.min().min(), df.max().max())\n",
    "# df = df.merge(df_meta, left_index=True, right_index=True)\n",
    "# ordered_subsystems = [\n",
    "#     \"Pentose phosphate pathway\",\n",
    "#     \"Glycolysis / Gluconeogenesis\",\n",
    "#     \"Galactose metabolism\",\n",
    "#     \"Transport extracellular\",\n",
    "#     \"Glycerophospholipid metabolism\",\n",
    "#     \"Ether lipid metabolism\",\n",
    "# ]\n",
    "# df = pd.concat(\n",
    "#     [\n",
    "#         df[df[\"subsystem\"] == subsystem].sort_values(by=\"pvalue\", ascending=False if subsystem == \"Glycolysis / Gluconeogenesis\" else True)\n",
    "#         for subsystem in ordered_subsystems\n",
    "#     ],\n",
    "#     axis=0,\n",
    "# )\n",
    "# significant_dataframes[key] = df.copy()\n",
    "# df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610dc29f-f123-457b-8d29-0d7461a189f7",
   "metadata": {},
   "source": [
    "## Export results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6cd72da-0f5b-47c3-bdf4-0cfdd3c54bd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"csv\"\n",
    "for key, df_main in significant_dataframes.items():\n",
    "    df_meta = df_main.loc[:, metadata_columns].copy()\n",
    "    df_data = df_main.loc[:, ~df_main.columns.isin(df_meta.columns)].copy()\n",
    "    for df_type, df in zip(\n",
    "        [\"data\", \"meta_row\", \"meta_col\"], [df_data, df_meta, df_metadata]\n",
    "    ):\n",
    "        filename = \"_\".join(\n",
    "            [\"MannWhiteney\" if len(key) == 2 else \"Kruskal\"]\n",
    "            + [g.split(\"_\")[-1] for g in key]\n",
    "            + [id_key]\n",
    "            + [df_type]\n",
    "        )\n",
    "        if use_group_means:\n",
    "            filename += \"_mean\"\n",
    "        group_results_dirpath_dict[group_name].mkdir(exist_ok=True)\n",
    "        filename = group_results_dirpath_dict[group_name] / filename\n",
    "        df.to_csv(\n",
    "            f\"{filename}.{ftype}\", sep=\"\\t\" if ftype == \"tsv\" else \",\", index=True\n",
    "        )\n",
    "    break"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
