{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fc9114c6-b2d1-48a1-8ba3-6686a1ea2449",
   "metadata": {},
   "source": [
    "# Compute statistically significant fluxes between groups - Mouse G6PD variants omics data\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2d694e76-31c0-4aba-add6-a07a00a5d04c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.3\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.13.4\n",
      "bio                                   1.8.0\n",
      "cobra                                0.29.1\n",
      "depinfo                               2.2.0\n",
      "gurobipy                             12.0.3\n",
      "matplotlib                           3.10.3\n",
      "matplotlib-venn                       1.1.2\n",
      "memote                               0.17.0\n",
      "networkx                                3.5\n",
      "notebook                              7.4.4\n",
      "openpyxl                              3.1.5\n",
      "pandas                                2.3.1\n",
      "pre-commit                            4.2.0\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.32.4\n",
      "scikit-learn                          1.7.0\n",
      "scipy                                1.16.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip          25.1\n",
      "setuptools 78.1.1\n",
      "wheel      0.45.1\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Windows 10-AMD64\n",
      "CPython  3.11.13\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import textwrap\n",
    "import warnings\n",
    "from collections import defaultdict\n",
    "from itertools import combinations\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    ensure_iterable,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    ")\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    DEFAULT_PREFIX_SUFFIX_VALUES,\n",
    "    DEFAULT_PROTEOME_COMPARTMENT,\n",
    "    EnzymeDilution,\n",
    "    add_relaxation_budget,\n",
    "    load_overlay_model,\n",
    "    plot_correlations,\n",
    ")\n",
    "from rbc_gem_utils.visualization import cmap_map\n",
    "from scipy.stats import kruskal, mannwhitneyu, spearmanr\n",
    "\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "384c972a-9f2f-4b88-a65c-017d1a6561d6",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "418ddc9c-b48e-400a-b5a1-e4e6bfe73bd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>100000000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1e-08\n",
       "upper_bound: 100000000.0\n",
       "processes: 127\n",
       "cache_directory: C:\\Users\\P7875\\AppData\\Local\\opencobra\\cobrapy\\Cache\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "# Set bound defaults much larger to prevent model loading issues\n",
    "COBRA_CONFIGURATION.bounds = (-1e-8, 1e8)\n",
    "COBRA_CONFIGURATION.tolerance = 1e-7\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3c3720a-d637-4115-8960-f48671ee725a",
   "metadata": {},
   "source": [
    "### Define organism, model, and dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5b08faa-240f-44ee-9d16-ddbdc556852f",
   "metadata": {},
   "outputs": [],
   "source": [
    "organism = \"Mouse\"\n",
    "model_id = \"RBC_GEM\"\n",
    "dataset_name = \"G6PDvariants\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c0ccb7e-c784-4b63-bfe9-7ebdce0e24b5",
   "metadata": {},
   "source": [
    "### Set variables for sample identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6435bfd2-9c1f-4f09-8a7a-3d0a0efba741",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For sample IDs\n",
    "timepoints = [\"Pre\", \"Post\", \"TD\"]\n",
    "phenotypes = [\"HumCan\", \"A\", \"MED\"]\n",
    "donor_re = re.compile(rf\"(?P<donor>({'|'.join(phenotypes)})(?P<num>\\d+))\")\n",
    "time_re = re.compile(rf\"(?P<time>{'|'.join(timepoints)})\")\n",
    "phenotype_re = re.compile(rf\"(?P<phenotype>({'|'.join(phenotypes)}))\")\n",
    "\n",
    "operations = \"|\".join([x.capitalize() for x in [\"mean\", \"median\"]])\n",
    "\n",
    "operation_re = re.compile(r\"(?P<op>\" + operations + r\")\\_(?P<group>\\w+)\")\n",
    "sample_id_re = re.compile(\n",
    "    r\"(?!\" + operations + r\")\" + donor_re.pattern + r\"\\_\" + time_re.pattern\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e478c85d-b0ed-466c-b6d3-e8dead52f018",
   "metadata": {},
   "source": [
    "### Set computation options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "52e59494-b23d-4775-8863-1baeb1f46194",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_computations = True\n",
    "verbose = True\n",
    "objective_reactions = [\"NaKt\"]\n",
    "\n",
    "enzyme_rxn_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.dilution\"]\n",
    "enzyme_met_prefix = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"prefix.metabolite\"]\n",
    "enzyme_met_suffix_total = DEFAULT_PREFIX_SUFFIX_VALUES[\"enzymes\"][\"suffix.total\"]\n",
    "comp_suffix = f\"_{DEFAULT_PROTEOME_COMPARTMENT}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "295bb5e1-67d6-4436-af4a-0f34befaa111",
   "metadata": {},
   "source": [
    "### Set figure options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "06ad3312-69a9-4bd9-92c4-a2d467bed985",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfc2d514-cb81-4968-95b6-95e758cd3e73",
   "metadata": {},
   "source": [
    "### Set paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0ccd462b-0c65-4b33-b745-6a4e74996f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set paths\n",
    "processed_data_dirpath = get_dirpath(use_temp=\"processed\") / organism / dataset_name\n",
    "\n",
    "overlay_dirpath = get_dirpath(\"analysis\") / \"OVERLAY\" / organism\n",
    "model_dirpath = overlay_dirpath / model_id\n",
    "results_dirpath = (\n",
    "    get_dirpath(use_temp=\"processed\") / model_id / \"OVERLAY\" / organism / dataset_name\n",
    ")\n",
    "\n",
    "sample_pcmodels_dirpath = results_dirpath / \"sample_pcmodels\"\n",
    "pcfva_results_dirpath = (\n",
    "    results_dirpath / \"pcFVA\" / \"_\".join((\"OBJ\", *objective_reactions))\n",
    ")\n",
    "# Objective reaction does not matter since correlations are computed\n",
    "# based on min and max fluxes and abundance, which are obtained when optimum is 0.\n",
    "corr_results_dirpath = results_dirpath / \"correlations\"\n",
    "# Ensure directory  exists\n",
    "corr_results_dirpath.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "806b8d0f-8ca2-48e4-9da1-7db464f01407",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "541f7b2c-ad0d-4404-899b-fa05a0bd7fa1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Set parameter LicenseID to value 2664191\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>268a967c850</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>10403</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>19806</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>826</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC at 0x268a967c850>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_cobra_model(filename=model_dirpath / f\"{model_id}.xml\")\n",
    "pcmodel = load_overlay_model(filename=model_dirpath / f\"{model_id}_PC.xml\")\n",
    "\n",
    "# Add relaxation budget to initial PC model to get names of relaxation reactions\n",
    "add_relaxation_budget(pcmodel, 0, verbose=False)\n",
    "pcmodel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49cf0307-bf28-4f10-8784-143d62c669a7",
   "metadata": {},
   "source": [
    "## Load pcFVA generated results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "537fccd8-52b6-44c9-83a0-af5c62b7fd2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.424793e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.054404e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.054404e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-2.054404e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3552995</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3552996</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3552997</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3552998</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3552999</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3553000 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            model        reactions  optimum  min           max\n",
       "0             RBC_GEM_PC_A10_Post        13DAMPPOX     0.00  0.0 -2.424793e-09\n",
       "1             RBC_GEM_PC_A10_Post        13DAMPPOX     0.50  0.0 -2.054404e-09\n",
       "2             RBC_GEM_PC_A10_Post        13DAMPPOX     0.90  0.0 -2.054404e-09\n",
       "3             RBC_GEM_PC_A10_Post        13DAMPPOX     0.99  0.0 -2.054404e-09\n",
       "4             RBC_GEM_PC_A10_Post        15KPGE1Ry     0.00  0.0  0.000000e+00\n",
       "...                           ...              ...      ...  ...           ...\n",
       "3552995  RBC_GEM_PC_Median_TD_MED  ZN2_HCO3_SELNIt     0.99  0.0  0.000000e+00\n",
       "3552996  RBC_GEM_PC_Median_TD_MED             ZN2t     0.00  0.0  0.000000e+00\n",
       "3552997  RBC_GEM_PC_Median_TD_MED             ZN2t     0.50  0.0  0.000000e+00\n",
       "3552998  RBC_GEM_PC_Median_TD_MED             ZN2t     0.90  0.0  0.000000e+00\n",
       "3552999  RBC_GEM_PC_Median_TD_MED             ZN2t     0.99  0.0  0.000000e+00\n",
       "\n",
       "[3553000 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load DataFrame of generated results\n",
    "df_pcfva_all = pd.read_csv(\n",
    "    pcfva_results_dirpath / f\"{pcmodel.id}_All_FVAsols.csv\",\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "df_pcfva_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63a4524-c95b-4980-a95f-5ba109e75025",
   "metadata": {},
   "source": [
    "## Create DataFrame for calculations and visualizations\n",
    "### Get maximum reaction fluxes and associated abundance values\n",
    "#### Get maximum reaction fluxes and ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3efaec1d-6db8-4b96-975b-67303d06d9ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [min, max, range]\n",
       "Index: []"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxns = model.reactions.list_attr(\"id\")\n",
    "df_max_flux_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "df_max_flux_per_model = df_max_flux_per_model.groupby(\n",
    "    [\"model\", \"reactions\", \"optimum\"]\n",
    ")[[\"min\", \"max\"]].agg(\n",
    "    {\n",
    "        \"min\": \"min\",  # Minimum reaction flux per model\n",
    "        \"max\": \"max\",  # Maximum reaction flux per model\n",
    "    }\n",
    ")\n",
    "# Address issues possibly caused by floating point precision, ideally a value that prevents any negative ranges\n",
    "df_max_flux_per_model.loc[\n",
    "    df_max_flux_per_model[\"max\"] < df_max_flux_per_model[\"min\"], [\"max\", \"min\"]\n",
    "] = [0, 0]\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_flux_per_model[\"max\"] = df_max_flux_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_flux_per_model[\"min\"] = df_max_flux_per_model[\"min\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_flux_per_model[\"range\"] = (\n",
    "    df_max_flux_per_model[\"max\"] - df_max_flux_per_model[\"min\"]\n",
    ")\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_flux_per_model[df_max_flux_per_model[\"range\"] < 0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e000d2d8-7a7c-4b02-b088-867e04ce7dbe",
   "metadata": {},
   "source": [
    "#### Get maximum \"enzyme\" abundances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4d6feeea-cfe4-4da9-99db-47f2729bd4e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [abundance]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rxns = pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, EnzymeDilution)\n",
    "    and x.id.endswith(f\"{enzyme_met_suffix_total}{comp_suffix}\")\n",
    ").list_attr(\"id\")\n",
    "df_max_abundance_per_model = df_pcfva_all[df_pcfva_all[\"reactions\"].isin(rxns)].copy()\n",
    "# Rename dilution reactions to match\n",
    "reaction_enzyme_map = {\n",
    "    enzyme_rid: enzyme_rid.replace(\n",
    "        f\"{enzyme_rxn_prefix}{enzyme_met_prefix}\", \"\"\n",
    "    ).replace(f\"{enzyme_met_suffix_total}{comp_suffix}\", \"\")\n",
    "    for enzyme_rid in df_max_abundance_per_model[\"reactions\"]\n",
    "}\n",
    "df_max_abundance_per_model[\"reactions\"] = df_max_abundance_per_model[\n",
    "    \"reactions\"\n",
    "].replace(reaction_enzyme_map)\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.groupby(\n",
    "    [\"model\", \"reactions\", \"optimum\"]\n",
    ")[[\"max\"]].max()\n",
    "# Address issues possibly caused by floating point precision, atol is ideally a value that prevents any negative ranges\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if x < 0 else x\n",
    ")\n",
    "atol = COBRA_CONFIGURATION.tolerance\n",
    "df_max_abundance_per_model[\"max\"] = df_max_abundance_per_model[\"max\"].apply(\n",
    "    lambda x: 0 if np.isclose(x, 0, atol=atol) else round(x, -int(np.log10(atol)))\n",
    ")\n",
    "df_max_abundance_per_model = df_max_abundance_per_model.rename(\n",
    "    {\"max\": \"abundance\"}, axis=1\n",
    ")\n",
    "# Ensure no negative values, if results appear then tolerance should be adjusted\n",
    "df_max_abundance_per_model[(df_max_abundance_per_model < 0).any(axis=1)]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95269b45-4f16-4d09-95c5-af4734b23d82",
   "metadata": {},
   "source": [
    "#### Merge DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "60602d4b-89ba-493b-ae29-55a35c669a31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636995</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636996</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636997</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636998</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636999</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1637000 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            model        reactions  optimum  min  max  range  \\\n",
       "0             RBC_GEM_PC_A10_Post        13DAMPPOX     0.00  0.0  0.0    0.0   \n",
       "1             RBC_GEM_PC_A10_Post        13DAMPPOX     0.50  0.0  0.0    0.0   \n",
       "2             RBC_GEM_PC_A10_Post        13DAMPPOX     0.90  0.0  0.0    0.0   \n",
       "3             RBC_GEM_PC_A10_Post        13DAMPPOX     0.99  0.0  0.0    0.0   \n",
       "4             RBC_GEM_PC_A10_Post        15KPGE1Ry     0.00  0.0  0.0    0.0   \n",
       "...                           ...              ...      ...  ...  ...    ...   \n",
       "1636995  RBC_GEM_PC_Median_TD_MED  ZN2_HCO3_SELNIt     0.99  0.0  0.0    0.0   \n",
       "1636996  RBC_GEM_PC_Median_TD_MED             ZN2t     0.00  0.0  0.0    0.0   \n",
       "1636997  RBC_GEM_PC_Median_TD_MED             ZN2t     0.50  0.0  0.0    0.0   \n",
       "1636998  RBC_GEM_PC_Median_TD_MED             ZN2t     0.90  0.0  0.0    0.0   \n",
       "1636999  RBC_GEM_PC_Median_TD_MED             ZN2t     0.99  0.0  0.0    0.0   \n",
       "\n",
       "         abundance  \n",
       "0              NaN  \n",
       "1              NaN  \n",
       "2              NaN  \n",
       "3              NaN  \n",
       "4              0.0  \n",
       "...            ...  \n",
       "1636995        0.0  \n",
       "1636996        0.0  \n",
       "1636997        0.0  \n",
       "1636998        0.0  \n",
       "1636999        0.0  \n",
       "\n",
       "[1637000 rows x 7 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_all = pd.merge(\n",
    "    df_max_flux_per_model,\n",
    "    df_max_abundance_per_model,\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    ")\n",
    "df_data_all = df_data_all.reset_index(drop=False)\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3094239-cea4-4899-8cba-84b0f3969ba3",
   "metadata": {},
   "source": [
    "### Identify donor, timepoints, and phenotypes for results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "04139aab-4c0c-4845-a65f-912ac235da32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636995</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636996</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636997</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636998</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636999</th>\n",
       "      <td>RBC_GEM_PC_Median_TD_MED</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1637000 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            model        reactions  optimum  min  max  range  \\\n",
       "0             RBC_GEM_PC_A10_Post        13DAMPPOX     0.00  0.0  0.0    0.0   \n",
       "1             RBC_GEM_PC_A10_Post        13DAMPPOX     0.50  0.0  0.0    0.0   \n",
       "2             RBC_GEM_PC_A10_Post        13DAMPPOX     0.90  0.0  0.0    0.0   \n",
       "3             RBC_GEM_PC_A10_Post        13DAMPPOX     0.99  0.0  0.0    0.0   \n",
       "4             RBC_GEM_PC_A10_Post        15KPGE1Ry     0.00  0.0  0.0    0.0   \n",
       "...                           ...              ...      ...  ...  ...    ...   \n",
       "1636995  RBC_GEM_PC_Median_TD_MED  ZN2_HCO3_SELNIt     0.99  0.0  0.0    0.0   \n",
       "1636996  RBC_GEM_PC_Median_TD_MED             ZN2t     0.00  0.0  0.0    0.0   \n",
       "1636997  RBC_GEM_PC_Median_TD_MED             ZN2t     0.50  0.0  0.0    0.0   \n",
       "1636998  RBC_GEM_PC_Median_TD_MED             ZN2t     0.90  0.0  0.0    0.0   \n",
       "1636999  RBC_GEM_PC_Median_TD_MED             ZN2t     0.99  0.0  0.0    0.0   \n",
       "\n",
       "         abundance donor  time phenotype  \n",
       "0              NaN   A10  Post         A  \n",
       "1              NaN   A10  Post         A  \n",
       "2              NaN   A10  Post         A  \n",
       "3              NaN   A10  Post         A  \n",
       "4              0.0   A10  Post         A  \n",
       "...            ...   ...   ...       ...  \n",
       "1636995        0.0  <NA>    TD       MED  \n",
       "1636996        0.0  <NA>    TD       MED  \n",
       "1636997        0.0  <NA>    TD       MED  \n",
       "1636998        0.0  <NA>    TD       MED  \n",
       "1636999        0.0  <NA>    TD       MED  \n",
       "\n",
       "[1637000 rows x 10 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_columns = [\"donor\", \"time\", \"phenotype\"]\n",
    "for key, search_re in zip(metadata_columns, [donor_re, time_re, phenotype_re]):\n",
    "    df_data_all[key] = df_data_all[\"model\"].apply(\n",
    "        lambda x: search_re.search(x).group(1) if search_re.search(x) else pd.NA\n",
    "    )\n",
    "df_data_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3902ff15-bf6a-43a9-ad82-d81742231410",
   "metadata": {},
   "source": [
    "## Compute statistically significant results between groups\n",
    "### Remove models based on data operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "76ce7d4c-834d-4b2f-b705-059839846fea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>model</th>\n",
       "      <th>reactions</th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "      <th>donor</th>\n",
       "      <th>time</th>\n",
       "      <th>phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>13DAMPPOX</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RBC_GEM_PC_A10_Post</td>\n",
       "      <td>15KPGE1Ry</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A10</td>\n",
       "      <td>Post</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401267</th>\n",
       "      <td>RBC_GEM_PC_MED9_TD</td>\n",
       "      <td>ZN2_HCO3_SELNIt</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MED9</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401268</th>\n",
       "      <td>RBC_GEM_PC_MED9_TD</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MED9</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401269</th>\n",
       "      <td>RBC_GEM_PC_MED9_TD</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MED9</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401270</th>\n",
       "      <td>RBC_GEM_PC_MED9_TD</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MED9</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1401271</th>\n",
       "      <td>RBC_GEM_PC_MED9_TD</td>\n",
       "      <td>ZN2t</td>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MED9</td>\n",
       "      <td>TD</td>\n",
       "      <td>MED</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1401272 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       model        reactions  optimum  min  max  range  \\\n",
       "0        RBC_GEM_PC_A10_Post        13DAMPPOX     0.00  0.0  0.0    0.0   \n",
       "1        RBC_GEM_PC_A10_Post        13DAMPPOX     0.50  0.0  0.0    0.0   \n",
       "2        RBC_GEM_PC_A10_Post        13DAMPPOX     0.90  0.0  0.0    0.0   \n",
       "3        RBC_GEM_PC_A10_Post        13DAMPPOX     0.99  0.0  0.0    0.0   \n",
       "4        RBC_GEM_PC_A10_Post        15KPGE1Ry     0.00  0.0  0.0    0.0   \n",
       "...                      ...              ...      ...  ...  ...    ...   \n",
       "1401267   RBC_GEM_PC_MED9_TD  ZN2_HCO3_SELNIt     0.99  0.0  0.0    0.0   \n",
       "1401268   RBC_GEM_PC_MED9_TD             ZN2t     0.00  0.0  0.0    0.0   \n",
       "1401269   RBC_GEM_PC_MED9_TD             ZN2t     0.50  0.0  0.0    0.0   \n",
       "1401270   RBC_GEM_PC_MED9_TD             ZN2t     0.90  0.0  0.0    0.0   \n",
       "1401271   RBC_GEM_PC_MED9_TD             ZN2t     0.99  0.0  0.0    0.0   \n",
       "\n",
       "         abundance donor  time phenotype  \n",
       "0              NaN   A10  Post         A  \n",
       "1              NaN   A10  Post         A  \n",
       "2              NaN   A10  Post         A  \n",
       "3              NaN   A10  Post         A  \n",
       "4              0.0   A10  Post         A  \n",
       "...            ...   ...   ...       ...  \n",
       "1401267        0.0  MED9    TD       MED  \n",
       "1401268        0.0  MED9    TD       MED  \n",
       "1401269        0.0  MED9    TD       MED  \n",
       "1401270        0.0  MED9    TD       MED  \n",
       "1401271        0.0  MED9    TD       MED  \n",
       "\n",
       "[1401272 rows x 10 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_data_samples = df_data_all[\n",
    "    [not bool(operation_re.search(x)) for x in df_data_all[\"model\"]]\n",
    "].copy()\n",
    "df_data_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a7f4389-594f-4c01-8743-94cbef5ffe54",
   "metadata": {},
   "source": [
    "### Create groups of sample models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "099f41d9-3091-46ef-9fb9-94e0431bfdf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_key = \"ALL\"\n",
    "model_groups = {all_key: list(df_data_samples[\"model\"].unique())}\n",
    "\n",
    "\n",
    "def create_group_of_models(df, groupby, verbose=False):\n",
    "    grouped = df.groupby(groupby)[\"model\"].agg(lambda x: list(x.unique()))\n",
    "    grouped = {\"_\".join(ensure_iterable(k)): v for k, v in grouped.to_dict().items()}\n",
    "    max_name_len = max([len(group_name) for group_name in list(grouped)])\n",
    "    if verbose:\n",
    "        for group_name, model_list in grouped.items():\n",
    "            spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "            print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "    return grouped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "325246aa-d8a5-4b14-8792-d1a535a6f017",
   "metadata": {},
   "source": [
    "#### Based on timepoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "13dc0a34-dc33-44c0-8624-044effe099ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Post:\t35 samples\n",
      "Pre: \t36 samples\n",
      "TD:  \t36 samples\n"
     ]
    }
   ],
   "source": [
    "grouped = create_group_of_models(\n",
    "    df_data_samples[[\"model\"] + metadata_columns], groupby=[\"time\"], verbose=verbose\n",
    ")\n",
    "model_groups.update(grouped)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c290b56e-0285-4f7a-84b1-0c6376dc545d",
   "metadata": {},
   "source": [
    "##### Based on timepoint and phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bd8b83f1-abda-4ade-b5bd-098b6885dd13",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Post_A:     \t12 samples\n",
      "Post_HumCan:\t11 samples\n",
      "Post_MED:   \t12 samples\n",
      "Pre_A:      \t12 samples\n",
      "Pre_HumCan: \t12 samples\n",
      "Pre_MED:    \t12 samples\n",
      "TD_A:       \t12 samples\n",
      "TD_HumCan:  \t12 samples\n",
      "TD_MED:     \t12 samples\n"
     ]
    }
   ],
   "source": [
    "grouped = create_group_of_models(\n",
    "    df_data_samples[[\"model\"] + metadata_columns],\n",
    "    groupby=[\"time\", \"phenotype\"],\n",
    "    verbose=verbose,\n",
    ")\n",
    "model_groups.update(grouped)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c472fbd2-8a5f-4e1e-a93e-901a30ee98ae",
   "metadata": {},
   "source": [
    "#### Based on phenotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bb437bf3-7040-4c20-9eb4-8c73c574d1ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A:     \t36 samples\n",
      "HumCan:\t35 samples\n",
      "MED:   \t36 samples\n"
     ]
    }
   ],
   "source": [
    "grouped = create_group_of_models(\n",
    "    df_data_samples[[\"model\"] + metadata_columns],\n",
    "    groupby=[\"phenotype\"],\n",
    "    verbose=verbose,\n",
    ")\n",
    "model_groups.update(grouped)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4fa0118-d6f0-4b5c-ba1c-a242cd7d0f7f",
   "metadata": {},
   "source": [
    "##### Based on phenotype and timepoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "09db8d24-a91a-4031-b3c6-c398c068f9af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A_Post:     \t12 samples\n",
      "A_Pre:      \t12 samples\n",
      "A_TD:       \t12 samples\n",
      "HumCan_Post:\t11 samples\n",
      "HumCan_Pre: \t12 samples\n",
      "HumCan_TD:  \t12 samples\n",
      "MED_Post:   \t12 samples\n",
      "MED_Pre:    \t12 samples\n",
      "MED_TD:     \t12 samples\n"
     ]
    }
   ],
   "source": [
    "grouped = create_group_of_models(\n",
    "    df_data_samples[[\"model\"] + metadata_columns],\n",
    "    groupby=[\"phenotype\", \"time\"],\n",
    "    verbose=verbose,\n",
    ")\n",
    "model_groups.update(grouped)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f798fa7e-2a9a-4664-afbe-b30fd94b1df0",
   "metadata": {},
   "source": [
    "### View model groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a1152679-d912-4768-802c-5d365e05c481",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Possible groups for analyses\n",
      "============================\n",
      "ALL:        \t107 samples\n",
      "Post:       \t35 samples\n",
      "Pre:        \t36 samples\n",
      "TD:         \t36 samples\n",
      "Post_A:     \t12 samples\n",
      "Post_HumCan:\t11 samples\n",
      "Post_MED:   \t12 samples\n",
      "Pre_A:      \t12 samples\n",
      "Pre_HumCan: \t12 samples\n",
      "Pre_MED:    \t12 samples\n",
      "TD_A:       \t12 samples\n",
      "TD_HumCan:  \t12 samples\n",
      "TD_MED:     \t12 samples\n",
      "A:          \t36 samples\n",
      "HumCan:     \t35 samples\n",
      "MED:        \t36 samples\n",
      "A_Post:     \t12 samples\n",
      "A_Pre:      \t12 samples\n",
      "A_TD:       \t12 samples\n",
      "HumCan_Post:\t11 samples\n",
      "HumCan_Pre: \t12 samples\n",
      "HumCan_TD:  \t12 samples\n",
      "MED_Post:   \t12 samples\n",
      "MED_Pre:    \t12 samples\n",
      "MED_TD:     \t12 samples\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>optimum</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th>model</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">13DAMPPOX</th>\n",
       "      <th>RBC_GEM_PC_A10_Post</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBC_GEM_PC_A10_Post</th>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBC_GEM_PC_A10_Post</th>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBC_GEM_PC_A10_Post</th>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE1Ry</th>\n",
       "      <th>RBC_GEM_PC_A10_Post</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2_HCO3_SELNIt</th>\n",
       "      <th>RBC_GEM_PC_MED9_TD</th>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">ZN2t</th>\n",
       "      <th>RBC_GEM_PC_MED9_TD</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBC_GEM_PC_MED9_TD</th>\n",
       "      <td>0.50</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBC_GEM_PC_MED9_TD</th>\n",
       "      <td>0.90</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RBC_GEM_PC_MED9_TD</th>\n",
       "      <td>0.99</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1401272 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     optimum  min  max  range  abundance\n",
       "reactions       model                                                   \n",
       "13DAMPPOX       RBC_GEM_PC_A10_Post     0.00  0.0  0.0    0.0        NaN\n",
       "                RBC_GEM_PC_A10_Post     0.50  0.0  0.0    0.0        NaN\n",
       "                RBC_GEM_PC_A10_Post     0.90  0.0  0.0    0.0        NaN\n",
       "                RBC_GEM_PC_A10_Post     0.99  0.0  0.0    0.0        NaN\n",
       "15KPGE1Ry       RBC_GEM_PC_A10_Post     0.00  0.0  0.0    0.0        0.0\n",
       "...                                      ...  ...  ...    ...        ...\n",
       "ZN2_HCO3_SELNIt RBC_GEM_PC_MED9_TD      0.99  0.0  0.0    0.0        0.0\n",
       "ZN2t            RBC_GEM_PC_MED9_TD      0.00  0.0  0.0    0.0        0.0\n",
       "                RBC_GEM_PC_MED9_TD      0.50  0.0  0.0    0.0        0.0\n",
       "                RBC_GEM_PC_MED9_TD      0.90  0.0  0.0    0.0        0.0\n",
       "                RBC_GEM_PC_MED9_TD      0.99  0.0  0.0    0.0        0.0\n",
       "\n",
       "[1401272 rows x 5 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Possible groups for analyses\\n============================\")\n",
    "max_name_len = max([len(group_name) for group_name in list(model_groups)])\n",
    "for group_name, model_list in model_groups.items():\n",
    "    spacepad = \"\".join([\" \"] * (max_name_len - len(group_name)))\n",
    "    print(f\"{group_name}:{spacepad}\\t{len(model_list)} samples\")\n",
    "\n",
    "df_data_for_analyses = df_data_samples.set_index([\"reactions\", \"model\"]).drop(\n",
    "    metadata_columns, axis=1\n",
    ")\n",
    "df_data_for_analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2684b9-37c0-462b-98a1-5d20012b4e93",
   "metadata": {},
   "source": [
    "#### Ensure groups exist and setup directory structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e89c9396-815f-450e-a20c-d58be95ac66d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expected directory structure\n",
      "============================\n",
      "ALL\n",
      "└── A\n",
      "│   └──A_Post\n",
      "│   └──A_Pre\n",
      "│   └──A_TD\n",
      "└── HumCan\n",
      "│   └──HumCan_Post\n",
      "│   └──HumCan_Pre\n",
      "│   └──HumCan_TD\n",
      "└── MED\n",
      "│   └──MED_Post\n",
      "│   └──MED_Pre\n",
      "│   └──MED_TD\n",
      "└── Post\n",
      "│   └──Post_A\n",
      "│   └──Post_HumCan\n",
      "│   └──Post_MED\n",
      "└── Pre\n",
      "│   └──Pre_A\n",
      "│   └──Pre_HumCan\n",
      "│   └──Pre_MED\n",
      "└── TD\n",
      "    └──TD_A\n",
      "    └──TD_HumCan\n",
      "    └──TD_MED\n"
     ]
    }
   ],
   "source": [
    "# New directories are created for main groups\n",
    "groups_dict = defaultdict(dict)\n",
    "group_items_list = [\n",
    "    [\"Pre\", \"Post\", \"TD\"],\n",
    "    [\"HumCan\", \"A\", \"MED\"],\n",
    "]\n",
    "# All main groups are created under the \"All\" directory, subgroups are created under each main group\n",
    "for item_list in group_items_list:\n",
    "    groups_dict[all_key].update({item: {} for item in item_list})\n",
    "    groups_dict[all_key].update(\n",
    "        {\n",
    "            item: sorted(\n",
    "                [\n",
    "                    group_name\n",
    "                    for group_name in model_groups\n",
    "                    if group_name.split(\"_\")[0] == item and group_name != item\n",
    "                ]\n",
    "            )\n",
    "            for item in item_list\n",
    "        }\n",
    "    )\n",
    "groups_dict[all_key].update(\n",
    "    {k: groups_dict[all_key][k] for k in sorted(sorted(groups_dict[all_key]))}\n",
    ")\n",
    "invalid_groups = [\n",
    "    group_name for group_name in groups_dict[all_key] if group_name not in model_groups\n",
    "]\n",
    "if any(invalid_groups):\n",
    "    raise KeyError(\n",
    "        f\"No group(s) found for `{invalid_groups}`. Model groups must be created first before correlation computations\"\n",
    "    )\n",
    "invalid_subgroups = [\n",
    "    subgroup\n",
    "    for group_values in groups_dict[all_key].values()\n",
    "    for subgroup in group_values\n",
    "    if subgroup not in model_groups\n",
    "]\n",
    "if any(invalid_subgroups):\n",
    "    raise KeyError(\n",
    "        f\"No subgroup(s) found for `{invalid_subgroups}`. Model groups must be created first before correlation computations\"\n",
    "    )\n",
    "\n",
    "\n",
    "header = \"Expected directory structure\"\n",
    "print(\"\\n\".join((header, \"=\" * len(header), all_key)))\n",
    "for idx, (group_name, subgroups) in enumerate(sorted(groups_dict[all_key].items())):\n",
    "    print(\"\\u2514\\u2500\\u2500\" + f\" {group_name}\")\n",
    "    vertical = \"\\u2502\" if idx != len(groups_dict[all_key]) - 1 else \" \"\n",
    "    for subgroup_name in sorted(subgroups):\n",
    "        print(vertical + \"   \\u2514\\u2500\\u2500\" + subgroup_name)\n",
    "\n",
    "group_results_dirpath_dict = {all_key: corr_results_dirpath}\n",
    "for group_name, subgroups in groups_dict[all_key].items():\n",
    "    group_results_dirpath_dict[group_name] = (\n",
    "        group_results_dirpath_dict[all_key] / group_name\n",
    "    )\n",
    "    group_results_dirpath_dict.update(\n",
    "        {\n",
    "            subgroup_name: group_results_dirpath_dict[group_name] / subgroup_name\n",
    "            for subgroup_name in subgroups\n",
    "        }\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b8fc2e6-5be8-4ac8-98ad-92d69335d316",
   "metadata": {},
   "source": [
    "#### Load subsystems and metabolic categories to enrich results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f7fff7e1-e706-4abf-adc9-6ee5a72bfd5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subsystem</th>\n",
       "      <th>category</th>\n",
       "      <th>kegg.pathway.name</th>\n",
       "      <th>kegg.pathway</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Alanine, aspartate and glutamate metabolism</td>\n",
       "      <td>hsa00250</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Arginine and proline metabolism</td>\n",
       "      <td>hsa00330</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>hsa00270</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Glycine, serine and threonine metabolism</td>\n",
       "      <td>hsa00260</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "      <td>Histidine metabolism</td>\n",
       "      <td>hsa00340</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>Other</td>\n",
       "      <td>Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>has00970</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Transport, extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Representative subsystem for all transport rea...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>5-fluorouracil metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>hsa00983</td>\n",
       "      <td>Subnetwork of KEGG pathway \"Drug metabolism - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>Azathioprine and 6-mercaptopurine metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td>Drug metabolism - other enzymes</td>\n",
       "      <td>hsa00983</td>\n",
       "      <td>Subnetwork of KEGG pathway \"Drug metabolism - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>Ribavirin metabolism</td>\n",
       "      <td>Other</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Curated category, from reactome R-HSA-9755088,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       subsystem               category  \\\n",
       "0    Alanine, aspartate and glutamate metabolism  Amino acid metabolism   \n",
       "1                Arginine and proline metabolism  Amino acid metabolism   \n",
       "2             Cysteine and methionine metabolism  Amino acid metabolism   \n",
       "3       Glycine, serine and threonine metabolism  Amino acid metabolism   \n",
       "4                           Histidine metabolism  Amino acid metabolism   \n",
       "..                                           ...                    ...   \n",
       "73                   Aminoacyl-tRNA biosynthesis                  Other   \n",
       "74                      Transport, extracellular    Transport reactions   \n",
       "75                     5-fluorouracil metabolism                  Other   \n",
       "76  Azathioprine and 6-mercaptopurine metabolism                  Other   \n",
       "77                          Ribavirin metabolism                  Other   \n",
       "\n",
       "                              kegg.pathway.name kegg.pathway  \\\n",
       "0   Alanine, aspartate and glutamate metabolism     hsa00250   \n",
       "1               Arginine and proline metabolism     hsa00330   \n",
       "2            Cysteine and methionine metabolism     hsa00270   \n",
       "3      Glycine, serine and threonine metabolism     hsa00260   \n",
       "4                          Histidine metabolism     hsa00340   \n",
       "..                                          ...          ...   \n",
       "73                  Aminoacyl-tRNA biosynthesis     has00970   \n",
       "74                                                             \n",
       "75              Drug metabolism - other enzymes     hsa00983   \n",
       "76              Drug metabolism - other enzymes     hsa00983   \n",
       "77                                                             \n",
       "\n",
       "                                                notes  \n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "..                                                ...  \n",
       "73                                                     \n",
       "74  Representative subsystem for all transport rea...  \n",
       "75  Subnetwork of KEGG pathway \"Drug metabolism - ...  \n",
       "76  Subnetwork of KEGG pathway \"Drug metabolism - ...  \n",
       "77  Curated category, from reactome R-HSA-9755088,...  \n",
       "\n",
       "[77 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystems_to_exclude = {\"Pseudoreactions\"}\n",
    "use_abbrevs = True\n",
    "abbreviations = {\n",
    "    \"Amino acid metabolism\": \"A\",\n",
    "    \"Carbohydrate metabolism\": \"C\",\n",
    "    \"Lipid metabolism\": \"L\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"V\",\n",
    "    \"Nucleotide metabolism\": \"N\",\n",
    "    \"Reactive species\": \"R\",\n",
    "    \"Transport reactions\": \"T\",\n",
    "    \"Other\": \"O\",\n",
    "}\n",
    "categories_to_keep = list(abbreviations)\n",
    "\n",
    "df_pathways = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / \"subsystems.tsv\", sep=\"\\t\", dtype=str\n",
    ").fillna(\"\")\n",
    "\n",
    "# Rename \"name\" to subsystem to match reaction attribute\n",
    "df_pathways = df_pathways.rename({\"name\": \"subsystem\"}, axis=1)\n",
    "# Group \"Metabolism of other amino acids\" with amino acids rather than treat as \"other\"\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].replace(\n",
    "    \"Metabolism of other amino acids\", \"Amino acid metabolism\"\n",
    ")\n",
    "\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].apply(\n",
    "    lambda x: (\"Other\" if x not in categories_to_keep else x)\n",
    ")\n",
    "df_pathways = df_pathways[~df_pathways[\"subsystem\"].isin(subsystems_to_exclude)].copy()\n",
    "subsystem_to_category_dict = df_pathways.set_index(\"subsystem\")[\"category\"].to_dict()\n",
    "df_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1281f9fb-3124-4f6b-95c6-ccc02b3c6733",
   "metadata": {},
   "source": [
    "## Compute significant results between groups\n",
    "#### Compare all subgroups at once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "3f37f27e-c5ab-4403-9bca-ec37a7208c97",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Groups to compare\n",
      "=================\n",
      "('Pre_HumCan', 'Pre_A', 'Pre_MED')\n",
      "('Pre_HumCan', 'Pre_A')\n",
      "('Pre_HumCan', 'Pre_MED')\n",
      "('Pre_A', 'Pre_MED')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>range</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th>model</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13DAMPPOX</th>\n",
       "      <th>RBC_GEM_PC_HumCan10_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE1Ry</th>\n",
       "      <th>RBC_GEM_PC_HumCan10_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE2Ry</th>\n",
       "      <th>RBC_GEM_PC_HumCan10_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGE3Ry</th>\n",
       "      <th>RBC_GEM_PC_HumCan10_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15KPGF1Ry</th>\n",
       "      <th>RBC_GEM_PC_HumCan10_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YYYTPAP</th>\n",
       "      <th>RBC_GEM_PC_MED9_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.979572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2Htex2</th>\n",
       "      <th>RBC_GEM_PC_MED9_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2_2HCO3t</th>\n",
       "      <th>RBC_GEM_PC_MED9_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2_HCO3_SELNIt</th>\n",
       "      <th>RBC_GEM_PC_MED9_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZN2t</th>\n",
       "      <th>RBC_GEM_PC_MED9_Pre</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117864 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         min  max  range  abundance\n",
       "reactions       model                                              \n",
       "13DAMPPOX       RBC_GEM_PC_HumCan10_Pre  0.0  0.0    0.0        NaN\n",
       "15KPGE1Ry       RBC_GEM_PC_HumCan10_Pre  0.0  0.0    0.0   0.000000\n",
       "15KPGE2Ry       RBC_GEM_PC_HumCan10_Pre  0.0  0.0    0.0   0.000000\n",
       "15KPGE3Ry       RBC_GEM_PC_HumCan10_Pre  0.0  0.0    0.0   0.000000\n",
       "15KPGF1Ry       RBC_GEM_PC_HumCan10_Pre  0.0  0.0    0.0   0.000000\n",
       "...                                      ...  ...    ...        ...\n",
       "YYYTPAP         RBC_GEM_PC_MED9_Pre      0.0  0.0    0.0   0.979572\n",
       "ZN2Htex2        RBC_GEM_PC_MED9_Pre      0.0  0.0    0.0   0.000000\n",
       "ZN2_2HCO3t      RBC_GEM_PC_MED9_Pre      0.0  0.0    0.0   0.000000\n",
       "ZN2_HCO3_SELNIt RBC_GEM_PC_MED9_Pre      0.0  0.0    0.0   0.000000\n",
       "ZN2t            RBC_GEM_PC_MED9_Pre      0.0  0.0    0.0   0.000000\n",
       "\n",
       "[117864 rows x 4 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "group_name = \"Pre\"\n",
    "optimum = 0\n",
    "value_to_compare = \"range\"\n",
    "compare_pairwise = True\n",
    "compare_all_groups = True\n",
    "ordered_group_to_compare = [f\"{group_name}_{phenotype}\" for phenotype in phenotypes]\n",
    "\n",
    "all_samples_for_comparison = [\n",
    "    value for g in ordered_group_to_compare for value in np.array(model_groups[g])\n",
    "]\n",
    "df_data_for_correlations = df_data_for_analyses.loc[\n",
    "    pd.IndexSlice[:, all_samples_for_comparison], :\n",
    "]\n",
    "df_data_for_correlations = df_data_for_correlations[\n",
    "    df_data_for_correlations[\"optimum\"] == optimum\n",
    "].drop(\"optimum\", axis=1)\n",
    "\n",
    "if value_to_compare == \"max\":\n",
    "    df_data_for_correlations[\"max\"] = (\n",
    "        df_data_for_correlations[[\"min\", \"max\"]].abs().max(axis=1)\n",
    "    )\n",
    "    df_data_for_correlations = df_data_for_correlations.drop(\"min\", axis=1)\n",
    "\n",
    "print(\"Groups to compare\\n=================\")\n",
    "if compare_all_groups:\n",
    "    print(tuple(ordered_group_to_compare))\n",
    "if compare_pairwise:\n",
    "    pairwise_group_combos = list(combinations(ordered_group_to_compare, 2))\n",
    "    for group in pairwise_group_combos:\n",
    "        print(group)\n",
    "df_data_for_correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "03749c24-92f5-4819-bf70-77778085e5c9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of different comparisons made: 4\n",
      "Groups compared\n",
      "===============\n",
      "('Pre_HumCan', 'Pre_A', 'Pre_MED')\n",
      "('Pre_HumCan', 'Pre_A')\n",
      "('Pre_HumCan', 'Pre_MED')\n",
      "('Pre_A', 'Pre_MED')\n"
     ]
    }
   ],
   "source": [
    "results_dict = defaultdict(dict)\n",
    "mannwhitneyu_for_pairs = True\n",
    "for rid in df_data_for_correlations.index.get_level_values(\"reactions\").unique():\n",
    "    df_data_rxn = df_data_for_correlations.loc[rid]\n",
    "    df_data_rxn_opt_value = df_data_rxn[value_to_compare].copy()\n",
    "    data_arrays = {\n",
    "        group_name: df_data_rxn_opt_value.loc[model_groups[group_name]].values\n",
    "        for group_name in ordered_group_to_compare\n",
    "    }\n",
    "    if compare_all_groups:\n",
    "        values = list(data_arrays.values())\n",
    "        unique_values = set(\n",
    "            [v for value_list in values for v in value_list if not np.isnan(v)]\n",
    "        )\n",
    "        if len(unique_values) <= 1:\n",
    "            # Skip variables that do not have any differences\n",
    "            results_dict[tuple(ordered_group_to_compare)][rid] = dict(\n",
    "                zip([\"statistic\", \"pvalue\"], [pd.NA, pd.NA])\n",
    "            )\n",
    "        else:\n",
    "            result = kruskal(*values, nan_policy=\"omit\")\n",
    "            results_dict[tuple(ordered_group_to_compare)][rid] = {\n",
    "                attr: getattr(result, attr) for attr in [\"statistic\", \"pvalue\"]\n",
    "            }\n",
    "    if compare_pairwise:\n",
    "        for combo in pairwise_group_combos:\n",
    "            values = [data_arrays[group] for group in combo]\n",
    "            unique_values = set(\n",
    "                [v for value_list in values for v in value_list if not np.isnan(v)]\n",
    "            )\n",
    "            if len(unique_values) <= 1:\n",
    "                # Skip variables that do not have any differences\n",
    "                result = dict(zip([\"statistic\", \"pvalue\"], [pd.NA, pd.NA]))\n",
    "            elif mannwhitneyu_for_pairs:\n",
    "                result = mannwhitneyu(*values, nan_policy=\"omit\")\n",
    "                result = {\n",
    "                    attr: getattr(result, attr) for attr in [\"statistic\", \"pvalue\"]\n",
    "                }\n",
    "            else:\n",
    "                result = kruskal(*values, nan_policy=\"omit\")\n",
    "                result = {\n",
    "                    attr: getattr(result, attr) for attr in [\"statistic\", \"pvalue\"]\n",
    "                }\n",
    "            results_dict[combo][rid] = result\n",
    "dataframes = {\n",
    "    key: pd.DataFrame.from_dict(values, orient=\"index\")\n",
    "    for key, values in results_dict.items()\n",
    "}\n",
    "print(f\"Number of different comparisons made: {len(dataframes)}\")\n",
    "print(\"Groups compared\\n===============\")\n",
    "for key in list(dataframes):\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a16a3b4-ff30-4ede-a251-5ef488ed13ab",
   "metadata": {},
   "source": [
    "### Determine significance using p-values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0864a7a5-d2c0-4b80-addf-830f9f36a424",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Pre_HumCan', 'Pre_A', 'Pre_MED')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HumCan1</th>\n",
       "      <th>HumCan2</th>\n",
       "      <th>HumCan3</th>\n",
       "      <th>HumCan4</th>\n",
       "      <th>HumCan5</th>\n",
       "      <th>HumCan6</th>\n",
       "      <th>HumCan7</th>\n",
       "      <th>HumCan8</th>\n",
       "      <th>HumCan9</th>\n",
       "      <th>HumCan10</th>\n",
       "      <th>...</th>\n",
       "      <th>MED9</th>\n",
       "      <th>MED10</th>\n",
       "      <th>MED11</th>\n",
       "      <th>MED12</th>\n",
       "      <th>name</th>\n",
       "      <th>stoichiometry</th>\n",
       "      <th>proteins</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OXOGAH</th>\n",
       "      <td>-0.505879</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>1.626253</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>2.531540</td>\n",
       "      <td>1.113161</td>\n",
       "      <td>1.696265</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>-0.505629</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>2-Oxoglutaramate amidohydrolase</td>\n",
       "      <td>2-oxoglutaramate + H2O --&gt; 2-Oxoglutarate + Am...</td>\n",
       "      <td>Nit2</td>\n",
       "      <td>0.02736</td>\n",
       "      <td>Alanine aspartate and glutamate metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GALK</th>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>3.449967</td>\n",
       "      <td>2.784602</td>\n",
       "      <td>1.736708</td>\n",
       "      <td>2.934428</td>\n",
       "      <td>-0.340423</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>Galactokinase</td>\n",
       "      <td>ATP + galactose --&gt; ADP + Alpha-D-Galactose 1-...</td>\n",
       "      <td>Galk1;Galk2</td>\n",
       "      <td>0.01263</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAL1PP</th>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>3.449967</td>\n",
       "      <td>2.784602</td>\n",
       "      <td>1.736708</td>\n",
       "      <td>2.934428</td>\n",
       "      <td>-0.340423</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>alpha-D-galactose 1-phosphatase</td>\n",
       "      <td>Alpha-D-Galactose 1-phosphate + H2O --&gt; galact...</td>\n",
       "      <td>Impa1;Impa2</td>\n",
       "      <td>0.01263</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G6PM</th>\n",
       "      <td>-0.881187</td>\n",
       "      <td>-0.270990</td>\n",
       "      <td>-1.280460</td>\n",
       "      <td>0.330973</td>\n",
       "      <td>-0.654370</td>\n",
       "      <td>-1.251614</td>\n",
       "      <td>0.666192</td>\n",
       "      <td>-1.302447</td>\n",
       "      <td>-0.417132</td>\n",
       "      <td>-0.469786</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.037303</td>\n",
       "      <td>0.105856</td>\n",
       "      <td>-0.331970</td>\n",
       "      <td>2.540066</td>\n",
       "      <td>Glucose-6-phosphate isomerase (anomerization)</td>\n",
       "      <td>alpha-D-Glucose 6-phosphate &lt;=&gt; beta-D-Glucose...</td>\n",
       "      <td>Gpi1</td>\n",
       "      <td>0.00812</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGMTa</th>\n",
       "      <td>-0.757153</td>\n",
       "      <td>-0.153843</td>\n",
       "      <td>-1.312542</td>\n",
       "      <td>0.276793</td>\n",
       "      <td>-0.702554</td>\n",
       "      <td>-1.391146</td>\n",
       "      <td>0.666374</td>\n",
       "      <td>-1.509916</td>\n",
       "      <td>-0.385171</td>\n",
       "      <td>-0.414374</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004752</td>\n",
       "      <td>0.119766</td>\n",
       "      <td>-0.344829</td>\n",
       "      <td>2.625484</td>\n",
       "      <td>Phosphoglucomutase (alpha-D-glucose 1-phosphate)</td>\n",
       "      <td>D-Glucose 1-phosphate &lt;=&gt; alpha-D-Glucose 6-ph...</td>\n",
       "      <td>Pgm1;Pgm2</td>\n",
       "      <td>0.01351</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGMT</th>\n",
       "      <td>-0.757152</td>\n",
       "      <td>-0.153843</td>\n",
       "      <td>-1.312541</td>\n",
       "      <td>0.276793</td>\n",
       "      <td>-0.702554</td>\n",
       "      <td>-1.391145</td>\n",
       "      <td>0.666374</td>\n",
       "      <td>-1.509916</td>\n",
       "      <td>-0.385170</td>\n",
       "      <td>-0.414373</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.004752</td>\n",
       "      <td>0.119766</td>\n",
       "      <td>-0.344829</td>\n",
       "      <td>2.625484</td>\n",
       "      <td>Phosphoglucomutase</td>\n",
       "      <td>D-Glucose 1-phosphate &lt;=&gt; beta-D-Glucose 6-pho...</td>\n",
       "      <td>Pgm2</td>\n",
       "      <td>0.01351</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G6PDH2</th>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>1.326458</td>\n",
       "      <td>2.001041</td>\n",
       "      <td>1.785031</td>\n",
       "      <td>2.375383</td>\n",
       "      <td>1.111049</td>\n",
       "      <td>3.205984</td>\n",
       "      <td>0.811825</td>\n",
       "      <td>0.461953</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>Glucose 6-phosphate dehydrogenase</td>\n",
       "      <td>beta-D-Glucose 6-phosphate + Nicotinamide aden...</td>\n",
       "      <td>G6pd2;G6pdx</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GND</th>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>1.326458</td>\n",
       "      <td>2.001041</td>\n",
       "      <td>1.785031</td>\n",
       "      <td>2.375383</td>\n",
       "      <td>1.111049</td>\n",
       "      <td>3.205984</td>\n",
       "      <td>0.811825</td>\n",
       "      <td>0.461953</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>Phosphogluconate dehydrogenase</td>\n",
       "      <td>6-phospho-D-gluconate + Nicotinamide adenine d...</td>\n",
       "      <td>Pgd</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGL</th>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>1.326458</td>\n",
       "      <td>2.001041</td>\n",
       "      <td>1.785031</td>\n",
       "      <td>2.375383</td>\n",
       "      <td>1.111049</td>\n",
       "      <td>3.205984</td>\n",
       "      <td>0.811825</td>\n",
       "      <td>0.461953</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>6-phospho-D-glucono-1,5-lactone + H2O &lt;=&gt; 6-ph...</td>\n",
       "      <td>Pgls</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPI</th>\n",
       "      <td>-0.561032</td>\n",
       "      <td>-0.561032</td>\n",
       "      <td>0.377273</td>\n",
       "      <td>0.720554</td>\n",
       "      <td>0.610631</td>\n",
       "      <td>0.911049</td>\n",
       "      <td>0.267655</td>\n",
       "      <td>1.333725</td>\n",
       "      <td>0.115528</td>\n",
       "      <td>-0.062657</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.561032</td>\n",
       "      <td>-0.561032</td>\n",
       "      <td>-0.561032</td>\n",
       "      <td>-0.561032</td>\n",
       "      <td>Ribose-5-phosphate isomerase</td>\n",
       "      <td>D-Ribulose 5-phosphate &lt;=&gt; Alpha-D-Ribose 5-ph...</td>\n",
       "      <td>Rpia</td>\n",
       "      <td>0.01399</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4c</th>\n",
       "      <td>-0.695753</td>\n",
       "      <td>-0.975472</td>\n",
       "      <td>-1.219804</td>\n",
       "      <td>-0.323804</td>\n",
       "      <td>-0.987994</td>\n",
       "      <td>-1.361652</td>\n",
       "      <td>0.727939</td>\n",
       "      <td>-1.677913</td>\n",
       "      <td>0.335838</td>\n",
       "      <td>0.125743</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.018030</td>\n",
       "      <td>-0.458944</td>\n",
       "      <td>-1.022421</td>\n",
       "      <td>0.406857</td>\n",
       "      <td>Deprotonation of ammonium</td>\n",
       "      <td>Ammonium &lt;=&gt; H+ + Ammonia</td>\n",
       "      <td></td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROTNDA</th>\n",
       "      <td>-0.700068</td>\n",
       "      <td>-0.979597</td>\n",
       "      <td>-1.205333</td>\n",
       "      <td>-0.328372</td>\n",
       "      <td>-0.979474</td>\n",
       "      <td>-1.357629</td>\n",
       "      <td>0.732102</td>\n",
       "      <td>-1.681560</td>\n",
       "      <td>0.330819</td>\n",
       "      <td>0.120870</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022807</td>\n",
       "      <td>-0.447717</td>\n",
       "      <td>-1.026514</td>\n",
       "      <td>0.401792</td>\n",
       "      <td>Nonenzymatic deamidation of L-asparaginyl prot...</td>\n",
       "      <td>H+ + L-asparaginyl-[protein] --&gt; Ammonium + [p...</td>\n",
       "      <td></td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Protein modification</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOt</th>\n",
       "      <td>-0.506216</td>\n",
       "      <td>-0.963827</td>\n",
       "      <td>-1.270763</td>\n",
       "      <td>-0.588956</td>\n",
       "      <td>-1.001087</td>\n",
       "      <td>-1.221146</td>\n",
       "      <td>0.558572</td>\n",
       "      <td>-1.510846</td>\n",
       "      <td>0.497794</td>\n",
       "      <td>0.108281</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.089459</td>\n",
       "      <td>-0.456309</td>\n",
       "      <td>-0.827720</td>\n",
       "      <td>0.019612</td>\n",
       "      <td>Transport of nitric oxide via diffusion</td>\n",
       "      <td>Nitric oxide &lt;=&gt; Nitric oxide (extracellular)</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2St</th>\n",
       "      <td>-0.506216</td>\n",
       "      <td>-0.963827</td>\n",
       "      <td>-1.270763</td>\n",
       "      <td>-0.588956</td>\n",
       "      <td>-1.001087</td>\n",
       "      <td>-1.221146</td>\n",
       "      <td>0.558572</td>\n",
       "      <td>-1.510846</td>\n",
       "      <td>0.497794</td>\n",
       "      <td>0.108281</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.089459</td>\n",
       "      <td>-0.456309</td>\n",
       "      <td>-0.827720</td>\n",
       "      <td>0.019612</td>\n",
       "      <td>Superoxide transport via diffusion</td>\n",
       "      <td>Superoxide &lt;=&gt; Superoxide (extracellular)</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2t</th>\n",
       "      <td>-0.506216</td>\n",
       "      <td>-0.963827</td>\n",
       "      <td>-1.270763</td>\n",
       "      <td>-0.588956</td>\n",
       "      <td>-1.001087</td>\n",
       "      <td>-1.221146</td>\n",
       "      <td>0.558572</td>\n",
       "      <td>-1.510846</td>\n",
       "      <td>0.497794</td>\n",
       "      <td>0.108281</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.089459</td>\n",
       "      <td>-0.456309</td>\n",
       "      <td>-0.827720</td>\n",
       "      <td>0.019612</td>\n",
       "      <td>Transport of O2</td>\n",
       "      <td>O2  &lt;=&gt; O2 (extracellular)</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Ot</th>\n",
       "      <td>-0.506216</td>\n",
       "      <td>-0.963827</td>\n",
       "      <td>-1.270763</td>\n",
       "      <td>-0.588956</td>\n",
       "      <td>-1.001087</td>\n",
       "      <td>-1.221146</td>\n",
       "      <td>0.558572</td>\n",
       "      <td>-1.510846</td>\n",
       "      <td>0.497797</td>\n",
       "      <td>0.108281</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.089459</td>\n",
       "      <td>-0.456309</td>\n",
       "      <td>-0.827720</td>\n",
       "      <td>0.019612</td>\n",
       "      <td>Transport of H2O</td>\n",
       "      <td>H2O &lt;=&gt; H2O (extracellular)</td>\n",
       "      <td>Aqp1;Aqp3</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH3t</th>\n",
       "      <td>-0.695753</td>\n",
       "      <td>-0.975472</td>\n",
       "      <td>-1.219804</td>\n",
       "      <td>-0.323804</td>\n",
       "      <td>-0.987994</td>\n",
       "      <td>-1.361652</td>\n",
       "      <td>0.727939</td>\n",
       "      <td>-1.677913</td>\n",
       "      <td>0.335837</td>\n",
       "      <td>0.125743</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.018030</td>\n",
       "      <td>-0.458944</td>\n",
       "      <td>-1.022421</td>\n",
       "      <td>0.406857</td>\n",
       "      <td>Transport of ammonia</td>\n",
       "      <td>Ammonia &lt;=&gt; Ammonia (extracellular)</td>\n",
       "      <td>Aqp1;Aqp3;Rhag;Rhd</td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4t</th>\n",
       "      <td>-0.695753</td>\n",
       "      <td>-0.975472</td>\n",
       "      <td>-1.219804</td>\n",
       "      <td>-0.323804</td>\n",
       "      <td>-0.987994</td>\n",
       "      <td>-1.361652</td>\n",
       "      <td>0.727939</td>\n",
       "      <td>-1.677913</td>\n",
       "      <td>0.335838</td>\n",
       "      <td>0.125743</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.018030</td>\n",
       "      <td>-0.458944</td>\n",
       "      <td>-1.022421</td>\n",
       "      <td>0.406857</td>\n",
       "      <td>Transport of ammonium</td>\n",
       "      <td>Ammonium &lt;=&gt; Ammonium (extracellular)</td>\n",
       "      <td>Aqp1;Aqp3;Rhag;Rhd</td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CO2t</th>\n",
       "      <td>-0.693577</td>\n",
       "      <td>-0.973386</td>\n",
       "      <td>-1.227020</td>\n",
       "      <td>-0.321510</td>\n",
       "      <td>-0.992236</td>\n",
       "      <td>-1.363635</td>\n",
       "      <td>0.725840</td>\n",
       "      <td>-1.676051</td>\n",
       "      <td>0.338338</td>\n",
       "      <td>0.128180</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.015640</td>\n",
       "      <td>-0.464552</td>\n",
       "      <td>-1.020349</td>\n",
       "      <td>0.409383</td>\n",
       "      <td>Transport of CO2</td>\n",
       "      <td>CO2 &lt;=&gt; CO2 (extracellular)</td>\n",
       "      <td>Aqp1;Rhag;Rhd</td>\n",
       "      <td>0.02036</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            HumCan1   HumCan2   HumCan3   HumCan4   HumCan5   HumCan6  \\\n",
       "reactions                                                               \n",
       "OXOGAH    -0.505879 -0.505879  1.626253 -0.505879  2.531540  1.113161   \n",
       "GALK      -0.340823 -0.340823 -0.340823 -0.340823  3.449967  2.784602   \n",
       "GAL1PP    -0.340823 -0.340823 -0.340823 -0.340823  3.449967  2.784602   \n",
       "G6PM      -0.881187 -0.270990 -1.280460  0.330973 -0.654370 -1.251614   \n",
       "PGMTa     -0.757153 -0.153843 -1.312542  0.276793 -0.702554 -1.391146   \n",
       "PGMT      -0.757152 -0.153843 -1.312541  0.276793 -0.702554 -1.391145   \n",
       "G6PDH2    -0.517405 -0.517405  1.326458  2.001041  1.785031  2.375383   \n",
       "GND       -0.517405 -0.517405  1.326458  2.001041  1.785031  2.375383   \n",
       "PGL       -0.517405 -0.517405  1.326458  2.001041  1.785031  2.375383   \n",
       "RPI       -0.561032 -0.561032  0.377273  0.720554  0.610631  0.911049   \n",
       "NH4c      -0.695753 -0.975472 -1.219804 -0.323804 -0.987994 -1.361652   \n",
       "PROTNDA   -0.700068 -0.979597 -1.205333 -0.328372 -0.979474 -1.357629   \n",
       "NOt       -0.506216 -0.963827 -1.270763 -0.588956 -1.001087 -1.221146   \n",
       "O2St      -0.506216 -0.963827 -1.270763 -0.588956 -1.001087 -1.221146   \n",
       "O2t       -0.506216 -0.963827 -1.270763 -0.588956 -1.001087 -1.221146   \n",
       "H2Ot      -0.506216 -0.963827 -1.270763 -0.588956 -1.001087 -1.221146   \n",
       "NH3t      -0.695753 -0.975472 -1.219804 -0.323804 -0.987994 -1.361652   \n",
       "NH4t      -0.695753 -0.975472 -1.219804 -0.323804 -0.987994 -1.361652   \n",
       "CO2t      -0.693577 -0.973386 -1.227020 -0.321510 -0.992236 -1.363635   \n",
       "\n",
       "            HumCan7   HumCan8   HumCan9  HumCan10  ...      MED9     MED10  \\\n",
       "reactions                                          ...                       \n",
       "OXOGAH     1.696265 -0.505879 -0.505629 -0.505879  ... -0.505879 -0.505879   \n",
       "GALK       1.736708  2.934428 -0.340423 -0.340823  ... -0.340823 -0.340823   \n",
       "GAL1PP     1.736708  2.934428 -0.340423 -0.340823  ... -0.340823 -0.340823   \n",
       "G6PM       0.666192 -1.302447 -0.417132 -0.469786  ... -0.037303  0.105856   \n",
       "PGMTa      0.666374 -1.509916 -0.385171 -0.414374  ... -0.004752  0.119766   \n",
       "PGMT       0.666374 -1.509916 -0.385170 -0.414373  ... -0.004752  0.119766   \n",
       "G6PDH2     1.111049  3.205984  0.811825  0.461953  ... -0.517405 -0.517405   \n",
       "GND        1.111049  3.205984  0.811825  0.461953  ... -0.517405 -0.517405   \n",
       "PGL        1.111049  3.205984  0.811825  0.461953  ... -0.517405 -0.517405   \n",
       "RPI        0.267655  1.333725  0.115528 -0.062657  ... -0.561032 -0.561032   \n",
       "NH4c       0.727939 -1.677913  0.335838  0.125743  ... -0.018030 -0.458944   \n",
       "PROTNDA    0.732102 -1.681560  0.330819  0.120870  ... -0.022807 -0.447717   \n",
       "NOt        0.558572 -1.510846  0.497794  0.108281  ... -0.089459 -0.456309   \n",
       "O2St       0.558572 -1.510846  0.497794  0.108281  ... -0.089459 -0.456309   \n",
       "O2t        0.558572 -1.510846  0.497794  0.108281  ... -0.089459 -0.456309   \n",
       "H2Ot       0.558572 -1.510846  0.497797  0.108281  ... -0.089459 -0.456309   \n",
       "NH3t       0.727939 -1.677913  0.335837  0.125743  ... -0.018030 -0.458944   \n",
       "NH4t       0.727939 -1.677913  0.335838  0.125743  ... -0.018030 -0.458944   \n",
       "CO2t       0.725840 -1.676051  0.338338  0.128180  ... -0.015640 -0.464552   \n",
       "\n",
       "              MED11     MED12  \\\n",
       "reactions                       \n",
       "OXOGAH    -0.505879 -0.505879   \n",
       "GALK      -0.340823 -0.340823   \n",
       "GAL1PP    -0.340823 -0.340823   \n",
       "G6PM      -0.331970  2.540066   \n",
       "PGMTa     -0.344829  2.625484   \n",
       "PGMT      -0.344829  2.625484   \n",
       "G6PDH2    -0.517405 -0.517405   \n",
       "GND       -0.517405 -0.517405   \n",
       "PGL       -0.517405 -0.517405   \n",
       "RPI       -0.561032 -0.561032   \n",
       "NH4c      -1.022421  0.406857   \n",
       "PROTNDA   -1.026514  0.401792   \n",
       "NOt       -0.827720  0.019612   \n",
       "O2St      -0.827720  0.019612   \n",
       "O2t       -0.827720  0.019612   \n",
       "H2Ot      -0.827720  0.019612   \n",
       "NH3t      -1.022421  0.406857   \n",
       "NH4t      -1.022421  0.406857   \n",
       "CO2t      -1.020349  0.409383   \n",
       "\n",
       "                                                        name  \\\n",
       "reactions                                                      \n",
       "OXOGAH                       2-Oxoglutaramate amidohydrolase   \n",
       "GALK                                           Galactokinase   \n",
       "GAL1PP                       alpha-D-galactose 1-phosphatase   \n",
       "G6PM           Glucose-6-phosphate isomerase (anomerization)   \n",
       "PGMTa       Phosphoglucomutase (alpha-D-glucose 1-phosphate)   \n",
       "PGMT                                      Phosphoglucomutase   \n",
       "G6PDH2                     Glucose 6-phosphate dehydrogenase   \n",
       "GND                           Phosphogluconate dehydrogenase   \n",
       "PGL                                6-phosphogluconolactonase   \n",
       "RPI                             Ribose-5-phosphate isomerase   \n",
       "NH4c                               Deprotonation of ammonium   \n",
       "PROTNDA    Nonenzymatic deamidation of L-asparaginyl prot...   \n",
       "NOt                  Transport of nitric oxide via diffusion   \n",
       "O2St                      Superoxide transport via diffusion   \n",
       "O2t                                          Transport of O2   \n",
       "H2Ot                                        Transport of H2O   \n",
       "NH3t                                    Transport of ammonia   \n",
       "NH4t                                   Transport of ammonium   \n",
       "CO2t                                        Transport of CO2   \n",
       "\n",
       "                                               stoichiometry  \\\n",
       "reactions                                                      \n",
       "OXOGAH     2-oxoglutaramate + H2O --> 2-Oxoglutarate + Am...   \n",
       "GALK       ATP + galactose --> ADP + Alpha-D-Galactose 1-...   \n",
       "GAL1PP     Alpha-D-Galactose 1-phosphate + H2O --> galact...   \n",
       "G6PM       alpha-D-Glucose 6-phosphate <=> beta-D-Glucose...   \n",
       "PGMTa      D-Glucose 1-phosphate <=> alpha-D-Glucose 6-ph...   \n",
       "PGMT       D-Glucose 1-phosphate <=> beta-D-Glucose 6-pho...   \n",
       "G6PDH2     beta-D-Glucose 6-phosphate + Nicotinamide aden...   \n",
       "GND        6-phospho-D-gluconate + Nicotinamide adenine d...   \n",
       "PGL        6-phospho-D-glucono-1,5-lactone + H2O <=> 6-ph...   \n",
       "RPI        D-Ribulose 5-phosphate <=> Alpha-D-Ribose 5-ph...   \n",
       "NH4c                               Ammonium <=> H+ + Ammonia   \n",
       "PROTNDA    H+ + L-asparaginyl-[protein] --> Ammonium + [p...   \n",
       "NOt            Nitric oxide <=> Nitric oxide (extracellular)   \n",
       "O2St               Superoxide <=> Superoxide (extracellular)   \n",
       "O2t                               O2  <=> O2 (extracellular)   \n",
       "H2Ot                             H2O <=> H2O (extracellular)   \n",
       "NH3t                     Ammonia <=> Ammonia (extracellular)   \n",
       "NH4t                   Ammonium <=> Ammonium (extracellular)   \n",
       "CO2t                             CO2 <=> CO2 (extracellular)   \n",
       "\n",
       "                     proteins   pvalue  \\\n",
       "reactions                                \n",
       "OXOGAH                   Nit2  0.02736   \n",
       "GALK              Galk1;Galk2  0.01263   \n",
       "GAL1PP            Impa1;Impa2  0.01263   \n",
       "G6PM                     Gpi1  0.00812   \n",
       "PGMTa               Pgm1;Pgm2  0.01351   \n",
       "PGMT                     Pgm2  0.01351   \n",
       "G6PDH2            G6pd2;G6pdx  0.00001   \n",
       "GND                       Pgd  0.00001   \n",
       "PGL                      Pgls  0.00001   \n",
       "RPI                      Rpia  0.01399   \n",
       "NH4c                           0.02384   \n",
       "PROTNDA                        0.02384   \n",
       "NOt                      Aqp1  0.02143   \n",
       "O2St                     Aqp1  0.02143   \n",
       "O2t                      Aqp1  0.02143   \n",
       "H2Ot                Aqp1;Aqp3  0.02143   \n",
       "NH3t       Aqp1;Aqp3;Rhag;Rhd  0.02384   \n",
       "NH4t       Aqp1;Aqp3;Rhag;Rhd  0.02384   \n",
       "CO2t            Aqp1;Rhag;Rhd  0.02036   \n",
       "\n",
       "                                            subsystem                 category  \n",
       "reactions                                                                       \n",
       "OXOGAH     Alanine aspartate and glutamate metabolism    Amino acid metabolism  \n",
       "GALK                             Galactose metabolism  Carbohydrate metabolism  \n",
       "GAL1PP                           Galactose metabolism  Carbohydrate metabolism  \n",
       "G6PM                     Glycolysis / Gluconeogenesis  Carbohydrate metabolism  \n",
       "PGMTa                    Glycolysis / Gluconeogenesis  Carbohydrate metabolism  \n",
       "PGMT                     Glycolysis / Gluconeogenesis  Carbohydrate metabolism  \n",
       "G6PDH2                      Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "GND                         Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "PGL                         Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "RPI                         Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "NH4c                                    Miscellaneous                    Other  \n",
       "PROTNDA                          Protein modification                    Other  \n",
       "NOt                           Transport extracellular      Transport reactions  \n",
       "O2St                          Transport extracellular      Transport reactions  \n",
       "O2t                           Transport extracellular      Transport reactions  \n",
       "H2Ot                          Transport extracellular      Transport reactions  \n",
       "NH3t                          Transport extracellular      Transport reactions  \n",
       "NH4t                          Transport extracellular      Transport reactions  \n",
       "CO2t                          Transport extracellular      Transport reactions  \n",
       "\n",
       "[19 rows x 42 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pvalue_sig = 0.05\n",
    "enzyme_reactions_only = False\n",
    "include_boundary_reactions = False\n",
    "remove_group_name_from_samples = True\n",
    "sort_by_subsystem = True\n",
    "standardize_by = \"mean\"\n",
    "use_group_means = False\n",
    "\n",
    "significant_dataframes = {}\n",
    "for met in model.metabolites.query(lambda x: x.compartment == \"e\"):\n",
    "    met.name += \" (extracellular)\"\n",
    "metadata_columns = [\n",
    "    \"name\",\n",
    "    \"stoichiometry\",\n",
    "    \"proteins\",\n",
    "    \"pvalue\",\n",
    "    \"subsystem\",\n",
    "    \"category\",\n",
    "]\n",
    "for key, df in dataframes.items():\n",
    "    df = df.dropna()\n",
    "    df = df[df[\"pvalue\"] <= pvalue_sig].drop(\"statistic\", axis=1)\n",
    "    if enzyme_reactions_only:\n",
    "        df_pivot = df_data_for_correlations.loc[\n",
    "            df.index, [\"abundance\", value_to_compare]\n",
    "        ].dropna(subset=\"abundance\")\n",
    "        df_pivot = df_pivot.drop(\"abundance\", axis=1)\n",
    "    else:\n",
    "        df_pivot = df_data_for_correlations.loc[df.index, value_to_compare]\n",
    "    if not include_boundary_reactions:\n",
    "        df_pivot = df_pivot[\n",
    "            ~df_pivot.index.isin(\n",
    "                model.reactions.query(lambda x: x.boundary).list_attr(\"id\"),\n",
    "                level=\"reactions\",\n",
    "            )\n",
    "        ]\n",
    "    df_pivot = df_pivot.reset_index(drop=False)\n",
    "    df_pivot = df_pivot.pivot(\n",
    "        columns=\"model\", index=\"reactions\", values=value_to_compare\n",
    "    )\n",
    "    df = pd.merge(df, df_pivot, left_index=True, right_index=True).sort_values(\"pvalue\")\n",
    "    df.index.name = \"reactions\"\n",
    "    df = df.reset_index(drop=False).set_index([\"reactions\", \"pvalue\"]).T\n",
    "    df = pd.concat(\n",
    "        [\n",
    "            # Sort index by donor number and subgroup while concatenating\n",
    "            df.loc[model_groups[g]].sort_index(\n",
    "                key=lambda x: [int(donor_re.search(v).group(\"num\")) for v in x.values]\n",
    "            )\n",
    "            for g in key\n",
    "        ],\n",
    "        axis=0,\n",
    "    )\n",
    "    # Remove pcmodel ID from sample names\n",
    "    df.index = [sample_id.replace(f\"{pcmodel.id}_\", \"\") for sample_id in df.index]\n",
    "    if remove_group_name_from_samples:\n",
    "        df.index = [\n",
    "            \"_\".join([x for x in sample_id.split(\"_\") if x != group_name])\n",
    "            for sample_id in df.index\n",
    "        ]\n",
    "\n",
    "    df = df.T.reset_index(drop=False)\n",
    "    # Enrich results\n",
    "    df[\"name\"] = [\n",
    "        r.name for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "    ]\n",
    "    df[\"stoichiometry\"] = [\n",
    "        r.build_reaction_string(use_metabolite_names=True)\n",
    "        for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "    ]\n",
    "    df[\"subsystem\"] = [\n",
    "        r.subsystem for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "    ]\n",
    "    df[\"category\"] = df[\"subsystem\"].replace(subsystem_to_category_dict)\n",
    "    df[\"proteins\"] = [\n",
    "        \";\".join(sorted([g.id for g in r.genes]))\n",
    "        for r in model.reactions.get_by_any(list(df[\"reactions\"].values))\n",
    "    ]\n",
    "    # Replace commas to prevent issues with CSV export\n",
    "    df[\"subsystem\"] = df[\"subsystem\"].apply(lambda x: x.replace(\",\", \"\"))\n",
    "    df[\"category\"] = df[\"category\"].apply(lambda x: x.replace(\",\", \"\"))\n",
    "    df[\"pvalue\"] = df[\"pvalue\"].apply(lambda x: round(x, 5))\n",
    "\n",
    "    df = df.set_index(\"reactions\")\n",
    "    if sort_by_subsystem:\n",
    "        df = df.sort_values(by=[\"category\", \"subsystem\", \"proteins\"])\n",
    "\n",
    "    df_meta = df.loc[:, metadata_columns].copy()\n",
    "    df_data = df.loc[:, ~df.columns.isin(df_meta.columns)].copy()\n",
    "\n",
    "    if use_group_means:\n",
    "        phenotypes_for_key = [k.split(\"_\")[-1] for k in key]\n",
    "        df_data = pd.concat(\n",
    "            [\n",
    "                df_data.loc[\n",
    "                    :,\n",
    "                    [\n",
    "                        phenotype_re.search(sample_id).group(\"phenotype\") == phenotype\n",
    "                        for sample_id in df_data.columns\n",
    "                    ],\n",
    "                ].mean(axis=1)\n",
    "                for phenotype in phenotypes_for_key\n",
    "            ],\n",
    "            axis=1,\n",
    "        )\n",
    "        df_data.columns = phenotypes_for_key\n",
    "    if standardize_by == \"mean\":\n",
    "        df_data = (\n",
    "            df_data.sub(df_data.mean(axis=1), axis=0)\n",
    "            .div(df_data.std(axis=1), axis=0)\n",
    "            .dropna(how=\"all\", axis=0)\n",
    "        )\n",
    "    elif standardize_by == \"median\":\n",
    "        df_data = (\n",
    "            (df_data.T - df_data.median(axis=1))\n",
    "            / (df_data.quantile(q=0.75, axis=1) - df_data.quantile(q=0.25, axis=1))\n",
    "        ).T\n",
    "    else:\n",
    "        pass\n",
    "    # Put dataframes back together for custom reordering\n",
    "    significant_dataframes[key] = df_data.merge(\n",
    "        df_meta, left_index=True, right_index=True\n",
    "    )\n",
    "    print(key)\n",
    "\n",
    "key = tuple(ordered_group_to_compare)\n",
    "# key = ('Pre_HumCan', 'Pre_A')\n",
    "significant_dataframes[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "ad7ac8fd-5bc2-412b-8c8e-0f418d73bdaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-0.5704479685135864 1.0348091417459664\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HumCan</th>\n",
       "      <th>A</th>\n",
       "      <th>MED</th>\n",
       "      <th>name</th>\n",
       "      <th>stoichiometry</th>\n",
       "      <th>proteins</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>subsystem</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>reactions</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>G6PDH2</th>\n",
       "      <td>1.034809</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>Glucose 6-phosphate dehydrogenase</td>\n",
       "      <td>beta-D-Glucose 6-phosphate + Nicotinamide aden...</td>\n",
       "      <td>G6pd2;G6pdx</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GND</th>\n",
       "      <td>1.034809</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>Phosphogluconate dehydrogenase</td>\n",
       "      <td>6-phospho-D-gluconate + Nicotinamide adenine d...</td>\n",
       "      <td>Pgd</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGL</th>\n",
       "      <td>1.034809</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>-0.517405</td>\n",
       "      <td>6-phosphogluconolactonase</td>\n",
       "      <td>6-phospho-D-glucono-1,5-lactone + H2O &lt;=&gt; 6-ph...</td>\n",
       "      <td>Pgls</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPI</th>\n",
       "      <td>0.386354</td>\n",
       "      <td>-0.389152</td>\n",
       "      <td>0.002797</td>\n",
       "      <td>Ribose-5-phosphate isomerase</td>\n",
       "      <td>D-Ribulose 5-phosphate &lt;=&gt; Alpha-D-Ribose 5-ph...</td>\n",
       "      <td>Rpia</td>\n",
       "      <td>0.01399</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GALK</th>\n",
       "      <td>0.681627</td>\n",
       "      <td>-0.340803</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>Galactokinase</td>\n",
       "      <td>ATP + galactose --&gt; ADP + Alpha-D-Galactose 1-...</td>\n",
       "      <td>Galk1;Galk2</td>\n",
       "      <td>0.01263</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAL1PP</th>\n",
       "      <td>0.681627</td>\n",
       "      <td>-0.340803</td>\n",
       "      <td>-0.340823</td>\n",
       "      <td>alpha-D-galactose 1-phosphatase</td>\n",
       "      <td>Alpha-D-Galactose 1-phosphate + H2O --&gt; galact...</td>\n",
       "      <td>Impa1;Impa2</td>\n",
       "      <td>0.01263</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G6PM</th>\n",
       "      <td>-0.570448</td>\n",
       "      <td>0.506711</td>\n",
       "      <td>0.063737</td>\n",
       "      <td>Glucose-6-phosphate isomerase (anomerization)</td>\n",
       "      <td>alpha-D-Glucose 6-phosphate &lt;=&gt; beta-D-Glucose...</td>\n",
       "      <td>Gpi1</td>\n",
       "      <td>0.00812</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGMTa</th>\n",
       "      <td>-0.562922</td>\n",
       "      <td>0.465742</td>\n",
       "      <td>0.097180</td>\n",
       "      <td>Phosphoglucomutase (alpha-D-glucose 1-phosphate)</td>\n",
       "      <td>D-Glucose 1-phosphate &lt;=&gt; alpha-D-Glucose 6-ph...</td>\n",
       "      <td>Pgm1;Pgm2</td>\n",
       "      <td>0.01351</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PGMT</th>\n",
       "      <td>-0.562922</td>\n",
       "      <td>0.465742</td>\n",
       "      <td>0.097180</td>\n",
       "      <td>Phosphoglucomutase</td>\n",
       "      <td>D-Glucose 1-phosphate &lt;=&gt; beta-D-Glucose 6-pho...</td>\n",
       "      <td>Pgm2</td>\n",
       "      <td>0.01351</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>Carbohydrate metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CO2t</th>\n",
       "      <td>-0.547445</td>\n",
       "      <td>0.567767</td>\n",
       "      <td>-0.020322</td>\n",
       "      <td>Transport of CO2</td>\n",
       "      <td>CO2 &lt;=&gt; CO2 (extracellular)</td>\n",
       "      <td>Aqp1;Rhag;Rhd</td>\n",
       "      <td>0.02036</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOt</th>\n",
       "      <td>-0.519685</td>\n",
       "      <td>0.555730</td>\n",
       "      <td>-0.036045</td>\n",
       "      <td>Transport of nitric oxide via diffusion</td>\n",
       "      <td>Nitric oxide &lt;=&gt; Nitric oxide (extracellular)</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2t</th>\n",
       "      <td>-0.519685</td>\n",
       "      <td>0.555730</td>\n",
       "      <td>-0.036045</td>\n",
       "      <td>Transport of O2</td>\n",
       "      <td>O2  &lt;=&gt; O2 (extracellular)</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>O2St</th>\n",
       "      <td>-0.519685</td>\n",
       "      <td>0.555730</td>\n",
       "      <td>-0.036045</td>\n",
       "      <td>Superoxide transport via diffusion</td>\n",
       "      <td>Superoxide &lt;=&gt; Superoxide (extracellular)</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2Ot</th>\n",
       "      <td>-0.519685</td>\n",
       "      <td>0.555730</td>\n",
       "      <td>-0.036045</td>\n",
       "      <td>Transport of H2O</td>\n",
       "      <td>H2O &lt;=&gt; H2O (extracellular)</td>\n",
       "      <td>Aqp1;Aqp3</td>\n",
       "      <td>0.02143</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH3t</th>\n",
       "      <td>-0.546947</td>\n",
       "      <td>0.567664</td>\n",
       "      <td>-0.020717</td>\n",
       "      <td>Transport of ammonia</td>\n",
       "      <td>Ammonia &lt;=&gt; Ammonia (extracellular)</td>\n",
       "      <td>Aqp1;Aqp3;Rhag;Rhd</td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4t</th>\n",
       "      <td>-0.546947</td>\n",
       "      <td>0.567664</td>\n",
       "      <td>-0.020717</td>\n",
       "      <td>Transport of ammonium</td>\n",
       "      <td>Ammonium &lt;=&gt; Ammonium (extracellular)</td>\n",
       "      <td>Aqp1;Aqp3;Rhag;Rhd</td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Transport extracellular</td>\n",
       "      <td>Transport reactions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OXOGAH</th>\n",
       "      <td>0.393407</td>\n",
       "      <td>0.112472</td>\n",
       "      <td>-0.505879</td>\n",
       "      <td>2-Oxoglutaramate amidohydrolase</td>\n",
       "      <td>2-oxoglutaramate + H2O --&gt; 2-Oxoglutarate + Am...</td>\n",
       "      <td>Nit2</td>\n",
       "      <td>0.02736</td>\n",
       "      <td>Alanine aspartate and glutamate metabolism</td>\n",
       "      <td>Amino acid metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NH4c</th>\n",
       "      <td>-0.546947</td>\n",
       "      <td>0.567664</td>\n",
       "      <td>-0.020717</td>\n",
       "      <td>Deprotonation of ammonium</td>\n",
       "      <td>Ammonium &lt;=&gt; H+ + Ammonia</td>\n",
       "      <td></td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Miscellaneous</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PROTNDA</th>\n",
       "      <td>-0.545926</td>\n",
       "      <td>0.567432</td>\n",
       "      <td>-0.021506</td>\n",
       "      <td>Nonenzymatic deamidation of L-asparaginyl prot...</td>\n",
       "      <td>H+ + L-asparaginyl-[protein] --&gt; Ammonium + [p...</td>\n",
       "      <td></td>\n",
       "      <td>0.02384</td>\n",
       "      <td>Protein modification</td>\n",
       "      <td>Other</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             HumCan         A       MED  \\\n",
       "reactions                                 \n",
       "G6PDH2     1.034809 -0.517405 -0.517405   \n",
       "GND        1.034809 -0.517405 -0.517405   \n",
       "PGL        1.034809 -0.517405 -0.517405   \n",
       "RPI        0.386354 -0.389152  0.002797   \n",
       "GALK       0.681627 -0.340803 -0.340823   \n",
       "GAL1PP     0.681627 -0.340803 -0.340823   \n",
       "G6PM      -0.570448  0.506711  0.063737   \n",
       "PGMTa     -0.562922  0.465742  0.097180   \n",
       "PGMT      -0.562922  0.465742  0.097180   \n",
       "CO2t      -0.547445  0.567767 -0.020322   \n",
       "NOt       -0.519685  0.555730 -0.036045   \n",
       "O2t       -0.519685  0.555730 -0.036045   \n",
       "O2St      -0.519685  0.555730 -0.036045   \n",
       "H2Ot      -0.519685  0.555730 -0.036045   \n",
       "NH3t      -0.546947  0.567664 -0.020717   \n",
       "NH4t      -0.546947  0.567664 -0.020717   \n",
       "OXOGAH     0.393407  0.112472 -0.505879   \n",
       "NH4c      -0.546947  0.567664 -0.020717   \n",
       "PROTNDA   -0.545926  0.567432 -0.021506   \n",
       "\n",
       "                                                        name  \\\n",
       "reactions                                                      \n",
       "G6PDH2                     Glucose 6-phosphate dehydrogenase   \n",
       "GND                           Phosphogluconate dehydrogenase   \n",
       "PGL                                6-phosphogluconolactonase   \n",
       "RPI                             Ribose-5-phosphate isomerase   \n",
       "GALK                                           Galactokinase   \n",
       "GAL1PP                       alpha-D-galactose 1-phosphatase   \n",
       "G6PM           Glucose-6-phosphate isomerase (anomerization)   \n",
       "PGMTa       Phosphoglucomutase (alpha-D-glucose 1-phosphate)   \n",
       "PGMT                                      Phosphoglucomutase   \n",
       "CO2t                                        Transport of CO2   \n",
       "NOt                  Transport of nitric oxide via diffusion   \n",
       "O2t                                          Transport of O2   \n",
       "O2St                      Superoxide transport via diffusion   \n",
       "H2Ot                                        Transport of H2O   \n",
       "NH3t                                    Transport of ammonia   \n",
       "NH4t                                   Transport of ammonium   \n",
       "OXOGAH                       2-Oxoglutaramate amidohydrolase   \n",
       "NH4c                               Deprotonation of ammonium   \n",
       "PROTNDA    Nonenzymatic deamidation of L-asparaginyl prot...   \n",
       "\n",
       "                                               stoichiometry  \\\n",
       "reactions                                                      \n",
       "G6PDH2     beta-D-Glucose 6-phosphate + Nicotinamide aden...   \n",
       "GND        6-phospho-D-gluconate + Nicotinamide adenine d...   \n",
       "PGL        6-phospho-D-glucono-1,5-lactone + H2O <=> 6-ph...   \n",
       "RPI        D-Ribulose 5-phosphate <=> Alpha-D-Ribose 5-ph...   \n",
       "GALK       ATP + galactose --> ADP + Alpha-D-Galactose 1-...   \n",
       "GAL1PP     Alpha-D-Galactose 1-phosphate + H2O --> galact...   \n",
       "G6PM       alpha-D-Glucose 6-phosphate <=> beta-D-Glucose...   \n",
       "PGMTa      D-Glucose 1-phosphate <=> alpha-D-Glucose 6-ph...   \n",
       "PGMT       D-Glucose 1-phosphate <=> beta-D-Glucose 6-pho...   \n",
       "CO2t                             CO2 <=> CO2 (extracellular)   \n",
       "NOt            Nitric oxide <=> Nitric oxide (extracellular)   \n",
       "O2t                               O2  <=> O2 (extracellular)   \n",
       "O2St               Superoxide <=> Superoxide (extracellular)   \n",
       "H2Ot                             H2O <=> H2O (extracellular)   \n",
       "NH3t                     Ammonia <=> Ammonia (extracellular)   \n",
       "NH4t                   Ammonium <=> Ammonium (extracellular)   \n",
       "OXOGAH     2-oxoglutaramate + H2O --> 2-Oxoglutarate + Am...   \n",
       "NH4c                               Ammonium <=> H+ + Ammonia   \n",
       "PROTNDA    H+ + L-asparaginyl-[protein] --> Ammonium + [p...   \n",
       "\n",
       "                     proteins   pvalue  \\\n",
       "reactions                                \n",
       "G6PDH2            G6pd2;G6pdx  0.00001   \n",
       "GND                       Pgd  0.00001   \n",
       "PGL                      Pgls  0.00001   \n",
       "RPI                      Rpia  0.01399   \n",
       "GALK              Galk1;Galk2  0.01263   \n",
       "GAL1PP            Impa1;Impa2  0.01263   \n",
       "G6PM                     Gpi1  0.00812   \n",
       "PGMTa               Pgm1;Pgm2  0.01351   \n",
       "PGMT                     Pgm2  0.01351   \n",
       "CO2t            Aqp1;Rhag;Rhd  0.02036   \n",
       "NOt                      Aqp1  0.02143   \n",
       "O2t                      Aqp1  0.02143   \n",
       "O2St                     Aqp1  0.02143   \n",
       "H2Ot                Aqp1;Aqp3  0.02143   \n",
       "NH3t       Aqp1;Aqp3;Rhag;Rhd  0.02384   \n",
       "NH4t       Aqp1;Aqp3;Rhag;Rhd  0.02384   \n",
       "OXOGAH                   Nit2  0.02736   \n",
       "NH4c                           0.02384   \n",
       "PROTNDA                        0.02384   \n",
       "\n",
       "                                            subsystem                 category  \n",
       "reactions                                                                       \n",
       "G6PDH2                      Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "GND                         Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "PGL                         Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "RPI                         Pentose phosphate pathway  Carbohydrate metabolism  \n",
       "GALK                             Galactose metabolism  Carbohydrate metabolism  \n",
       "GAL1PP                           Galactose metabolism  Carbohydrate metabolism  \n",
       "G6PM                     Glycolysis / Gluconeogenesis  Carbohydrate metabolism  \n",
       "PGMTa                    Glycolysis / Gluconeogenesis  Carbohydrate metabolism  \n",
       "PGMT                     Glycolysis / Gluconeogenesis  Carbohydrate metabolism  \n",
       "CO2t                          Transport extracellular      Transport reactions  \n",
       "NOt                           Transport extracellular      Transport reactions  \n",
       "O2t                           Transport extracellular      Transport reactions  \n",
       "O2St                          Transport extracellular      Transport reactions  \n",
       "H2Ot                          Transport extracellular      Transport reactions  \n",
       "NH3t                          Transport extracellular      Transport reactions  \n",
       "NH4t                          Transport extracellular      Transport reactions  \n",
       "OXOGAH     Alanine aspartate and glutamate metabolism    Amino acid metabolism  \n",
       "NH4c                                    Miscellaneous                    Other  \n",
       "PROTNDA                          Protein modification                    Other  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# key = tuple(ordered_group_to_compare)\n",
    "# df = significant_dataframes[key]\n",
    "# df = df.loc[:, ~df.columns.isin([\"proteins\", \"pvalue\", \"subsystem\", \"category\", \"name\", \"stoichiometry\"])].T\n",
    "# df.index = pd.MultiIndex.from_tuples([(x, phenotype_re.search(x).group(1)) for x in df.index])\n",
    "# df = df.groupby(level=1).mean().T\n",
    "# df = df[[\"HumCan\", \"A\", \"MED\"]]\n",
    "\n",
    "# print(df.min().min(), df.max().max())\n",
    "# df = df.merge(df_meta, left_index=True, right_index=True)\n",
    "# ordered_subsystems = [\n",
    "#     \"Pentose phosphate pathway\",\n",
    "#     \"Galactose metabolism\",\n",
    "#     \"Glycolysis / Gluconeogenesis\",\n",
    "#     \"Transport extracellular\",\n",
    "#     \"Alanine aspartate and glutamate metabolism\",\n",
    "#     \"Miscellaneous\",\n",
    "#     \"Protein modification\",\n",
    "# ]\n",
    "# df = pd.concat(\n",
    "#     [\n",
    "#         df[df[\"subsystem\"] == subsystem].sort_values(by=\"pvalue\")\n",
    "#         for subsystem in ordered_subsystems\n",
    "#     ],\n",
    "#     axis=0,\n",
    "# )\n",
    "# significant_dataframes[key] = df\n",
    "\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "59521b3e-ab82-42aa-aae1-9df6fe676d39",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "('Pre_HumCan', 'Pre_A')",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyError\u001b[39m                                  Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[27]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m key = (\u001b[33m\"\u001b[39m\u001b[33mPre_HumCan\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mPre_A\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m df_main = \u001b[43msignificant_dataframes\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n\u001b[32m      3\u001b[39m df_main.loc[[\u001b[33m\"\u001b[39m\u001b[33mO2t\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mO2St\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mNOt\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mH2Ot\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mNH4t\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mNH3t\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mCO2t\u001b[39m\u001b[33m\"\u001b[39m], \u001b[33m\"\u001b[39m\u001b[33msubsystem\u001b[39m\u001b[33m\"\u001b[39m] = (\n\u001b[32m      4\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mTransport gas\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      5\u001b[39m )\n\u001b[32m      6\u001b[39m df_main.loc[\n\u001b[32m      7\u001b[39m     [\n\u001b[32m      8\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mFAt_hs_3_0\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m   (...)\u001b[39m\u001b[32m     20\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33msubsystem\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     21\u001b[39m ] = \u001b[33m\"\u001b[39m\u001b[33mTransport other\u001b[39m\u001b[33m\"\u001b[39m\n",
      "\u001b[31mKeyError\u001b[39m: ('Pre_HumCan', 'Pre_A')"
     ]
    }
   ],
   "source": [
    "# key = (\"Pre_HumCan\", \"Pre_A\")\n",
    "# df_main = significant_dataframes[key]\n",
    "# df_main.loc[[\"O2t\", \"O2St\", \"NOt\", \"H2Ot\", \"NH4t\", \"NH3t\", \"CO2t\"], \"subsystem\"] = (\n",
    "#     \"Transport gas\"\n",
    "# )\n",
    "# df_main.loc[\n",
    "#     [\n",
    "#         \"FAt_hs_3_0\",\n",
    "#         \"3DGt\",\n",
    "#         \"HEMATINABCte\",\n",
    "#         \"E217BGLCRABCte\",\n",
    "#         \"5FLURAABCte\",\n",
    "#         \"URATEABCte\",\n",
    "#         \"Clt\",\n",
    "#         \"NO3t\",\n",
    "#         \"LDOPAt\",\n",
    "#         \"5FLURAt\",\n",
    "#         \"E217BGLCRte\",\n",
    "#     ],\n",
    "#     \"subsystem\",\n",
    "# ] = \"Transport other\"\n",
    "# ordered_subsystems = [\n",
    "#     \"Pentose phosphate pathway\",\n",
    "#     \"Galactose metabolism\",\n",
    "#     \"Glycolysis / Gluconeogenesis\",\n",
    "#     \"Pyruvate metabolism\",\n",
    "#     \"Protein modification\",\n",
    "#     \"Reactive species formation and detoxification\",\n",
    "#     \"Miscellaneous\",\n",
    "#     \"Cysteine and methionine metabolism\",\n",
    "#     \"Glutathione metabolism\",\n",
    "#     \"Phenylalanine tyrosine and tryptophan metabolism\",\n",
    "#     \"Transport other\",\n",
    "#     \"Transport gas\",\n",
    "#     \"Transport extracellular\",\n",
    "# ]\n",
    "# df_main = pd.concat(\n",
    "#     [\n",
    "#         (\n",
    "#             df_main[df_main[\"subsystem\"] == subsystem]\n",
    "#             if subsystem != \"Protein modification\"\n",
    "#             else df_main[df_main[\"subsystem\"] == subsystem].iloc[::-1]\n",
    "#         )\n",
    "#         for subsystem in ordered_subsystems\n",
    "#     ],\n",
    "#     axis=0,\n",
    "# )\n",
    "# significant_dataframes[key] = df_main.copy()\n",
    "# df_main"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610dc29f-f123-457b-8d29-0d7461a189f7",
   "metadata": {},
   "source": [
    "## Export results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6cd72da-0f5b-47c3-bdf4-0cfdd3c54bd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "ftype = \"csv\"\n",
    "for key, df_main in significant_dataframes.items():\n",
    "    df_meta = df_main.loc[:, metadata_columns].copy()\n",
    "    df_data = df_main.loc[:, ~df_main.columns.isin(df_meta.columns)].copy()\n",
    "    for df_type, df in zip([\"data\", \"meta\"], [df_data, df_meta]):\n",
    "        filename = \"_\".join(\n",
    "            [\"MannWhiteney\" if mannwhitneyu_for_pairs and len(key) == 2 else \"Kruskal\"]\n",
    "            + [g.split(\"_\")[-1] for g in key]\n",
    "            + [df_type]\n",
    "        )\n",
    "        if use_group_means:\n",
    "            filename += \"_mean\"\n",
    "        filename = group_results_dirpath_dict[group_name] / filename\n",
    "        df.to_csv(\n",
    "            f\"{filename}.{ftype}\", sep=\"\\t\" if ftype == \"tsv\" else \",\", index=True\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbe84078-6a73-41ee-80b2-568bf120839c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
