{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e3208d0-39f0-4d23-8d86-6de6fa19fd8b",
   "metadata": {},
   "source": [
    "# Create Proteome-Constrained RBC model via OVERLAY workflow \n",
    "This notebook facilitates the construction of a proteome constrained model (\"pcModel\") via the OVERLAY methodology.\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d968a45f-c732-43c5-bfc4-1114f8d8cb01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2025-11-21\n",
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.1\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "beautifulsoup4                       4.12.3\n",
      "bio                                   1.6.2\n",
      "cobra                                0.29.0\n",
      "depinfo                               2.2.0\n",
      "kaleido                               0.2.1\n",
      "matplotlib                            3.8.2\n",
      "memote                               0.17.0\n",
      "networkx                              3.2.1\n",
      "notebook                              7.0.7\n",
      "openpyxl                              3.1.2\n",
      "pandas                                2.2.0\n",
      "pre-commit                            3.6.0\n",
      "pyvis                                 0.3.2\n",
      "rbc-gem-utils[database,network,vis] missing\n",
      "requests                             2.31.0\n",
      "scipy                                1.12.0\n",
      "seaborn                              0.13.2\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip        23.3.1\n",
      "setuptools 68.2.2\n",
      "wheel      0.41.2\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Darwin  24.3.0-x86_64\n",
      "CPython        3.11.7\n"
     ]
    }
   ],
   "source": [
    "import itertools\n",
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "\n",
    "import gurobipy as gp\n",
    "import pandas as pd\n",
    "from rbc_gem_utils import (COBRA_CONFIGURATION, GEM_NAME, build_string,\n",
    "                           get_annotation_df, read_cobra_model, show_versions,\n",
    "                           split_string, write_cobra_model)\n",
    "from rbc_gem_utils.analysis.overlay import (\n",
    "    ATTR_SUBCLASS_DICT, DEFAULT_COMPARTMENT_CONSTRAINT_PREFIX,\n",
    "    DEFAULT_CONCENTRATION_BOUND, DEFAULT_CONSTRAINT_PREFIX,\n",
    "    DEFAULT_ENZYME_FORWARD_SUFFIX, DEFAULT_ENZYME_REVERSE_SUFFIX,\n",
    "    DEFAULT_ENZYME_TOTAL_SUFFIX, DEFAULT_ISOFORM_CONSTRAINT_PREFIX,\n",
    "    DEFAULT_KEFF, Enzyme, EnzymeDilution, Protein, ProteinDilution,\n",
    "    ProteomeBudget, ProteomeBudgetDilution, add_relaxation_budget,\n",
    "    construct_pcmodel_from_tables, create_complex_table, create_enzyme_table,\n",
    "    create_protein_table, create_sequence_table)\n",
    "from rbc_gem_utils.util import strip_plural\n",
    "\n",
    "gp.setParam(\"OutputFlag\", 0)\n",
    "gp.setParam(\"LogToConsole\", 0)\n",
    "\n",
    "# Show versions of notebook\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2c19f01-8e9f-448e-bcbb-64fb84358e1b",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28d822cd-8d55-40c1-81d3-34165bf0f24b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION.solver = \"gurobi\"\n",
    "COBRA_CONFIGURATION.bounds = (-1e3, 1e3)\n",
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae972b5d-74b6-4530-a55f-14b45fefd4c6",
   "metadata": {},
   "source": [
    "## Load RBC model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66ac203d-e751-4070-9920-52277287b79d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14cdec250</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14cdec250>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_name = \"RBComics\"\n",
    "data_path = Path(\"data\").resolve()\n",
    "models_path = Path(\"models\").resolve()\n",
    "figures_path = Path(\"figures\").resolve()\n",
    "dataset_path = Path(dataset_name).resolve()\n",
    "\n",
    "imagetype = \"svg\"\n",
    "transparent = True\n",
    "save_figures = True\n",
    "\n",
    "ftype = \"xml\"\n",
    "model = read_cobra_model(models_path / f\"{GEM_NAME.replace('-', '_')}.{ftype}\")\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c38a3d73-7dee-4165-ac00-a21f45d98ea1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes      820\n",
      "uniprot    820\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Q9NPC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS1</td>\n",
       "      <td>P49588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AASDHPPT</td>\n",
       "      <td>Q9NRN7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA1</td>\n",
       "      <td>O95477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>Q8IZY2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>815</th>\n",
       "      <td>YES1</td>\n",
       "      <td>P07947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>816</th>\n",
       "      <td>ZDHHC2</td>\n",
       "      <td>Q9UIJ5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>817</th>\n",
       "      <td>ZDHHC20</td>\n",
       "      <td>Q5W0Z9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>818</th>\n",
       "      <td>ZDHHC3</td>\n",
       "      <td>Q9NYG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>819</th>\n",
       "      <td>ZDHHC5</td>\n",
       "      <td>Q9C0B5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>820 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genes uniprot\n",
       "0      A4GALT  Q9NPC4\n",
       "1       AARS1  P49588\n",
       "2    AASDHPPT  Q9NRN7\n",
       "3       ABCA1  O95477\n",
       "4       ABCA7  Q8IZY2\n",
       "..        ...     ...\n",
       "815      YES1  P07947\n",
       "816    ZDHHC2  Q9UIJ5\n",
       "817   ZDHHC20  Q5W0Z9\n",
       "818    ZDHHC3  Q9NYG2\n",
       "819    ZDHHC5  Q9C0B5\n",
       "\n",
       "[820 rows x 2 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_type = \"genes\"\n",
    "mapping_key = \"uniprot\"\n",
    "annotation_cols = [mapping_key]\n",
    "\n",
    "df_model_mappings = (\n",
    "    get_annotation_df(model.genes, annotation_cols)\n",
    "    .rename({\"id\": annotation_type}, axis=1)\n",
    "    .dropna(subset=[mapping_key])\n",
    ")\n",
    "for col in df_model_mappings.columns:\n",
    "    df_model_mappings[col] = df_model_mappings[col].apply(lambda x: split_string(x))\n",
    "    df_model_mappings = df_model_mappings.explode(col).drop_duplicates().dropna()\n",
    "df_model_mappings = df_model_mappings.sort_values(annotation_type)\n",
    "\n",
    "print(df_model_mappings.nunique(dropna=True))\n",
    "df_model_mappings = df_model_mappings.reset_index(drop=True)\n",
    "df_model_mappings"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dae5bf0-27c9-43c3-9e99-b0278f4e5491",
   "metadata": {},
   "source": [
    "## Assemble data for PC-model\n",
    "### Load protein data\n",
    "#### Protein amino acid sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "77b24e89-a6d8-40d1-b830-032681929c07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['P00167' 'P00387' 'P02730' 'P05089' 'P08397' 'P19367' 'P22303' 'P30613'\n",
      " 'Q8N0V5' 'Q9H0P0']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot</th>\n",
       "      <th>uniprot.isoform</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>canonical</th>\n",
       "      <th>erythroid</th>\n",
       "      <th>backup</th>\n",
       "      <th>avoid</th>\n",
       "      <th>keywords.erythroid</th>\n",
       "      <th>keywords.backup</th>\n",
       "      <th>keywords.avoid</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sequence.length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>A0AVT1-1</td>\n",
       "      <td>A0AVT1-1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...</td>\n",
       "      <td>1052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>A0AVT1-2</td>\n",
       "      <td>A0AVT1-2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MLKNFALLGVGTSKEKGMITVTDPDLIEKSNLNRQFLFRPHHIQKP...</td>\n",
       "      <td>578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>A0AVT1-3</td>\n",
       "      <td>A0AVT1-3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...</td>\n",
       "      <td>389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0AVT1</td>\n",
       "      <td>A0AVT1-4</td>\n",
       "      <td>A0AVT1-4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...</td>\n",
       "      <td>340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A4D126</td>\n",
       "      <td>A4D126-1</td>\n",
       "      <td>A4D126-1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MEAGPPGSARPAEPGPCLSGQRGADHTASASLQSVAGTEPGRHPQA...</td>\n",
       "      <td>451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1848</th>\n",
       "      <td>Q9Y6M4</td>\n",
       "      <td>Q9Y6M4-3</td>\n",
       "      <td>Q9Y6M4-3</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MENKKKDKDKSDDRMARPSGRSGHNTRGTGSSSSGVLMVGPNFRVG...</td>\n",
       "      <td>423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1849</th>\n",
       "      <td>Q9Y6M4</td>\n",
       "      <td>Q9Y6M4-4</td>\n",
       "      <td>Q9Y6M4-4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MENKKKDKDKSDDRMARPSGRSGHNTRGTGSSSSGVLMVGPNFRVG...</td>\n",
       "      <td>424</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850</th>\n",
       "      <td>Q9Y6M4</td>\n",
       "      <td>Q9Y6M4-5</td>\n",
       "      <td>Q9Y6M4-5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MKSRAPQLHLEYRFYKQLGSGDGIPQVYYFGPCGKYNAMVLELLGP...</td>\n",
       "      <td>348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>Q9Y6M4</td>\n",
       "      <td>Q9Y6M4-6</td>\n",
       "      <td>Q9Y6M4-6</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MVLELLGPSLEDLFDLCDRTFSLKTVLMIAIQLISRMEYVHSKNLI...</td>\n",
       "      <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1852</th>\n",
       "      <td>Q9Y6M5</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>Q9Y6M5</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>MGCWGRNRGRLLCMLALTFMFMVLEVVVSRVTSSLAMLSDSFHMLS...</td>\n",
       "      <td>507</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1853 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     uniprot uniprot.isoform sequence.id  canonical  erythroid  backup  avoid  \\\n",
       "0     A0AVT1        A0AVT1-1    A0AVT1-1       True      False   False  False   \n",
       "1     A0AVT1        A0AVT1-2    A0AVT1-2      False      False   False  False   \n",
       "2     A0AVT1        A0AVT1-3    A0AVT1-3      False      False   False  False   \n",
       "3     A0AVT1        A0AVT1-4    A0AVT1-4      False      False   False  False   \n",
       "4     A4D126        A4D126-1    A4D126-1       True      False   False  False   \n",
       "...      ...             ...         ...        ...        ...     ...    ...   \n",
       "1848  Q9Y6M4        Q9Y6M4-3    Q9Y6M4-3      False      False   False  False   \n",
       "1849  Q9Y6M4        Q9Y6M4-4    Q9Y6M4-4      False      False   False  False   \n",
       "1850  Q9Y6M4        Q9Y6M4-5    Q9Y6M4-5      False      False   False  False   \n",
       "1851  Q9Y6M4        Q9Y6M4-6    Q9Y6M4-6      False      False   False  False   \n",
       "1852  Q9Y6M5            <NA>      Q9Y6M5       True      False   False  False   \n",
       "\n",
       "     keywords.erythroid keywords.backup keywords.avoid  \\\n",
       "0                  <NA>            <NA>           <NA>   \n",
       "1                  <NA>            <NA>           <NA>   \n",
       "2                  <NA>            <NA>           <NA>   \n",
       "3                  <NA>            <NA>           <NA>   \n",
       "4                  <NA>            <NA>           <NA>   \n",
       "...                 ...             ...            ...   \n",
       "1848               <NA>            <NA>           <NA>   \n",
       "1849               <NA>            <NA>           <NA>   \n",
       "1850               <NA>            <NA>           <NA>   \n",
       "1851               <NA>            <NA>           <NA>   \n",
       "1852               <NA>            <NA>           <NA>   \n",
       "\n",
       "                                               sequence  sequence.length  \n",
       "0     MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...             1052  \n",
       "1     MLKNFALLGVGTSKEKGMITVTDPDLIEKSNLNRQFLFRPHHIQKP...              578  \n",
       "2     MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...              389  \n",
       "3     MEGSEPVAAHQGEEASCSSWGTGSTNKNLPIMSTASVEIDDALYSR...              340  \n",
       "4     MEAGPPGSARPAEPGPCLSGQRGADHTASASLQSVAGTEPGRHPQA...              451  \n",
       "...                                                 ...              ...  \n",
       "1848  MENKKKDKDKSDDRMARPSGRSGHNTRGTGSSSSGVLMVGPNFRVG...              423  \n",
       "1849  MENKKKDKDKSDDRMARPSGRSGHNTRGTGSSSSGVLMVGPNFRVG...              424  \n",
       "1850  MKSRAPQLHLEYRFYKQLGSGDGIPQVYYFGPCGKYNAMVLELLGP...              348  \n",
       "1851  MVLELLGPSLEDLFDLCDRTFSLKTVLMIAIQLISRMEYVHSKNLI...              311  \n",
       "1852  MGCWGRNRGRLLCMLALTFMFMVLEVVVSRVTSSLAMLSDSFHMLS...              507  \n",
       "\n",
       "[1853 rows x 12 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_isoforms_sequences = pd.read_csv(\n",
    "    data_path / \"uniprot_isoforms_sequences.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(pd.NA)\n",
    "print(df_isoforms_sequences[df_isoforms_sequences[\"erythroid\"]][\"uniprot\"].unique())\n",
    "df_isoforms_sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc4d1592-c3e0-41a2-b52f-cfb4699da14b",
   "metadata": {},
   "source": [
    "#### Determine protein isoforms and associated sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "441ed5b2-a8a5-4f8d-ab82-590d0f52e2b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "canonical    820\n",
      "erythroid     10\n",
      "backup        93\n",
      "avoid         22\n",
      "dtype: int64\n",
      "Total: 887\n",
      "\n",
      "canonical    807\n",
      "erythroid     10\n",
      "backup        53\n",
      "avoid          2\n",
      "dtype: int64\n",
      "Total: 835\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Q9NPC4</td>\n",
       "      <td>Q9NPC4</td>\n",
       "      <td>MSKPPDLLLRLLRGAPRQRVCTLFIIGFKFTFFVSIMIYWHVVGEP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AARS1</td>\n",
       "      <td>P49588</td>\n",
       "      <td>P49588-1</td>\n",
       "      <td>MDSTLTASEIRQRFIDFFKRNEHTYVHSSATIPLDDPTLLFANAGM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AASDHPPT</td>\n",
       "      <td>Q9NRN7</td>\n",
       "      <td>Q9NRN7-1</td>\n",
       "      <td>MVFPAKRFCLVPSMEGVRWAFSCGTWLPSRAEWLLAVRSIQPEEKE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCA1</td>\n",
       "      <td>O95477</td>\n",
       "      <td>O95477</td>\n",
       "      <td>MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>Q8IZY2</td>\n",
       "      <td>Q8IZY2-1</td>\n",
       "      <td>MAFWTQLMLLLWKNFMYRRRQPVQLLVELLWPLFLFFILVAVRHSH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830</th>\n",
       "      <td>YES1</td>\n",
       "      <td>P07947</td>\n",
       "      <td>P07947</td>\n",
       "      <td>MGCIKSKENKSPAIKYRPENTPEPVSTSVSHYGAEPTTVSPCPSSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>831</th>\n",
       "      <td>ZDHHC2</td>\n",
       "      <td>Q9UIJ5</td>\n",
       "      <td>Q9UIJ5</td>\n",
       "      <td>MAPSGPGSSARRRCRRVLYWIPVVFITLLLGWSYYAYAIQLCIVSM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>832</th>\n",
       "      <td>ZDHHC20</td>\n",
       "      <td>Q5W0Z9</td>\n",
       "      <td>Q5W0Z9-1</td>\n",
       "      <td>MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVFTIFGNE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>ZDHHC3</td>\n",
       "      <td>Q9NYG2</td>\n",
       "      <td>Q9NYG2-1</td>\n",
       "      <td>MMLIPTHHFRNIERKPEYLQPEKCVPPPYPGPVGTMWFIRDGCGIA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>ZDHHC5</td>\n",
       "      <td>Q9C0B5</td>\n",
       "      <td>Q9C0B5-1</td>\n",
       "      <td>MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLYVSPA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>835 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genes uniprot sequence.id  \\\n",
       "0      A4GALT  Q9NPC4      Q9NPC4   \n",
       "1       AARS1  P49588    P49588-1   \n",
       "2    AASDHPPT  Q9NRN7    Q9NRN7-1   \n",
       "3       ABCA1  O95477      O95477   \n",
       "4       ABCA7  Q8IZY2    Q8IZY2-1   \n",
       "..        ...     ...         ...   \n",
       "830      YES1  P07947      P07947   \n",
       "831    ZDHHC2  Q9UIJ5      Q9UIJ5   \n",
       "832   ZDHHC20  Q5W0Z9    Q5W0Z9-1   \n",
       "833    ZDHHC3  Q9NYG2    Q9NYG2-1   \n",
       "834    ZDHHC5  Q9C0B5    Q9C0B5-1   \n",
       "\n",
       "                                              sequence  \n",
       "0    MSKPPDLLLRLLRGAPRQRVCTLFIIGFKFTFFVSIMIYWHVVGEP...  \n",
       "1    MDSTLTASEIRQRFIDFFKRNEHTYVHSSATIPLDDPTLLFANAGM...  \n",
       "2    MVFPAKRFCLVPSMEGVRWAFSCGTWLPSRAEWLLAVRSIQPEEKE...  \n",
       "3    MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...  \n",
       "4    MAFWTQLMLLLWKNFMYRRRQPVQLLVELLWPLFLFFILVAVRHSH...  \n",
       "..                                                 ...  \n",
       "830  MGCIKSKENKSPAIKYRPENTPEPVSTSVSHYGAEPTTVSPCPSSS...  \n",
       "831  MAPSGPGSSARRRCRRVLYWIPVVFITLLLGWSYYAYAIQLCIVSM...  \n",
       "832  MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVFTIFGNE...  \n",
       "833  MMLIPTHHFRNIERKPEYLQPEKCVPPPYPGPVGTMWFIRDGCGIA...  \n",
       "834  MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLYVSPA...  \n",
       "\n",
       "[835 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Erythroid first, then canonical to enable drop_duplicates to keep erythroid over canonical\n",
    "df_model_isoforms_sequences = (\n",
    "    pd.concat(\n",
    "        (\n",
    "            df_isoforms_sequences[df_isoforms_sequences[\"erythroid\"]],\n",
    "            df_isoforms_sequences[df_isoforms_sequences[\"canonical\"]],\n",
    "            df_isoforms_sequences[df_isoforms_sequences[\"backup\"]],\n",
    "        ),\n",
    "        axis=0,\n",
    "    )\n",
    "    .fillna(pd.NA)\n",
    "    .drop_duplicates()\n",
    "    .sort_values(\n",
    "        [\"uniprot\", \"erythroid\", \"uniprot.isoform\"], ascending=[True, False, True]\n",
    "    )\n",
    ")\n",
    "\n",
    "print(\n",
    "    df_model_isoforms_sequences[[\"canonical\", \"erythroid\", \"backup\", \"avoid\"]].sum(\n",
    "        axis=0\n",
    "    )\n",
    ")\n",
    "print(f\"Total: {len(df_model_isoforms_sequences)}\")\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"uniprot\",\n",
    "        \"uniprot.isoform\",\n",
    "        \"sequence.id\",\n",
    "        \"sequence\",\n",
    "        \"sequence.length\",\n",
    "        \"canonical\",\n",
    "        \"erythroid\",\n",
    "        \"backup\",\n",
    "        \"avoid\",\n",
    "    ],\n",
    "].reset_index(drop=True)\n",
    "\n",
    "\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences.copy()\n",
    "df_model_isoforms_sequences[\"keep\"] = df_model_isoforms_sequences[\"canonical\"].values\n",
    "to_avoid = df_model_isoforms_sequences[df_model_isoforms_sequences[\"avoid\"]][\n",
    "    \"uniprot\"\n",
    "].to_dict()\n",
    "df_model_isoforms_sequences.loc[\n",
    "    list(to_avoid),\n",
    "    \"keep\",\n",
    "] = False\n",
    "\n",
    "df_possible_backups = df_model_isoforms_sequences[\n",
    "    df_model_isoforms_sequences[\"uniprot\"].isin(list(to_avoid.values()))\n",
    "]\n",
    "df_possible_backups = df_possible_backups[~df_possible_backups[\"avoid\"]]\n",
    "df_model_isoforms_sequences.loc[\n",
    "    list(df_possible_backups.index),\n",
    "    \"keep\",\n",
    "] = True\n",
    "df_model_isoforms_sequences.loc[\n",
    "    df_model_isoforms_sequences[df_model_isoforms_sequences[\"erythroid\"]].index,\n",
    "    \"keep\",\n",
    "] = True\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences[\n",
    "    df_model_isoforms_sequences[\"keep\"]\n",
    "]\n",
    "\n",
    "lost_ids = set(df_isoforms_sequences[\"uniprot\"].unique()).difference(\n",
    "    set(df_model_isoforms_sequences[\"uniprot\"].unique())\n",
    ")\n",
    "if lost_ids:\n",
    "    lost_ids = df_isoforms_sequences[df_isoforms_sequences[\"uniprot\"].isin(lost_ids)]\n",
    "    df_model_isoforms_sequences = pd.concat(\n",
    "        (df_model_isoforms_sequences, lost_ids[lost_ids[\"canonical\"]]), axis=0\n",
    "    )\n",
    "print()\n",
    "print(\n",
    "    df_model_isoforms_sequences[[\"canonical\", \"erythroid\", \"backup\", \"avoid\"]].sum(\n",
    "        axis=0\n",
    "    )\n",
    ")\n",
    "print(f\"Total: {len(df_model_isoforms_sequences)}\")\n",
    "\n",
    "df_model_isoforms_sequences = df_model_isoforms_sequences.loc[\n",
    "    :, [\"uniprot\", \"sequence.id\", \"sequence\"]\n",
    "].copy()\n",
    "df_sequence_data = (\n",
    "    df_model_mappings.merge(\n",
    "        df_model_isoforms_sequences, left_on=\"uniprot\", right_on=\"uniprot\"\n",
    "    )\n",
    "    .loc[:, [\"genes\", \"uniprot\", \"sequence.id\", \"sequence\"]]\n",
    "    .copy()\n",
    ")\n",
    "df_sequence_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3e2ea22-65c2-40b1-866e-956a86de3ed9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of proteins: 12\n",
      "Number of isoforms: 27\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>protein.id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACOT7</td>\n",
       "      <td>O00154</td>\n",
       "      <td>O00154-4</td>\n",
       "      <td>MSGPDVETPSAIQICRIMRPDDANVAGNVHGGTILKMIEEAGAIIS...</td>\n",
       "      <td>ACOT7_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARG1</td>\n",
       "      <td>P05089</td>\n",
       "      <td>P05089-2</td>\n",
       "      <td>MSAKSRTIGIIGAPFSKGQPRGGVEEGPTVLRKAGLLEKLKEQVTQ...</td>\n",
       "      <td>ARG1_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CYB5A</td>\n",
       "      <td>P00167</td>\n",
       "      <td>P00167-2</td>\n",
       "      <td>MAEQSDEAVKYYTLEEIQKHNHSKSTWLILHHKVYDLTKFLEEHPG...</td>\n",
       "      <td>CYB5A_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CYB5R3</td>\n",
       "      <td>P00387</td>\n",
       "      <td>P00387-2</td>\n",
       "      <td>MKLFQRSTPAITLESPDIKYPLRLIDREIISHDTRRFRFALPSPQH...</td>\n",
       "      <td>CYB5R3_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCNT2</td>\n",
       "      <td>Q8N0V5</td>\n",
       "      <td>Q8N0V5-3</td>\n",
       "      <td>MNFWRYCFFAFTLLSVVIFVRFYSSQLSPPKSYEKLNSSSERYFRK...</td>\n",
       "      <td>GCNT2_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>830</th>\n",
       "      <td>SLC12A7</td>\n",
       "      <td>Q9Y666</td>\n",
       "      <td>Q9Y666-1</td>\n",
       "      <td>MPTNFTVVPVEAHADGGGDETAERTEAPGTPEGPEPERPSPGDGNP...</td>\n",
       "      <td>SLC12A7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>831</th>\n",
       "      <td>SLC22A7</td>\n",
       "      <td>Q9Y694</td>\n",
       "      <td>Q9Y694-1</td>\n",
       "      <td>MGFEELLEQVGGFGPFQLRNVALLALPRVLLPLHFLLPIFLAAVPA...</td>\n",
       "      <td>SLC22A7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>832</th>\n",
       "      <td>CLIC4</td>\n",
       "      <td>Q9Y696</td>\n",
       "      <td>Q9Y696</td>\n",
       "      <td>MALSMPLNGLKEEDKEPLIELFVKAGSDGESIGNCPFSQRLFMILW...</td>\n",
       "      <td>CLIC4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>833</th>\n",
       "      <td>CSNK1G3</td>\n",
       "      <td>Q9Y6M4</td>\n",
       "      <td>Q9Y6M4-1</td>\n",
       "      <td>MENKKKDKDKSDDRMARPSGRSGHNTRGTGSSSSGVLMVGPNFRVG...</td>\n",
       "      <td>CSNK1G3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>834</th>\n",
       "      <td>SLC30A1</td>\n",
       "      <td>Q9Y6M5</td>\n",
       "      <td>Q9Y6M5</td>\n",
       "      <td>MGCWGRNRGRLLCMLALTFMFMVLEVVVSRVTSSLAMLSDSFHMLS...</td>\n",
       "      <td>SLC30A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>820 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genes uniprot sequence.id  \\\n",
       "0      ACOT7  O00154    O00154-4   \n",
       "2       ARG1  P05089    P05089-2   \n",
       "4      CYB5A  P00167    P00167-2   \n",
       "6     CYB5R3  P00387    P00387-2   \n",
       "8      GCNT2  Q8N0V5    Q8N0V5-3   \n",
       "..       ...     ...         ...   \n",
       "830  SLC12A7  Q9Y666    Q9Y666-1   \n",
       "831  SLC22A7  Q9Y694    Q9Y694-1   \n",
       "832    CLIC4  Q9Y696      Q9Y696   \n",
       "833  CSNK1G3  Q9Y6M4    Q9Y6M4-1   \n",
       "834  SLC30A1  Q9Y6M5      Q9Y6M5   \n",
       "\n",
       "                                              sequence protein.id  \n",
       "0    MSGPDVETPSAIQICRIMRPDDANVAGNVHGGTILKMIEEAGAIIS...    ACOT7_4  \n",
       "2    MSAKSRTIGIIGAPFSKGQPRGGVEEGPTVLRKAGLLEKLKEQVTQ...     ARG1_2  \n",
       "4    MAEQSDEAVKYYTLEEIQKHNHSKSTWLILHHKVYDLTKFLEEHPG...    CYB5A_2  \n",
       "6    MKLFQRSTPAITLESPDIKYPLRLIDREIISHDTRRFRFALPSPQH...   CYB5R3_2  \n",
       "8    MNFWRYCFFAFTLLSVVIFVRFYSSQLSPPKSYEKLNSSSERYFRK...    GCNT2_3  \n",
       "..                                                 ...        ...  \n",
       "830  MPTNFTVVPVEAHADGGGDETAERTEAPGTPEGPEPERPSPGDGNP...    SLC12A7  \n",
       "831  MGFEELLEQVGGFGPFQLRNVALLALPRVLLPLHFLLPIFLAAVPA...    SLC22A7  \n",
       "832  MALSMPLNGLKEEDKEPLIELFVKAGSDGESIGNCPFSQRLFMILW...      CLIC4  \n",
       "833  MENKKKDKDKSDDRMARPSGRSGHNTRGTGSSSSGVLMVGPNFRVG...    CSNK1G3  \n",
       "834  MGCWGRNRGRLLCMLALTFMFMVLEVVVSRVTSSLAMLSDSFHMLS...    SLC30A1  \n",
       "\n",
       "[820 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_key = \"uniprot\"\n",
    "protein_id_key = (\n",
    "    \"sequence.id.genes\"  # genes, uniprot, sequence.id, or sequence.id.genes are best,\n",
    ")\n",
    "unique_gene_to_protein_map = True\n",
    "isoform_transform = False\n",
    "df_copy_numbers_data = None\n",
    "df_protein_data = create_sequence_table(\n",
    "    df_sequences=df_sequence_data,\n",
    "    mapping_key=mapping_key,\n",
    "    isoform_transform=isoform_transform,\n",
    ")\n",
    "ordered_isoform_ids = df_sequence_data[df_sequence_data[\"uniprot\"].duplicated(False)][\n",
    "    \"sequence.id\"\n",
    "]\n",
    "df_isoforms = df_protein_data[\n",
    "    df_protein_data[\"sequence.id\"].isin(ordered_isoform_ids)\n",
    "].copy()\n",
    "print(f\"Number of proteins: {len(df_isoforms[mapping_key].unique())}\")\n",
    "print(f\"Number of isoforms: {len(df_isoforms['sequence.id'].unique())}\")\n",
    "df_protein_data = df_protein_data.set_index(\"sequence.id\")\n",
    "df_protein_data = pd.concat(\n",
    "    (\n",
    "        df_protein_data.loc[ordered_isoform_ids],\n",
    "        df_protein_data.loc[df_protein_data.index.difference(ordered_isoform_ids)],\n",
    "    ),\n",
    "    axis=0,\n",
    ")\n",
    "df_protein_data = df_protein_data.reset_index(drop=False)\n",
    "df_protein_data = df_protein_data.loc[\n",
    "    :, [\"genes\", \"uniprot\", \"sequence.id\", \"sequence\"]\n",
    "].copy()\n",
    "# print(df_isoforms[mapping_key])\n",
    "if protein_id_key == \"sequence.id.genes\":\n",
    "    protein_id_key = \"protein.id\"\n",
    "    sequence_id_updates = df_model_mappings.set_index(\"uniprot\")[\"genes\"].to_dict()\n",
    "    df_protein_data[\"protein.id\"] = df_protein_data[\"sequence.id\"].apply(\n",
    "        lambda seq_id: \"_\".join(\n",
    "            [sequence_id_updates.get(x, x) for x in seq_id.split(\"-\")]\n",
    "        )\n",
    "    )\n",
    "    df_isoforms[\"protein.id\"] = df_isoforms[\"sequence.id\"].apply(\n",
    "        lambda seq_id: \"_\".join(\n",
    "            [sequence_id_updates.get(x, x) for x in seq_id.split(\"-\")]\n",
    "        )\n",
    "    )\n",
    "    ids_to_fix = df_protein_data[\n",
    "        ~df_protein_data[\"sequence.id\"].isin(df_isoforms[\"sequence.id\"])\n",
    "    ].index\n",
    "    df_protein_data.loc[ids_to_fix, \"protein.id\"] = df_protein_data.loc[\n",
    "        ids_to_fix, \"protein.id\"\n",
    "    ].apply(lambda x: x.split(\"_\")[0])\n",
    "\n",
    "# Use to remove duplicates\n",
    "if unique_gene_to_protein_map:\n",
    "    df_protein_data = df_protein_data.drop_duplicates(\n",
    "        subset=[\"uniprot\"],\n",
    "        keep=\"first\",\n",
    "    )\n",
    "    protein_id_key = \"genes\"\n",
    "\n",
    "df_protein_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb9eff5e-0f50-463e-a82c-40c4565e16b6",
   "metadata": {},
   "source": [
    "###  List all unique proteins, complexes, and enzymes\n",
    "#### Option 1: Initialize draft tables\n",
    "1. The draft tables are created and used to initialize the draft PC-model.\n",
    "    * The protein table can be used to initialize proteins and their molar weight ($\\textbf{d}$ vector).\n",
    "    * The complex table can be used to initialize complexes with their subunit stoichiometry ($\\textbf{C}$ matrix).\n",
    "        * All stoichiometric coefficients are initialized at a value of one.\n",
    "    * The enzyme table can be used to initialize enzymes with their effective rate constants ($\\textbf{K}_\\mathrm{eff}$ matrix).\n",
    "        * All $k_\\mathrm{eff}$ values are initialized at average rate constant of 65 $s^{-1}$  (or equivalently, 234000 $hr^{-1})$.\n",
    "\n",
    "2. The draft tables are made to be facilitate curation and data replacement. Therefore, the draft PC-model is exported with the draft tables. \n",
    "3. A refined PC-model can be created using the curated tables. \n",
    "\n",
    "#### Option 2: Load tables from files\n",
    "4. The formation of a draft model can be skipped if the curated tables already exist. They can be loaded."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "18560139-b484-4255-bffa-deed41fa153a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pcmodel_tables = {}\n",
    "replace_compartments = {\n",
    "    # Cytosol:extracellular --> plasma membrane\n",
    "    \"c\": \"c\",\n",
    "    \"ce\": \"pm\",\n",
    "    \"e\": \"e\",  # Most extracellular reactions that occur are due to proteins bound to the external side of them membrane.\n",
    "}\n",
    "\n",
    "# Convert all protein compartments to one compartment\n",
    "simplify_compartments = True\n",
    "prefix = True\n",
    "optional_columns = True\n",
    "\n",
    "# Enzyme values for new tables\n",
    "max_weight_fraction = 100\n",
    "enzyme_keff_base = DEFAULT_KEFF\n",
    "enzyme_forward_suffix = DEFAULT_ENZYME_FORWARD_SUFFIX\n",
    "enzyme_reverse_suffix = DEFAULT_ENZYME_REVERSE_SUFFIX\n",
    "enzyme_total_suffix = DEFAULT_ENZYME_TOTAL_SUFFIX\n",
    "\n",
    "dict_of_id_keys = {\n",
    "    \"proteins\": protein_id_key,\n",
    "    \"complexes\": None,\n",
    "    \"enzymes\": \"reactions\",\n",
    "}\n",
    "\n",
    "# Provide filepaths to speed up creation process files, comment out to generate from scratch\n",
    "filepaths = {\n",
    "    \"proteins\": data_path / \"pcmodel_proteins.tsv\",\n",
    "    \"complexes\": data_path / \"pcmodel_complexes.tsv\",\n",
    "    \"enzymes\": data_path / \"pcmodel_enzymes.tsv\",\n",
    "    # \"complex_keffs\": data_path / \"pcmodel_complex_keffs.tsv\",\n",
    "    # \"enzyme_keffs\": data_path / \"pcmodel_enzyme_keffs.tsv\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "792a4744-149f-4fe3-a274-bfb7c2cb97b1",
   "metadata": {},
   "source": [
    "##### Proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "acb342b9-c941-47ab-b80f-5b11fec4fb24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded from main RBC-GEM file\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>sequence.id</th>\n",
       "      <th>sequence</th>\n",
       "      <th>compartment</th>\n",
       "      <th>proteins</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>protein_A4GALT</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Q9NPC4</td>\n",
       "      <td>Q9NPC4</td>\n",
       "      <td>MSKPPDLLLRLLRGAPRQRVCTLFIIGFKFTFFVSIMIYWHVVGEP...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_A4GALT_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_AARS1</th>\n",
       "      <td>AARS1</td>\n",
       "      <td>P49588</td>\n",
       "      <td>P49588-1</td>\n",
       "      <td>MDSTLTASEIRQRFIDFFKRNEHTYVHSSATIPLDDPTLLFANAGM...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_AARS1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_AASDHPPT</th>\n",
       "      <td>AASDHPPT</td>\n",
       "      <td>Q9NRN7</td>\n",
       "      <td>Q9NRN7-1</td>\n",
       "      <td>MVFPAKRFCLVPSMEGVRWAFSCGTWLPSRAEWLLAVRSIQPEEKE...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_AASDHPPT_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_ABCA1</th>\n",
       "      <td>ABCA1</td>\n",
       "      <td>O95477</td>\n",
       "      <td>O95477</td>\n",
       "      <td>MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_ABCA1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_ABCA7</th>\n",
       "      <td>ABCA7</td>\n",
       "      <td>Q8IZY2</td>\n",
       "      <td>Q8IZY2-1</td>\n",
       "      <td>MAFWTQLMLLLWKNFMYRRRQPVQLLVELLWPLFLFFILVAVRHSH...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_ABCA7_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_YES1</th>\n",
       "      <td>YES1</td>\n",
       "      <td>P07947</td>\n",
       "      <td>P07947</td>\n",
       "      <td>MGCIKSKENKSPAIKYRPENTPEPVSTSVSHYGAEPTTVSPCPSSS...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_YES1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_ZDHHC2</th>\n",
       "      <td>ZDHHC2</td>\n",
       "      <td>Q9UIJ5</td>\n",
       "      <td>Q9UIJ5</td>\n",
       "      <td>MAPSGPGSSARRRCRRVLYWIPVVFITLLLGWSYYAYAIQLCIVSM...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_ZDHHC2_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_ZDHHC20</th>\n",
       "      <td>ZDHHC20</td>\n",
       "      <td>Q5W0Z9</td>\n",
       "      <td>Q5W0Z9-1</td>\n",
       "      <td>MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVFTIFGNE...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_ZDHHC20_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_ZDHHC3</th>\n",
       "      <td>ZDHHC3</td>\n",
       "      <td>Q9NYG2</td>\n",
       "      <td>Q9NYG2-1</td>\n",
       "      <td>MMLIPTHHFRNIERKPEYLQPEKCVPPPYPGPVGTMWFIRDGCGIA...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_ZDHHC3_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_ZDHHC5</th>\n",
       "      <td>ZDHHC5</td>\n",
       "      <td>Q9C0B5</td>\n",
       "      <td>Q9C0B5-1</td>\n",
       "      <td>MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLYVSPA...</td>\n",
       "      <td>pc</td>\n",
       "      <td>protein_ZDHHC5_pc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>820 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     genes uniprot sequence.id  \\\n",
       "protein                                          \n",
       "protein_A4GALT      A4GALT  Q9NPC4      Q9NPC4   \n",
       "protein_AARS1        AARS1  P49588    P49588-1   \n",
       "protein_AASDHPPT  AASDHPPT  Q9NRN7    Q9NRN7-1   \n",
       "protein_ABCA1        ABCA1  O95477      O95477   \n",
       "protein_ABCA7        ABCA7  Q8IZY2    Q8IZY2-1   \n",
       "...                    ...     ...         ...   \n",
       "protein_YES1          YES1  P07947      P07947   \n",
       "protein_ZDHHC2      ZDHHC2  Q9UIJ5      Q9UIJ5   \n",
       "protein_ZDHHC20    ZDHHC20  Q5W0Z9    Q5W0Z9-1   \n",
       "protein_ZDHHC3      ZDHHC3  Q9NYG2    Q9NYG2-1   \n",
       "protein_ZDHHC5      ZDHHC5  Q9C0B5    Q9C0B5-1   \n",
       "\n",
       "                                                           sequence  \\\n",
       "protein                                                               \n",
       "protein_A4GALT    MSKPPDLLLRLLRGAPRQRVCTLFIIGFKFTFFVSIMIYWHVVGEP...   \n",
       "protein_AARS1     MDSTLTASEIRQRFIDFFKRNEHTYVHSSATIPLDDPTLLFANAGM...   \n",
       "protein_AASDHPPT  MVFPAKRFCLVPSMEGVRWAFSCGTWLPSRAEWLLAVRSIQPEEKE...   \n",
       "protein_ABCA1     MACWPQLRLLLWKNLTFRRRQTCQLLLEVAWPLFIFLILISVRLSY...   \n",
       "protein_ABCA7     MAFWTQLMLLLWKNFMYRRRQPVQLLVELLWPLFLFFILVAVRHSH...   \n",
       "...                                                             ...   \n",
       "protein_YES1      MGCIKSKENKSPAIKYRPENTPEPVSTSVSHYGAEPTTVSPCPSSS...   \n",
       "protein_ZDHHC2    MAPSGPGSSARRRCRRVLYWIPVVFITLLLGWSYYAYAIQLCIVSM...   \n",
       "protein_ZDHHC20   MAPWTLWRCCQRVVGWVPVLFITFVVVWSYYAYVVELCVFTIFGNE...   \n",
       "protein_ZDHHC3    MMLIPTHHFRNIERKPEYLQPEKCVPPPYPGPVGTMWFIRDGCGIA...   \n",
       "protein_ZDHHC5    MPAESGKRFKPSKYVPVSAAAIFLVGATTLFFAFTCPGLSLYVSPA...   \n",
       "\n",
       "                 compartment             proteins  \n",
       "protein                                            \n",
       "protein_A4GALT            pc    protein_A4GALT_pc  \n",
       "protein_AARS1             pc     protein_AARS1_pc  \n",
       "protein_AASDHPPT          pc  protein_AASDHPPT_pc  \n",
       "protein_ABCA1             pc     protein_ABCA1_pc  \n",
       "protein_ABCA7             pc     protein_ABCA7_pc  \n",
       "...                      ...                  ...  \n",
       "protein_YES1              pc      protein_YES1_pc  \n",
       "protein_ZDHHC2            pc    protein_ZDHHC2_pc  \n",
       "protein_ZDHHC20           pc   protein_ZDHHC20_pc  \n",
       "protein_ZDHHC3            pc    protein_ZDHHC3_pc  \n",
       "protein_ZDHHC5            pc    protein_ZDHHC5_pc  \n",
       "\n",
       "[820 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"proteins\"\n",
    "# Otherwise try using main RBC-GEM files to make model proteins\n",
    "try:\n",
    "    df_proteins = pd.read_csv(filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    print(\"Loaded from main RBC-GEM file\")\n",
    "except (FileNotFoundError, KeyError):\n",
    "    # Otherwise, make from scratch\n",
    "    df_proteins = create_protein_table(\n",
    "        model,\n",
    "        df_protein_data=df_protein_data,\n",
    "        id_key=dict_of_id_keys.get(table_type),\n",
    "        prefix=prefix,\n",
    "        optional_columns=optional_columns,\n",
    "        annotation_columns=[\n",
    "            \"uniprot\",\n",
    "        ],\n",
    "        replace_compartments=replace_compartments,\n",
    "    )\n",
    "    print(\"Created new table\")\n",
    "    # Create column for identifiers if None exists, or if compartments were replaced\n",
    "    if not isoform_transform:\n",
    "        df_proteins[df_proteins[table_type].duplicated(False)]\n",
    "else:\n",
    "    df_proteins = df_protein_data.merge(\n",
    "        df_proteins[[\"uniprot\", \"compartment\"]],\n",
    "        left_on=\"uniprot\",\n",
    "        right_on=\"uniprot\",\n",
    "        how=\"left\",\n",
    "    )\n",
    "    df_proteins[\"protein\"] = df_proteins[protein_id_key].apply(lambda x: f\"protein_{x}\")\n",
    "    df_proteins = df_proteins.drop(\"protein.id\", axis=1)\n",
    "    df_proteins = df_proteins[\n",
    "        df_proteins[\"genes\"].isin(model.genes.list_attr(\"id\"))\n",
    "    ].reset_index(drop=True)\n",
    "\n",
    "if simplify_compartments:\n",
    "    df_proteins = df_proteins.groupby([\"genes\", \"protein\"]).agg(\n",
    "        lambda values: \";\".join(\n",
    "            [str(value) for value in list(values.dropna().unique())]\n",
    "        )\n",
    "    )\n",
    "    df_proteins[\"compartment\"] = \"pc\"\n",
    "    df_proteins = df_proteins.reset_index(drop=False)\n",
    "\n",
    "df_proteins[\"proteins\"] = df_proteins[[strip_plural(table_type), \"compartment\"]].apply(\n",
    "    lambda x: \"_\".join(x.values), axis=1\n",
    ")\n",
    "df_proteins = df_proteins.set_index(strip_plural(table_type))\n",
    "pcmodel_tables[table_type] = df_proteins.copy()\n",
    "df_proteins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "610036dd-56de-4d9b-8a5b-147d569624ed",
   "metadata": {},
   "source": [
    "##### Complexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7190d981-1b98-486b-9e06-8c9344d6fd63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded from main RBC-GEM file\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>subunits</th>\n",
       "      <th>compartment</th>\n",
       "      <th>reactions</th>\n",
       "      <th>genes</th>\n",
       "      <th>coefficients</th>\n",
       "      <th>cofactors</th>\n",
       "      <th>notes</th>\n",
       "      <th>complexes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>complex</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_A4GALT</th>\n",
       "      <td>protein_A4GALT_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>A4GALTc</td>\n",
       "      <td>A4GALT</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_A4GALT_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_AARS1</th>\n",
       "      <td>protein_AARS1_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ALATRS</td>\n",
       "      <td>AARS1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_AARS1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_AASDHPPT</th>\n",
       "      <td>protein_AASDHPPT_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>AASDHPPT</td>\n",
       "      <td>AASDHPPT</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_AASDHPPT_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_HOMOTETRAMER_ABCA1</th>\n",
       "      <td>protein_ABCA1_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PCFLOPt;PSFLOPt;SMFLOPt</td>\n",
       "      <td>ABCA1</td>\n",
       "      <td>4</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_HOMOTETRAMER_ABCA1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_ABCA7</th>\n",
       "      <td>protein_ABCA7_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PCFLOPt;PEFLOPt;PSFLOPt;SMFLOPt</td>\n",
       "      <td>ABCA7</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_ABCA7_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_YES1</th>\n",
       "      <td>protein_YES1_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTYK</td>\n",
       "      <td>YES1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_YES1_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_ZDHHC20</th>\n",
       "      <td>protein_ZDHHC20_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>ZDHHC20</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>ASSUMED MONOMER</td>\n",
       "      <td>cplx_MONOMER_ZDHHC20_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_ZDHHC2</th>\n",
       "      <td>protein_ZDHHC2_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>ZDHHC2</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_ZDHHC2_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_ZDHHC3</th>\n",
       "      <td>protein_ZDHHC3_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>ZDHHC3</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_ZDHHC3_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cplx_MONOMER_ZDHHC5</th>\n",
       "      <td>protein_ZDHHC5_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0</td>\n",
       "      <td>ZDHHC5</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>cplx_MONOMER_ZDHHC5_pc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>887 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    subunits compartment  \\\n",
       "complex                                                    \n",
       "cplx_MONOMER_A4GALT        protein_A4GALT_pc          pc   \n",
       "cplx_MONOMER_AARS1          protein_AARS1_pc          pc   \n",
       "cplx_MONOMER_AASDHPPT    protein_AASDHPPT_pc          pc   \n",
       "cplx_HOMOTETRAMER_ABCA1     protein_ABCA1_pc          pc   \n",
       "cplx_MONOMER_ABCA7          protein_ABCA7_pc          pc   \n",
       "...                                      ...         ...   \n",
       "cplx_MONOMER_YES1            protein_YES1_pc          pc   \n",
       "cplx_MONOMER_ZDHHC20      protein_ZDHHC20_pc          pc   \n",
       "cplx_MONOMER_ZDHHC2        protein_ZDHHC2_pc          pc   \n",
       "cplx_MONOMER_ZDHHC3        protein_ZDHHC3_pc          pc   \n",
       "cplx_MONOMER_ZDHHC5        protein_ZDHHC5_pc          pc   \n",
       "\n",
       "                                                         reactions     genes  \\\n",
       "complex                                                                        \n",
       "cplx_MONOMER_A4GALT                                        A4GALTc    A4GALT   \n",
       "cplx_MONOMER_AARS1                                          ALATRS     AARS1   \n",
       "cplx_MONOMER_AASDHPPT                                     AASDHPPT  AASDHPPT   \n",
       "cplx_HOMOTETRAMER_ABCA1                    PCFLOPt;PSFLOPt;SMFLOPt     ABCA1   \n",
       "cplx_MONOMER_ABCA7                 PCFLOPt;PEFLOPt;PSFLOPt;SMFLOPt     ABCA7   \n",
       "...                                                            ...       ...   \n",
       "cplx_MONOMER_YES1                                           PROTYK      YES1   \n",
       "cplx_MONOMER_ZDHHC20     PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0   ZDHHC20   \n",
       "cplx_MONOMER_ZDHHC2      PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0    ZDHHC2   \n",
       "cplx_MONOMER_ZDHHC3      PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0    ZDHHC3   \n",
       "cplx_MONOMER_ZDHHC5      PROTCSAT_14_0;PROTCSAT_16_0;PROTCSAT_18_0    ZDHHC5   \n",
       "\n",
       "                        coefficients cofactors            notes  \\\n",
       "complex                                                           \n",
       "cplx_MONOMER_A4GALT                1            ASSUMED MONOMER   \n",
       "cplx_MONOMER_AARS1                 1                              \n",
       "cplx_MONOMER_AASDHPPT              1                              \n",
       "cplx_HOMOTETRAMER_ABCA1            4                              \n",
       "cplx_MONOMER_ABCA7                 1            ASSUMED MONOMER   \n",
       "...                              ...       ...              ...   \n",
       "cplx_MONOMER_YES1                  1            ASSUMED MONOMER   \n",
       "cplx_MONOMER_ZDHHC20               1            ASSUMED MONOMER   \n",
       "cplx_MONOMER_ZDHHC2                1                              \n",
       "cplx_MONOMER_ZDHHC3                1                              \n",
       "cplx_MONOMER_ZDHHC5                1                              \n",
       "\n",
       "                                          complexes  \n",
       "complex                                              \n",
       "cplx_MONOMER_A4GALT          cplx_MONOMER_A4GALT_pc  \n",
       "cplx_MONOMER_AARS1            cplx_MONOMER_AARS1_pc  \n",
       "cplx_MONOMER_AASDHPPT      cplx_MONOMER_AASDHPPT_pc  \n",
       "cplx_HOMOTETRAMER_ABCA1  cplx_HOMOTETRAMER_ABCA1_pc  \n",
       "cplx_MONOMER_ABCA7            cplx_MONOMER_ABCA7_pc  \n",
       "...                                             ...  \n",
       "cplx_MONOMER_YES1              cplx_MONOMER_YES1_pc  \n",
       "cplx_MONOMER_ZDHHC20        cplx_MONOMER_ZDHHC20_pc  \n",
       "cplx_MONOMER_ZDHHC2          cplx_MONOMER_ZDHHC2_pc  \n",
       "cplx_MONOMER_ZDHHC3          cplx_MONOMER_ZDHHC3_pc  \n",
       "cplx_MONOMER_ZDHHC5          cplx_MONOMER_ZDHHC5_pc  \n",
       "\n",
       "[887 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"complexes\"\n",
    "try:\n",
    "    df_complexes = pd.read_csv(filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    print(\"Loaded from main RBC-GEM file\")\n",
    "except (FileNotFoundError, KeyError):\n",
    "    genes_to_proteins = (\n",
    "        pcmodel_tables[\"proteins\"]\n",
    "        .groupby([\"genes\"], as_index=True)[\"proteins\"]\n",
    "        .agg(lambda x: build_string(list(x)))\n",
    "        .to_dict()\n",
    "    )\n",
    "    cofactor_genes = {}\n",
    "    # Create table\n",
    "    df_complexes = create_complex_table(\n",
    "        model,\n",
    "        genes_to_proteins=genes_to_proteins,\n",
    "        cofactor_genes=cofactor_genes,\n",
    "        id_key=dict_of_id_keys.get(table_type),\n",
    "        prefix=prefix,\n",
    "        optional_columns=optional_columns,\n",
    "        annotation_columns=[\n",
    "            # \"uniprot\"\n",
    "        ],\n",
    "        replace_compartments=replace_compartments,\n",
    "    )\n",
    "    print(\"Created new table\")\n",
    "\n",
    "else:\n",
    "    df_complexes = df_complexes.drop(\"molar_mass\", axis=1)\n",
    "    df_complexes = df_complexes[\n",
    "        df_complexes[\"genes\"].apply(\n",
    "            lambda genes: all([model.genes.has_id(gene) for gene in genes.split(\";\")])\n",
    "        )\n",
    "    ]\n",
    "    df_complexes[\"reactions\"] = df_complexes[\"reactions\"].apply(\n",
    "        lambda reactions: \";\".join(\n",
    "            [r for r in reactions.split(\";\") if model.reactions.has_id(r)]\n",
    "        )\n",
    "    )\n",
    "    df_complexes = df_complexes[df_complexes[\"reactions\"] != \"\"]\n",
    "    df_complexes = df_complexes.loc[\n",
    "        :,\n",
    "        [\n",
    "            \"complex\",\n",
    "            \"subunits\",\n",
    "            \"compartment\",\n",
    "            \"reactions\",\n",
    "            \"genes\",\n",
    "            \"coefficients\",\n",
    "            \"cofactors\",\n",
    "            \"notes\",\n",
    "        ],\n",
    "    ]\n",
    "# Address isoform mapping to complexes\n",
    "isoforms_map = defaultdict(list)\n",
    "complex_name_update = defaultdict(list)\n",
    "for x in df_proteins[df_proteins[\"genes\"].duplicated(False)].index:\n",
    "    isoforms_map[x.rsplit(\"_\", maxsplit=1)[0]].append(x)\n",
    "    complex_name_update[x.rsplit(\"_\", maxsplit=1)[0].replace(\"protein_\", \"\")].append(\n",
    "        x.replace(\"protein_\", \"\")\n",
    "    )\n",
    "df_isoforms_complexes = df_complexes[\n",
    "    df_complexes[\"subunits\"].apply(\n",
    "        lambda proteins: bool(set(isoforms_map).intersection(proteins.split(\";\")))\n",
    "    )\n",
    "]\n",
    "df_updated_rows = []\n",
    "for _, row in df_isoforms_complexes.iterrows():\n",
    "    complex_names = [\n",
    "        complex_name\n",
    "        for complex_name in itertools.product(\n",
    "            *[complex_name_update.get(c, [c]) for c in row[\"complex\"].split(\"_\")]\n",
    "        )\n",
    "    ]\n",
    "    combos = [\n",
    "        list(combo)\n",
    "        for combo in itertools.product(\n",
    "            *[\n",
    "                isoforms_map.get(protein, [protein])\n",
    "                for protein in row[\"subunits\"].split(\";\")\n",
    "            ]\n",
    "        )\n",
    "    ]\n",
    "\n",
    "    for complex_name, combo in zip(complex_names, combos):\n",
    "        new_row = row.to_dict()\n",
    "        new_row[\"complex\"] = \"_\".join(complex_name)\n",
    "        new_row[\"subunits\"] = \";\".join(combo)\n",
    "        df_updated_rows.append(new_row)\n",
    "\n",
    "df_complexes = pd.concat(\n",
    "    (\n",
    "        df_complexes[~df_complexes.index.isin(df_isoforms_complexes.index)],\n",
    "        pd.DataFrame(df_updated_rows),\n",
    "    ),\n",
    "    axis=0,\n",
    ")\n",
    "\n",
    "\n",
    "if simplify_compartments:\n",
    "    df_complexes = df_complexes.groupby([\"subunits\", \"complex\"]).agg(\n",
    "        lambda values: \";\".join(\n",
    "            [str(value) for value in list(values.dropna().unique())]\n",
    "        )\n",
    "    )\n",
    "    df_complexes[\"compartment\"] = \"pc\"\n",
    "    df_complexes = df_complexes.reset_index(drop=False)\n",
    "\n",
    "df_complexes[\"complexes\"] = df_complexes[\n",
    "    [strip_plural(table_type), \"compartment\"]\n",
    "].apply(lambda x: \"_\".join(x.values), axis=1)\n",
    "df_complexes[\"subunits\"] = df_complexes[[\"subunits\", \"compartment\"]].apply(\n",
    "    lambda values: \";\".join(\n",
    "        [\n",
    "            \"_\".join((x, values[\"compartment\"])) if not x.endswith(\"_pc\") else x\n",
    "            for x in values[\"subunits\"].split(\";\")\n",
    "        ]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "df_complexes = df_complexes.set_index(strip_plural(table_type))\n",
    "pcmodel_tables[table_type] = df_complexes.copy()\n",
    "df_complexes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5388d72-79df-4444-886c-82ca4ff0d7ba",
   "metadata": {},
   "source": [
    "##### Enzymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3ec3f525-31d4-41a4-bd1e-7b0c6242f572",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded from main RBC-GEM file\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>complexes</th>\n",
       "      <th>compartment</th>\n",
       "      <th>reactions</th>\n",
       "      <th>direction</th>\n",
       "      <th>enzymes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO3ARGH_fwd</th>\n",
       "      <td>cplx_COMPLEX_AGT_ANPEP_pc;cplx_COMPLEX_AGT_DPP...</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO3ARGH</td>\n",
       "      <td>forward</td>\n",
       "      <td>enzyme_ANGIO3ARGH_fwd_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO3ARGH_rev</th>\n",
       "      <td>cplx_COMPLEX_AGT_ANPEP_pc;cplx_COMPLEX_AGT_DPP...</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO3ARGH</td>\n",
       "      <td>reverse</td>\n",
       "      <td>enzyme_ANGIO3ARGH_rev_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2ASPH_fwd</th>\n",
       "      <td>cplx_COMPLEX_AGT_DNPEP_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO2ASPH</td>\n",
       "      <td>forward</td>\n",
       "      <td>enzyme_ANGIO2ASPH_fwd_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2ASPH_rev</th>\n",
       "      <td>cplx_COMPLEX_AGT_DNPEP_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO2ASPH</td>\n",
       "      <td>reverse</td>\n",
       "      <td>enzyme_ANGIO2ASPH_rev_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_ANGIO2DDP_fwd</th>\n",
       "      <td>cplx_COMPLEX_AGT_DPP3_pc</td>\n",
       "      <td>pc</td>\n",
       "      <td>ANGIO2DDP</td>\n",
       "      <td>forward</td>\n",
       "      <td>enzyme_ANGIO2DDP_fwd_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_14_0_rev</th>\n",
       "      <td>cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_14_0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>enzyme_PROTCSAT_14_0_rev_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_16_0_fwd</th>\n",
       "      <td>cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_16_0</td>\n",
       "      <td>forward</td>\n",
       "      <td>enzyme_PROTCSAT_16_0_fwd_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_16_0_rev</th>\n",
       "      <td>cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_16_0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>enzyme_PROTCSAT_16_0_rev_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_18_0_fwd</th>\n",
       "      <td>cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_18_0</td>\n",
       "      <td>forward</td>\n",
       "      <td>enzyme_PROTCSAT_18_0_fwd_pc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enzyme_PROTCSAT_18_0_rev</th>\n",
       "      <td>cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...</td>\n",
       "      <td>pc</td>\n",
       "      <td>PROTCSAT_18_0</td>\n",
       "      <td>reverse</td>\n",
       "      <td>enzyme_PROTCSAT_18_0_rev_pc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4362 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  complexes  \\\n",
       "enzyme                                                                        \n",
       "enzyme_ANGIO3ARGH_fwd     cplx_COMPLEX_AGT_ANPEP_pc;cplx_COMPLEX_AGT_DPP...   \n",
       "enzyme_ANGIO3ARGH_rev     cplx_COMPLEX_AGT_ANPEP_pc;cplx_COMPLEX_AGT_DPP...   \n",
       "enzyme_ANGIO2ASPH_fwd                             cplx_COMPLEX_AGT_DNPEP_pc   \n",
       "enzyme_ANGIO2ASPH_rev                             cplx_COMPLEX_AGT_DNPEP_pc   \n",
       "enzyme_ANGIO2DDP_fwd                               cplx_COMPLEX_AGT_DPP3_pc   \n",
       "...                                                                     ...   \n",
       "enzyme_PROTCSAT_14_0_rev  cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...   \n",
       "enzyme_PROTCSAT_16_0_fwd  cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...   \n",
       "enzyme_PROTCSAT_16_0_rev  cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...   \n",
       "enzyme_PROTCSAT_18_0_fwd  cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...   \n",
       "enzyme_PROTCSAT_18_0_rev  cplx_MONOMER_ZDHHC2_pc;cplx_MONOMER_ZDHHC20_pc...   \n",
       "\n",
       "                         compartment      reactions direction  \\\n",
       "enzyme                                                          \n",
       "enzyme_ANGIO3ARGH_fwd             pc     ANGIO3ARGH   forward   \n",
       "enzyme_ANGIO3ARGH_rev             pc     ANGIO3ARGH   reverse   \n",
       "enzyme_ANGIO2ASPH_fwd             pc     ANGIO2ASPH   forward   \n",
       "enzyme_ANGIO2ASPH_rev             pc     ANGIO2ASPH   reverse   \n",
       "enzyme_ANGIO2DDP_fwd              pc      ANGIO2DDP   forward   \n",
       "...                              ...            ...       ...   \n",
       "enzyme_PROTCSAT_14_0_rev          pc  PROTCSAT_14_0   reverse   \n",
       "enzyme_PROTCSAT_16_0_fwd          pc  PROTCSAT_16_0   forward   \n",
       "enzyme_PROTCSAT_16_0_rev          pc  PROTCSAT_16_0   reverse   \n",
       "enzyme_PROTCSAT_18_0_fwd          pc  PROTCSAT_18_0   forward   \n",
       "enzyme_PROTCSAT_18_0_rev          pc  PROTCSAT_18_0   reverse   \n",
       "\n",
       "                                              enzymes  \n",
       "enzyme                                                 \n",
       "enzyme_ANGIO3ARGH_fwd        enzyme_ANGIO3ARGH_fwd_pc  \n",
       "enzyme_ANGIO3ARGH_rev        enzyme_ANGIO3ARGH_rev_pc  \n",
       "enzyme_ANGIO2ASPH_fwd        enzyme_ANGIO2ASPH_fwd_pc  \n",
       "enzyme_ANGIO2ASPH_rev        enzyme_ANGIO2ASPH_rev_pc  \n",
       "enzyme_ANGIO2DDP_fwd          enzyme_ANGIO2DDP_fwd_pc  \n",
       "...                                               ...  \n",
       "enzyme_PROTCSAT_14_0_rev  enzyme_PROTCSAT_14_0_rev_pc  \n",
       "enzyme_PROTCSAT_16_0_fwd  enzyme_PROTCSAT_16_0_fwd_pc  \n",
       "enzyme_PROTCSAT_16_0_rev  enzyme_PROTCSAT_16_0_rev_pc  \n",
       "enzyme_PROTCSAT_18_0_fwd  enzyme_PROTCSAT_18_0_fwd_pc  \n",
       "enzyme_PROTCSAT_18_0_rev  enzyme_PROTCSAT_18_0_rev_pc  \n",
       "\n",
       "[4362 rows x 5 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_type = \"enzymes\"\n",
    "try:\n",
    "    df_enzymes = pd.read_csv(filepaths[table_type], sep=\"\\t\", index_col=None)\n",
    "    print(\"Loaded from main RBC-GEM file\")\n",
    "except (FileNotFoundError, KeyError):\n",
    "    complexes_to_reactions = (\n",
    "        pcmodel_tables[\"complexes\"].set_index(\"complexes\")[\"reactions\"].to_dict()\n",
    "    )\n",
    "    df_enzymes = create_enzyme_table(\n",
    "        model,\n",
    "        complexes_to_reactions=complexes_to_reactions,\n",
    "        enzyme_keff_base=enzyme_keff_base,\n",
    "        enzyme_forward_suffix=enzyme_forward_suffix,\n",
    "        enzyme_reverse_suffix=enzyme_reverse_suffix,\n",
    "        id_key=dict_of_id_keys.get(table_type),\n",
    "        prefix=prefix,\n",
    "        optional_columns=optional_columns,\n",
    "        annotation_columns=[\n",
    "            # \"uniprot\"\n",
    "        ],\n",
    "        replace_compartments=replace_compartments,\n",
    "    )\n",
    "    print(\"Created new table\")\n",
    "    if replace_compartments:\n",
    "        df_enzymes[\"compartment\"] = df_enzymes[\"compartment\"].replace(\n",
    "            replace_compartments\n",
    "        )\n",
    "else:\n",
    "    df_enzymes = df_enzymes.loc[\n",
    "        :, [\"complexes\", \"compartment\", \"reactions\", \"enzyme\", \"direction\"]\n",
    "    ]\n",
    "\n",
    "if simplify_compartments:\n",
    "    df_enzymes = df_enzymes.groupby([\"complexes\", \"enzyme\"]).agg(\n",
    "        lambda values: \";\".join(\n",
    "            [str(value) for value in list(values.dropna().unique())]\n",
    "        )\n",
    "    )\n",
    "    df_enzymes[\"compartment\"] = \"pc\"\n",
    "    df_enzymes = df_enzymes.reset_index(drop=False)\n",
    "\n",
    "df_enzymes[table_type] = df_enzymes[[strip_plural(table_type), \"compartment\"]].apply(\n",
    "    lambda x: \"_\".join(x.values), axis=1\n",
    ")\n",
    "df_enzymes = df_enzymes.set_index(strip_plural(table_type))\n",
    "pcmodel_tables[table_type] = df_enzymes.copy()\n",
    "df_enzymes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aee3dbe-cb7b-40a7-b3e6-56c9a238df06",
   "metadata": {},
   "source": [
    "## Create PC-model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bf7b1787-95e1-4848-a4cc-d4c79245c81c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of metabolites: 2157\n",
      "Number of proteins: 820\n",
      "Number of complexes: 887\n",
      "Number of enzymes: 6543\n",
      "Forward variable: 2181\n",
      "Reverse variable: 2181\n",
      "Summation variable : 2181\n",
      "Number of proteome budget: 1\n",
      "\n",
      "Number of reactions: 3275\n",
      "Number of protein dilutions: 820\n",
      "Number of complex formation reactions: 887\n",
      "Number of complex dilutions: 887\n",
      "Number of enzyme formation reactions: 6384\n",
      "Number of enzyme dilutions: 6543\n",
      "Forward variable: 2181\n",
      "Reverse variable: 2181\n",
      "Summation variable : 2181\n",
      "Number of proteome budget demand: 1\n",
      "\n",
      "Number of non-zero complex rate constants: 6384\n",
      "Number of non-zero enzyme rate constants: 4362\n"
     ]
    }
   ],
   "source": [
    "protein_table = pcmodel_tables[\"proteins\"].reset_index(drop=False)\n",
    "complex_table = pcmodel_tables[\"complexes\"].reset_index(drop=False)\n",
    "enzyme_table = pcmodel_tables[\"enzymes\"].reset_index(drop=False)\n",
    "max_weight_fraction = 100\n",
    "\n",
    "pcmodel, final_pcmodel_tables = construct_pcmodel_from_tables(\n",
    "    model,\n",
    "    protein_table=protein_table,\n",
    "    complex_table=complex_table,\n",
    "    enzyme_table=enzyme_table,\n",
    "    max_weight_fraction=max_weight_fraction,\n",
    "    enzyme_keff_base=enzyme_keff_base,\n",
    "    enzyme_forward_suffix=enzyme_forward_suffix,\n",
    "    enzyme_reverse_suffix=enzyme_reverse_suffix,\n",
    "    enzyme_total_suffix=enzyme_total_suffix,\n",
    "    include_complex_dilutions=True,  # Relaxes constraints areound complexes. Recommend to start, can be set to zero later or removed entirely\n",
    "    irrev_rxn_complex_keff=0,  # Set as None to ignore, small number to keep in model, 0 to remove from complex-enzyme mapping\n",
    ")\n",
    "if simplify_compartments:\n",
    "    pcmodel.compartments = {\"pc\": \"protein compartment\"}\n",
    "# Print summary\n",
    "for attr, subclass_dict in ATTR_SUBCLASS_DICT.items():\n",
    "    n = len(\n",
    "        getattr(pcmodel, attr).query(\n",
    "            lambda x: not isinstance(x, tuple(subclass_dict.values()))\n",
    "        )\n",
    "    )\n",
    "    print(f\"Number of {attr}: {n}\")\n",
    "    for key, subcls in subclass_dict.items():\n",
    "        obj_list = getattr(pcmodel, attr).query(lambda x: isinstance(x, subcls))\n",
    "        n = len(obj_list)\n",
    "        print(f\"Number of {key}: {n}\")\n",
    "        if subcls in (Enzyme, EnzymeDilution):\n",
    "            print(\n",
    "                f\"Forward variable: {len(obj_list.query(lambda x: enzyme_forward_suffix in x.id))}\"\n",
    "            )\n",
    "            print(\n",
    "                f\"Reverse variable: {len(obj_list.query(lambda x: enzyme_reverse_suffix in x.id))}\"\n",
    "            )\n",
    "            print(\n",
    "                f\"Summation variable : {len(obj_list.query(lambda x: enzyme_total_suffix in x.id))}\"\n",
    "            )\n",
    "    print()\n",
    "\n",
    "keff_table = final_pcmodel_tables[\"enzymes\"].copy()\n",
    "keff_table[\"direction\"] = keff_table[\"reactions\"].apply(\n",
    "    lambda rid: model.reactions.get_by_id(rid).reaction\n",
    ")\n",
    "keff_table[\"direction\"] = keff_table[\"direction\"].apply(\n",
    "    lambda x: x.replace(\"<=>\", \"-->\")\n",
    ")\n",
    "keff_table[\"direction\"] = keff_table[[\"enzyme\", \"direction\"]].apply(\n",
    "    lambda x: (\n",
    "        x[\"direction\"].replace(\"-->\", \"<--\")\n",
    "        if x[\"enzyme\"].endswith(DEFAULT_ENZYME_REVERSE_SUFFIX)\n",
    "        else x[\"direction\"]\n",
    "    ),\n",
    "    axis=1,\n",
    ")\n",
    "keff_table[\"complexes\"] = keff_table[\"complexes\"].apply(lambda x: x.split(\";\"))\n",
    "keff_table[\"complex_keff\"] = keff_table[\"complex_keff\"].apply(lambda x: x.split(\";\"))\n",
    "keff_table = keff_table.explode([\"complexes\", \"complex_keff\"])\n",
    "keff_table[\"complex\"] = keff_table[[\"complexes\", \"compartment\"]].apply(\n",
    "    lambda x: x[\"complexes\"].replace(f\"_{x['compartment']}\", \"\"), axis=1\n",
    ")\n",
    "keff_table = keff_table.groupby([\"enzyme\", \"complex\"], as_index=False).agg(\n",
    "    lambda x: list(x.unique())[0]\n",
    ")\n",
    "keff_table = keff_table.loc[\n",
    "    :,\n",
    "    [\n",
    "        \"enzyme\",\n",
    "        \"enzyme_keff\",\n",
    "        \"complex\",\n",
    "        \"complex_keff\",\n",
    "        \"compartment\",\n",
    "        \"reactions\",\n",
    "        \"direction\",\n",
    "    ],\n",
    "]\n",
    "complex_keff_table = keff_table.drop(\"enzyme_keff\", axis=1).drop_duplicates()\n",
    "enzyme_keff_table = (\n",
    "    keff_table.groupby([\"enzyme\", \"enzyme_keff\"], as_index=False)[\n",
    "        [\"reactions\", \"direction\"]\n",
    "    ]\n",
    "    .agg(lambda x: list(x.unique())[0])\n",
    "    .drop_duplicates()\n",
    ")\n",
    "final_pcmodel_tables[\"complex_keffs\"] = complex_keff_table\n",
    "final_pcmodel_tables[\"enzyme_keffs\"] = enzyme_keff_table\n",
    "\n",
    "n_cplx_keff = len(\n",
    "    complex_keff_table[complex_keff_table[\"complex_keff\"].astype(float) != 0]\n",
    ")\n",
    "print(f\"Number of non-zero complex rate constants: {n_cplx_keff}\")\n",
    "\n",
    "n_enzyme_keff = len(\n",
    "    enzyme_keff_table[enzyme_keff_table[\"enzyme_keff\"].astype(float) != 0]\n",
    ")\n",
    "print(f\"Number of non-zero enzyme rate constants: {n_enzyme_keff}\")\n",
    "\n",
    "\n",
    "for table_type, df_table in final_pcmodel_tables.items():\n",
    "    df_table.to_csv(data_path / f\"pcmodel_{table_type}.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e7554cd-dbf7-4286-a29e-81281e44f767",
   "metadata": {},
   "source": [
    "### Formulate additional protein constraints\n",
    "#### Address isoforms and compartments with additional constraints\n",
    "For isoforms and/or compartments, place an additional constraint such that the total sum of all isoforms does not exceed the measured concentraiton value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ae04a412-75e3-4ca8-b629-46f8fafcb243",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_table = pcmodel_tables[\"proteins\"]\n",
    "mapping_key = \"uniprot\"\n",
    "df_additional_constraints = protein_table[protein_table[mapping_key].duplicated(False)]\n",
    "df_additional_constraints = df_additional_constraints.groupby(\n",
    "    [\n",
    "        \"genes\",\n",
    "        mapping_key,\n",
    "    ],\n",
    "    as_index=False,\n",
    ").agg(lambda x: list(x))\n",
    "if not df_additional_constraints.empty:\n",
    "    if \"lower_bound\" in df_additional_constraints.columns:\n",
    "        df_additional_constraints[\"lower_bound\"] = df_additional_constraints[\n",
    "            \"lower_bound\"\n",
    "        ].apply(min)\n",
    "    if \"upper_bound\" in df_additional_constraints.columns:\n",
    "        df_additional_constraints[\"upper_bound\"] = df_additional_constraints[\n",
    "            \"upper_bound\"\n",
    "        ].apply(max)\n",
    "\n",
    "data = {}\n",
    "for idx, row in df_additional_constraints.iterrows():\n",
    "    # Technically, always one gene but refers to genes attribute\n",
    "    genes = row[\"genes\"]\n",
    "    uniprot = model.genes.get_by_id(genes).annotation.get(mapping_key, \"\")\n",
    "    proteins = split_string(row.get(\"proteins\"))\n",
    "    proteins = pcmodel.metabolites.get_by_any(proteins)\n",
    "    is_compartment = len({p.compartment for p in proteins}) > 1\n",
    "    is_isoform = (\n",
    "        len(\n",
    "            {\n",
    "                p.id.replace(f\"_{p.compartment}\", \"\").split(\n",
    "                    \"_\",\n",
    "                )[-1]\n",
    "                for p in proteins\n",
    "                if p.id.replace(f\"_{p.compartment}\", \"\")\n",
    "                .split(\n",
    "                    \"_\",\n",
    "                )[-1]\n",
    "                .isnumeric()\n",
    "            }\n",
    "        )\n",
    "        > 1\n",
    "    )\n",
    "    if is_compartment and not is_isoform:\n",
    "        default_prefix = DEFAULT_COMPARTMENT_CONSTRAINT_PREFIX\n",
    "    elif is_isoform and not is_compartment:\n",
    "        default_prefix = DEFAULT_ISOFORM_CONSTRAINT_PREFIX\n",
    "    else:\n",
    "        default_prefix = DEFAULT_CONSTRAINT_PREFIX\n",
    "    constraint_id = row.get(\"constraints\", f\"{default_prefix}{genes}\")\n",
    "    lower_bound = float(row.get(\"lower_bound\")) if row.get(\"lower_bound\") else 0\n",
    "    upper_bound = (\n",
    "        float(row.get(\"upper_bound\"))\n",
    "        if row.get(\"upper_bound\")\n",
    "        else DEFAULT_CONCENTRATION_BOUND\n",
    "    )\n",
    "    protein_dilutions = [\n",
    "        reaction\n",
    "        for protein in proteins\n",
    "        for reaction in list(protein.reactions)\n",
    "        if reaction.id.endswith(protein.id)\n",
    "    ]\n",
    "    # \"ISOCONS\" is short for \"ISOFORM CONSTRAINT\"\n",
    "    # \"COMPCONS\" is short for \"COMPARTMENT CONSTRAINT\"\n",
    "    # \"CONS\" for general constraint\n",
    "    data[idx] = {\n",
    "        \"constraints\": constraint_id,\n",
    "        \"genes\": genes,\n",
    "        \"proteins\": build_string([p.id for p in proteins]),\n",
    "        \"reactions\": build_string([p.id for p in protein_dilutions]),\n",
    "        # Assume sum of isoforms is a constant, works well with proteomic measurements that do not distinguish\n",
    "        \"coefficients\": \";\".join([str(1) for p in protein_dilutions]),\n",
    "        \"lower_bound\": lower_bound,\n",
    "        \"upper_bound\": upper_bound,\n",
    "        \"unit\": \"nmol / gDW\",\n",
    "        mapping_key: uniprot,\n",
    "    }\n",
    "df_additional_constraints = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "df_additional_constraints.to_csv(\n",
    "    data_path / f\"constraints_proteins_{pcmodel.id}.tsv\", sep=\"\\t\", index=False\n",
    ")\n",
    "df_additional_constraints"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f7be1c-67b8-4ebb-92ee-bb53b312d30f",
   "metadata": {},
   "source": [
    "## Add additional protein constraints to model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "df3bb685-f825-4f2a-bc0d-66d50606d73c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "try:\n",
    "    df_additional_constraints = pd.read_csv(\n",
    "        data_path / f\"constraints_proteins_{pcmodel.id}.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "    )\n",
    "except (FileNotFoundError, pd.errors.EmptyDataError):\n",
    "    df_additional_constraints = pd.DataFrame()\n",
    "else:\n",
    "    if not df_additional_constraints.empty:\n",
    "        for constraint_id, row in df_additional_constraints.set_index(\n",
    "            \"constraints\"\n",
    "        ).iterrows():\n",
    "            reactions = pcmodel.reactions.get_by_any(row[\"reactions\"].split(\";\"))\n",
    "            coefficients = row[\"coefficients\"].split(\";\")\n",
    "            abundance = sum(\n",
    "                [\n",
    "                    int(coeff) * reaction.flux_expression\n",
    "                    for reaction, coeff in zip(reactions, coefficients)\n",
    "                ]\n",
    "            )\n",
    "            lower_bound = float(row.get(\"lower_bound\")) if row.get(\"lower_bound\") else 0\n",
    "            upper_bound = (\n",
    "                float(row.get(\"upper_bound\"))\n",
    "                if row.get(\"upper_bound\")\n",
    "                else DEFAULT_CONCENTRATION_BOUND\n",
    "            )\n",
    "            if constraint_id in pcmodel.constraints:\n",
    "                # TODO warn\n",
    "                pcmodel.remove_cons_vars(pcmodel.constraints[constraint_id])\n",
    "            additional_constraint = pcmodel.problem.Constraint(\n",
    "                abundance,\n",
    "                name=constraint_id,\n",
    "                lb=lower_bound,\n",
    "                ub=upper_bound,\n",
    "            )\n",
    "            pcmodel.add_cons_vars(additional_constraint)\n",
    "\n",
    "df_additional_constraints"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9398f43f-208a-413a-850b-e78e37a0025f",
   "metadata": {},
   "source": [
    "### Annotate proteins with UniProt IDs and sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "950fa420-ace9-4354-a71e-6e2e49833bc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_to_uniprot = protein_table.set_index(\"proteins\")[\"uniprot\"].to_dict()\n",
    "protein_to_sequence = protein_table.set_index(\"proteins\")[\"sequence\"].to_dict()\n",
    "\n",
    "for protein_dilution in pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, ProteinDilution)\n",
    "):\n",
    "    if protein_to_uniprot.get(f\"{protein_dilution.id}\".replace(\"PROTDL_\", \"\")):\n",
    "        protein_dilution.annotation[\"uniprot\"] = protein_to_uniprot[\n",
    "            f\"{protein_dilution.id}\".replace(\"PROTDL_\", \"\")\n",
    "        ]\n",
    "        protein_dilution.annotation[\"uniprot.aa_sequence\"] = protein_to_sequence[\n",
    "            f\"{protein_dilution.id}\".replace(\"PROTDL_\", \"\")\n",
    "        ]\n",
    "\n",
    "for protein in pcmodel.metabolites.query(lambda x: isinstance(x, Protein)):\n",
    "    if protein_to_uniprot.get(protein.id):\n",
    "        protein.annotation[\"uniprot\"] = protein_to_uniprot[protein.id]\n",
    "        protein.annotation[\"uniprot.aa_sequence\"] = protein_to_sequence[protein.id]\n",
    "\n",
    "\n",
    "enzyme_to_reaction = enzyme_table.set_index(\"enzymes\")[\"reactions\"].to_dict()\n",
    "for enzyme_dilution in pcmodel.reactions.query(lambda x: isinstance(x, EnzymeDilution)):\n",
    "    if enzyme_to_reaction.get(f\"{enzyme_dilution.id}\".replace(\"ENZDL_\", \"\")):\n",
    "        # Make it easier to use later\n",
    "        enzyme_dilution.annotation[\"reaction\"] = enzyme_to_reaction[\n",
    "            f\"{enzyme_dilution.id}\".replace(\"ENZDL_\", \"\")\n",
    "        ]\n",
    "\n",
    "for enzyme in pcmodel.metabolites.query(lambda x: isinstance(x, Enzyme)):\n",
    "    if enzyme_to_reaction.get(enzyme.id):\n",
    "        enzyme.annotation[\"reaction\"] = enzyme_to_reaction[enzyme.id]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75b5ed10-611d-44a1-b924-cf3badb5a39f",
   "metadata": {},
   "source": [
    "### Set proteome budget constraints for low abundance and high-abundance proteomes\n",
    "* RBCs are enucleated, terminally differentiated cells that are composed of 95% to 98% Hb by dry mass (mass of all the constituents of a cell in the absence of water)\n",
    "    * PMID: 13429433, PMID: 13999462, PMID: 21796773, **PMID: 34378368**\n",
    "* Therefore, remove hemoglobin from the low abundance proteome budget constraint and create a new constraint specific to hemoglobin abundance.\n",
    "* Assume 90% minimum of dry mass is hemoglobin, and up to 10% of dry mass are other proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "954117e4-76e8-4035-a16d-94289415015c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PBDL_hemoglobin_budget: hemoglobin_budget --> \n",
      "PROTDL_protein_HBA_pc:  --> 0.015259421999999979 hemoglobin_budget + protein_HBA_pc\n",
      "PROTDL_protein_HBB_pc:  --> 0.015997243999999983 hemoglobin_budget + protein_HBB_pc\n",
      "PROTDL_protein_HBD_pc:  --> 0.016056335999999984 hemoglobin_budget + protein_HBD_pc\n",
      "PROTDL_protein_HBE1_pc:  --> 0.016204680999999985 hemoglobin_budget + protein_HBE1_pc\n",
      "PROTDL_protein_HBG1_pc:  --> 0.016127253999999987 hemoglobin_budget + protein_HBG1_pc\n",
      "PROTDL_protein_HBG2_pc:  --> 0.016125281999999987 hemoglobin_budget + protein_HBG2_pc\n",
      "PROTDL_protein_HBM_pc:  --> 0.015614928999999987 hemoglobin_budget + protein_HBM_pc\n",
      "PROTDL_protein_HBQ1_pc:  --> 0.015507637999999985 hemoglobin_budget + protein_HBQ1_pc\n",
      "PROTDL_protein_HBZ_pc:  --> 0.015637878999999983 hemoglobin_budget + protein_HBZ_pc\n",
      "\n",
      "PBDL_proteome_budget: proteome_budget --> total_budget\n",
      "PBDL_hemoglobin_budget: hemoglobin_budget --> total_budget\n",
      "PBDL_total_budget: total_budget --> \n"
     ]
    }
   ],
   "source": [
    "# Proteome budget for low abundance proteins\n",
    "PBDL_proteome_budget = pcmodel.reactions.get_by_id(\"PBDL_proteome_budget\")\n",
    "PBDL_proteome_budget.id = \"PBDL_proteome_budget\"\n",
    "PBDL_proteome_budget.name = \"Proteome budget demand (Low abundance)\"\n",
    "\n",
    "proteome_budget = pcmodel.metabolites.get_by_id(\"proteome_budget\")\n",
    "proteome_budget.id = \"proteome_budget\"\n",
    "proteome_budget.name = \"Proteome Budget Constraint (Low abundance)\"\n",
    "\n",
    "# Proteome budget for hemoglobin\n",
    "PBDL_hemoglobin_budget = PBDL_proteome_budget.copy()\n",
    "hemoglobin_budget = list(PBDL_hemoglobin_budget.metabolites).pop()\n",
    "PBDL_hemoglobin_budget.id = \"PBDL_hemoglobin_budget\"\n",
    "PBDL_hemoglobin_budget.name = \"Proteome budget demand (Hemoglobin)\"\n",
    "\n",
    "hemoglobin_budget.id = \"hemoglobin_budget\"\n",
    "hemoglobin_budget.name = \"Hemoglobin Budget Constraint\"\n",
    "\n",
    "PBDL_hemoglobin_budget.bounds = (900, 1000)\n",
    "PBDL_proteome_budget.bounds = (0.0, 100)\n",
    "\n",
    "pcmodel.add_reactions([PBDL_hemoglobin_budget])\n",
    "\n",
    "remove_from_low_abundance_budget = [\n",
    "    \"HBA\",\n",
    "    \"HBB\",\n",
    "    \"HBD\",\n",
    "    \"HBE1\",\n",
    "    \"HBG1\",\n",
    "    \"HBG2\",\n",
    "    \"HBM\",\n",
    "    \"HBQ1\",\n",
    "    \"HBZ\",\n",
    "]\n",
    "for reaction in proteome_budget.reactions:\n",
    "    if any(\n",
    "        [f\"protein_{gid}\" in reaction.id for gid in remove_from_low_abundance_budget]\n",
    "    ):\n",
    "        reaction.add_metabolites(\n",
    "            {hemoglobin_budget: reaction.get_coefficient(proteome_budget)}\n",
    "        )\n",
    "        reaction.subtract_metabolites(\n",
    "            {proteome_budget: reaction.get_coefficient(proteome_budget)}\n",
    "        )\n",
    "\n",
    "for reaction in sorted(\n",
    "    pcmodel.metabolites.get_by_id(\"hemoglobin_budget\").reactions, key=lambda x: x.id\n",
    "):\n",
    "    print(reaction)\n",
    "\n",
    "# Add total budget constraint for hemoglobin and protein\n",
    "pcmodel.add_metabolites(\n",
    "    [\n",
    "        ProteomeBudget(\n",
    "            id=\"total_budget\",\n",
    "            name=\"Total Budget Constraint\",\n",
    "            compartment=proteome_budget.compartment,\n",
    "        )\n",
    "    ]\n",
    ")\n",
    "total_budget = pcmodel.metabolites.get_by_id(\"total_budget\")\n",
    "PBDL_total_budget = f\"PBDL_{total_budget.id}\"\n",
    "pcmodel.add_reactions(\n",
    "    [\n",
    "        ProteomeBudgetDilution(\n",
    "            id=PBDL_total_budget,\n",
    "            name=\"Total budget demand\",\n",
    "            lower_bound=0,\n",
    "            upper_bound=1000,\n",
    "        )\n",
    "    ]\n",
    ")\n",
    "PBDL_total_budget = pcmodel.reactions.get_by_id(PBDL_total_budget)\n",
    "print()\n",
    "for reaction in pcmodel.reactions.query(\n",
    "    lambda x: isinstance(x, ProteomeBudgetDilution)\n",
    "):\n",
    "    if PBDL_total_budget.id == reaction.id:\n",
    "        reaction.add_metabolites({total_budget: -1}, combine=False)\n",
    "    else:\n",
    "        reaction.add_metabolites({total_budget: 1}, combine=False)\n",
    "    print(reaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17b852f8-6444-40e7-bed3-27741a75842e",
   "metadata": {},
   "source": [
    "### Ensure model can be optimized for glucose uptake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2548b619-2bf2-4923-96ca-b6136398e614",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "COt                        -92.169711\n",
       "DPGM                       -23.042428\n",
       "ENO                         23.042428\n",
       "EX_co_e                    -92.169711\n",
       "EX_h2o2_e                  -23.042428\n",
       "EX_k_e                     -92.169711\n",
       "EX_na1_e                   138.254567\n",
       "EX_pi_e                     46.084856\n",
       "EX_pyr_e                    23.042428\n",
       "FAt_hs_3_0                  23.042428\n",
       "H2O2t                      -23.042428\n",
       "HB23DPGB                   -23.042428\n",
       "HBCOB                       92.169711\n",
       "METOXR                      23.042428\n",
       "NaKt                        46.084856\n",
       "PGK                         23.042428\n",
       "PGM                         23.042428\n",
       "PIt                         46.084856\n",
       "PYK                         23.042428\n",
       "PYR_PPAtex                  23.042428\n",
       "SK_carboxyhb_c              92.169711\n",
       "SK_hb4_23dpg_c             -23.042428\n",
       "SK_k_c                      92.169711\n",
       "SK_na1_c                  -138.254567\n",
       "SK_protres_metSox__SL_c     23.042428\n",
       "SK_protres_met__L_c        -23.042428\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "objective_rxns = [\"NaKt\"]\n",
    "pcmodel.objective = sum(\n",
    "    [pcmodel.reactions.get_by_id(rid).flux_expression for rid in objective_rxns]\n",
    ")\n",
    "pcsol = pcmodel.optimize()\n",
    "pcsol.fluxes.loc[\n",
    "    [r.id for r in model.reactions if r.id in pcsol.fluxes[pcsol.fluxes != 0].index]\n",
    "].sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "376fde0a-9090-4c06-9849-937303d72538",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PROTDL_protein_ATP1A3_pc       236.229799\n",
       "PROTDL_protein_ATP1B3_pc       236.229799\n",
       "PROTDL_protein_ENO3_pc         172.206954\n",
       "PROTDL_protein_HBE1_pc       55431.263739\n",
       "PROTDL_protein_HBZ_pc          112.166981\n",
       "PROTDL_protein_PGAM1_pc        238.598927\n",
       "PROTDL_protein_PGK1_pc          49.248098\n",
       "PROTDL_protein_PKM_pc          677.784457\n",
       "PROTDL_protein_SLC16A7_pc      186.384576\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcsol.fluxes.loc[\n",
    "    [\n",
    "        r.id\n",
    "        for r in pcmodel.reactions.query(lambda x: isinstance(x, ProteinDilution))\n",
    "        if r.id in pcsol.fluxes[pcsol.fluxes != 0].index\n",
    "    ]\n",
    "].sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5f91db47-c712-4090-998c-d9e528c40975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ENZDL_enzyme_DPGM_fwd_pc             98.471913\n",
       "ENZDL_enzyme_DPGM_total_pc           98.471913\n",
       "ENZDL_enzyme_ENO_rev_pc              98.471913\n",
       "ENZDL_enzyme_ENO_total_pc            98.471913\n",
       "ENZDL_enzyme_HB23DPGB_fwd_pc         98.471913\n",
       "ENZDL_enzyme_HB23DPGB_total_pc       98.471913\n",
       "ENZDL_enzyme_HBCOB_rev_pc           393.887654\n",
       "ENZDL_enzyme_HBCOB_total_pc         393.887654\n",
       "ENZDL_enzyme_NaKt_rev_pc            196.943827\n",
       "ENZDL_enzyme_NaKt_total_pc          196.943827\n",
       "ENZDL_enzyme_PGK_rev_pc              98.471913\n",
       "ENZDL_enzyme_PGK_total_pc            98.471913\n",
       "ENZDL_enzyme_PGM_rev_pc              98.471913\n",
       "ENZDL_enzyme_PGM_total_pc            98.471913\n",
       "ENZDL_enzyme_PYK_rev_pc              98.471913\n",
       "ENZDL_enzyme_PYK_total_pc            98.471913\n",
       "ENZDL_enzyme_PYR_PPAtex_rev_pc       98.471913\n",
       "ENZDL_enzyme_PYR_PPAtex_total_pc     98.471913\n",
       "Name: fluxes, dtype: float64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcsol.fluxes.loc[\n",
    "    [\n",
    "        r.id\n",
    "        for r in pcmodel.reactions.query(lambda x: isinstance(x, EnzymeDilution))\n",
    "        if r.id in pcsol.fluxes[pcsol.fluxes != 0].index\n",
    "    ]\n",
    "].sort_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a01f53-12d7-464b-b633-762fb818f823",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d995d1dc-1355-4ace-b469-3067a9932791",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Regular model\n",
    "write_cobra_model(model, filename=models_path / f\"{model}.xml\")\n",
    "write_cobra_model(model, filename=models_path / f\"{model}.json\")\n",
    "\n",
    "# Protein constrained  without curated keffs\n",
    "write_cobra_model(pcmodel, filename=models_path / f\"{pcmodel}.xml\")\n",
    "write_cobra_model(pcmodel, filename=models_path / f\"{pcmodel}.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aa5c7c65-f808-4970-af4d-fc43f20ab8bd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14cdec250</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>2157</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>3275</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM at 0x14cdec250>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f9ed6456-9e04-414a-9dec-e1aa13c0fcaf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>15c84e550</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>10410</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>18799</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC at 0x15c84e550>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcmodel"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0839ecb1-89cd-47e7-acd1-a9839c82b0f6",
   "metadata": {},
   "source": [
    "### Create PC-model representative of RBC-Omics\n",
    "Can only be done after pcFVA results are generated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "fb2e37c2-f7eb-4981-9624-616a5814af25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>RBC_GEM_PC_RBComics</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>1532d0410</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>10411</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>19620</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>820</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>78</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space, protein compartment</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model RBC_GEM_PC_RBComics at 0x1532d0410>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_reaction_bounds = pd.read_csv(\n",
    "    dataset_path / f\"{pcmodel.id}_{dataset_name}_reaction_bounds.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=\"reactions\",\n",
    ")\n",
    "df_reaction_bounds = df_reaction_bounds.rename(\n",
    "    {\"minimum\": \"lower_bound\", \"maximum\": \"upper_bound\"}, axis=1\n",
    ")\n",
    "\n",
    "pcmodel_dataset_parameterized = pcmodel.copy()\n",
    "pcmodel_dataset_parameterized.id += f\"_{dataset_name}\"\n",
    "add_relaxation_budget(pcmodel_dataset_parameterized, 0, verbose=False)\n",
    "for rid, bounds in df_reaction_bounds.iterrows():\n",
    "    reaction = pcmodel_dataset_parameterized.reactions.get_by_id(rid)\n",
    "    reaction.bounds = bounds\n",
    "\n",
    "\n",
    "# Protein constrained  without curated keffs\n",
    "write_cobra_model(\n",
    "    pcmodel_dataset_parameterized,\n",
    "    filename=models_path / f\"{pcmodel_dataset_parameterized}.xml\",\n",
    ")\n",
    "write_cobra_model(\n",
    "    pcmodel_dataset_parameterized,\n",
    "    filename=models_path / f\"{pcmodel_dataset_parameterized}.json\",\n",
    ")\n",
    "pcmodel_dataset_parameterized"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
