{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "735fdc9a-80ee-40c6-9214-fa09bdb0643e",
   "metadata": {},
   "source": [
    "# Model Content Comparison \n",
    "The purpose of this notebook is to compare the current version of the RBC-GEM to earlier versions.\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1443fbd2-1eeb-4c5d-b0e1-0b72016b6d6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import requests\n",
    "import seaborn as sns\n",
    "from cobra.manipulation import rename_genes\n",
    "from rbc_gem_utils import GEM_NAME, GEM_URL, MODEL_PATH, read_cobra_model, show_versions\n",
    "from rbc_gem_utils.visualization import cmap_map\n",
    "\n",
    "# Display versions of last time notebook ran and worked\n",
    "show_versions()\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00a763f5-135b-459b-a584-99ebb493f029",
   "metadata": {},
   "source": [
    "## Load RBC-GEM models\n",
    "### Current model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ff1a81f-9552-41d7-a933-f12d22dee040",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = Path(\"data\").resolve()\n",
    "models_path = Path(\"models\").resolve()\n",
    "figures_path = Path(\"figures\").resolve()\n",
    "version = \"1.2.0\"\n",
    "\n",
    "imagetype = \"png\"\n",
    "transparent = True\n",
    "save_figures = True\n",
    "\n",
    "ftype = \"xml\"\n",
    "current_model = read_cobra_model(models_path / f\"{GEM_NAME.replace('-', '_')}.{ftype}\")\n",
    "\n",
    "current_version = version\n",
    "\n",
    "current_model.metabolites.sort()\n",
    "current_model.genes.sort()\n",
    "current_model.reactions.sort()\n",
    "current_model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd0e660c-5f71-48a8-bae3-dfb35fe4b920",
   "metadata": {},
   "source": [
    "### Earlier models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6538447-8eac-4046-a40e-fb2e33f5a1f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Probably a better way to do this instead of erroring out.\n",
    "download = False\n",
    "version = \"0.2.0\"\n",
    "ftype = \"xml\"\n",
    "\n",
    "filepath = models_path / f\"{GEM_NAME.replace('-', '_')}_{version}.{ftype}\"\n",
    "\n",
    "if download:\n",
    "    response = requests.get(f\"{GEM_URL}/{version}/{MODEL_PATH}/{GEM_NAME}.{ftype}\")\n",
    "    response.raise_for_status()\n",
    "\n",
    "    # Write file\n",
    "    # Is there a better way of checking whether binary file?\n",
    "    if not response.encoding:\n",
    "        with open(filepath, \"wb\") as file:\n",
    "            file.write(response.content)\n",
    "    else:\n",
    "        with open(filepath, \"w\") as file:\n",
    "            file.write(response.text)\n",
    "\n",
    "prev_model_020 = read_cobra_model(filepath)\n",
    "prev_model_020.metabolites.sort()\n",
    "prev_model_020.genes.sort()\n",
    "prev_model_020.reactions.sort()\n",
    "prev_model_020.groups.sort()\n",
    "# For comparison\n",
    "group = prev_model_020.groups.get_by_id(\"Transport reactions\")\n",
    "group.id = \"Transport, extracellular\"\n",
    "\n",
    "\n",
    "# For comparison rename genes and remove splice isoforms\n",
    "rename_genes(\n",
    "    prev_model_020,\n",
    "    {g.id: g.annotation.get(\"hgnc.symbol\") for g in prev_model_020.genes},\n",
    ")\n",
    "prev_model_020.repair()\n",
    "prev_model_020"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04e940e3-83d6-49ac-b7c5-b1a838081a14",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(prev_model_020.reactions.query(lambda x: x.subsystem != \"Pseudoreactions\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2272078e-ceee-409d-8250-794299c85d7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_list = [prev_model_020, current_model]\n",
    "model_data = defaultdict(dict)\n",
    "\n",
    "save_figures = True\n",
    "transparent = False\n",
    "imagetype = \"svg\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58b6065d-395a-46a5-b164-fee7d8431ccc",
   "metadata": {},
   "source": [
    "### Comparison of subsystems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b63064cc-67e9-4600-aab6-ba337332fcac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pathways = pd.read_csv(\n",
    "    f\"{data_path}/subsystems.tsv\", sep=\"\\t\", index_col=None, dtype=str\n",
    ")\n",
    "df_pathways = df_pathways.fillna(\"\")\n",
    "# Group \"Metabolism of other amino acids\" with amino acids rather than treat as \"other\"\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].replace(\n",
    "    \"Metabolism of other amino acids\", \"Amino acid metabolism\"\n",
    ")\n",
    "df_pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0959b568-a93e-4e0f-998d-70c283b0c634",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Categories that should be excluded from the figure\n",
    "categories_to_exclude = {\"Pseudoreactions\", \"Model total\"}\n",
    "# Main categories in figure and abbreviations, all unmapped categories are mapped to \"Other\"\n",
    "categories_to_keep = {\n",
    "    \"Amino acid metabolism\": mpl.cm.spring,\n",
    "    \"Carbohydrate metabolism\": mpl.cm.Greens,\n",
    "    \"Lipid metabolism\": mpl.cm.Blues,\n",
    "    \"Metabolism of cofactors and vitamins\": mpl.cm.summer,\n",
    "    \"Nucleotide metabolism\": mpl.cm.winter,\n",
    "    \"Reactive species\": mpl.cm.Reds,\n",
    "    \"Transport reactions\": mpl.cm.Purples,\n",
    "    \"Other\": mpl.cm.gray_r,\n",
    "}\n",
    "use_abbrevs = True\n",
    "abbrevs = {\n",
    "    \"Amino acid metabolism\": \"A\",\n",
    "    \"Carbohydrate metabolism\": \"C\",\n",
    "    \"Lipid metabolism\": \"L\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"V\",\n",
    "    \"Nucleotide metabolism\": \"N\",\n",
    "    \"Reactive species\": \"R\",\n",
    "    \"Transport reactions\": \"T\",\n",
    "    \"Other\": \"O\",\n",
    "}\n",
    "barsize = 0.8\n",
    "fontsize = 12\n",
    "cmax = 0.8\n",
    "cmin = 0.3\n",
    "\n",
    "\n",
    "colormaps_normal = dict(\n",
    "    zip(\n",
    "        categories_to_keep,\n",
    "        [\n",
    "            cmap_map(lambda x: x * 1, categories_to_keep[k])(cmax)\n",
    "            for k in categories_to_keep\n",
    "        ],\n",
    "    )\n",
    ")\n",
    "colormaps_light = dict(\n",
    "    zip(\n",
    "        categories_to_keep,\n",
    "        [\n",
    "            cmap_map(lambda x: x / 2 + 0.5, categories_to_keep[k])(cmax)\n",
    "            for k in categories_to_keep\n",
    "        ],\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c0c4417-5d90-497f-8992-224b9d54dadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "categories = [\"Removed\", \"Unchanged\", \"Added\"]\n",
    "colors_dict = dict(\n",
    "    zip(categories, [\"xkcd:light blue\", \"xkcd:purple\", \"xkcd:light red\"])\n",
    ")\n",
    "attribute_types = [\"reactions\", \"metabolites\", \"genes\"]\n",
    "edgecolor = \"black\"\n",
    "linewidth = 0.5\n",
    "data = defaultdict(dict)\n",
    "for attribute_type in attribute_types:\n",
    "    try:\n",
    "        df_removed = pd.read_csv(\n",
    "            f\"{data_path}/{attribute_type}_removed.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index_col=None,\n",
    "            dtype=str,\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "    if attribute_type == \"reactions\":\n",
    "        obj_lists = [\n",
    "            getattr(model, attribute_type).query(lambda x: not x.boundary)\n",
    "            for model in model_list\n",
    "        ]\n",
    "    else:\n",
    "        obj_lists = [getattr(model, attribute_type) for model in model_list]\n",
    "\n",
    "    shared = obj_lists[0].query(lambda x: x.id not in df_removed[attribute_type].values)\n",
    "    removed = df_removed[attribute_type].values\n",
    "    new = obj_lists[1].query(lambda x: x.id not in set(shared.list_attr(\"id\")))\n",
    "    data[attribute_type].update(\n",
    "        dict(zip(categories, list(map(len, [removed, shared, new]))))\n",
    "    )\n",
    "\n",
    "\n",
    "df_data = pd.DataFrame.from_dict(data, orient=\"index\")\n",
    "df_data.index = df_data.index.str.capitalize()\n",
    "\n",
    "fig, ax_general = plt.subplots(figsize=(6, 2))\n",
    "total = df_data.cumsum(axis=1)\n",
    "offsets = df_data[categories[0]].values\n",
    "labels = [s.capitalize() for s in attribute_types]\n",
    "tick_pos = [0.05, 0.11, 0.17]\n",
    "for name, column in df_data.items():\n",
    "    ax_general.barh(\n",
    "        tick_pos,\n",
    "        column.values,\n",
    "        height=0.045,\n",
    "        left=total[name] - column.values - offsets,\n",
    "        color=colors_dict[name],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=linewidth,\n",
    "    )\n",
    "    ax_general.set_yticks(tick_pos, labels=attribute_types)\n",
    "    ax_general.set_yticklabels([s.capitalize() for s in attribute_types])\n",
    "\n",
    "# Remove spines\n",
    "ax_general.spines[\"right\"].set_visible(False)\n",
    "ax_general.spines[\"top\"].set_visible(False)\n",
    "ax_general.spines[\"left\"].set_visible(False)\n",
    "ax_general.axvline(0, linestyle=\":\", color=\"black\", alpha=0.8, linewidth=linewidth)\n",
    "xticks = list(np.linspace(0, 2400, 7))\n",
    "ax_general.set_xticks(xticks)\n",
    "ax_general.xaxis.set_major_formatter(lambda x, pos: int(abs(x)))\n",
    "ax_general.yaxis.set_tick_params(left=False, labelsize=\"xx-large\")\n",
    "ax_general.xaxis.set_tick_params(labelsize=\"x-large\")\n",
    "legend = ax_general.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(\n",
    "            label=label, facecolor=color, edgecolor=edgecolor, linewidth=linewidth\n",
    "        )\n",
    "        for label, color in colors_dict.items()\n",
    "    ],\n",
    "    frameon=False,\n",
    "    edgecolor=\"black\",\n",
    "    ncol=3,\n",
    "    loc=\"lower center\",\n",
    "    bbox_to_anchor=(0.4, 1, 0, 0),\n",
    "    fontsize=\"xx-large\",\n",
    "    handlelength=1,\n",
    ")\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "print(\n",
    "    f\"Number of unique genes in {prev_model_020.id}:\\t\\t{len([x.id for x in prev_model_020.genes])}\\n\"\n",
    "    f\"Number of unique metabolites in {prev_model_020.id}:\\t{len(set([x.id.replace(f'_{x.compartment}', '') for x in prev_model_020.metabolites]))}\\n\"\n",
    "    f\"Number of biochemical reactions in {prev_model_020.id}:\\t{len(prev_model_020.reactions.query(lambda x: x.subsystem != 'Pseudoreactions'))}\\n\"\n",
    ")\n",
    "\n",
    "print(\n",
    "    f\"Number of unique genes in {current_model.id}:\\t\\t{len([x.id for x in current_model.genes])}\\n\"\n",
    "    f\"Number of unique metabolites in {current_model.id}:\\t{len(set([x.id.replace(f'_{x.compartment}', '') for x in current_model.metabolites]))}\\n\"\n",
    "    f\"Number of biochemical reactions in {current_model.id}:\\t{len(current_model.reactions.query(lambda x: x.subsystem != 'Pseudoreactions'))}\\n\"\n",
    ")\n",
    "if save_figures:\n",
    "    fig.savefig(\n",
    "        figures_path / f\"Fig1_Panel_GeneralComparision.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        dpi=None if imagetype != \"png\" else 600,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b00a579-2159-4008-aae3-6da94899da97",
   "metadata": {},
   "outputs": [],
   "source": [
    "for idx, model in enumerate(model_list):\n",
    "    data = defaultdict(dict)\n",
    "    all_reactions = set()\n",
    "    df_cat_subsystems = df_pathways.groupby(\"category\")[\"name\"].agg(lambda x: list(x))\n",
    "    for category, subsystem_list in df_cat_subsystems.items():\n",
    "        if category not in categories_to_keep and category not in categories_to_exclude:\n",
    "            category = \"Other\"\n",
    "\n",
    "        if category not in data:\n",
    "            data[category] = {\"Reactions\": set(), \"Metabolites\": set(), \"Genes\": set()}\n",
    "        # print([x for x in subsystem_list if not x in model.groups])\n",
    "        subsystem_list = [x for x in subsystem_list if x in model.groups]\n",
    "        reactions = [\n",
    "            reaction\n",
    "            for group in model.groups.get_by_any(subsystem_list)\n",
    "            for reaction in group.members\n",
    "        ]\n",
    "        for reaction in reactions:\n",
    "            all_reactions.update([reaction])\n",
    "            data[category][\"Reactions\"].update([reaction])\n",
    "            data[category][\"Metabolites\"].update(reaction.metabolites)\n",
    "            data[category][\"Genes\"].update(reaction.genes)\n",
    "\n",
    "    data[\"Model total\"] = {\n",
    "        \"Reactions\": model.reactions,\n",
    "        \"Metabolites\": model.metabolites,\n",
    "        \"Genes\": model.genes,\n",
    "    }\n",
    "    df_data = pd.DataFrame(\n",
    "        {\n",
    "            key: {k: len(x) for k, x in value_dict.items()}\n",
    "            for key, value_dict in data.items()\n",
    "        }\n",
    "    ).T.sort_index()\n",
    "    df_data = df_data.loc[df_data.index.difference(categories_to_exclude)]\n",
    "\n",
    "    model_data[model.id][f\"{attribute_type}-cat\"] = df_data\n",
    "    print(model.id, \"\\n\", df_data, \"\\n\")\n",
    "df_comparison = pd.merge(\n",
    "    model_data[model_list[0].id][f\"{attribute_type}-cat\"],\n",
    "    model_data[model_list[1].id][f\"{attribute_type}-cat\"],\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    suffixes=(f\"_{model_list[0].id}\", f\"_{model_list[1].id}\"),\n",
    ")\n",
    "df_comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c648822-8845-466e-aa05-8e8ee782f65e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
