{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "871638fb-3cdc-42f3-bae4-30e763de4a81",
   "metadata": {},
   "source": [
    "# Visualize model content\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66a4c857-00ec-4d9e-94ee-c428f4b11cf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import textwrap\n",
    "from collections import defaultdict\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from rbc_gem_utils import (\n",
    "    GEM_NAME,\n",
    "    build_string,\n",
    "    get_annotation_df,\n",
    "    get_dirpath,\n",
    "    read_cobra_model,\n",
    "    show_versions,\n",
    "    split_string,\n",
    ")\n",
    "from rbc_gem_utils.database.tcdb import TCDB_DB_TAG\n",
    "from rbc_gem_utils.visualization import cmap_map\n",
    "\n",
    "# Display versions of last time notebook ran and worked\n",
    "show_versions()\n",
    "plt.rcParams[\"font.family\"] = \"Arial\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b58a91dd-d23e-412b-8b08-43f607542db9",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0f605c2-0c6c-4a09-af9c-8f2342a2d9ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use XML for groups\n",
    "model_dirpath = get_dirpath(\"model\")\n",
    "model = read_cobra_model(filename=model_dirpath / f\"{GEM_NAME}.xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d775e32d-6926-45f3-aab0-c373b28bc755",
   "metadata": {},
   "source": [
    "## Categorize reactions by KEGG Pathways"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5303d81b-da5c-45e7-9600-c2260dc77552",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Turned into an xlsx where keys are sheet names and values are DataFrames representing the sheet\n",
    "all_figure_data = {}\n",
    "imagetype = \"svg\"\n",
    "transparent = True\n",
    "save_figures = True\n",
    "\n",
    "results_path = get_dirpath(use_temp=\"processed\") / model.id / \"ModelContentOverview\"\n",
    "results_path.mkdir(exist_ok=True, parents=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "338ad25d-9d76-4263-9829-29c4d32a84c7",
   "metadata": {},
   "source": [
    "### Load pathway mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9f87562-e0f7-402b-96e9-1f09639f3763",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pathways = pd.read_csv(\n",
    "    get_dirpath(\"curation\") / \"subsystems.tsv\", sep=\"\\t\", index_col=0, dtype=str\n",
    ")\n",
    "df_pathways = df_pathways.fillna(\"\").reset_index()\n",
    "all_figure_data[\"PathwayMapping\"] = df_pathways.copy()\n",
    "# Group \"Metabolism of other amino acids\" with amino acids rather than treat as \"other\"\n",
    "df_pathways[\"category\"] = df_pathways[\"category\"].replace(\n",
    "    \"Metabolism of other amino acids\", \"Amino acid metabolism\"\n",
    ")\n",
    "df_pathways"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772633f5-bce0-40e1-bd9a-74ba7694ae56",
   "metadata": {},
   "source": [
    "### Visualization options"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bba2ce5f-f04c-4086-9a04-c7210615f6ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Categories that should be excluded from the figure\n",
    "categories_to_exclude = {\"Pseudoreactions\", \"Model total\"}\n",
    "# Main categories in figure and abbreviations, all unmapped categories are mapped to \"Other\"\n",
    "categories_to_keep = {\n",
    "    \"Amino acid metabolism\": mpl.cm.spring,\n",
    "    \"Carbohydrate metabolism\": mpl.cm.Greens,\n",
    "    \"Lipid metabolism\": mpl.cm.Blues,\n",
    "    \"Metabolism of cofactors and vitamins\": mpl.cm.summer,\n",
    "    \"Nucleotide metabolism\": mpl.cm.winter,\n",
    "    \"Reactive species\": mpl.cm.Reds,\n",
    "    \"Transport reactions\": mpl.cm.Purples,\n",
    "    \"Other\": mpl.cm.gray_r,\n",
    "}\n",
    "use_abbrevs = True\n",
    "abbrevs = {\n",
    "    \"Amino acid metabolism\": \"A\",\n",
    "    \"Carbohydrate metabolism\": \"C\",\n",
    "    \"Lipid metabolism\": \"L\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"V\",\n",
    "    \"Nucleotide metabolism\": \"N\",\n",
    "    \"Reactive species\": \"R\",\n",
    "    \"Transport reactions\": \"T\",\n",
    "    \"Other\": \"O\",\n",
    "}\n",
    "barsize = 0.8\n",
    "cmax = 0.8\n",
    "cmin = 0.15\n",
    "\n",
    "\n",
    "colormaps_normal = dict(\n",
    "    zip(\n",
    "        categories_to_keep,\n",
    "        [\n",
    "            cmap_map(lambda x: x * 1, categories_to_keep[k])(cmax)\n",
    "            for k in categories_to_keep\n",
    "        ],\n",
    "    )\n",
    ")\n",
    "colormaps_light = dict(\n",
    "    zip(\n",
    "        categories_to_keep,\n",
    "        [\n",
    "            cmap_map(lambda x: x / 2 + 0.5, categories_to_keep[k])(cmax)\n",
    "            for k in categories_to_keep\n",
    "        ],\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ff20294-7ec0-4c38-8c65-887f7d9734f8",
   "metadata": {},
   "source": [
    "#### Create tables and individual figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "761acfd2-7e95-49c5-9c31-f9677511f9c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = defaultdict(dict)\n",
    "all_reactions = set()\n",
    "df_cat_subsystems = df_pathways.groupby(\"category\")[\"name\"].agg(lambda x: list(x))\n",
    "for category, subsystem_list in df_cat_subsystems.items():\n",
    "    if category not in categories_to_keep and category not in categories_to_exclude:\n",
    "        category = \"Other\"\n",
    "\n",
    "    if category not in data:\n",
    "        data[category] = {\"Reactions\": set(), \"Metabolites\": set(), \"Genes\": set()}\n",
    "    reactions = [\n",
    "        reaction\n",
    "        for group in model.groups.get_by_any(subsystem_list)\n",
    "        for reaction in group.members\n",
    "    ]\n",
    "    for reaction in reactions:\n",
    "        all_reactions.update([reaction])\n",
    "        data[category][\"Reactions\"].update([reaction])\n",
    "        data[category][\"Metabolites\"].update(reaction.metabolites)\n",
    "        data[category][\"Genes\"].update(reaction.genes)\n",
    "\n",
    "data[\"Model total\"] = {\n",
    "    \"Reactions\": model.reactions,\n",
    "    \"Metabolites\": model.metabolites,\n",
    "    \"Genes\": model.genes,\n",
    "}\n",
    "df_data = pd.DataFrame(\n",
    "    {\n",
    "        key: {k: len(x) for k, x in value_dict.items()}\n",
    "        for key, value_dict in data.items()\n",
    "    }\n",
    ").T.sort_index()\n",
    "\n",
    "# Remove columns to exlcude:\n",
    "for to_exclude in categories_to_exclude:\n",
    "    del data[to_exclude]\n",
    "df_data = df_data.sort_values(\"Reactions\", ascending=False)\n",
    "if use_abbrevs:\n",
    "    df_data[\"abbrevs\"] = abbrevs\n",
    "all_figure_data[\"General\"] = df_data.copy()\n",
    "df_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dd2f4f3-4601-47d1-a293-3288545836f2",
   "metadata": {},
   "source": [
    "### Subsystems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b944aee0-5319-4ec6-8c88-973a717111ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_subsystem_data = defaultdict(dict)\n",
    "reactions = set()\n",
    "idx = 0\n",
    "for category, subsystem_list in df_cat_subsystems.items():\n",
    "    if category in categories_to_exclude:\n",
    "        continue\n",
    "    if category not in categories_to_keep:\n",
    "        category = \"Other\"\n",
    "\n",
    "    for group in model.groups.get_by_any(subsystem_list):\n",
    "        df_subsystem_data[idx][\"name\"] = group.name\n",
    "        df_subsystem_data[idx][\"category\"] = category\n",
    "        df_subsystem_data[idx][\"size\"] = len(group.members)\n",
    "        reactions.update(group.members)\n",
    "        idx += 1\n",
    "df_subsystem_data = pd.DataFrame.from_dict(df_subsystem_data, orient=\"index\")\n",
    "# QC check: Ensure mappings worked correctly and there is a color per category\n",
    "# QC check Total number of reactions (excluding pseudoreactions) match sum of the group sizes\n",
    "assert (\n",
    "    len(model.reactions.query(lambda x: not x.subsystem == \"Pseudoreactions\"))\n",
    "    == df_subsystem_data[\"size\"].sum()\n",
    ")\n",
    "df_subsystem_data = df_subsystem_data.sort_values(\n",
    "    by=[\"category\", \"size\"], ascending=[True, False]\n",
    ").reset_index(drop=True)\n",
    "all_figure_data[\"Subsystem Sizes\"] = df_subsystem_data.copy()\n",
    "\n",
    "# Set colors\n",
    "df_subsystem_data_color = df_subsystem_data.set_index(\"name\")\n",
    "colors = {}\n",
    "for category, (names, size) in (\n",
    "    df_subsystem_data.groupby([\"category\"])\n",
    "    .agg({\"name\": lambda x: list(x), \"size\": \"size\"})\n",
    "    .iterrows()\n",
    "):\n",
    "    colors.update(\n",
    "        dict(\n",
    "            zip(\n",
    "                names, categories_to_keep[category](np.linspace(cmax - 0.1, cmin, size))\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "df_subsystem_data_color[\"color\"] = colors\n",
    "df_subsystem_data_color = df_subsystem_data_color.reset_index(drop=False)\n",
    "df_subsystem_data_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b90721e0-093b-42de-8c2b-22258eef99e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_category_data = df_subsystem_data.groupby(\"category\")[\"size\"].sum()\n",
    "assert df_category_data.nunique() == len(categories_to_keep)\n",
    "if use_abbrevs:\n",
    "    df_category_data = pd.concat(\n",
    "        (df_category_data, pd.Series(abbrevs, name=\"abbrev\")), axis=1\n",
    "    )\n",
    "    df_category_data.index.name = \"category\"\n",
    "df_category_data = df_category_data.sort_index(ascending=True).reset_index(drop=False)\n",
    "all_figure_data[\"Category Sizes\"] = df_data.copy()\n",
    "# # # Add colors to table\n",
    "df_category_data_color = df_category_data.set_index(\"category\")\n",
    "df_category_data_color[\"color\"] = {\n",
    "    key: tuple(value(cmax)) for key, value in categories_to_keep.items()\n",
    "}\n",
    "df_category_data_color = df_category_data_color.reset_index(drop=False)\n",
    "df_category_data_color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e18817e2-d0e1-4d26-a9c5-4066fc1272fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_ring0, df_ring1 = df_category_data_color, df_subsystem_data_color\n",
    "\n",
    "fig_ring, ax_ring = plt.subplots(1, 1, figsize=(8, 8))\n",
    "\n",
    "wedgesize = 0.3\n",
    "radius = 0.5\n",
    "startangle = -30\n",
    "ring_idx = 1\n",
    "linewidth = 0.75\n",
    "edgecolor = \"k\"\n",
    "wedges, labels = ax_ring.pie(\n",
    "    df_ring1[\"size\"].values,\n",
    "    radius=radius + (ring_idx * wedgesize),\n",
    "    colors=df_ring1[\"color\"].values,\n",
    "    # labels=df_ring1[\"name\"].values,\n",
    "    wedgeprops=dict(\n",
    "        width=wedgesize, linewidth=linewidth, edgecolor=edgecolor, clip_on=False\n",
    "    ),\n",
    "    textprops={\"fontsize\": \"large\"},\n",
    "    startangle=startangle,\n",
    ")\n",
    "\n",
    "ring_idx = 0\n",
    "\n",
    "# Use for nice legend\n",
    "legend_labels_dict = {\n",
    "    \"Amino acid metabolism\": \"Amino\\nAcids\\n(A)\",\n",
    "    \"Carbohydrate metabolism\": \"\\n       Carbohydrates\\n       (C)\",\n",
    "    \"Lipid metabolism\": \"\\n\\nLipids\\n(L)\",\n",
    "    \"Metabolism of cofactors and vitamins\": \"Cofactors  \\nand Vitamins  \\n(V)    \",\n",
    "    \"Nucleotide metabolism\": \"\\nNucleotides  \\n(N)  \",\n",
    "    \"Other\": \"    Other\\n    (O)\",\n",
    "    \"Reactive species\": \"\\nReactive\\nSpecies\\n(R)\",\n",
    "    \"Transport reactions\": \"Transport\\nReactions\\n(T)\",\n",
    "}\n",
    "wedges, labels = ax_ring.pie(\n",
    "    df_ring0[\"size\"].values,\n",
    "    radius=radius + (ring_idx * wedgesize),\n",
    "    colors=df_ring0[\"color\"].values,\n",
    "    labels=[f\"{legend_labels_dict[x]}\" for x in df_ring0[\"category\"].values],\n",
    "    wedgeprops=dict(\n",
    "        width=wedgesize, linewidth=linewidth, edgecolor=edgecolor, clip_on=False\n",
    "    ),\n",
    "    textprops={\"fontsize\": 18, \"va\": \"center\", \"ha\": \"center\"},\n",
    "    startangle=startangle,\n",
    "    labeldistance=2.0,\n",
    ")\n",
    "\n",
    "fig_ring.tight_layout()\n",
    "if save_figures:\n",
    "    fig_ring.savefig(\n",
    "        results_path / f\"ModelContent_PanelA.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca7391e1-caef-47d4-9441-a9781607a3b1",
   "metadata": {},
   "source": [
    "#### Transporter categorized"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9e8ebba-e30f-488f-addc-fe92b990a4d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "database_dirpath = get_dirpath(\"database\", TCDB_DB_TAG)\n",
    "\n",
    "df_tcdb_data = pd.read_csv(\n",
    "    database_dirpath / f\"TCDB_{GEM_NAME}.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "    dtype=str,\n",
    ").fillna(\"\")\n",
    "df_tcdb_data = df_tcdb_data.loc[\n",
    "    :, [\"tcdb\", \"subfamily\", \"family\", \"subclass\", \"class\", \"fam_name\"]\n",
    "].drop_duplicates()\n",
    "\n",
    "groupby_key = \"class\"\n",
    "groupby_regex = {\n",
    "    \"class\": r\"^\\w+\",\n",
    "    \"subclass\": r\"^\\w+\\W\\w+\",\n",
    "    \"family\": r\"^\\w+\\W\\w+\\W\\w+\",\n",
    "    \"subfamily\": r\"^\\w+\\W\\w+\\W\\w+\\W\\w+\",\n",
    "    \"tcdb\": r\"^\\w+\\W\\w+\\W\\w+\\W\\w+\\W\\w+\",\n",
    "}[groupby_key]\n",
    "\n",
    "tcdb_mappings = {\n",
    "    \"class\": {\n",
    "        \"1\": \"Channels/Pores\",\n",
    "        \"2\": \"Electrochemical Potential-driven Transporters\",\n",
    "        \"3\": \"Primary Active Transporters\",\n",
    "        \"4\": \"Group Translocators\",\n",
    "        \"5\": \"Transmembrane Electron Carriers\",\n",
    "        \"8\": \"Accessory Factors Involved in Transport\",\n",
    "        \"9\": \"Incompletely Characterized Transport Systems\",\n",
    "    },\n",
    "    \"subclass\": {},\n",
    "    \"family\": df_tcdb_data.set_index(\"family\")[\"fam_name\"]\n",
    "    .replace(\"\", float(\"nan\"))\n",
    "    .dropna()\n",
    "    .drop_duplicates()\n",
    "    .to_dict(),\n",
    "    \"subfamily\": {},\n",
    "}[groupby_key]\n",
    "\n",
    "df_tcdb_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c8630fa-b4cd-4619-90a4-be2cf6a3b55d",
   "metadata": {},
   "outputs": [],
   "source": [
    "transport_reactions = model.reactions.query(\n",
    "    lambda x: x in data[\"Transport reactions\"][\"Reactions\"]\n",
    ")\n",
    "\n",
    "df_transport_reactions = pd.DataFrame(\n",
    "    {\"genes\": {r.id: {g.id for g in list(r.genes)} for r in transport_reactions}}\n",
    ")\n",
    "df_transport_reactions.index.name = \"reactions\"\n",
    "df_transport_reactions = df_transport_reactions.explode(\"genes\").reset_index(drop=False)\n",
    "\n",
    "\n",
    "df_model_genes = get_annotation_df(model.genes, [\"ncbigene\", \"tcdb\"]).rename(\n",
    "    {\"id\": \"genes\"}, axis=1\n",
    ")\n",
    "df_model_genes = df_model_genes[df_model_genes[\"tcdb\"].notna()]\n",
    "df_model_genes[\"tcdb\"] = df_model_genes[\"tcdb\"].apply(\n",
    "    lambda x: split_string(x) if isinstance(x, str) else x\n",
    ")\n",
    "df_model_genes = df_model_genes.explode(\"tcdb\")\n",
    "df_model_genes = df_model_genes.sort_values(by=\"tcdb\").reset_index(drop=True)\n",
    "\n",
    "df_transport_reactions = (\n",
    "    df_transport_reactions.merge(\n",
    "        df_model_genes, left_on=\"genes\", right_on=\"genes\", how=\"left\"\n",
    "    )\n",
    "    .merge(df_tcdb_data, left_on=\"tcdb\", right_on=\"tcdb\", how=\"left\")\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "\n",
    "df_transport_reactions = df_transport_reactions.loc[\n",
    "    :, [\"reactions\", \"genes\", \"tcdb\", \"subfamily\", \"family\", \"subclass\", \"class\"]\n",
    "]\n",
    "df_transport_reactions = df_transport_reactions.drop_duplicates()\n",
    "df_transport_reactions = df_transport_reactions.copy()\n",
    "\n",
    "df_transport_reactions[groupby_key] = (\n",
    "    df_transport_reactions[\"tcdb\"].str.findall(groupby_regex).explode()\n",
    ")\n",
    "df_transport_reactions[groupby_key] = df_transport_reactions[groupby_key].fillna(\n",
    "    \"Unclassified\"\n",
    ")\n",
    "df_transport_reactions = pd.merge(\n",
    "    df_transport_reactions.groupby(groupby_key)[[\"reactions\", \"genes\"]].agg(\n",
    "        lambda x: x.dropna().nunique()\n",
    "    ),\n",
    "    df_transport_reactions.groupby(groupby_key)[[\"reactions\", \"genes\"]].agg(\n",
    "        lambda x: build_string(x.dropna().unique())\n",
    "    ),\n",
    "    left_index=True,\n",
    "    right_index=True,\n",
    "    how=\"left\",\n",
    "    suffixes=(\"\", \".ids\"),\n",
    ")\n",
    "all_figure_data[f\"Transporters_{groupby_key}\"] = df_transport_reactions.reset_index(\n",
    "    drop=False\n",
    ").copy()\n",
    "df_transport_reactions = df_transport_reactions.drop(\n",
    "    labels=[c for c in df_transport_reactions.columns if c.endswith(\".ids\")], axis=1\n",
    ").astype(int)\n",
    "df_transport_reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06c54ba1-c9b8-4509-a9f8-45c33264360b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_transport_reactions.copy()\n",
    "\n",
    "barsize = 0.7\n",
    "edgecolor = \"black\"\n",
    "edgewidth = 1\n",
    "columns = [\"genes\", \"reactions\"]\n",
    "labels = df.index\n",
    "\n",
    "\n",
    "totals = df.sum(axis=0)\n",
    "df = round(df / totals * 100, 3)\n",
    "df = df[::-1].T\n",
    "df = df.loc[columns]\n",
    "\n",
    "color_dict = dict(\n",
    "    zip(\n",
    "        labels,\n",
    "        categories_to_keep[\"Transport reactions\"](np.linspace(cmax, 0.05, len(labels))),\n",
    "    )\n",
    ")\n",
    "\n",
    "fig_transport, ax_transport = plt.subplots(1, 1, figsize=(7, 6))\n",
    "sns.despine(fig_transport)\n",
    "offset = np.zeros(len(df.index))\n",
    "for idx, (label, series) in enumerate(df.items()):\n",
    "    rects = ax_transport.bar(\n",
    "        [x.capitalize() for x in series.index],\n",
    "        series.values,\n",
    "        barsize,\n",
    "        bottom=offset,\n",
    "        color=color_dict[label],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "    # if idx == len(df.columns) - 1:\n",
    "    #     ax.bar_label(rects, labels=totals.values, label_type='edge', color=\"black\",  fmt='{:,.0f}', padding=1)\n",
    "\n",
    "\n",
    "ticks = np.linspace(0, 100, 5)\n",
    "ax_transport.set_yticks(ticks, [f\"{v:.0f}%\" for v in ticks])\n",
    "ax_transport.set_yticklabels(\n",
    "    ax_transport.get_yticklabels(), fontdict={\"size\": \"xx-large\"}\n",
    ")\n",
    "ax_transport.xaxis.set_tick_params(labelsize=\"xx-large\")\n",
    "fig_transport.suptitle(\"Transport Reactions (T)\", fontsize=20, y=0.975)\n",
    "legend = ax_transport.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(\n",
    "            facecolor=color,\n",
    "            label=tcdb_mappings.get(label, label),\n",
    "            edgecolor=edgecolor,\n",
    "            linewidth=edgewidth,\n",
    "        )\n",
    "        for label, color in color_dict.items()\n",
    "    ],\n",
    "    labels=[\n",
    "        (\n",
    "            \"\\n    \".join(\n",
    "                textwrap.wrap(f\"{label}. \" + tcdb_mappings.get(label, label), 30)\n",
    "            )\n",
    "            if label != \"Unclassified\"\n",
    "            else label\n",
    "        )\n",
    "        for label in list(color_dict)\n",
    "    ],\n",
    "    loc=\"center left\",\n",
    "    bbox_to_anchor=(1, 0.5),\n",
    "    frameon=False,\n",
    "    fontsize=\"x-large\",\n",
    "    labelspacing=1.5,\n",
    ")\n",
    "for patch in legend.get_patches():\n",
    "    patch.set_height(20)\n",
    "    patch.set_y(-5)\n",
    "fig_transport.tight_layout()\n",
    "if save_figures:\n",
    "    fig_transport.savefig(\n",
    "        results_path / f\"ModelContent_PanelB.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34d46c14-6a93-4e43-a5f0-7cbf49c6c6e9",
   "metadata": {},
   "source": [
    "### Reaction content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6498f2d1-cf0b-47fd-a65c-ea5ff3553ce6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_reactions = defaultdict(dict)\n",
    "\n",
    "for category in categories_to_keep:\n",
    "    reactions = data[category][\"Reactions\"]\n",
    "    gene_associated = {rxn for rxn in reactions if rxn.gene_reaction_rule}\n",
    "    known_spontaneous = {\n",
    "        rxn\n",
    "        for rxn in reactions\n",
    "        if int(float(rxn.annotation.get(\"spontaneous\", 0)))\n",
    "        and rxn not in gene_associated\n",
    "    }\n",
    "    unknown = {\n",
    "        rxn\n",
    "        for rxn in reactions\n",
    "        if not rxn in gene_associated and not rxn in known_spontaneous\n",
    "    }\n",
    "    df_reactions[category][\"Gene associated\"] = len(gene_associated)\n",
    "    df_reactions[category][\"Spontaneous\"] = len(known_spontaneous)\n",
    "    df_reactions[category][\"Unknown\"] = len(unknown)\n",
    "    df_reactions[category][\"Total\"] = len(reactions)\n",
    "\n",
    "    df_reactions[category][\"Gene associated.ids\"] = build_string(\n",
    "        sorted([x.id for x in gene_associated])\n",
    "    )\n",
    "    df_reactions[category][\"Spontaneous.ids\"] = build_string(\n",
    "        sorted([x.id for x in known_spontaneous])\n",
    "    )\n",
    "    df_reactions[category][\"Unknown.ids\"] = build_string(\n",
    "        sorted([x.id for x in unknown])\n",
    "    )\n",
    "\n",
    "df_reactions = pd.DataFrame.from_dict(df_reactions).T\n",
    "if use_abbrevs:\n",
    "    df_reactions[\"abbrevs\"] = abbrevs\n",
    "# QC check to ensure no reactions are missing\n",
    "assert (\n",
    "    df_reactions[[\"Gene associated\", \"Spontaneous\", \"Unknown\"]].sum(axis=1)\n",
    "    == df_reactions[\"Total\"]\n",
    ").all()\n",
    "assert df_reactions[\"Total\"].sum() == len(\n",
    "    model.reactions.query(lambda x: not x.subsystem == \"Pseudoreactions\")\n",
    ")\n",
    "df_reactions = df_reactions.sort_values(by=\"Total\", ascending=True)\n",
    "\n",
    "all_figure_data[\"Reactions\"] = df_reactions.copy()\n",
    "df_reactions = df_reactions.drop(\n",
    "    labels=[c for c in df_reactions.columns if c.endswith(\".ids\") or c == \"abbrevs\"],\n",
    "    axis=1,\n",
    ").astype(int)\n",
    "df_reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d9aac08-88cf-48af-bd29-b4126ece7b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_reactions.copy()\n",
    "\n",
    "fontsize = 12\n",
    "percentage_col = \"Gene associated\"\n",
    "only_total = False\n",
    "colormaps_dict = {\n",
    "    \"Gene associated\": colormaps_normal,\n",
    "    \"Spontaneous\": colormaps_light,\n",
    "    \"Unknown\": colormaps_light,\n",
    "}\n",
    "hatches_dict = {\n",
    "    \"Gene associated\": None,\n",
    "    \"Spontaneous\": \"///\",\n",
    "    \"Unknown\": None,\n",
    "}\n",
    "barsize = 0.75\n",
    "edgecolor = \"black\"\n",
    "edgewidth = 0.5\n",
    "padding = 2\n",
    "barchart_figsize = (5, 3)\n",
    "\n",
    "fig_reactions, ax_rxns = plt.subplots(1, 1, figsize=barchart_figsize)\n",
    "sns.despine(fig_reactions)\n",
    "\n",
    "offset = np.zeros(len(df.index))\n",
    "columns_to_plot = (\n",
    "    [\"Total\"] if only_total else [col for col in df.columns if col != \"Total\"]\n",
    ")\n",
    "for idx, (label, series) in enumerate(df[columns_to_plot].items()):\n",
    "    rects = ax_rxns.barh(\n",
    "        [\n",
    "            f\"({abbrevs.get(x)})\" if abbrevs.get(x) and use_abbrevs else x\n",
    "            for x in series.index\n",
    "        ],\n",
    "        series.values,\n",
    "        height=barsize,\n",
    "        hatch=hatches_dict.get(label, None),\n",
    "        left=offset,\n",
    "        label=f\"% - {percentage_col}\",\n",
    "        color=[colormaps_dict.get(label, colormaps_normal)[k] for k in series.index],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "\n",
    "percents = [\n",
    "    \"{:,.1f}%\".format(v) for v in 100 * (df[percentage_col] / df[\"Total\"]).values\n",
    "]\n",
    "ax_rxns.bar_label(\n",
    "    rects,\n",
    "    labels=percents,\n",
    "    label_type=\"edge\",\n",
    "    color=\"black\",\n",
    "    padding=padding,\n",
    "    fontsize=fontsize,\n",
    ")\n",
    "ax_rxns.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "    ],\n",
    "    labels=[\"/// - Spontaneous\", f\"% - {percentage_col}\"],\n",
    "    frameon=False,\n",
    "    loc=\"lower right\",\n",
    "    fontsize=fontsize,\n",
    ")\n",
    "ax_rxns.set_xlabel(\"Number of reactions\", fontsize=fontsize)\n",
    "ax_rxns.yaxis.set_tick_params(labelsize=fontsize)\n",
    "if save_figures:\n",
    "    fig_reactions.savefig(\n",
    "        results_path / f\"ModelContent_PanelC.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3fc95d0-0baa-4a92-a710-31f5fb2e9f35",
   "metadata": {},
   "source": [
    "### Metabolite content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "740ae86a-8c10-467a-9ae5-e6853d3569ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_metabolites = defaultdict(dict)\n",
    "\n",
    "for category in categories_to_keep:\n",
    "    metabolites = data[category][\"Metabolites\"]\n",
    "    # metabolites = set([met for met in metabolites if met.compartment != \"e\"])\n",
    "    # Genes that are not unique to one category\n",
    "    shared = {\n",
    "        x\n",
    "        for x in metabolites\n",
    "        if any(\n",
    "            [\n",
    "                set(x.reactions).intersection(data[cat][\"Reactions\"])\n",
    "                for cat in data.keys()\n",
    "                if cat != category\n",
    "            ]\n",
    "        )\n",
    "    }\n",
    "    unique = {x for x in metabolites if not x in shared}\n",
    "    df_metabolites[category][\"Shared\"] = len(shared)\n",
    "    df_metabolites[category][\"Unique\"] = len(unique)\n",
    "    df_metabolites[category][\"Total\"] = len(metabolites)\n",
    "\n",
    "    df_metabolites[category][\"Shared.ids\"] = build_string(\n",
    "        sorted([x.id for x in shared])\n",
    "    )\n",
    "    df_metabolites[category][\"Unique.ids\"] = build_string(\n",
    "        sorted([x.id for x in unique])\n",
    "    )\n",
    "\n",
    "# To DataFrame\n",
    "df_metabolites = pd.DataFrame.from_dict(df_metabolites).T\n",
    "if use_abbrevs:\n",
    "    df_metabolites[\"abbrevs\"] = abbrevs\n",
    "# QC check to ensure no genes are missing\n",
    "assert (\n",
    "    df_metabolites[[\"Shared\", \"Unique\"]].sum(axis=1) == df_metabolites[\"Total\"]\n",
    ").all()\n",
    "df_metabolites = df_metabolites.sort_values(by=\"Total\", ascending=True)\n",
    "all_figure_data[\"Metabolites\"] = df_metabolites.copy()\n",
    "df_metabolites = df_metabolites.drop(\n",
    "    labels=[c for c in df_metabolites.columns if c.endswith(\".ids\") or c == \"abbrevs\"],\n",
    "    axis=1,\n",
    ").astype(int)\n",
    "df_metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "006aa8cb-a2df-4db7-a322-40c491570880",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_metabolites.copy()\n",
    "\n",
    "fontsize = 12\n",
    "percentage_col = \"Unique\"\n",
    "only_total = False\n",
    "colormaps_dict = {\n",
    "    \"Unique\": colormaps_normal,\n",
    "    \"Shared\": colormaps_light,\n",
    "    \"Total\": colormaps_normal,\n",
    "}\n",
    "hatches_dict = {}\n",
    "barsize = 0.75\n",
    "edgecolor = \"black\"\n",
    "edgewidth = 0.5\n",
    "padding = 2\n",
    "barchart_figsize = (5, 3)\n",
    "\n",
    "fig_metabolites, ax_mets = plt.subplots(1, 1, figsize=barchart_figsize)\n",
    "sns.despine(fig_metabolites)\n",
    "\n",
    "offset = np.zeros(len(df.index))\n",
    "columns_to_plot = [\"Total\"] if only_total else [\"Unique\", \"Shared\"]\n",
    "for idx, (label, series) in enumerate(df[columns_to_plot].items()):\n",
    "    rects = ax_mets.barh(\n",
    "        [\n",
    "            f\"({abbrevs.get(x)})\" if abbrevs.get(x) and use_abbrevs else x\n",
    "            for x in series.index\n",
    "        ],\n",
    "        series.values,\n",
    "        height=barsize,\n",
    "        hatch=hatches_dict.get(label, None),\n",
    "        left=offset,\n",
    "        label=f\"% - {percentage_col}\",\n",
    "        color=[colormaps_dict.get(label, colormaps_normal)[k] for k in series.index],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "\n",
    "percents = [\n",
    "    \"{:,.1f}%\".format(v) for v in 100 * (df[percentage_col] / df[\"Total\"]).values\n",
    "]\n",
    "ax_mets.bar_label(\n",
    "    rects,\n",
    "    labels=percents,\n",
    "    label_type=\"edge\",\n",
    "    color=\"black\",\n",
    "    padding=padding,\n",
    "    fontsize=fontsize,\n",
    ")\n",
    "ax_mets.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "    ],\n",
    "    labels=[f\"% - {percentage_col}\"],\n",
    "    frameon=False,\n",
    "    loc=\"lower right\",\n",
    "    fontsize=fontsize,\n",
    ")\n",
    "ax_mets.set_xlabel(\"Number of metabolites\", fontsize=fontsize)\n",
    "ax_mets.yaxis.set_tick_params(labelsize=fontsize)\n",
    "if save_figures:\n",
    "    fig_metabolites.savefig(\n",
    "        results_path / f\"ModelContent_PanelD.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3b60906-2935-479e-b9ce-007036e05878",
   "metadata": {},
   "source": [
    "### Gene content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbdb1533-63b2-44f8-9a5d-dd70dc66e247",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_genes = defaultdict(dict)\n",
    "\n",
    "for category in categories_to_keep:\n",
    "    genes = data[category][\"Genes\"]\n",
    "    # Genes that are not unique to one category\n",
    "    shared = {\n",
    "        x\n",
    "        for x in genes\n",
    "        if any(\n",
    "            [\n",
    "                set(x.reactions).intersection(data[cat][\"Reactions\"])\n",
    "                for cat in data.keys()\n",
    "                if cat != category\n",
    "            ]\n",
    "        )\n",
    "    }\n",
    "    unique = {x for x in genes if not x in shared}\n",
    "    df_genes[category][\"Shared\"] = len(shared)\n",
    "    df_genes[category][\"Unique\"] = len(unique)\n",
    "    df_genes[category][\"Total\"] = len(genes)\n",
    "\n",
    "    df_genes[category][\"Shared.ids\"] = build_string(sorted([x.id for x in shared]))\n",
    "    df_genes[category][\"Unique.ids\"] = build_string(sorted([x.id for x in unique]))\n",
    "\n",
    "# To DataFrame\n",
    "df_genes = pd.DataFrame.from_dict(df_genes).T\n",
    "if use_abbrevs:\n",
    "    df_genes[\"abbrevs\"] = abbrevs\n",
    "# QC check to ensure no genes are missing\n",
    "assert (df_genes[[\"Shared\", \"Unique\"]].sum(axis=1) == df_genes[\"Total\"]).all()\n",
    "df_genes = df_genes.sort_values(by=\"Total\", ascending=True)\n",
    "all_figure_data[\"Genes\"] = df_genes.copy()\n",
    "df_genes = df_genes.drop(\n",
    "    labels=[c for c in df_genes.columns if c.endswith(\".ids\") or c == \"abbrevs\"], axis=1\n",
    ").astype(int)\n",
    "df_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f6da53c-ff69-4e9a-8e29-73061e661de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df_genes.copy()\n",
    "\n",
    "fontsize = 12\n",
    "percentage_col = \"Unique\"\n",
    "only_total = False\n",
    "colormaps_dict = {\n",
    "    \"Unique\": colormaps_normal,\n",
    "    \"Shared\": colormaps_light,\n",
    "    \"Total\": colormaps_normal,\n",
    "}\n",
    "hatches_dict = {}\n",
    "barsize = 0.75\n",
    "edgecolor = \"black\"\n",
    "edgewidth = 0.5\n",
    "padding = 2\n",
    "barchart_figsize = (5, 3)\n",
    "\n",
    "fig_genes, ax_genes = plt.subplots(1, 1, figsize=barchart_figsize)\n",
    "sns.despine(fig_genes)\n",
    "\n",
    "offset = np.zeros(len(df.index))\n",
    "columns_to_plot = [\"Total\"] if only_total else [\"Unique\", \"Shared\"]\n",
    "for idx, (label, series) in enumerate(df[columns_to_plot].items()):\n",
    "    rects = ax_genes.barh(\n",
    "        [\n",
    "            f\"({abbrevs.get(x)})\" if abbrevs.get(x) and use_abbrevs else x\n",
    "            for x in series.index\n",
    "        ],\n",
    "        series.values,\n",
    "        height=barsize,\n",
    "        hatch=hatches_dict.get(label, None),\n",
    "        left=offset,\n",
    "        label=f\"% - {percentage_col}\",\n",
    "        color=[colormaps_dict.get(label, colormaps_normal)[k] for k in series.index],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "\n",
    "percents = [\n",
    "    \"{:,.1f}%\".format(v) for v in 100 * (df[percentage_col] / df[\"Total\"]).values\n",
    "]\n",
    "ax_genes.bar_label(\n",
    "    rects,\n",
    "    labels=percents,\n",
    "    label_type=\"edge\",\n",
    "    color=\"black\",\n",
    "    padding=padding,\n",
    "    fontsize=fontsize,\n",
    ")\n",
    "ax_genes.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "    ],\n",
    "    labels=[f\"% - {percentage_col}\"],\n",
    "    frameon=False,\n",
    "    loc=\"lower right\",\n",
    "    fontsize=fontsize,\n",
    ")\n",
    "ax_genes.set_xlabel(\"Number of genes\", fontsize=fontsize)\n",
    "ax_genes.yaxis.set_tick_params(labelsize=fontsize)\n",
    "if save_figures:\n",
    "    fig_genes.savefig(\n",
    "        results_path / f\"ModelContent_PanelE_{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97257335-0d7b-4328-acf5-b163ce08e82d",
   "metadata": {},
   "source": [
    "## Export Figures and Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d2a0a2-43ff-49c5-a001-f1c8cc22fdd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(all_figure_data.keys())\n",
    "with pd.ExcelWriter(results_path / f\"ModelContentOverviewData.xlsx\") as writer:\n",
    "    for sheet_name, df in all_figure_data.items():\n",
    "        df.to_excel(writer, sheet_name=sheet_name, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bb44be6-76b6-413f-8b01-463b09864876",
   "metadata": {},
   "source": [
    "### Aggregate Panels for visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4441faa-d7ef-41c0-86c9-f20d27f7d445",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(3, 1, figsize=(5, 10))\n",
    "sns.despine(fig)\n",
    "(ax_rxns, ax_mets, ax_genes) = axes.flatten()\n",
    "\n",
    "\n",
    "only_total = False\n",
    "barsize = 0.75\n",
    "edgecolor = \"black\"\n",
    "edgewidth = 1\n",
    "padding = 2\n",
    "barchart_figsize = (5, 3)\n",
    "\n",
    "\n",
    "## Reaction content\n",
    "df = df_reactions.copy()\n",
    "percentage_col = \"Gene associated\"\n",
    "colormaps_dict = {\n",
    "    \"Gene associated\": colormaps_normal,\n",
    "    \"Spontaneous\": colormaps_light,\n",
    "    \"Unknown\": colormaps_light,\n",
    "}\n",
    "hatches_dict = {\n",
    "    \"Gene associated\": None,\n",
    "    \"Spontaneous\": \"///\",\n",
    "    \"Unknown\": None,\n",
    "}\n",
    "\n",
    "offset = np.zeros(len(df.index))\n",
    "columns_to_plot = (\n",
    "    [\"Total\"] if only_total else [col for col in df.columns if col != \"Total\"]\n",
    ")\n",
    "for idx, (label, series) in enumerate(df[columns_to_plot].items()):\n",
    "    rects = ax_rxns.barh(\n",
    "        [\n",
    "            f\"({abbrevs.get(x)})\" if abbrevs.get(x) and use_abbrevs else x\n",
    "            for x in series.index\n",
    "        ],\n",
    "        series.values,\n",
    "        height=barsize,\n",
    "        hatch=hatches_dict.get(label, None),\n",
    "        left=offset,\n",
    "        label=f\"% - {percentage_col}\",\n",
    "        color=[colormaps_dict.get(label, colormaps_normal)[k] for k in series.index],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "\n",
    "ax_rxns.bar_label(\n",
    "    rects,\n",
    "    labels=[\n",
    "        \"{:,.0f}%\".format(v) for v in 100 * (df[percentage_col] / df[\"Total\"]).values\n",
    "    ],\n",
    "    label_type=\"edge\",\n",
    "    color=\"black\",\n",
    "    padding=padding,\n",
    "    fontsize=\"x-large\",\n",
    ")\n",
    "ax_rxns.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "    ],\n",
    "    labels=[\"/// - Spontaneous\", f\"% - {percentage_col}\"],\n",
    "    frameon=False,\n",
    "    loc=\"lower right\",\n",
    "    fontsize=\"x-large\",\n",
    ")\n",
    "attribute_type = \"reactions\"\n",
    "ax_rxns.set_xlabel(f\"Number of {attribute_type}\", fontsize=\"xx-large\")\n",
    "ax_rxns.xaxis.set_tick_params(labelsize=\"x-large\")\n",
    "ax_rxns.set_xticks([0, 200, 400, 600, 800], minor=False)\n",
    "ax_rxns.set_xticks([100, 300, 500, 700], minor=True)\n",
    "ax_rxns.yaxis.set_tick_params(labelsize=\"xx-large\")\n",
    "\n",
    "\n",
    "## Metabolite content\n",
    "df = df_metabolites.copy()\n",
    "percentage_col = \"Unique\"\n",
    "colormaps_dict = {\n",
    "    \"Unique\": colormaps_normal,\n",
    "    \"Shared\": colormaps_light,\n",
    "    \"Total\": colormaps_normal,\n",
    "}\n",
    "hatches_dict = {}\n",
    "\n",
    "offset = np.zeros(len(df.index))\n",
    "columns_to_plot = [\"Total\"] if only_total else [\"Unique\", \"Shared\"]\n",
    "for idx, (label, series) in enumerate(df[columns_to_plot].items()):\n",
    "    rects = ax_mets.barh(\n",
    "        [\n",
    "            f\"({abbrevs.get(x)})\" if abbrevs.get(x) and use_abbrevs else x\n",
    "            for x in series.index\n",
    "        ],\n",
    "        series.values,\n",
    "        height=barsize,\n",
    "        hatch=hatches_dict.get(label, None),\n",
    "        left=offset,\n",
    "        label=f\"% - {percentage_col}\",\n",
    "        color=[colormaps_dict.get(label, colormaps_normal)[k] for k in series.index],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "\n",
    "ax_mets.bar_label(\n",
    "    rects,\n",
    "    labels=[\n",
    "        \"{:,.0f}%\".format(v) for v in 100 * (df[percentage_col] / df[\"Total\"]).values\n",
    "    ],\n",
    "    label_type=\"edge\",\n",
    "    color=\"black\",\n",
    "    padding=padding,\n",
    "    fontsize=\"x-large\",\n",
    ")\n",
    "ax_mets.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "    ],\n",
    "    labels=[f\"% - {percentage_col}\"],\n",
    "    frameon=False,\n",
    "    loc=\"lower right\",\n",
    "    fontsize=\"x-large\",\n",
    ")\n",
    "attribute_type = \"metabolites\"\n",
    "ax_mets.set_xlabel(f\"Number of {attribute_type}\", fontsize=\"xx-large\")\n",
    "ax_mets.xaxis.set_tick_params(labelsize=\"x-large\")\n",
    "ax_mets.set_xticks([0, 200, 400, 600, 800], minor=False)\n",
    "ax_mets.set_xticks([100, 300, 500, 700], minor=True)\n",
    "ax_mets.yaxis.set_tick_params(labelsize=\"xx-large\")\n",
    "\n",
    "\n",
    "## Gene content\n",
    "df = df_genes.copy()\n",
    "percentage_col = \"Unique\"\n",
    "colormaps_dict = {\n",
    "    \"Unique\": colormaps_normal,\n",
    "    \"Shared\": colormaps_light,\n",
    "    \"Total\": colormaps_normal,\n",
    "}\n",
    "hatches_dict = {}\n",
    "\n",
    "offset = np.zeros(len(df.index))\n",
    "columns_to_plot = [\"Total\"] if only_total else [\"Unique\", \"Shared\"]\n",
    "for idx, (label, series) in enumerate(df[columns_to_plot].items()):\n",
    "    rects = ax_genes.barh(\n",
    "        [\n",
    "            f\"({abbrevs.get(x)})\" if abbrevs.get(x) and use_abbrevs else x\n",
    "            for x in series.index\n",
    "        ],\n",
    "        series.values,\n",
    "        height=barsize,\n",
    "        hatch=hatches_dict.get(label, None),\n",
    "        left=offset,\n",
    "        label=f\"% - {percentage_col}\",\n",
    "        color=[colormaps_dict.get(label, colormaps_normal)[k] for k in series.index],\n",
    "        edgecolor=edgecolor,\n",
    "        linewidth=edgewidth,\n",
    "    )\n",
    "    offset += series.values\n",
    "\n",
    "ax_genes.bar_label(\n",
    "    rects,\n",
    "    labels=[\n",
    "        \"{:,.0f}%\".format(v) for v in 100 * (df[percentage_col] / df[\"Total\"]).values\n",
    "    ],\n",
    "    label_type=\"edge\",\n",
    "    color=\"black\",\n",
    "    padding=padding,\n",
    "    fontsize=\"x-large\",\n",
    ")\n",
    "ax_genes.legend(\n",
    "    handles=[\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "        mpl.patches.Patch(color=\"white\"),\n",
    "    ],\n",
    "    labels=[f\"% - {percentage_col}\"],\n",
    "    frameon=False,\n",
    "    loc=\"lower right\",\n",
    "    fontsize=\"x-large\",\n",
    ")\n",
    "attribute_type = \"genes\"\n",
    "ax_genes.set_xlabel(f\"Number of {attribute_type}\", fontsize=\"xx-large\")\n",
    "ax_genes.xaxis.set_tick_params(labelsize=\"x-large\")\n",
    "ax_genes.set_xticks([0, 50, 100, 150], minor=False)\n",
    "ax_genes.set_xticks([25, 75, 125, 175], minor=True)\n",
    "ax_genes.yaxis.set_tick_params(labelsize=\"xx-large\")\n",
    "fig.align_ylabels([ax_rxns, ax_mets, ax_genes])\n",
    "fig.tight_layout()\n",
    "if save_figures:\n",
    "    fig.savefig(\n",
    "        results_path / f\"ModelContent_PanelCDE.{imagetype}\",\n",
    "        transparent=transparent,\n",
    "        format=imagetype,\n",
    "    )\n",
    "fig;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7d1ad7e-612f-4704-85c4-87518304bc94",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
