{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "291414c1-9b6f-4338-b0bc-13c8958d259d",
   "metadata": {},
   "source": [
    "# RBC-GEM 0.1.1 --> 0.2.0\n",
    "## Update iAB-RBC-283 model \n",
    "\n",
    "The purpose of this notebook is to update the iAB-RBC-283 reconstruction for the following purposes:\n",
    "1. To serve as a base/draft model for the initial expansion of iAB-RBC-283 to RBC-GEM.\n",
    "2. To serve as a model for comparison to the new reconstruction.\n",
    "3. To link to the HumanGEM reconstruction annotations where possible in the model.\n",
    "\n",
    "No stoichiometric changes are made in this minor update, however BIGG identifiers are altered in order to enable better comparision to the RBC-GEM. Furthermore, associations to Recon3D and HumanGEM are made here, as subsequent changes past 0.2.0 will be in an effort to modernize to current COBRA standards before iterative expansion.\n",
    "\n",
    "RBC-GEM 0.2.0 will remain identified as iAB-RBC-283 in order to help future comparisions.\n",
    "\n",
    "Bordbar, A., Jamshidi, N. & Palsson, B.O. iAB-RBC-283: A proteomically derived knowledge-base of erythrocyte metabolism that can be used to simulate its physiological and patho-physiological states. BMC Syst Biol 5, 110 (2011). https://doi.org/10.1186/1752-0509-5-110\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "54ef02a3-2f02-4072-a04d-36e2689794c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Package Information\n",
      "-------------------\n",
      "rbc-gem-utils 0.0.1\n",
      "\n",
      "Dependency Information\n",
      "----------------------\n",
      "cobra      0.29.0\n",
      "depinfo     2.2.0\n",
      "matplotlib  3.8.2\n",
      "memote     0.16.1\n",
      "notebook    7.0.6\n",
      "requests   2.31.0\n",
      "scipy      1.11.4\n",
      "seaborn    0.13.0\n",
      "\n",
      "Build Tools Information\n",
      "-----------------------\n",
      "pip        23.3.1\n",
      "setuptools 68.2.2\n",
      "wheel      0.41.2\n",
      "\n",
      "Platform Information\n",
      "--------------------\n",
      "Darwin  22.6.0-x86_64\n",
      "CPython        3.12.0\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter, defaultdict\n",
    "\n",
    "import pandas as pd\n",
    "import cobra\n",
    "from cobra.core import Group\n",
    "\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION, \n",
    "    ROOT_PATH, \n",
    "    show_versions,\n",
    "    read_rbc_model, write_rbc_model, \n",
    "    read_cobra_model,\n",
    ")\n",
    "from rbc_gem_utils.util import split_string, build_string\n",
    "from rbc_gem_utils.qc import standardardize_metabolite_formulas\n",
    "\n",
    "# Display versions of last time notebook ran and worked\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6abc5fbf-6345-4620-a0a2-1e23fc6dd504",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "51309b09-8cd3-47f8-a7a4-db996ff17e8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <td><strong>Attribute</strong></td>\n",
       "      <td><strong>Description</strong></td>\n",
       "      <td><strong>Value</strong></td>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td><pre>solver</pre></td>\n",
       "      <td>Mathematical optimization solver</td>\n",
       "      <td>gurobi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>tolerance</pre></td>\n",
       "        <td>General solver tolerance (feasibility, integrality, etc.)</td>\n",
       "        <td>1e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>lower_bound</pre></td>\n",
       "        <td>Default reaction lower bound</td>\n",
       "        <td>-1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>upper_bound</pre></td>\n",
       "        <td>Default reaction upper bound</td>\n",
       "        <td>1000.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>processes</pre></td>\n",
       "        <td>Number of parallel processes</td>\n",
       "        <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_directory</pre></td>\n",
       "        <td>Path for the model cache</td>\n",
       "        <td>/Users/zhaiman/Library/Caches/cobrapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>max_cache_size</pre></td>\n",
       "        <td>Maximum cache size in bytes</td>\n",
       "        <td>104857600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td><pre>cache_expiration</pre></td>\n",
       "        <td>Model cache expiration time in seconds (if any)</td>\n",
       "        <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "\n",
       "solver: gurobi\n",
       "tolerance: 1e-07\n",
       "lower_bound: -1000.0\n",
       "upper_bound: 1000.0\n",
       "processes: 15\n",
       "cache_directory: /Users/zhaiman/Library/Caches/cobrapy\n",
       "max_cache_size: 104857600\n",
       "cache_expiration: None"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28d59af5-dfc2-4ef6-a9e2-b1911f3b0032",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 0.1.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1c91c721-5625-477c-bfc0-dee5c8642fbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iAB_RBC_283</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14a29c890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>342</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>469</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>346</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>41</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iAB_RBC_283 at 0x14a29c890>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = read_rbc_model('xml')\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79f0984b-4328-4625-b8c6-5cac3f291493",
   "metadata": {},
   "source": [
    "## Extract identifiers from HumanGEM\n",
    "\n",
    "Use HumanGEM (1.18.0) to get the MetabolicAtlas identifiers for reactions and metabolites.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ac74f778-d659-4f2a-8e84-8dd4daf9bbad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>HumanGEM</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14b339ee0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>8456</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>12995</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>2889</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>148</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*MAR13082 - 1.0*MAR13082_reverse_11d67</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>Cytosol, Extracellular, Lysosome, Endoplasmic reticulum, Mitochondria, Peroxisome, Golgi apparatus, Nucleus, Inner mitochondria</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model HumanGEM at 0x14b339ee0>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HumanGEM = read_cobra_model(f\"{ROOT_PATH}/data/raw/Human-GEM.xml\")\n",
    "HumanGEM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b50a5606-37f7-4439-9a60-9c51676db717",
   "metadata": {},
   "source": [
    "### Map model BiGG ID metabolites to MetabolicAtlas via HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "69eda8e4-baef-4d14-8c73-9b3413dd7dbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of metabolites in RBC-GEM (excluding compartments): 267\n",
      "Number of metabolites that could be found: 205\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bigg</th>\n",
       "      <th>metatlas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13dpg</td>\n",
       "      <td>MAM00247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>23dpg</td>\n",
       "      <td>MAM00569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2kmb</td>\n",
       "      <td>MAM01016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2pg</td>\n",
       "      <td>MAM00674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>35cgmp</td>\n",
       "      <td>MAM01433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200</th>\n",
       "      <td>xmp</td>\n",
       "      <td>MAM03150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>201</th>\n",
       "      <td>xu5p__D</td>\n",
       "      <td>MAM01761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>202</th>\n",
       "      <td>xylt</td>\n",
       "      <td>MAM03155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>xylu__D</td>\n",
       "      <td>MAM01759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>xylu__L</td>\n",
       "      <td>MAM02425</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>205 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        bigg  metatlas\n",
       "0      13dpg  MAM00247\n",
       "1      23dpg  MAM00569\n",
       "2       2kmb  MAM01016\n",
       "3        2pg  MAM00674\n",
       "4     35cgmp  MAM01433\n",
       "..       ...       ...\n",
       "200      xmp  MAM03150\n",
       "201  xu5p__D  MAM01761\n",
       "202     xylt  MAM03155\n",
       "203  xylu__D  MAM01759\n",
       "204  xylu__L  MAM02425\n",
       "\n",
       "[205 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_metabolite_ids = {metabolite.id.replace(f\"_{metabolite.compartment}\", \"\") for metabolite in model.metabolites}\n",
    "bigg_metabolites = set()\n",
    "human_gem_mapping = {}\n",
    "for metabolite in HumanGEM.metabolites:\n",
    "    bigg_ids = metabolite.annotation.get(\"bigg.metabolite\")\n",
    "    if not bigg_ids:\n",
    "        continue\n",
    "    if isinstance(bigg_ids, str):\n",
    "        bigg_ids = [bigg_ids]\n",
    "    bigg_ids = set(bigg_ids)\n",
    "    bigg_metabolites.update(bigg_ids)\n",
    "    for bigg_id in bigg_ids:\n",
    "        if bigg_id in human_gem_mapping:\n",
    "            human_gem_mapping[bigg_id] += [metabolite.id[:-1]]\n",
    "        else:\n",
    "            human_gem_mapping[bigg_id] = [metabolite.id[:-1]]\n",
    "\n",
    "\n",
    "print(f\"Number of metabolites in RBC-GEM (excluding compartments): {len(model_metabolite_ids)}\")\n",
    "\n",
    "intersection = sorted(bigg_metabolites.intersection(model_metabolite_ids))\n",
    "print(f\"Number of metabolites that could be found: {len(intersection)}\")\n",
    "print()\n",
    "\n",
    "id_mapping_dict = {}\n",
    "for rbc_metabolite in intersection:\n",
    "    metabolite_ids = set(human_gem_mapping.get(rbc_metabolite, []))\n",
    "    if len(metabolite_ids) > 1:\n",
    "        print(f'\"{rbc_metabolite}\": \"{build_string(metabolite_ids)}\",')\n",
    "    id_mapping_dict[rbc_metabolite] = build_string(metabolite_ids)\n",
    "\n",
    "manual_updates_and_corrections = {\n",
    "    # Selected reactions below needed corrections.\n",
    " \n",
    "}\n",
    "\n",
    "id_mapping_dict.update(manual_updates_and_corrections)\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"bigg\", \"metatlas\"]\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b95018-d02f-47ee-9d3c-bb8fecc1b6b0",
   "metadata": {},
   "source": [
    "### Map model BiGG ID reactions to MetabolicAtlas via HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d67ebee9-9d84-4001-af42-d5bc218105ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of reactions in RBC-GEM (excluding boundaries): 469\n",
      "Number of reactions that could be found: 276\n",
      "\n",
      "Check CAATPS for false mappings\n",
      "Check CAT for false mappings\n",
      "Check CHLPCTD for false mappings\n",
      "Check CHOLK for false mappings\n",
      "Check CHOLt4 for false mappings\n",
      "Check COt for false mappings\n",
      "Check CYStec for false mappings\n",
      "Check CYTK1 for false mappings\n",
      "Check EX_ac_e for false mappings\n",
      "Check EX_adrnl_e for false mappings\n",
      "Check EX_ala__L_e for false mappings\n",
      "Check EX_arg__L_e for false mappings\n",
      "Check EX_chol_e for false mappings\n",
      "Check EX_cl_e for false mappings\n",
      "Check EX_dopa_e for false mappings\n",
      "Check EX_fe2_e for false mappings\n",
      "Check EX_gal_e for false mappings\n",
      "Check EX_gam_e for false mappings\n",
      "Check EX_glc__D_e for false mappings\n",
      "Check EX_gln__L_e for false mappings\n",
      "Check EX_h2o_e for false mappings\n",
      "Check EX_h_e for false mappings\n",
      "Check EX_hco3_e for false mappings\n",
      "Check EX_k_e for false mappings\n",
      "Check EX_lnlc_e for false mappings\n",
      "Check EX_met__L_e for false mappings\n",
      "Check EX_nac_e for false mappings\n",
      "Check EX_nh4_e for false mappings\n",
      "Check EX_nrpphr_e for false mappings\n",
      "Check EX_ocdcea_e for false mappings\n",
      "Check EX_phe__L_e for false mappings\n",
      "Check EX_pyr_e for false mappings\n",
      "Check EX_ribflv_e for false mappings\n",
      "Check EX_thmmp_e for false mappings\n",
      "Check FBA for false mappings\n",
      "Check FBP26 for false mappings\n",
      "Check FUM for false mappings\n",
      "Check G6PDH2r for false mappings\n",
      "Check GALKr for false mappings\n",
      "Check GAMt1r for false mappings\n",
      "Check GAPD for false mappings\n",
      "Check GLNt4 for false mappings\n",
      "Check MEPIVESSte for false mappings\n",
      "Check NADK for false mappings\n",
      "Check NH4t3r for false mappings\n",
      "Check PPPGO for false mappings\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bigg</th>\n",
       "      <th>metatlas</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3MOXTYRESSte</td>\n",
       "      <td>MAR11306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4PYRDX</td>\n",
       "      <td>MAR08103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5AOPt2</td>\n",
       "      <td>MAR11307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACALDt</td>\n",
       "      <td>MAR04948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACGAM2E</td>\n",
       "      <td>MAR04527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>TMDPPK</td>\n",
       "      <td>MAR04206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>UNK3</td>\n",
       "      <td>MAR05391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>UPPDC1</td>\n",
       "      <td>MAR04750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>XYLK</td>\n",
       "      <td>MAR04595</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>XYLTD_D</td>\n",
       "      <td>MAR04593</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>322 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             bigg  metatlas\n",
       "0    3MOXTYRESSte  MAR11306\n",
       "1          4PYRDX  MAR08103\n",
       "2          5AOPt2  MAR11307\n",
       "3          ACALDt  MAR04948\n",
       "4         ACGAM2E  MAR04527\n",
       "..            ...       ...\n",
       "317        TMDPPK  MAR04206\n",
       "318          UNK3  MAR05391\n",
       "319        UPPDC1  MAR04750\n",
       "320          XYLK  MAR04595\n",
       "321       XYLTD_D  MAR04593\n",
       "\n",
       "[322 rows x 2 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_reaction_ids = {reaction.id for reaction in model.reactions}\n",
    "\n",
    "bigg_reactions = set()\n",
    "human_gem_mapping = {}\n",
    "for reaction in HumanGEM.reactions:\n",
    "    bigg_ids = reaction.annotation.get(\"bigg.reaction\")\n",
    "    if not bigg_ids:\n",
    "        continue\n",
    "    if isinstance(bigg_ids, str):\n",
    "        bigg_ids = [bigg_ids]\n",
    "    bigg_ids = set(bigg_ids)\n",
    "    bigg_reactions.update(bigg_ids)\n",
    "    for bigg_id in bigg_ids:\n",
    "        if bigg_id in human_gem_mapping:\n",
    "            human_gem_mapping[bigg_id] += [reaction.id]\n",
    "        else:\n",
    "            human_gem_mapping[bigg_id] = [reaction.id]\n",
    "\n",
    "\n",
    "print(f\"Number of reactions in RBC-GEM (excluding boundaries): {len(model_reaction_ids)}\")\n",
    "\n",
    "intersection = sorted(bigg_reactions.intersection(model_reaction_ids))\n",
    "print(f\"Number of reactions that could be found: {len(intersection)}\")\n",
    "print()\n",
    "id_mapping_dict = {}\n",
    "for rbc_reaction in intersection:\n",
    "    reaction_ids = human_gem_mapping.get(rbc_reaction, [])\n",
    "    if len(reaction_ids) > 1:\n",
    "        print(f\"Check {rbc_reaction} for false mappings\")\n",
    "    id_mapping_dict[rbc_reaction] = build_string(reaction_ids)\n",
    "\n",
    "manual_updates_and_corrections = {\n",
    "    # Selected reactions below needed corrections.\n",
    "    \"CAATPS\": \"MAR07629\",\n",
    "    \"CAT\": \"MAR03980\",\n",
    "    \"CHLPCTD\": \"MAR00638\",\n",
    "    \"CHOLK\": \"MAR00636\",\n",
    "    \"CHOLt4\": \"MAR07734\",\n",
    "    \"COt\": \"MAR07798\",\n",
    "    \"CYStec\": \"MAR05084\",\n",
    "    \"CYTK1\": \"MAR04024\",\n",
    "    \"EX_ac_e\": \"MAR09086\",\n",
    "    \"EX_adrnl_e\": \"MAR09095\",\n",
    "    \"EX_ala__L_e\": \"MAR09061\",\n",
    "    \"EX_arg__L_e\": \"MAR09066\",\n",
    "    \"EX_chol_e\": \"MAR09083\",\n",
    "    \"EX_cl_e\": \"MAR09150\",\n",
    "    \"EX_dopa_e\": \"MAR09092\",\n",
    "    \"EX_fe2_e\": \"MAR09076\",\n",
    "    \"EX_gal_e\": \"MAR09140\",\n",
    "    \"EX_gam_e\": \"MAR09168\",\n",
    "    \"EX_glc__D_e\": \"MAR09034\",\n",
    "    \"EX_gln__L_e\": \"MAR09063\",\n",
    "    \"EX_h2o_e\": \"MAR09047\",\n",
    "    \"EX_h_e\": \"MAR09079\",\n",
    "    \"EX_hco3_e\": \"MAR09078\",\n",
    "    \"EX_k_e\": \"MAR09081\",\n",
    "    \"EX_lnlc_e\": \"MAR09035\",\n",
    "    \"EX_met__L_e\": \"MAR09042\",\n",
    "    \"EX_nac_e\": \"MAR09142\",\n",
    "    \"EX_nh4_e\": \"MAR11420\",\n",
    "    \"EX_nrpphr_e\": \"MAR09093\",\n",
    "    \"EX_ocdcea_e\": \"MAR00650\",\n",
    "    \"EX_phe__L_e\": \"MAR09043\",\n",
    "    \"EX_pyr_e\": \"MAR09133\",\n",
    "    \"EX_ribflv_e\": \"MAR09143\",\n",
    "    \"EX_thmmp_e\": \"MAR09105\",\n",
    "    \"FBA\": \"MAR04375\",\n",
    "    \"FBP26\": \"MAR04706\",\n",
    "    \"FUM\": \"MAR04408\",\n",
    "    \"G6PDH2r\": \"MAR08971\",\n",
    "    \"GALKr\": \"MAR04130\",\n",
    "    \"GAMt1r\": \"MAR04996\",\n",
    "    \"GAPD\": \"MAR04373\",\n",
    "    \"GLNt4\": \"MAR05308\",\n",
    "    \"MEPIVESSte\": \"MAR08922\",\n",
    "    \"NADK\": \"MAR04269\",\n",
    "    \"NH4t3r\": \"MAR01534\",\n",
    "    \"PPPGO\": \"MAR11316\",\n",
    "    # Selected reactions needed manaul updates\n",
    "    \"ACP1_FMN\": \"MAR06507\",\n",
    "    \"ADRNLtu\": \"MAR09192\",\n",
    "    \"ARD\": \"MAR05389\",\n",
    "    \"ARGN\": \"MAR03816\",\n",
    "    \"BILIRBU\": \"MAR11321\",\n",
    "    \"C160CPT2rbc\": \"MAR02626\",\n",
    "    \"C181CPT2rbc\": \"MAR11310\",\n",
    "    \"CHLP\": \"MAR08424\",\n",
    "    \"DGULND\": \"MAR08353\",\n",
    "    \"DHAAt1r\": \"MAR08846\",\n",
    "    \"DOPAMT\": \"MAR06763\",\n",
    "    \"DPGM\": \"MAR04371\",\n",
    "    \"DPGase\": \"MAR04372\",\n",
    "    \"ENOPH\": \"MAR05387\",\n",
    "    \"ETHAt\": \"MAR07896\",\n",
    "    \"GALOR\": \"MAR08766\",\n",
    "    \"GALT\": \"MAR08767\",\n",
    "    \"GLCt1\": \"MAR05029\",\n",
    "    \"GMPR\": \"MAR04419\",\n",
    "    \"GPDDA1\": \"MAR00635\",\n",
    "    \"GULND\": \"MAR06537\",\n",
    "    \"HCO3_CLt\": \"MAR06525\",\n",
    "    \"LEUKTRA4t\": \"MAR06254\",\n",
    "    \"LEUKTRB4t\": \"MAR06255\",\n",
    "    \"LNLCCPT2rbc\": \"MAR02742\",\n",
    "    \"LTA4H\": \"MAR01080\",\n",
    "    \"MDRPD\": \"MAR05386\",\n",
    "    \"MI1345PP\": \"MAR06565\",\n",
    "    \"MI145PK\": \"MAR06563\",\n",
    "    \"MI145PP\": \"MAR06560\",\n",
    "    \"MTRI\": \"MAR05385\",\n",
    "    \"NADPN\": \"MAR07627\",\n",
    "    \"NMNHYD\": \"MAR04264\",\n",
    "    \"ORNDC\": \"MAR04212\",\n",
    "    \"RNMK\": \"MAR04265\",\n",
    "    \"SALMCOM\": \"MAR06750\",\n",
    "    \"SALMCOM2\": \"MAR06746\",\n",
    "    \"SBTD_D2\": \"MAR04315\",\n",
    "    \"SBTR\": \"MAR04316\",\n",
    "    \"SPMDtex2\": \"MAR04994\",\n",
    "    \"TDP\": \"MAR04208\",\n",
    "    \"THMTP\": \"MAR04207\",\n",
    "    \"TMDPK\": \"MAR04204\",\n",
    "    \"TMDPPK\": \"MAR04206\",\n",
    "    \"UNK3\": \"MAR05391\",\n",
    "    \"UPPDC1\": \"MAR04750\",\n",
    "    \"XYLK\": \"MAR04595\",\n",
    "    \"XYLTD_D\": \"MAR04593\",\n",
    "}\n",
    "\n",
    "id_mapping_dict.update(manual_updates_and_corrections)\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"bigg\", \"metatlas\"]\n",
    "# id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001284e3-012f-4e65-9592-16a9ab8fa2e5",
   "metadata": {},
   "source": [
    "## Update genes and gene reaction rules\n",
    "1. Update gene IDs to match RECON3D with inclusion of NCBI Gene ID.\n",
    "2. Manually update genes without NCBI numbers and genes lumped together.\n",
    "3. Remove identical genes from `DAGK` gene reaction rule\n",
    "4. Sort gene reaction components\n",
    "3. Update gene reaction rules accordingly for new genes and identifiers\n",
    "\n",
    "#### Generate gene ID mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a80353b5-0b80-43ac-84ea-3095e3c7780d",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dict = {}\n",
    "manual_updates = {}\n",
    "for gene in model.genes:\n",
    "    ncbigene = gene.annotation.get(\"ncbigene\")\n",
    "    if ncbigene is None:\n",
    "        # NCBI Gene ID is not annotated, will need manual updating\n",
    "        continue\n",
    "    gname, isonum = gene.id.split(\"_\")\n",
    "    id_mapping_dict[gene.id] = f\"{ncbigene}_{isonum}\"\n",
    "\n",
    "# Annotations mapped in Recon3D but not iAB-RBC-283\n",
    "id_mapping_dict[\"ThmtP_AT1\"] = \"79178_AT1\"\n",
    "id_mapping_dict[\"Cgi_14_AT1\"] = \"51005_AT1\"\n",
    "id_mapping_dict[\"Rhag_AT1\"] = \"6005_AT1\"\n",
    "id_mapping_dict[\"Rhbg_AT1\"] = \"57127_AT1\"\n",
    "id_mapping_dict[\"Abcc4_AT1\"] = \"10257_AT1\"\n",
    "id_mapping_dict[\"Slc2a1_AT1\"] = \"6513_AT1\"\n",
    "id_mapping_dict[\"Slc2a2_AT1\"] = \"6514_AT1\"\n",
    "id_mapping_dict[\"Slc2a3_AT1\"] = \"6515_AT1\"\n",
    "id_mapping_dict[\"Slc2a4_AT1\"] = \"6517_AT1\"\n",
    "id_mapping_dict[\"Slc2a5_AT1\"] = \"6518_AT1\"\n",
    "id_mapping_dict[\"Slc2a7_AT1\"] = \"155184_AT1\"\n",
    "id_mapping_dict[\"Slc2a8_AT1\"] = \"29988_AT1\"\n",
    "id_mapping_dict[\"Slc2a11_AT1\"] = \"66035_AT1\"\n",
    "id_mapping_dict[\"Slc4a1_AT1\"] = \"6521_AT1\"\n",
    "id_mapping_dict[\"Slc5a1_AT1\"] = \"6523_AT1\"\n",
    "id_mapping_dict[\"Slc5a3_AT1\"] = \"6526_AT1\"\n",
    "id_mapping_dict[\"Slc5a5_AT1\"] = \"6528_AT1\"\n",
    "id_mapping_dict[\"Slc12a7_AT1\"] = \"10723_AT1\"\n",
    "id_mapping_dict[\"Slc14a1_AT1\"] = \"6563_AT1\"\n",
    "id_mapping_dict[\"Slc29a1_AT1\"] = \"2030_AT1\"\n",
    "id_mapping_dict[\"Slc29a2_AT1\"] = \"3177_AT1\"\n",
    "id_mapping_dict[\"Flj22761_AT1\"] = \"80201_AT1\"\n",
    "\n",
    "# Pseudogene, does not seem to have annotations mapped \n",
    "id_mapping_dict[\"Gkp3_AT1\"] = \"2713_AT1\"\n",
    "\n",
    "# Rename genes using ID map, then add lumped genes to ID map and \n",
    "cobra.manipulation.modify.rename_genes(model, id_mapping_dict)\n",
    "\n",
    "# Two seperate genes lumped into one\n",
    "id_mapping_dict[\"Gucy1A2B2_AT1\"] = \"2977_AT1 and 2974_AT1\"\n",
    "id_mapping_dict[\"Gucy1A2B3_AT1\"] = \"2977_AT1 and 2983_AT1\"\n",
    "id_mapping_dict[\"Gucy1A3B3_AT1\"] = \"2982_AT1 and 2983_AT1\"\n",
    "id_mapping_dict[\"Gucy1A3B2_AT1\"] = \"2982_AT1 and 2974_AT1\"\n",
    "id_mapping_dict[\"NME12_AT1\"] = \"4830_AT1 and 4831_AT1\"\n",
    "id_mapping_dict[\"NME12_AT2\"] = \"4830_AT2 and 4831_AT1\"\n",
    "id_mapping_dict[\"Mat2ab1\"] = \"4144_AT1 and 27430_AT1\"\n",
    "id_mapping_dict[\"Mat2ab2\"] = \"4144_AT1 and 27430_AT2\"\n",
    "id_mapping_dict[\"Atp1a1b1\"] = \"476_AT1 and 481_AT1\"\n",
    "id_mapping_dict[\"Atp1a1b2\"] = \"476_AT1 and 482_AT1\"\n",
    "id_mapping_dict[\"Atp1a1b3\"] = \"476_AT1 and 483_AT1\"\n",
    "id_mapping_dict[\"Atp1a1b4\"] = \"476_AT1 and 23439_AT1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a7f4fc68-e1b3-4388-a682-d933e4c3ae3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>geneRetired</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>54981_AT1</td>\n",
       "      <td>Nrk1_AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6120_AT2</td>\n",
       "      <td>Rpe_AT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6120_AT1</td>\n",
       "      <td>Rpe_AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22934_AT1</td>\n",
       "      <td>Rpia_AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>118881_AT1</td>\n",
       "      <td>Comtd1_AT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343</th>\n",
       "      <td>482_AT1</td>\n",
       "      <td>Atp1a1b2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>476_AT1</td>\n",
       "      <td>Atp1a1b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>483_AT1</td>\n",
       "      <td>Atp1a1b3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>476_AT1</td>\n",
       "      <td>Atp1a1b4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>23439_AT1</td>\n",
       "      <td>Atp1a1b4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>358 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          genes geneRetired\n",
       "0     54981_AT1    Nrk1_AT1\n",
       "1      6120_AT2     Rpe_AT2\n",
       "2      6120_AT1     Rpe_AT1\n",
       "3     22934_AT1    Rpia_AT1\n",
       "4    118881_AT1  Comtd1_AT1\n",
       "..          ...         ...\n",
       "343     482_AT1    Atp1a1b2\n",
       "344     476_AT1    Atp1a1b3\n",
       "344     483_AT1    Atp1a1b3\n",
       "345     476_AT1    Atp1a1b4\n",
       "345   23439_AT1    Atp1a1b4\n",
       "\n",
       "[358 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"geneRetired\", \"genes\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "id_mapping_df[\"genes\"] = id_mapping_df[\"genes\"].str.split(\" and \")\n",
    "id_mapping_df = id_mapping_df.explode(\"genes\")\n",
    "\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/deprecatedIdentifiers/replacedGenes.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"genes\", \"geneRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[previous_id_mapping_df[\"genes\"] == retiring]\n",
    "    retired_set_of_ids = set([retiring])\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        retired_set_of_ids.update(previously_retired[\"geneRetired\"].apply(split_string).item())\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"geneRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/replacedGenes.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "175e6c8f-df31-416c-bd3a-b2a90ea26070",
   "metadata": {},
   "source": [
    "#### Seperate \"Lumped\" genes into individual genes\n",
    "These are reactions containing GPRs that have both \"AND\" and \"OR\" in them. \n",
    "The \"AND\" genes are lumped together, likely due to software compability with early COBRA versions that had difficulty parsing rules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e1e52615-9524-4511-bf11-63a6c8ccd20f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Reaction GUACYC: Recon3D top, iAB-RBC-283 bottom\n",
    "# 4881_AT1 or 4882_AT1 or 4882_AT2 or   2984_AT1 or   3000_AT1 or   2986_AT1 or (2977_AT1 and 2974_AT1) or (2977_AT1 and 2983_AT1) or (2982_AT1 and 2974_AT1) or (2982_AT1 and 2983_AT1)\n",
    "# Npr1_AT1 or Npr2_AT1 or Npr2_AT2 or Gucy2C_AT1 or Gucy2D_AT1 or Gucy2F_AT1 or (    Gucy1A2B2_AT1    ) or (    Gucy1A2B3_AT1    ) or (    Gucy1A3B2_AT1    ) or (    Gucy1A3B3_AT1    )\n",
    "reaction = model.reactions.get_by_id(\"GUACYC\")\n",
    "reaction.gene_reaction_rule = \"(2974_AT1 and 2977_AT1) or (2974_AT1 and 2982_AT1) or (2977_AT1 and 2983_AT1) or (2982_AT1 and 2983_AT1) or 2984_AT1 or 2986_AT1 or 3000_AT1 or 4881_AT1 or 4882_AT1 or 4882_AT2\"\n",
    "\n",
    "# (4830_AT1 and 4831_AT1) or (4830_AT2 and 4831_AT1) or 4832_AT1 or 10201_AT1 or 29922_AT1 or 29922_AT2\n",
    "# (      NME12_AT1      ) or (      NME12_AT2      ) or NME3_AT1 or  NME6_AT1 or  NME7_AT1 or  NME7_AT2\n",
    "for reaction in [\"NDPK1\", \"NDPK2\", \"NDPK3\"]:\n",
    "    reaction = model.reactions.get_by_id(reaction)\n",
    "    reaction.gene_reaction_rule = \"(4830_AT1 and 4831_AT1) or (4830_AT2 and 4831_AT1) or 4832_AT1 or 10201_AT1 or 29922_AT1 or 29922_AT2\"\n",
    "\n",
    "#  4143_AT1 or (4144_AT1 and 27430_AT1) or (4144_AT1 and 27430_AT2)\n",
    "# Mat1a_AT1 or (       Mat2ab1        ) or (       Mat2ab2        )\n",
    "reaction = model.reactions.get_by_id(\"METAT\")\n",
    "reaction.gene_reaction_rule = \"4143_AT1 or (4144_AT1 and 27430_AT1) or (4144_AT1 and 27430_AT2)\"\n",
    "\n",
    "# (476_AT1 and 481_AT1) or (476_AT1 and  482_AT1) or (476_AT1 and 483_AT1) or (476_AT1 and 23439_AT1)\n",
    "# (     Atp1a1b1      ) or (     Atp1a1b2       ) or (     Atp1a1b3      ) or (     Atp1a1b4      )\n",
    "reaction = model.reactions.get_by_id(\"NaKt\")\n",
    "reaction.gene_reaction_rule = \"(476_AT1 and 481_AT1) or (476_AT1 and 482_AT1) or (476_AT1 and 483_AT1) or (476_AT1 and 23439_AT1)\"\n",
    "\n",
    "# Remove old lumped genes from model\n",
    "cobra.manipulation.delete.remove_genes(\n",
    "    model,\n",
    "    gene_list=model.genes.query(lambda x: not x.reactions),\n",
    "    remove_reactions=False\n",
    ")\n",
    "model.genes.query(lambda x: not x.reactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3d4ce24-87e2-4501-ac99-717d12aa2b69",
   "metadata": {},
   "source": [
    "#### Fix gene reaction rules with duplicate genes\n",
    "The `DAGK` reactions have an extra gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ad7bb3a9-d0dc-48e8-82e2-90a83582543b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reaction: DAGK_hs_16_0_16_0: Counter({'1607_AT1': 1})\n",
      "Reaction: DAGK_hs_16_0_18_1: Counter({'1607_AT1': 1})\n",
      "Reaction: DAGK_hs_16_0_18_2: Counter({'1607_AT1': 1})\n",
      "Reaction: DAGK_hs_18_1_18_1: Counter({'1607_AT1': 1})\n",
      "Reaction: DAGK_hs_18_1_18_2: Counter({'1607_AT1': 1})\n",
      "Reaction: DAGK_hs_18_2_16_0: Counter({'1607_AT1': 1})\n",
      "Reaction: DAGK_hs_18_2_18_1: Counter({'1607_AT1': 1})\n"
     ]
    }
   ],
   "source": [
    "# Auto fixed\n",
    "for reaction in model.reactions.query(lambda x: x.gene_reaction_rule):\n",
    "    or_split = reaction.gene_reaction_rule.split(\" or \")\n",
    "    if Counter(or_split) != Counter(set(or_split)):\n",
    "        # Identify extra gene\n",
    "        extra = Counter(or_split) - Counter(set(or_split))\n",
    "        if extra:\n",
    "            print(f\"Reaction: {reaction.id}: {extra}\")\n",
    "        for gene, num_extra in extra.items():\n",
    "            reaction.gene_reaction_rule = reaction.gene_reaction_rule.replace(f\"{gene} or \", \"\", num_extra)\n",
    "\n",
    "# # DAGK reactions have an extra gene\n",
    "# # Manual fix\n",
    "# model.reactions.get_by_id(\"DAGK_hs_16_0_16_0\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_16_0_18_1\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_16_0_18_2\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_1_18_1\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_1_18_2\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_2_16_0\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_2_18_1\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7752d2fe-94ea-419a-9643-0d18f86dece4",
   "metadata": {},
   "source": [
    "#### Sort GPRs in reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "46689b14-31c1-4580-9b41-838b92fdeb8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_gpr_reactions = [\n",
    "    r for r in model.reactions \n",
    "    if \"or\" in r.gene_reaction_rule and not \"and\" in r.gene_reaction_rule]\n",
    "\n",
    "and_gpr_reactions = [\n",
    "    r for r in model.reactions \n",
    "    if \"and\" in r.gene_reaction_rule and not \"or\" in r.gene_reaction_rule]\n",
    "\n",
    "for join_str, reactions in zip([\" or \", \" and \"], [or_gpr_reactions, and_gpr_reactions]):\n",
    "    for r in reactions:\n",
    "        gpr_dict = defaultdict(list)\n",
    "        for i in r.gene_reaction_rule.split(join_str):\n",
    "            gpr_dict[int(i.split(\"_\")[0])] += [i.split(\"_\")[1]]\n",
    "        genes = r.genes\n",
    "        r.gene_reaction_rule = join_str.join([f\"{key}_{v}\" for key in sorted(gpr_dict) for v in sorted(gpr_dict[key])])\n",
    "        try:\n",
    "            assert genes == r.genes\n",
    "        except AssertionError:\n",
    "            raise AssertionError(f\"A change occured in the genes for reaction {r.id} that should not have occured.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eff1b45f-ba45-4025-84e8-52fc96cfab13",
   "metadata": {},
   "source": [
    "## Update metabolites\n",
    "1. Update metabolites to have the corresponding MetabolicAtlas annotation\n",
    "2. Update BiGG identifiers and retire old ones\n",
    "\n",
    "### Update metabolite annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5a15f709-c07e-4a23-9e49-6174da4cfc43",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "for _, row in id_mapping_df.iterrows():\n",
    "    bigg_id, metatlas = row[[\"bigg\", \"metatlas\"]]\n",
    "    for model_metabolite in model.metabolites.query(lambda x: x.id.replace(f\"_{x.compartment}\", \"\") == bigg_id):\n",
    "        model_metabolite.annotation[\"metatlas\"] = f\"{metatlas}{model_metabolite.compartment}\"\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accb6c13-a4c6-447c-b110-95041bc49cee",
   "metadata": {},
   "source": [
    "### Update metabolite identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d1139c66-a3b3-4ad7-a5d5-5987334b9482",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dict = {\n",
    "    # Add L/D stereochemistry to ID\n",
    "    \"adrnl\": \"adrnl__L\",\n",
    "    'ahcys': 'ahcys__L',\n",
    "    'amet': 'amet__L',\n",
    "    'dhdascb': 'dhdascb__L',\n",
    "    \"mepi\": \"mepi__L\",\n",
    "    'nrpphr': 'nrpphr__L',\n",
    "    'orn': 'orn__L',\n",
    "\n",
    "     # ID change to help distinguish L-dopa (3,4-dihydroxyphenylalanine) and L-dopamine, D-Dopa and D-dopamine\n",
    "    'dopa': 'dpam__L',\n",
    "     # ID change to help distinguish the 'gamma' in gamma-L-glutamyl-L-cysteine\n",
    "    'glucys': 'gglucys__L',\n",
    "    # ID change to make it clear that the species is (R)-S-lactoylglutathione, with (R) stereoconfirmation and S-bond\n",
    "    'lgt__S': 'slgth__R',\n",
    "    # ID change to make it clear that the species is UDP bound to a D-glucose\n",
    "    'udpg': 'udpglc__D',\n",
    "    # Increased name specificity for cyclic nucleotides\n",
    "    'camp': '35camp',\n",
    "    \n",
    "    # Lipid ID changes to help distinguish stereochemistry on species and without use of common names.\n",
    "    # Fatty Acids and Conjugates [FA01]\t(FA)\n",
    "    'hdca': 'FA_hs_16_0',\n",
    "    'ocdcea': 'FA_hs_18_9Z',\n",
    "    'lnlc': 'FA_hs_18_9Z12Z',\n",
    "\n",
    "    # Fatty acyl CoAs [FA0705] (CoA)\n",
    "    'pmtcoa': 'FAcoa_hs_16_0',\n",
    "    'odecoa': 'FAcoa_hs_18_9Z',\n",
    "    'lnlccoa': 'FAcoa_hs_18_9Z12Z',\n",
    "\n",
    "    # Fatty acyl carnitines [FA0707] (CAR)\n",
    "    'pmtcrn': 'FAcrn_hs_16_0',\n",
    "    'odecrn': 'FAcrn_hs_18_9Z',\n",
    "    'lnlccrn': 'FAcrn_hs_18_9Z12Z',\n",
    "\n",
    "    # Diacylglycerols [GL0201] (DG)\n",
    "    'dag_hs_16_0_18_1': 'dag_hs_16_0_18_9Z',\n",
    "    'dag_hs_16_0_18_2': 'dag_hs_16_0_18_9Z12Z',\n",
    "    'dag_hs_18_1_18_1': 'dag_hs_18_9Z_18_9Z',\n",
    "    'dag_hs_18_1_18_2': 'dag_hs_18_9Z_18_9Z12Z',\n",
    "    'dag_hs_18_2_16_0': 'dag_hs_18_9Z12Z_16_0',\n",
    "    'dag_hs_18_2_18_1': 'dag_hs_18_9Z12Z_18_9Z',\n",
    "\n",
    "    # Diacylglycerophosphates [GP1001] (PA)\n",
    "    'pa_hs_16_0_18_1': 'pa_hs_16_0_18_9Z',\n",
    "    'pa_hs_16_0_18_2': 'pa_hs_16_0_18_9Z12Z',\n",
    "    'pa_hs_18_1_18_1': 'pa_hs_18_9Z_18_9Z',\n",
    "    'pa_hs_18_1_18_2': 'pa_hs_18_9Z_18_9Z12Z',\n",
    "    'pa_hs_18_2_16_0': 'pa_hs_18_9Z12Z_16_0',\n",
    "    'pa_hs_18_2_18_1': 'pa_hs_18_9Z12Z_18_9Z',\n",
    "    # Lysophospholipids\tPrefix L\n",
    "    'alpa_hs_16_0': 'lpa_hs_16_0',\n",
    "    'alpa_hs_18_1': 'lpa_hs_18_9Z',\n",
    "    'alpa_hs_18_2': 'lpa_hs_18_9Z12Z',\n",
    "\n",
    "    # Diacylglycerophosphocholines [GP0101] PC\n",
    "    'pchol_hs_16_0_16_0': 'pc_hs_16_0_16_0',\n",
    "    'pchol_hs_16_0_18_1': 'pc_hs_16_0_18_9Z',\n",
    "    'pchol_hs_16_0_18_2': 'pc_hs_16_0_18_9Z12Z',\n",
    "    'pchol_hs_18_1_18_1': 'pc_hs_18_9Z_18_9Z',\n",
    "    'pchol_hs_18_1_18_2': 'pc_hs_18_9Z_18_9Z12Z',\n",
    "    'pchol_hs_18_2_16_0': 'pc_hs_18_9Z12Z_16_0',\n",
    "    'pchol_hs_18_2_18_1': 'pc_hs_18_9Z12Z_18_9Z',\n",
    "    # Lysophospholipids\tPrefix L\n",
    "    'lpchol_hs_16_0': 'lpc_hs_16_0',\n",
    "    'lpchol_hs_18_1': 'lpc_hs_18_9Z',\n",
    "    'lpchol_hs_18_2': 'lpc_hs_18_9Z12Z',\n",
    "\n",
    "    # Diacylglycerophosphoethanolamines [GP0201] PE\n",
    "    'pe_hs_16_0_18_1': 'pe_hs_16_0_18_9Z',\n",
    "    'pe_hs_16_0_18_2': 'pe_hs_16_0_18_9Z12Z',\n",
    "    'pe_hs_18_1_18_1': 'pe_hs_18_9Z_18_9Z',\n",
    "    'pe_hs_18_1_18_2': 'pe_hs_18_9Z_18_9Z12Z',\n",
    "    'pe_hs_18_2_16_0': 'pe_hs_18_9Z12Z_16_0',\n",
    "    'pe_hs_18_2_18_1': 'pe_hs_18_9Z12Z_18_9Z',\n",
    "\n",
    "    # Diacylglycerophosphoinositols [GP0601] PI\n",
    "    'pail_hs_16_0_18_1': 'pail_hs_16_0_18_9Z',\n",
    "    'pail_hs_16_0_18_2': 'pail_hs_16_0_18_9Z12Z',\n",
    "    'pail_hs_18_1_18_1': 'pail_hs_18_9Z_18_9Z',\n",
    "    'pail_hs_18_1_18_2': 'pail_hs_18_9Z_18_9Z12Z',\n",
    "    'pail_hs_18_2_16_0': 'pail_hs_18_9Z12Z_16_0',\n",
    "    'pail_hs_18_2_18_1': 'pail_hs_18_9Z12Z_18_9Z',\n",
    "    # Diacylglycerophosphoinositol monophosphates [GP0701] PIP[4’]\n",
    "    'pail4p_hs_16_0_18_1': 'pail4p_hs_16_0_18_9Z',\n",
    "    'pail4p_hs_16_0_18_2': 'pail4p_hs_16_0_18_9Z12Z',\n",
    "    'pail4p_hs_18_1_18_1': 'pail4p_hs_18_9Z_18_9Z',\n",
    "    'pail4p_hs_18_1_18_2': 'pail4p_hs_18_9Z_18_9Z12Z',\n",
    "    'pail4p_hs_18_2_16_0': 'pail4p_hs_18_9Z12Z_16_0',\n",
    "    'pail4p_hs_18_2_18_1': 'pail4p_hs_18_9Z12Z_18_9Z',\n",
    "    # Diacylglycerophosphoinositol bisphosphates [GP0801] PIP2[4’,5’]\n",
    "    'pail45p_hs_16_0_18_1': 'pail45p_hs_16_0_18_9Z',\n",
    "    'pail45p_hs_16_0_18_2': 'pail45p_hs_16_0_18_9Z12Z',\n",
    "    'pail45p_hs_18_1_18_1': 'pail45p_hs_18_9Z_18_9Z',\n",
    "    'pail45p_hs_18_1_18_2': 'pail45p_hs_18_9Z_18_9Z12Z',\n",
    "    'pail45p_hs_18_2_16_0': 'pail45p_hs_18_9Z12Z_16_0',\n",
    "    'pail45p_hs_18_2_18_1': 'pail45p_hs_18_9Z12Z_18_9Z',\n",
    "\n",
    "    # CDP-diacylglycerols [GP1301] CDP-DG\n",
    "    'cdpdag_hs_16_0_18_1': 'cdpdag_hs_16_0_18_9Z',\n",
    "    'cdpdag_hs_16_0_18_2': 'cdpdag_hs_16_0_18_9Z12Z',\n",
    "    'cdpdag_hs_18_1_18_1': 'cdpdag_hs_18_9Z_18_9Z',\n",
    "    'cdpdag_hs_18_1_18_2': 'cdpdag_hs_18_9Z_18_9Z12Z',\n",
    "    'cdpdag_hs_18_2_16_0': 'cdpdag_hs_18_9Z12Z_16_0',\n",
    "    'cdpdag_hs_18_2_18_1': 'cdpdag_hs_18_9Z12Z_18_9Z',\n",
    "\n",
    "}\n",
    "for old_met, new_met in id_mapping_dict.items():\n",
    "    for model_metabolite in model.metabolites.query(lambda x: x.id.replace(f\"_{x.compartment}\", \"\") == old_met):\n",
    "        model_metabolite.id = f\"{new_met}_{model_metabolite.compartment}\"\n",
    "model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0dd99f29-d8e1-46fb-9037-05f01e23cbbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mets</th>\n",
       "      <th>metRetired</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adrnl__L</td>\n",
       "      <td>adrnl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ahcys__L</td>\n",
       "      <td>ahcys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>amet__L</td>\n",
       "      <td>amet</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dhdascb__L</td>\n",
       "      <td>dhdascb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mepi__L</td>\n",
       "      <td>mepi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>cdpdag_hs_16_0_18_9Z12Z</td>\n",
       "      <td>cdpdag_hs_16_0_18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>cdpdag_hs_18_9Z_18_9Z</td>\n",
       "      <td>cdpdag_hs_18_1_18_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>cdpdag_hs_18_9Z_18_9Z12Z</td>\n",
       "      <td>cdpdag_hs_18_1_18_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>cdpdag_hs_18_9Z12Z_16_0</td>\n",
       "      <td>cdpdag_hs_18_2_16_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>cdpdag_hs_18_9Z12Z_18_9Z</td>\n",
       "      <td>cdpdag_hs_18_2_18_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        mets           metRetired\n",
       "0                   adrnl__L                adrnl\n",
       "1                   ahcys__L                ahcys\n",
       "2                    amet__L                 amet\n",
       "3                 dhdascb__L              dhdascb\n",
       "4                    mepi__L                 mepi\n",
       "..                       ...                  ...\n",
       "71   cdpdag_hs_16_0_18_9Z12Z  cdpdag_hs_16_0_18_2\n",
       "72     cdpdag_hs_18_9Z_18_9Z  cdpdag_hs_18_1_18_1\n",
       "73  cdpdag_hs_18_9Z_18_9Z12Z  cdpdag_hs_18_1_18_2\n",
       "74   cdpdag_hs_18_9Z12Z_16_0  cdpdag_hs_18_2_16_0\n",
       "75  cdpdag_hs_18_9Z12Z_18_9Z  cdpdag_hs_18_2_18_1\n",
       "\n",
       "[76 rows x 2 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"metRetired\", \"mets\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"mets\", \"metRetired\"]]\n",
    "    retired_set_of_ids = set([retiring])\n",
    "    id_mapping_df.loc[idx, \"metRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/replacedMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c281ba7-924b-4a16-bfff-c88d3cdf7251",
   "metadata": {},
   "source": [
    "## Update reactions\n",
    "1. Update reactions to have the corresponding MetabolicAtlas annotation\n",
    "2. Change the NADH demand reaction annoation to prevent false identification as a boundary reaction.\n",
    "3. Update BiGG identifiers and retire old ones\n",
    "### Update reaction annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5c218e64-cdbc-43a4-b353-3a479ce07371",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "for _, row in id_mapping_df.iterrows():\n",
    "    bigg_id, metatlas = row[[\"bigg\", \"metatlas\"]]\n",
    "    reaction = model.reactions.get_by_id(bigg_id)\n",
    "    reaction.annotation[\"metatlas\"] = metatlas\n",
    "\n",
    "reaction = model.reactions.get_by_id(\"DM_nadh\")\n",
    "reaction.annotation[\"sbo\"] = \"SBO:0000176\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6666c72-5623-4e14-86d2-cd3492dfac72",
   "metadata": {},
   "source": [
    "### Update reaction identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c6af1928-a0dd-4b9e-b646-7ca43c95949d",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dict = {\n",
    "\n",
    "    # Change NADH demand to NADHload\n",
    "    'DM_nadh': 'NADHload',\n",
    "    # Updated to reflect stoichiometry of 2:1\n",
    "    \"CAATPS\": \"CAATPS2\",\n",
    "    # Addition of `x` or `y` to indiciate NADH and NADPH cofactors\n",
    "    'BILIRED': 'BILIREDy',\n",
    "    'GTHOr': 'GTHOy',\n",
    "    \"XYLTD_D\": \"XYLTD_Dx\",\n",
    "    \"XYLUR\": \"XYLUR_Ly\",\n",
    "\n",
    "    # Updated for increased specificity\n",
    "    'ACP1_FMN': 'FMNPH',\n",
    "    'DOPAMT': 'LDPAMMT',\n",
    "    'GGLUCT': 'GGLUCTC',\n",
    "    'MI1345PP': 'MI1345P5PP',\n",
    "    'MI145PK': 'MI145P3K',\n",
    "    'MI145PP': 'MI145P5PP',\n",
    "    'NP1': 'PNP2',\n",
    "    'NT5C': 'NICHYD',\n",
    "    'PDE1': 'PDEA',\n",
    "    'UMPK': 'UMPK1',\n",
    "    'UPP3S': 'UPPG3S',\n",
    "    'UPPDC1': 'UPPG3DC',\n",
    "    \n",
    "    '4PYRDX': '4PYRDXABCte',\n",
    "    'CAMPt': 'CAMPABCte',\n",
    "    'CGMPt': 'CGMPABCte',\n",
    "    'GTHOXti2': 'GTHOXABCte',\n",
    "    'RIBFLVt3o': 'RIBFLVABCte',\n",
    "    'HCO3_CLt': 'HCO3_Cltex',\n",
    "    \n",
    "    'ADRNLtu': 'ADRNLt',\n",
    "    'ACt2r': 'ACt2',\n",
    "    'GLYt7_211_r': 'GLY_Cl_2Nat',\n",
    "    'L_LACt2r': 'L_LACt2',\n",
    "    'NCAMUP': 'NCAMt',\n",
    "    'NRPPHRtu': 'NRPPHRt',\n",
    "    'DOPAtu': 'LDPAMt',\n",
    "\n",
    "    # Removal of incorrect/unclear suffixes\n",
    "    'AHCi': 'AHC',\n",
    "    'G6PDH2r': 'G6PDH2',\n",
    "    'GALKr': 'GALK',\n",
    "    'GALUi': 'GALU',\n",
    "    'GTHPi': 'GTHP',\n",
    "    'ICDHyr': 'ICDHy',\n",
    "    'NNATr': 'NNAT',\n",
    "    'PDX5POi': 'PDX5PO',\n",
    "\n",
    "    'ARGt5r': 'ARGtec',\n",
    "    'SPRMt2i': 'SPRMt2',\n",
    "    'THMtrbc': 'THMt',\n",
    "\n",
    "    \n",
    "    # Changed for preparation of additional lipids\n",
    "    'C160CPT1': 'CRNAT_16_0',\n",
    "    'C160CPT2rbc': 'CRNAT_16_0rbc',\n",
    "    'C181CPT1': 'CRNAT_18_9Z',\n",
    "    'C181CPT2rbc': 'CRNAT_18_9Zrbc',\n",
    "    'LNLCCPT1': 'CRNAT_18_9Z12Z',\n",
    "    'LNLCCPT2rbc': 'CRNAT_18_9Z12Zrbc',\n",
    "    \n",
    "    'FACOAL160': 'FACOAL_16_0',\n",
    "    'FACOAL181': 'FACOAL_18_9Z',\n",
    "    'FACOAL1821': 'FACOAL_18_9Z12Z',\n",
    "    \n",
    "    'HDCAt': 'FAt_16_0',\n",
    "    'OCDCEAt': 'FAt_18_9Z',\n",
    "    'LNLCt': 'FAt_18_9Z12Z',\n",
    "\n",
    "\n",
    "    # Lysophospatidate (LPA) acyl transferase\n",
    "    'AGPAT1_16_0_16_0': 'LPAAT_16_0_16_0',\n",
    "    'AGPAT1_16_0_18_1': 'LPAAT_16_0_18_9Z',\n",
    "    'AGPAT1_16_0_18_2': 'LPAAT_16_0_18_9Z12Z',\n",
    "    'AGPAT1_18_1_18_1': 'LPAAT_18_9Z_18_9Z',\n",
    "    'AGPAT1_18_1_18_2': 'LPAAT_18_9Z_18_9Z12Z',\n",
    "    'AGPAT1_18_2_16_0': 'LPAAT_18_9Z12Z_16_0',\n",
    "    'AGPAT1_18_2_18_1': 'LPAAT_18_9Z12Z_18_9Z',\n",
    "\n",
    "    # Removal of \"r\"\n",
    "    'CDIPTr_16_0_16_0': 'CDIPT_16_0_16_0',\n",
    "    'CDIPTr_16_0_18_1': 'CDIPT_16_0_18_9Z',\n",
    "    'CDIPTr_16_0_18_2': 'CDIPT_16_0_18_9Z12Z',\n",
    "    'CDIPTr_18_1_18_1': 'CDIPT_18_9Z_18_9Z',\n",
    "    'CDIPTr_18_1_18_2': 'CDIPT_18_9Z_18_9Z12Z',\n",
    "    'CDIPTr_18_2_16_0': 'CDIPT_18_9Z12Z_16_0',\n",
    "    'CDIPTr_18_2_18_1': 'CDIPT_18_9Z12Z_18_9Z',\n",
    "\n",
    "    'CDS_16_0_18_1': 'CDS_16_0_18_9Z',\n",
    "    'CDS_16_0_18_2': 'CDS_16_0_18_9Z12Z',\n",
    "    'CDS_18_1_18_1': 'CDS_18_9Z_18_9Z',\n",
    "    'CDS_18_1_18_2': 'CDS_18_9Z_18_9Z12Z',\n",
    "    'CDS_18_2_16_0': 'CDS_18_9Z12Z_16_0',\n",
    "    'CDS_18_2_18_1': 'CDS_18_9Z12Z_18_9Z',\n",
    "\n",
    "    'CEPTC_16_0_18_1': 'CEPTC_16_0_18_9Z',\n",
    "    'CEPTC_16_0_18_2': 'CEPTC_16_0_18_9Z12Z',\n",
    "    'CEPTC_18_1_18_1': 'CEPTC_18_9Z_18_9Z',\n",
    "    'CEPTC_18_1_18_2': 'CEPTC_18_9Z_18_9Z12Z',\n",
    "    'CEPTC_18_2_16_0': 'CEPTC_18_9Z12Z_16_0',\n",
    "    'CEPTC_18_2_18_1': 'CEPTC_18_9Z12Z_18_9Z',\n",
    "    \n",
    "    'CEPTE_16_0_18_1': 'CEPTE_16_0_18_9Z',\n",
    "    'CEPTE_16_0_18_2': 'CEPTE_16_0_18_9Z12Z',\n",
    "    'CEPTE_18_1_18_1': 'CEPTE_18_9Z_18_9Z',\n",
    "    'CEPTE_18_1_18_2': 'CEPTE_18_9Z_18_9Z12Z',\n",
    "    'CEPTE_18_2_16_0': 'CEPTE_18_9Z12Z_16_0',\n",
    "    'CEPTE_18_2_18_1': 'CEPTE_18_9Z12Z_18_9Z',\n",
    "\n",
    "    'DAGK_hs_16_0_16_0': 'DAGK_16_0_16_0',\n",
    "    'DAGK_hs_16_0_18_1': 'DAGK_16_0_18_9Z',\n",
    "    'DAGK_hs_16_0_18_2': 'DAGK_16_0_18_9Z12Z',\n",
    "    'DAGK_hs_18_1_18_1': 'DAGK_18_9Z_18_9Z',\n",
    "    'DAGK_hs_18_1_18_2': 'DAGK_18_9Z_18_9Z12Z',\n",
    "    'DAGK_hs_18_2_16_0': 'DAGK_18_9Z12Z_16_0',\n",
    "    'DAGK_hs_18_2_18_1': 'DAGK_18_9Z12Z_18_9Z',\n",
    "\n",
    "    # Glycerol-3-phosphate acyltransferase (GPAT)\n",
    "    'GPAM_hs_16_0': 'GPAT_16_0',\n",
    "    'GPAM_hs_18_1': 'GPAT_18_9Z',\n",
    "    'GPAM_hs_18_2': 'GPAT_18_9Z12Z',\n",
    "\n",
    "    # Phosphpolipase B acting on the LPC \n",
    "    'LPASE_16_0': 'LPCLPLB_16_0',\n",
    "    'LPASE_18_1': 'LPCLPLB_18_9Z',\n",
    "    'LPASE_18_2': 'LPCLPLB_18_9Z12Z',\n",
    "\n",
    "    'PI45P5P_16_0_18_1': 'PI45P5P_16_0_18_9Z',\n",
    "    'PI45P5P_16_0_18_2': 'PI45P5P_16_0_18_9Z12Z',\n",
    "    'PI45P5P_18_1_18_1': 'PI45P5P_18_9Z_18_9Z',\n",
    "    'PI45P5P_18_1_18_2': 'PI45P5P_18_9Z_18_9Z12Z',\n",
    "    'PI45P5P_18_2_16_0': 'PI45P5P_18_9Z12Z_16_0',\n",
    "    'PI45P5P_18_2_18_1': 'PI45P5P_18_9Z12Z_18_9Z',\n",
    "    \n",
    "    'PI45PLC_16_0_18_1': 'PI45PLC_16_0_18_9Z',\n",
    "    'PI45PLC_16_0_18_2': 'PI45PLC_16_0_18_9Z12Z',\n",
    "    'PI45PLC_18_1_18_1': 'PI45PLC_18_9Z_18_9Z',\n",
    "    'PI45PLC_18_1_18_2': 'PI45PLC_18_9Z_18_9Z12Z',\n",
    "    'PI45PLC_18_2_16_0': 'PI45PLC_18_9Z12Z_16_0',\n",
    "    'PI45PLC_18_2_18_1': 'PI45PLC_18_9Z12Z_18_9Z',\n",
    "    \n",
    "    'PI4P5K_16_0_18_1': 'PI4P5K_16_0_18_9Z',\n",
    "    'PI4P5K_16_0_18_2': 'PI4P5K_16_0_18_9Z12Z',\n",
    "    'PI4P5K_18_1_18_1': 'PI4P5K_18_9Z_18_9Z',\n",
    "    'PI4P5K_18_1_18_2': 'PI4P5K_18_9Z_18_9Z12Z',\n",
    "    'PI4P5K_18_2_16_0': 'PI4P5K_18_9Z12Z_16_0',\n",
    "    'PI4P5K_18_2_18_1': 'PI4P5K_18_9Z12Z_18_9Z',\n",
    "    \n",
    "    'PI4PLC_16_0_18_1': 'PI4PLC_16_0_18_9Z',\n",
    "    'PI4PLC_16_0_18_2': 'PI4PLC_16_0_18_9Z12Z',\n",
    "    'PI4PLC_18_1_18_1': 'PI4PLC_18_9Z_18_9Z',\n",
    "    'PI4PLC_18_1_18_2': 'PI4PLC_18_9Z_18_9Z12Z',\n",
    "    'PI4PLC_18_2_16_0': 'PI4PLC_18_9Z12Z_16_0',\n",
    "    'PI4PLC_18_2_18_1': 'PI4PLC_18_9Z12Z_18_9Z',\n",
    "    \n",
    "    'PI4PP_16_0_18_1': 'PI4PP_16_0_18_9Z',\n",
    "    'PI4PP_16_0_18_2': 'PI4PP_16_0_18_9Z12Z',\n",
    "    'PI4PP_18_1_18_1': 'PI4PP_18_9Z_18_9Z',\n",
    "    'PI4PP_18_1_18_2': 'PI4PP_18_9Z_18_9Z12Z',\n",
    "    'PI4PP_18_2_16_0': 'PI4PP_18_9Z12Z_16_0',\n",
    "    'PI4PP_18_2_18_1': 'PI4PP_18_9Z12Z_18_9Z',\n",
    "\n",
    "    # PI 4-kinase\n",
    "    'PIK4_16_0_16_0': 'PI4K_16_0_16_0',\n",
    "    'PIK4_16_0_18_1': 'PI4K_16_0_18_9Z',\n",
    "    'PIK4_16_0_18_2': 'PI4K_16_0_18_9Z12Z',\n",
    "    'PIK4_18_1_18_1': 'PI4K_18_9Z_18_9Z',\n",
    "    'PIK4_18_1_18_2': 'PI4K_18_9Z_18_9Z12Z',\n",
    "    'PIK4_18_2_16_0': 'PI4K_18_9Z12Z_16_0',\n",
    "    'PIK4_18_2_18_1': 'PI4K_18_9Z12Z_18_9Z',\n",
    "    \n",
    "    'PIPLC_16_0_18_1': 'PIPLC_16_0_18_9Z',\n",
    "    'PIPLC_16_0_18_2': 'PIPLC_16_0_18_9Z12Z',\n",
    "    'PIPLC_18_1_18_1': 'PIPLC_18_9Z_18_9Z',\n",
    "    'PIPLC_18_1_18_2': 'PIPLC_18_9Z_18_9Z12Z',\n",
    "    'PIPLC_18_2_16_0': 'PIPLC_18_9Z12Z_16_0',\n",
    "    'PIPLC_18_2_18_1': 'PIPLC_18_9Z12Z_18_9Z',\n",
    "    # PC Phospholipase A2\n",
    "    'PLA2_2_16_0_16_0': 'PCPLA2_16_0_16_0',\n",
    "    'PLA2_2_16_0_18_1': 'PCPLA2_16_0_18_9Z',\n",
    "    'PLA2_2_16_0_18_2': 'PCPLA2_16_0_18_9Z12Z',\n",
    "    'PLA2_2_18_1_18_1': 'PCPLA2_18_9Z_18_9Z',\n",
    "    'PLA2_2_18_1_18_2': 'PCPLA2_18_9Z_18_9Z12Z',\n",
    "    'PLA2_2_18_2_16_0': 'PCPLA2_18_9Z12Z_16_0',\n",
    "    'PLA2_2_18_2_18_1': 'PCPLA2_18_9Z12Z_18_9Z',\n",
    "    # Phosphatidate (PA) phosphatase \n",
    "    'PPAP_16_0_16_0': 'PAPP_16_0_16_0',\n",
    "    'PPAP_16_0_18_1': 'PAPP_16_0_18_9Z',\n",
    "    'PPAP_16_0_18_2': 'PAPP_16_0_18_9Z12Z',\n",
    "    'PPAP_18_1_18_1': 'PAPP_18_9Z_18_9Z',\n",
    "    'PPAP_18_1_18_2': 'PAPP_18_9Z_18_9Z12Z',\n",
    "    'PPAP_18_2_16_0': 'PAPP_18_9Z12Z_16_0',\n",
    "    'PPAP_18_2_18_1': 'PAPP_18_9Z12Z_18_9Z',\n",
    "    # To match metabolite changes\n",
    "    'SK_pchol_hs_16_0_16_0_c': 'SK_pc_hs_16_0_16_0_c',\n",
    "    'SK_pchol_hs_16_0_18_1_c': 'SK_pc_hs_16_0_18_9Z_c',\n",
    "    'SK_pchol_hs_16_0_18_2_c': 'SK_pc_hs_16_0_18_9Z12Z_c',\n",
    "    'SK_pchol_hs_18_1_18_1_c': 'SK_pc_hs_18_9Z_18_9Z_c',\n",
    "    'SK_pchol_hs_18_1_18_2_c': 'SK_pc_hs_18_9Z_18_9Z12Z_c',\n",
    "    'SK_pchol_hs_18_2_16_0_c': 'SK_pc_hs_18_9Z12Z_16_0_c',\n",
    "    'SK_pchol_hs_18_2_18_1_c': 'SK_pc_hs_18_9Z12Z_18_9Z_c',\n",
    "    'SK_pe_hs_16_0_16_0_c': 'SK_pe_hs_16_0_16_0_c',\n",
    "    'SK_pe_hs_16_0_18_1_c': 'SK_pe_hs_16_0_18_9Z_c',\n",
    "    'SK_pe_hs_16_0_18_2_c': 'SK_pe_hs_16_0_18_9Z12Z_c',\n",
    "    'SK_pe_hs_18_1_18_1_c': 'SK_pe_hs_18_9Z_18_9Z_c',\n",
    "    'SK_pe_hs_18_1_18_2_c': 'SK_pe_hs_18_9Z_18_9Z12Z_c',\n",
    "    'SK_pe_hs_18_2_16_0_c': 'SK_pe_hs_18_9Z12Z_16_0_c',\n",
    "    'SK_pe_hs_18_2_18_1_c': 'SK_pe_hs_18_9Z12Z_18_9Z_c',\n",
    "    # Exchanges\n",
    "    'EX_adrnl_e': 'EX_adrnl__L_e',\n",
    "    'EX_camp_e': 'EX_35camp_e',\n",
    "    'EX_dhdascb_e': 'EX_dhdascb__L_e',\n",
    "    'EX_dopa_e': 'EX_dpam__L_e',\n",
    "    'EX_nrpphr_e': 'EX_nrpphr__L_e',\n",
    "    'EX_mepi_e': 'EX_mepi__L_e',\n",
    "    'EX_hdca_e': 'EX_FA_hs_16_0_e',\n",
    "    'EX_ocdcea_e': 'EX_FA_hs_18_9Z_e',\n",
    "    'EX_lnlc_e': 'EX_FA_hs_18_9Z12Z_e',\n",
    "    # To match naming conventions, intracellular demands are irreversible while sinks are reversible.\n",
    "    'SK_adprbp_c': 'DM_adprbp_c',\n",
    "    'SK_mi1345p_c': 'DM_mi1345p_c',\n",
    "    'SK_mi134p_c': 'DM_mi134p_c',\n",
    "    'SK_mi145p_c': 'DM_mi145p_c',\n",
    "    'SK_mi14p_c': 'DM_mi14p_c',\n",
    "}\n",
    "\n",
    "for old_rxn, new_rxn in id_mapping_dict.items():\n",
    "    for model_reaction in model.reactions.query(lambda x: x.id == old_rxn):\n",
    "        model_reaction.id = new_rxn\n",
    "model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b0486ab5-0525-48a0-a854-998d27dd2bc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rxns</th>\n",
       "      <th>rxnRetired</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NADHload</td>\n",
       "      <td>DM_nadh</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAATPS2</td>\n",
       "      <td>CAATPS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BILIREDy</td>\n",
       "      <td>BILIRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTHOy</td>\n",
       "      <td>GTHOr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XYLTD_Dx</td>\n",
       "      <td>XYLTD_D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>DM_adprbp_c</td>\n",
       "      <td>SK_adprbp_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>DM_mi1345p_c</td>\n",
       "      <td>SK_mi1345p_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>DM_mi134p_c</td>\n",
       "      <td>SK_mi134p_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>DM_mi145p_c</td>\n",
       "      <td>SK_mi145p_c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>DM_mi14p_c</td>\n",
       "      <td>SK_mi14p_c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             rxns    rxnRetired\n",
       "0        NADHload       DM_nadh\n",
       "1         CAATPS2        CAATPS\n",
       "2        BILIREDy       BILIRED\n",
       "3           GTHOy         GTHOr\n",
       "4        XYLTD_Dx       XYLTD_D\n",
       "..            ...           ...\n",
       "179   DM_adprbp_c   SK_adprbp_c\n",
       "180  DM_mi1345p_c  SK_mi1345p_c\n",
       "181   DM_mi134p_c   SK_mi134p_c\n",
       "182   DM_mi145p_c   SK_mi145p_c\n",
       "183    DM_mi14p_c    SK_mi14p_c\n",
       "\n",
       "[184 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"rxnRetired\", \"rxns\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/deprecatedIdentifiers/replacedReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"rxns\", \"rxnRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[previous_id_mapping_df[\"rxns\"] == retiring]\n",
    "    retired_set_of_ids = set([retiring])\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        retired_set_of_ids.update(previously_retired[\"rxnRetired\"].apply(split_string).item())\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"rxnRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/replacedReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10101aba-02a3-4e30-a805-b7e7b0e751f9",
   "metadata": {},
   "source": [
    "#### Update boundary identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "cf3e6e40-6f22-4be0-9415-76afbc8bbcae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Type</th>\n",
       "      <th>Lower Bound</th>\n",
       "      <th>Upper Bound</th>\n",
       "      <th>Unit</th>\n",
       "      <th>References</th>\n",
       "      <th>Notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EX_5aop_e</td>\n",
       "      <td>5 Amino 4 oxopentanoate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.020000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:900140</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EX_acald_e</td>\n",
       "      <td>Acetaldehyde exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-10.000000</td>\n",
       "      <td>10.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:28460</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EX_acnam_e</td>\n",
       "      <td>N-Acetylneuraminate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:12765793</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EX_ade_e</td>\n",
       "      <td>Adenine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.014000</td>\n",
       "      <td>0.010000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:2141093</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EX_adn_e</td>\n",
       "      <td>Adenosine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.010000</td>\n",
       "      <td>0.014000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:2141093</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EX_adrnl__L_e</td>\n",
       "      <td>Adrenaline exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.037800</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:8737076</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>EX_arg__L_e</td>\n",
       "      <td>L-Arginine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.115200</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:6817948, PMID:11950212</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>EX_chol_e</td>\n",
       "      <td>Choline exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.012000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:5651769</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>EX_cys__L_e</td>\n",
       "      <td>L-Cysteine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.080000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:6109287</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>EX_dhdascb__L_e</td>\n",
       "      <td>Dehydroascorbate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.111100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:9586809, PMID:6708818</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EX_dpam__L_e</td>\n",
       "      <td>Dopamine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.032400</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:8737076</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EX_etha_e</td>\n",
       "      <td>Ethanolamine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.012000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:5651769</td>\n",
       "      <td>Estimated to Choline Levels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EX_fru_e</td>\n",
       "      <td>D-Fructose exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.007500</td>\n",
       "      <td>1.120000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:9166863</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EX_fum_e</td>\n",
       "      <td>Fumarate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.250000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:7138916</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>EX_gal_e</td>\n",
       "      <td>D-Galactose exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.316900</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:722542</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>EX_gam_e</td>\n",
       "      <td>D-Glucosamine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000010</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:12765793</td>\n",
       "      <td>Estimate to N-Acetylneuraminate Levels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>EX_glc__D_e</td>\n",
       "      <td>D-Glucose exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-1.120000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:2141093</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>EX_gly_e</td>\n",
       "      <td>Glycine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.080000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:7320944</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>EX_glyc_e</td>\n",
       "      <td>Glycerol exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.012000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:826280</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>EX_FA_hs_16_0_e</td>\n",
       "      <td>Hexadecanoate (n-C16:0) exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.012000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:826280</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>EX_leuktrA4_e</td>\n",
       "      <td>Leukotriene A4 exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000002</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:2168394</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>EX_FA_hs_18_9Z12Z_e</td>\n",
       "      <td>Linoleic acid (all cis C18:2) exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.006000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:826280</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>EX_mal__L_e</td>\n",
       "      <td>L-Malate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.500000</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:7138916</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>EX_man_e</td>\n",
       "      <td>D-Mannose exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.010000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:9169093</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>EX_met__L_e</td>\n",
       "      <td>L-Methionine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.010000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:1275912</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>EX_nac_e</td>\n",
       "      <td>Nicotinate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.001000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:4231915</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>EX_ncam_e</td>\n",
       "      <td>Nicotinamide exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:4231915</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>EX_nrpphr__L_e</td>\n",
       "      <td>Norepinephrine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.028800</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:8737076</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>EX_FA_hs_18_9Z_e</td>\n",
       "      <td>Octadecenoate (n-C18:1) exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.006000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:826280</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>EX_orot_e</td>\n",
       "      <td>Orotate exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.010000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:2141093</td>\n",
       "      <td>Estimated to Adenine uptake</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>EX_ptrc_e</td>\n",
       "      <td>Putrescine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115200</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:8679659, PMID:1484149</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>EX_pydam_e</td>\n",
       "      <td>Pyridoxamine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:7440576</td>\n",
       "      <td>Estimated to other B6 derivative levels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>EX_pydx_e</td>\n",
       "      <td>Pyridoxal exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000155</td>\n",
       "      <td>0.000228</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:7440576</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>EX_pydxn_e</td>\n",
       "      <td>Pyridoxine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000128</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:7440576</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>EX_spmd_e</td>\n",
       "      <td>Spermidine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.080000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:8679659, PMID:1484149</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>EX_sprm_e</td>\n",
       "      <td>Spermine exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:1484149</td>\n",
       "      <td>Enzyme properties estimated from other polyamines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>EX_thm_e</td>\n",
       "      <td>Thiamin exchange</td>\n",
       "      <td>exchange</td>\n",
       "      <td>-0.000060</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>mmol / (hr * L cells)</td>\n",
       "      <td>PMID:2283678</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ID                                    Name      Type  \\\n",
       "0             EX_5aop_e        5 Amino 4 oxopentanoate exchange  exchange   \n",
       "1            EX_acald_e                   Acetaldehyde exchange  exchange   \n",
       "2            EX_acnam_e            N-Acetylneuraminate exchange  exchange   \n",
       "3              EX_ade_e                        Adenine exchange  exchange   \n",
       "4              EX_adn_e                      Adenosine exchange  exchange   \n",
       "5         EX_adrnl__L_e                     Adrenaline exchange  exchange   \n",
       "6           EX_arg__L_e                     L-Arginine exchange  exchange   \n",
       "7             EX_chol_e                        Choline exchange  exchange   \n",
       "8           EX_cys__L_e                     L-Cysteine exchange  exchange   \n",
       "9       EX_dhdascb__L_e               Dehydroascorbate exchange  exchange   \n",
       "10         EX_dpam__L_e                       Dopamine exchange  exchange   \n",
       "11            EX_etha_e                   Ethanolamine exchange  exchange   \n",
       "12             EX_fru_e                     D-Fructose exchange  exchange   \n",
       "13             EX_fum_e                       Fumarate exchange  exchange   \n",
       "14             EX_gal_e                    D-Galactose exchange  exchange   \n",
       "15             EX_gam_e                  D-Glucosamine exchange  exchange   \n",
       "16          EX_glc__D_e                      D-Glucose exchange  exchange   \n",
       "17             EX_gly_e                        Glycine exchange  exchange   \n",
       "18            EX_glyc_e                       Glycerol exchange  exchange   \n",
       "19      EX_FA_hs_16_0_e        Hexadecanoate (n-C16:0) exchange  exchange   \n",
       "20        EX_leuktrA4_e                 Leukotriene A4 exchange  exchange   \n",
       "21  EX_FA_hs_18_9Z12Z_e  Linoleic acid (all cis C18:2) exchange  exchange   \n",
       "22          EX_mal__L_e                       L-Malate exchange  exchange   \n",
       "23             EX_man_e                      D-Mannose exchange  exchange   \n",
       "24          EX_met__L_e                   L-Methionine exchange  exchange   \n",
       "25             EX_nac_e                     Nicotinate exchange  exchange   \n",
       "26            EX_ncam_e                   Nicotinamide exchange  exchange   \n",
       "27       EX_nrpphr__L_e                 Norepinephrine exchange  exchange   \n",
       "28     EX_FA_hs_18_9Z_e        Octadecenoate (n-C18:1) exchange  exchange   \n",
       "29            EX_orot_e                        Orotate exchange  exchange   \n",
       "30            EX_ptrc_e                     Putrescine exchange  exchange   \n",
       "31           EX_pydam_e                   Pyridoxamine exchange  exchange   \n",
       "32            EX_pydx_e                      Pyridoxal exchange  exchange   \n",
       "33           EX_pydxn_e                     Pyridoxine exchange  exchange   \n",
       "34            EX_spmd_e                     Spermidine exchange  exchange   \n",
       "35            EX_sprm_e                       Spermine exchange  exchange   \n",
       "36             EX_thm_e                        Thiamin exchange  exchange   \n",
       "\n",
       "    Lower Bound  Upper Bound                   Unit  \\\n",
       "0     -0.020000     0.000000  mmol / (hr * L cells)   \n",
       "1    -10.000000    10.000000  mmol / (hr * L cells)   \n",
       "2     -0.000037     0.000000  mmol / (hr * L cells)   \n",
       "3     -0.014000     0.010000  mmol / (hr * L cells)   \n",
       "4     -0.010000     0.014000  mmol / (hr * L cells)   \n",
       "5     -0.037800     0.000000  mmol / (hr * L cells)   \n",
       "6     -0.115200     0.000000  mmol / (hr * L cells)   \n",
       "7     -0.012000     0.000000  mmol / (hr * L cells)   \n",
       "8     -0.080000     0.000000  mmol / (hr * L cells)   \n",
       "9     -0.111100     0.000000  mmol / (hr * L cells)   \n",
       "10    -0.032400     0.000000  mmol / (hr * L cells)   \n",
       "11    -0.012000     0.000000  mmol / (hr * L cells)   \n",
       "12    -0.007500     1.120000  mmol / (hr * L cells)   \n",
       "13    -0.250000     0.500000  mmol / (hr * L cells)   \n",
       "14    -0.316900     0.000000  mmol / (hr * L cells)   \n",
       "15    -0.000010     0.000000  mmol / (hr * L cells)   \n",
       "16    -1.120000     0.000000  mmol / (hr * L cells)   \n",
       "17    -0.080000     0.000000  mmol / (hr * L cells)   \n",
       "18    -0.012000     0.000000  mmol / (hr * L cells)   \n",
       "19    -0.012000     0.000000  mmol / (hr * L cells)   \n",
       "20    -0.000002     0.000000  mmol / (hr * L cells)   \n",
       "21    -0.006000     0.000000  mmol / (hr * L cells)   \n",
       "22    -0.500000     0.250000  mmol / (hr * L cells)   \n",
       "23    -0.010000     0.000000  mmol / (hr * L cells)   \n",
       "24    -0.010000     0.000000  mmol / (hr * L cells)   \n",
       "25    -0.001000     0.000000  mmol / (hr * L cells)   \n",
       "26     0.000000     0.001000  mmol / (hr * L cells)   \n",
       "27    -0.028800     0.000000  mmol / (hr * L cells)   \n",
       "28    -0.006000     0.000000  mmol / (hr * L cells)   \n",
       "29    -0.010000     0.000000  mmol / (hr * L cells)   \n",
       "30     0.000000     0.115200  mmol / (hr * L cells)   \n",
       "31    -0.000100     0.000000  mmol / (hr * L cells)   \n",
       "32    -0.000155     0.000228  mmol / (hr * L cells)   \n",
       "33    -0.000128     0.000255  mmol / (hr * L cells)   \n",
       "34     0.000000     0.080000  mmol / (hr * L cells)   \n",
       "35     0.000000     0.040000  mmol / (hr * L cells)   \n",
       "36    -0.000060     0.000000  mmol / (hr * L cells)   \n",
       "\n",
       "                     References  \\\n",
       "0                   PMID:900140   \n",
       "1                    PMID:28460   \n",
       "2                 PMID:12765793   \n",
       "3                  PMID:2141093   \n",
       "4                  PMID:2141093   \n",
       "5                  PMID:8737076   \n",
       "6   PMID:6817948, PMID:11950212   \n",
       "7                  PMID:5651769   \n",
       "8                  PMID:6109287   \n",
       "9    PMID:9586809, PMID:6708818   \n",
       "10                 PMID:8737076   \n",
       "11                 PMID:5651769   \n",
       "12                 PMID:9166863   \n",
       "13                 PMID:7138916   \n",
       "14                  PMID:722542   \n",
       "15                PMID:12765793   \n",
       "16                 PMID:2141093   \n",
       "17                 PMID:7320944   \n",
       "18                  PMID:826280   \n",
       "19                  PMID:826280   \n",
       "20                 PMID:2168394   \n",
       "21                  PMID:826280   \n",
       "22                 PMID:7138916   \n",
       "23                 PMID:9169093   \n",
       "24                 PMID:1275912   \n",
       "25                 PMID:4231915   \n",
       "26                 PMID:4231915   \n",
       "27                 PMID:8737076   \n",
       "28                  PMID:826280   \n",
       "29                 PMID:2141093   \n",
       "30   PMID:8679659, PMID:1484149   \n",
       "31                 PMID:7440576   \n",
       "32                 PMID:7440576   \n",
       "33                 PMID:7440576   \n",
       "34   PMID:8679659, PMID:1484149   \n",
       "35                 PMID:1484149   \n",
       "36                 PMID:2283678   \n",
       "\n",
       "                                                Notes  \n",
       "0                                                      \n",
       "1                                                      \n",
       "2                                                      \n",
       "3                                                      \n",
       "4                                                      \n",
       "5                                                      \n",
       "6                                                      \n",
       "7                                                      \n",
       "8                                                      \n",
       "9                                                      \n",
       "10                                                     \n",
       "11                        Estimated to Choline Levels  \n",
       "12                                                     \n",
       "13                                                     \n",
       "14                                                     \n",
       "15             Estimate to N-Acetylneuraminate Levels  \n",
       "16                                                     \n",
       "17                                                     \n",
       "18                                                     \n",
       "19                                                     \n",
       "20                                                     \n",
       "21                                                     \n",
       "22                                                     \n",
       "23                                                     \n",
       "24                                                     \n",
       "25                                                     \n",
       "26                                                     \n",
       "27                                                     \n",
       "28                                                     \n",
       "29                        Estimated to Adenine uptake  \n",
       "30                                                     \n",
       "31            Estimated to other B6 derivative levels  \n",
       "32                                                     \n",
       "33                                                     \n",
       "34                                                     \n",
       "35  Enzyme properties estimated from other polyamines  \n",
       "36                                                     "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_bounds = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/parameterization/iAB-RBC-283_Bordbar2011.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ").fillna(\"\")\n",
    "\n",
    "for idx, id_str in enumerate(df_bounds[\"ID\"].values):\n",
    "    if id_str in id_mapping_dict:\n",
    "        df_bounds.loc[idx, \"ID\"] = id_mapping_dict[id_str]\n",
    "\n",
    "df_bounds.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/iAB-RBC-283_Bordbar2011.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "df_bounds"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e96a25b6-7c0f-425a-abf0-53c8a5f52154",
   "metadata": {},
   "source": [
    "#### Subsystem updates\n",
    "\n",
    "All subsystems are updated to the corresponding group in HumanGEM.\n",
    "\n",
    "Notable exceptions are as follows:\n",
    "* \"ROS detoxification\" is now \"Reactive species formation and detoxification\"\n",
    "* \"Urea Cycle\" is combined with \"Arginine and proline metabolism\"\n",
    "* \"Oxidative Phosphorylation\" contains only `PPA` reaction, which is also the only cytosolic reaction in the HumanGEM subsystem. As reaction could be placed into a variety of categories, it was placed in the \"Nucleotide metabolism\" subsystem in which inorganic diphosphate is generated.\n",
    "* Subsystems that contains \"Biosynthesis\" were replaced with the more generic term \"metabolism\", better reflecting RBCs. Includes the following:\n",
    "    * \"Phenylalanine, tyrosine and tryptophan biosynthesis\"\n",
    "* \"Tyrosine metabolism\" was grouped into \"Phenylalanine, tyrosine and tryptophan metabolism\".\n",
    "* \"Eicosanoid Metabolism contains only `LTA4H` reaction, which becomes part of the `Leukotriene` subsystem in HumanGEM.\n",
    "* 'Intracellular source/sink' Extracellular exchange are grouped under Pseudoreactions rather than of `Exchange/demand reactions`, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "76db05c7-160e-4176-ad9b-387a1813af79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Group Alanine, aspartate and glutamate metabolism at 0x14b3714c0>,\n",
       " <Group Amino sugar and nucleotide sugar metabolism at 0x14b371610>,\n",
       " <Group Arginine and proline metabolism at 0x14b370c50>,\n",
       " <Group Ascorbate and aldarate metabolism at 0x14b370fe0>,\n",
       " <Group Carnitine shuttle (cytosolic) at 0x14b371e20>,\n",
       " <Group Cysteine and methionine metabolism at 0x14b370a40>,\n",
       " <Group Fatty acid activation (cytosolic) at 0x14b3711f0>,\n",
       " <Group Fructose and mannose metabolism at 0x14b371be0>,\n",
       " <Group Galactose metabolism at 0x14b371580>,\n",
       " <Group Glutathione metabolism at 0x14b3715e0>,\n",
       " <Group Glycerolipid metabolism at 0x150ac8d70>,\n",
       " <Group Glycerophospholipid metabolism at 0x14b371ca0>,\n",
       " <Group Glycolysis / Gluconeogenesis at 0x14b3713d0>,\n",
       " <Group Inositol phosphate metabolism at 0x14b371550>,\n",
       " <Group Leukotriene metabolism at 0x14b3717c0>,\n",
       " <Group Miscellaneous at 0x14b3718e0>,\n",
       " <Group Nicotinate and nicotinamide metabolism at 0x14b371d30>,\n",
       " <Group Nucleotide metabolism at 0x14b267fb0>,\n",
       " <Group Pentose and glucuronate interconversions at 0x14b371490>,\n",
       " <Group Pentose phosphate pathway at 0x14b371760>,\n",
       " <Group Phenylalanine, tyrosine and tryptophan metabolism at 0x14b371b20>,\n",
       " <Group Porphyrin metabolism at 0x14b3719a0>,\n",
       " <Group Pseudoreactions at 0x14b371790>,\n",
       " <Group Purine metabolism at 0x14b371bb0>,\n",
       " <Group Pyrimidine metabolism at 0x14b371400>,\n",
       " <Group Pyruvate metabolism at 0x14b371940>,\n",
       " <Group Reactive species formation and detoxification at 0x14b3715b0>,\n",
       " <Group Riboflavin metabolism at 0x14b3716d0>,\n",
       " <Group Starch and sucrose metabolism at 0x14b371a00>,\n",
       " <Group Thiamine metabolism at 0x14b3719d0>,\n",
       " <Group Transport reactions at 0x14b371700>,\n",
       " <Group Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism at 0x14b371a30>,\n",
       " <Group Vitamin B6 metabolism at 0x14b3717f0>]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsystem_mapping = {\n",
    "    # Special\n",
    "    'Oxidative Phosphorylation': 'Nucleotide metabolism',\n",
    "    'Diacylglycerol Synthesis': 'Glycerolipid metabolism',\n",
    "    'ROS Detoxification': 'Reactive species formation and detoxification',\n",
    "    'Eicosanoid Metabolism': 'Leukotriene metabolism',\n",
    "\n",
    "    'Intracellular source/sink': 'Pseudoreactions',\n",
    "    'Extracellular exchange': 'Pseudoreactions',\n",
    "    # Straightforward\n",
    "    ## Carbohydrate\n",
    "    'Glycolysis/Gluconeogenesis': 'Glycolysis / Gluconeogenesis',\n",
    "    'Galactose metabolism': 'Galactose metabolism',\n",
    "    'Inositol Phosphate Metabolism': 'Inositol phosphate metabolism',\n",
    "    'Ascorbate and Aldarate Metabolism': 'Ascorbate and aldarate metabolism',\n",
    "\n",
    "    'Transport, Extracellular': 'Transport reactions',\n",
    "    'Vitamin B6 Metabolism': 'Vitamin B6 metabolism',\n",
    "    'Urea cycle/amino group metabolism': 'Arginine and proline metabolism',\n",
    "    'Porphyrin metabolism': 'Porphyrin metabolism',\n",
    "    'Methionine Salvage': 'Cysteine and methionine metabolism',\n",
    "    'Thiamine Metabolism': 'Thiamine metabolism',\n",
    "    'Miscellaneous': 'Miscellaneous',\n",
    "    'Pentose Phosphate Pathway': 'Pentose phosphate pathway',\n",
    "    'Riboflavin Metabolism': 'Riboflavin metabolism',\n",
    "    'Arginine and Proline Metabolism': 'Arginine and proline metabolism',\n",
    "    'Glutamate metabolism': 'Alanine, aspartate and glutamate metabolism',\n",
    "    'Salvage Pathway': 'Purine metabolism',\n",
    "    'Phenylalanine metabolism': 'Phenylalanine, tyrosine and tryptophan metabolism',\n",
    "    'Tyrosine metabolism': 'Phenylalanine, tyrosine and tryptophan metabolism',\n",
    "    'Nucleotides': 'Nucleotide metabolism',\n",
    "    'Heme Degradation': 'Porphyrin metabolism',\n",
    "    'Heme Biosynthesis': 'Porphyrin metabolism',\n",
    "    'Fructose and Mannose Metabolism': 'Fructose and mannose metabolism',\n",
    "    'Purine Catabolism': 'Purine metabolism',\n",
    "    'Aminosugar Metabolism': 'Amino sugar and nucleotide sugar metabolism',\n",
    "    'Glycerophospholipid Metabolism': 'Glycerophospholipid metabolism',\n",
    "    'Pyrimidine Catabolism': 'Pyrimidine metabolism',\n",
    "    'Citric Acid Cycle': 'Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism',\n",
    "    'Pyrimidine Biosynthesis': 'Pyrimidine metabolism',\n",
    "    'Pyruvate Metabolism': 'Pyruvate metabolism',\n",
    "    'NAD Metabolism': 'Nicotinate and nicotinamide metabolism',\n",
    "    'Pentose and Glucuronate Interconversions': 'Pentose and glucuronate interconversions',\n",
    "    'Carnitine shuttle': 'Carnitine shuttle (cytosolic)',\n",
    "    'Glutathione Metabolism': 'Glutathione metabolism',\n",
    "    'Methionine Metabolism': 'Cysteine and methionine metabolism',\n",
    "    'Fatty acid activation': 'Fatty acid activation (cytosolic)',\n",
    "    'Starch and Sucrose Metabolism': 'Starch and sucrose metabolism',\n",
    "}\n",
    "\n",
    "individual_subsystem_updates = {\n",
    "    \"Arginine and proline metabolism\": [\"MTAP\"],\n",
    "    \"Purine metabolism\": [\"ADK1\", \"ADNCYC\", \"ADNK1\", \"AMPDA\", \"PDEA\", \"GUACYC\",\n",
    "                          \"NDPK1\", \"IMPD\", \"GK1\", \"NTD7\", \"NTD9\", \"GMPR\", \"GMPS2\"],\n",
    "    \"Pyrimidine metabolism\": [\"UMPK1\", \"NDPK2\", \"NDPK3\"],\n",
    "    \"Pyruvate metabolism\": [\"ALDD2x\"],\n",
    "    \"Nucleotide metabolism\": [\"AP4AH1\"],\n",
    "    \"Miscellaneous\": [\"BANDMT\"],\n",
    "    \n",
    "    \n",
    "}\n",
    "for subsystem, reaction_id_list in individual_subsystem_updates.items():\n",
    "    reaction_list = model.reactions.get_by_any(reaction_id_list)\n",
    "    for reaction in reaction_list:\n",
    "        reaction.subsystem = subsystem\n",
    "\n",
    "model.remove_groups(list(model.groups))\n",
    "for old_subsystem, new_subsystem in subsystem_mapping.items():\n",
    "    reaction_list = model.reactions.query(lambda x: x.subsystem == old_subsystem)\n",
    "    for reaction in reaction_list:\n",
    "        reaction.subsystem = new_subsystem\n",
    "    if new_subsystem not in model.groups:\n",
    "        group = Group(id=new_subsystem, name=new_subsystem, members=reaction_list)\n",
    "        model.add_groups([group])\n",
    "    else:\n",
    "        group = model.groups.get_by_id(new_subsystem).add_members(reaction_list)\n",
    "\n",
    "for subsystem, reaction_id_list in individual_subsystem_updates.items():\n",
    "    reaction_list = model.reactions.get_by_any(reaction_id_list)\n",
    "    if subsystem not in model.groups:\n",
    "        group = Group(id=subsystem, name=subsystem, members=reaction_list)\n",
    "        model.add_groups([group])\n",
    "    else:\n",
    "        group = model.groups.get_by_id(subsystem).add_members(reaction_list)\n",
    "\n",
    "model.groups.sort()\n",
    "model.groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "222f5a0d-f7c8-4a62-85c4-04ea5c52d499",
   "metadata": {},
   "source": [
    "### Standardize metabolite formulas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1d2f379e-1216-40aa-a6ef-1529323aa1da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Standardizing formula for `band_c`\n",
      "Standardizing formula for `bandmt_c`\n"
     ]
    }
   ],
   "source": [
    "metabolite_formulas = dict(zip(\n",
    "    model.metabolites.list_attr(\"id\"), \n",
    "    model.metabolites.list_attr(\"formula\")\n",
    "))\n",
    "standardized = standardardize_metabolite_formulas(metabolite_formulas)\n",
    "\n",
    "for mid, updated_formula in standardized.items():\n",
    "    if metabolite_formulas[mid] != updated_formula:\n",
    "        print(f\"Standardizing formula for `{mid}`\")\n",
    "        model_metabolite = model.metabolites.get_by_id(mid)\n",
    "        model_metabolite.formula = updated_formula"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f939329a-022d-4a39-94f4-7f0ba639358f",
   "metadata": {},
   "source": [
    "## Export updated model\n",
    "### Version: 0.2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d237f4db-3ee3-4fd6-96f9-68b78290d4ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>iAB_RBC_283</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>14a29c890</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>342</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>469</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>349</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>33</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*NaKt - 1.0*NaKt_reverse_db47e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>cytosol, extracellular space</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model iAB_RBC_283 at 0x14a29c890>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_rbc_model(model, \"all\")\n",
    "model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
