{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "291414c1-9b6f-4338-b0bc-13c8958d259d",
   "metadata": {},
   "source": [
    "# RBC-GEM 0.1.1 --> 0.2.0\n",
    "## Update iAB-RBC-283 model \n",
    "\n",
    "The purpose of this notebook is to update the iAB-RBC-283 reconstruction for the following purposes:\n",
    "1. To serve as a base/draft model for the initial expansion of iAB-RBC-283 to RBC-GEM.\n",
    "2. To serve as a model for comparison to the new reconstruction.\n",
    "3. To link to the HumanGEM reconstruction annotations where possible in the model.\n",
    "\n",
    "No stoichiometric changes are made in this minor update, however BIGG identifiers are altered in order to enable better comparision to the RBC-GEM. Furthermore, associations to Recon3D and HumanGEM are made here, as subsequent changes past 0.2.0 will be in an effort to modernize to current COBRA standards before iterative expansion.\n",
    "\n",
    "RBC-GEM 0.2.0 will remain identified as iAB-RBC-283 in order to help future comparisions.\n",
    "\n",
    "Bordbar, A., Jamshidi, N. & Palsson, B.O. iAB-RBC-283: A proteomically derived knowledge-base of erythrocyte metabolism that can be used to simulate its physiological and patho-physiological states. BMC Syst Biol 5, 110 (2011). https://doi.org/10.1186/1752-0509-5-110\n",
    "\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54ef02a3-2f02-4072-a04d-36e2689794c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter, defaultdict\n",
    "\n",
    "import cobra\n",
    "import pandas as pd\n",
    "from cobra.core import Group\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    ROOT_PATH,\n",
    "    read_cobra_model,\n",
    "    read_rbc_model,\n",
    "    show_versions,\n",
    "    write_rbc_model,\n",
    ")\n",
    "from rbc_gem_utils.qc import standardardize_metabolite_formulas\n",
    "from rbc_gem_utils.util import build_string, split_string\n",
    "\n",
    "# Display versions of last time notebook ran and worked\n",
    "show_versions()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6abc5fbf-6345-4620-a0a2-1e23fc6dd504",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51309b09-8cd3-47f8-a7a4-db996ff17e8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28d59af5-dfc2-4ef6-a9e2-b1911f3b0032",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 0.1.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c91c721-5625-477c-bfc0-dee5c8642fbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = read_rbc_model(\"xml\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79f0984b-4328-4625-b8c6-5cac3f291493",
   "metadata": {},
   "source": [
    "## Extract identifiers from HumanGEM\n",
    "\n",
    "Use HumanGEM (1.18.0) to get the MetabolicAtlas identifiers for reactions and metabolites.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac74f778-d659-4f2a-8e84-8dd4daf9bbad",
   "metadata": {},
   "outputs": [],
   "source": [
    "HumanGEM = read_cobra_model(f\"{ROOT_PATH}/data/raw/Human-GEM.xml\")\n",
    "HumanGEM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b50a5606-37f7-4439-9a60-9c51676db717",
   "metadata": {},
   "source": [
    "### Map model BiGG ID metabolites to MetabolicAtlas via HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69eda8e4-baef-4d14-8c73-9b3413dd7dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_metabolite_ids = {\n",
    "    metabolite.id.replace(f\"_{metabolite.compartment}\", \"\")\n",
    "    for metabolite in model.metabolites\n",
    "}\n",
    "bigg_metabolites = set()\n",
    "human_gem_mapping = {}\n",
    "for metabolite in HumanGEM.metabolites:\n",
    "    bigg_ids = metabolite.annotation.get(\"bigg.metabolite\")\n",
    "    if not bigg_ids:\n",
    "        continue\n",
    "    if isinstance(bigg_ids, str):\n",
    "        bigg_ids = [bigg_ids]\n",
    "    bigg_ids = set(bigg_ids)\n",
    "    bigg_metabolites.update(bigg_ids)\n",
    "    for bigg_id in bigg_ids:\n",
    "        if bigg_id in human_gem_mapping:\n",
    "            human_gem_mapping[bigg_id] += [metabolite.id[:-1]]\n",
    "        else:\n",
    "            human_gem_mapping[bigg_id] = [metabolite.id[:-1]]\n",
    "\n",
    "\n",
    "print(\n",
    "    f\"Number of metabolites in RBC-GEM (excluding compartments): {len(model_metabolite_ids)}\"\n",
    ")\n",
    "\n",
    "intersection = sorted(bigg_metabolites.intersection(model_metabolite_ids))\n",
    "print(f\"Number of metabolites that could be found: {len(intersection)}\")\n",
    "print()\n",
    "\n",
    "id_mapping_dict = {}\n",
    "for rbc_metabolite in intersection:\n",
    "    metabolite_ids = set(human_gem_mapping.get(rbc_metabolite, []))\n",
    "    if len(metabolite_ids) > 1:\n",
    "        print(f'\"{rbc_metabolite}\": \"{build_string(metabolite_ids)}\",')\n",
    "    id_mapping_dict[rbc_metabolite] = build_string(metabolite_ids)\n",
    "\n",
    "manual_updates_and_corrections = {\n",
    "    # Selected reactions below needed corrections.\n",
    "}\n",
    "\n",
    "id_mapping_dict.update(manual_updates_and_corrections)\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"bigg\", \"metatlas\"]\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86b95018-d02f-47ee-9d3c-bb8fecc1b6b0",
   "metadata": {},
   "source": [
    "### Map model BiGG ID reactions to MetabolicAtlas via HumanGEM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d67ebee9-9d84-4001-af42-d5bc218105ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_reaction_ids = {reaction.id for reaction in model.reactions}\n",
    "\n",
    "bigg_reactions = set()\n",
    "human_gem_mapping = {}\n",
    "for reaction in HumanGEM.reactions:\n",
    "    bigg_ids = reaction.annotation.get(\"bigg.reaction\")\n",
    "    if not bigg_ids:\n",
    "        continue\n",
    "    if isinstance(bigg_ids, str):\n",
    "        bigg_ids = [bigg_ids]\n",
    "    bigg_ids = set(bigg_ids)\n",
    "    bigg_reactions.update(bigg_ids)\n",
    "    for bigg_id in bigg_ids:\n",
    "        if bigg_id in human_gem_mapping:\n",
    "            human_gem_mapping[bigg_id] += [reaction.id]\n",
    "        else:\n",
    "            human_gem_mapping[bigg_id] = [reaction.id]\n",
    "\n",
    "\n",
    "print(\n",
    "    f\"Number of reactions in RBC-GEM (excluding boundaries): {len(model_reaction_ids)}\"\n",
    ")\n",
    "\n",
    "intersection = sorted(bigg_reactions.intersection(model_reaction_ids))\n",
    "print(f\"Number of reactions that could be found: {len(intersection)}\")\n",
    "print()\n",
    "id_mapping_dict = {}\n",
    "for rbc_reaction in intersection:\n",
    "    reaction_ids = human_gem_mapping.get(rbc_reaction, [])\n",
    "    if len(reaction_ids) > 1:\n",
    "        print(f\"Check {rbc_reaction} for false mappings\")\n",
    "    id_mapping_dict[rbc_reaction] = build_string(reaction_ids)\n",
    "\n",
    "manual_updates_and_corrections = {\n",
    "    # Selected reactions below needed corrections.\n",
    "    \"CAATPS\": \"MAR07629\",\n",
    "    \"CAT\": \"MAR03980\",\n",
    "    \"CHLPCTD\": \"MAR00638\",\n",
    "    \"CHOLK\": \"MAR00636\",\n",
    "    \"CHOLt4\": \"MAR07734\",\n",
    "    \"COt\": \"MAR07798\",\n",
    "    \"CYStec\": \"MAR05084\",\n",
    "    \"CYTK1\": \"MAR04024\",\n",
    "    \"EX_ac_e\": \"MAR09086\",\n",
    "    \"EX_adrnl_e\": \"MAR09095\",\n",
    "    \"EX_ala__L_e\": \"MAR09061\",\n",
    "    \"EX_arg__L_e\": \"MAR09066\",\n",
    "    \"EX_chol_e\": \"MAR09083\",\n",
    "    \"EX_cl_e\": \"MAR09150\",\n",
    "    \"EX_dopa_e\": \"MAR09092\",\n",
    "    \"EX_fe2_e\": \"MAR09076\",\n",
    "    \"EX_gal_e\": \"MAR09140\",\n",
    "    \"EX_gam_e\": \"MAR09168\",\n",
    "    \"EX_glc__D_e\": \"MAR09034\",\n",
    "    \"EX_gln__L_e\": \"MAR09063\",\n",
    "    \"EX_h2o_e\": \"MAR09047\",\n",
    "    \"EX_h_e\": \"MAR09079\",\n",
    "    \"EX_hco3_e\": \"MAR09078\",\n",
    "    \"EX_k_e\": \"MAR09081\",\n",
    "    \"EX_lnlc_e\": \"MAR09035\",\n",
    "    \"EX_met__L_e\": \"MAR09042\",\n",
    "    \"EX_nac_e\": \"MAR09142\",\n",
    "    \"EX_nh4_e\": \"MAR11420\",\n",
    "    \"EX_nrpphr_e\": \"MAR09093\",\n",
    "    \"EX_ocdcea_e\": \"MAR00650\",\n",
    "    \"EX_phe__L_e\": \"MAR09043\",\n",
    "    \"EX_pyr_e\": \"MAR09133\",\n",
    "    \"EX_ribflv_e\": \"MAR09143\",\n",
    "    \"EX_thmmp_e\": \"MAR09105\",\n",
    "    \"FBA\": \"MAR04375\",\n",
    "    \"FBP26\": \"MAR04706\",\n",
    "    \"FUM\": \"MAR04408\",\n",
    "    \"G6PDH2r\": \"MAR08971\",\n",
    "    \"GALKr\": \"MAR04130\",\n",
    "    \"GAMt1r\": \"MAR04996\",\n",
    "    \"GAPD\": \"MAR04373\",\n",
    "    \"GLNt4\": \"MAR05308\",\n",
    "    \"MEPIVESSte\": \"MAR08922\",\n",
    "    \"NADK\": \"MAR04269\",\n",
    "    \"NH4t3r\": \"MAR01534\",\n",
    "    \"PPPGO\": \"MAR11316\",\n",
    "    # Selected reactions needed manaul updates\n",
    "    \"ACP1_FMN\": \"MAR06507\",\n",
    "    \"ADRNLtu\": \"MAR09192\",\n",
    "    \"ARD\": \"MAR05389\",\n",
    "    \"ARGN\": \"MAR03816\",\n",
    "    \"BILIRBU\": \"MAR11321\",\n",
    "    \"C160CPT2rbc\": \"MAR02626\",\n",
    "    \"C181CPT2rbc\": \"MAR11310\",\n",
    "    \"CHLP\": \"MAR08424\",\n",
    "    \"DGULND\": \"MAR08353\",\n",
    "    \"DHAAt1r\": \"MAR08846\",\n",
    "    \"DOPAMT\": \"MAR06763\",\n",
    "    \"DPGM\": \"MAR04371\",\n",
    "    \"DPGase\": \"MAR04372\",\n",
    "    \"ENOPH\": \"MAR05387\",\n",
    "    \"ETHAt\": \"MAR07896\",\n",
    "    \"GALOR\": \"MAR08766\",\n",
    "    \"GALT\": \"MAR08767\",\n",
    "    \"GLCt1\": \"MAR05029\",\n",
    "    \"GMPR\": \"MAR04419\",\n",
    "    \"GPDDA1\": \"MAR00635\",\n",
    "    \"GULND\": \"MAR06537\",\n",
    "    \"HCO3_CLt\": \"MAR06525\",\n",
    "    \"LEUKTRA4t\": \"MAR06254\",\n",
    "    \"LEUKTRB4t\": \"MAR06255\",\n",
    "    \"LNLCCPT2rbc\": \"MAR02742\",\n",
    "    \"LTA4H\": \"MAR01080\",\n",
    "    \"MDRPD\": \"MAR05386\",\n",
    "    \"MI1345PP\": \"MAR06565\",\n",
    "    \"MI145PK\": \"MAR06563\",\n",
    "    \"MI145PP\": \"MAR06560\",\n",
    "    \"MTRI\": \"MAR05385\",\n",
    "    \"NADPN\": \"MAR07627\",\n",
    "    \"NMNHYD\": \"MAR04264\",\n",
    "    \"ORNDC\": \"MAR04212\",\n",
    "    \"RNMK\": \"MAR04265\",\n",
    "    \"SALMCOM\": \"MAR06750\",\n",
    "    \"SALMCOM2\": \"MAR06746\",\n",
    "    \"SBTD_D2\": \"MAR04315\",\n",
    "    \"SBTR\": \"MAR04316\",\n",
    "    \"SPMDtex2\": \"MAR04994\",\n",
    "    \"TDP\": \"MAR04208\",\n",
    "    \"THMTP\": \"MAR04207\",\n",
    "    \"TMDPK\": \"MAR04204\",\n",
    "    \"TMDPPK\": \"MAR04206\",\n",
    "    \"UNK3\": \"MAR05391\",\n",
    "    \"UPPDC1\": \"MAR04750\",\n",
    "    \"XYLK\": \"MAR04595\",\n",
    "    \"XYLTD_D\": \"MAR04593\",\n",
    "}\n",
    "\n",
    "id_mapping_dict.update(manual_updates_and_corrections)\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"bigg\", \"metatlas\"]\n",
    "# id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "001284e3-012f-4e65-9592-16a9ab8fa2e5",
   "metadata": {},
   "source": [
    "## Update genes and gene reaction rules\n",
    "1. Update gene IDs to match RECON3D with inclusion of NCBI Gene ID.\n",
    "2. Manually update genes without NCBI numbers and genes lumped together.\n",
    "3. Remove identical genes from `DAGK` gene reaction rule\n",
    "4. Sort gene reaction components\n",
    "3. Update gene reaction rules accordingly for new genes and identifiers\n",
    "\n",
    "#### Generate gene ID mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a80353b5-0b80-43ac-84ea-3095e3c7780d",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dict = {}\n",
    "manual_updates = {}\n",
    "for gene in model.genes:\n",
    "    ncbigene = gene.annotation.get(\"ncbigene\")\n",
    "    if ncbigene is None:\n",
    "        # NCBI Gene ID is not annotated, will need manual updating\n",
    "        continue\n",
    "    gname, isonum = gene.id.split(\"_\")\n",
    "    id_mapping_dict[gene.id] = f\"{ncbigene}_{isonum}\"\n",
    "\n",
    "# Annotations mapped in Recon3D but not iAB-RBC-283\n",
    "id_mapping_dict[\"ThmtP_AT1\"] = \"79178_AT1\"\n",
    "id_mapping_dict[\"Cgi_14_AT1\"] = \"51005_AT1\"\n",
    "id_mapping_dict[\"Rhag_AT1\"] = \"6005_AT1\"\n",
    "id_mapping_dict[\"Rhbg_AT1\"] = \"57127_AT1\"\n",
    "id_mapping_dict[\"Abcc4_AT1\"] = \"10257_AT1\"\n",
    "id_mapping_dict[\"Slc2a1_AT1\"] = \"6513_AT1\"\n",
    "id_mapping_dict[\"Slc2a2_AT1\"] = \"6514_AT1\"\n",
    "id_mapping_dict[\"Slc2a3_AT1\"] = \"6515_AT1\"\n",
    "id_mapping_dict[\"Slc2a4_AT1\"] = \"6517_AT1\"\n",
    "id_mapping_dict[\"Slc2a5_AT1\"] = \"6518_AT1\"\n",
    "id_mapping_dict[\"Slc2a7_AT1\"] = \"155184_AT1\"\n",
    "id_mapping_dict[\"Slc2a8_AT1\"] = \"29988_AT1\"\n",
    "id_mapping_dict[\"Slc2a11_AT1\"] = \"66035_AT1\"\n",
    "id_mapping_dict[\"Slc4a1_AT1\"] = \"6521_AT1\"\n",
    "id_mapping_dict[\"Slc5a1_AT1\"] = \"6523_AT1\"\n",
    "id_mapping_dict[\"Slc5a3_AT1\"] = \"6526_AT1\"\n",
    "id_mapping_dict[\"Slc5a5_AT1\"] = \"6528_AT1\"\n",
    "id_mapping_dict[\"Slc12a7_AT1\"] = \"10723_AT1\"\n",
    "id_mapping_dict[\"Slc14a1_AT1\"] = \"6563_AT1\"\n",
    "id_mapping_dict[\"Slc29a1_AT1\"] = \"2030_AT1\"\n",
    "id_mapping_dict[\"Slc29a2_AT1\"] = \"3177_AT1\"\n",
    "id_mapping_dict[\"Flj22761_AT1\"] = \"80201_AT1\"\n",
    "\n",
    "# Pseudogene, does not seem to have annotations mapped\n",
    "id_mapping_dict[\"Gkp3_AT1\"] = \"2713_AT1\"\n",
    "\n",
    "# Rename genes using ID map, then add lumped genes to ID map and\n",
    "cobra.manipulation.modify.rename_genes(model, id_mapping_dict)\n",
    "\n",
    "# Two seperate genes lumped into one\n",
    "id_mapping_dict[\"Gucy1A2B2_AT1\"] = \"2977_AT1 and 2974_AT1\"\n",
    "id_mapping_dict[\"Gucy1A2B3_AT1\"] = \"2977_AT1 and 2983_AT1\"\n",
    "id_mapping_dict[\"Gucy1A3B3_AT1\"] = \"2982_AT1 and 2983_AT1\"\n",
    "id_mapping_dict[\"Gucy1A3B2_AT1\"] = \"2982_AT1 and 2974_AT1\"\n",
    "id_mapping_dict[\"NME12_AT1\"] = \"4830_AT1 and 4831_AT1\"\n",
    "id_mapping_dict[\"NME12_AT2\"] = \"4830_AT2 and 4831_AT1\"\n",
    "id_mapping_dict[\"Mat2ab1\"] = \"4144_AT1 and 27430_AT1\"\n",
    "id_mapping_dict[\"Mat2ab2\"] = \"4144_AT1 and 27430_AT2\"\n",
    "id_mapping_dict[\"Atp1a1b1\"] = \"476_AT1 and 481_AT1\"\n",
    "id_mapping_dict[\"Atp1a1b2\"] = \"476_AT1 and 482_AT1\"\n",
    "id_mapping_dict[\"Atp1a1b3\"] = \"476_AT1 and 483_AT1\"\n",
    "id_mapping_dict[\"Atp1a1b4\"] = \"476_AT1 and 23439_AT1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7f4fc68-e1b3-4388-a682-d933e4c3ae3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"geneRetired\", \"genes\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "id_mapping_df[\"genes\"] = id_mapping_df[\"genes\"].str.split(\" and \")\n",
    "id_mapping_df = id_mapping_df.explode(\"genes\")\n",
    "\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/deprecatedIdentifiers/replacedGenes.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"genes\", \"geneRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[\n",
    "        previous_id_mapping_df[\"genes\"] == retiring\n",
    "    ]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        retired_set_of_ids.update(\n",
    "            previously_retired[\"geneRetired\"].apply(split_string).item()\n",
    "        )\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"geneRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/replacedGenes.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "175e6c8f-df31-416c-bd3a-b2a90ea26070",
   "metadata": {},
   "source": [
    "#### Seperate \"Lumped\" genes into individual genes\n",
    "These are reactions containing GPRs that have both \"AND\" and \"OR\" in them. \n",
    "The \"AND\" genes are lumped together, likely due to software compability with early COBRA versions that had difficulty parsing rules."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1e52615-9524-4511-bf11-63a6c8ccd20f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reaction GUACYC: Recon3D top, iAB-RBC-283 bottom\n",
    "# 4881_AT1 or 4882_AT1 or 4882_AT2 or   2984_AT1 or   3000_AT1 or   2986_AT1 or (2977_AT1 and 2974_AT1) or (2977_AT1 and 2983_AT1) or (2982_AT1 and 2974_AT1) or (2982_AT1 and 2983_AT1)\n",
    "# Npr1_AT1 or Npr2_AT1 or Npr2_AT2 or Gucy2C_AT1 or Gucy2D_AT1 or Gucy2F_AT1 or (    Gucy1A2B2_AT1    ) or (    Gucy1A2B3_AT1    ) or (    Gucy1A3B2_AT1    ) or (    Gucy1A3B3_AT1    )\n",
    "reaction = model.reactions.get_by_id(\"GUACYC\")\n",
    "reaction.gene_reaction_rule = \"(2974_AT1 and 2977_AT1) or (2974_AT1 and 2982_AT1) or (2977_AT1 and 2983_AT1) or (2982_AT1 and 2983_AT1) or 2984_AT1 or 2986_AT1 or 3000_AT1 or 4881_AT1 or 4882_AT1 or 4882_AT2\"\n",
    "\n",
    "# (4830_AT1 and 4831_AT1) or (4830_AT2 and 4831_AT1) or 4832_AT1 or 10201_AT1 or 29922_AT1 or 29922_AT2\n",
    "# (      NME12_AT1      ) or (      NME12_AT2      ) or NME3_AT1 or  NME6_AT1 or  NME7_AT1 or  NME7_AT2\n",
    "for reaction in [\"NDPK1\", \"NDPK2\", \"NDPK3\"]:\n",
    "    reaction = model.reactions.get_by_id(reaction)\n",
    "    reaction.gene_reaction_rule = \"(4830_AT1 and 4831_AT1) or (4830_AT2 and 4831_AT1) or 4832_AT1 or 10201_AT1 or 29922_AT1 or 29922_AT2\"\n",
    "\n",
    "#  4143_AT1 or (4144_AT1 and 27430_AT1) or (4144_AT1 and 27430_AT2)\n",
    "# Mat1a_AT1 or (       Mat2ab1        ) or (       Mat2ab2        )\n",
    "reaction = model.reactions.get_by_id(\"METAT\")\n",
    "reaction.gene_reaction_rule = (\n",
    "    \"4143_AT1 or (4144_AT1 and 27430_AT1) or (4144_AT1 and 27430_AT2)\"\n",
    ")\n",
    "\n",
    "# (476_AT1 and 481_AT1) or (476_AT1 and  482_AT1) or (476_AT1 and 483_AT1) or (476_AT1 and 23439_AT1)\n",
    "# (     Atp1a1b1      ) or (     Atp1a1b2       ) or (     Atp1a1b3      ) or (     Atp1a1b4      )\n",
    "reaction = model.reactions.get_by_id(\"NaKt\")\n",
    "reaction.gene_reaction_rule = \"(476_AT1 and 481_AT1) or (476_AT1 and 482_AT1) or (476_AT1 and 483_AT1) or (476_AT1 and 23439_AT1)\"\n",
    "\n",
    "# Remove old lumped genes from model\n",
    "cobra.manipulation.delete.remove_genes(\n",
    "    model,\n",
    "    gene_list=model.genes.query(lambda x: not x.reactions),\n",
    "    remove_reactions=False,\n",
    ")\n",
    "model.genes.query(lambda x: not x.reactions)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3d4ce24-87e2-4501-ac99-717d12aa2b69",
   "metadata": {},
   "source": [
    "#### Fix gene reaction rules with duplicate genes\n",
    "The `DAGK` reactions have an extra gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad7bb3a9-d0dc-48e8-82e2-90a83582543b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Auto fixed\n",
    "for reaction in model.reactions.query(lambda x: x.gene_reaction_rule):\n",
    "    or_split = reaction.gene_reaction_rule.split(\" or \")\n",
    "    if Counter(or_split) != Counter(set(or_split)):\n",
    "        # Identify extra gene\n",
    "        extra = Counter(or_split) - Counter(set(or_split))\n",
    "        if extra:\n",
    "            print(f\"Reaction: {reaction.id}: {extra}\")\n",
    "        for gene, num_extra in extra.items():\n",
    "            reaction.gene_reaction_rule = reaction.gene_reaction_rule.replace(\n",
    "                f\"{gene} or \", \"\", num_extra\n",
    "            )\n",
    "\n",
    "# # DAGK reactions have an extra gene\n",
    "# # Manual fix\n",
    "# model.reactions.get_by_id(\"DAGK_hs_16_0_16_0\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_16_0_18_1\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_16_0_18_2\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_1_18_1\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_1_18_2\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_2_16_0\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\"\n",
    "# model.reactions.get_by_id(\"DAGK_hs_18_2_18_1\").gene_reaction_rule = \"1606_AT1 or 1607_AT1 or 1607_AT2 or 1608_AT1 or 1609_AT1 or 8525_AT1 or 8525_AT2 or 8525_AT3 or 8526_AT1 or 8527_AT1 or 8527_AT2 or 9162_AT1 or 160851_AT1 or 160851_AT2\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7752d2fe-94ea-419a-9643-0d18f86dece4",
   "metadata": {},
   "source": [
    "#### Sort GPRs in reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46689b14-31c1-4580-9b41-838b92fdeb8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "or_gpr_reactions = [\n",
    "    r\n",
    "    for r in model.reactions\n",
    "    if \"or\" in r.gene_reaction_rule and not \"and\" in r.gene_reaction_rule\n",
    "]\n",
    "\n",
    "and_gpr_reactions = [\n",
    "    r\n",
    "    for r in model.reactions\n",
    "    if \"and\" in r.gene_reaction_rule and not \"or\" in r.gene_reaction_rule\n",
    "]\n",
    "\n",
    "for join_str, reactions in zip(\n",
    "    [\" or \", \" and \"], [or_gpr_reactions, and_gpr_reactions]\n",
    "):\n",
    "    for r in reactions:\n",
    "        gpr_dict = defaultdict(list)\n",
    "        for i in r.gene_reaction_rule.split(join_str):\n",
    "            gpr_dict[int(i.split(\"_\")[0])] += [i.split(\"_\")[1]]\n",
    "        genes = r.genes\n",
    "        r.gene_reaction_rule = join_str.join(\n",
    "            [f\"{key}_{v}\" for key in sorted(gpr_dict) for v in sorted(gpr_dict[key])]\n",
    "        )\n",
    "        try:\n",
    "            assert genes == r.genes\n",
    "        except AssertionError:\n",
    "            raise AssertionError(\n",
    "                f\"A change occured in the genes for reaction {r.id} that should not have occured.\"\n",
    "            )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eff1b45f-ba45-4025-84e8-52fc96cfab13",
   "metadata": {},
   "source": [
    "## Update metabolites\n",
    "1. Update metabolites to have the corresponding MetabolicAtlas annotation\n",
    "2. Update BiGG identifiers and retire old ones\n",
    "\n",
    "### Update metabolite annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a15f709-c07e-4a23-9e49-6174da4cfc43",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "for _, row in id_mapping_df.iterrows():\n",
    "    bigg_id, metatlas = row[[\"bigg\", \"metatlas\"]]\n",
    "    for model_metabolite in model.metabolites.query(\n",
    "        lambda x: x.id.replace(f\"_{x.compartment}\", \"\") == bigg_id\n",
    "    ):\n",
    "        model_metabolite.annotation[\"metatlas\"] = (\n",
    "            f\"{metatlas}{model_metabolite.compartment}\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "accb6c13-a4c6-447c-b110-95041bc49cee",
   "metadata": {},
   "source": [
    "### Update metabolite identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1139c66-a3b3-4ad7-a5d5-5987334b9482",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dict = {\n",
    "    # Add L/D stereochemistry to ID\n",
    "    \"adrnl\": \"adrnl__L\",\n",
    "    \"ahcys\": \"ahcys__L\",\n",
    "    \"amet\": \"amet__L\",\n",
    "    \"dhdascb\": \"dhdascb__L\",\n",
    "    \"mepi\": \"mepi__L\",\n",
    "    \"nrpphr\": \"nrpphr__L\",\n",
    "    \"orn\": \"orn__L\",\n",
    "    # ID change to help distinguish L-dopa (3,4-dihydroxyphenylalanine) and L-dopamine, D-Dopa and D-dopamine\n",
    "    \"dopa\": \"dpam__L\",\n",
    "    # ID change to help distinguish the 'gamma' in gamma-L-glutamyl-L-cysteine\n",
    "    \"glucys\": \"gglucys__L\",\n",
    "    # ID change to make it clear that the species is (R)-S-lactoylglutathione, with (R) stereoconfirmation and S-bond\n",
    "    \"lgt__S\": \"slgth__R\",\n",
    "    # ID change to make it clear that the species is UDP bound to a D-glucose\n",
    "    \"udpg\": \"udpglc__D\",\n",
    "    # Increased name specificity for cyclic nucleotides\n",
    "    \"camp\": \"35camp\",\n",
    "    # Lipid ID changes to help distinguish stereochemistry on species and without use of common names.\n",
    "    # Fatty Acids and Conjugates [FA01]\t(FA)\n",
    "    \"hdca\": \"FA_hs_16_0\",\n",
    "    \"ocdcea\": \"FA_hs_18_9Z\",\n",
    "    \"lnlc\": \"FA_hs_18_9Z12Z\",\n",
    "    # Fatty acyl CoAs [FA0705] (CoA)\n",
    "    \"pmtcoa\": \"FAcoa_hs_16_0\",\n",
    "    \"odecoa\": \"FAcoa_hs_18_9Z\",\n",
    "    \"lnlccoa\": \"FAcoa_hs_18_9Z12Z\",\n",
    "    # Fatty acyl carnitines [FA0707] (CAR)\n",
    "    \"pmtcrn\": \"FAcrn_hs_16_0\",\n",
    "    \"odecrn\": \"FAcrn_hs_18_9Z\",\n",
    "    \"lnlccrn\": \"FAcrn_hs_18_9Z12Z\",\n",
    "    # Diacylglycerols [GL0201] (DG)\n",
    "    \"dag_hs_16_0_18_1\": \"dag_hs_16_0_18_9Z\",\n",
    "    \"dag_hs_16_0_18_2\": \"dag_hs_16_0_18_9Z12Z\",\n",
    "    \"dag_hs_18_1_18_1\": \"dag_hs_18_9Z_18_9Z\",\n",
    "    \"dag_hs_18_1_18_2\": \"dag_hs_18_9Z_18_9Z12Z\",\n",
    "    \"dag_hs_18_2_16_0\": \"dag_hs_18_9Z12Z_16_0\",\n",
    "    \"dag_hs_18_2_18_1\": \"dag_hs_18_9Z12Z_18_9Z\",\n",
    "    # Diacylglycerophosphates [GP1001] (PA)\n",
    "    \"pa_hs_16_0_18_1\": \"pa_hs_16_0_18_9Z\",\n",
    "    \"pa_hs_16_0_18_2\": \"pa_hs_16_0_18_9Z12Z\",\n",
    "    \"pa_hs_18_1_18_1\": \"pa_hs_18_9Z_18_9Z\",\n",
    "    \"pa_hs_18_1_18_2\": \"pa_hs_18_9Z_18_9Z12Z\",\n",
    "    \"pa_hs_18_2_16_0\": \"pa_hs_18_9Z12Z_16_0\",\n",
    "    \"pa_hs_18_2_18_1\": \"pa_hs_18_9Z12Z_18_9Z\",\n",
    "    # Lysophospholipids\tPrefix L\n",
    "    \"alpa_hs_16_0\": \"lpa_hs_16_0\",\n",
    "    \"alpa_hs_18_1\": \"lpa_hs_18_9Z\",\n",
    "    \"alpa_hs_18_2\": \"lpa_hs_18_9Z12Z\",\n",
    "    # Diacylglycerophosphocholines [GP0101] PC\n",
    "    \"pchol_hs_16_0_16_0\": \"pc_hs_16_0_16_0\",\n",
    "    \"pchol_hs_16_0_18_1\": \"pc_hs_16_0_18_9Z\",\n",
    "    \"pchol_hs_16_0_18_2\": \"pc_hs_16_0_18_9Z12Z\",\n",
    "    \"pchol_hs_18_1_18_1\": \"pc_hs_18_9Z_18_9Z\",\n",
    "    \"pchol_hs_18_1_18_2\": \"pc_hs_18_9Z_18_9Z12Z\",\n",
    "    \"pchol_hs_18_2_16_0\": \"pc_hs_18_9Z12Z_16_0\",\n",
    "    \"pchol_hs_18_2_18_1\": \"pc_hs_18_9Z12Z_18_9Z\",\n",
    "    # Lysophospholipids\tPrefix L\n",
    "    \"lpchol_hs_16_0\": \"lpc_hs_16_0\",\n",
    "    \"lpchol_hs_18_1\": \"lpc_hs_18_9Z\",\n",
    "    \"lpchol_hs_18_2\": \"lpc_hs_18_9Z12Z\",\n",
    "    # Diacylglycerophosphoethanolamines [GP0201] PE\n",
    "    \"pe_hs_16_0_18_1\": \"pe_hs_16_0_18_9Z\",\n",
    "    \"pe_hs_16_0_18_2\": \"pe_hs_16_0_18_9Z12Z\",\n",
    "    \"pe_hs_18_1_18_1\": \"pe_hs_18_9Z_18_9Z\",\n",
    "    \"pe_hs_18_1_18_2\": \"pe_hs_18_9Z_18_9Z12Z\",\n",
    "    \"pe_hs_18_2_16_0\": \"pe_hs_18_9Z12Z_16_0\",\n",
    "    \"pe_hs_18_2_18_1\": \"pe_hs_18_9Z12Z_18_9Z\",\n",
    "    # Diacylglycerophosphoinositols [GP0601] PI\n",
    "    \"pail_hs_16_0_18_1\": \"pail_hs_16_0_18_9Z\",\n",
    "    \"pail_hs_16_0_18_2\": \"pail_hs_16_0_18_9Z12Z\",\n",
    "    \"pail_hs_18_1_18_1\": \"pail_hs_18_9Z_18_9Z\",\n",
    "    \"pail_hs_18_1_18_2\": \"pail_hs_18_9Z_18_9Z12Z\",\n",
    "    \"pail_hs_18_2_16_0\": \"pail_hs_18_9Z12Z_16_0\",\n",
    "    \"pail_hs_18_2_18_1\": \"pail_hs_18_9Z12Z_18_9Z\",\n",
    "    # Diacylglycerophosphoinositol monophosphates [GP0701] PIP[4’]\n",
    "    \"pail4p_hs_16_0_18_1\": \"pail4p_hs_16_0_18_9Z\",\n",
    "    \"pail4p_hs_16_0_18_2\": \"pail4p_hs_16_0_18_9Z12Z\",\n",
    "    \"pail4p_hs_18_1_18_1\": \"pail4p_hs_18_9Z_18_9Z\",\n",
    "    \"pail4p_hs_18_1_18_2\": \"pail4p_hs_18_9Z_18_9Z12Z\",\n",
    "    \"pail4p_hs_18_2_16_0\": \"pail4p_hs_18_9Z12Z_16_0\",\n",
    "    \"pail4p_hs_18_2_18_1\": \"pail4p_hs_18_9Z12Z_18_9Z\",\n",
    "    # Diacylglycerophosphoinositol bisphosphates [GP0801] PIP2[4’,5’]\n",
    "    \"pail45p_hs_16_0_18_1\": \"pail45p_hs_16_0_18_9Z\",\n",
    "    \"pail45p_hs_16_0_18_2\": \"pail45p_hs_16_0_18_9Z12Z\",\n",
    "    \"pail45p_hs_18_1_18_1\": \"pail45p_hs_18_9Z_18_9Z\",\n",
    "    \"pail45p_hs_18_1_18_2\": \"pail45p_hs_18_9Z_18_9Z12Z\",\n",
    "    \"pail45p_hs_18_2_16_0\": \"pail45p_hs_18_9Z12Z_16_0\",\n",
    "    \"pail45p_hs_18_2_18_1\": \"pail45p_hs_18_9Z12Z_18_9Z\",\n",
    "    # CDP-diacylglycerols [GP1301] CDP-DG\n",
    "    \"cdpdag_hs_16_0_18_1\": \"cdpdag_hs_16_0_18_9Z\",\n",
    "    \"cdpdag_hs_16_0_18_2\": \"cdpdag_hs_16_0_18_9Z12Z\",\n",
    "    \"cdpdag_hs_18_1_18_1\": \"cdpdag_hs_18_9Z_18_9Z\",\n",
    "    \"cdpdag_hs_18_1_18_2\": \"cdpdag_hs_18_9Z_18_9Z12Z\",\n",
    "    \"cdpdag_hs_18_2_16_0\": \"cdpdag_hs_18_9Z12Z_16_0\",\n",
    "    \"cdpdag_hs_18_2_18_1\": \"cdpdag_hs_18_9Z12Z_18_9Z\",\n",
    "}\n",
    "for old_met, new_met in id_mapping_dict.items():\n",
    "    for model_metabolite in model.metabolites.query(\n",
    "        lambda x: x.id.replace(f\"_{x.compartment}\", \"\") == old_met\n",
    "    ):\n",
    "        model_metabolite.id = f\"{new_met}_{model_metabolite.compartment}\"\n",
    "model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd99f29-d8e1-46fb-9037-05f01e23cbbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"metRetired\", \"mets\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"mets\", \"metRetired\"]]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    id_mapping_df.loc[idx, \"metRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/replacedMetabolites.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c281ba7-924b-4a16-bfff-c88d3cdf7251",
   "metadata": {},
   "source": [
    "## Update reactions\n",
    "1. Update reactions to have the corresponding MetabolicAtlas annotation\n",
    "2. Change the NADH demand reaction annoation to prevent false identification as a boundary reaction.\n",
    "3. Update BiGG identifiers and retire old ones\n",
    "### Update reaction annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c218e64-cdbc-43a4-b353-3a479ce07371",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/interim/BiGGMetAtlasReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "for _, row in id_mapping_df.iterrows():\n",
    "    bigg_id, metatlas = row[[\"bigg\", \"metatlas\"]]\n",
    "    reaction = model.reactions.get_by_id(bigg_id)\n",
    "    reaction.annotation[\"metatlas\"] = metatlas\n",
    "\n",
    "reaction = model.reactions.get_by_id(\"DM_nadh\")\n",
    "reaction.annotation[\"sbo\"] = \"SBO:0000176\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6666c72-5623-4e14-86d2-cd3492dfac72",
   "metadata": {},
   "source": [
    "### Update reaction identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6af1928-a0dd-4b9e-b646-7ca43c95949d",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dict = {\n",
    "    # Change NADH demand to NADHload\n",
    "    \"DM_nadh\": \"NADHload\",\n",
    "    # Updated to reflect stoichiometry of 2:1\n",
    "    \"CAATPS\": \"CAATPS2\",\n",
    "    # Addition of `x` or `y` to indiciate NADH and NADPH cofactors\n",
    "    \"BILIRED\": \"BILIREDy\",\n",
    "    \"GTHOr\": \"GTHOy\",\n",
    "    \"XYLTD_D\": \"XYLTD_Dx\",\n",
    "    \"XYLUR\": \"XYLUR_Ly\",\n",
    "    # Updated for increased specificity\n",
    "    \"ACP1_FMN\": \"FMNPH\",\n",
    "    \"DOPAMT\": \"LDPAMMT\",\n",
    "    \"GGLUCT\": \"GGLUCTC\",\n",
    "    \"MI1345PP\": \"MI1345P5PP\",\n",
    "    \"MI145PK\": \"MI145P3K\",\n",
    "    \"MI145PP\": \"MI145P5PP\",\n",
    "    \"NP1\": \"PNP2\",\n",
    "    \"NT5C\": \"NICHYD\",\n",
    "    \"PDE1\": \"PDEA\",\n",
    "    \"UMPK\": \"UMPK1\",\n",
    "    \"UPP3S\": \"UPPG3S\",\n",
    "    \"UPPDC1\": \"UPPG3DC\",\n",
    "    \"4PYRDX\": \"4PYRDXABCte\",\n",
    "    \"CAMPt\": \"CAMPABCte\",\n",
    "    \"CGMPt\": \"CGMPABCte\",\n",
    "    \"GTHOXti2\": \"GTHOXABCte\",\n",
    "    \"RIBFLVt3o\": \"RIBFLVABCte\",\n",
    "    \"HCO3_CLt\": \"HCO3_Cltex\",\n",
    "    \"ADRNLtu\": \"ADRNLt\",\n",
    "    \"ACt2r\": \"ACt2\",\n",
    "    \"GLYt7_211_r\": \"GLY_Cl_2Nat\",\n",
    "    \"L_LACt2r\": \"L_LACt2\",\n",
    "    \"NCAMUP\": \"NCAMt\",\n",
    "    \"NRPPHRtu\": \"NRPPHRt\",\n",
    "    \"DOPAtu\": \"LDPAMt\",\n",
    "    # Removal of incorrect/unclear suffixes\n",
    "    \"AHCi\": \"AHC\",\n",
    "    \"G6PDH2r\": \"G6PDH2\",\n",
    "    \"GALKr\": \"GALK\",\n",
    "    \"GALUi\": \"GALU\",\n",
    "    \"GTHPi\": \"GTHP\",\n",
    "    \"ICDHyr\": \"ICDHy\",\n",
    "    \"NNATr\": \"NNAT\",\n",
    "    \"PDX5POi\": \"PDX5PO\",\n",
    "    \"ARGt5r\": \"ARGtec\",\n",
    "    \"SPRMt2i\": \"SPRMt2\",\n",
    "    \"THMtrbc\": \"THMt\",\n",
    "    # Changed for preparation of additional lipids\n",
    "    \"C160CPT1\": \"CRNAT_16_0\",\n",
    "    \"C160CPT2rbc\": \"CRNAT_16_0rbc\",\n",
    "    \"C181CPT1\": \"CRNAT_18_9Z\",\n",
    "    \"C181CPT2rbc\": \"CRNAT_18_9Zrbc\",\n",
    "    \"LNLCCPT1\": \"CRNAT_18_9Z12Z\",\n",
    "    \"LNLCCPT2rbc\": \"CRNAT_18_9Z12Zrbc\",\n",
    "    \"FACOAL160\": \"FACOAL_16_0\",\n",
    "    \"FACOAL181\": \"FACOAL_18_9Z\",\n",
    "    \"FACOAL1821\": \"FACOAL_18_9Z12Z\",\n",
    "    \"HDCAt\": \"FAt_16_0\",\n",
    "    \"OCDCEAt\": \"FAt_18_9Z\",\n",
    "    \"LNLCt\": \"FAt_18_9Z12Z\",\n",
    "    # Lysophospatidate (LPA) acyl transferase\n",
    "    \"AGPAT1_16_0_16_0\": \"LPAAT_16_0_16_0\",\n",
    "    \"AGPAT1_16_0_18_1\": \"LPAAT_16_0_18_9Z\",\n",
    "    \"AGPAT1_16_0_18_2\": \"LPAAT_16_0_18_9Z12Z\",\n",
    "    \"AGPAT1_18_1_18_1\": \"LPAAT_18_9Z_18_9Z\",\n",
    "    \"AGPAT1_18_1_18_2\": \"LPAAT_18_9Z_18_9Z12Z\",\n",
    "    \"AGPAT1_18_2_16_0\": \"LPAAT_18_9Z12Z_16_0\",\n",
    "    \"AGPAT1_18_2_18_1\": \"LPAAT_18_9Z12Z_18_9Z\",\n",
    "    # Removal of \"r\"\n",
    "    \"CDIPTr_16_0_16_0\": \"CDIPT_16_0_16_0\",\n",
    "    \"CDIPTr_16_0_18_1\": \"CDIPT_16_0_18_9Z\",\n",
    "    \"CDIPTr_16_0_18_2\": \"CDIPT_16_0_18_9Z12Z\",\n",
    "    \"CDIPTr_18_1_18_1\": \"CDIPT_18_9Z_18_9Z\",\n",
    "    \"CDIPTr_18_1_18_2\": \"CDIPT_18_9Z_18_9Z12Z\",\n",
    "    \"CDIPTr_18_2_16_0\": \"CDIPT_18_9Z12Z_16_0\",\n",
    "    \"CDIPTr_18_2_18_1\": \"CDIPT_18_9Z12Z_18_9Z\",\n",
    "    \"CDS_16_0_18_1\": \"CDS_16_0_18_9Z\",\n",
    "    \"CDS_16_0_18_2\": \"CDS_16_0_18_9Z12Z\",\n",
    "    \"CDS_18_1_18_1\": \"CDS_18_9Z_18_9Z\",\n",
    "    \"CDS_18_1_18_2\": \"CDS_18_9Z_18_9Z12Z\",\n",
    "    \"CDS_18_2_16_0\": \"CDS_18_9Z12Z_16_0\",\n",
    "    \"CDS_18_2_18_1\": \"CDS_18_9Z12Z_18_9Z\",\n",
    "    \"CEPTC_16_0_18_1\": \"CEPTC_16_0_18_9Z\",\n",
    "    \"CEPTC_16_0_18_2\": \"CEPTC_16_0_18_9Z12Z\",\n",
    "    \"CEPTC_18_1_18_1\": \"CEPTC_18_9Z_18_9Z\",\n",
    "    \"CEPTC_18_1_18_2\": \"CEPTC_18_9Z_18_9Z12Z\",\n",
    "    \"CEPTC_18_2_16_0\": \"CEPTC_18_9Z12Z_16_0\",\n",
    "    \"CEPTC_18_2_18_1\": \"CEPTC_18_9Z12Z_18_9Z\",\n",
    "    \"CEPTE_16_0_18_1\": \"CEPTE_16_0_18_9Z\",\n",
    "    \"CEPTE_16_0_18_2\": \"CEPTE_16_0_18_9Z12Z\",\n",
    "    \"CEPTE_18_1_18_1\": \"CEPTE_18_9Z_18_9Z\",\n",
    "    \"CEPTE_18_1_18_2\": \"CEPTE_18_9Z_18_9Z12Z\",\n",
    "    \"CEPTE_18_2_16_0\": \"CEPTE_18_9Z12Z_16_0\",\n",
    "    \"CEPTE_18_2_18_1\": \"CEPTE_18_9Z12Z_18_9Z\",\n",
    "    \"DAGK_hs_16_0_16_0\": \"DAGK_16_0_16_0\",\n",
    "    \"DAGK_hs_16_0_18_1\": \"DAGK_16_0_18_9Z\",\n",
    "    \"DAGK_hs_16_0_18_2\": \"DAGK_16_0_18_9Z12Z\",\n",
    "    \"DAGK_hs_18_1_18_1\": \"DAGK_18_9Z_18_9Z\",\n",
    "    \"DAGK_hs_18_1_18_2\": \"DAGK_18_9Z_18_9Z12Z\",\n",
    "    \"DAGK_hs_18_2_16_0\": \"DAGK_18_9Z12Z_16_0\",\n",
    "    \"DAGK_hs_18_2_18_1\": \"DAGK_18_9Z12Z_18_9Z\",\n",
    "    # Glycerol-3-phosphate acyltransferase (GPAT)\n",
    "    \"GPAM_hs_16_0\": \"GPAT_16_0\",\n",
    "    \"GPAM_hs_18_1\": \"GPAT_18_9Z\",\n",
    "    \"GPAM_hs_18_2\": \"GPAT_18_9Z12Z\",\n",
    "    # Phosphpolipase B acting on the LPC\n",
    "    \"LPASE_16_0\": \"LPCLPLB_16_0\",\n",
    "    \"LPASE_18_1\": \"LPCLPLB_18_9Z\",\n",
    "    \"LPASE_18_2\": \"LPCLPLB_18_9Z12Z\",\n",
    "    \"PI45P5P_16_0_18_1\": \"PI45P5P_16_0_18_9Z\",\n",
    "    \"PI45P5P_16_0_18_2\": \"PI45P5P_16_0_18_9Z12Z\",\n",
    "    \"PI45P5P_18_1_18_1\": \"PI45P5P_18_9Z_18_9Z\",\n",
    "    \"PI45P5P_18_1_18_2\": \"PI45P5P_18_9Z_18_9Z12Z\",\n",
    "    \"PI45P5P_18_2_16_0\": \"PI45P5P_18_9Z12Z_16_0\",\n",
    "    \"PI45P5P_18_2_18_1\": \"PI45P5P_18_9Z12Z_18_9Z\",\n",
    "    \"PI45PLC_16_0_18_1\": \"PI45PLC_16_0_18_9Z\",\n",
    "    \"PI45PLC_16_0_18_2\": \"PI45PLC_16_0_18_9Z12Z\",\n",
    "    \"PI45PLC_18_1_18_1\": \"PI45PLC_18_9Z_18_9Z\",\n",
    "    \"PI45PLC_18_1_18_2\": \"PI45PLC_18_9Z_18_9Z12Z\",\n",
    "    \"PI45PLC_18_2_16_0\": \"PI45PLC_18_9Z12Z_16_0\",\n",
    "    \"PI45PLC_18_2_18_1\": \"PI45PLC_18_9Z12Z_18_9Z\",\n",
    "    \"PI4P5K_16_0_18_1\": \"PI4P5K_16_0_18_9Z\",\n",
    "    \"PI4P5K_16_0_18_2\": \"PI4P5K_16_0_18_9Z12Z\",\n",
    "    \"PI4P5K_18_1_18_1\": \"PI4P5K_18_9Z_18_9Z\",\n",
    "    \"PI4P5K_18_1_18_2\": \"PI4P5K_18_9Z_18_9Z12Z\",\n",
    "    \"PI4P5K_18_2_16_0\": \"PI4P5K_18_9Z12Z_16_0\",\n",
    "    \"PI4P5K_18_2_18_1\": \"PI4P5K_18_9Z12Z_18_9Z\",\n",
    "    \"PI4PLC_16_0_18_1\": \"PI4PLC_16_0_18_9Z\",\n",
    "    \"PI4PLC_16_0_18_2\": \"PI4PLC_16_0_18_9Z12Z\",\n",
    "    \"PI4PLC_18_1_18_1\": \"PI4PLC_18_9Z_18_9Z\",\n",
    "    \"PI4PLC_18_1_18_2\": \"PI4PLC_18_9Z_18_9Z12Z\",\n",
    "    \"PI4PLC_18_2_16_0\": \"PI4PLC_18_9Z12Z_16_0\",\n",
    "    \"PI4PLC_18_2_18_1\": \"PI4PLC_18_9Z12Z_18_9Z\",\n",
    "    \"PI4PP_16_0_18_1\": \"PI4PP_16_0_18_9Z\",\n",
    "    \"PI4PP_16_0_18_2\": \"PI4PP_16_0_18_9Z12Z\",\n",
    "    \"PI4PP_18_1_18_1\": \"PI4PP_18_9Z_18_9Z\",\n",
    "    \"PI4PP_18_1_18_2\": \"PI4PP_18_9Z_18_9Z12Z\",\n",
    "    \"PI4PP_18_2_16_0\": \"PI4PP_18_9Z12Z_16_0\",\n",
    "    \"PI4PP_18_2_18_1\": \"PI4PP_18_9Z12Z_18_9Z\",\n",
    "    # PI 4-kinase\n",
    "    \"PIK4_16_0_16_0\": \"PI4K_16_0_16_0\",\n",
    "    \"PIK4_16_0_18_1\": \"PI4K_16_0_18_9Z\",\n",
    "    \"PIK4_16_0_18_2\": \"PI4K_16_0_18_9Z12Z\",\n",
    "    \"PIK4_18_1_18_1\": \"PI4K_18_9Z_18_9Z\",\n",
    "    \"PIK4_18_1_18_2\": \"PI4K_18_9Z_18_9Z12Z\",\n",
    "    \"PIK4_18_2_16_0\": \"PI4K_18_9Z12Z_16_0\",\n",
    "    \"PIK4_18_2_18_1\": \"PI4K_18_9Z12Z_18_9Z\",\n",
    "    \"PIPLC_16_0_18_1\": \"PIPLC_16_0_18_9Z\",\n",
    "    \"PIPLC_16_0_18_2\": \"PIPLC_16_0_18_9Z12Z\",\n",
    "    \"PIPLC_18_1_18_1\": \"PIPLC_18_9Z_18_9Z\",\n",
    "    \"PIPLC_18_1_18_2\": \"PIPLC_18_9Z_18_9Z12Z\",\n",
    "    \"PIPLC_18_2_16_0\": \"PIPLC_18_9Z12Z_16_0\",\n",
    "    \"PIPLC_18_2_18_1\": \"PIPLC_18_9Z12Z_18_9Z\",\n",
    "    # PC Phospholipase A2\n",
    "    \"PLA2_2_16_0_16_0\": \"PCPLA2_16_0_16_0\",\n",
    "    \"PLA2_2_16_0_18_1\": \"PCPLA2_16_0_18_9Z\",\n",
    "    \"PLA2_2_16_0_18_2\": \"PCPLA2_16_0_18_9Z12Z\",\n",
    "    \"PLA2_2_18_1_18_1\": \"PCPLA2_18_9Z_18_9Z\",\n",
    "    \"PLA2_2_18_1_18_2\": \"PCPLA2_18_9Z_18_9Z12Z\",\n",
    "    \"PLA2_2_18_2_16_0\": \"PCPLA2_18_9Z12Z_16_0\",\n",
    "    \"PLA2_2_18_2_18_1\": \"PCPLA2_18_9Z12Z_18_9Z\",\n",
    "    # Phosphatidate (PA) phosphatase\n",
    "    \"PPAP_16_0_16_0\": \"PAPP_16_0_16_0\",\n",
    "    \"PPAP_16_0_18_1\": \"PAPP_16_0_18_9Z\",\n",
    "    \"PPAP_16_0_18_2\": \"PAPP_16_0_18_9Z12Z\",\n",
    "    \"PPAP_18_1_18_1\": \"PAPP_18_9Z_18_9Z\",\n",
    "    \"PPAP_18_1_18_2\": \"PAPP_18_9Z_18_9Z12Z\",\n",
    "    \"PPAP_18_2_16_0\": \"PAPP_18_9Z12Z_16_0\",\n",
    "    \"PPAP_18_2_18_1\": \"PAPP_18_9Z12Z_18_9Z\",\n",
    "    # To match metabolite changes\n",
    "    \"SK_pchol_hs_16_0_16_0_c\": \"SK_pc_hs_16_0_16_0_c\",\n",
    "    \"SK_pchol_hs_16_0_18_1_c\": \"SK_pc_hs_16_0_18_9Z_c\",\n",
    "    \"SK_pchol_hs_16_0_18_2_c\": \"SK_pc_hs_16_0_18_9Z12Z_c\",\n",
    "    \"SK_pchol_hs_18_1_18_1_c\": \"SK_pc_hs_18_9Z_18_9Z_c\",\n",
    "    \"SK_pchol_hs_18_1_18_2_c\": \"SK_pc_hs_18_9Z_18_9Z12Z_c\",\n",
    "    \"SK_pchol_hs_18_2_16_0_c\": \"SK_pc_hs_18_9Z12Z_16_0_c\",\n",
    "    \"SK_pchol_hs_18_2_18_1_c\": \"SK_pc_hs_18_9Z12Z_18_9Z_c\",\n",
    "    \"SK_pe_hs_16_0_16_0_c\": \"SK_pe_hs_16_0_16_0_c\",\n",
    "    \"SK_pe_hs_16_0_18_1_c\": \"SK_pe_hs_16_0_18_9Z_c\",\n",
    "    \"SK_pe_hs_16_0_18_2_c\": \"SK_pe_hs_16_0_18_9Z12Z_c\",\n",
    "    \"SK_pe_hs_18_1_18_1_c\": \"SK_pe_hs_18_9Z_18_9Z_c\",\n",
    "    \"SK_pe_hs_18_1_18_2_c\": \"SK_pe_hs_18_9Z_18_9Z12Z_c\",\n",
    "    \"SK_pe_hs_18_2_16_0_c\": \"SK_pe_hs_18_9Z12Z_16_0_c\",\n",
    "    \"SK_pe_hs_18_2_18_1_c\": \"SK_pe_hs_18_9Z12Z_18_9Z_c\",\n",
    "    # Exchanges\n",
    "    \"EX_adrnl_e\": \"EX_adrnl__L_e\",\n",
    "    \"EX_camp_e\": \"EX_35camp_e\",\n",
    "    \"EX_dhdascb_e\": \"EX_dhdascb__L_e\",\n",
    "    \"EX_dopa_e\": \"EX_dpam__L_e\",\n",
    "    \"EX_nrpphr_e\": \"EX_nrpphr__L_e\",\n",
    "    \"EX_mepi_e\": \"EX_mepi__L_e\",\n",
    "    \"EX_hdca_e\": \"EX_FA_hs_16_0_e\",\n",
    "    \"EX_ocdcea_e\": \"EX_FA_hs_18_9Z_e\",\n",
    "    \"EX_lnlc_e\": \"EX_FA_hs_18_9Z12Z_e\",\n",
    "    # To match naming conventions, intracellular demands are irreversible while sinks are reversible.\n",
    "    \"SK_adprbp_c\": \"DM_adprbp_c\",\n",
    "    \"SK_mi1345p_c\": \"DM_mi1345p_c\",\n",
    "    \"SK_mi134p_c\": \"DM_mi134p_c\",\n",
    "    \"SK_mi145p_c\": \"DM_mi145p_c\",\n",
    "    \"SK_mi14p_c\": \"DM_mi14p_c\",\n",
    "}\n",
    "\n",
    "for old_rxn, new_rxn in id_mapping_dict.items():\n",
    "    for model_reaction in model.reactions.query(lambda x: x.id == old_rxn):\n",
    "        model_reaction.id = new_rxn\n",
    "model.repair()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0486ab5-0525-48a0-a854-998d27dd2bc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"rxnRetired\", \"rxns\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/deprecatedIdentifiers/replacedReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"rxns\", \"rxnRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[\n",
    "        previous_id_mapping_df[\"rxns\"] == retiring\n",
    "    ]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        retired_set_of_ids.update(\n",
    "            previously_retired[\"rxnRetired\"].apply(split_string).item()\n",
    "        )\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"rxnRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "\n",
    "id_mapping_df.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/replacedReactions.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10101aba-02a3-4e30-a805-b7e7b0e751f9",
   "metadata": {},
   "source": [
    "#### Update boundary identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf3e6e40-6f22-4be0-9415-76afbc8bbcae",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_bounds = pd.read_csv(\n",
    "    f\"{ROOT_PATH}/data/parameterization/iAB-RBC-283_Bordbar2011.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    ").fillna(\"\")\n",
    "\n",
    "for idx, id_str in enumerate(df_bounds[\"ID\"].values):\n",
    "    if id_str in id_mapping_dict:\n",
    "        df_bounds.loc[idx, \"ID\"] = id_mapping_dict[id_str]\n",
    "\n",
    "df_bounds.to_csv(\n",
    "    f\"{ROOT_PATH}/data/processed/iAB-RBC-283_Bordbar2011.tsv\",\n",
    "    sep=\"\\t\",\n",
    ")\n",
    "df_bounds"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "e96a25b6-7c0f-425a-abf0-53c8a5f52154",
   "metadata": {},
   "source": [
    "#### Subsystem updates\n",
    "\n",
    "All subsystems are updated to the corresponding group in HumanGEM.\n",
    "\n",
    "Notable exceptions are as follows:\n",
    "* \"ROS detoxification\" is now \"Reactive species formation and detoxification\"\n",
    "* \"Urea Cycle\" is combined with \"Arginine and proline metabolism\"\n",
    "* \"Oxidative Phosphorylation\" contains only `PPA` reaction, which is also the only cytosolic reaction in the HumanGEM subsystem. As reaction could be placed into a variety of categories, it was placed in the \"Nucleotide metabolism\" subsystem in which inorganic diphosphate is generated.\n",
    "* Subsystems that contains \"Biosynthesis\" were replaced with the more generic term \"metabolism\", better reflecting RBCs. Includes the following:\n",
    "    * \"Phenylalanine, tyrosine and tryptophan biosynthesis\"\n",
    "* \"Tyrosine metabolism\" was grouped into \"Phenylalanine, tyrosine and tryptophan metabolism\".\n",
    "* \"Eicosanoid Metabolism contains only `LTA4H` reaction, which becomes part of the `Leukotriene` subsystem in HumanGEM.\n",
    "* 'Intracellular source/sink' Extracellular exchange are grouped under Pseudoreactions rather than of `Exchange/demand reactions`, "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76db05c7-160e-4176-ad9b-387a1813af79",
   "metadata": {},
   "outputs": [],
   "source": [
    "subsystem_mapping = {\n",
    "    # Special\n",
    "    \"Oxidative Phosphorylation\": \"Nucleotide metabolism\",\n",
    "    \"Diacylglycerol Synthesis\": \"Glycerolipid metabolism\",\n",
    "    \"ROS Detoxification\": \"Reactive species formation and detoxification\",\n",
    "    \"Eicosanoid Metabolism\": \"Leukotriene metabolism\",\n",
    "    \"Intracellular source/sink\": \"Pseudoreactions\",\n",
    "    \"Extracellular exchange\": \"Pseudoreactions\",\n",
    "    # Straightforward\n",
    "    ## Carbohydrate\n",
    "    \"Glycolysis/Gluconeogenesis\": \"Glycolysis / Gluconeogenesis\",\n",
    "    \"Galactose metabolism\": \"Galactose metabolism\",\n",
    "    \"Inositol Phosphate Metabolism\": \"Inositol phosphate metabolism\",\n",
    "    \"Ascorbate and Aldarate Metabolism\": \"Ascorbate and aldarate metabolism\",\n",
    "    \"Transport, Extracellular\": \"Transport reactions\",\n",
    "    \"Vitamin B6 Metabolism\": \"Vitamin B6 metabolism\",\n",
    "    \"Urea cycle/amino group metabolism\": \"Arginine and proline metabolism\",\n",
    "    \"Porphyrin metabolism\": \"Porphyrin metabolism\",\n",
    "    \"Methionine Salvage\": \"Cysteine and methionine metabolism\",\n",
    "    \"Thiamine Metabolism\": \"Thiamine metabolism\",\n",
    "    \"Miscellaneous\": \"Miscellaneous\",\n",
    "    \"Pentose Phosphate Pathway\": \"Pentose phosphate pathway\",\n",
    "    \"Riboflavin Metabolism\": \"Riboflavin metabolism\",\n",
    "    \"Arginine and Proline Metabolism\": \"Arginine and proline metabolism\",\n",
    "    \"Glutamate metabolism\": \"Alanine, aspartate and glutamate metabolism\",\n",
    "    \"Salvage Pathway\": \"Purine metabolism\",\n",
    "    \"Phenylalanine metabolism\": \"Phenylalanine, tyrosine and tryptophan metabolism\",\n",
    "    \"Tyrosine metabolism\": \"Phenylalanine, tyrosine and tryptophan metabolism\",\n",
    "    \"Nucleotides\": \"Nucleotide metabolism\",\n",
    "    \"Heme Degradation\": \"Porphyrin metabolism\",\n",
    "    \"Heme Biosynthesis\": \"Porphyrin metabolism\",\n",
    "    \"Fructose and Mannose Metabolism\": \"Fructose and mannose metabolism\",\n",
    "    \"Purine Catabolism\": \"Purine metabolism\",\n",
    "    \"Aminosugar Metabolism\": \"Amino sugar and nucleotide sugar metabolism\",\n",
    "    \"Glycerophospholipid Metabolism\": \"Glycerophospholipid metabolism\",\n",
    "    \"Pyrimidine Catabolism\": \"Pyrimidine metabolism\",\n",
    "    \"Citric Acid Cycle\": \"Tricarboxylic acid cycle and glyoxylate/dicarboxylate metabolism\",\n",
    "    \"Pyrimidine Biosynthesis\": \"Pyrimidine metabolism\",\n",
    "    \"Pyruvate Metabolism\": \"Pyruvate metabolism\",\n",
    "    \"NAD Metabolism\": \"Nicotinate and nicotinamide metabolism\",\n",
    "    \"Pentose and Glucuronate Interconversions\": \"Pentose and glucuronate interconversions\",\n",
    "    \"Carnitine shuttle\": \"Carnitine shuttle (cytosolic)\",\n",
    "    \"Glutathione Metabolism\": \"Glutathione metabolism\",\n",
    "    \"Methionine Metabolism\": \"Cysteine and methionine metabolism\",\n",
    "    \"Fatty acid activation\": \"Fatty acid activation (cytosolic)\",\n",
    "    \"Starch and Sucrose Metabolism\": \"Starch and sucrose metabolism\",\n",
    "}\n",
    "\n",
    "individual_subsystem_updates = {\n",
    "    \"Arginine and proline metabolism\": [\"MTAP\"],\n",
    "    \"Purine metabolism\": [\n",
    "        \"ADK1\",\n",
    "        \"ADNCYC\",\n",
    "        \"ADNK1\",\n",
    "        \"AMPDA\",\n",
    "        \"PDEA\",\n",
    "        \"GUACYC\",\n",
    "        \"NDPK1\",\n",
    "        \"IMPD\",\n",
    "        \"GK1\",\n",
    "        \"NTD7\",\n",
    "        \"NTD9\",\n",
    "        \"GMPR\",\n",
    "        \"GMPS2\",\n",
    "    ],\n",
    "    \"Pyrimidine metabolism\": [\"UMPK1\", \"NDPK2\", \"NDPK3\"],\n",
    "    \"Pyruvate metabolism\": [\"ALDD2x\"],\n",
    "    \"Nucleotide metabolism\": [\"AP4AH1\"],\n",
    "    \"Miscellaneous\": [\"BANDMT\"],\n",
    "}\n",
    "for subsystem, reaction_id_list in individual_subsystem_updates.items():\n",
    "    reaction_list = model.reactions.get_by_any(reaction_id_list)\n",
    "    for reaction in reaction_list:\n",
    "        reaction.subsystem = subsystem\n",
    "\n",
    "model.remove_groups(list(model.groups))\n",
    "for old_subsystem, new_subsystem in subsystem_mapping.items():\n",
    "    reaction_list = model.reactions.query(lambda x: x.subsystem == old_subsystem)\n",
    "    for reaction in reaction_list:\n",
    "        reaction.subsystem = new_subsystem\n",
    "    if new_subsystem not in model.groups:\n",
    "        group = Group(id=new_subsystem, name=new_subsystem, members=reaction_list)\n",
    "        model.add_groups([group])\n",
    "    else:\n",
    "        group = model.groups.get_by_id(new_subsystem).add_members(reaction_list)\n",
    "\n",
    "for subsystem, reaction_id_list in individual_subsystem_updates.items():\n",
    "    reaction_list = model.reactions.get_by_any(reaction_id_list)\n",
    "    if subsystem not in model.groups:\n",
    "        group = Group(id=subsystem, name=subsystem, members=reaction_list)\n",
    "        model.add_groups([group])\n",
    "    else:\n",
    "        group = model.groups.get_by_id(subsystem).add_members(reaction_list)\n",
    "\n",
    "model.groups.sort()\n",
    "model.groups"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "222f5a0d-f7c8-4a62-85c4-04ea5c52d499",
   "metadata": {},
   "source": [
    "### Standardize metabolite formulas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d2f379e-1216-40aa-a6ef-1529323aa1da",
   "metadata": {},
   "outputs": [],
   "source": [
    "metabolite_formulas = dict(\n",
    "    zip(model.metabolites.list_attr(\"id\"), model.metabolites.list_attr(\"formula\"))\n",
    ")\n",
    "standardized = standardardize_metabolite_formulas(metabolite_formulas)\n",
    "\n",
    "for mid, updated_formula in standardized.items():\n",
    "    if metabolite_formulas[mid] != updated_formula:\n",
    "        print(f\"Standardizing formula for `{mid}`\")\n",
    "        model_metabolite = model.metabolites.get_by_id(mid)\n",
    "        model_metabolite.formula = updated_formula"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f939329a-022d-4a39-94f4-7f0ba639358f",
   "metadata": {},
   "source": [
    "## Export updated model\n",
    "### Version: 0.2.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d237f4db-3ee3-4fd6-96f9-68b78290d4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_rbc_model(model, \"all\")\n",
    "model"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
