{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ba9c3d2c-72da-416c-8612-8a3f91eee20a",
   "metadata": {},
   "source": [
    "# RBC-GEM 1.2.0 Updates\n",
    "## Setup\n",
    "### Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "194c1f88-6d9b-40d5-b468-86da5bc66f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from warnings import warn\n",
    "\n",
    "import pandas as pd\n",
    "from cobra.core import Gene, Metabolite, Reaction\n",
    "from rbc_gem_utils import (\n",
    "    COBRA_CONFIGURATION,\n",
    "    CURATION_PATH,\n",
    "    INTERIM_PATH,\n",
    "    ROOT_PATH,\n",
    "    build_string,\n",
    "    read_rbc_model,\n",
    "    split_string,\n",
    "    write_rbc_model,\n",
    ")\n",
    "from rbc_gem_utils.qc import (\n",
    "    reset_reaction_bounds,\n",
    "    reset_subsystem_groups,\n",
    "    standardardize_metabolite_formulas,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "945a5ce1-932d-4622-a823-c54b76b48d99",
   "metadata": {},
   "source": [
    "### Define configuration\n",
    "#### COBRA Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c9d3e47-f2f1-44e6-9333-4adec35ef021",
   "metadata": {},
   "outputs": [],
   "source": [
    "COBRA_CONFIGURATION"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f967b496-7302-49a9-a474-45b13dc70980",
   "metadata": {},
   "source": [
    "## Load RBC-GEM model\n",
    "### Version: 1.1.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "787d69b8-f885-4cda-9598-505f2b72b8c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = read_rbc_model(filetype=\"yml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a1c443b-d26f-41e6-a674-30a1c91900db",
   "metadata": {},
   "source": [
    "### Deprecate identifiers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91e9978b-8421-45e9-8260-35b86b855dfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "id_mapping_dicts = {\n",
    "    \"metabolites\": {\n",
    "        \"PPP2CA_leu__L_c\": \"pp2a_leu__L_c\",\n",
    "        \"PPP2CA_leumeo_c\": \"pp2a_leumeo_c\",\n",
    "        # Protein ID updates required for protein constrained modeling compatibility\n",
    "        \"protein_aclys__L_c\": \"protres_aclys__L_c\",\n",
    "        \"protein_admarg__L_c\": \"protres_admarg__L_c\",\n",
    "        \"protein_arg__L_c\": \"protres_arg__L_c\",\n",
    "        \"protein_arg__L_e\": \"protres_arg__L_e\",\n",
    "        \"protein_asn__L_c\": \"protres_asn__L_c\",\n",
    "        \"protein_asp__D_c\": \"protres_asp__D_c\",\n",
    "        \"protein_asp__L_c\": \"protres_asp__L_c\",\n",
    "        \"protein_aspmeo__D_c\": \"protres_aspmeo__D_c\",\n",
    "        \"protein_cys__L_c\": \"protres_cys__L_c\",\n",
    "        \"protein_frulys_c\": \"protres_frulys_c\",\n",
    "        \"protein_frulys3p_c\": \"protres_frulys3p_c\",\n",
    "        \"protein_frulys6p_c\": \"protres_frulys6p_c\",\n",
    "        \"protein_gln__L_c\": \"protres_gln__L_c\",\n",
    "        \"protein_glu__L_c\": \"protres_glu__L_c\",\n",
    "        \"protein_glu5meo__L_c\": \"protres_glu5meo__L_c\",\n",
    "        \"protein_gludpam_c\": \"protres_gludpam_c\",\n",
    "        \"protein_gluhista_c\": \"protres_gluhista_c\",\n",
    "        \"protein_glunpphr_c\": \"protres_glunpphr_c\",\n",
    "        \"protein_glusrtn_c\": \"protres_glusrtn_c\",\n",
    "        \"protein_his__L_c\": \"protres_his__L_c\",\n",
    "        \"protein_hxdcacys_c\": \"protres_hxdcacys_c\",\n",
    "        \"protein_isoasp__D_c\": \"protres_isoasp__D_c\",\n",
    "        \"protein_isoasp__L_c\": \"protres_isoasp__L_c\",\n",
    "        \"protein_isoaspmeo__L_c\": \"protres_isoaspmeo__L_c\",\n",
    "        \"protein_lacarg_c\": \"protres_lacarg_c\",\n",
    "        \"protein_laccys_c\": \"protres_laccys_c\",\n",
    "        \"protein_laclys_c\": \"protres_laclys_c\",\n",
    "        \"protein_lys__L_c\": \"protres_lys__L_c\",\n",
    "        \"protein_lysglu_protein_c\": \"protres_lysglu_protres_c\",\n",
    "        \"protein_mmarg__L_c\": \"protres_mmarg__L_c\",\n",
    "        \"protein_Nproshispi_c\": \"protres_Nproshispi_c\",\n",
    "        \"protein_Ntelehispi_c\": \"protres_Ntelehispi_c\",\n",
    "        \"protein_Nwadpribarg__L_e\": \"protres_Nwadpribarg__L_e\",\n",
    "        \"protein_oadpribser_c\": \"protres_oadpribser_c\",\n",
    "        \"protein_ocdcacys_c\": \"protres_ocdcacys_c\",\n",
    "        \"protein_pepproW0_c\": \"protres_pepproW0_c\",\n",
    "        \"protein_pepproW180_c\": \"protres_pepproW180_c\",\n",
    "        \"protein_rbllys_c\": \"protres_rbllys_c\",\n",
    "        \"protein_rbllys3p_c\": \"protres_rbllys3p_c\",\n",
    "        \"protein_rbllys5p_c\": \"protres_rbllys5p_c\",\n",
    "        \"protein_sadpribcys_c\": \"protres_sadpribcys_c\",\n",
    "        \"protein_sdmarg__L_c\": \"protres_sdmarg__L_c\",\n",
    "        \"protein_ser__L_c\": \"protres_ser__L_c\",\n",
    "        \"protein_ser3oacgam_c\": \"protres_ser3oacgam_c\",\n",
    "        \"protein_serpi_c\": \"protres_serpi_c\",\n",
    "        \"protein_serTAg_c\": \"protres_serTAg_c\",\n",
    "        \"protein_serTnAg_c\": \"protres_serTnAg_c\",\n",
    "        \"protein_sucim__D_c\": \"protres_sucim__D_c\",\n",
    "        \"protein_sucim__L_c\": \"protres_sucim__L_c\",\n",
    "        \"protein_thr__L_c\": \"protres_thr__L_c\",\n",
    "        \"protein_thr3oacgam_c\": \"protres_thr3oacgam_c\",\n",
    "        \"protein_thrpi_c\": \"protres_thrpi_c\",\n",
    "        \"protein_thrTAg_c\": \"protres_thrTAg_c\",\n",
    "        \"protein_thrTnAg_c\": \"protres_thrTnAg_c\",\n",
    "        \"protein_ttdcacys_c\": \"protres_ttdcacys_c\",\n",
    "        \"protein_tyr__L_c\": \"protres_tyr__L_c\",\n",
    "        \"protein_tyrpi_c\": \"protres_tyrpi_c\",\n",
    "    },\n",
    "    \"genes\": {},\n",
    "    \"reactions\": {\n",
    "        \"ClPItex\": \"Cl_PItex\",\n",
    "        \"NO22Natex\": \"NO2_2Natex\",\n",
    "        \"CYSL\": \"CYSLNET\",\n",
    "        \"CYSTCYSL\": \"CYSTL1\",\n",
    "        \"SELCYSTL\": \"SELCYSTL1\",\n",
    "        \"SELMETHL\": \"SELMETHL1\",\n",
    "        \"LCMTPPP2CA\": \"LCMTPP2A\",\n",
    "        \"PPMEPPP2CA\": \"PPMEPP2A\",\n",
    "        \"GALt1r\": \"GAL_Dt\",\n",
    "        \"EX_protein_Nwadpribarg__L_e\": \"EX_protres_Nwadpribarg__L_e\",\n",
    "        \"EX_protein_arg__L_e\": \"EX_protres_arg__L_e\",\n",
    "        \"IMRBVPT\": \"RBVIMPT\",\n",
    "        # Cleanup of nucleotide IDs --> NTD(D)(N)MP\n",
    "        \"NTD1\": \"NTDDUMP\",\n",
    "        \"NTD2\": \"NTDUMP\",\n",
    "        \"NTD3\": \"NTDDCMP\",\n",
    "        \"NTD4\": \"NTDCMP\",\n",
    "        \"NTD5\": \"NTDDTMP\",\n",
    "        \"NTD6\": \"NTDDAMP\",\n",
    "        \"NTD7\": \"NTDAMP\",\n",
    "        \"NTD8\": \"NTDDGMP\",\n",
    "        \"NTD9\": \"NTDGMP\",\n",
    "        \"NTD10\": \"NTDXMP\",\n",
    "        \"NTD11\": \"NTDIMP\",\n",
    "        \"NTD12\": \"NTDDIMP\",\n",
    "        \"NTD15\": \"NTDPSUMP\",\n",
    "        \"2NTDA\": \"NTD2AMP\",\n",
    "        \"2NTDC\": \"NTD2CMP\",\n",
    "        \"2NTDG\": \"NTD2GMP\",\n",
    "        \"2NTDU\": \"NTD2UMP\",\n",
    "        \"NTDAe\": \"NTDAMPe\",\n",
    "        \"NDPAe\": \"NDPADPe\",\n",
    "        \"NTDADPe\": \"NDPADPe\",\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "473c1a5d-5b44-4028-affe-9b059fa50c3f",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc76e7fa-8a1e-4f6e-bb74-4d4745e7f189",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "id_mapping_dict = id_mapping_dicts[attribute_type]\n",
    "\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"metRetired\", \"mets\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    ROOT_PATH\n",
    "    / \"data\"\n",
    "    / \"deprecatedIdentifiers\"\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"mets\", \"metRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[\n",
    "        previous_id_mapping_df[\"mets\"] == retiring\n",
    "    ]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        retired_set_of_ids.update(\n",
    "            previously_retired[\"metRetired\"].apply(split_string).item()\n",
    "        )\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"metRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "for old, new in id_mapping_dict.items():\n",
    "    try:\n",
    "        metabolite = model.metabolites.get_by_id(old)\n",
    "    except KeyError:\n",
    "        print(f\"Could not map {old} to new ID.\")\n",
    "    else:\n",
    "        try:\n",
    "            metabolite.id = id_mapping_dict[metabolite.id]\n",
    "        except ValueError as e:\n",
    "            warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n",
    "            model.remove_metabolites([metabolite])\n",
    "\n",
    "model.repair()\n",
    "\n",
    "id_mapping_df = pd.concat((id_mapping_df, previous_id_mapping_df), axis=0)\n",
    "id_mapping_df = (\n",
    "    id_mapping_df.drop_duplicates()\n",
    "    .sort_values(\"mets\", ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "id_mapping_df.to_csv(\n",
    "    ROOT_PATH\n",
    "    / \"data\"\n",
    "    / \"deprecatedIdentifiers\"\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e053b43e-9c15-4c91-8032-68bf5f5ca30b",
   "metadata": {},
   "source": [
    "#### Reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3ce4c3a-e273-491d-a597-f7847cd19479",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"reactions\"\n",
    "id_mapping_dict = id_mapping_dicts[attribute_type]\n",
    "\n",
    "\n",
    "id_mapping_df = pd.DataFrame.from_dict(id_mapping_dict, orient=\"index\")\n",
    "id_mapping_df = id_mapping_df.reset_index(drop=False)\n",
    "id_mapping_df.columns = [\"rxnRetired\", \"rxns\"]\n",
    "id_mapping_df = id_mapping_df.loc[:, id_mapping_df.columns[::-1]]\n",
    "\n",
    "previous_id_mapping_df = pd.read_csv(\n",
    "    ROOT_PATH\n",
    "    / \"data\"\n",
    "    / \"deprecatedIdentifiers\"\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ")\n",
    "\n",
    "for idx, row in id_mapping_df.iterrows():\n",
    "    new_id, retiring = row[[\"rxns\", \"rxnRetired\"]]\n",
    "    previously_retired = previous_id_mapping_df[\n",
    "        previous_id_mapping_df[\"rxns\"] == retiring\n",
    "    ]\n",
    "    retired_set_of_ids = {retiring}\n",
    "    if not previously_retired.empty:\n",
    "        # Get all previously retired IDs\n",
    "        print(previously_retired)\n",
    "        retired_set_of_ids.update(\n",
    "            previously_retired[\"rxnRetired\"].apply(split_string).item()\n",
    "        )\n",
    "        # Pulling the ID out of retirement\n",
    "        if new_id in retired_set_of_ids:\n",
    "            retired_set_of_ids.remove(new_id)\n",
    "        retired_set_of_ids.add(retiring)\n",
    "    id_mapping_df.loc[idx, \"rxnRetired\"] = build_string(retired_set_of_ids, sep=\";\")\n",
    "\n",
    "for old, new in id_mapping_dict.items():\n",
    "    try:\n",
    "        reaction = model.reactions.get_by_id(old)\n",
    "    except KeyError:\n",
    "        print(f\"Could not map {old} to new ID.\")\n",
    "    else:\n",
    "        try:\n",
    "            reaction.id = id_mapping_dict[reaction.id]\n",
    "        except ValueError as e:\n",
    "            warn(f\"{e}, this is expected if removing/renaming a duplicate\")\n",
    "            model.remove_reactions([reaction])\n",
    "\n",
    "model.repair()\n",
    "\n",
    "id_mapping_df = pd.concat((id_mapping_df, previous_id_mapping_df), axis=0)\n",
    "id_mapping_df = (\n",
    "    id_mapping_df.drop_duplicates()\n",
    "    .sort_values(\"rxns\", ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "id_mapping_df.to_csv(\n",
    "    ROOT_PATH\n",
    "    / \"data\"\n",
    "    / \"deprecatedIdentifiers\"\n",
    "    / f\"{attribute_type}_deprecatedIdentifiers.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index=False,\n",
    ")\n",
    "id_mapping_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "624e20c3-db6d-4ec6-96fc-9600cc998e1f",
   "metadata": {},
   "source": [
    "### Update model based on curation and proteomic evidence\n",
    "* Add metabolites first, then genes and lastly, reactions\n",
    "* Add metabolites with mass and charge balanced formulas (ChemAxon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03deaf1e-18e7-4ce7-ae1b-597650a70cd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataframes_updated = {}\n",
    "dataframes_evidence = {}\n",
    "dataframes_removed = {}\n",
    "overwrite = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b32955-a0d5-45c6-81e4-183760ce44d0",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e073dc66-8abb-48e3-9c70-7dc447806c2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "\n",
    "df_updated = pd.read_csv(\n",
    "    ROOT_PATH / CURATION_PATH / f\"{attribute_type}_updated_1.2.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "\n",
    "attr_cols = [\"metabolites\", \"name\", \"formula\", \"charge\", \"compartment\"]\n",
    "for idx, row in df_updated.iterrows():\n",
    "    mid, name, formula, charge, compartment = row[attr_cols]\n",
    "    if not model.metabolites.has_id(mid):\n",
    "        # Add reaction to model\n",
    "        model.add_metabolites([Metabolite(mid)])\n",
    "        # print(mid)\n",
    "\n",
    "    metabolite = model.metabolites.get_by_id(mid)\n",
    "    metabolite.name = name\n",
    "    metabolite.formula = formula\n",
    "    metabolite.charge = int(charge)\n",
    "    metabolite.compartment = compartment\n",
    "\n",
    "    annotations_dict = row[~row.index.isin(attr_cols)].to_dict()\n",
    "    notes = annotations_dict.pop(\"notes\")\n",
    "    if notes:\n",
    "        metabolite.notes.update({\"notes\": str(notes)})\n",
    "    annotations_dict = {\n",
    "        k: v\n",
    "        for k, v in annotations_dict.items()\n",
    "        if not k.startswith(\"metabolomic\") and v\n",
    "    }\n",
    "    metabolite.annotation.update(annotations_dict)\n",
    "\n",
    "met_formulas = standardardize_metabolite_formulas(\n",
    "    dict(zip(model.metabolites.list_attr(\"id\"), model.metabolites.list_attr(\"formula\")))\n",
    ")\n",
    "for mid, formula in met_formulas.items():\n",
    "    model.metabolites.get_by_id(mid).formula = formula\n",
    "\n",
    "\n",
    "dataframes_updated[attribute_type] = df_updated\n",
    "try:\n",
    "    df_previous_evidence = pd.read_csv(\n",
    "        ROOT_PATH / CURATION_PATH / f\"{attribute_type}_evidence.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0)\n",
    "df_evidence[attribute_type] = df_evidence[attribute_type].replace(\n",
    "    id_mapping_dicts[attribute_type]\n",
    ")\n",
    "df_evidence = (\n",
    "    df_evidence.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")[df_previous_evidence.columns]\n",
    "\n",
    "dataframes_evidence[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "142935ef-76b4-4ccb-8ba6-2a3f999c2f2b",
   "metadata": {},
   "source": [
    "#### Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "852b2f5f-4195-464b-b880-63ada6769a84",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"genes\"\n",
    "\n",
    "df_updated = pd.read_csv(\n",
    "    ROOT_PATH / CURATION_PATH / f\"{attribute_type}_updated_1.2.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "\n",
    "attr_cols = [\"genes\"]\n",
    "for idx, row in df_updated.iterrows():\n",
    "    # Between these three annotation fields, should be likely that the rest can get extracted from UniProt.\n",
    "    gid = row[attr_cols[0]]\n",
    "    if not model.genes.has_id(gid):\n",
    "        gene = Gene(gid)\n",
    "        model.genes.extend([gene])\n",
    "        # print(gene.id)\n",
    "    gene = model.genes.get_by_id(gid)\n",
    "    annotations_dict = row[~row.index.isin(attr_cols)].to_dict()\n",
    "    annotations_dict = {\n",
    "        k: v for k, v in annotations_dict.items() if not k.startswith(\"proteomic\") and v\n",
    "    }\n",
    "    gene.annotation.update(annotations_dict)\n",
    "\n",
    "\n",
    "dataframes_updated[attribute_type] = df_updated\n",
    "try:\n",
    "    df_previous_evidence = pd.read_csv(\n",
    "        ROOT_PATH / CURATION_PATH / f\"{attribute_type}_evidence.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "\n",
    "df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0)\n",
    "df_evidence[attribute_type] = df_evidence[attribute_type].replace(\n",
    "    id_mapping_dicts[attribute_type]\n",
    ")\n",
    "df_evidence = (\n",
    "    df_evidence.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")[df_previous_evidence.columns]\n",
    "\n",
    "dataframes_evidence[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3b1b3476-7c28-4164-a425-87ea22f92085",
   "metadata": {},
   "source": [
    "#### Reactions\n",
    "* Addition of hemoglobin glycation\n",
    "* All intracellular reactions converted to use/consume NH4, which is dominant at pH 7.25. Transport and protonation reactions for NH3 kept.\n",
    "* Add NEDD8 neddylation and fix ubiquitin reaction GPRs accordingly\n",
    "* Reversibility updates for phosphoribosyltransferase reactions\n",
    "* GPR updates according to Complex Portal and proteomic data\n",
    "    * Add PRMT5 methylsome complex\n",
    "    * Add PI3K class III complex proteins\n",
    "    * Multiple Ubiquination complexes\n",
    "    * Multiple VCP complexes and other ATPases\n",
    "* Add IDH2 and MDH2 to GPRs\n",
    "* Add SLC22A4 and SLC22A16 GPRs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff19985d-2a62-4283-8da0-6b8177927b92",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"reactions\"\n",
    "\n",
    "df_updated = pd.read_csv(\n",
    "    ROOT_PATH / CURATION_PATH / f\"{attribute_type}_updated_1.2.0.tsv\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=None,\n",
    ").fillna(\"\")\n",
    "attr_cols = [\"reactions\", \"name\", \"reaction\", \"gene reaction rule\", \"subsystem\"]\n",
    "for idx, row in df_updated.iterrows():\n",
    "    rid, name, reaction_str, gpr, subsystem = row[attr_cols]\n",
    "    try:\n",
    "        if not model.reactions.has_id(rid):\n",
    "            # Add reaction to model\n",
    "            model.add_reactions([Reaction(rid)])\n",
    "    except ValueError as e:\n",
    "        print(idx, rid, name)\n",
    "        raise e\n",
    "\n",
    "    reaction = model.reactions.get_by_id(rid)\n",
    "    reaction.build_reaction_from_string(reaction_str)\n",
    "    reaction.name = name\n",
    "    reaction.gene_reaction_rule = gpr\n",
    "    reaction.subsystem = subsystem\n",
    "\n",
    "    annotations_dict = row[~row.index.isin(attr_cols)].to_dict()\n",
    "    notes = annotations_dict.pop(\"notes\")\n",
    "    if notes:\n",
    "        reaction.notes.update({\"notes\": str(notes)})\n",
    "    if \"proteomic evidence (#studies)\" in annotations_dict:\n",
    "        annotations_dict.pop(\"proteomic evidence (#studies)\")\n",
    "    annotations_dict = {\n",
    "        k: v for k, v in annotations_dict.items() if not k.startswith(\"proteomic\") and v\n",
    "    }\n",
    "    reaction.annotation.update(annotations_dict)\n",
    "\n",
    "\n",
    "dataframes_updated[attribute_type] = df_updated\n",
    "try:\n",
    "    df_previous_evidence = pd.read_csv(\n",
    "        ROOT_PATH / CURATION_PATH / f\"{attribute_type}_evidence.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_previous_evidence = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "df_evidence = pd.concat((df_updated, df_previous_evidence), axis=0)\n",
    "df_evidence[attribute_type] = df_evidence[attribute_type].replace(\n",
    "    id_mapping_dicts[attribute_type]\n",
    ")\n",
    "df_evidence = (\n",
    "    df_evidence.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(\n",
    "        [\"subsystem\", \"reactions\"],\n",
    "        ascending=[True, True],\n",
    "    )\n",
    "    .reset_index(drop=True)\n",
    ")[df_previous_evidence.columns]\n",
    "\n",
    "dataframes_evidence[attribute_type] = df_evidence\n",
    "dataframes_updated[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64d45450-d19e-4e09-b76b-8ad6471b8153",
   "metadata": {},
   "source": [
    "### Refine model through removing items\n",
    "#### Reactions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d434a25-de3f-4e6b-98be-bb902e3092eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"reactions\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        ROOT_PATH / CURATION_PATH / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for reaction in df_removed[attribute_type]:\n",
    "    try:\n",
    "        reaction = getattr(model, attribute_type).get_by_id(reaction)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(reaction)\n",
    "\n",
    "model.remove_reactions(to_remove)\n",
    "# Clean up removal file before archival\n",
    "df_removed = (\n",
    "    df_removed.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "    ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "        df_removed[attribute_type].values\n",
    "    )\n",
    "]\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7279505a-5e9f-44f3-b4ba-33e9b957d6e2",
   "metadata": {},
   "source": [
    "#### Genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8987ef03-a8fb-433e-8396-99a6d9bd6d1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"genes\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        ROOT_PATH / CURATION_PATH / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for gene in df_removed[attribute_type]:\n",
    "    try:\n",
    "        gene = model.genes.get_by_id(gene)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    to_remove.append(gene)\n",
    "model.genes -= to_remove\n",
    "\n",
    "# Clean up removal file before archival\n",
    "df_removed = (\n",
    "    df_removed.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "for gene in model.genes:\n",
    "    if gene.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {gene}\")\n",
    "\n",
    "dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "    ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "        df_removed[attribute_type].values\n",
    "    )\n",
    "]\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eab7e51b-9730-439a-a0ce-271b65d72dcd",
   "metadata": {},
   "source": [
    "#### Metabolites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d9b992c-9399-45aa-921f-c415db9e0178",
   "metadata": {},
   "outputs": [],
   "source": [
    "attribute_type = \"metabolites\"\n",
    "try:\n",
    "    df_removed = pd.read_csv(\n",
    "        ROOT_PATH / CURATION_PATH / f\"{attribute_type}_removed.tsv\",\n",
    "        sep=\"\\t\",\n",
    "        index_col=None,\n",
    "        dtype=str,\n",
    "    )\n",
    "except FileNotFoundError:\n",
    "    df_removed = pd.DataFrame([], columns=[attribute_type], dtype=str)\n",
    "\n",
    "to_remove = []\n",
    "for metabolite in df_removed[attribute_type]:\n",
    "    try:\n",
    "        metabolite = model.metabolites.get_by_id(metabolite)\n",
    "    except KeyError:\n",
    "        continue\n",
    "    model.remove_metabolites([metabolite])\n",
    "model.remove_metabolites(to_remove)\n",
    "\n",
    "# Clean up removal file before archival\n",
    "df_removed = (\n",
    "    df_removed.drop_duplicates(subset=[attribute_type])\n",
    "    .sort_values(attribute_type, ascending=True)\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "for metabolite in model.metabolites:\n",
    "    if metabolite.reactions:\n",
    "        continue\n",
    "    print(f\"Orphaned: {metabolite}\")\n",
    "\n",
    "dataframes_evidence[attribute_type] = dataframes_evidence[attribute_type][\n",
    "    ~dataframes_evidence[attribute_type][attribute_type].isin(\n",
    "        df_removed[attribute_type].values\n",
    "    )\n",
    "]\n",
    "dataframes_removed[attribute_type] = df_removed\n",
    "dataframes_removed[attribute_type]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cfe3007-e6fe-4fdd-97a7-4f764d13ae57",
   "metadata": {},
   "outputs": [],
   "source": [
    "for attribute_type in [\"reactions\", \"genes\", \"metabolites\"]:\n",
    "    df_evidence = dataframes_evidence[attribute_type]\n",
    "    df_removed = dataframes_removed[attribute_type]\n",
    "    if overwrite:\n",
    "        df_evidence.to_csv(\n",
    "            ROOT_PATH / CURATION_PATH / f\"{attribute_type}_evidence.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "        )\n",
    "        df_removed.to_csv(\n",
    "            ROOT_PATH / CURATION_PATH / f\"{attribute_type}_removed.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "        )\n",
    "    else:\n",
    "        df_evidence.to_csv(\n",
    "            ROOT_PATH / INTERIM_PATH / f\"{attribute_type}_evidence.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "        )\n",
    "        df_removed.to_csv(\n",
    "            ROOT_PATH / INTERIM_PATH / f\"{attribute_type}_removed.tsv\",\n",
    "            sep=\"\\t\",\n",
    "            index=False,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "886da88c-00c0-4638-974f-2766333f4d7b",
   "metadata": {},
   "source": [
    "### Ensure all metabolites, genes, and reactions exist\n",
    "If removed (e.g., a duplicate), will show up in missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c56939d1-4a0c-4f68-b200-b84d7da345d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_metabolites = set(model.metabolites.list_attr(\"id\")).symmetric_difference(\n",
    "    dataframes_evidence[\"metabolites\"][\"metabolites\"].values\n",
    ")\n",
    "\n",
    "missing_genes = set(model.genes.list_attr(\"id\")).symmetric_difference(\n",
    "    dataframes_evidence[\"genes\"][\"genes\"].values\n",
    ")\n",
    "\n",
    "missing_reactions = set(\n",
    "    model.reactions.query(lambda x: not x.subsystem == \"Pseudoreactions\").list_attr(\n",
    "        \"id\"\n",
    "    )\n",
    ").symmetric_difference(dataframes_evidence[\"reactions\"][\"reactions\"].values)\n",
    "\n",
    "print(len(missing_genes))\n",
    "print(len(missing_metabolites))\n",
    "print(len(missing_reactions))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f237f06-e5c8-49d2-ae2f-c2f9a67c8fde",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in missing_metabolites:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "382efe39-c1a3-43b8-a129-853ad91020b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in missing_genes:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dd0a570-1026-4b40-8e95-1b270900a61a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in missing_reactions:\n",
    "    print(x)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc83e690-0829-4236-9161-165a400539cb",
   "metadata": {},
   "source": [
    "### Check for extra metabolites, genes, and reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "834b542d-2669-43ab-b212-4d875f54558c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for met in model.metabolites.query(lambda x: not len(x.reactions)):\n",
    "    print(f\"{met.id}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bc9de67-ade2-46d6-b873-7f392e1c2e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for gene in model.genes.query(lambda x: not len(x.reactions)):\n",
    "    print(gene.id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f764f34d-4465-44df-b973-9ea11b0fb93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for reaction in model.reactions.query(lambda x: not len(x.metabolites)):\n",
    "    print(reaction.id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a0c06ba-7a50-4dfc-962c-fa8044219372",
   "metadata": {},
   "source": [
    "### Cleanup model values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "285cb1ab-9039-447a-a567-3bcf51bdc0fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "for mid, row in dataframes_evidence[\"metabolites\"].set_index(\"metabolites\").iterrows():\n",
    "    if gid in missing_metabolites:\n",
    "        continue\n",
    "    metabolite = model.metabolites.get_by_id(mid)\n",
    "    metabolite.name = row[\"name\"]\n",
    "    metabolite.formula = row[\"formula\"]\n",
    "    metabolite.charge = int(row[\"charge\"])\n",
    "    metabolite.compartment = row[\"compartment\"]\n",
    "    metabolite.annotation[\"metabolite\"] = metabolite.id.replace(\n",
    "        f\"_{metabolite.compartment}\", \"\"\n",
    "    )\n",
    "\n",
    "for rid, row in dataframes_evidence[\"reactions\"].set_index(\"reactions\").iterrows():\n",
    "    if rid in missing_reactions:\n",
    "        continue\n",
    "    reaction = model.reactions.get_by_id(rid)\n",
    "    reaction.name = row[\"name\"]\n",
    "\n",
    "for gid, row in dataframes_evidence[\"genes\"].set_index(\"genes\").iterrows():\n",
    "    if gid in missing_genes:\n",
    "        continue\n",
    "    gene = model.genes.get_by_id(gid)\n",
    "    gene.name = row[\"hgnc.symbol\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ac24451-ec72-4565-8cb8-8fc57c6ee2d9",
   "metadata": {},
   "source": [
    "#### Add boundary reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b424c4c7-dd7d-4473-b9e4-f0b69b9278b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "boundaries = {\n",
    "    # All exchange boundary reactions added\n",
    "    \"exchange\": model.metabolites.query(lambda x: x.compartment == \"e\").list_attr(\"id\"),\n",
    "    # Intracellular demands, only used for accumulation is allowed for a compound\n",
    "    \"demand\": [],\n",
    "    # Intracellular sinks, only used for when a source is needed for a compound\n",
    "    \"sink\": [\n",
    "        # Globin/Hemoglobin\n",
    "        \"oxyhb_c\",\n",
    "        \"hb4_23dpg_c\",\n",
    "        \"hb_hco2_c\",\n",
    "        \"globin_c\",\n",
    "        \"hbsno_c\",\n",
    "        \"carboxyhb_c\",\n",
    "        \"cclglobin_c\",\n",
    "        \"hemedegprods_c\",\n",
    "        \"hba1c_c\",\n",
    "        # Amino acids\n",
    "        # AA protein Residues\n",
    "        \"protres_arg__L_c\",\n",
    "        \"protres_asn__L_c\",\n",
    "        \"protres_asp__L_c\",\n",
    "        \"protres_cys__L_c\",\n",
    "        \"protres_gln__L_c\",\n",
    "        \"protres_glu__L_c\",\n",
    "        \"protres_his__L_c\",\n",
    "        \"protres_lys__L_c\",\n",
    "        \"protres_met__L_c\",\n",
    "        \"protres_ser__L_c\",\n",
    "        \"protres_thr__L_c\",\n",
    "        \"protres_tyr__L_c\",\n",
    "        \"protres_tyr__L_c\",\n",
    "        # Other AA residues\n",
    "        \"protres_asp__D_c\",\n",
    "        \"protres_isoasp__L_c\",\n",
    "        \"protres_isoasp__D_c\",\n",
    "        # Phosphorylated residues\n",
    "        \"protres_Nproshispi_c\",\n",
    "        \"protres_Ntelehispi_c\",\n",
    "        \"protres_serpi_c\",\n",
    "        \"protres_thrpi_c\",\n",
    "        \"protres_tyrpi_c\",\n",
    "        # Acetylated residues\n",
    "        \"protres_aclys__L_c\",\n",
    "        # Glycosylated residues\n",
    "        \"protres_ser3oacgam_c\",\n",
    "        \"protres_thr3oacgam_c\",\n",
    "        \"protres_serTAg_c\",\n",
    "        \"protres_thrTAg_c\",\n",
    "        # Glycated residues\n",
    "        \"protres_frulys_c\",\n",
    "        \"protres_rbllys_c\",\n",
    "        # Methylated residues\n",
    "        \"protres_admarg__L_c\",\n",
    "        \"protres_sdmarg__L_c\",\n",
    "        # Lipidated residues\n",
    "        \"protres_ttdcacys_c\",\n",
    "        \"protres_hxdcacys_c\",\n",
    "        \"protres_ocdcacys_c\",\n",
    "        # Oxidized residues\n",
    "        \"protres_metSox__SL_c\",\n",
    "        # Nitrosylated residues\n",
    "        \"protres_snocys__L_c\",\n",
    "        # ADP-ribosylated residues\n",
    "        \"protres_oadpribser_c\",\n",
    "        \"protres_sadpribcys_c\",\n",
    "        # Amine\n",
    "        \"protres_gludpam_c\",\n",
    "        \"protres_gluhista_c\",\n",
    "        \"protres_glunpphr_c\",\n",
    "        \"protres_glusrtn_c\",\n",
    "        \"protres_glu5meo__L_c\",\n",
    "        \"protres_lysglu_protres_c\",\n",
    "        # Ubiquitin\n",
    "        \"polyubb_c\",\n",
    "        \"ubiquitin_c\",\n",
    "        \"accprot_monoubiqlys_c\",\n",
    "        \"accprot_ubiqlys_c\",\n",
    "        \"accprot_lys__L_c\",\n",
    "        \"cullin_lys__L_c\",\n",
    "        \"cullin_nedd8lys_c\",\n",
    "        \"nedd8_c\",\n",
    "        # Small ions\n",
    "        \"na1_c\",\n",
    "        \"k_c\",\n",
    "        \"ca2_c\",\n",
    "        \"hno_c\",\n",
    "        \"co3r_c\",\n",
    "        # Vitamin E\n",
    "        \"avite1_c\",\n",
    "        \"avite1qn_c\",\n",
    "        # 'Redoxins'\n",
    "        \"prdx2crd_c\",\n",
    "        \"prdx2cso3_c\",\n",
    "        \"grdx2crd_c\",\n",
    "        \"grdx2cox_c\",\n",
    "        # CoA\n",
    "        # 'FAcoa_10_DC_c',\n",
    "        # 'FAcoa_12_DC_c',\n",
    "        # 'FAcoa_16_DC_c',\n",
    "        # 'FAcoa_4_DC_c',\n",
    "        # 'FAcoa_5_2EDC_c',\n",
    "        # 'FAcoa_5_DC_c',\n",
    "        # 'FAcoa_6_DC_c',\n",
    "        # 'FAcoa_7_DC_c',\n",
    "        # 'FAcoa_8_DC_c',\n",
    "        # 'FAcoa_5_3M3OH__S_c',\n",
    "        # 'FAcoa_hs_10_3OH__S_c',\n",
    "        # 'FAcoa_hs_12_3OH__S_c',\n",
    "        # 'FAcoa_hs_14_3OH__S_c',\n",
    "        # 'FAcoa_hs_14_5E8Z3OH__S_c',\n",
    "        # 'FAcoa_hs_14_7Z3OH__S_c',\n",
    "        # 'FAcoa_hs_16_3OH__R_c',\n",
    "        # 'FAcoa_hs_16_3OH__S_c',\n",
    "        # 'FAcoa_hs_16_7E10Z3OH__S_c',\n",
    "        # 'FAcoa_hs_16_9Z3OH__S_c',\n",
    "        # 'FAcoa_hs_17_3OH__R_c',\n",
    "        # 'FAcoa_hs_18_3OH__R_c',\n",
    "        # 'FAcoa_hs_18_3OH__S_c',\n",
    "        # 'FAcoa_hs_18_9Z12Z3OH__S_c',\n",
    "        # 'FAcoa_hs_18_9Z3OH__S_c',\n",
    "        # 'FAcoa_hs_19_3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_13Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_20_9Z3OH__R_c',\n",
    "        # 'FAcoa_hs_21_3OH__R_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_11Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_13Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_3OH__R_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z19Z3OH__R_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z3OH__R_c',\n",
    "        # 'FAcoa_hs_23_3OH__R_c',\n",
    "        # 'FAcoa_hs_24_12Z15Z18Z21Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_15Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_3OH__R_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z21Z3OH__R_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z3OH__R_c',\n",
    "        # 'FAcoa_hs_26_17Z3OH__R_c',\n",
    "        # 'FAcoa_hs_26_3OH__R_c',\n",
    "        # 'FAcoa_hs_3_3OH__S_c',\n",
    "        # 'FAcoa_hs_4_3OH__R_c',\n",
    "        # 'FAcoa_hs_6_3OH__S_c',\n",
    "        # 'FAcoa_hs_7_3OH__S_c',\n",
    "        # 'FAcoa_hs_8_3OH__S_c',\n",
    "        # 'FAcoa_hs_9_3OH__S_c',\n",
    "        # 'FAcoa_5_2E2M_c',\n",
    "        # 'FAcoa_5_2M_c',\n",
    "        # 'FAcoa_hs_10_2E6Z_c',\n",
    "        # 'FAcoa_hs_10_2E_c',\n",
    "        # 'FAcoa_hs_12_2E_c',\n",
    "        # 'FAcoa_hs_14_2E_c',\n",
    "        # 'FAcoa_hs_16_2E_c',\n",
    "        # 'FAcoa_hs_17_2E_c',\n",
    "        # 'FAcoa_hs_18_2E_c',\n",
    "        # 'FAcoa_hs_19_2E_c',\n",
    "        # 'FAcoa_hs_20_2E11Z14Z17Z_c',\n",
    "        # 'FAcoa_hs_20_2E11Z14Z_c',\n",
    "        # 'FAcoa_hs_20_2E11Z_c',\n",
    "        # 'FAcoa_hs_20_2E13Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z14Z17Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z14Z_c',\n",
    "        # 'FAcoa_hs_20_2E8Z11Z_c',\n",
    "        # 'FAcoa_hs_20_2E9Z_c',\n",
    "        # 'FAcoa_hs_20_2E_c',\n",
    "        # 'FAcoa_hs_21_2E_c',\n",
    "        # 'FAcoa_hs_22_2E10Z13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E10Z13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E11Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E13Z_c',\n",
    "        # 'FAcoa_hs_22_2E7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcoa_hs_22_2E7Z10Z13Z16Z_c',\n",
    "        # 'FAcoa_hs_22_2E_c',\n",
    "        # 'FAcoa_hs_23_2E_c',\n",
    "        # 'FAcoa_hs_24_2E12Z15Z18Z21Z_c',\n",
    "        # 'FAcoa_hs_24_2E15Z_c',\n",
    "        # 'FAcoa_hs_24_2E9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcoa_hs_24_2E9Z12Z15Z18Z_c',\n",
    "        # 'FAcoa_hs_24_2E_c',\n",
    "        # 'FAcoa_hs_26_2E17Z_c',\n",
    "        # 'FAcoa_hs_26_2E_c',\n",
    "        # 'FAcoa_hs_3_2E_c',\n",
    "        # 'FAcoa_hs_4_2E_c',\n",
    "        # 'FAcoa_hs_6_2E_c',\n",
    "        # 'FAcoa_hs_8_2E_c',\n",
    "        # 'FAcoa_hs_14_5E8Z_c',\n",
    "        # 'FAcoa_4_2M_c',\n",
    "        # 'FAcoa_hs_16_3O_c',\n",
    "        # 'FAcoa_hs_17_3O_c',\n",
    "        # 'FAcoa_hs_18_3O_c',\n",
    "        # 'FAcoa_hs_19_3O_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z17Z3O_c',\n",
    "        # 'FAcoa_hs_20_11Z14Z3O_c',\n",
    "        # 'FAcoa_hs_20_11Z3O_c',\n",
    "        # 'FAcoa_hs_20_13Z3O_c',\n",
    "        # 'FAcoa_hs_20_3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z17Z3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z14Z3O_c',\n",
    "        # 'FAcoa_hs_20_8Z11Z3O_c',\n",
    "        # 'FAcoa_hs_20_9Z3O_c',\n",
    "        # 'FAcoa_hs_21_3O_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_10Z13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_22_11Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_22_13Z3O_c',\n",
    "        # 'FAcoa_hs_22_3O_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z19Z3O_c',\n",
    "        # 'FAcoa_hs_22_7Z10Z13Z16Z3O_c',\n",
    "        # 'FAcoa_hs_23_3O_c',\n",
    "        # 'FAcoa_hs_24_12Z15Z18Z21Z3O_c',\n",
    "        # 'FAcoa_hs_24_15Z3O_c',\n",
    "        # 'FAcoa_hs_24_3O_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z21Z3O_c',\n",
    "        # 'FAcoa_hs_24_9Z12Z15Z18Z3O_c',\n",
    "        # 'FAcoa_hs_26_17Z3O_c',\n",
    "        # 'FAcoa_hs_26_3O_c',\n",
    "        # 'dmnoncoa_c',\n",
    "        # 'dmhptcoa_c',\n",
    "        # # Carnitine\n",
    "        # 'FAcrn_10_DC_c',\n",
    "        # 'FAcrn_12_DC_c',\n",
    "        # 'FAcrn_16_DC_c',\n",
    "        # 'FAcrn_4_2M_c',\n",
    "        # 'FAcrn_4_DC_c',\n",
    "        # 'FAcrn_5_2E2M_c',\n",
    "        # 'FAcrn_5_2EDC_c',\n",
    "        # 'FAcrn_5_2M_c',\n",
    "        # 'FAcrn_5_3M3OH__S_c',\n",
    "        # 'FAcrn_5_DC_c',\n",
    "        # 'FAcrn_6_DC_c',\n",
    "        # 'FAcrn_7_DC_c',\n",
    "        # 'FAcrn_8_DC_c',\n",
    "        # 'FAcrn_hs_10_0_c',\n",
    "        # 'FAcrn_hs_10_2E6Z_c',\n",
    "        # 'FAcrn_hs_10_2E_c',\n",
    "        # 'FAcrn_hs_10_3OH__S_c',\n",
    "        # 'FAcrn_hs_11_0_c',\n",
    "        # 'FAcrn_hs_12_0_c',\n",
    "        # 'FAcrn_hs_12_2E_c',\n",
    "        # 'FAcrn_hs_12_3OH__S_c',\n",
    "        # 'FAcrn_hs_13_0_c',\n",
    "        # 'FAcrn_hs_14_0_c',\n",
    "        # 'FAcrn_hs_14_2E_c',\n",
    "        # 'FAcrn_hs_14_3OH__S_c',\n",
    "        # 'FAcrn_hs_14_5E8Z3OH__S_c',\n",
    "        # 'FAcrn_hs_14_5E8Z_c',\n",
    "        # 'FAcrn_hs_14_5Z_c',\n",
    "        # 'FAcrn_hs_14_7Z3OH__S_c',\n",
    "        # 'FAcrn_hs_14_7Z_c',\n",
    "        # 'FAcrn_hs_14_9Z_c',\n",
    "        # 'FAcrn_hs_15_0_c',\n",
    "        # 'FAcrn_hs_16_0_c',\n",
    "        # 'FAcrn_hs_16_2E_c',\n",
    "        # 'FAcrn_hs_16_3OH__S_c',\n",
    "        # 'FAcrn_hs_16_7E10Z3OH__S_c',\n",
    "        # 'FAcrn_hs_16_7Z_c',\n",
    "        # 'FAcrn_hs_16_9Z3OH__S_c',\n",
    "        # 'FAcrn_hs_16_9Z_c',\n",
    "        # 'FAcrn_hs_17_0_c',\n",
    "        # 'FAcrn_hs_17_10Z_c',\n",
    "        # 'FAcrn_hs_17_9Z_c',\n",
    "        # 'FAcrn_hs_18_0_c',\n",
    "        # 'FAcrn_hs_18_11Z_c',\n",
    "        # 'FAcrn_hs_18_13Z_c',\n",
    "        # 'FAcrn_hs_18_2E_c',\n",
    "        # 'FAcrn_hs_18_3OH__S_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z12Z15Z_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z12Z_c',\n",
    "        # 'FAcrn_hs_18_6Z9Z_c',\n",
    "        # 'FAcrn_hs_18_7Z_c',\n",
    "        # 'FAcrn_hs_18_9E_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z15Z_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z3OH__S_c',\n",
    "        # 'FAcrn_hs_18_9Z12Z_c',\n",
    "        # 'FAcrn_hs_18_9Z3OH__S_c',\n",
    "        # 'FAcrn_hs_18_9Z_c',\n",
    "        # 'FAcrn_hs_19_0_c',\n",
    "        # 'FAcrn_hs_20_0_c',\n",
    "        # 'FAcrn_hs_20_11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_11Z_c',\n",
    "        # 'FAcrn_hs_20_13Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_5Z8Z11Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z14Z17Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z14Z_c',\n",
    "        # 'FAcrn_hs_20_8Z11Z_c',\n",
    "        # 'FAcrn_hs_20_9Z_c',\n",
    "        # 'FAcrn_hs_21_0_c',\n",
    "        # 'FAcrn_hs_22_0_c',\n",
    "        # 'FAcrn_hs_22_10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_11Z_c',\n",
    "        # 'FAcrn_hs_22_13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_13Z_c',\n",
    "        # 'FAcrn_hs_22_4Z7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_4Z7Z10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_22_7Z10Z13Z16Z19Z_c',\n",
    "        # 'FAcrn_hs_22_7Z10Z13Z16Z_c',\n",
    "        # 'FAcrn_hs_23_0_c',\n",
    "        # 'FAcrn_hs_24_0_c',\n",
    "        # 'FAcrn_hs_24_12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_15Z_c',\n",
    "        # 'FAcrn_hs_24_6Z9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_6Z9Z12Z15Z18Z_c',\n",
    "        # 'FAcrn_hs_24_9Z12Z15Z18Z21Z_c',\n",
    "        # 'FAcrn_hs_24_9Z12Z15Z18Z_c',\n",
    "        # 'FAcrn_hs_26_0_c',\n",
    "        # 'FAcrn_hs_26_17Z_c',\n",
    "        # 'FAcrn_hs_3_0_c',\n",
    "        # 'FAcrn_hs_3_2E_c',\n",
    "        # 'FAcrn_hs_3_3OH__S_c',\n",
    "        # 'FAcrn_hs_4_0_c',\n",
    "        # 'FAcrn_hs_4_2E_c',\n",
    "        # 'FAcrn_hs_4_3OH__R_c',\n",
    "        # 'FAcrn_hs_5_0_c',\n",
    "        # 'FAcrn_hs_6_0_c',\n",
    "        # 'FAcrn_hs_6_2E_c',\n",
    "        # 'FAcrn_hs_6_3OH__S_c',\n",
    "        # 'FAcrn_hs_7_0_c',\n",
    "        # 'FAcrn_hs_7_3OH__S_c',\n",
    "        # 'FAcrn_hs_8_0_c',\n",
    "        # 'FAcrn_hs_8_2E_c',\n",
    "        # 'FAcrn_hs_8_3OH__S_c',\n",
    "        # 'FAcrn_hs_9_0_c',\n",
    "        # 'FAcrn_hs_9_3OH__S_c',\n",
    "        # 'acrn_c',\n",
    "        # 'dmhptcrn_c',\n",
    "        # 'dmnoncrn_c',\n",
    "        # 'malcrn_c',\n",
    "        # tRNA\n",
    "        # 'trnaala_c',\n",
    "        # 'trnaarg_c',\n",
    "        # 'trnaasn_c',\n",
    "        # 'trnaasp_c',\n",
    "        # 'trnacys_c',\n",
    "        # 'trnagln_c',\n",
    "        # 'trnaglu_c',\n",
    "        # 'trnagly_c',\n",
    "        # 'trnahis_c',\n",
    "        # 'trnaile_c',\n",
    "        # 'trnaleu_c',\n",
    "        # 'trnalys_c',\n",
    "        # 'trnamet_c',\n",
    "        # 'trnaphe_c',\n",
    "        # 'trnapro_c',\n",
    "        # 'trnaser_c',\n",
    "        # 'trnathr_c',\n",
    "        # 'trnatrp_c',\n",
    "        # 'trnatyr_c',\n",
    "        # 'trnaval_c',\n",
    "        # 'alatrna_c',\n",
    "        # 'argtrna_c',\n",
    "        # 'asntrna_c',\n",
    "        # 'asptrna_c',\n",
    "        # 'cystrna_c',\n",
    "        # 'glntrna_c',\n",
    "        # 'glutrna_c',\n",
    "        # 'glytrna_c',\n",
    "        # 'histrna_c',\n",
    "        # 'iletrna_c',\n",
    "        # 'leutrna_c',\n",
    "        # 'lystrna_c',\n",
    "        # 'mettrna_c',\n",
    "        # 'phetrna_c',\n",
    "        # 'protrna_c',\n",
    "        # 'sertrna_c',\n",
    "        # 'thrtrna_c',\n",
    "        # 'trptrna_c',\n",
    "        # 'tyrtrna_c',\n",
    "        # 'valtrna_c',\n",
    "        # Sugar\n",
    "        # '2ddglcn_c',\n",
    "        # '3dfru_c',\n",
    "        # Nucleotides\n",
    "        # \"23camp_c\",\n",
    "        # \"23ccmp_c\",\n",
    "        # \"23cgmp_c\",\n",
    "        # \"23cump_c\",\n",
    "        # \"dctp_c\",\n",
    "        # \"dgtp_c\",\n",
    "        # \"datp_c\",\n",
    "        # \"dttp_c\",\n",
    "        # 'dutp_c',\n",
    "        # \"psump_c\",\n",
    "        # \"psi_c\",\n",
    "        # Sourced from somewhere/\n",
    "        # Drains to somewhere/accumulates\n",
    "        # 'so2gth_c',\n",
    "        # \"dh15kprostge1_c\",\n",
    "        # \"dh15kprostge2_c\",\n",
    "        # \"dh15kprostge3_c\",\n",
    "        # \"dh15kprostgf1_c\",\n",
    "        # \"dh15kprostgf2_c\",\n",
    "        # \"dh15kprostgf3_c\",\n",
    "        # 'polyadprib2_c',\n",
    "        # 'polyadprib1_c',\n",
    "    ],\n",
    "}\n",
    "default_closed = []\n",
    "for btype, met_list in boundaries.items():\n",
    "    for met in met_list:\n",
    "        met = model.metabolites.get_by_id(met)\n",
    "        try:\n",
    "            reaction = model.add_boundary(met, type=btype)\n",
    "        except ValueError:\n",
    "            rid = {\n",
    "                \"exchange\": f\"EX_{met}\",\n",
    "                \"demand\": f\"DM_{met}\",\n",
    "                \"sink\": f\"SK_{met}\",\n",
    "            }[btype]\n",
    "            reaction = model.reactions.get_by_id(rid)\n",
    "            reaction.name = f\"{met.name} {btype}\"\n",
    "\n",
    "        if met in default_closed:\n",
    "            reaction.lower_bound = 0\n",
    "\n",
    "for reaction in model.boundary:\n",
    "    reaction.subsystem = \"Pseudoreactions\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff186a49-66a1-444c-92dc-0f1eca0464d9",
   "metadata": {},
   "source": [
    "#### Reset subsystem groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5037b248-ee81-442e-aa81-dd8acd25380f",
   "metadata": {},
   "outputs": [],
   "source": [
    "reset_subsystem_groups(model)\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e2ba5b6-68bb-4160-bcfa-58fe424d8025",
   "metadata": {},
   "source": [
    "### Check mass balancing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d47506ac-f31a-4730-93dc-95fef5cf1986",
   "metadata": {},
   "outputs": [],
   "source": [
    "for reaction in model.reactions:\n",
    "    if reaction.boundary:\n",
    "        continue\n",
    "    try:\n",
    "        if reaction.check_mass_balance():\n",
    "            print(reaction)\n",
    "            print(reaction.check_mass_balance())\n",
    "            print()\n",
    "    except:\n",
    "        print({m.id: m.charge for m in reaction.metabolites})\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18d61d2b-a4a0-4ac8-84c6-68bd7b421790",
   "metadata": {},
   "source": [
    "### Set bounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc710bb4-325e-4db2-bd30-c3595b60052e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reset_reaction_bounds(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa717ddf-cb70-4e15-839d-80e245c113dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.flux_analysis.variability import find_blocked_reactions\n",
    "\n",
    "blocked_reactions = find_blocked_reactions(model, open_exchanges=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ef012b-a0b7-4d18-b054-f262db4b7b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Number of blocked reactions: {len(blocked_reactions)}\")\n",
    "mets_one_reaction = model.metabolites.query(\n",
    "    lambda x: len(x.reactions) == 1 and not any([r for r in x.reactions if r.boundary])\n",
    ")\n",
    "mets_one_reaction = sorted(mets_one_reaction.list_attr(\"id\"))\n",
    "print(f\"Number of metabolites with one associated reaction {len(mets_one_reaction)}\")\n",
    "print()\n",
    "for reaction in model.reactions.get_by_any(sorted(blocked_reactions)):\n",
    "    # Expected blocked\n",
    "    # if any([\n",
    "    #     x in reaction.id\n",
    "    #     for x in [\"3OHAA\", \"BKAACP\", 'EACPR1', 'CRNAT', 'FACOA3OHD', '3OXAACPR1', \"ACPH\", 'FACOA2ERy', '3OXAACPR', 'FACOA3ORy']\n",
    "    # ]):\n",
    "    #     continue\n",
    "    print(reaction)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cad316c-ec10-461b-9c38-4e1bdeed2db5",
   "metadata": {},
   "source": [
    "### Ensure correct types before export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea115b87-ffa6-4773-8207-08701a8d03b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SBML will not export charges correctly if they are float\n",
    "for metabolite in model.metabolites:\n",
    "    metabolite.charge = int(metabolite.charge)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77e6dc2-ed20-455b-aadd-8a03db5666f1",
   "metadata": {},
   "source": [
    "### Export model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742aeb63-f41b-4280-8fc0-9a79e9410e3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if overwrite:\n",
    "    write_rbc_model(model, filetype={\"xml\", \"json\"})\n",
    "else:\n",
    "    write_rbc_model(\n",
    "        model, filetype={\"xml\", \"json\"}, directory=f\"{ROOT_PATH}{INTERIM_PATH}\"\n",
    "    )\n",
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "978d93a7-23dc-47ee-b9b9-f330fdfa8f05",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Genes: {len(set([x.id for x in model.genes]))}\")\n",
    "print(f\"Metabolites (all): {len(set([x.id for x in model.metabolites]))}\")\n",
    "nmets_unique = len({x.id.replace(f\"_{x.compartment}\", \"\") for x in model.metabolites})\n",
    "print(f\"Metabolites (unique): {nmets_unique}\")\n",
    "print(\n",
    "    f\"Reactions: {len(set([x.id for x in model.reactions.query(lambda x: not x.boundary)]))}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "607db495-a807-44fa-b747-447e8218180c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
